<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html>
<head>
	<meta http-equiv="content-type" content="text/html; charset=iso-8859-1"/>
	<title>The Biostar Handbook: 2nd Edition</title>
	<meta name="generator" content="LibreOffice 6.0.7.3 (Linux)"/>
	<meta name="created" content="2024-04-11T11:37:42"/>
	<meta name="changed" content="2024-04-11T11:37:42"/>
	<meta name="AppVersion" content="12.0000"/>
	<meta name="DocSecurity" content="0"/>
	<meta name="HyperlinksChanged" content="false"/>
	<meta name="LinksUpToDate" content="false"/>
	<meta name="ScaleCrop" content="false"/>
	<meta name="ShareDoc" content="false"/>
	<style type="text/css">
		@page { size: 8.5in 11in; margin-left: 1.25in; margin-right: 1.25in; margin-top: 1in; margin-bottom: 1in }
		p { margin-top: 0.13in; margin-bottom: 0.13in; direction: ltr; text-align: left; orphans: 2; widows: 2 }
		p.compact { margin-top: 0.03in; margin-bottom: 0.03in }
		h1 { margin-top: 0.33in; margin-bottom: 0in; direction: ltr; color: #345a8a; text-align: left; page-break-inside: avoid; orphans: 2; widows: 2 }
		h1.western { font-family: "Calibri", serif; font-size: 16pt }
		h1.cjk { font-family: ; font-size: 16pt }
		h1.ctl { font-family: ; font-size: 16pt }
		h2 { margin-bottom: 0in; direction: ltr; color: #4f81bd; text-align: left; page-break-inside: avoid; orphans: 2; widows: 2 }
		h2.western { font-family: "Calibri", serif; font-size: 16pt }
		h2.cjk { font-family: ; font-size: 16pt }
		h2.ctl { font-family: ; font-size: 16pt }
		h3 { margin-top: 0.14in; margin-bottom: 0in; direction: ltr; color: #4f81bd; text-align: left; page-break-inside: avoid; orphans: 2; widows: 2 }
		h3.western { font-family: "Calibri", serif }
		h3.cjk { font-family:  }
		h3.ctl { font-family:  }
		h4 { margin-top: 0.14in; margin-bottom: 0in; direction: ltr; color: #4f81bd; text-align: left; page-break-inside: avoid; orphans: 2; widows: 2 }
		h4.western { font-family: "Calibri", serif }
		h4.cjk { font-family:  }
		h4.ctl { font-family:  }
		p.block-text-western { margin-top: 0.07in; margin-bottom: 0.07in; font-family: "Calibri", serif; font-size: 10pt }
		p.block-text-cjk { margin-top: 0.07in; margin-bottom: 0.07in; font-family: ; font-size: 10pt }
		p.block-text-ctl { margin-top: 0.07in; margin-bottom: 0.07in; font-family: ; font-size: 10pt; font-weight: bold }
		a:link { color: #4f81bd; text-decoration: none }
	</style>
</head>
<body lang="en-US" link="#4f81bd" dir="ltr">
<p align="center" style="margin-top: 0.33in; margin-bottom: 0.17in; page-break-inside: avoid; page-break-after: avoid">
<font color="#345a8a"><font face="Calibri, serif"><font size="5" style="font-size: 18pt"><b>The
Biostar Handbook: 2nd Edition</b></font></font></font></p>
<p align="center" style="margin-top: 0in; margin-bottom: 0in; page-break-inside: avoid; page-break-after: avoid">
2021-12-13T02:54:53Z</p>
<p class="first-paragraph"><img src="0000_html_e565f55b1babbdb8.png" name="Picture" align="bottom" hspace="12" width="560" height="734" border="0"/>
</p>
<p><br/>
<br/>

</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="welcome-to-the-biostar-handbook"></a>1
Welcome to the Biostar Handbook</h1>
<p class="first-paragraph">Last updated on <b>December 12, 2021</b>,
ISBN: <b>978-0-578-80435-4</b></p>
<p><img src="0000_html_40b61c0b7f7cb3d1.png" name="Image1" align="bottom" hspace="12" width="180" height="236" border="0"/>
</p>
<p>The Biostar Handbook introduces readers to bioinformatics, the
scientific discipline at the intersection of biology, computer
science, and statistical data analytics dedicated to the digital
processing of genomic information.</p>
<p>The Biostar Handbook has been developed, improved and refined over
more than a half decade in a research university setting while used
in an accredited Ph.D.&nbsp;level training program. The contents of
this book have provided the analytical foundation to thousands of
students, many of whom have become full-time bioinformaticians and
work at the most innovative companies in the world.</p>
<h2 class="western"><a name="the-biostar-handbook-collection"></a>1.1
The Biostar Handbook Collection</h2>
<p class="first-paragraph">The Biostar Handbook is being reworked
into separate, more manageable volumes of study.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/">The
	Biostar Handbook</a></font> - An introduction to Bioinformatics as a
	scientific field.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/scripting/index.html">The
	Art of Bioinformatics Scripting</a></font> - Learn advanced Unix and
	Bash scripting skills.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html">RNA-Seq
	by Example</a></font> - Master RNA-Seq data analysis.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/corona/index.html">Corona
	Virus Genome Analysis</a></font> - Advanced topics devoted to the
	study of the Corona Virus.</p>
</ul>
<p class="first-paragraph">Access to new books and new content is
included with your subscription!</p>
<p class="compact"><a href="https://www.biostarhandbook.com/"></a><img src="0000_html_c7cfc27144d70375.png" name="Image2" align="bottom" hspace="12" width="114" height="150" border="0"/>
<a href="https://www.biostarhandbook.com/"></a></p>
<p><font color="#4f81bd"><a href="https://www.biostarhandbook.com/">The
Biostar Handbook</a></font></p>
<p class="compact"><a href="https://www.biostarhandbook.com/books/scripting/index.html"></a><img src="0000_html_1755df69f71fec9f.png" name="Image3" align="bottom" hspace="12" width="116" height="150" border="0"/>
<a href="https://www.biostarhandbook.com/books/scripting/index.html"></a></p>
<p><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/scripting/index.html">The
Art of Bioinformatics Scripting</a></font></p>
<p class="compact"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html"></a><img src="0000_html_e1642cf5ddff8fc3.png" name="Image4" align="bottom" hspace="12" width="116" height="150" border="0"/>
<a href="https://www.biostarhandbook.com/books/rnaseq/index.html"></a></p>
<p><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html">RNA-Seq
by Example</a></font></p>
<p class="compact"><a href="https://www.biostarhandbook.com/books/corona/index.html"></a><img src="0000_html_a8c36a3c62dae3e0.png" name="Image5" align="bottom" hspace="12" width="116" height="150" border="0"/>
<a href="https://www.biostarhandbook.com/books/corona/index.html"></a></p>
<p><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/corona/index.html">Corona
Virus Genome Analysis</a></font></p>
<h2 class="western"><a name="get-notified"></a>Get notified</h2>
<p class="first-paragraph">A few times a year we send out emails
summarizing the changes made to the book</p>
<p class="compact">Email    
</p>
<p>Enter your email and press subscribe above to receive these
notifications!</p>
<h2 class="western"><a name="online-course"></a>1.2 Online course</h2>
<p class="first-paragraph">Access to the online courses is included
with the book, for the duration of your subscription.</p>
<p><a href="https://www.biostarhandbook.com/edu/course/6/"></a><img src="0000_html_1379cae292489f57.png" name="Image6" align="bottom" hspace="12" width="560" height="69" border="0"/>
<a href="https://www.biostarhandbook.com/edu/course/6/"></a></p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/6/">Course:
	Applied Bioinformatics</a></font></p>
</ul>
<p class="first-paragraph">The course, reworked for the Fall of 2021,
tracks the <b>BMMB 852: Applied Bioinformatics</b> graduate-level
course at Penn State.</p>
<h2 class="western"><a name="previous-courses"></a>1.3 Previous
courses</h2>
<p class="first-paragraph">Please note that the following courses are
to be considered <i>outdated</i> and are listed here only for
reference:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/1/">Learn
	Bioinformatics in 100 hours (2017)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/2/">Python
	Programming (2018)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/3/">Bioinformatics
	Cookbook (2018)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/5/">Learn
	Bioinformatics the Right Way (2019)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/edu/course/4/">Bioinformatics
	data analysis (2019)</a></font></p>
</ul>
<p class="first-paragraph">Though valuable and contain useful
content, the links within our older courses are not kept up to date
with the changes in the book.</p>
<h2 class="western"><a name="how-to-download-the-book"></a>1.4 How to
download the book?</h2>
<p class="first-paragraph">The book is available to registered users.
The latest versions can be downloaded from:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="">Biostar
	Handbook, 2nd Edition, PDF</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="">Biostar
	Handbook, 2nd Edition, eBook</a></font></p>
</ul>
<p class="first-paragraph">We recommend reading and accessing the
book via the website! The web version will always contain the most
recent additions and most up-to-date content. A few times a year we
send out emails that describe the new additions.</p>
<p>In addition, in the web version of the book all images have larger
resolution than that shown on the page itself. To see an image in
full resolution right-click then select &ldquo;Open image in new
tab&rdquo;.</p>
<h2 class="western"><a name="new-feature-shell-casting-january-2020"></a>
1.5 New feature: Shell casting (January 2020)</h2>
<p class="first-paragraph">The book now includes <b>shell casts</b>
to demonstrate command line usage. The <b>shell casts</b> are not
videos! These animations are textual, captured from a terminal
exactly as typed, with sharp visibility, will zoom with your screen
font sizes, may be paused at any time, and may be selected for
copy-paste from the screen! Here is an example:</p>
<p><br/>
<br/>

</p>
<p>Shell casts are only visible via the web. PDF and ebooks do not
have these animations embedded in them. Animations are a new feature
in 2020. As we re-work the book we will be adding <b>shell casts</b>
into each chapter.</p>
<h2 class="western"><a name="how-was-the-book-developed"></a>1.6 How
was the book developed?</h2>
<p class="first-paragraph">We have been teaching bioinformatics and
programming courses to life scientists for many years now. We are
also the developers and maintainers of <font color="#4f81bd"><a href="https://www.biostars.org/">Biostars:
Bioinformatics Question and Answer</a></font> website the leading
resource for helping bioinformatics scientists with their data
analysis questions.</p>
<p>We wrote this book based on these multi-year experiences in
training students and interacting with scientists that needed help to
complete their analyses. We are uniquely in tune with the challenges
and complexities of applying bioinformatics methods to real problems,
and we&rsquo;ve designed this book to help readers overcome these
challenges and go further than they have ever imagined.</p>
<h2 class="western"><a name="how-is-this-book-different"></a>1.7 How
is this book different?</h2>
<p class="first-paragraph">We use a Question/Answer based format to
explain concepts and demonstrate methods simply and practically. We
found this to be the most efficient training method for the topics
that we cover.</p>
<h2 class="western"><a name="who-is-a-biostar"></a>1.8 Who is a
Biostar?</h2>
<p class="first-paragraph">It is not a what; it is a who. And it
could be you. It is the person whose responsibilities go beyond clear
boundaries. This book is for them.</p>
<p><img src="0000_html_2c28dd2bf9a20668.png" name="Image7" align="bottom" hspace="12" width="560" height="259" border="0"/>
</p>
<p>Visualization inspired by a blog post from 2014 by Anthony Fejes:
<font color="#4f81bd">Who is a bioinformatician?</font></p>
<h2 class="western"><a name="manage-your-account"></a>Manage your
account</h2>
<p class="first-paragraph">If you are logged in you may manage your
account by clicking the link button below.</p>
<p><font color="#4f81bd"><a href="">Access Your Account</a></font></p>
<p>You may also change your email or log out via this page.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="about-the-author"></a>2 About the author</h1>
<p class="first-paragraph"><img src="0000_html_a989ae769d6d3e1c.jpg" name="Image8" align="bottom" hspace="12" width="96" height="96" border="0"/>
</p>
<p>Dr.&nbsp;Istv&aacute;n Albert is renowned for being at the
forefront of the Bioinformatics frontier and pushing the field
forward in new directions.</p>
<p>He helped pioneer the <font color="#4f81bd"><a href="https://usegalaxy.org/">Galaxy
Bioinformatics Platform</a></font>, an open source web-based tool
that allows users to perform and share data-intensive biomedical
research. He also developed <font color="#4f81bd"><a href="https://scfbm.biomedcentral.com/articles/10.1186/1751-0473-3-16">BooleanNet</a></font>,
a biological system simulation software and the <font color="#4f81bd"><a href="http://bioinformatics.oxfordjournals.org/content/24/10/1305.long">GeneTrack</a></font>
software platform that automates and processes large sets of
biological data.</p>
<p>Currently, Dr.&nbsp;Albert is the lead software developer,
maintainer and administrator of the website, <font color="#4f81bd"><a href="https://www.biostars.org/">Biostars:
Bioinformatics Questions and Answers</a></font>, the most
comprehensive information source in Bioinformatics that is visited by
millions of people every year.</p>
<p>Besides being well-known for his <font color="#4f81bd"><a href="https://scholar.google.com/citations?hl=en&amp;user=hordfUUAAAAJ&amp;view_op=list_works&amp;sortby=pubdate">scientific
innovations</a></font> in three different scientific fields: Physics,
Computer Science and Biology with publications that gathered over 10
thousand citations, Dr.&nbsp;Albert is a celebrated educator. The
courses he develops are consistently highly ranked and revered by
students.</p>
<p>In 2014, Istv&aacute;n was awarded the <font color="#4f81bd"><a href="http://bmb.psu.edu/about/news-articles/istvan-albert-recognized-for-outstanding-teaching">Paul
M. Althouse Outstanding Teaching Award</a></font> at Penn State for
the superb quality of his lectures. The lecture information that has
garnered him such high praise is the very same material that you will
find in the <font color="#4f81bd"><a href="https://www.biostarhandbook.com/">Biostar
Handbook</a></font>.</p>
<p>Dr.&nbsp;Albert is currently a <font color="#4f81bd"><a href="https://www.ialbert.me/">Research
Professor of Bioinformatics</a></font> at Pennsylvania State
University and the Director of the Bioinformatics Consulting Center
at this institute. Istv&aacute;n established this cutting-edge
research facility where he has advised and assisted dozens of
brilliant young scientists in their ambitious endeavors. He also
leads the <font color="#4f81bd"><a href="http://www.worldcampus.psu.edu/degrees-and-certificates/applied-bioinformatics-certificate/overview">Graduate
Certificate Program in Applied Bioinformatics</a></font>, a
comprehensive Bioinformatics training course offered through the <font color="#4f81bd"><a href="http://www.worldcampus.psu.edu/">Penn
State World Campus</a></font> that allows anyone, anywhere, to attain
in-depth knowledge of this exciting new field.</p>
<p>The Handbook&rsquo;s main author and editor is <font color="#4f81bd"><a href="https://www.ialbert.me/">Dr.&nbsp;Istvan
Albert</a></font>.</p>
<ul>
	<li/>
<p class="compact">Personal website: <font color="#4f81bd"><a href="https://www.ialbert.me/">https:/www.ialbert.me</a></font></p>
	<li/>
<p class="compact">Email: <font color="#4f81bd"><a href="mailto:%20istvan.albert@gmail.com">istvan.albert@gmail.com</a></font></p>
</ul>
<h3 class="western"><a name="contributors"></a>2.0.1 Contributors</h3>
<ul>
	<li/>
<p class="compact">Aswathy Sebastian, MS, Bioinformatics
	Analyst, Bioinformatics Consulting Center, Pennsylvania State
	University, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ralbert.me/">Reka
	Albert</a></font>, Ph.D., Professor of Physics and Biology,
	Department of Physics, Pennsylvania State University, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://dbhi.chop.edu/index.php/people/40-people/90-jeremy-leipzig-ms.html">Jeremy
	Leipzig</a></font>, MS, Senior Data Integration Analyst, The
	Children&rsquo;s Hospital of Philadelphia, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://bioinformatics.unc.edu/personnel/">Hemant
	Kelkar</a></font>, Ph.D., Research Associate Professor, Department
	of Genetics and Center for Bioinformatics, University of North
	Carolina, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.linkedin.com/in/ming-tang-40650014">Ming
	Tang</a></font>, Ph.D., Computational biologist in Genomic Medicine
	Department, MD Anderson Cancer Center, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.linkedin.com/in/ramakrishnanrs">Ram
	Srinivasan</a></font>, MS, Bioinformatician, Icahn School of
	Medicine and Mount Sinai, New York, NY, USA</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://shenwei.me/">Wei
	Shen</a></font>, Ph.D.&nbsp;student, Third Military Medical
	University, China.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.uantwerpen.be/en/staff/wouter-decoster/">Wouter
	De Coster</a></font>, Ph.D.&nbsp;Student, University of Antwerp,
	Belgium</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.linkedin.com/in/madelainegogol">Madelaine
	Gogol</a></font>, BS, Programmer Analyst, Stowers Institute, Kansas
	City, MO, USA</p>
	<li/>
<p class="compact">John Dale, retired physician, biohacker,
	Ham Radio enthusiast(VE7QJD), commercial pilot, Nelson, BC, Canada</p>
</ul>
<h3 class="western"><a name="can-i-contribute-to-the-book"></a>2.0.2
Can I contribute to the book?</h3>
<p class="first-paragraph">Join our effort to build a comprehensive
and up to date compendium for genomic data analysis.</p>
<p>It is a simple process that works through GitHub via simple,
text-based, Markdown files. The only permission we ask from authors
are the rights to publish their content under our own terms (see
below).</p>
<p>Authors retain re-publishing rights for the material that they are
the principal author of and may re-distribute the content that they
have created for this book under any other terms of their choosing.</p>
<h3 class="western"><a name="what-are-the-licensing-terms"></a>2.0.3
What are the licensing terms?</h3>
<p class="first-paragraph">The Handbook&rsquo;s content is
copyrighted material owned by <font color="#4f81bd"><a href="https://www.biostargenomics.com/">Biostar
Genomics LLC</a></font>. Republishing any part of the Handbook is
permitted only on a limited basis and must follow the terms of the
<font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Fair_use">Fair
Use</a></font> policy unless other, prior agreements have been made.</p>
<h3 class="western"><a name="what-can-i-reuse-from-the-book"></a>2.0.4
What can I reuse from the book?</h3>
<p class="first-paragraph">If you have purchased the book or have had
at any time an account on this site that means that you have license
to apply, modify and reuse information from the book as it pertains
to your own work and research.</p>
<h3 class="western"><a name="terms-of-use"></a>2.0.5 Terms of use</h3>
<p class="first-paragraph">You may not share your login information
with others and you may not re-distribute the book, or any part of
the book to others.</p>
<h3 class="western"><a name="citations"></a>2.0.6 Citations</h3>
<p class="first-paragraph">In general, when it comes to citations we
recommend that you cite the authors of the tools and techniques that
are using. We also welcome acknowledgements of the book in your work.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="why-bioinformatics"></a>3 Why
bioinformatics?</h1>
<p class="first-paragraph">In the modern world, it often seems that
the age of exploration is over. We have been on the Moon, there are
research bases on Antarctica, we can look at atoms. You may ask
yourself are there wonders waiting to be discovered? After all, it
may seem that all we are left with is to refine and improve on prior
observations.</p>
<p>I have good news for you all. The information age has opened up an
entirely new domain of science, that of understanding what life
itself consists of. We are all beginners in this new science, we are
all starting from the same knowledge and principles. Everyone has the
chance to make unique and fascinating discoveries using little more
than their own computer.</p>
<h2 class="western"><a name="what-is-this-book-about"></a>3.1 What is
this book about?</h2>
<p class="first-paragraph"><img src="0000_html_b55e7d94cba14b13.png" name="Image9" align="bottom" hspace="12" width="180" height="236" border="0"/>
</p>
<p>The answers to some of the greatest questions of life lie within
ourselves. Bioinformatics is a new science created from the marriage
of Data Science and Biology. Through this emerging field of study,
scientists are able to find and decode hidden information in our very
own genes, allowing us to understand what none before us have known.</p>
<p>This book teaches you practical skills that will allow you to
enter this fast expanding industry. Beginning with fundamental
concepts such as understanding data formats and how analysis is done
in general and what conclusions can be drawn from data, the Handbook
eases you into the limitless world of possibilities.</p>
<p>With the help of the book, you will be reproducing results from
realistic data analysis scenarios such as genome assembly or gene
expression analysis. The methods and tools you will find in the
Handbook were refined in world-class research facilities and will
allow you to break into this new field and tackle some of the most
significant challenges we are facing in the scientific frontier of
the 21st century.</p>
<h2 class="western"><a name="what-is-covered-in-the-book"></a>3.2
What is covered in the book?</h2>
<p class="first-paragraph">The Handbook is divided into several
sections. We cover both the foundations and their applications to
realistic data analysis scenarios.</p>
<p>Bioinformatics analysis concepts:</p>
<ul>
	<li/>
<p class="compact">Data formats and repositories.</p>
	<li/>
<p class="compact">Sequence alignments.</p>
	<li/>
<p class="compact">Data visualization.</p>
	<li/>
<p class="compact">Unix command line usage.</p>
</ul>
<p class="first-paragraph">Bioinformatics protocols:</p>
<ul>
	<li/>
<p class="compact">Genome variation and SNP calling.</p>
	<li/>
<p class="compact">RNA-seq and gene expression analysis</p>
	<li/>
<p class="compact">Genome assembly</p>
	<li/>
<p class="compact">Metagenomics classification</p>
	<li/>
<p class="compact">ChIP-Seq analysis</p>
</ul>
<p class="first-paragraph">Software tool usage:</p>
<ul>
	<li/>
<p class="compact">Using short read aligners</p>
	<li/>
<p class="compact">Using quality control tools</p>
	<li/>
<p class="compact">Manipulating sequence data</p>
</ul>
<p class="first-paragraph">The table of contents on the left allows
you to jump to the corresponding sections.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="what-is-bioinformatics"></a>4 What is
bioinformatics?</h1>
<p class="first-paragraph">Bioinformatics is a new,
computationally-oriented Life Science domain. Its primary goal is to
make sense of the information stored within living organisms.
Bioinformatics relies on and combines concepts and approaches from
biology, computer science, and data analysis. Bioinformaticians
evaluate and define their success primarily in terms of the new
insights they produce about biological processes through digitally
parsing <b>genomic</b> information.</p>
<p class="block-text-western">Bioinformatics is a data science that
investigates how information is <i>stored within</i> and <i>processed
by</i> living organisms.</p>
<h2 class="western"><a name="how-has-bioinformatics-changed"></a>4.1
How has bioinformatics changed?</h2>
<p class="first-paragraph">In its early days&ndash;&ndash;perhaps
until the beginning of the 2000s&ndash;&ndash;bioinformatics was
synonymous with <b>sequence analysis</b>. Scientists typically
obtained just a few DNA sequences, then analyzed them for various
properties. Today, sequence analysis is still central to the work of
bioinformaticians, but it has also grown well beyond it.</p>
<p>In the mid-2000s, the so-called <i>next-generation</i>,
<i>high-throughput</i> sequencing instruments (such as the Illumina
HiSeq) made it possible to measure the full genomic content of a cell
in a single experimental run. With that, the quantity of data shot up
immensely as scientists were able to capture a snapshot of <i>everything</i>
that is DNA-related.</p>
<p>These new technologies have transformed bioinformatics into an
entirely new field of <i>data science</i> that builds on the
&ldquo;classical bioinformatics&rdquo; to process, investigate, and
summarize massive data sets of extraordinary complexity.</p>
<h2 class="western"><a name="what-subfields-of-bioinformatics-exist"></a>
4.2 What subfields of bioinformatics exist?</h2>
<p class="first-paragraph">DNA sequencing as initially valued for
revealing the DNA content of a cell. It may come as a surprise to
many, however, that the most significant promise for the future of
bioinformatics might lie in other applications. In general, most
bioinformatics problems fall under one of four categories:</p>
<ol>
	<li/>
<p class="compact"><b>Classification</b>: determining the
	species composition of a population of organisms</p>
	<li/>
<p class="compact"><b>Assembly</b>: establishing the
	nucleotide composition of genomes</p>
	<li/>
<p class="compact"><b>Resequencing</b>: identifying mutations
	and variations in genomes</p>
	<li/>
<p class="compact"><b>Quantification</b>: using DNA sequencing
	to measure the functional characteristics of a cell</p>
</ol>
<p class="first-paragraph">The <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Human_Genome_Project">Human
Genome Project</a></font> fell squarely in the <b>assembly</b>
category. Since its completion, scientists have assembled the genomes
of thousands of others species. The genomes of many millions of
species, however, remain entirely unknown.</p>
<p>Studies that attempt to identify changes relative to known genomes
fall into the <b>resequencing</b> field of study. DNA mutations and
variants may cause phenotypic changes like emerging diseases,
changing fitness, different survival rates, and many others. For
example, there are several ongoing efforts to compile all variants
present in the human genome&ndash;&ndash;these efforts would fall
into the resequencing category. Thanks to the work of
bioinformaticians, massive computing efforts are underway to produce
clinically valuable information from the knowledge gained through
resequencing.</p>
<p>Living micro-organisms surround us, and we coexist with them in
complex collectives that can only survive by maintaining
interdependent harmony. <b>Classifying</b> these mostly-unknown
species of micro-organisms by their genetic material is a
fast-growing subfield of bioinformatics.</p>
<p>Finally, and perhaps most unexpectedly, bioinformatics methods can
help us better understand biological processes, like gene
expressions, through <b>quantification</b>. In these protocols, the
sequencing procedures are used to determine the relative abundances
of various DNA fragments that were made to correlate with other
biological processes.</p>
<p>Over the decades biologists have become experts at manipulating
DNA and are now able to co-opt the many naturally-occurring molecular
processes to copy, translate, and reproduce DNA molecules and connect
these actions to biological processes. Sequencing has opened a new
window into this world, new methods and sequence manipulations are
continuously discovered. The various methods are typically named as
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Something-Seq</font></font>
for example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNA-Seq</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Chip-Seq</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">RAD-Seq</font></font>
to reflect what mechanism was captured/connected to sequencing. For
example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNA-Seq</font></font>
reveals the abundance of RNA by turning it into DNA via reverse
transcription. Sequencing this construct allows for simultaneously
measuring the expression levels of all genes of a cell. The for
example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">RAD-Seq</font></font>
uses restriction enzymes to cleave DNA at specific locations and only
the fragments around these locations are then sequenced. This method
produces very high coverage around these sites, thus is suited for
population genetics studies.</p>
<h2 class="western"><a name="is-there-a-list-of-functional-assays-used-in-bioinformatics"></a>
4.3 Is there a list of functional assays used in bioinformatics?</h2>
<p class="first-paragraph">In the Life Sciences, an <b>assay</b> is
an investigative procedure used to <i>assess</i> or measure the
presence, amount, or function of some target (like a DNA fragment).
<font color="#4f81bd"><a href="https://liorpachter.wordpress.com/">Dr.&nbsp;Lior
Pachter</a></font>, professor of Mathematics at Caltech, maintains a
list of &ldquo;functional genomics&rdquo; assay technologies on the
page called <font color="#4f81bd"><a href="https://liorpachter.wordpress.com/seq/">Star-Seq</a></font>.</p>
<p>All of these techniques fall into the <b>quantification</b>
category. Each assay uses DNA sequencing to quantify another measure,
and many are examples of connecting DNA abundances to various
biological processes.</p>
<p>Notably, the list now contains nearly 100 technologies. Many
people, us included, believe that these applications of sequencing
are of greater importance and impact than identifying the base
composition of genomes.</p>
<p>Below are some examples of the assay technologies on Dr.&nbsp;Pachter&rsquo;s
list:</p>
<p><img src="0000_html_9e9403a20d7d58cd.png" name="Image10" align="bottom" hspace="12" width="560" height="585" border="0"/>
</p>
<h2 class="western"><a name="but-what-is-bioinformatics-really"></a>4.4
But what is bioinformatics, <i>really</i>?</h2>
<p class="first-paragraph">So now that you know what bioinformatics
is all about, you&rsquo;re probably wondering what it&rsquo;s like to
practice it day-in-day-out as a bioinformatician. The truth is, it&rsquo;s
not easy. Just take a look at this &ldquo;Biostar Quote of the Day&rdquo;
from Brent Pedersen in <font color="#4f81bd"><a href="https://www.biostars.org/p/63336/">Very
Bad Things</a></font>:</p>
<p class="block-text-western">I&rsquo;ve been doing bioinformatics
for about 10 years now. I used to joke with a friend of mine that
most of our work was converting between file formats. We don&rsquo;t
joke about that anymore.</p>
<p class="first-paragraph">Jokes aside, modern bioinformatics relies
heavily on file and data processing. The data sets are large and
contain complex interconnected information. A bioinformatician&rsquo;s
job is to simplify massive datasets and search them for the
information that is relevant for the given study. Essentially,
bioinformatics is the art of finding the needle in the haystack.</p>
<h2 class="western"><a name="are-bioinformaticians-data-janitors"></a>
4.5 Are bioinformaticians data janitors?</h2>
<p class="first-paragraph">Oh, yes. But then, make no mistake, all
data scientists are. It is not a unique feature of this particular
field.</p>
<p>I used to get worked up about the problems with data, but then I
read the New York Times opinion piece: <font color="#4f81bd"><a href="http://www.nytimes.com/2014/08/18/technology/for-big-data-scientists-hurdle-to-insights-is-janitor-work.html?_r=0">For
Big-Data Scientists, &lsquo;Janitor Work&rsquo; Is Key Hurdle to
Insights</a></font>, where they state:</p>
<p><img src="0000_html_ec5858d64262e514.png" name="Image11" align="bottom" hspace="12" width="560" height="41" border="0"/>
</p>
<p class="block-text-western">[&hellip;] Yet far too much handcrafted
work &mdash; what data scientists call &ldquo;data wrangling,&rdquo;
&ldquo;data munging&rdquo; and &ldquo;data janitor work&rdquo; &mdash;
is still required. Data scientists, according to interviews and
expert estimates, spend from 50 percent to 80 percent of their time
mired in this more mundane labor of collecting and preparing unruly
digital data, before it can be explored for useful nuggets [&hellip;]</p>
<p class="first-paragraph">You will have problems like that all the
time. For example, you&rsquo;ll download a file from an official
database like NCBI. It turns to be a large file with 1.5 million
entries. But you won&rsquo;t be able to use it, programs that ought
to work fine with data of this size crash spectacularly. You&rsquo;ll
scratch you head, dive in, few hours later it turns out that one
single line, out of the 1.5 million, has the wrong number of fields -
one line has the wrong format! Now you&rsquo;ll need to find a way to
fix or get rid of that single line - then everything works again.</p>
<p>It helps to be prepared and know what to expect.</p>
<h2 class="western"><a name="is-creativity-required-to-succeed"></a>4.6
Is creativity required to succeed?</h2>
<p class="first-paragraph">Bioinformatics requires a dynamic,
creative approach. Protocols should be viewed as guidelines, not as
rules that guarantee success. Following protocols by the letter is
usually entirely counterproductive. At best, doing so leads to
sub-optimal outcomes; at worst, it can produce misinformation that
spells the end of a research project.</p>
<p>Living organisms operate in immense complexity. Bioinformaticians
need to recognize this complexity, respond dynamically to variations,
and understand when methods and protocols are not suited to a data
set. The myriad complexities and challenges of venturing at the
frontiers of scientific knowledge always require creativity,
sensitivity, and imagination. Bioinformatics is no exception.</p>
<p>Unfortunately, the misconception that bioinformatics is a
<i>procedural</i> skill that anyone can quickly add to their toolkit
rather than a scientific domain in its own right can lead some people
to underestimate the value of a bioinformatician&rsquo;s
contributions to the success of a project.</p>
<p>As observed in the Nature paper <font color="#4f81bd"><a href="http://www.nature.com/news/core-services-reward-bioinformaticians-1.17251">Core
services: Reward bioinformaticians</a></font></p>
<p class="block-text-western">Biological data will continue to pile
up unless those who analyze it are recognized as creative
collaborators in need of career paths.</p>
<p class="first-paragraph">Bioinformatics requires multiple skill
sets, extensive practice, and familiarity with multiple analytical
frameworks. Proper training, a solid foundation and an in-depth
understanding of concepts are required of anyone who wishes to
develop the particular creativity needed to succeed in this field.</p>
<p>This need for creativity and the necessity for a bioinformatician
to think &ldquo;outside the box&rdquo; is what this Handbook aims to
teach. We don&rsquo;t just want to list instructions: &ldquo;do this,
do that.&rdquo; We want to help you establish that robust and
reliable foundation that allows you to be creative when (not if) that
time comes.</p>
<h2 class="western"><a name="are-analyses-all-alike"></a>4.7 Are
analyses all alike?</h2>
<p class="first-paragraph">Most bioinformatics projects start with a
&ldquo;standardized&rdquo; plan, like the ones you&rsquo;ll find in
this Handbook. However, these plans are never set in stone. Depending
on the types and features of observations and results of analyses,
additional tasks inevitably deviate from the original plan to account
for variances observed in the data. Frequently, the studies need
substantive customization.</p>
<p>Again, as the authors of <font color="#4f81bd"><a href="http://www.nature.com/news/core-services-reward-bioinformaticians-1.17251">Core
services: Reward bioinformaticians</a></font> note the following:</p>
<p class="block-text-western">&ldquo;No project was identical, and we
were surprised at how common one-off requests were. There were a few
routine procedures that many people wanted, such as finding genes
expressed in a disease. But 79% of techniques applied to fewer than
20% of the projects. In other words, most researchers came to the
bioinformatics core seeking customized analysis, not a standardized
package.&rdquo;</p>
<p class="first-paragraph">In summary, almost no two analyses are
precisely the same. Also, it is quite common for projects to deviate
from the standardized workflow substantially.</p>
<h2 class="western"><a name="should-life-scientists-know-bioinformatics"></a>
4.8 Should life scientists know bioinformatics?</h2>
<p class="first-paragraph"><b>Yes!</b></p>
<p>The results of bioinformatic analyses are relevant for most areas
of study inside the life sciences. Even if a scientist isn&rsquo;t
performing the analysis themselves, they need to be familiar with how
bioinformatics operates so they can accurately interpret and
incorporate the findings of bioinformaticians into their work. All
scientists informing their research with bioinformatic insights
should understand how it works by studying its principles, methods,
and limitations&ndash;&ndash;the majority of which is available for
you in this Handbook.</p>
<p>We believe that this book is of great utility even for those that
don&rsquo;t plan to run the analysis themselves.</p>
<h2 class="western"><a name="what-type-of-computer-is-required"></a>4.9
What type of computer is required?</h2>
<p class="first-paragraph">All tools and methods presented in this
book have been tested and will run on all three major operating
systems: <b>MacOS</b>, <b>Linux</b> and <b>Windows 10</b>. See the
<font color="#4f81bd"><a href="#ch009.xhtml#computer-setup">Computer
Setup</a></font> page.</p>
<p><img src="0000_html_6a2e44e8c50d3db8.png" name="Image12" align="bottom" hspace="12" width="560" height="78" border="0"/>
</p>
<p>For best results, Windows 10 users will need to join the <font color="#4f81bd"><a href="https://insider.windows.com/">Windows
Insider</a></font> program (a free service offered by Microsoft) that
will allow them to install the newest release of &ldquo;Windows
Subsystem for Linux (WSL)&rdquo;</p>
<h2 class="western"><a name="is-there-data-with-the-book"></a>4.10 Is
there data with the book?</h2>
<p class="first-paragraph">Yes, we have a separate data site at
<font color="#4f81bd"><a href="http://data.biostarhandbook.com/">http://data.biostarhandbook.com</a></font>.
Various chapters will refer to content distributed from this site.</p>
<h2 class="western"><a name="who-is-the-book-for"></a>4.11 Who is the
book for?</h2>
<p class="first-paragraph">The Biostar Handbook provides training and
practical instructions for students and scientists interested in data
analysis methodologies of genome-related studies. Our goal is to
enable readers to perform analyses on data obtained from high
throughput DNA sequencing instruments.</p>
<p>All of the Handbook&rsquo;s content is designed to be simple,
brief, and geared towards practical application.</p>
<h2 class="western"><a name="is-bioinformatics-hard-to-learn"></a>4.12
Is bioinformatics hard to learn?</h2>
<p class="first-paragraph">Bioinformatics engages the distinct fields
of biology, computer science, and statistical data analysis.
Practitioners must navigate the various philosophies, terminologies,
and research priorities of these three domains of science while
keeping up with the ongoing advances of each.</p>
<p>Its position at the intersection of these fields might make
bioinformatics more challenging than other scientific subdisciplines,
but it also means that you&rsquo;re exploring the frontiers of
scientific knowledge, and few things are more rewarding than that!</p>
<h2 class="western"><a name="can-i-learn-bioinformatics-from-this-book"></a>
4.13 Can I learn bioinformatics from this book?</h2>
<p class="first-paragraph"><b>Yes you can!</b></p>
<p>The questions and answers in the Handbook have been carefully
selected to provide you with steady, progressive, accumulating levels
of knowledge. Think of each question/answer pair as a small,
well-defined unit of instruction that builds on the previous ones.</p>
<ul>
	<li/>
<p class="compact">Reading this book will teach you what
	bioinformatics is all about.</p>
	<li/>
<p class="compact">Running the code will demonstrate you the
	skills you need to perform the analyses.</p>
</ul>
<h2 class="western"><a name="how-long-will-it-take-me-to-learn-bioinformatics"></a>
4.14 How long will it take me to learn bioinformatics?</h2>
<p class="first-paragraph">About <b>100</b> hours.</p>
<p>Of course, a more accurate answer depends on your background
preparation, and each person is different. Prior training in at least
one of the three fields that bioinformatics builds upon (biology,
computer science, and data analysis) is recommended. The time
required to master all skills also depends on how you plan to use
them. Solving larger and more complex data problems will require more
advanced skills, which need more time to develop fully.</p>
<p>That being said, based on several years of evaluating trainees in
the field, we have come to believe that an active student would be
able to perform publication quality analyses after dedicating about
<b>100</b> hours of study.</p>
<p>This is what this book is really about &ndash; to help you put
those <b>100</b> hours to good use.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="biology-for-bioinformaticians"></a>5
Biology for bioinformaticians</h1>
<p class="first-paragraph">Biology is a domain of the Life Sciences,
which include, but are not limited to, organic chemistry, ecology,
botany, zoology, physiology, and so on. Biology seeks to understand
the structures, functions, origins, interactions, and taxonomies of
living organisms.</p>
<p>As a science, biology is still relatively immature. It is
eclectic, encompassing several scientific domains that are each
insufficiently understood and described.</p>
<p>Organisms are immensely complex and always changing. For
centuries, we haven&rsquo;t had precise enough tools to measure,
describe, or understand the full extent of their complexity. Digital
technologies are changing this. Bioinformatics is at the frontier of
these changes, and its potential contributions to Biology and the
Life Sciences more broadly are quite exciting. As bioinformatics and
other innovative methodologies advance, we expect Life Sciences to
mature and develop rich, accurate vocabularies and models to
understand and describe living organisms.</p>
<p>For these reasons, current concepts and definitions in the Life
Sciences are still just approximations. While they can be
sufficiently accurate in some contexts, they may be woefully
inadequate in others. In what almost sounds like a &ldquo;scientific
fight club&rdquo;, the &ldquo;rules&rdquo; of bioinformatics are as
follows:</p>
<ol>
	<li/>
<p class="compact">There are no &ldquo;universal&rdquo; rules.</p>
	<li/>
<p class="compact">Every seemingly fundamental paradigm has
	one or more exceptions.</p>
	<li/>
<p class="compact">The validity of a bioinformatics method
	depends on unknown characteristics of the data.</p>
	<li/>
<p class="compact">Biology is always more complicated than you
	think, <i>even after taking this rule into account</i>.</p>
</ol>
<p class="first-paragraph">Below, we have attempted to describe the
biological concepts that we deem important to understand the types of
information encoded in data. Each of our definitions is short and
terse, and we recommend additional self-guided study of each concept
of interest.</p>
<h2 class="western"><a name="what-is-dna"></a>5.1 What is DNA?</h2>
<p class="first-paragraph"><b>DNA</b> stands for Deoxyribo Nucleic
Acid.</p>
<p>It is a macromolecule (a molecule constructed of smaller
molecules) that carries the genetic instructions required for the
development, functioning and reproduction of all known living
organisms. In eukaryotic organisms (like animals, plants, and fungi),
DNA is present in the nucleus of each cell. In prokaryotic organisms
(single-celled organisms like bacteria and mitochondria), DNA is
present in the cell&rsquo;s cytoplasm.</p>
<h3 class="western"><a name="what-is-dna-composed-of"></a>5.1.1 What
is DNA composed of?</h3>
<p class="first-paragraph">DNA is made up of two strands of smaller
molecules coiled around each other in a double-helix structure. If
you uncoiled DNA, you could imagine it looking somewhat like a
ladder. Ladders have two important parts: the poles on either side
and the rungs that you climb. The &ldquo;poles&rdquo; of DNA are made
of alternating molecules of deoxyribose (sugar) and phosphate. While
they provide the structure, it&rsquo;s the &ldquo;rungs&rdquo; of the
DNA that are most important for bioinformatics.</p>
<p>To understand the &ldquo;rungs&rdquo; of the DNA, imagine a ladder
split in half down the middle so that each half is a pole with a
bunch of half-rungs sticking out. In DNA, these &ldquo;half-rungs&rdquo;
are a molecule called nucleotides. Genetic information is encoded
into DNA by the order, or sequence, in which these nucleotides occur.
The &ldquo;ladder&rdquo; of DNA is held together by the bonds between
each &ldquo;half-rung.&rdquo;</p>
<h3 class="western"><a name="what-are-nucleotides"></a>5.1.2 What are
nucleotides?</h3>
<p class="first-paragraph">Nucleotides are the building blocks of
nucleic acids (DNA and RNA&ndash;we&rsquo;ll get to that one later
on). In DNA, there are four types of nucleotide: Adenine, Cytosine,
Guanine, and Thymine. Because the order in which they occur encodes
the information biologists try to understand, we refer to them by
their first letter, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T,</font></font>
respectively.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
  Adenine</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
  Cytosine</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">G
  Guanine</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">T
  Thymine</font></font></p>
<p class="first-paragraph">Back to the ladder analogy. &ldquo;Nucleotide&rdquo;
is the <i>type</i> of molecule that makes up each half-rung of our
ladder. Because they function as the units that encode genetic
information, each letter is also called a base.</p>
<p>For an example of how we represent the sequence of DNA bases using
these letters, if we took the DNA of the <i>Bacillus anthracis</i>
bacteria that causes Anthrax disease, unfolded the double-helix into
a ladder, and then split the ladder in two, the top half (the forward
strand&ndash;we&rsquo;ll get to that in a bit) would be written like
this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATTTTTTCTTGTTTTTTATATCCACAAACTCTTTT</font></font></p>
<h3 class="western"><a name="what-are-base-pairs"></a>5.1.3 What are
base pairs?</h3>
<p class="first-paragraph">When you put both sides of the ladder back
together, each rung is made by the bond between two bases. When they
are bonded, we call them a base pair (or sometimes &ldquo;bp&rdquo;).</p>
<p>When it comes to base pairs, it&rsquo;s important to remember that
Adenine only ever bonds with Thymine, and Guanine with Cytosine. In
the ladder analogy, this means that each rung will be denoted as
&ldquo;A-T,&rdquo; &ldquo;T-A,&rdquo; &ldquo;G-C,&rdquo; or &ldquo;C-G.&rdquo;</p>
<p>Certain groups of nucleotides that share some common attribute may
be designated by so called <i>ambiguity</i> codes, for example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">W</font></font>
stands for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y
    Pyrimidine (C or T) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R
    Purine (A or G) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">W
    Weak (A or T)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S
    Strong (G or C) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">K
    Keto (T or G) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">M
    Amino (C or A)     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">D
    A, G, T (not C - remember as after C)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">V
    A, C, G (not T - remember as after T/U - We'll get to &quot;U&quot;
soon) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H
    A, C, T (not G - remember as after G)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">B
    C, G, T (not A - remember as after A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N
    Any base</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
    Gap     </font></font>
</p>
<h3 class="western"><a name="what-are-dna-strands"></a>5.1.4 What are
DNA strands?</h3>
<p class="first-paragraph">Remember when we split the DNA &ldquo;ladder&rdquo;
in half down the middle so that each side was a pole with half-rungs
sticking out? These two halves are called strands. To distinguish the
two strands, scientists label one the forward strand and the second,
the reverse strand. Above, we gave the example of the forward strand
in the <i>Bacillus anthracis</i> bacteria. Here&rsquo;s what it looks
like paired with its reverse strand:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">forward
--&gt; ATATTTTTTCTTGTTTTTTATATCCACAAACTCTTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           |||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           TATAAAAAAGAACAAAAAATATAGGTGTTTGAGAAAA &lt;-- reverse</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           </font></font>
</p>
<p class="first-paragraph">The lines that connect the bases on either
side denote a basepair relationship.</p>
<p>&ldquo;Forward&rdquo; and &ldquo;reverse&rdquo; are just labels.
The choice of labeling is arbitrary and does not depend on any
inherent property of the DNA. The forward strand is not &ldquo;special.&rdquo;
Scientists decide which to call &ldquo;forward&rdquo; and which to
call &ldquo;reverse&rdquo; when they first analyze the DNA of an
organism. Even though the decision is arbitrary, it&rsquo;s important
to maintain consistency with that decision for the sake of clear
communication.</p>
<p>The forward and reverse strands may also be denoted with different
terms. For example, in some datasets you may find them labeled as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>.
They might also be called <i>top</i> and <i>bottom</i> strands, or
even <i>Watson</i> and <i>Crick</i> strands.</p>
<p>In our opinion, these variances are needlessly confusing. Please
avoid referring to strands with any other terms than <i>forward</i>
and <i>reverse</i>.</p>
<h2 class="western"><a name="is-there-a-directionality-of-dna"></a>5.2
Is there a directionality of DNA?</h2>
<p class="first-paragraph">Yes, there is a directionality to the DNA
determined by the polarity of the molecules. This direction runs in
opposite ways for each strand. Typically, we indicate this polarity
with arrows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATTTTTTCTTGTTTTTTATATCCACAAACTCTTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TATAAAAAAGAACAAAAAATATAGGTGTTTGAGAAAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                             &lt;------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                          </font></font>
</p>
<p class="first-paragraph">Most biological mechanisms (but not all)
take place on a single strand of the DNA and in the direction of the
arrow. Hence, sequences of the DNA above will be &ldquo;seen&rdquo;
by the biochemical machinery as either:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATTTTTTCTTGTTTTTTATATCCACAAACTCTTTT</font></font></p>
<p class="first-paragraph">or as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAAGAGTTTGTGGATATAAAAAACAAGAAAAAATAT</font></font></p>
<p class="first-paragraph">This last sequence is called the <i>reverse
complement</i> of the first and is formed by reversing the order of
the letters then swapping <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
and swapping <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
(and vice-versa).</p>
<p>Hence a DNA sequence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAACT</font></font>
may need to be considered:</p>
<ul>
	<li/>
<p class="compact">in reverse, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TCAAA</font></font></p>
	<li/>
<p class="compact">as a complement, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTGA</font></font></p>
	<li/>
<p class="compact">as a reverse-complement, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGTTT</font></font></p>
</ul>
<h2 class="western"><a name="what-is-senseantisense"></a>5.3 What is
sense/antisense?</h2>
<p class="first-paragraph">When a process occurs in the expected
direction, its directionality may be called <i>sense</i>; if it is
going against the normal direction, its directionality may be called
<i>anti-sense</i>.</p>
<p>It is very important not to collate the concepts of
<i>forward</i>/<i>reverse</i> with <i>sense</i>/<i>anti-sense</i>, as
these concepts are completely unrelated. The <i>sense/anti-sense</i>
is relative to a sequence&rsquo;s direction; the sequence, in turn,
may come from a <i>forward</i> or <i>reverse</i> strand.</p>
<h2 class="western"><a name="what-is-dna-sequencing"></a>5.4 What is
DNA sequencing?</h2>
<p class="first-paragraph">DNA sequencing is the &ldquo;catch-all&rdquo;
terminology that describes all the processes for identifying the
composition of a DNA macromolecule.</p>
<p>The results of a DNA sequencing process are data files stored in
an unprocessed format, typically either <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ,</font></font>
or unaligned <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
(called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">uBAM</font></font>)
files. Most published papers also store their data in repositories
that can be downloaded for reanalysis.</p>
<h2 class="western"><a name="what-gets-sequenced"></a>5.5 What gets
sequenced?</h2>
<p class="first-paragraph">It is essential to note that instruments
do not directly sequence the DNA in its original form. Sequencing
requires a laboratory process that transforms the original DNA into a
so-called &ldquo;sequencing library&rdquo; - an artificial construct
based on the original DNA. The process of creating this sequencing
library introduces a wide variety of limitations and artificial
properties into the results. Also, the method of building the
sequencing library will also limit the information that can be
learned about the original DNA molecule.</p>
<p>Most (perhaps all) life scientists use the term &ldquo;sequencing&rdquo;
over-generously and are often under the assumption that it produces
more precise information than what it can deliver.</p>
<h2 class="western"><a name="what-is-a-genome"></a>5.6 What is a
genome?</h2>
<p class="first-paragraph">A genome is all of an organism&rsquo;s DNA
sequence. Each cell typically contains a copy of the entire genome.
More accurately, each cell has one or more nearly identical copies of
the genome. Replication is the process of duplicating the genome when
a cell divides.</p>
<p>While, due to complementarity, the number of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
nucleotides and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
nucleotides is equal, the relative ratios of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AT</font></font>
pairs vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CG</font></font>
pairs may be very different. Some genomes may contain more <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AT</font></font>
pairs while others may contain more <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CG</font></font>
pairs.</p>
<h2 class="western"><a name="what-is-a-genomes-purpose"></a>5.7 What
is a genome&rsquo;s purpose?</h2>
<p class="first-paragraph">The genome contains the information that
makes the functioning of an organism possible. In cellular organisms,
for example, the genome has regions that contain the instructions for
making proteins. These are typically called &ldquo;coding regions&rdquo;.</p>
<p>The genome may also have regions used to produce molecules other
than proteins, as well as regions that regulate the rates by which
other processes take place.</p>
<p>All genomes are subject to evolutionary principles. Hence, some
(or even substantial) parts of a healthy genome may be non-functional
and may no longer serve any obvious purpose. Some percent of a genome
may consist of copies of various kinds of interspersed, repeat
sequences. At some point, biologists labeled these regions &ldquo;junk
DNA&rdquo;, but the term has since become a lightning rod of
controversy.</p>
<p>Identifying non-functional regions has turned out to be more
difficult, and quite more controversial than biologists originally
anticipated. The so-called <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/C-value#C-value_paradox">C
Value Paradox</a></font> captures the observation that the size of a
genome does not directly determine the genome&rsquo;s complexity.</p>
<h2 class="western"><a name="how-big-is-a-genome"></a>5.8 How big is
a genome?</h2>
<p class="first-paragraph">Functional genomes range from as short as
the 300 base pairs of the prions that cause the mad-cow disease to as
long as the 150 billion base pairs of <i>Paris japonica</i>, a rare,
perennial, star-like flower from Japan. It is common to refer to
genome sizes in terms of kilo-bases (thousands), mega-bases
(millions), and giga-bases (billions).</p>
<p>Here are some other genome sizes:</p>
<ul>
	<li/>
<p class="compact">Ebola virus genome: 18 thousand basepairs
	(18Kb)</p>
	<li/>
<p class="compact">E.coli bacteria genome: 4 million basepairs
	(4Mb)</p>
	<li/>
<p class="compact">Baker&rsquo;s yeast fungus genome: 12
	million basepairs (12Mb)</p>
	<li/>
<p class="compact">Fruit fly genome: 120 million basepairs
	(120Mb)</p>
	<li/>
<p class="compact">Human genome: 3 billion basepairs (3Gb)</p>
	<li/>
<p class="compact">Some salamander species: 120 billion
	basepairs (120 Gb)</p>
</ul>
<h2 class="western"><a name="what-is-rna"></a>5.9 What is RNA?</h2>
<p class="first-paragraph">Whereas DNA is the blueprint, RNA is a
smaller interval of this plan translated into a new kind of molecule.
This molecule is similar to DNA in that it differs only owing to a
few chemical modifications that change its properties relative to
DNA. For example, in RNA, the base <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
(Thymine) is replaced by the nucleotide <font face="Consolas, serif"><font size="2" style="font-size: 11pt">U</font></font>
(Uracil).</p>
<p>RNA is a polymeric, single-stranded molecule that often performs a
function and is believed to exist transiently. Unlike DNA, there are
many classes of RNA: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mRNA</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tRNA</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rRNA,</font></font>
and many others. DNA is continuously present in the cell, whereas RNA
degrades quickly over time (minutes).</p>
<h2 class="western"><a name="how-does-a-genome-function"></a>5.10 How
does a genome function?</h2>
<p class="first-paragraph">The genome has numerous functions, most of
which are too complicated (or insufficiently studied) to describe
accurately. Instead, scientists employ a &ldquo;biological narrative&rdquo;
to describe genomic functions and processes.</p>
<p>The &ldquo;narrative story&rdquo; is made very simple by design,
often accompanied by a simplified illustration of events: step 1
followed by step 2 followed by step 3, etc. While this helps
initially, it also runs of risk of explaining concepts in an
oversimplified manner. To give you an example of how this looks in
practice, here is a &ldquo;biological narrative&rdquo; of what is
perhaps the most studied genomic phenomenon: <i>primary mRNA
transcription</i> in eukaryotic cells.</p>
<p class="block-text-western">&ldquo;The cell begins by transcribing
a&rdquo;gene&rdquo; (see later) into an RNA molecule. After
transcription, pieces of the RNA are cut out and discarded in a
process called &ldquo;splicing.&rdquo; Each discarded piece is
referred to as an <i>intron</i>. Each piece between consecutive
introns is known as an <i>exon</i>, and the RNA molecule with the
introns removed is known as messenger RNA, or <i>mRNA</i>.</p>
<p class="block-text-western">Part of the cell&rsquo;s method for
identifying introns is the presence of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AG,</font></font>
the so-called &ldquo;splice signals&rdquo; that usually occur as the
first and last dinucleotides of an intron. But the mere presence of
these dinucleotides is not sufficient for splicing to occur. Perhaps
35% of human genes are alternatively spliced, meaning that under
different circumstances, various combinations of exons are selected.&rdquo;</p>
<p class="first-paragraph">Remember, again, that the above
explanation is a simplistic description of a far more complex
phenomenon. Any single detail of it opens a series of yet
unexplainable questions: why is it that only a small subset of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AG</font></font>
pairs cause splicing and how come the polymerase that initiates a
splicing event at a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
knows that there will be a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AG</font></font>
later? Or perhaps it does not know and it is all random - but that
would have other implications. You see how the rabbit hole only gets
deeper and deeper.</p>
<h2 class="western"><a name="what-is-a-protein"></a>5.11 What is a
protein?</h2>
<p class="first-paragraph">A protein is a three-dimensional
macromolecule built from a series of so-called &ldquo;amino acid&rdquo;
molecules that can form a 3D structure. There are 20 kinds of amino
acids that can form a protein; these are labeled as letters in the
&ldquo;alphabet&rdquo; of protein sequences, where each letter is an
amino acid. The &ldquo;alphabetical order of the amino acids alphabet
is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ARNDCEQGHILKMFPSTWYZ</font></font></p>
<p class="first-paragraph">Where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
stands for Alanine, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">R</font></font>
for Arginine, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
for Asparagine <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/DNA_codon_table">and
so on</a></font></p>
<p>Proteins can be described by their sequence alone. But in our
current state of understanding, the sequence itself is typically
insufficient to fully determine the protein&rsquo;s 3D structure or
function.</p>
<p>Whereas DNA and mRNA usually carry <i>information</i>, proteins
are the actual, physical building blocks of life. Every living
organism is built out of proteins and functions via the interactions
of proteins that are being continuously produced.</p>
<p>A short series (less than 40) of amino acids without a
well-defined 3D structure is called a polypeptide (peptides).</p>
<h2 class="western"><a name="how-are-proteins-made"></a>5.12 How are
proteins made?</h2>
<p class="first-paragraph">The process of reading <i>DNA</i> and
creating <i>mRNA</i> out of it is called <i>transcription</i>. Then,
in eukaryotes, the <i>mRNA</i> is transported out of the nucleus (to
the cell&rsquo;s cytoplasm), where it is converted to a protein
sequence in a process called <i>translation</i>. Multiple proteins
(even hundreds) may be translated from a single mRNA molecule.</p>
<ul>
	<li/>
<p class="compact">transcription: DNA &ndash;&gt; mRNA</p>
	<li/>
<p class="compact">translation: mRNA &ndash;&gt; Protein</p>
</ul>
<p class="first-paragraph">To perform the translation, the mRNA is
partitioned into units of three consecutive letters, each called a
<i>codon</i>. A <i>codon</i> is then translated via a <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/DNA_codon_table">translation
table</a></font> into an <i>amino acid</i>. A protein is a sequence
of amino acids.</p>
<p>There is a genetic code for translating a codon into an amino
acid. For example, the codon <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TCA</font></font>
(or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">UGA</font></font>
if we describe it as an RNA sequence) codes for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">S</font></font>,
the amino acid <i>Serine</i>. There are 64 combinations of codons but
only 20 amino acids. Some codons code for the same amino acid, this
is called the <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Codon_degeneracy">codon
degeneracy</a></font>. For example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGU,
CGC, CGA, CGG, AGA, AGG --&gt; Arginine (Arg/R)</font></font></p>
<p class="first-paragraph">Most (but not all) amino acids are encoded
by more than one codon. A few codons, in most cases <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TAA</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TAG</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGA</font></font>
are the so called <i>stop codons</i>; translation stops when one of
them is reached.</p>
<p>The translation process begins with the so-called start codon <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATG</font></font>
that corresponds to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>
(the <i>Methionine</i> amino acid).</p>
<p>For example the subunit B of the Cholera Toxin protein complex
responsible for the massive, watery diarrhea characteristic of
cholera infection has the following sequence:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MIKLKFGVFFTVLLSSAYAHGTPQNITDLCAEYHNTQIYTLNDKIFSYTESLAGKREMAI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ITFKNGAIFQVEVPGSQHIDSQKKAIERMKDTLRIAYLTEAKVEKLCVWNNKTPHAIAAI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SMAN</font></font></p>
<p class="first-paragraph">It starts with an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>,
the Methionine, which as we just stated is also the start codon. You
may think all protein sequences should start with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>.</p>
<p>But wait, remember how every rule has exceptions? That applies
here as well. While the overwhelming majority of proteins start with
an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>
it is possible for this <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>
to get cut off during later processes.</p>
<h2 class="western"><a name="what-is-an-orf"></a>5.13 What is an ORF?</h2>
<p class="first-paragraph">An <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORF</font></font>,
or <i>open reading frame</i>, is a sequence of at least, say, 100
consecutive codons without a stop codon.</p>
<p>It is important also to note that while there is a &ldquo;standard&rdquo;
translation table, some organisms may use slight variations of it.
These are called <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/List_of_genetic_codes">genetic
codes</a></font>. When translating from genome to protein, the use of
the correct genetic code is essential.</p>
<h2 class="western"><a name="what-is-a-gene"></a>5.14 What is a gene?</h2>
<p class="first-paragraph">In our experience, the word gene is one of
the most misused words in the entirety of biology. It is usually
interpreted quite differently by different scientists. Some think of
genes as the DNA that codes for proteins; others think genes may
include upstream elements, and so on. The &ldquo;official&rdquo;
definition of the <font color="#4f81bd"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000704">term
gene</a></font> in the Sequence Ontology is</p>
<p class="block-text-western">A region (or regions) that includes all
of the sequence elements necessary to encode a functional transcript.
A gene may include regulatory regions, transcribed regions and other
functional sequence regions.</p>
<h2 class="western"><a name="do-genomes-have-other-features"></a>5.15
Do genomes have other features?</h2>
<p class="first-paragraph">Genomic regions may have a wide variety of
functions. Here are a few often-studied types:</p>
<h3 class="western"><a name="untranslated-regions"></a>5.15.1
Untranslated regions</h3>
<p class="first-paragraph">The region of the mRNA before the start
codon (or the corresponding genomic region) is called the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5'
UTR</font></font> (5 prime UTR), or untranslated region; the portion
from the stop codon to the start of the poly-A tail is the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3'
UTR</font></font> (three prime UTR).</p>
<h3 class="western"><a name="promoter-regions"></a>5.15.2 Promoter
regions</h3>
<p class="first-paragraph">The genomic region just before the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5'
UTR</font></font> may contain patterns of nucleotides, called the
promoter, that are used to position the molecular machinery that
performs the transcription to RNA. For about 60% of human genes, the
promoter occurs in a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CpG</font></font>
island: a region of DNA where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
are far more frequent than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>.
Other patterns in the DNA tell the cell when (how frequently and in
what tissues) to transcribe the gene; that is, they regulate
transcription. A pattern that increases the frequency of
transcription operations is an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">enhancer</font></font>,
while one that decreases the rate is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">silencer</font></font>.</p>
<h3 class="western"><a name="cpg-islands"></a>5.15.3 CpG islands</h3>
<p class="first-paragraph">CpG islands are regions of DNA where a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
(cytosine) nucleotide is followed by a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
guanine nucleotide in the linear sequence of bases along its 5&rsquo;
-&gt; 3&rsquo; direction. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CpG</font></font>
is shorthand for 5&rsquo;&mdash;C&mdash;phosphate&mdash;G&mdash;3&rsquo;,
that is, cytosine and guanine separated a phosphate; phosphate links
any two nucleosides together in DNA. Promoter regions may contain CpG
islands.</p>
<p>Cytosines in CpG dinucleotides can be methylated. In turn,
methylated cytosines within a gene may change the gene&rsquo;s
expression, a mechanism that is part of a larger field of science
studying gene regulation, called epigenetics.</p>
<p>Below is a section of a CpG island from chromosome 1 of the human
genome. It has 30 CpGs and GC% is 77.14.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;gi|568815597:36306860-36307069
Homo sapiens chromosome 1, GRCh38.p7 Primary
Assembly</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGGGCTCCGGAGAGGCGCGGAGGCCGCGCTGTGCGCGCCGCCGAGGTGAGCGCAAGGGCGGGGACGGGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCCGGTGGGCGGGTGCACGGAGCCAGTGCGACCCCGGCGTCTCCGGCTCTTAGTGACGGGCGCGGCTCTG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGCGGGACCTCGGGGCCGCCCTGCGGTCTGTGATTGGTTCTCGAGTGCAATGCTCCGCCCTGGGGCGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<p class="first-paragraph">We obtained the above via:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_000001.11 -format fasta -seq_start=36306860
-seq_stop=36307069</font></font></p>
<p class="first-paragraph">This tool will be covered in later
sections.</p>
<h3 class="western"><a name="enhancers"></a>5.15.4 Enhancers</h3>
<p class="first-paragraph">Enhancers help a gene turn on and regulate
where, when, and how much the gene should be expressed for the cell
to function properly. One enhancer can regulate many genes, and one
gene can have many enhancers. Since DNA is folded and crumpled up in
a cell, an enhancer can be far away from a gene it regulates in terms
of distance measured along the DNA strand, but physically very close
to the gene promoter. The best way to identify a sequence as an
enhancer is to experimentally disrupt it and observe gene expression
changes. Since this is cumbersome in practice, we use proxies like
histone modifications (chemical decorations on the structural
backbone upon which DNA is coiled) to identify enhancers. Silencers
work similarly &ndash; they can be far from a gene, many-to-many
relationships, hard to find &ndash; but cause a gene to be less,
rather than more, expressed.</p>
<h2 class="western"><a name="what-is-homology"></a>5.16 What is
homology?</h2>
<p class="first-paragraph">Two regions of DNA that evolved from the
same sequence (through processes of duplication of genomic regions
and separation of two species) are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">homologous</font></font>,
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">homologs</font></font>
of one another.</p>
<p>More specifically, regions in the genomes of two species that are
descended from the same area in a common ancestor&rsquo;s genome are
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">orthologs</font></font>.
These regions are said to be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">orthologous</font></font>.</p>
<p>On the other hand, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">paralogous</font></font>
sequences or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">paralogs</font></font>
were separated by duplication of a genomic region within the same
genome.</p>
<p>Standard pet peeve of many scientists:</p>
<p class="block-text-western">Homology is not a synonym of sequence
similarity!</p>
<p class="block-text-western">Homologous sequences are usually
similar to one another, but similarity of sequences does not indicate
homology.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-is-bioinformatics-practiced"></a>6
How is bioinformatics practiced?</h1>
<p class="first-paragraph">Bioinformatics requires a broad skill set.
The diverse tasks undertaken by bioinformaticians can be roughly
divided into three tiers:</p>
<p><b>1. Data management</b>.</p>
<p>Data management requires accessing, combining, converting,
manipulating, storing, and annotating data. It requires routine data
quality checks, summarizing large amounts of information, and
automating existing methods.</p>
<p><b>2. Primary data analysis</b>.</p>
<p>Analysis of the data requires running alignments, variation
callers, and RNA-Seq quantification, as well as finding lists of
genes. Strategically, the bioinformatician must anticipate potential
pitfalls of planned analyses, find alternatives, and adapt and
customize the methods to the data.</p>
<p><b>3. Data interpretation</b>.</p>
<p>Data management and analysis are meaningless without accurate and
insightful interpretation. Bioinformaticians discover or support
biological hypotheses via the results of their analyses, and so they
must be able to interpret their findings in the context of ongoing
scientific discourse.</p>
<h2 class="western"><a name="what-is-the-recommended-computer-for-bioinformatics"></a>
6.1 What is the recommended computer for bioinformatics?</h2>
<p class="first-paragraph">In our experience, the most productive
setup for a bioinformatician is using a Mac OS based computer to
develop and test the methods and then using a high-performance Linux
workstation&ndash;&ndash;or cluster&ndash;&ndash;to execute these
pipelines on the data. With the release of Windows 10 Creator Edition
(accessible for users that are part of the Windows Insider program),
a fully functional Linux Bash shell can be installed and run on
Windows. See the <font color="#4f81bd"><a href="#ch009.xhtml#computer-setup">Computer
Setup</a></font> page for more details.</p>
<h2 class="western"><a name="how-much-computing-power-do-we-need"></a>
6.2 How much computing power do we need?</h2>
<p class="first-paragraph">Bioinformatics methodologies are improving
at a rapid pace. A regular workstation is often all you need to
analyze data from a typical, high-volume sequencing run (hundreds of
millions of measurements). For example, you could handle most RNA-Seq
data analysis required for a given day by using only the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hisat2</font></font>
aligner on a standard iMac computer (32GB RAM and eight cores used in
parallel). More substantial analyses like genome assembly, however,
typically require more of memory than what is generally available on
a standard desktop computer.</p>
<p>As a rule, low-quality data (contamination, incorrect sample prep,
etc.) takes substantially longer to analyze than high-quality data.</p>
<h2 class="western"><a name="does-learning-bioinformatics-need-massive-computing-power"></a>
6.3 Does learning bioinformatics need massive computing power?</h2>
<p class="first-paragraph"><b>No!</b></p>
<p>Mastering bioinformatics requires nothing more than a standard,
UNIX-ready laptop. You can learn bioinformatics using only a $100
Chromebook.</p>
<p>Of course, the computational needs for applying the bioinformatics
methods on particular problems will depend on the amount and scale of
data being analyzed. A $100 Chromebook has the tools one needs but
probably doesn&rsquo;t have the compute power, storage or memory to
run analyses of the massive datasets that may need to be processed.</p>
<p>You might be surprised, however, just how much can be done on just
a Chromebook!</p>
<p>Remember this; you don&rsquo;t need an expensive or powerful
system to become one of the best bioinformaticians on the planet! You
can learn, practice and gain a profound understanding of this entire
field on any simple, Unix capable computing device.</p>
<h2 class="western"><a name="what-about-the-cloud"></a>6.4 What about
the cloud?</h2>
<p class="first-paragraph">Cloud computing is becoming an
increasingly common platform for bioinformatics, in particular for
efforts that allow bioinformaticians to &ldquo;bring the tools to the
data&rdquo; - applying your algorithms to massive sequence databases.
Cloud services such as Amazon Web Services also enable anyone to host
web applications and services on powerful, accessible machines while
only paying for the compute time they use.</p>
<p>Running a cloud service involves learning about object stores,
virtual private clouds, file systems, and security. Building and
maintaining a cloud instance requires you become something like a
part-time systems administrator, which can distract you from learning
bioinformatics. In later chapters, we will introduce both cloud
computing and cloud-based stacks - commonly referred to as
software-as-a-service (SaaS), platform-as-a-service (PaaS, and
infrastructure-as-a-service (IaaS). These offerings, combined with
<i>containerized</i> applications and analyses, have become valued
for their potential to provide a scalable platform for reproducible
research.</p>
<p>If you are comfortable with these concepts, a cloud-based service
can be a companion to this guide, but learning bioinformatics <i>does
not require</i> the cloud.</p>
<h2 class="western"><a name="do-i-need-to-know-unix-to-do-bioinformatics"></a>
6.5 Do I need to know Unix to do bioinformatics?</h2>
<p class="first-paragraph">Bioinformatics has primarily been
developed via freely available tools written on the Unix platform.
The vast majority of new advances are published with software written
for Unix-based operating systems.</p>
<p>It&rsquo;s unlikely you&rsquo;ll be in a position for long-term
career advancement as a bioinformatician without a basic
understanding of the command line. Starting with 2016, the Windows
platform has begun to offer the so-called &ldquo;Bash for Windows&rdquo;
application that allows the Windows 10 operating system to run almost
all of the Unix-specific software.</p>
<p>The good news is that using Unix is not that complicated. While
somewhat unforgiving, the Unix philosophy is logical and precise. In
our experience, even people with no prior computational background
can become confident Unix users in a matter of weeks.</p>
<h2 class="western"><a name="do-i-need-to-learn-a-programming-language"></a>
6.6 Do I need to learn a programming language?</h2>
<p class="first-paragraph">Yes &ndash;&ndash; an introductory level
of programming ability (in any programming language) will be
necessary.</p>
<p>Fortunately, programming languages function by similar underlying
logic &ndash;&ndash; even when they seem entirely different on the
surface. While symbols, syntax, and procedures may differ among
programming languages, an introductory level of programming ability
in any language is necessary to understand the thought processes
required for computational analysis.</p>
<p>We do cover programming concepts in this book. See the section
titled <font color="#4f81bd"><a href="#ch006.xhtml#someprogramming">Some
programming required</a></font> and the corresponding subsections for
a primer.</p>
<h2 class="western"><a name="are-there-alternatives-to-using-unix"></a>
6.7 Are there alternatives to using Unix?</h2>
<p class="first-paragraph">Alternatives to Unix are limited but do
exist. They fall into two categories:</p>
<h3 class="western"><a name="software-that-provides-a-web-interface-to-the-command-line-tools"></a>
6.7.1 Software that provides a web interface to the command line
tools</h3>
<p class="first-paragraph">These software systems run the same
command line tools that one could install into a Unix-based system,
but the interfaces to these tools are graphical and provide better
information &ldquo;discoverability.&rdquo;</p>
<ol>
	<li/>
<p class="compact">Galaxy (open source)</p>
	<li/>
<p class="compact">GenePattern (open source)</p>
	<li/>
<p class="compact">BaseSpace (commercial)</p>
	<li/>
<p class="compact">DNA Nexus (commercial)</p>
	<li/>
<p class="compact">Seven Bridges Genomics (commercial)</p>
</ol>
<h3 class="western"><a name="systems-that-offer-custom-implementations-of-bioinformatics-methods."></a>
6.7.2 Systems that offer custom implementations of bioinformatics
methods.</h3>
<p class="first-paragraph">These can be standalone software that runs
on your local system &ndash;&ndash; they are compatible with
web-based applications. They often run open-source tools via a GUI.</p>
<ol>
	<li/>
<p class="compact">CLC Genomics Workbench</p>
	<li/>
<p class="compact">Partek</p>
	<li/>
<p class="compact">Golden Helix</p>
	<li/>
<p class="compact">SoftGenetics</p>
	<li/>
<p class="compact">DNA Star</p>
	<li/>
<p class="compact">Geneious</p>
</ol>
<h2 class="western"><a name="what-is-bioconductor"></a>6.8 What is
Bioconductor?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="https://www.bioconductor.org/">Bioconductor</a></font>
is an open-source software project for the analysis and comprehension
of genomic data. Bioconductor is based primarily on the <font color="#4f81bd"><a href="https://www.r-project.org/">R
statistical programming language</a></font></p>
<p>Users of Bioconductor typically write scripts using R that make
calls to functions defined in the Bioconductor library. For example,
a Bioconductor-based script that runs an RNA-Seq data analysis could
look like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">biocLite(&quot;DESeq&quot;)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">library(DESeq)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">count
= read.table(&quot;stdin&quot;, header=TRUE, row.names=1 )</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cond1
= c(&quot;control&quot;, &quot;control&quot;, &quot;control&quot;)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cond2
= c(&quot;treatment&quot;, &quot;treatment&quot;, &quot;treatment&quot;)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conds
= factor(c(cond1, cond2))</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cdata
= newCountDataSet(count, conds)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esize
= estimateSizeFactors(cdata)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">edisp
= estimateDispersions(esize)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rdata
= nbinomTest(edisp, &quot;control&quot;,
&quot;treatment&quot;)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">write.table(rdata,
file=&quot;&quot;, sep=&quot;\t&quot;, row.name=FALSE, quote=FALSE)</font></font></p>
<h2 class="western"><a name="what-is-galaxy"></a>6.9 What is Galaxy?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://www.usegalaxy.org/">Galaxy</a></font>
is a web-based scientific workflow and data analysis platform that
aims to make computational biology accessible to research scientists
with no computer programming experience.</p>
<p>Galaxy can generate and present a web-based user interface to
existing command-line tools, making them available to those unable to
run the tools themselves.</p>
<p><b>Important:</b> Galaxy is not a data analysis tool! It is an
<b>Interface</b> and a <b>Platform</b> that provides computational
capacity that runs <b>command line</b> bioinformatics software on the
users&rsquo; behalf. Galaxy makes it seem simple to launch tools and
manage data &ndash; but it is the user&rsquo;s responsibility to
understand the underlying concepts and use these tools correctly.</p>
<p><b>Pet peeve</b>: <i>Please don&rsquo;t say I used Galaxy to
analyze data</i>. That is like saying I used a computer to do
bioinformatics. As stated above Galaxy is an interface and a platform
that uses tools developed by scientists around the world. Recognize
the tools authors, cite the publications for the software, then
mention and recognize Galaxy.</p>
<p>Galaxy can also serve as graduation moment in a bioinformaticians&rsquo;
life. We&rsquo;ve never met an expert bioinformatician that would
voluntarily want to use Galaxy for their own needs. You see the
&ldquo;ease&rdquo; of use of Galaxy is, in the end, nothing more than
<i>an illusion</i>. A skilled scientist will always prefer to specify
their work via set words such as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem file1.fq file2.fq | samtools view -F 68 | samtools sort &gt;
result.bam</font></font></p>
<p class="first-paragraph">instead of describing it as a series of
clicks in an interface.</p>
<p>In our opinion, your comfort level of using Galaxy relative to
using the command line quantifies your skill level. The moment you
feel the drag, the slowdown, the limitations and that
&ldquo;straight-jacket&rdquo; feel on your creativity that Galaxy
imposes on you is the moment where you have graduated from the
<i>Elementary School of Galaxy</i> and you&rsquo;ve entered the
<i>Academy of the Command Line</i>.</p>
<p>But by all means, if Galaxy works for you keep using it, it is a
progression step.</p>
<p>And let&rsquo;s make one thing clear, the above sentiment is valid
for just about all &ldquo;easy to use&rdquo; graphical bioinformatics
data analysis platforms. There may be a time when you too will
recognize that &ldquo;easy-to-use&rdquo; is, in reality, a
restriction, a curse, it means that you are limited in what you can
do and that you will have to operate inside of a glass box with
invisible walls.</p>
<p>Rant over :-). Back to our regularly scheduled content.</p>
<p>Galaxy uses a three-panel system where the left panel lists the
tools, the middle panel describes the tools, and the right panel
displays user data.</p>
<p class="compact"><img src="0000_html_9c4a77da19186e1b.png" name="Image13" align="bottom" hspace="12" width="560" height="349" border="0"/>
</p>
<p class="compact">Galaxy Interface</p>
<h2 class="western"><a name="what-is-basespace"></a>6.10 What is
BaseSpace?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://www.illumina.com/informatics/research/sequencing-data-analysis-management/basespace.html">BaseSpace</a></font>
is a cloud-based genomics analysis and storage platform provided by
Illumina that directly integrates with all Illumina sequencers.</p>
<p>The system can operate only with sequencing data produced by
Illumina and only when the data is in the formats generated by the
sequencers.</p>
<h2 class="western"><a name="are-commercial-bioinformatics-software-packages-expensive"></a>
6.11 Are commercial bioinformatics software packages expensive?</h2>
<p class="first-paragraph">From an individual, consumer-level
perspective, the cost of commercial bioinformatics software is high
&ndash;&ndash; typically measured in thousands of dollars. The price
tag means that few people could or would buy them directly.</p>
<p>For larger organizations, the software costs may pay off quickly
in productivity gains as long as the software is chosen carefully to
match specific research needs.</p>
<p>Most academicians and open-source developers treat commercial
software with some distrust. In our opinion, this distrust may be
unwarranted. Many commercial tools cater to specific needs and need
to be operated by individuals with diverse levels of training. Also,
the goals and resource allocations in academia are typically very
different than those of private organizations.</p>
<h2 class="western"><a name="should-i-freelance-as-a-bioinformatician"></a>
6.12 Should I freelance as a bioinformatician?</h2>
<p class="first-paragraph">In our opinion, the answer is <b>no</b>
unless the terms are generous or your clients appear to recognize the
potential complexities that you may run into.</p>
<p>Successful bioinformatics analysis requires more than skill alone:
the information content of the data is a crucial second piece. The
world&rsquo;s best bioinformatician won&rsquo;t be able to extract
information that is not present in the data and may have trouble
demonstrating that they have done their best.</p>
<p>Our experiences line up with the observations in <font color="#4f81bd"><a href="http://www.nature.com/news/core-services-reward-bioinformaticians-1.17251">Core
services: Reward bioinformaticians, Nature, (2015)</a></font> that
state that analyses always end up being substantially more
complicated than anticipated.</p>
<p>It is unexpectedly difficult to estimate the effort that a project
requires accurately.</p>
<p class="block-text-western"><b>Amusing story</b></p>
<p class="block-text-western">I once got called up by a company
offering custom RNA-Seq analysis. It was a quite basic service as far
as analyses go. They ran the Tuxedo pipeline and provided the files
it generated. Once we established what the service consisted of I
asked how much it cost. The answer was surprising:</p>
<p class="block-text-western">&ldquo;For educational institutions, we
charge <i>two fifty</i> per sample.&rdquo;</p>
<p class="block-text-western">I was both stupefied and elated by this
answer. Here I am, I thought I knew all about how to estimate the
cost of bioinformatics analyses. Yet I am unable to decide what <i>two
fifty</i> meant. A typical RNA-Seq at that time might have had ten
samples. So:</p>
<ol>
	<li/>
<p class="block-text-western">Was it $2.50 per sample, that
	would sum up to a ridiculously low $25, that adds up to less than
	the cost of owning the computing power for an analysis of this type.</p>
	<li/>
<p class="block-text-western">Or was it $250 per sample, that
	would sum up to a ridiculously high $2500, for executing a
	straightforward command line tool that at most would take an hour to
	set up and maybe some hours to run.</p>
</ol>
<p class="block-text-western">I meekly inquired: <i>What is two
fifty? Is it $2.50?</i></p>
<p class="block-text-western">*** glacial pause **</p>
<p class="block-text-western">Answer: <i>No it is $250</i></p>
<p class="first-paragraph">Moral of the story, if someone pays $250
per sample, you should take the job. If someone pays $2.50 then
better don&rsquo;t.</p>
<h2 class="western"><a name="what-do-bioinformaticians-look-like"></a>
6.13 What do bioinformaticians look like?</h2>
<p class="first-paragraph">Handsome of course!</p>
<p>Here are some participants at the NSF-funded &ldquo;Analyzing Next
Generations Sequencing Data&rdquo; workshop at Michigan State
University (2014). Graduate students, postdocs, researchers, faculty,
were all represented.</p>
<p class="compact"><img src="0000_html_a228336f6a33d228.jpg" name="Image14" align="bottom" hspace="12" width="560" height="420" border="0"/>
</p>
<p class="compact">MSU Workshop</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-solve-it"></a>7 How to solve it</h1>
<p class="first-paragraph">This section has been inspired by the book
titled <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/How_to_Solve_It">How
to solve it</a></font> by George P&oacute;lya describing methods of
problem-solving. His advice can be summarized via:</p>
<ol>
	<li/>
<p class="compact">First, you have to understand the problem.</p>
	<li/>
<p class="compact">After understanding, then make a plan.</p>
	<li/>
<p class="compact">Carry out the plan.</p>
	<li/>
<p class="compact">Look back on your work. How could it be
	better?</p>
</ol>
<p class="first-paragraph">The advice is general and timeless, but
how does it apply to Bioinformatics?</p>
<h2 class="western"><a name="what-is-holistic-data-analysis"></a>7.1
What is holistic data analysis?</h2>
<p class="first-paragraph">In our opinion, one of the most essential
characteristics of the analysis is to think about it &ldquo;holistically&rdquo;
&ndash; considering all steps that need to be performed.</p>
<p>The best analogy is drawing - usually artists will start with a
rough draft. In time, they will add more detail and accuracy to their
work. The same applies to any data analysis - do not get bogged down
with the initial details - make it first work on a rough scale - so
that you see all the moving components that will make up your final
result. Only when you have all the parts together will you have a
firm understanding of the strengths and weaknesses of the entire
process.</p>
<p>Imagine and visualize what your result will need to look like and
work toward that goal.</p>
<p><img src="0000_html_7c56d79a58c810d6.png" name="Image15" align="bottom" hspace="12" width="560" height="134" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-perform-a-holistic-analysis"></a>
7.2 How do I perform a holistic analysis?</h2>
<p class="first-paragraph">The most straightforward approach is to
isolate a small section of your data and build your analysis around
that. Don&rsquo;t start with processing hundreds of gigabytes of data
against genomes with billions of bases.</p>
<p>Start small and stay in control. You will be able to build
pipelines that run in minutes (or even seconds), and you will be
continually aware of what is taking place, what problems are cropping
up, and what the results look like. You&rsquo;ll be able to identify
very early if something is amiss. Your analysis will run quickly even
on a small laptop.</p>
<p>Do not try to optimize a process too early. Build your pipeline to
be flexible and robust that you can rerun with other techniques if
you need to. You will only really understand the problems that you
need to solve when you get near the end.</p>
<p>How to select these subsets? Here are some examples we employ in
this book:</p>
<ol>
	<li/>
<p class="compact">Download just a subset of the data. For
	instance in all our cases that use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
	we limit data to values between 10,000 or 1 million reads. You can
	tweak these numbers to find that balance where the results are quick
	and allow you to understand and iterate over the results.</p>
	<li/>
<p class="compact">Build random samples out of your data.
	Tools such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqtk</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit</font></font>
	and others will let you randomly select a much smaller subset of
	data.</p>
	<li/>
<p class="compact">Use a smaller segment of the target genome.
	If you study large genomes, use just a single chromosome of it. For
	example, build your pipeline but execute it using only human
	chromosome 22 instead of the entire human genome. Then swap out the
	reference file when you are ready to do it all.</p>
	<li/>
<p class="compact">Generate data with known properties. We
	have a chapter on simulating data - and it is, in fact, the best way
	to understand the features and limits of the various processes that
	you encounter. When you analyze data with known properties, you can
	validate and understand how your methods work.</p>
</ol>
<h2 class="western"><a name="what-are-the-rules-of-a-bioinformatics-analysis"></a>
7.3 What are the rules of a bioinformatics analysis?</h2>
<p class="first-paragraph">In our center, written on the whiteboard
is the motto and credo of how we believe progress is made in this
field:</p>
<p><img src="0000_html_fe63640682f72891.jpg" name="Image16" align="bottom" hspace="12" width="560" height="282" border="0"/>
</p>
<ol>
	<li/>
<p class="compact">Make it work.</p>
	<li/>
<p class="compact">Make it fast.</p>
	<li/>
<p class="compact">Make it simple.</p>
</ol>
<h2 class="western"><a name="what-does-make-it-work-mean"></a>7.4
What does &ldquo;make it work&rdquo; mean?</h2>
<p class="first-paragraph">Well, it means get the entire process
moving - not just the beginning - the whole thing. It will be hacky,
it will be squeaky, it may need to be held together with duct tape.
Go all the way! Find a way to reduce your problem so that it finishes
in minutes, perhaps to a maximum one hour. Then you can iterate on
it.</p>
<h2 class="western"><a name="why-is-fast-so-important"></a>7.5 Why is
fast so important?</h2>
<p class="first-paragraph">On the one hand, doing it fast sort of
make sense. Who wouldn&rsquo;t want to do things fast?</p>
<p>But there is more to it than just getting home by six. It is about
the way we think about problems in the subconscious. When a process
takes three days, we instinctively think up questions where we factor
in the time it takes to get the answer. We try to ask more
&ldquo;difficult&rdquo; questions that are &ldquo;worth&rdquo; the
time. If a solution takes just two minutes, we can maintain mental
focus on the problem, and we will investigate it with a very
different approach.</p>
<p>So make your data small, analyze it quickly and ask quick and easy
questions - but lots of them. You will get much further.</p>
<h2 class="western"><a name="what-does-simple-mean"></a>7.6 What does
&ldquo;simple&rdquo; mean?</h2>
<p class="first-paragraph">Simple means identifying the commonality
of the processes, and generalizing the solution to take fewer steps.
It is very likely that once you understand the solution, once it
works and once it is fast, you can find better ways to do the achieve
the same results with even fewer assumptions and requirements. Your
building blocks should be like &ldquo;lego blocks&rdquo; that are
reusable for building any other structure. Turn a specific solution
into a generic one and you&rsquo;ll find yourself immensely more
productive.</p>
<h2 class="western"><a name="how-to-deal-with-anxiety-and-stress"></a>
7.7 How to deal with anxiety and stress?</h2>
<p class="first-paragraph">It may feel strange to talk about anxiety
in a data analysis book. But make no mistake about it - it is
something many of us experience.</p>
<p>Bioinformaticians often work alone because other members of the
group do not have sufficient expertise to contribute. In that case
the data analyst alone needs to make dozens if not hundreds of
decisions and they alone are responsible for applying these
correctly.</p>
<p>In addition most tools can be cryptic and terse, with features
that you may know nothing about. When you save the commands that you
type for posterity (and yes, you should), anyone at any time can
review and criticize these, and any error you make will be visible
and apparent. Being the sole decision maker of this analysis puts the
burden on you alone.</p>
<p>With time the added pressure of this sole responsibility can wear
you down psychologically. After a while, you won&rsquo;t remember why
you made certain decisions, and you might have increasing doubts and
fears. When many months later the paper is accepted and the
publication comes around you may ask yourself: <i>Did I do it right?
What if now, that many people will look at it, they will instantly
find and point out that trivial and glaring mistake that renders
everything invalid?</i>.</p>
<p>One of the saddest, public, episodes of science are described in
the article <font color="#4f81bd"><a href="https://www.washingtonpost.com/business/economy/doubts-about-johns-hopkins-research-have-gone-unanswered-scientist-says/2013/03/11/52822cba-7c84-11e2-82e8-61a46c2cde3d_story.html">Doubts
about Johns Hopkins research have gone unanswered, scientist says</a></font>
a story that follows what we believe a situation where the added
pressure of being to sole data analyst of a significant paper has led
to a tragic outcome. Notably, a recent article points out that
<font color="#4f81bd"><a href="http://www.sciencemag.org/careers/2017/04/phd-students-face-significant-mental-health-challenges">Ph.D.&nbsp;students
face significant mental health challenges</a></font>. The risk of
being affected by the [impostor syndrome] is high.</p>
<p><a href="https://www.washingtonpost.com/business/economy/doubts-about-johns-hopkins-research-have-gone-unanswered-scientist-says/2013/03/11/52822cba-7c84-11e2-82e8-61a46c2cde3d_story.html"></a><img src="0000_html_64cf5138079f894a.png" name="Image17" align="bottom" hspace="12" width="560" height="77" border="0"/>
<a href="https://www.washingtonpost.com/business/economy/doubts-about-johns-hopkins-research-have-gone-unanswered-scientist-says/2013/03/11/52822cba-7c84-11e2-82e8-61a46c2cde3d_story.html"></a></p>
<p>See also: <font color="#4f81bd"><a href="https://caselaw.findlaw.com/md-court-of-appeals/1854627.html">Court
of Appeals of Maryland: DANIEL YUAN v. JOHNS HOPKINS UNIVERSITY</a></font>
Decided: March 29, 2017</p>
<p>For me the stories above are eerily familiar - I wish I had known
about them sooner. You see, earlier in my career I&rsquo;ve struggled
with similar issues. I&rsquo;ve rarely enjoyed the publication of
what later turned out to be some of my most cited papers - in fact;
I&rsquo;ve dreaded the publication date - I saw it as the date when
everyone finds out that I made a mistake, it is all wrong and that I
am incompetent.</p>
<p>Today I know better - and I tell you the same. If you do your due
diligence, follow good practices, keep it simple, keep it straight,
everything will be ok, it does not matter that much anyway.</p>
<p>And you know what? If you have messed up because among the
hundreds of commands you&rsquo;ve entered by accident</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -f 266</font></font></p>
<p class="first-paragraph">instead of the &ldquo;correct&rdquo;
command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -f 226</font></font></p>
<p class="first-paragraph">then whose fault is it? Do you see how
hard it is to catch this error? Especially when it might be buried
among hundreds of equally unreadable statements. I believe that above
it is the absurdity of the tool that bears all responibility. A
well-designed tool should never allow you to make errors of this type
and this magnitude accidentally. As you will see later mistyping a
single digit will radically alter the entire meaning of the command,
yet there is no way to recognize that from just looking at it. The
number that you see there needs to be interpreted as the binary bits
encoded into that integer <font face="Consolas, serif"><font size="2" style="font-size: 11pt">226
= 11100010</font></font> vs.&nbsp;<font face="Consolas, serif"><font size="2" style="font-size: 11pt">266
= 100001010</font></font>.</p>
<p>Gee, thank why you <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtool</font></font>
creators (sarcasm) for providing me with all the rope I need to shoot
myself into the foot!</p>
<p>Psychology says that cursing helps: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
you suck and worse, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAM</font></font>
format is amusingly ill-designed. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATK</font></font>?
That tool is bonkers &hellip; I won&rsquo;t go on, but the list is
practically endless. Bioinformatics, unfortunately, represents all
too well the disconnect of the Ivory Tower from reality, where the
tool makers are &ldquo;too smart for their own good&rdquo; and create
software that is fragile, brittle, difficult to use then trivial to
misuse, and frankly often just plain wrong.</p>
<p>So once you get that off your chest, you can go back to enjoy data
analysis - and if you do mess up&hellip; well, rest assured many,
many others did too. You may have been alone in the lab, but when it
comes to people that published at least one incorrect analysis, you
aren&rsquo;t alone.</p>
<p>There are a lot of us out there.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-not-to-waste-your-time"></a>8 How
not to waste your time</h1>
<p class="block-text-western">Disclaimer: This section is optional.
It contains an opinionated take on the current state of bionformatics
computation, enumerates a few fallacies and false prophecies that
scientists have to continually put up with. No specific data analysis
strategy is covered in this section. You may just as well revisit
this page at a later time.</p>
<h2 class="western"><a name="too-many-rube-goldberg-machines"></a>8.1
Too many Rube-Goldberg machines</h2>
<p class="first-paragraph">Every time I tried to use the tools and
techniques listed below, I found that I&rsquo;ve wasted a lot of time
and gained no benefit at all. I want to save you the trouble.</p>
<p><img src="0000_html_5851c710e273c848.png" name="Image18" align="bottom" hspace="12" width="560" height="118" border="0"/>
</p>
<p>I have come to believe that the solutions below are the
bioinformatics equivalents of <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Rube_Goldberg">Rube
Goldberg machines</a></font>, basically overcomplicated gadgets
performing what should be straightforward tasks.</p>
<p>Your solutions will always be more natural and more
straightforward when done and automated with a simple shell script or
a makefile.</p>
<h2 class="western"><a name="the-false-peaks-of-biology"></a>8.2 The
False Peaks of Biology</h2>
<p class="block-text-western">In mountaineering, a false peak or
false summit is a peak that appears to be the pinnacle of the
mountain, but upon reaching, it turns out the summit is higher. False
peaks can have significant effects on climbers&rsquo; psychological
states by inducing feelings of dashed hopes or even failure.</p>
<p class="first-paragraph">There is a simple explanation of why the
techniques and methods listed on this page even exist. Some are even
supported by continually renewed, multi-million dollar governmental
research grants.</p>
<p>You see, the vast majority of life scientists mistakenly believe
that the primary limitation to understanding biology is their
inability to run some existing software. All these scientists believe
that if they could get over this first hurdle, the <i>software
mountain</i> that they see right in front of them, everything would
start making sense and that they would be well on their way to making
scientific discoveries. They believe all that because they are
<i>computationally illiterate</i> - they have never been taught
computation. Understandably so, since the field of data-oriented
biology is new, going back perhaps ten years.</p>
<p>Compared to the real challenges of bioinformatics, the computing
obstacle that they experience is nothing more than a <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/False_peak">false
peak</a></font>:</p>
<p><img src="0000_html_615081c4d76a5d74.png" name="Image19" align="bottom" hspace="12" width="560" height="84" border="0"/>
</p>
<p>The false peaks of biology are responsible for deeply misguided
efforts like CWL, CyVerse, and Galaxy. These solutions create the
illusion of providing people with computational skills. In reality,
the use of these tools does not provide any of the training required
to scale the next summit. Ironically, the mere existence of these
methods hinders science instead of promoting discovery.</p>
<p>The challenges of biology are part of a much larger <b>computational
challenge</b>.</p>
<p>We lack the proper representation, terminology, and interpretation
of the phenomena that we observe. We have to continually adapt and
make decisions then choose the correct methods then apply them
appropriately based on a deeper understanding of what we did.</p>
<p>Becoming successful in bioinformatics means to be <b>computationally
proficient</b>. Only then you can scale the subsequent computational
peaks independently, rather than using the crutches and dead-end
solutions that well-meaning, yet computationally illiterate
scientists have built for you.</p>
<p>You are in the right place, though, from this book, you&rsquo;ll
learn computation the way Mother Nature intended it to work.</p>
<h2 class="western"><a name="dont-let-architecture-astronauts-scare-you"></a>
8.3 Don&rsquo;t Let Architecture Astronauts Scare You</h2>
<p class="first-paragraph"><img src="0000_html_71da9203c0467e2a.png" name="Image20" align="bottom" hspace="12" width="560" height="75" border="0"/>
</p>
<p>We all want to adopt the best practices, keep up with development,
embrace progress. But we have to be careful when doing so. Someone
being smarter than us does not always mean they are doing more
intelligent things. Let me refer you to the seminal blog titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.joelonsoftware.com/2001/04/21/dont-let-architecture-astronauts-scare-you/">Don&rsquo;t
	Let Architecture Astronauts Scare You</a></font></p>
</ul>
<p class="first-paragraph">In that blog Joel Spolsky makes the
following salient point:</p>
<p class="block-text-western">When great thinkers think about
problems, they start to see patterns.</p>
<p class="block-text-western">[&hellip;]</p>
<p class="block-text-western">Sometimes smart thinkers just don&rsquo;t
know when to stop, and they create these absurd, all-encompassing,
high-level pictures of the universe that are all good and fine but
don&rsquo;t actually mean anything at all.</p>
<p class="block-text-western">[&hellip;]</p>
<p class="block-text-western">These are the people I call
<b>Architecture Astronauts</b>. It&rsquo;s very hard to get them to
write code or design programs because they won&rsquo;t stop thinking
about <b>Architecture</b>.</p>
<p class="first-paragraph">If you want actually to get things done,
you have to recognize the astronauts early on.</p>
<p>Bioinformatics is fertile ground for these astronauts. Biologists
are not experts in computing; they all &ldquo;want to believe.&rdquo;
They always hold out the hope out there is a silver bullet, that a
single and magical solution explains everything. As an example just
about every biologist seems to hope that their incredibly complicated
phenotype can be traced back to a few SNPs.</p>
<p>No wonder that <b>Architecture Astronauts</b> thrive in these
environments.</p>
<h2 class="western"><a name="the-curse-of-docker"></a>8.4 The Curse
of Docker</h2>
<p class="first-paragraph"><img src="0000_html_da425514a037097d.png" name="Image21" align="bottom" hspace="12" width="560" height="84" border="0"/>
</p>
<p>What is <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Docker_(software)">Docker</a></font>?
Wikipedia says:</p>
<p class="block-text-western">Docker is a set of
platform-as-a-service (PaaS) products that use OS-level
virtualization to deliver software in packages called containers.
Containers are isolated from one another and bundle their own
software, libraries, and configuration files; they can communicate
with each other through well-defined channels. All containers are run
by a single operating-system kernel and are thus more lightweight
than virtual machines.</p>
<p class="first-paragraph">I&rsquo;ll make this brief by formulating
a recommendation that I will call: &ldquo;The curse of Docker&rdquo;:</p>
<p>In my experience when you hear &ldquo;docker&rdquo; mentioned in
any positive way in the context of bioinformatics, it is exceedingly
likely that at least one (and often all) of the following statements
below are true:</p>
<ol>
	<li/>
<p class="compact">The person that uttered those words does
	not understand <b>any bioinformatics at all</b>.</p>
	<li/>
<p class="compact">The person that recommends Docker will
	directly or indirectly benefit from you using Docker. They are
	hoping to lock you into a solution that only they can provide, thus
	you become dependent on them.</p>
	<li/>
<p class="compact">The bioinformatics software that needs
	Docker is usually terrible, poorly written, full of bugs and overall
	follows awful software practices. That&rsquo;s why it requires
	Docker! Software written by smart people does not need Docker. By
	using Docker, you are selecting for worse.</p>
	<li/>
<p class="compact">The more time you spend with Docker, the
	worse off you&rsquo;ll be. You dig a hole that will only get deeper
	(see <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Sunk_cost">the
	sunken cost fallacy</a></font>).</p>
</ol>
<p class="first-paragraph"><b>Biostar Theorem of the Day</b></p>
<p class="compact">Any software that needs Docker can be replaced by
better, faster, and simpler software that does not need Docker.</p>
<p>Thus, if at any time, if you hear the word <b>Docker</b>
recommended in any bioinformatics context, run away as fast as
possible. Don&rsquo;t stop, don&rsquo;t look back - or you&rsquo;ll
never get anything done.</p>
<h2 class="western"><a name="outlook-is-hazy-amazon-compute-cloud"></a>
8.5 Outlook is hazy: Amazon Compute Cloud</h2>
<p class="first-paragraph"><img src="0000_html_5ba1a82e4c8398bf.png" name="Image22" align="bottom" hspace="12" width="560" height="84" border="0"/>
</p>
<p>Everything written for the section titled <font color="#4f81bd"><a href="#ch008.xhtml#curse">The
Curse of Docker</a></font> applies the same way.</p>
<p>Using either Docker or the Amazon Compute Cloud will demand
excessive efforts and putting up with massive information overhead -
all gratuitous and nonsensical, all driven by corporate politics,
commercial interests, and bureaucratic agendas.</p>
<p>You&rsquo;ll be neither smarter nor understand anything more about
biology than you did before embarking on understanding these
processes.</p>
<p>The people that teach you Docker and Amazon Cloud <i>before</i>
teaching you bioinformatics cause you more intellectual harm than
good.</p>
<h2 class="western"><a name="the-curse-of-large-scale-data-analysis-platforms"></a>
8.6 The curse of large scale data analysis platforms</h2>
<p class="first-paragraph">When someone promises you everything, the
most likely outcome is that you get nothing.</p>
<p><img src="0000_html_4dffa778a805f261.jpg" name="Image23" align="bottom" hspace="12" width="560" height="105" border="0"/>
</p>
<p>For example, take <font color="#4f81bd"><a href="https://www.cyverse.org/">CyVerse</a></font>,
funded by the National Science Foundation&rsquo;s Directorate for
Biological Sciences note how they claim:</p>
<p class="block-text-western">CyVerse provides life scientists with
powerful computational infrastructure to handle huge datasets and
complex analyses, thus enabling data-driven discovery.</p>
<p class="first-paragraph">Astronautics is quite strong here;
buzzwords abound, amusingly, they even have <font color="#4f81bd"><a href="https://www.cyverse.org/atmosphere">Atmosphere</a></font>
product.</p>
<p>Note how the page discussing the terms of <font color="#4f81bd"><a href="https://www.cyverse.org/use-logo">using
their logo</a></font> is more expansive than the explanation of a
typical data analysis protocol.</p>
<p>Poke around the site; note how they don&rsquo;t teach you how to
to do something yourself. You cannot learn anything meaningful about
science or bioinformatics. It is always about click here, click
there, no lasting knowledge, all about clicking, as if their
interface and whatever half-baked approach they have listed there
would be the perfect data analysis process. Not surprisingly and
quite tellingly: <b>they love Docker</b>, they offer regular training
for Docker! Like minds think alike.</p>
<p><i>At no point in their rambling, are they close to anything that
could be considered a rational scientific thought process. Everyone
visiting the site is now dumber for having clicked the link that led
them there. They deserve no points, and may God have mercy on their
soul.</i></p>
<h2 class="western"><a name="the-new-york-times-test"></a>8.7 The New
York Times Test</h2>
<p class="first-paragraph"><img src="0000_html_62f36eefa3d4f36e.png" name="Image24" align="bottom" hspace="12" width="560" height="84" border="0"/>
</p>
<p>Some of you are looking for projects that you can graduate from.
Some promising plans and research directions do not pan out. Through
tireless work and dedication, I have discovered a fool-proof method
that has high accuracy in predicting if a scientific discovery is
false.</p>
<p class="compact">95% of life-science oriented discoveries mentioned
in a <font color="#4f81bd"><a href="https://www.nytimes.com/">New
York Times</a></font> news release are false.</p>
<p>So there you have a straightforward way to pick winning projects.</p>
<h2 class="western"><a name="designed-by-a-committee-common-workflow-language-cwl"></a>
8.8 Designed by a committee: Common Workflow Language (CWL)</h2>
<p class="first-paragraph"><img src="0000_html_7808ccf132614670.png" name="Image25" align="bottom" hspace="12" width="560" height="84" border="0"/>
</p>
<p>I was giving a presentation on scripting, showed off recipes that
automated the re-analysis of entire publications in less than ten
lines of code. A biologist in the audience asked me the following
question:</p>
<p>-Wouldn&rsquo;t it have been a better option to use the Common
Workflow Language?</p>
<p>I asked him in what way would the CWL be of help. It turns out he
did not know anything at all about the CWL - just that he thought I
would be better off had I used it.</p>
<p>Wow! I thought to myself. Now that&rsquo;s excellent marketing!
They sold a biologist on CWL, with him being unable to articulate a
single benefit of it.</p>
<p>Ok, so let&rsquo;s talk CWL. By their own definition:</p>
<p class="block-text-western">The <font color="#4f81bd"><a href="https://www.commonwl.org/">Common
Workflow Language (CWL)</a></font> is an open standard for describing
analysis workflows and tools in a way that makes them portable and
scalable across a variety of software and hardware environments, from
workstations to cluster, cloud, and high-performance computing (HPC)
environments. CWL is designed to meet the needs of data-intensive
science, such as Bioinformatics, Medical Imaging, Astronomy, High
Energy Physics, and Machine Learning.</p>
<p class="first-paragraph">Let&rsquo;s see how it works in practice.
Imagine that you want to print &ldquo;Hello World&rdquo; to the
screen. Normally, in a shell you could do it like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Hello World&quot;</font></font></p>
<p class="first-paragraph">and it would print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Hello
World</font></font></p>
<p class="first-paragraph">Done!</p>
<p>If you were <i>to solve</i> this same task via the <font color="#4f81bd"><a href="https://www.commonwl.org/user_guide/02-1st-example/index.html">Common
Workflow Language</a></font> you would first need to write a document
called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hello.cwl</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!/usr/bin/env
cwl-runner</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cwlVersion:
v1.0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">class:
CommandLineTool</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">baseCommand:
echo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">inputs:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 message:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   type: string</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   inputBinding:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     position: 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">outputs:
[]</font></font></p>
<p class="first-paragraph">In the next step, you would need to create
another file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo-job.yml</font></font>
that contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">message:
Hello world!</font></font></p>
<p class="first-paragraph">Finally, in the next step you would
<i>execute</i> the code with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cwl-runner
1st-tool.cwl echo-job.yml</font></font></p>
<p class="first-paragraph">that in turn will print the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[job
1st-tool.cwl] /tmp/tmpmM5S_1$ echo \</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   'Hello world!'</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Hello
world!</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[job
1st-tool.cwl] completed success</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">{}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Final
process status is success</font></font></p>
<p class="first-paragraph">Compare the two solutions! I find this
latter both ridiculous and frankly ineffectual.</p>
<p>Of course, some could (and I am sure many will) argue that the
example above is <i>too simple</i> and that in a <i>realistic
scenario</i> CWL will pay other dividends. Alas no.</p>
<p>My experience is the opposite. It is only downhill from here on
out.</p>
<p>On the other hand, do you know what works well with <b>CWL</b>?
<b>Docker</b> and <b>Cloud Computing</b>!</p>
<h2 class="western"><a name="time-to-pay-the-iron-price"></a>8.9 Time
to pay the Iron Price</h2>
<p class="first-paragraph">I could go on, but I think I made enough
enemies as is.</p>
<p>I have been laughing writing this chapter, but I know I am going
to <i>pay the iron price</i> for this little moment of fun.</p>
<p>Scientists have long memories!</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-set-up-your-computer"></a>9 1.
How to set up your computer</h1>
<p class="first-paragraph">Practicing bioinformatics will require the
installation of software tools distributed in various formats:</p>
<ol>
	<li/>
<p class="compact">Binary &ldquo;executable&rdquo; code that
	can be run directly on the target computer.</p>
	<li/>
<p class="compact">&ldquo;Source code&rdquo; that needs to be
	&ldquo;compiled&rdquo; to create a &ldquo;binary&rdquo; program.</p>
	<li/>
<p class="compact">Programs that require the presence of
	another programming language: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">java</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">python</font></font>,
	or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">perl</font></font>.</p>
</ol>
<h2 class="western"><a name="how-do-i-set-up-my-computer"></a>9.1 How
do I set up my computer?</h2>
<p class="first-paragraph">Follow each of the steps below. We may
instruct you to visit other pages but after you do visit those make
sure to come back to this page.</p>
<p>Ensure that you followed each of the steps and the corresponding
subsections.</p>
<ol>
	<li/>
<p class="compact"><b>Prepare your computer.</b></p>
	<li/>
<p class="compact"><b>Set up the terminal</b></p>
	<li/>
<p class="compact"><b>Install conda and activate bioconda.</b></p>
	<li/>
<p class="compact"><b>Install the software.</b></p>
</ol>
<h2 class="western"><a name="is-this-going-to-be-difficult"></a>9.2
Is this going to be difficult?</h2>
<p class="first-paragraph">Well, truth to be told it can be a bit
tedious. But you only need to this once, so bear with it.</p>
<p>The installation process may take a while depending on your
network speed and compute power, it can go from 5 minutes to an hour.
Along the way the process will print copious amounts of information
on your screen. While you don&rsquo;t have to watch the cryptic
sentences go by in real time, the process may, at times stop and ask
you permissions or some questions (yes/no). The choice is to select
or type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">yes</font></font>
unless otherwise instructed by the book.</p>
<p>You may need to enter a password, this is either the Linux user
password or your Mac user password. In each case, you must have
administrative privileges.</p>
<p class="block-text-western"><b>Important</b>: Follow the
instructions in order!</p>
<h2 class="western"><a name="how-do-i-prepare-my-computer"></a>9.3
How do I prepare my computer?</h2>
<p class="first-paragraph">Before installing bioinformatic software,
you have to prepare your computer for the tasks ahead.</p>
<p>We have separate instructions for MacOS, Linux and Windows
computers. Follow only the instructions for your platform.</p>
<p>There will be multiple steps that you have to follow, the
processes may take some time, possibly hours. Make sure to complete
the system setup before venturing forth:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch126.xhtml#setup-macos">Setting
	up a Mac OS computer</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch128.xhtml#setup-windows">Setting
	up a Windows 10 computer</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch127.xhtml#setup-linux">Setting
	up a Linux computer</a></font></p>
</ul>
<h2 class="western"><a name="the-next-step"></a>9.4 The next step</h2>
<p class="first-paragraph">Once you followed the computer setup above
continue on with <font color="#4f81bd"><a href="#ch010.xhtml#terminal-setup">How
to set up the terminal</a></font>.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-initialize-the-terminal"></a>10
2. How to initialize the terminal</h1>
<p class="first-paragraph">The &ldquo;Terminal&rdquo; is the
component of the graphical user interface responsible for entering
and executing text commands. Think of it as a more direct window into
the &ldquo;guts&rdquo; of the computer.</p>
<p>Every analysis will be performed in a terminal. On Windows this
means opening the Ubuntu application, on Mac OS or Linux it means
opening the Terminal application. When logging into a computer
remotely it means the SHH connection to the remote system.</p>
<p>Now depending on the defaults of your system your Terminal may
look like this:</p>
<p><img src="0000_html_3b17bfccb0a44984.png" name="Image26" align="bottom" hspace="12" width="560" height="365" border="0"/>
</p>
<p>Hard to read and comprehend. What we really want is something that
looks more like the following:</p>
<p><img src="0000_html_2f36de05b0f3b844.png" name="Image27" align="bottom" hspace="12" width="560" height="367" border="0"/>
</p>
<p>It is exceedingly important that the fonts and presentation of the
terminal are legible, readable and easy to see. Many times small
differences in the way commands are invoked have huge effects on the
results, thus we want to ensure that the terminal is as readable as
possible.</p>
<p>On a Mac you go to the <i>Terminal-&gt;Preferences</i> and set the
<i>Novel</i> color scheme as default, then increase the font sizes.
Now open a new terminal. It should look nice and clean.</p>
<p>On Windows you need to right click the panel and set colors for
foreground and background as well as font sizes. The color scheme I
use has the background hexcolor of `<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#DEDBC6</font></font>
and text color of black <font face="Consolas, serif"><font size="2" style="font-size: 11pt">#000000</font></font>.</p>
<h2 class="western"><a name="pin-your-terminal-to-the-taskbar-or-dock"></a>
10.1 Pin your terminal to the taskbar or dock</h2>
<p class="first-paragraph">You will make use of the terminal a lot,
we recommend that you pin it to the desktop, taskbar or dock. On Mac
OS make sure to pin the <i>Rosetta Terminal</i> to the dock.</p>
<h2 class="western"><a name="what-does-the-prompt-show"></a>10.2 What
does the prompt show?</h2>
<p class="first-paragraph">The bash prompt is that the text/symbol
that appears on each line. Once you follow the instruction on the
next page, our installation script will customize the prompt on your
terminal to give you information on the context you are working in.</p>
<p><img src="0000_html_6866e55205fe4a18.png" name="Image28" align="bottom" hspace="12" width="560" height="158" border="0"/>
</p>
<p>Initially, when you only work on a single system and in a single
directory, all that information seems unnecessary and perhaps
overbearing.</p>
<p>Once you gain experience, perhaps you use multiple remote
computers, the information in the prompt becomes extraordinarily
useful. That information will be your first line of defense in
avoiding some of the most dangerous mistakes you can make: <i>running
a command on the wrong computer or in the wrong directory</i>. We set
you with the right prompt to train your eyes and to help you avoid
mistakes.</p>
<p>While the above customization will happen once you run the
installation on the next page, we mention it here as it belongs to
the topic of customization.</p>
<h2 class="western"><a name="what-is-the-next-step"></a>10.3 What is
the next step?</h2>
<p class="first-paragraph">Once your system is set up properly
continue on with the section <font color="#4f81bd"><a href="#ch011.xhtml#software-installation">How
to install software</a></font>.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-install-software"></a>11 3. How
to install software</h1>
<p class="first-paragraph">As of Dec 1st, 2021, we have developed an
automated approach to bootstrap your system with minimal input.
Follow a three-step process:</p>
<ol>
	<li/>
<p class="compact">Run the script and wait for it to finish.</p>
	<li/>
<p class="compact">Close the terminal and open a new one.</p>
	<li/>
<p class="compact">Run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">doctor.py</font></font>
	to verify that the installation was successful.</p>
</ol>
<p class="block-text-western">Since the script is a new development,
it may have &ldquo;issues&rdquo;. If you encounter any problems,
please report them to us on the <font color="#4f81bd"><a href="https://github.com/biostars/biostar-handbook-issues">Biostar
Handbook Issue Tracker</a></font></p>
<p class="first-paragraph">The commands should be attempted after you
have already completed the instruction on preparing your system as
indicated in <font color="#4f81bd"><a href="#ch009.xhtml#computer-setup">How
to set up your computer</a></font>.</p>
<h2 class="western"><a name="quick-install"></a>11.1 Quick install</h2>
<p class="first-paragraph">Copy the following command to your
terminal:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://data.biostarhandbook.com/install.sh | bash</font></font></p>
<p class="first-paragraph">The installation script will download and
install all the software you need while producing copious amounts of
output to the screen. Depending on various factors, such as internet
speed, hard drive performance, etc., the process may take from 3
minutes (when we ran it on the latest MacBook Air) to less than 20
minutes (about six years old Windows-based Linux system).</p>
<p>Wait for the process to complete. Completion will be indicated via
finishing with a message to the screen:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Biostar Handbook installation complete!</font></font></p>
<p class="first-paragraph">Your computer setup should be complete.</p>
<h2 class="western"><a name="apply-settings-to-terminal"></a>11.2
Apply settings to terminal</h2>
<p class="first-paragraph">Once the installation process completes,
the changes need to be applied to your terminal.</p>
<p>The easiest way to load the new settings is to close then re-open
the terminal.</p>
<p class="compact">Once re-opened, the terminal should look nicer
with a better-looking prompt.</p>
<p>Alternatively, you may also apply the changes in the current
terminal by running:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">source
~/.bash_profile</font></font></p>
<h2 class="western"><a name="validate-the-installation"></a>11.3
Validate the installation</h2>
<p class="first-paragraph">After re-opening the terminal, you can
verify that the installation was successful by first activating the
bioinformatics environment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font></p>
<p class="first-paragraph">then, after entering the above, type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">doctor.py</font></font></p>
<p class="first-paragraph">The command should produce output similar
to:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Doctor! Doctor! Give me the news.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Checking 15 symptoms...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
bwa ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
datamash ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
fastqc -h ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
hisat2 ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seqret -h ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
featureCounts ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
efetch ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
esearch ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
samtools ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
fastq-dump ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
bowtie2 ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
bcftools ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seqtk ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seqkit ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
bio ... OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
You are doing well!</font></font></p>
<p class="first-paragraph">If you get the above, you are good to go.</p>
<h2 class="western"><a name="run-a-realistic-analysis"></a>11.4 Run a
realistic analysis</h2>
<p class="first-paragraph">Your system is now ready and itching to do
something useful! Are you eager to put your system to the test? Of
course, you are. Who wouldn&rsquo;t be?</p>
<p>Let&rsquo;s run a realistic analysis. <i>This is not a toy</i>!
The pipeline below produces results on par with findings in published
research. First, let&rsquo;s make a new directory for the analysis:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a new directory for the analysis.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p work</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Switch to that directory</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
work</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Activate your environment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font></p>
<p class="first-paragraph">Obtain our <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Makefile</font></font>
that runs a variant calling analysis, reproducing findings regarding
the variations occurring in the Ebola virus genome between 1972 and
2014. The data and paper are discussed in more detail in the chapter
titled <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font>
originally published in Nature (2014):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the makefile</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-s http://data.biostarhandbook.com/make/snpcall.mk &gt; Makefile</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run the makefile</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">make</font></font></p>
<p class="first-paragraph">it will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">USAGE:</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 make vcf         # generates a VCF file</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 make clean       # removes test data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 make realclean   # removes all intermediate files</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 make data        # downloads the full dataset</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DEFAULTS:</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  ACC=AF086833  SRR=SRR1553425  CPU=4  LIMIT=100000</font></font></p>
<p class="first-paragraph">Run the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">vcf</font></font>
step (if your computer has fewer than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>
cores, say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>,
then add <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CPU=2</font></font>
to the command):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">make
vcf</font></font></p>
<p class="first-paragraph">the system will churn for a while; it will
look a bit like this (click to start animation):</p>
<p><br/>
<br/>

</p>
<p>If you knew what the files were, you could immediately make
several observations regarding the nature, frequency, and effect of
mutations on the genome. Well, that&rsquo;s what you&rsquo;re here to
learn, so don&rsquo;t worry, we&rsquo;ll get there!</p>
<p>Visualizing the results generated above would look like this:</p>
<p><img src="0000_html_545c9f92b24b9429.png" name="Image29" align="bottom" hspace="12" width="560" height="332" border="0"/>

Whoo-hoo! In less than a few minutes, a wealth of information was
generated. What previously was an empty directory is now filled with
goodies. We got:</p>
<ul>
	<li/>
<p class="compact">The viral sequence files as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>.</p>
	<li/>
<p class="compact">The viral annotation files as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF</font></font>.</p>
	<li/>
<p class="compact">Sequencing data in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
	format obtained from the Short Read Archive (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA</font></font>)</p>
	<li/>
<p class="compact">Quality controlled sequenced data with
	trimmed and quality filtered reads using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic</font></font>.</p>
	<li/>
<p class="compact">An alignment <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
	file generated with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
	short read aligner, filtered, sorted, and indexed with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>.</p>
	<li/>
<p class="compact">A VCF file generated with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
	that lists the variants of the 2014 data relative to 1972 strain.</p>
	<li/>
<p class="compact">An annotated VCF file generated with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff</font></font>
	that describes the effect each variant has on the protein-coding
	regions.</p>
</ul>
<p class="first-paragraph">And it did all that with no external
intervention. It may be rerun at any time, and it may be rerun on any
other genome or data. That is the power automation at the command
line! You can open the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Makefile</font></font>
in a text editor to see, adjust, tweak, or change the commands that
run the pipeline.</p>
<h2 class="western"><a name="lets-check-the-mutations-for-omicron"></a>
11.5 Let&rsquo;s check the mutations for Omicron</h2>
<p class="first-paragraph">Here is how easy it is to adapt this same
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Makefile</font></font>
based pipeline to other data. Suppose you want to check how many
variants are there between the original SAR-COV-2 and newly obtained
sequencing data suspected to represent the Omicron variant. Run the
same <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Makefile</font></font>
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">make
vcf ACC=NC_045512  SRR=SRR17054510  CPU=4</font></font></p>
<p class="first-paragraph">Boom! In a minute, it is all done. The
primary challenge was identifying the correct accession numbers:
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR17054510</font></font>.
The analysis itself will take around just minutes to run!</p>
<p>That&rsquo;s what this book is all about! How to generate these
results, and, of course, just as importantly, what do they all mean?</p>
<p>The Biostar Handbook will take you one step at a time in the right
direction. Stick with us, and you&rsquo;ll be coding up better
pipelines in no time.</p>
<h2 class="western"><a name="what-are-environments"></a>11.6 What are
environments?</h2>
<p class="first-paragraph">Modern software packages are
interdependent. Each program may rely on the presence of one or more
other software. In turn, each required software may also need another
program to be present - and so on - down the rabbit hole we go.</p>
<p>Then, things can get even more complicated. Software A may only
inter-operate appropriately with a specific version of software B.
For that reason, there may be conflicts between the different
requirements. These conflicts may be solved by creating isolated
installations where only the compatible software is grouped.</p>
<p>The isolated installations are called <i>environments</i>. To set
up and manage our environment, we have set up your system to use the
<font color="#4f81bd"><a href="#ch011.xhtml#setup-conda">conda</a></font>
environment manager.</p>
<p><font color="#4f81bd"><a href="https://conda.io/miniconda.html">Conda</a></font>
is an open-source package management system and environment
management system for installing multiple versions of software
packages and their dependencies and switching quickly between them.
Conda works on Linux, macOS, and Windows, and it can package and
distribute just about any software.</p>
<h2 class="western"><a name="what-is-bioconda"></a>11.7 What is
bioconda?</h2>
<p class="first-paragraph">We will also make use of a repository
called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioconda</font></font>
customized to install bioinformatics packages. Our automated script
has already installed <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioconda</font></font>
on your system. The only thing you need to remember is to activate
the bioinformatics environment in each new terminal you open.</p>
<p>Every time you open a new terminal, you will need to enter the
following command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font></p>
<p class="first-paragraph">That&rsquo;s it.</p>
<h2 class="western"><a name="no-seriously.-activate-the-environment."></a>
11.8 No, seriously. Activate the environment.</h2>
<p class="first-paragraph">One of the most common errors that
students make is forgetting to activate the environment. If you get
the error:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-bash:
bwa: command not found</font></font></p>
<p class="first-paragraph">It most likely means you forgot to
activate the environment in the current terminal. Do so with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font></p>
<p class="first-paragraph">Take solace in knowing that we all forgot
to activate the environment at some point in our work. You will
probably too.</p>
<p><img src="0000_html_52ba03004cd3b3de.jpg" name="Image30" align="bottom" hspace="12" width="560" height="183" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-next-step-1"></a>11.9 What
is the next step?</h2>
<p class="first-paragraph">Once your system is set up properly,
continue on with the section <font color="#4f81bd"><a href="#ch012.xhtml#manage-environments">How
to manage environments</a></font>.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-manage-environments"></a>12 4.
How to manage environments</h1>
<p class="first-paragraph">Installing multiple software packages into
the same environment can lead to trouble. For a while, you may get
by, but at some point, conflicts will arise. Annoyingly often, when
installing a new tool and existing and working software gets replaced
with an older version. When you run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install</font></font>, it waits for confirmation to proceed, and it
tells you about the changes that it will perform. Alas, occasionally,
dozens of packages get installed, and it is hard to keep track of all
modifications.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo</font></font>
environment we created contains most tools used in the book, but
occasionally you may need to install missing software even for some
of our examples.</p>
<p>The recommended approach is to be minimalistic and to install only
the software you need. You may create several environments, each with
its own software packages.</p>
<p>For example, if you know that a project needs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic</font></font>,
create a dedicated environment for it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a new environment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create -y -n heatshock</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Activate the new environment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate heatshock</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install the software</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install fastqc bwa bcftools trimmomatic</font></font></p>
<p class="first-paragraph">You may list the installation steps in a
separate script or a Makefile.</p>
<h2 class="western"><a name="how-can-i-see-my-environments"></a>12.1
How can I see my environments?</h2>
<p class="first-paragraph">The following command lists all
environments:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
info --envs</font></font></p>
<p class="first-paragraph">on my system currently shows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">base
                    /home/ialbert/miniconda3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo
                 /home/ialbert/miniconda3/envs/bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">heatshock
            *  /home/ialbert/miniconda3/envs/heatshock</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">test
                    /home/ialbert/miniconda3/envs/test</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>
indicates the current environment.</p>
<h2 class="western"><a name="do-not-run-sudo"></a>12.2 Do not run
sudo!</h2>
<p class="first-paragraph">On Linux, at some point, you may type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedops</font></font></p>
<p class="first-paragraph">and you may get the following &ldquo;helpful&rdquo;
response:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Command
'bedops' not found, but can be installed with:</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
apt install bedops</font></font></p>
<p class="first-paragraph"><b>Do not follow the recommendation above!</b></p>
<p>It bypasses the bioinformatics environment and runs the risk of
irrevocably breaking your system. Never use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo</font></font>
command to install bioinformatics software.</p>
<p>Instead, whenever you want to install software that may not exist
on your system, use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install bedops</font></font></p>
<p class="first-paragraph">Now view the effect that the above command
has on your system. If you see that many other packages are affected
(especially downgraded), then the best choice would be to create a
new environment and install the software there.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create --name bedops</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bedops</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install bedops</font></font></p>
<p class="first-paragraph">You may create as many environments as
necessary. The more software an environment has, the more likely it
is that you run into conflicts.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-choose-a-text-editor"></a>13 5.
How to choose a text editor</h1>
<p class="first-paragraph">Many tasks will ask you to view, edit or
generate text files.</p>
<p>To create, investigate and modify these files you will need a
proper <b>text</b> editor. And no, Microsoft Word is not a text
editor. Using Word to edit text will eventually cause (devious)
errors. It&rsquo;s best if you learn to use a proper text editor from
the start.</p>
<p>Keep in mind that a substantial number of bioinformatics file
formats use the so-called &ldquo;tab&rdquo; character to separate
columns. Thus the need to use an editor that supports working with
tab-delimited data.</p>
<h2 class="western"><a name="what-features-should-my-text-editor-have"></a>
13.1 What features should my text editor have?</h2>
<p class="first-paragraph">Essential features that your editor needs
to have:</p>
<ul>
	<li/>
<p class="compact">The editor needs to have the ability to
	show so-called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tab</font></font>
	characters versus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">space</font></font>
	characters. Both are invisible, one-byte characters, that by eye
	will look like one or more spaces. Data that, for some reason, end
	up using a tab instead of space (or vice versa) will cause
	incredibly confusing errors. You will have a hard time tracking down
	these errors unless you can visualize the two normally invisible
	characters.</p>
	<li/>
<p class="compact">The editor should be able to display line
	numbers. Frequently this feature lets you identify the line where
	the error occurs.</p>
	<li/>
<p class="compact">The editor has to be able to switch line
	endings. Line endings have variants: Unix, Windows. As a rule, all
	your files should always have Unix line endings, and you should
	easily be able to convert line endings to Unix format.</p>
</ul>
<h2 class="western"><a name="which-text-editor-to-choose"></a>13.2
Which text editor to choose?</h2>
<p class="first-paragraph">The choice of an editor is somewhat
personal preference. You may already have an editor that you like.
Text editors that recommend come in two &ldquo;flavors&rdquo;:</p>
<ul>
	<li/>
<p class="compact">Programming Editors: <font color="#4f81bd"><a href="https://code.visualstudio.com/">Visual
	Studio Code</a></font>, <font color="#4f81bd">PyCharm</font> that
	are a little more complicated but support a wide variety of
	productivity increasing features.</p>
	<li/>
<p class="compact">Generic Text Editors: Komodo Edit, Sublime
	Text 2, <font color="#4f81bd"><a href="https://notepad-plus-plus.org/">NotePad++</a></font>
	that smaller downloads, sometimes run only on certain platforms and
	require fewer resources to run.</p>
</ul>
<p class="first-paragraph">Either type of editor will work for
bioinformatics.</p>
<p>We recommend <font color="#4f81bd"><a href="https://code.visualstudio.com/">Visual
Studio Code (VS Code)</a></font> as a great all around choice for
every operating system. It runs on every platform and provides a wide
array of functionality that can support you as you develop your
skills. Among the many excellent features, Visual Studio Code
integrates well with Linux on Windows and it can even launch <font color="#4f81bd"><a href="https://code.visualstudio.com/docs/remote/wsl">Linux
commands</a></font>.</p>
<p>Here is the VS Code interface with a light theme:</p>
<p><img src="0000_html_756989dae0d85835.png" name="Image31" align="bottom" hspace="12" width="560" height="356" border="0"/>

Visual Studio Code tips:</p>
<ul>
	<li/>
<p class="compact">Launch the &ldquo;Command Palette&rdquo;
	with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Ctrl+Shift+P</font></font>
	(On Windows) or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Command+Shift+P</font></font>
	then type to search for the command you need</p>
	<li/>
<p class="compact">Information on <font color="#4f81bd"><a href="https://code.visualstudio.com/docs/getstarted/keybindings">how
	to change key bindings</a></font></p>
</ul>
<p class="first-paragraph">If you&rsquo;d prefer a simpler
alternative the following options are also excellent:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://notepad-plus-plus.org/">NotePad++</a></font>
	- only works on Windows but it is one of best editors ever created.
	Use Notepad++ for Windows.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.activestate.com/komodo-ide/downloads/edit">Komodo
	Edit</a></font> - works on all operating systems (macOS, Widows and
	Linux)</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.sublimetext.com/2">Sublime
	Text 2</a></font> - all operating systems.</p>
</ul>
<p class="first-paragraph">I use <font color="#4f81bd">PyCharm</font>
for text editing and programming. I chose this editor before Visual
Studio Code was developed, if I were starting today I&rsquo;d most
likely choose VS Code.</p>
<h2 class="western"><a name="viewing-whitespace"></a>13.3 Viewing
whitespace</h2>
<p class="first-paragraph">Occasionally mysterious errors are caused
by incorrect column separators in the file. Both the space and tab
characters can look the same, but in bioinformatics the programs that
read the file usually need all columns separated by tabs. When
copy-pasting data from a webpage the tabs may be turned into spaces.</p>
<p>Here is what it looks like. The words on both rows appear to be
separated by the same characters when in reality the two lines
differ.</p>
<p><img src="0000_html_a35e4d385992e709.png" name="Image32" align="bottom" hspace="12" width="560" height="57" border="0"/>
</p>
<p>On the first line, three spaces separate the words. On the second
line, a single tab character separates the two words.</p>
<p>Now look at the same file file with &ldquo;View Whitespace&rdquo;
option turned on. The so-called &ldquo;white-space&rdquo; content
between the words is now made visible.</p>
<p><img src="0000_html_11a70c4c3f934d1f.png" name="Image33" align="bottom" hspace="12" width="560" height="57" border="0"/>
</p>
<p>VS Code will show the whitespace by default.</p>
<h2 class="western"><a name="what-gets-inserted-when-you-press-tab"></a>
13.4 What gets inserted when you press TAB?</h2>
<p class="first-paragraph">Depending on how your text editor is set
up, it may, by default, insert multiple spaces even when you press
the TAB key. You have to explicitly configure VS Code and Komodo Edit
to enter the TAB character when you press the TAB key. This behavior
is immensely confusing (and annoying) at first. Hence make sure to
set up your text editor to insert TAB when you press TAB. In some
cases this is the desired behavior (for example when programming in
Python). When entering data or editing a data file one mistaken TAB
can mess up your entire file.</p>
<p>When using <b>Visual Studio Code</b> you can launch the so called
&ldquo;Command Palette&rdquo; with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Ctrl+Shift+P</font></font>
(On Windows) or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Command+Shift+P</font></font>
(on the Mac) then search for the command that changes indentation:</p>
<p><img src="0000_html_d0903a6b66890234.png" name="Image34" align="bottom" hspace="12" width="560" height="113" border="0"/>
</p>
<p>In Komodo Edit you need to navigate to preferences and tick a
checkbox to make it do what it is supposed to.</p>
<p><img src="0000_html_19dae6ce192e1b4c.png" name="Image35" align="bottom" hspace="12" width="560" height="560" border="0"/>
</p>
<h2 class="western"><a name="watch-the-line-endings-on-windows"></a>13.5
Watch the line endings on Windows!</h2>
<p class="first-paragraph">Having Windows line endings instead of
Unix line endings will lead to various mysterious errors. Press play
below to see how the error may manifest itself:</p>
<p><br/>
<br/>

</p>
<p>Currently there are two kinds of line endings in use: UNIX (LF)
style and Windows style (CRLF) line termination characters. If you
are running Linux from Windows and use a Windows text editor from
Windows you have to ensure that the files that you create will use
UNIX line endings. This is typically a setting on the file, and is
only necessary when you start a new file.</p>
<p>When using <b>Visual Studio Code</b> you can change on the line
ending by clicking on line feed indicator:</p>
<p><img src="0000_html_85f8f2c3c716b51a.png" name="Image36" align="bottom" hspace="12" width="560" height="81" border="0"/>
</p>
<p>Below we link to a short explanation:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://blog.codinghorror.com/the-great-newline-schism/">The
	Great Newline Schism</a></font></p>
</ul>
<p class="first-paragraph">Google the concept if you want to learn
more.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-fix-problems"></a>14 6. How to
fix problems</h1>
<p class="first-paragraph">Occasionally your installation may break
for various reasons. Perhaps you have installed different tools,
explored other software that changed your setup. Sometimes you may
find that even after uninstalling tools your system may exhibit
various errors that were not present before.</p>
<h2 class="western"><a name="solution-1.-redo-it-again"></a>14.1
Solution 1. Redo it again</h2>
<p class="first-paragraph">It is typically quite challenging to fix
these so in our experience the best solution is to start anew, make a
new environment and reinstall everything. For us, this usually solves
the problem. The steps to do a full reinstall by removing the
previous environment and resetting everything are:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
deactivate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
remove --name bioinfo --all</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://data.biostarhandbook.com/install.sh | bash</font></font></p>
<p class="first-paragraph">Alternatively see if you can create a new
environment with a different name and install into that.</p>
<h2 class="western"><a name="how-do-i-report-installation-problems"></a>
14.2 How do I report installation problems?</h2>
<p class="first-paragraph">While we are doing our best to keep the
book up-to-date certain problems may go unnoticed for short periods
of time.</p>
<p>Twice a year the installation instructions are tested in a
classroom setting where wide variety of computing platforms are in
use: Windows, Mac and Linux. So we know the instructions are tested
twice a year in a wide range of systems, but even so problems may
still occur from time to time.</p>
<p>The book has an issue tracker that you may consult to see whether
other people had problems similar to yours. This is also the place
where you can report new issues and problems that appear to have not
been discussed yet:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/biostars/biostar-handbook-issues">Biostar
	Handbook Issue Tracker</a></font></p>
</ul>
<p class="first-paragraph">The tracker above may also be used to
report errors or inconsistencies in the content of the book.</p>
<h2 class="western"><a name="how-do-i-use-conda-in-general"></a>14.3
How do I use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
in general?</h2>
<p class="first-paragraph">Once <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
is installed as above you can create an environments, for example we
created the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo</font></font>
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create -y --name foo python=3.8</font></font></p>
<p class="first-paragraph">To activate the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">foo</font></font>
environment (this needs to be done in each new window you open) type
:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate foo</font></font></p>
<p class="first-paragraph">You may create multiple environments on
your system, sometimes creating new enviroments is necessary to run
software with different requirements and dependencies.</p>
<p>Note: in prior versions of conda the command to activate
environment was <font face="Consolas, serif"><font size="2" style="font-size: 11pt">source
activate bioinfo</font></font>, if you see this advice remember to
replace is with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate</font></font></p>
<p>You can list all existing environments with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
info --envs</font></font></p>
<p class="first-paragraph">You can learn more about managing
environments on the <font color="#4f81bd"><a href="https://conda.io/docs/using/envs.html">conda
webpage</a></font></p>
<h2 class="western"><a name="how-do-i-install-a-new-tool"></a>14.4
How do I install a new tool?</h2>
<p class="first-paragraph">While our instructions install all that
tools that are used in the book, you may occasionally find yourself
needing a program that we did not include. In that case, your first
approach would be to see if <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
already covers the tool you want to install. Most of the time you may
be able to install into the environment <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install toolname</font></font></p>
<p class="first-paragraph">This command will collect the requirements
and it will ask you if you want to proceed. Read the output carefully
and make a note of any warning it indicates. It is possible that some
tools cannot operate together. In those cases, you can create a new
environment just for the new tool.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create --name mynewthing -y</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate mynewthing</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install toolname</font></font></p>
<p class="first-paragraph">To run the new tool you&rsquo;d have to be
in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mynewthing</font></font>
environment. To avoid having to switch back and forth, you may open a
terminal and activate the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo</font></font>
environment then open another terminal and activate the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mynewthing</font></font>
environment there.</p>
<p>Now in some cases, software needs to be installed from the
&ldquo;source&rdquo; code. Source code installation typically
involves downloading the source code, unpacking the software and
following the installation instructions. Sometimes this means running
either <font face="Consolas, serif"><font size="2" style="font-size: 11pt">make</font></font>
or some other command. We do list the source code installation for a
few tools in the appendix.</p>
<h2 class="western"><a name="how-do-i-upgrade-a-tool"></a>14.5 How do
I upgrade a tool?</h2>
<p class="first-paragraph">Tools may have updated versions released.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Upgrade a package.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
upgrade toolname</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-unix"></a>15
Introduction to Unix</h1>
<p class="first-paragraph">The terms <b>&ldquo;Unix&rdquo;</b> and
<b>&ldquo;Command Line&rdquo;</b> are used interchangeably both in
this Handbook and in the real world. Technically speaking, Unix is a
class of operating system that originally used the command line as
the primary mode of interaction with the user. As computer
technologies have progressed, the two terms have come to refer to the
same thing.</p>
<h2 class="western"><a name="what-is-the-command-line"></a>15.1 What
is the command line?</h2>
<p class="first-paragraph">People interact with computers via
so-called <b>user interfaces</b>. Clicking an icon, scrolling with
the mouse, and entering text into a window are all user interfaces by
which we instruct the computer to perform a specific task.</p>
<p>When it comes to data analysis, one of the most &ldquo;ancient&rdquo;
of interfaces&ndash;&ndash;the text interface&ndash;&ndash;is still
the most powerful way of passing instructions to the computer.</p>
<p>In a nutshell, instead of clicking and pressing buttons you will
be entering<br/>
<i>words</i> and <i>symbols</i> that will act as
direct commands and will instruct the computer to perform various
processes.</p>
<h2 class="western"><a name="what-does-the-command-line-look-like"></a>
15.2 What does the command line look like?</h2>
<p class="first-paragraph">For the uninitiated command line
instructions may look like a magical incantation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
sraids.txt | parallel fastq-dump -O sra --split-files {} </font></font>
</p>
<p class="first-paragraph">In a way it is modern magic - with just a
few words we can unleash complex computations that would be
impossible to perform in any other way. With the command line, we can
slay the dragons and still be home by six.</p>
<p>As for the above &ndash; and don&rsquo;t worry if you can&rsquo;t
follow this yet &ndash; it is a way to download all the published
sequencing data that corresponds to the ids stored in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sraids.txt</font></font>
file. The command as listed above invokes three other commands: a
funny one named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>,
a more logical sounding one <font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font>
and a mysterious <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>.</p>
<p>This series of commands will read the content of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sraids.txt</font></font>
and will feed that line by line into the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font>
program that in turn will launch as many <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
programs as there are CPU cores on the machine. This parallel
execution typically speeds up processing a great deal as most
computers can execute more than one process at a time.</p>
<h2 class="western"><a name="what-are-the-advantages-of-the-command-line"></a>
15.3 What are the advantages of the command line?</h2>
<p class="first-paragraph">The command line presents a word by word
representation of a series of actions that are explicit, shareable,
repeatable and automatable. The ability to express actions with words
allows us to organize these actions in the most appropriate way to
solve every problem.</p>
<h2 class="western"><a name="what-are-the-disadvantages-of-the-command-line"></a>
15.4 What are the disadvantages of the command line?</h2>
<p class="first-paragraph">Running command line tools requires
additional training and a more indepth understanding of how computers
operate. Do note that this knowledge is quite valuable beyond the
domain of bioinformatics.</p>
<h2 class="western"><a name="is-knowing-the-command-line-necessary"></a>
15.5 Is knowing the command line necessary?</h2>
<p class="first-paragraph">Yes. While it is entirely possible to
perform many bioinformatics analyses without running a command line
tool most studies will benefit immensely from using a Unix like
environment. Most importantly understanding Unix develops a way of
thinking that is well suited to solving bioinformatics problems.</p>
<h2 class="western"><a name="is-the-command-line-hard-to-learn"></a>15.6
Is the command line hard to learn?</h2>
<p class="first-paragraph">That is not the right way to think about
it.</p>
<p>Understand that learning the command line (and programming in
general) is not just a matter of learning the concepts. Instead, it
is primarily about learning to think a certain way - to decompose a
problem into very simple steps each of which can be easily solved.
Some people have an affinity to this &ndash; they will pick it up
faster, while others will need more practice. In our observation,
prior preparation/experience does not affect the speed at which
people can learn the command line.</p>
<p>At the same time, you have to realize that you can only succeed at
changing how you think when you perform these actions. Besides it
also means that you can&rsquo;t just learn it all in a day or even a
week. So keep at it for a little while longer building your skills.
In the end, it is only a matter of time and practice.</p>
<p>Try to actively make use of command line tools by integrating them
into your daily work. Making use of command line tools is usually
quite easy to do since Unix tools are useful for just about any type
of data analysis.</p>
<h2 class="western"><a name="do-other-people-also-make-many-mistakes"></a>
15.7 Do other people also make many mistakes?</h2>
<p class="first-paragraph">When you start out using the command line
you will continuously make errors. That can be very frustrating. You
could ask yourself, &ldquo;Do other people make lots of mistakes
too?&rdquo;</p>
<p>Yes, we all do. We correct these errors faster.</p>
<p>Your skill should be measured not just in making fewer mistakes
but primarily in how quickly you fix the errors that you do make. So
get on with making mistakes, that&rsquo;s how you know you are on the
right path.</p>
<h2 class="western"><a name="how-much-unix-do-i-need-to-know-to-be-able-to-progress"></a>
15.8 How much Unix do I need to know to be able to progress?</h2>
<p class="first-paragraph">A beginner level Unix expertise is
sufficient to get started and to perform the majority of analyses
that we cover in the book.</p>
<p>You should have a firm grasp of the following:</p>
<ol>
	<li/>
<p class="compact">Directory navigation: what the directory
	tree is, how to navigate and move around with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font></p>
	<li/>
<p class="compact">Absolute and relative paths: how to access
	files located in directories</p>
	<li/>
<p class="compact">What simple Unix commands do: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font></p>
	<li/>
<p class="compact">Getting help: how to find out more on what
	a unix command does</p>
	<li/>
<p class="compact">What are &ldquo;flags&rdquo;: how to
	customize typical unix programs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
	vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
	-l</font></font></p>
	<li/>
<p class="compact">Shell redirection: what is the standard
	input and output, how to &ldquo;pipe&rdquo; or redirect the output
	of one program into the input of the other</p>
</ol>
<h2 class="western"><a name="how-do-i-access-the-command-line"></a>15.9
How do I access the command line?</h2>
<p class="first-paragraph">When using graphical user interfaces the
command line shell is accessed via an application called &ldquo;Terminal&rdquo;.
When the &ldquo;Terminal&rdquo; runs it launches a so-called &ldquo;shell&rdquo;
that by default opens into a folder that is called your &ldquo;home&rdquo;
directory.</p>
<h2 class="western"><a name="what-is-a-shell"></a>15.10 What is a
shell?</h2>
<p class="first-paragraph">The shell is an interface that interprets
and executes the commands that entered into it via words. There are
different types of shells, the most commonly used shell is called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash</font></font>.
Below we run the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>
command in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
'Hello World!'</font></font></p>
<p class="first-paragraph">and we see it executed in a Mac OSX
Terminal below. The default terminal that opens up for you will
probably look a bit different. There are options to change the fonts,
background colors etc. Besides, our so-called command prompt (the
line where you enter the commands), is probably simpler than yours.
See the <font color="#4f81bd"><a href="#ch129.xhtml#bash-profile">Setup
profile</a></font> for details on how to change that.</p>
<p><img src="0000_html_7e128acc6bc7a2be.png" name="Image37" align="bottom" hspace="12" width="560" height="247" border="0"/>
</p>
<p>It is essential to recognize that the shell is not just a way to
launch commands but a <b>full programming language</b> that provides
many powerful constructs to streamline its use. Alas, initially these
powers tend to get in the way and cause some frustrations. Suppose
that instead of</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
'Hello World!'</font></font></p>
<p class="first-paragraph">you&rsquo;ve enclosed the text with double
quotes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Hello World!&quot;</font></font></p>
<p class="first-paragraph">now instead of echoing the output, it will
produce:</p>
<p><img src="0000_html_8f575c3122e32182.png" name="Image38" align="bottom" hspace="12" width="560" height="247" border="0"/>
</p>
<p>Maddeningly enough the error occurs only if the string contains
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>
character as the last element. For example, this works just fine.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Hello World&quot;</font></font></p>
<p class="first-paragraph">Understandably getting mysterious errors
when writing seemingly simple text can be quite confusing. The error
occurred because in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash</font></font>
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>
character allows you to rerun a previous command. For example</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">!e</font></font></p>
<p class="first-paragraph">Will rerun the last command that starts
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">e</font></font>.
In this case the first <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
'Hello World!'</font></font>. The feature was added as help to avoid
retyping long commands. Tip: try <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!!</font></font>
see what happens.</p>
<p>Then, as it turns out in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash</font></font>
text enclosed within double quotes is interpreted differently than
text enclosed in single quotes. Single quoted content is passed down
exactly as written, whereas in double-quoted text some letters that
have may special meaning like: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font>
(and others) will be interpreted as bash commands. Initially this
gets in the way, as you don&rsquo;t know which letter is &ldquo;special&rdquo;.
Later this feature allows you to build more powerful commands.</p>
<p><b>Note:</b> As a rule, in bash type single quotes if you want the
text the stay the same, and type double quotes when the text needs to
be modified later on.</p>
<p class="block-text-western">Gotcha&rsquo;s like these are numerous.
Everyone gets tripped up at some time. Google is your friend. Bash
has been around for a long time (decades) and did not change much.
Every problem you run into has most likely been experienced by others
as well, and will have solutions written out in great detail.</p>
<h2 class="western"><a name="wait-there-is-more"></a>15.11 Wait there
is more</h2>
<p class="first-paragraph">As it turns out sometimes</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Hello World!&quot;</font></font></p>
<p class="first-paragraph">will print</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Hello
World!</font></font></p>
<p class="first-paragraph">as expected. If you were to put the
command into a file (script):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Check out the mix of single and double quotes here.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
'echo &quot;Hello World!&quot;' &gt; script.sh</font></font></p>
<p class="first-paragraph">now if you were to run it like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
script.sh</font></font></p>
<p class="first-paragraph">It works without raising an error. But
why?</p>
<p>You see, the &ldquo;interactive&rdquo; bash terminal tries to
assist you in minimize typing. A feature called &ldquo;history
substitution&rdquo; is turned on when you type into bash, but it is
turned off when you run a script via bash. Doing so allows you to use
the event shortcuts when you are typing, but you don&rsquo;t need to
be use these in a script.</p>
<p>You can turn off the history substitution from the command line
yourself. And on some systems this is done automatically in the bash
configuration file. Typing:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
+H</font></font></p>
<p class="first-paragraph">will turn off history substitution. It is
a bit odd that we need to use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font>
sign to turn a feature off, and a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>
to turn the feature back on. But we&rsquo;ll live with that.</p>
<h2 class="western"><a name="what-is-the-best-way-to-learn-unix"></a>15.12
What is the best way to learn Unix?</h2>
<p class="first-paragraph">We offer a number of tutorials right here
in this book.</p>
<p>In addition there are excellent online tutorials, some free others
at cost, each with a different take and philosophy on how to teach
the concepts. We recommend that you try out a few and stick with the
one that seems to work the best for you.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.codecademy.com/learn/learn-the-command-line">CodeAcademy:
	Learn the command line</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://swcarpentry.github.io/shell-novice/">Software
	Carpentry: The Unix shell</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://korflab.ucdavis.edu/bootcamp.html">Command
	line bootcamp</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://korflab.ucdavis.edu/Unix_and_Perl/current.html">Unix
	and Perl Primer for Biologists</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.learnenough.com/command-line-tutorial">Learn
	Enough Command Line to Be Dangerous</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://cli.learncodethehardway.org/book/">The
	Command Line Crash Course</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://learnxinyminutes.com/docs/bash/">Learn
	Bash in Y minutes</a></font></p>
</ul>
<p class="first-paragraph">And there are many other options.</p>
<h2 class="western"><a name="how-do-i-troubleshoot-errors"></a>15.13
How do I troubleshoot errors?</h2>
<p class="first-paragraph">If you don&rsquo;t know how to do or fix
something try Googling it. Put the relevant section of the error
message into your search query. You will find that other people will
have had the same problem and someone will have told them what the
answer was. Google will likely direct you to one of these sites</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://askubuntu.com/">Ask
	Ubuntu</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://stackoverflow.com/">StackOverflow</a></font></p>
</ul>
<h2 class="western"><a name="where-can-i-learn-more-about-the-shell"></a>
15.14 Where can I learn more about the shell?</h2>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://files.fosswire.com/2007/08/fwunixref.pdf">Command
	Reference</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://explainshell.com/">Explain
	Shell</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-unix-bootcamp"></a>16 The Unix
bootcamp</h1>
<p class="first-paragraph">The following document has been adapted
from the <font color="#4f81bd"><a href="http://korflab.ucdavis.edu/bootcamp.html">Command-line
Bootcamp</a></font> by <font color="#4f81bd"><a href="http://www.keithbradnam.com/">Keith
Bradnam</a></font> licensed via Creative Commons Attribution 4.0
International License. The original content has been substantially
reworked, abbreviated and simplified.</p>
<h2 class="western"><a name="introduction"></a>16.1 Introduction</h2>
<p class="first-paragraph">This &lsquo;bootcamp&rsquo; is intended to
provide the reader with a basic overview of essential Unix/Linux
commands that will allow them to navigate a file system and move,
copy, edit files. It will also introduce a brief overview of some
&lsquo;power&rsquo; commands in Unix.</p>
<h2 class="western"><a name="why-unix"></a>16.2 Why Unix?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://en.wikipedia.org/wiki/Unix">Unix
operating system</a></font> has been around since 1969. Back then,
there was no such thing as a graphical user interface. You typed
everything. It may seem archaic to use a keyboard to issue commands
today, but it&rsquo;s much easier to automate keyboard tasks than
mouse tasks. There are several variants of Unix (including <font color="#4f81bd"><a href="http://en.wikipedia.org/wiki/Linux">Linux</a></font>),
though the differences do not matter much for most basic functions.</p>
<p>Increasingly, the raw output of biological research exists as <i>in
silico</i> data, usually in the form of large text files. Unix is
particularly suited to working with such files and has several
powerful (and flexible) commands that can process your data for you.
The real strength of learning Unix is that most of these commands can
be combined in an almost unlimited fashion. So if you can learn just
five Unix commands, you will be able to do a lot more than just five
things.</p>
<h2 class="western"><a name="typeset-conventions"></a>16.3 Typeset
Conventions</h2>
<p class="first-paragraph">Command-line examples that you are meant
to type into a terminal window will be shown indented in a
constant-width font, e.g.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-lrh</font></font></p>
<p class="first-paragraph">The lessons from this point onwards will
assume very little apart from the following:</p>
<ol>
	<li/>
<p class="compact">You have access to a Unix/Linux system</p>
	<li/>
<p class="compact">You know how to launch a terminal program
	on that system</p>
	<li/>
<p class="compact">You have a home directory where you can
	create/edit new files</p>
</ol>
<h2 class="western"><a name="the-terminal"></a>16.4 1. The Terminal</h2>
<p class="first-paragraph">A <i>terminal</i> is the common name for
the program that does two main things. It allows you to type input to
the computer (i.e.&nbsp;run programs, move/view files etc.) and it
allows you to see output from those programs. All Unix machines will
have a terminal program available.</p>
<p>Open the terminal application. You should now see something that
looks like the following:</p>
<p><img src="0000_html_2f36de05b0f3b844.png" name="Image39" align="bottom" hspace="12" width="560" height="367" border="0"/>
</p>
<p>Your terminal may not look like the image above. Depending on how
your system is set up, it may even look like the image below:</p>
<p><img src="0000_html_3b17bfccb0a44984.png" name="Image40" align="bottom" hspace="12" width="560" height="365" border="0"/>
</p>
<p>If your terminal is not easily readable customize it. Make the
fonts BIGGER, make the background lighter. On a Mac select
Preferences, on Windows right click the top bar and set more
appropriate colors and font.</p>
<p>Understand that learning Unix is also a &ldquo;physical&rdquo;
skill that requires a sort of of hand-eye coordination. You have to
teach your body to cooperate. You have to be comfortable, you have to
see the commands clearly and with no effort. The presence of a single
character in a multiline pipeline may change the meaning of the
entire process. Ensure that you have minimal distractions and that
the text in your terminal is eminently legible.</p>
<p>There will be many situations where it will be useful to have
multiple terminals open and it will be a matter of preference as to
whether you want to have multiple windows, or one window with
multiple tabs (there are typically keyboard shortcuts for switching
between windows, or moving between tabs).</p>
<h2 class="western"><a name="your-first-unix-command"></a>16.5 2.
Your first Unix command</h2>
<p class="first-paragraph">It is important to note that you will
always be <i>inside</i> a single directory when using the terminal.
The default behavior is that when you open a new terminal you start
in your own <i>home</i> directory (containing files and directories
that only you can modify). To see what files and directories are in
your home directory, you need to use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command. This command lists the contents of a directory. If we run
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command we should see something like:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></p>
<p class="first-paragraph">prints (this depends on what computer you
use):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Applications
Desktop Documents Downloads</font></font></p>
<p class="first-paragraph">There are four things that you should note
here:</p>
<ol>
	<li/>
<p class="compact">You will probably see different output to
	what is shown here, it depends on your operating system and computer
	setup. Don&rsquo;t worry about that for now.</p>
	<li/>
<p class="compact">In your case, you may see a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font>
	symbol. This is called the <b>Unix command prompt</b>. Note that the
	command prompt might not look the same on every Unix system. We do
	not include the command prompt in our examples so that you can
	copy-paste code. In general you use the command prompt to be able to
	tell that the computer is waiting on commands to be entered.</p>
	<li/>
<p class="compact">The output of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
	command lists two types of objects: files and directories. We&rsquo;ll
	learn how to tell them apart later on.</p>
	<li/>
<p class="compact">After the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
	command finishes it produces a new command prompt, ready for you to
	type your next command.</p>
</ol>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command is used to list the contents of <i>any</i> directory, not
necessarily the one that you are currently in. Try the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
/bin</font></font></p>
<p class="first-paragraph">it will print a lot of program names,
among them,</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
pwd mv</font></font></p>
<h2 class="western"><a name="the-unix-tree"></a>16.6 3: The Unix tree</h2>
<p class="first-paragraph">Looking at directories from within a Unix
terminal can often seem confusing. But bear in mind that these
directories are exactly the same type of folders that you can see if
you use any graphical file browser. From the <i>root</i> level (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font>)
there are usually a dozen or so directories. You can treat the root
directory like any other, e.g.&nbsp;you can list its contents:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
/</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bin
  dev   initrd.img      lib64       mnt   root  software  tmp 
vmlinuz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">boot
 etc   initrd.img.old  lost+found  opt   run   srv       usr 
vmlinuz.old</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">data
 home  lib             media       proc  sbin  sys       var</font></font></p>
<p class="first-paragraph">You might notice some of these names
appear in different colors. Many Unix systems will display files and
directories differently by default. Other colors may be used for
special types of files. When you log in to a computer, you are
typically placed in your home directory, which is often inside a
directory called &lsquo;Users&rsquo; or &lsquo;home&rsquo;.</p>
<h2 class="western"><a name="finding-out-where-you-are"></a>16.7 4:
Finding out where you are</h2>
<p class="first-paragraph">There may be many hundreds of directories
on any Unix machine, so how do you know which one you are in? The
command <font color="#4f81bd"><a href="http://en.wikipedia.org/wiki/Pwd">pwd</a></font>
will print the working directory. That&rsquo;s pretty much all this
command does:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert</font></font></p>
<p class="first-paragraph">When you log in to a Unix computer, you
are typically placed into your <i>home</i> directory. In this
example, after we log in, we are placed in a directory called
&lsquo;ialbert&rsquo; which itself is a <i>subdirectory</i> of
another directory called &lsquo;Users&rsquo;. Conversely, &lsquo;Users&rsquo;
is the <i>parent</i> directory of &lsquo;ialbert&rsquo;. The first
forward slash that appears in a list of directory names always refers
to the top level directory of the file system (known as the <i>root
directory</i>). The remaining forward slash (between &lsquo;Users&rsquo;
and &lsquo;ialbert&rsquo;) delimits the various parts of the
directory hierarchy. If you ever get &lsquo;lost&rsquo; in Unix,
remember the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
command.</p>
<p>As you learn Unix you will frequently type commands that don&rsquo;t
seem to work. Most of the time this will be because you are in the
wrong directory, so it&rsquo;s a really good habit to get used to
running the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
command a lot.</p>
<h2 class="western"><a name="making-new-directories"></a>16.8 5:
Making new directories</h2>
<p class="first-paragraph">If we want to make a new directory
(e.g.&nbsp;to store some lecture related files), we can use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font>
command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
edu</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font></p>
<p class="first-paragraph">shows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">edu
</font></font>
</p>
<h2 class="western"><a name="getting-from-a-to-b"></a>16.9 6: Getting
from &lsquo;A&rsquo; to &lsquo;B&rsquo;</h2>
<p class="first-paragraph">We are in the home directory on the
computer but we want to to work in the new <font face="Consolas, serif"><font size="2" style="font-size: 11pt">edu</font></font>
directory. To change directories in Unix, we use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font>
command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
edu</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu</font></font></p>
<p class="first-paragraph">Let&rsquo;s make two new subdirectories
and navigate into them:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
lecture</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
lecture</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu/lecture</font></font></p>
<p class="first-paragraph">then make a data directory:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
data/</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu/lecture/data</font></font></p>
<p class="first-paragraph">We created the two directories in separate
steps, but it is possible to use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font>
command in way to do this all in one step.</p>
<p>Like most Unix commands, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font>
supports <i>command-line options</i> which let you alter its behavior
and functionality. Command-like options are &ndash;&nbsp;as the name
suggests &ndash;&nbsp;optional arguments that are placed after the
command name. They often take the form of single letters (following a
dash). If we had used the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-p</font></font>
option of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font>
command we could have done this in one step. E.g.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/edu/lecture/docs</font></font></p>
<p class="first-paragraph"><i><b>Note the spaces either side of the
</b></i><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i><b>-p</b></i></font></font><i><b>!</b></i></p>
<h2 class="western"><a name="the-root-directory"></a>16.10 7: The
root directory</h2>
<p class="first-paragraph">Let&rsquo;s change directory to the root
directory:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
/</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
Users</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
ialbert</font></font></p>
<p class="first-paragraph">In this case, we may as well have just
changed directory in one go:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
/Users/ialbert</font></font></p>
<p class="first-paragraph">The leading <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font>
is incredibly important. The following two commands are very
different:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
/step1/step2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
step1/step2/</font></font></p>
<p class="first-paragraph">The first command specifies a so called
<i>absolute path</i> . It says go the root directory (folder) then
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step1</font></font>
folder then the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step2</font></font>
folder that opens from the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step1</font></font>
directory. Importantly there can only be one <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/step1/step2</font></font>
directory on any Unix system.</p>
<p>The second command specifies a so called <i>relative path</i>. It
says that from the <i>current location</i> go to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step1</font></font>
directory then the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step2</font></font>
directory.</p>
<p>There can potentially be many <font face="Consolas, serif"><font size="2" style="font-size: 11pt">step1/step2</font></font>
directories that open from different directories.</p>
<p>Learn and understand the difference between these two commands.
Initially it is very easy to miss the leading <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font>
yet it is essential and fundamentally important. With a little time
and practice your eyes will be trained and you will quickly notice
the difference.</p>
<h2 class="western"><a name="navigating-upwards-in-the-unix-filesystem"></a>
16.11 8: Navigating upwards in the Unix filesystem</h2>
<p class="first-paragraph">Frequently, you will find that you want to
go &lsquo;upwards&rsquo; one level in the directory hierarchy. Two
dots <font face="Consolas, serif"><font size="2" style="font-size: 11pt">..</font></font>
are used in Unix to refer to the <i>parent</i> directory of wherever
you are. Every directory has a parent except the root level of the
computer. Let&rsquo;s go into the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">lecture</font></font>
directory and then navigate up two levels:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
edu</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu
</font></font>
</p>
<p class="first-paragraph">but now:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
..</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert</font></font></p>
<p class="first-paragraph">What if you wanted to navigate up <i>two</i>
levels in the file system in one go? Use two sets of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">..</font></font>
operator, separated by a forward slash:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
../..</font></font></p>
<h2 class="western"><a name="absolute-and-relative-paths"></a>16.12
9: Absolute and relative paths</h2>
<p class="first-paragraph">Using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
..</font></font> allows us to change directory <i>relative</i> to
where we are now. You can also always change to a directory based on
its <i>absolute</i> location. E.g. if you are working in the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/edu/lecture</font></font>
directory and you want to change to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/edu/tmp</font></font>
directory, then you could do either of the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Go down one level. Use a relative path.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
../tmp</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">For me it prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu/tmp</font></font></p>
<p class="first-paragraph">or:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Use an absolute path.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~/edu/tmp</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">the last command also prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/edu/tmp</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~</font></font>
is a shorthand notation that gets substituted to my home directory
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/</font></font>.
You can think of it as a mixture of relative and absolute paths. It
is a full path, but it expands to different locations for different
people. If I were to specify the path as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This is an &quot;even more&quot; absolute path.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
/Users/ialbert/edu/tmp</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font></p>
<p class="first-paragraph">Then the example would only work on my
computer (or other computers where the home directory is also named
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ialbert</font></font>).
Using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~</font></font>
allows us to get the best of both worlds. It is an absolute path with
a little bit of <i>relativity</i> in it.</p>
<p>All examples achieve the same thing, but the later examples
require that you know about more about the location of your directory
of interest. It is important that you understand the slight
differences here, as the absolute and relative paths are important
concepts that you will have make use of frequently.</p>
<p>Sometimes it is quicker to change directories using the relative
path, and other times it will be quicker to use the absolute path.</p>
<h2 class="western"><a name="finding-your-way-back-home"></a>16.13
10: Finding your way back home</h2>
<p class="first-paragraph">Remember that the command prompt shows you
the name of the directory that you are currently in, and that when
you are in your home directory it shows you a tilde character (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~</font></font>)
instead? This is because Unix uses the tilde character as a
short-hand way of [specifying a home directory][home directory].</p>
<p>See what happens when you try the following commands (use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
command after each one to confirm the results):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
/ </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~ </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font></p>
<p class="first-paragraph">Hopefully, you should find that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~</font></font> do the same thing, i.e.&nbsp;they take you back to
your home directory (from wherever you were). You will frequently
want to jump straight back to your home directory, and typing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font>
is a very quick way to get there.</p>
<p>You can also use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~</font></font>
as a quick way of navigating into subdirectories of your home
directory when your current directory is somewhere else. I.e. the
quickest way of navigating from anywhere on the filesystem to your
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">edu/lecture</font></font>
directory is as follows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/edu/lecture</font></font></p>
<h2 class="western"><a name="making-the-ls-command-more-useful"></a>16.14
11: Making the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command more useful</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">..</font></font>
operator that we saw earlier can also be used with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command, e.g.&nbsp;you can list directories that are &lsquo;above&rsquo;
you:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/edu/lecture/</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
../../</font></font></p>
<p class="first-paragraph">Time to learn another useful command-line
option. If you add the letter &lsquo;l&rsquo; to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command it will give you a longer output compared to the default:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l ~</font></font></p>
<p class="first-paragraph">prints (system-dependent):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">drwx------
  5 ialbert  staff   170B May 13 15:24 Applications</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">drwx------+
 7 ialbert  staff   238B Aug 20 05:51 Desktop</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">drwx------+
30 ialbert  staff   1.0K Aug 12 13:12 Documents</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">drwx------+
14 ialbert  staff   476B Aug 24 10:43 Downloads</font></font></p>
<p class="first-paragraph">For each file or directory we now see more
information (including file ownership and modification times). The
&lsquo;d&rsquo; at the start of each line indicates that these are
directories. There are many, many different options for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
command. Try out the following (against any directory of your choice)
to see how the output changes.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-R </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l -t -r </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-lh</font></font></p>
<p class="first-paragraph">Note that the last example combines
multiple options but only uses one dash. This is a very common way of
specifying multiple command-line options. You may be wondering what
some of these options are doing. It is time to learn about Unix
documentation&hellip;.</p>
<h2 class="western"><a name="man-pages"></a>16.15 12: Man pages</h2>
<p class="first-paragraph">If every Unix command has so many options,
you might be wondering how you find out what they are and what they
do. Well, thankfully every Unix command has an associated &lsquo;manual&rsquo;
that you can access by using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font>
command. E.g.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ls </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
cd</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
man </font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Yes, even the man command has a manual page</i></font></font></font></p>
<p class="first-paragraph">When you are using the man command, press
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">space</font></font>
to scroll down a page, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">b</font></font>
to go back a page, or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">q</font></font>
to quit. You can also use the up and down arrows to scroll a line at
a time. The man command is actually using another Unix program, a
text viewer called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>,
which we&rsquo;ll come to later on.</p>
<h2 class="western"><a name="removing-directories"></a>16.16 13:
Removing directories</h2>
<p class="first-paragraph">We now have a few (empty) directories that
we should remove. To do this use the [rmdir][] command. This will
only remove empty directories, so it is quite safe to use. If you
want to know more about this command (or any Unix command), then
remember that you can just look at its man page.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/edu/lecture/</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rmdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
..</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rmdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
lecture</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font>
</p>
<p class="first-paragraph"><b>Note, you have to be outside a
directory before you can remove it with </b><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>rmdir</b></font></font></p>
<h2 class="western"><a name="using-tab-completion"></a>16.17 14:
Using tab completion</h2>
<p class="first-paragraph">Saving keystrokes may not seem important
now, but the longer that you spend typing in a terminal window, the
happier you will be if you can reduce the time you spend at the
keyboard. Especially as prolonged typing is not good for your body.
So the best Unix tip to learn early on is that you can [tab
complete][] the names of files and programs on most Unix systems.
Type enough letters to uniquely identify the name of a file,
directory, or program and press tab &ndash; Unix will do the rest.
E.g. if you type &lsquo;tou&rsquo; and then press tab, Unix should
autocomplete the word to &lsquo;touch&rsquo; (this is a command which
we will learn more about in a minute). In this case, tab completion
will occur because there are no other Unix commands that start with
&lsquo;tou&rsquo;. If pressing tab doesn&rsquo;t do anything, then
you have not have typed enough unique characters. In this case
pressing tab <i>twice</i> will show you all possible completions.
This trick can save you a LOT of typing!</p>
<p>Navigate to your home directory, and then use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font>
command to change to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">edu</font></font>
directory. Use tab completion to complete directory name. If there
are no other directories starting with &lsquo;e&rsquo; in your home
directory, then you should only need to type &lsquo;cd&rsquo; + &lsquo;e&rsquo;
+ &lsquo;tab&rsquo;.</p>
<p class="block-text-western"><b>Tab completion will make your life
easier and make you more productive!</b></p>
<p class="first-paragraph">Another great time-saver is that Unix
stores a list of all the commands that you have typed in each login
session. You can access this list by using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">history</font></font>
command or more simply by using the up and down arrows to access
anything from your history. So if you type a long command but make a
mistake, press the up arrow and then you can use the left and right
arrows to move the cursor in order to make a change.</p>
<h2 class="western"><a name="creating-empty-files-with-the-touch-command"></a>
16.18 15: Creating empty files with the touch command</h2>
<p class="first-paragraph">The following sections will deal with Unix
commands that help us to work with files, i.e.&nbsp;copy files
to/from places, move files, rename files, remove files, and most
importantly, look at files. First, we need to have some files to play
with. The Unix command <font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font>
will let us create a new, empty file. The touch command does other
things too, but for now we just want a couple of files to work with.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
edu</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
heaven.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
earth.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">earth.txt
 heaven.txt</font></font></p>
<h2 class="western"><a name="moving-files"></a>16.19 16: Moving files</h2>
<p class="first-paragraph">Now, let&rsquo;s assume that we want to
move these files to a new directory (&lsquo;temp&rsquo;). We will do
this using the Unix [mv][] (move) command. Remember to use tab
completion:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
heaven.txt temp</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
earth.txt temp</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp</font></font></p>
<p class="first-paragraph">For the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>
command, we always have to specify a source file (or directory) that
we want to move, and then specify a target location. If we had wanted
to, we could have moved both files in one go by typing any of the
following commands:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*.txt temp </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*t temp </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*ea* temp</font></font></p>
<p class="first-paragraph">The asterisk <font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>
acts as a <i>wild-card character</i>, essentially meaning &lsquo;match
anything&rsquo;. The second example works because there are no other
files or directories in the directory that end with the letter &lsquo;t&rsquo;
(if there were, then they would be moved too). Likewise, the third
example works because only those two files contain the letters &lsquo;ea&rsquo;
in their names. Using wild-card characters can save you a lot of
typing.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">?</font></font>
character is also a wild-card but with a slightly different meaning.
See if you can work out what it does.</p>
<h2 class="western"><a name="renaming-files"></a>16.20 17: Renaming
files</h2>
<p class="first-paragraph">In the earlier example, the destination
for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>
command was a directory name (temp). So we moved a file from its
source location to a target location, but note that the target could
have also been a (different) file name, rather than a directory. E.g.
let&rsquo;s make a new file and move it whilst renaming it at the
same time:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
rags</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
rags temp/riches</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp/</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">earth.txt
 heaven.txt  riches</font></font></p>
<p class="first-paragraph">In this example we create a new file
(&lsquo;rags&rsquo;) and move it to a new location and in the process
change the name (to &lsquo;riches&rsquo;). So <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>
can rename a file as well as move it. The logical extension of this
is using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>
to rename a file without moving it. You may also have access to a
tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rename</font></font>,
type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">man
rename</font></font> for more information.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp/riches temp/rags</font></font></p>
<h2 class="western"><a name="moving-directories"></a>16.21 18: Moving
directories</h2>
<p class="first-paragraph">It is important to understand that as long
as you have specified a &lsquo;source&rsquo; and a &lsquo;target&rsquo;
location when you are moving a file, then it doesn&rsquo;t matter
what your <i>current</i> directory is. You can move or copy things
within the same directory or between different directories regardless
of whether you are in any of those directories. Moving directories is
just like moving files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp temp2</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
temp2</font></font></p>
<h2 class="western"><a name="removing-files"></a>16.22 19: Removing
files</h2>
<p class="first-paragraph">You&rsquo;ve seen how to remove a
directory with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rmdir</font></font>
command, but <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rmdir</font></font>
won&rsquo;t remove directories if they contain any files. So how can
we remove the files we have created (inside <font face="Consolas, serif"><font size="2" style="font-size: 11pt">temp</font></font>)?
In order to do this, we will have to use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>
(remove) command.</p>
<p class="block-text-western"><b>Please read the next section VERY
carefully. Misuse of the </b><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>rm</b></font></font><b>
command can lead to needless death &amp; destruction</b></p>
<p class="first-paragraph">Potentially, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>
is a very dangerous command; if you delete something with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>,
you will not get it back! It is possible to delete everything in your
home directory (all directories and subdirectories) with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>.
That is why it is such a dangerous command.</p>
<p>Let me repeat that last part again. It is possible to delete EVERY
file you have ever created with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>
command. Are you scared yet? You should be. Luckily there is a way of
making <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>
a little bit safer. We can use it with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-i</font></font>
command-line option which will ask for confirmation before deleting
anything (remember to use tab-completion):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
temp</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm
-i earth.txt heaven.txt rags</font></font></p>
<p class="first-paragraph">will ask permission for each step:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm:
remove regular empty file 'earth.txt'? y</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm:
remove regular empty file 'heaven.txt'? y</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm:
remove regular empty file 'rags'? y</font></font></p>
<p class="first-paragraph">We could have simplified this step by
using a wild-card (e.g.&nbsp;<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm
-i *.txt</font></font>) or we could have made things more complex by
removing each file with a separate <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font>
command.</p>
<h2 class="western"><a name="copying-files"></a>16.23 20: Copying
files</h2>
<p class="first-paragraph">Copying files with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cp</font></font>
(copy) command has a similar syntax as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font>,
but the file will remain at the source and be copied to the target
location. Remember to always specify a source and a target location.
Let&rsquo;s create a new file and make a copy of it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file1</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cp</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file1 file2</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font></p>
<p class="first-paragraph">What if we wanted to copy files from a
different directory to our current directory? Let&rsquo;s put a file
in our home directory (specified by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~</font></font>,
remember) and copy it to the lecture directory (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/edu/lecture</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">touch</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/edu/file3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cp</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/edu/file3 ~/edu/lecture/</font></font></p>
<p class="first-paragraph">In Unix, the current directory can be
represented by a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
(dot) character. You will often use for copying files to the
directory that you are in. Compare the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
. </font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
./</font></font></p>
<p class="first-paragraph">In this case, using the dot is somewhat
pointless because <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
will already list the contents of the current directory by default.
Also note how the trailing slash is optional.</p>
<h2 class="western"><a name="copying-directories"></a>16.24 21:
Copying directories</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cp</font></font>
command also allows us (with the use of a command-line option) to
copy entire directories. Use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">man
cp</font></font> to see how the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-R</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-r</font></font>
options let you copy a directory <i>recursively</i>.</p>
<h2 class="western"><a name="viewing-files-with-less-or-more"></a>16.25
22: Viewing files with less (or more)</h2>
<p class="first-paragraph">So far we have covered listing the
contents of directories and moving/copying/deleting either files or
directories. Now we will quickly cover how you can look at files. The
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">more</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>
commands let you view (but not edit) text files. We will use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>
command to put some text in a file and then view it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Call me Ishmael.&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Call
me Ishmael.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;Call me Ishmael.&quot; &gt; opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">opening_lines.txt</font></font></p>
<p class="first-paragraph">we can view the content of the file with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">more
opening_lines.txt</font></font></p>
<p class="first-paragraph">On its own, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>
isn&rsquo;t a very exciting Unix command. It just echoes text back to
the screen. But we can redirect that text into an output file by
using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>
symbol. This allows for something called file <i>redirection</i>.</p>
<p class="block-text-western"><b>Careful when using file redirection
(&gt;), it will overwrite any existing file of the same name</b></p>
<p class="first-paragraph">When you are using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">more</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>,
you can bring up a page of help commands by pressing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">h</font></font>,
scroll forward a page by pressing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">space</font></font>,
or go forward or backwards one line at a time by pressing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">j</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>.
To exit <font face="Consolas, serif"><font size="2" style="font-size: 11pt">more</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>,
press <font face="Consolas, serif"><font size="2" style="font-size: 11pt">q</font></font>
(for quit). The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">more</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>
programs also do about a million other useful things (including text
searching).</p>
<h2 class="western"><a name="viewing-files-with-cat"></a>16.26 23:
Viewing files with cat</h2>
<p class="first-paragraph">Let&rsquo;s add another line to the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;The primroses were over.&quot; &gt;&gt; opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
opening_lines.txt</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Call
me Ishmael.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">The
primroses were over.</font></font></p>
<p class="first-paragraph">Notice that we use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;&gt;</font></font>
and not just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>.
This operator will <b>append</b> to a file. If we only used <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>,
we would end up overwriting the file. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>
command displays the contents of the file (or files) and then returns
you to the command line. Unlike <font face="Consolas, serif"><font size="2" style="font-size: 11pt">less</font></font>
you have no control on how you view that text (or what you do with
it). It is a very simple, but sometimes useful, command. You can use
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>
to quickly combine multiple files or, if you wanted to, make a copy
of an existing file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
opening_lines.txt </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file_copy.txt</font></font></p>
<h2 class="western"><a name="counting-characters-in-a-file"></a>16.27
24: Counting characters in a file</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">opening_lines.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">total</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-rw-rw-r--</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
1 ubuntu ubuntu 42 Jun 15 04:13 opening_lines.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 7 42 opening_lines.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
opening_lines.txt</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l</font></font> option shows us a long listing, which includes the
size of the file in bytes (in this case &lsquo;42&rsquo;). Another
way of finding this out is by using Unix&rsquo;s <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font>
command (word count). By default this tells you many lines, words,
and characters are in a specified file (or files), but you can use
command-line options to give you just one of those statistics (in
this case we count lines with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc
-l</font></font>).</p>
<h2 class="western"><a name="editing-small-text-files-with-nano"></a>16.28
25: Editing small text files with nano</h2>
<p class="first-paragraph">Nano is a lightweight editor installed on
most Unix systems. There are many more powerful editors (such as
&lsquo;emacs&rsquo; and &lsquo;vi&rsquo;), but these have steep
learning curves. Nano is very simple. You can edit (or create) files
by typing:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
opening_lines.txt</font></font></p>
<p class="first-paragraph">You should see the following appear in
your terminal:</p>
<p><img src="0000_html_90d77b46447cc8b9.png" name="Image41" align="bottom" hspace="12" width="560" height="338" border="0"/>
</p>
<p>The bottom of the nano window shows you a list of simple commands
which are all accessible by typing &lsquo;Control&rsquo; plus a
letter. E.g. Control + X exits the program.</p>
<h2 class="western"><a name="the-path-environment-variable"></a>16.29
26: The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font>
environment variable</h2>
<p class="first-paragraph">One other use of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>
command is for displaying the contents of something known as
<i>environment variables</i>. These contain user-specific or
system-wide values that either reflect simple pieces of information
(your username), or lists of useful locations on the file system.
Some examples:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$USER</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ialbert</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$HOME</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">/usr/local</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/sbin:</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/usr/local</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/bin:</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/bin</font></font></p>
<p class="first-paragraph">The last one shows the content of the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font>
environment variable, which displays a &mdash;&nbsp;colon separated
&mdash;&nbsp;list of directories that are expected to contain
programs that you can run. This includes all of the Unix commands
that you have seen so far. These are files that live in directories
which are run like programs (e.g.&nbsp;<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
is just a special type of file in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/bin</font></font>
directory).</p>
<p>Knowing how to change your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font>
to include custom directories can be necessary sometimes (e.g.&nbsp;if
you install some new bioinformatics software in a non-standard
location).</p>
<h2 class="western"><a name="matching-lines-in-files-with-grep"></a>16.30
27: Matching lines in files with grep</h2>
<p class="first-paragraph">Use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font>
to add the following lines to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">opening_lines.txt</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Now
is the winter of our discontent.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">All
children, except one, grow up.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">The
Galactic Empire was dying.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">In
a hole in the ground there lived a hobbit.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It
was a pleasure to burn.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It
was a bright, cold day in April, and the clocks were striking
thirteen.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It
was love at first sight.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">I
am an invisible man.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It
was the day my grandmother exploded.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">When
he was nearly thirteen, my brother Jem got his arm badly broken at
the elbow.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Marley
was dead, to begin with.</font></font></p>
<p class="first-paragraph">You will often want to search files to
find lines that match a certain pattern. The Unix command <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
does this (and much more). The following examples show how you can
use grep&rsquo;s command-line options to:</p>
<ul>
	<li/>
<p class="compact">show lines that match a specified pattern</p>
	<li/>
<p class="compact">ignore case when matching (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-i</font></font>)</p>
	<li/>
<p class="compact">only match whole words (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-w</font></font>)</p>
	<li/>
<p class="compact">show lines that don&rsquo;t match a pattern
	(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-v</font></font>)</p>
	<li/>
<p class="compact">Use wildcard characters and other patterns
	to allow for alternatives (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>,
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[]</font></font>)</p>
</ul>
<p class="first-paragraph">Show lines that match the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">was</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">The</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Galactic Empire was dying.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was a pleasure to burn.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was a bright, cold day in April, and the clocks were striking
thirteen.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was love at first sight.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">It</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was the day my grandmother exploded.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">When</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
he was nearly thirteen, my brother Jem got his arm badly broken at
the elbow.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Marley</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was dead, to begin with.</font></font></p>
<p class="first-paragraph">Use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
--color=AUTO was opening_lines.txt</font></font></p>
<p class="first-paragraph">to highlight the match.</p>
<p>Show lines that do not match the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">was</font></font>
:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v was opening_lines.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Now</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
is the winter of our discontent.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">All</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
children, except one, grow up.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">In</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
a hole in the ground there lived a hobbit.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">I</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
am an invisible man.</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
has a great many options and applications.</p>
<h2 class="western"><a name="combining-unix-commands-with-pipes"></a>16.31
28: Combining Unix commands with pipes</h2>
<p class="first-paragraph">One of the most powerful features of Unix
is that you can send the output from one command or program to any
other command (as long as the second command accepts input of some
sort). We do this by using what is known as a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pipe</font></font>.
This is implemented using the &lsquo;|&rsquo; character (which is a
character which always seems to be on different keys depending on the
keyboard that you are using). Think of the pipe as simply connecting
two Unix programs. Here&rsquo;s an example which introduces some new
Unix commands:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was opening_lines.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
316</i></font></font></font><br/>
<br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
was opening_lines.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-n 3 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
130</i></font></font></font></p>
<p class="first-paragraph">The first use of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
searches the specified file for lines matching &lsquo;was&rsquo;. It
sends the lines that match through a pipe to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font>
program. We use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c</font></font>
option to just count characters in the matching lines (316).</p>
<p>The second example first sends the output of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
to the Unix <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font>
command. This sorts a file alphanumerically by default. The sorted
output is sent to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font>
command which by default shows the first 10 lines of a file. We use
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-n</font></font>
option of this command to only show 3 lines. These 3 lines are then
sent to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font>
command as before.</p>
<p class="block-text-western"><b>Whenever making a long pipe, test
each step as you build it!</b></p>
<h2 class="western"><a name="miscellaneous-unix-power-commands"></a>16.32
Miscellaneous Unix power commands</h2>
<p class="first-paragraph">The following examples introduce some
other Unix commands, and show how they could be used to work on a
fictional file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">file.txt</font></font>.
Remember, you can always learn more about these Unix commands from
their respective man pages with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font>
command. These are not all real world cases, but rather show the
diversity of Unix command-line tools:</p>
<ul>
	<li/>
<p class="compact">View the penultimate (second-to-last) 10
	lines of a file (by piping <font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tail</font></font>
	commands):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#tail
-n 20 gives the last 20 lines of the file, and piping that to head
will show the first 10 lines of the last 20.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tail</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-n 20 file.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<ul>
	<li/>
<p class="compact">Show the lines of a file that begin with a
	start codon (ATG) (the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">^</font></font>
	matches patterns at the start of a line):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;^ATG&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file.txt</font></font></p>
<ul>
	<li/>
<p class="compact">Cut out the 3rd column of a tab-delimited
	text file and sort it to only show unique lines (i.e.&nbsp;remove
	duplicates):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 file.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-u</font></font></p>
<ul>
	<li/>
<p class="compact">Count how many lines in a file contain the
	words &lsquo;cat&rsquo; or &lsquo;bat&rsquo; (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c</font></font>
	option of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
	counts lines):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'[bc]at'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file.txt</font></font></p>
<ul>
	<li/>
<p class="compact">Turn lower-case text into upper-case (using
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tr</font></font>
	command to &lsquo;transliterate&rsquo;):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tr</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'a-z'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'A-Z'</font></font></font></p>
<ul>
	<li/>
<p class="compact">Change all occurrences of &lsquo;Chr1&rsquo;
	to &lsquo;Chromosome 1&rsquo; and write changed output to a new file
	(using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sed</font></font>
	command):</p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sed</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'s/Chr1/Chromosome
1/'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
file2.txt</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="defense-against-the-dark-arts"></a>17
Defense against the Dark Arts</h1>
<p class="first-paragraph">When you start out, errors will abound.</p>
<p>That&rsquo;s ok, we make errors all the time as well.</p>
<p>Watch any expert bioinformatician at work, they&rsquo;ll make
error all the time. Their experience allows them to recover from
errors quickly without a complete breakdown of the process. That
should be your aim to, read to learn and identify why an error
occurs. Errors are no big deal, happen all the time.</p>
<p class="block-text-western"><b>Note</b>: Study the error messages.
The errors are usually quite verbose. Typically a single sentences
indicate what actually went wrong. All other errors may be a
consequence of the first error.</p>
<p class="first-paragraph">Initially the following two curses will
follow you around:</p>
<ol>
	<li/>
<p class="compact">The <b>curse of doom</b>: My program does
	not run!!!!</p>
	<li/>
<p class="compact">The <b>curse of agony</b>: My file does not
	exist!!!!</p>
</ol>
<p class="first-paragraph">Learn to deal with each, and you&rsquo;ll
be well on your way to become productive.</p>
<h3 class="western"><a name="the-curse-of-doom-my-program-does-not-run"></a>
17.0.1 1.The curse of doom: My program does not run!</h3>
<p class="first-paragraph">Suppose you read about a new method that
works like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">brangelina
--shift 300 --report 200</font></font></p>
<p class="first-paragraph">But when you type it and press enter it
prints this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-bash:
brangelina: command not found</font></font></p>
<p class="first-paragraph">We call the error above the <i>curse of
doom</i>. In your terminal it might look like this:</p>
<p><br/>
<br/>

</p>
<p><b>Why am I getting the curse of doom?</b></p>
<p>The most likely explanation for getting the error:</p>
<ul>
	<li/>
<p class="compact">You have not activated the proper conda
	environment: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
	activate bioinfo</font></font></p>
	<li/>
<p class="compact">The name is not exactly what you typed (Use
	TAB completion!).</p>
	<li/>
<p class="compact">The program is not installed.</p>
</ul>
<p class="first-paragraph">Make sure that you have followed all
instructions as described in the <font color="#4f81bd"><a href="https://www.biostarhandbook.com/computer-setup.html">Computer
setup</a></font> chapter.</p>
<h3 class="western"><a name="the-curse-of-agony-my-file-does-not-exist"></a>
17.0.2 2. The curse of agony: My file does not exist!</h3>
<p class="first-paragraph">Suppose you type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
bennifer.txt</font></font></p>
<p class="first-paragraph">then it prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat:
bennifer.txt: No such file or directory</font></font></p>
<p class="first-paragraph">We call these types of errors the <i>curse
of agony</i>. In your terminal it might look like this:</p>
<p><br/>
<br/>

</p>
<p><b>Why am I getting the curse of agony?</b></p>
<p>The most likely explanation for getting the error:</p>
<ul>
	<li/>
<p class="compact">The file does not exist.</p>
	<li/>
<p class="compact">The file exists but is somewhere else on
	your computer.</p>
	<li/>
<p class="compact">The name is not exactly what you typed (Use
	TAB completion!).</p>
</ul>
<p class="first-paragraph">Visit the chapter called <font color="#4f81bd"><a href="https://www.biostarhandbook.com/the-unix-bootcamp.html">Unix
Bootcamp</a></font> and make sure that you understand file paths,
relative and absolute paths, and the Unix tree.</p>
<p>The reasons listed above solve 90% of problems (perhaps 100% for
beginners).</p>
<h3 class="western"><a name="both-errors-at-the-same-time"></a>17.0.3
3. Both errors at the same time</h3>
<p class="first-paragraph">If you are really down on your luck you
may also end up with both curses at the same time. Annoyingly,
depending on how you invoke the commands you may get the errors
manifest in different order:</p>
<p>For example if you type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">brangelina
--input bennifer.txt</font></font></p>
<p class="first-paragraph">the command will stop with the <i>curse of
doom</i>. Invoking it as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">brangelina
&lt; bennifer.txt</font></font></p>
<p class="first-paragraph">will lash you with the <i>curse of agony</i>.
What is confusing here is that depending on how you use the commands
either of the errors shows up first, like so:</p>
<p><br/>
<br/>

</p>
<p>Then once you correct the first error, the second error manifests
itself. Needless to say, super annoying.</p>
<p>Defense: use the TAB-charm to complete program and file names. It
keeps evil spirits at bay.</p>
<h2 class="western"><a name="bonus-round-windows-woes"></a>17.1 4.
Bonus round: Windows Woes</h2>
<p class="first-paragraph">This usually happens on Windows only, but
it may strike the unwary at any moment. Suppose you have a beautiful
well designed script that contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Stop on errors.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-uex</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Print the greeting.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
hello world</font></font></p>
<p class="first-paragraph">but then when you run it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
magic.sh</font></font></p>
<p class="first-paragraph">instead of the desired incantation you
get:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
command not found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
invalid optionset: -</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set:
usage: set [--abefhkmnptuvxBCHP] [-o option] [arg ...]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
command not found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">hello
world</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
command not found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
command not found</font></font></p>
<p class="first-paragraph">The code all looks correct but when run,
it won&rsquo;t work at all. The answer is that your file was created
with Windows style line endings and those are different from Unix
style line endings. See the section <font color="#4f81bd"><a href="#ch013.xhtml#choose-editor">Choose
a text editor</a></font>.</p>
<p>As for the solution, you need to set the line-ending to be Unix.
The task is usually changing a settings (the exact location depends
on editor and version sometimes).</p>
<h2 class="western"><a name="what-if-i-am-sure-that-i-typed-the-commands-correctly"></a>
17.2 What if I am sure that I typed the commands correctly?</h2>
<p class="compact">Often you are <b>absolutely certain</b> that you
did everything right, yet you are still getting the error. In those
cases the answer is that you <b>only think</b> you are doing it the
way you are supposed to.</p>
<p>Surprisingly often your eyes and brain will conspire against you,
they might not let you see that tiny mistake that you made. It
happens to me all the time. Sometimes I see only what I want to see.
In those cases the best course of action is to clear your view:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">clear</font></font></p>
<p class="first-paragraph">This clears the terminal and mind and with
that all preconceptions you might have about what the command should
look like. Now <i>retype everything</i> from the very beginning! No
copy-pasta! Build your commands out one piece at a time, keep
pressing TAB and make use of autocomplete.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="data-analysis-with-unix"></a>18 Data
analysis with Unix</h1>
<p class="first-paragraph">In this section, we will use simple Unix
tools to obtain and investigate a genome data file that represents
all the annotated features of the yeast genome.</p>
<p>This section is not meant to be a full introduction to all the
utility of the Unix command line.</p>
<p>What we attempt to show you are the concepts, methods and the
manner in which Unix tools are typically employed to probe and poke
at data.</p>
<h2 class="western"><a name="what-directory-should-i-use"></a>18.1
What directory should I use?</h2>
<p class="first-paragraph">When you open a terminal, it starts in a
specific location (directory, folder) on your computer. You can, of
course, move around and navigate to other places, but you have to
remember that every action that you take is always relative to the
location from which you are giving the command.</p>
<p>For each project, lecture or exploration create a separate folder
that you use only for that task. Starting with an empty directory
will help you keep track of the files you generate at each step.</p>
<p>Never start typing commands from your home directory. Tools create
several ancillary files, and you do not want these mixed in with your
other work-related data.</p>
<p>Your two BFFs (best friends forever) are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>.
These to commands will tell you the most important information:</p>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
	will tell you <i>where</i> you are.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
	will tell you <i>what</i> you have there.</p>
</ol>
<p class="first-paragraph">Especially at the beginning a lot of
confusion is caused by not knowing these two pieces of information.
When in doubt ask yourself:</p>
<ol>
	<li/>
<p class="compact">Where am I on my computer?</p>
	<li/>
<p class="compact">What do I have in the current folder?</p>
</ol>
<h2 class="western"><a name="where-are-we-getting-the-data-from"></a>18.2
Where are we getting the data from?</h2>
<p class="first-paragraph">We&rsquo;ll visit the <font color="#4f81bd"><a href="http://www.yeastgenome.org/">Saccharomyces
Genome Database (SGD)</a></font> webpage. What you will notice that
while clicking around &ldquo;makes sense&rdquo; it is not easy to
properly describe where you would get a given dataset. All too often
you will hear people say: <i>&ldquo;I&rsquo;ve downloaded data from
the XYZ database&rdquo;</i> as if there was only one data to be
downloaded. Often you are left scratching your head as to what
exactly did they download.</p>
<p><img src="0000_html_89841b65b5cf4460.png" name="Image42" align="bottom" hspace="12" width="560" height="161" border="0"/>
</p>
<p class="block-text-western"><b>Note</b>: Websites may change
unexpectedly and without notice. When a site changes the links may
break. To avoid the problems caused by changes outside of our
control, we host some of the data on the handbook website.</p>
<p class="first-paragraph">SGD like many other archives has a
notorious habit of capriciously (and nonchalantly) changing the
location of the data. This promptly breaks links to any training
resorces. Visit the main website if the links don&rsquo;t seem to
work. Do also let us know at by opening an issue at <font color="#4f81bd"><a href="https://github.com/biostars/biostar-handbook-issues">Biostar
Handbook Issues</a></font>. As an example the domain has recently
changed from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">downloads.yeastgenome.org</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sgd-archive.yeastgenome.org</font></font>.</p>
<p>Visit the <font color="#4f81bd"><a href="http://sgd-archive.yeastgenome.org/">&ldquo;Download&rdquo;</a></font>
link then <font color="#4f81bd"><a href="http://sgd-archive.yeastgenome.org/curation/">&ldquo;Curated
Data&rdquo;</a></font> then <b>&ldquo;Chromosomal Features&rdquo;.</b>
From that section we want to investigate the file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SGD_features.tab</font></font>
that is described as <i>&ldquo;Chromosomal features, coordinates,
gene descriptions.&rdquo;</i>. Let&rsquo;s get it into our terminal.</p>
<p><img src="0000_html_d2b396a54fcc3ff1.png" name="Image43" align="bottom" hspace="12" width="560" height="129" border="0"/>
</p>
<p>The first thing you need to do is obtain the URL of this file.
While hovering over the link right click and select &ldquo;Copy link
address.&rdquo; You now have the location of the data.</p>
<p class="block-text-western">Note: Since the publication of the book
the file naming and structure of the site above has changed, not just
once but several times. For that reason we could not directly link to
the data source as it keeps changing and breaking the links. To avoid
the problems we had to download and mirror the file ourselves to
ensure that the links work as we claim them to:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/data/SGD_features.tab">http://data.biostarhandbook.com/data/SGD_features.tab</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/data/SGD_features.README">http://data.biostarhandbook.com/data/SGD_features.README</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-obtain-a-data-file-that-is-online"></a>
18.3 How do I obtain a data file that is online?</h2>
<p class="first-paragraph">We can use either the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font>
command line tool (see <font color="#4f81bd"><a href="https://daniel.haxx.se/docs/curl-vs-wget.html">Curl
vs wget</a></font>) to download the content of URLs.</p>
<p>You may use the link above, but for simplicity and to avoid
potential errors outside of our control, we also host these files
ourselves; thus you can get them from the handbook data site.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/data/SGD_features.tab">http://data.biostarhandbook.com/data/SGD_features.tab</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/data/SGD_features.README">http://data.biostarhandbook.com/data/SGD_features.README</a></font></p>
</ul>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Create a work directory for a unix tutorial work.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
unixtut</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Switch to the new directory you will work in.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
unixtut</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get the data from SGD.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/data/SGD_features.tab </font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Also get the README file.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/data/SGD_features.README</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Quick check. Where are you and what have you got?</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font></p>
<p class="first-paragraph">Simple Unix tools let us answer a wide
variety of questions.</p>
<h3 class="western"><a name="what-are-flags-parameters"></a>18.3.1
What are flags (parameters)?</h3>
<p class="first-paragraph">A &ldquo;flag&rdquo; in Unix terminology
is a parameter added to the command. See for example</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></p>
<p class="first-paragraph">versus</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l</font></font></p>
<p class="first-paragraph">Typically flags make programs behave
differently or report their information in different ways.</p>
<h3 class="western"><a name="how-to-find-out-what-flags-are-available"></a>
18.3.2 How to find out what flags are available?</h3>
<p class="first-paragraph">You can use the manual to learn more about
a command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">man
ls</font></font></p>
<p class="first-paragraph">will produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NAME</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ls -- list directory contents</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SYNOPSIS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ls [-ABCFGHLOPRSTUW@abcdefghiklmnopqrstuwx1] [file
...]</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DESCRIPTION</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 For each operand that names a file of a type other than directory,
ls</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 displays its name as well as any requested, associated information. 
For</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 each operand that names a file of type directory, ls displays the
names</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 of files contained within that directory, as well as any requested,
asso-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ciated information.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l      (The lowercase letter ``ell''.)  List in long format.  (See</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        below.)  If the output is to a terminal, a total sum for all
the</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        file sizes is output on a line before the long listing.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
...</font></font></p>
<p class="first-paragraph">We can see above that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
takes the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-l</font></font>
parameter (letter ell, not to be confused with the number 1 as the
two can look the same depending on the fonts used by your screen).
The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-l</font></font>
parameter makes <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font>
report the long form.</p>
<p>You don&rsquo;t need to remember all flags, not even the majority
of them. Most of the time we use just a few, in some cases when you
find yourself needing some specific use of a tool you may want to
investigate the manual, perhaps the tool already supports the
behavior that you seek.</p>
<h3 class="western"><a name="what-if-the-tools-does-not-have-manual-page"></a>
18.3.3 What if the tools does not have manual page?</h3>
<p class="first-paragraph">Not all tools that come with Unix will
have a manual. For other software the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-h</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-help</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--help</font></font>
command will typically print instructions for their use.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">hisat2
--help</font></font></p>
<p class="first-paragraph">If the help flag is not recognized you
will need to find the manual for the software.</p>
<h3 class="western"><a name="what-are-flag-formats"></a>18.3.4 What
are flag formats?</h3>
<p class="first-paragraph">Traditionally Unix tools use two flag
forms:</p>
<ul>
	<li/>
<p class="compact">short form: single minus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>
	then one letter, like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g</font></font>,
	&rsquo;-v`</p>
	<li/>
<p class="compact">long form: double minus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--</font></font>
	then a word, like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--genome</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--verbosity</font></font></p>
</ul>
<p class="first-paragraph">In each case the parameter may stand alone
as a toggle (on/off) or may take additional values after the flag. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-p</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-p
100</font></font></p>
<p>Now some bioinformatics tools do not follow this tradition and use
a single <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>
character for both short and long options. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-genome</font></font>.</p>
<h3 class="western"><a name="is-using-flags-important"></a>18.3.5 Is
using flags important?</h3>
<p class="first-paragraph">Using flags is a essential to using Unix.
Bioinformatics tools typically take a large number of parameters
specified via flags. The correctness of the results critically
depends on the proper use of these parameters and flags.</p>
<h3 class="western"><a name="how-do-i-view-the-data-one-page-at-a-time"></a>
18.3.6 How do I view the data one page at a time?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">more
SGD_features.README</font></font></p>
<p class="first-paragraph">press &ldquo;space&rdquo; to move forward,
&ldquo;b&rdquo; to move backward, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">q</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ESC</font></font>
to exit.</p>
<h3 class="western"><a name="how-do-i-open-a-stream-from-a-data"></a>18.3.7
How do I open a stream from a data?</h3>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>
command concatenates one or more files and stars producing them on
the output stream:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab</font></font></p>
<p class="first-paragraph">The command above will print the contents
of the file to the current output stream which happens to be your
terminal window.</p>
<p>Tips:</p>
<ul>
	<li/>
<p class="compact">Use the manual <font face="Consolas, serif"><font size="2" style="font-size: 11pt">man</font></font>
	for more information on commands.</p>
	<li/>
<p class="compact">If you want to rerun a command you can use
	the up-arrow key to recall the previous entry.</p>
	<li/>
<p class="compact">You can press TAB to ask your shell to try
	to complete a file name. It is convenient! Always try to have the
	shell fill in the filename for you, that way you will never mistype
	these.</p>
</ul>
<h3 class="western"><a name="how-many-lines-does-the-file-have"></a>18.3.8
How many lines does the file have?</h3>
<p class="first-paragraph">You can pipe the output of you stream into
another program rather than the screen. Use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">|</font></font>
character to connect the programs. For example, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font>
program is the word counter.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font>
</p>
<p class="first-paragraph">prints the number of lines, words, and
characters in the stream:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16454
 425719 3264490</font></font></p>
<p class="first-paragraph">You can customize what the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font>
program does:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | wc -l</font></font></p>
<p class="first-paragraph">so that it will only print the number of
lines:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16454</font></font></p>
<p class="first-paragraph">Now many programs also operate on files
directly like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc
-l SGD_features.tab</font></font></p>
<p class="first-paragraph">The command above is fully equivalent to:</p>
<p>cat SGD_features.tab | wc -l</p>
<p>In general, it is a better option to open a stream with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>
then pipe the flow into the next program. Later you will see that it
is easier to design, build and understand more complex pipelines when
the data stream is opened at the beginning as a separate step.</p>
<p>Some people have strong negative feelings about this second form.
They, feel that it is redundant, it even has a name: <font color="#4f81bd"><a href="http://porkmail.org/era/unix/award.html">Useless
Use of Cat</a></font>. For them, it feels superfluous, and indeed it
is ever so slightly less efficient as it opens two processes instead
of just one.</p>
<p>The main problem with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc
-l SGD_features.tab</font></font> that it makes the process less
reusable if you wanted to run another tool with the same file you&rsquo;d
modify the beginning of the command, whereas when using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>
you can delete the last command and add a new one.</p>
<h3 class="western"><a name="how-does-the-file-start"></a>18.3.9 How
does the file start?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">S000002143
     ORF     Dubious YAL069W                 chromosome 1           
1       335     649     W               1996-07-31      1996-07-31   
  Dubious open reading frame; unlikely to encode a functional
protein, based on available experimental and comparative sequence
data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S000031098
     CDS                                     YAL069W         1      
335     649     W               1996-07-31      1996-07-31</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S000028594
     ORF     Dubious YAL068W-A                       chromosome 1    
       1       538     792     W               2003-07-29     
2003-07-29      Dubious open reading frame; unlikely to encode a
functional protein, based on available experimental and comparative
sequence data; identified by gene-trapping, microarray-based
expression analysis, and genome-wide homology searching</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S000031372
     CDS                                     YAL068W-A              
1       538     792     W               2003-07-29      2003-07-29</font></font></p>
<h3 class="western"><a name="which-lines-match-a-specific-pattern"></a>
18.3.10 Which lines match a specific pattern?</h3>
<p class="first-paragraph">Suppose you wanted to find information on
gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAL060W</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab  </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAL060W </font></font>
</p>
<h3 class="western"><a name="how-can-i-color-the-pattern-that-is-matched"></a>
18.3.11 How can I color the pattern that is matched?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAL060W --color=always </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<h3 class="western"><a name="how-can-we-tell-how-many-lines-do-not-match-a-given-pattern"></a>
18.3.12 How can we tell how many lines DO NOT match a given pattern?</h3>
<p class="first-paragraph">Adding the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-v</font></font>
flag to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
reverses the action of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
it shows the lines that do not match.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v Dubious </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font></p>
<h3 class="western"><a name="how-do-i-store-the-results-in-a-new-file"></a>
18.3.13 How do I store the results in a new file?</h3>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>
character is the redirection.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAL060W </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
match.tab</font></font></p>
<p class="first-paragraph">now check how many files do you have:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">match.tab</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SGD_features.README</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SGD_features.tab</font></font></p>
<h3 class="western"><a name="how-many-lines-in-the-file-match-the-word-gene"></a>
18.3.14 How many lines in the file match the word gene?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
gene </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font></p>
<p class="first-paragraph">Important: note how the above matches the
word gene anywhere on the line.</p>
<h3 class="western"><a name="how-do-i-select-genes"></a>18.3.15 How
do I select genes?</h3>
<p class="first-paragraph">It looks like this file uses the feature
type (column 2) ORF for protein-coding genes. You will need to cut
the second field (by tabs).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | head</font></font></p>
<p class="first-paragraph">Build your commands one step at a time,
always checking that you are on the right track:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | grep ORF | head</font></font></p>
<h3 class="western"><a name="how-many-genes-are-there"></a>18.3.16
How many genes are there?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | grep ORF | wc -l</font></font></p>
<h3 class="western"><a name="can-i-select-multiple-columns"></a>18.3.17
Can I select multiple columns?</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2,3,4 | grep ORF | head</font></font></p>
<p class="first-paragraph">What does this do?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2,3,4 | grep ORF | grep -v Dubious | wc -l</font></font></p>
<h2 class="western"><a name="how-many-feature-types-are-in-this-data"></a>
18.4 How many feature types are in this data?</h2>
<p class="first-paragraph">We are going to use this data a lot, so
place it into a separate file for now.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SGD_features.tab </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 2 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
types.txt</font></font></p>
<p class="first-paragraph">Sorting places identical consecutive
entries next to one another.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
types.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">Find unique words. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font>
command collapses consecutive identical words into one.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
types.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">Using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c</font></font>
flag to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font>
will not only collapse consecutive entries it will print their
counts.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
types.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">352
ARS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">196
ARS_consensus_sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 6 blocked_reading_frame</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7074
CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
16 centromere</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
16 centromere_DNA_Element_I</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
16 centromere_DNA_Element_II</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
16 centromere_DNA_Element_III</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 8 external_transcribed_spacer_region</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
24 five_prime_UTR_intron</font></font></p>
<p class="first-paragraph">To find out the most frequent unique
elements, we need to sort the output of uniq.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
types.txt | sort | uniq -c | sort -rn | head</font></font></p>
<p class="first-paragraph">it now prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">7074
CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">6604
ORF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
484 noncoding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
383 long_terminal_repeat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
377 intron</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
352 ARS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
299 tRNA_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
196 ARS_consensus_sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 91 transposable_element_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 77 snoRNA_gene</font></font></p>
<h2 class="western"><a name="the-single-most-useful-unix-pattern"></a>
18.5 The single most useful Unix pattern</h2>
<p class="first-paragraph">The pattern <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c | sort -rn</font></font> is perhaps the most useful simple
unix command pattern that you will ever learn.</p>
<p>Time and again we are surprised by just how many applications it
has, and how frequently problems can be solved by sorting, collapsing
identical values, then resorting by the collapsed counts.</p>
<p>The skill of using Unix is not just that of understanding the
commands themselves. It is more about recognizing when a pattern,
such as the one that we show above, is the solution to the problem
that you wish to solve. The easiest way to learn to apply these
patterns is by looking at how others solve problems, then adapting it
to your needs.</p>
<h2 class="western"><a name="the-sort-uniq-count-tool"></a>18.6 The
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count</font></font>
tool</h2>
<p class="first-paragraph">There is one fundamental flaw with the way
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c | sort -rn</font></font> pattern works. The output is space
delimited and padded. The spaces and padding make operating on the
results surprisingly challenging. To be consistent with the other
UNIX tools the output should be tab delimited yet for some obscure,
historical reason <font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq
-c</font></font> pads with spaces.</p>
<p>The good news is that the pattern is common and useful that the
entrez-direct package offers two commands that streamline the
&ldquo;sort-uniq-sort&rdquo; pattern. The commands are called:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count</font></font></p>
<p class="first-paragraph">and:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font></p>
<p class="first-paragraph">The above combine the utilities of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c</font></font> and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c | sort -rn</font></font> into single commands and will
produce tab delimited outputs. The two commands are equivalent:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Original pattern.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
types.txt | sort | uniq -c | sort -rn | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Improved pattern.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
types.txt | sort-uniq-count-rank | head</font></font></p>
<p class="first-paragraph">the latter produces tab delimited output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">7074
   CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">6604
   ORF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">484
noncoding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">383
long_terminal_repeat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">377
intron</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">352
ARS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">299
tRNA_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">196
ARS_consensus_sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">91
 transposable_element_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">77
 snoRNA_gene</font></font></p>
<p class="first-paragraph">If you forget what these new programs are
called type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font>
and double tap the TAB key. When I do that on my system it shows me
the following</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
                 sort-uniq-count-rank  sortbyname.sh</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count
      sortBed</font></font></p>
<h2 class="western"><a name="one-liners"></a>18.7 One-liners</h2>
<p class="first-paragraph">We could have also done the last step in
one single step to find out what are the most frequent unique types
of features in this file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | sort-uniq-count-rank | head</font></font></p>
<p class="first-paragraph">The commands above are what we will call a
pipeline, or a segment of a pipeline.</p>
<p>When you see a command like the one above, rebuild it yourself,
one step at a time from the left to right:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab |  head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 |  head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | sort | head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | sort-uniq-count | head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | sort-uniq-count-rank | head</font></font></p>
<p class="first-paragraph">When you understand each step, you have
mastered the pattern.</p>
<h2 class="western"><a name="tip-use-sort-uniq-count-rank-everywhere"></a>
18.8 Tip: use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font>
everywhere</h2>
<p class="first-paragraph">You may read other examples in other data
analysis resources that use the</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq | sort</font></font></p>
<p class="first-paragraph">pattern. In most cases your code will
become much simpler if you replace that with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font></p>
<p class="first-paragraph">but do remember that the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font>
program gets installed only when you install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">entrez-direct</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install entrez-direct</font></font></p>
<h2 class="western"><a name="how-does-sort-uniq-count-rank-work"></a>18.9
How does <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font>
work?</h2>
<p class="first-paragraph">It is actually a quite simple bash script
that uses some helpers to operate on the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c | sort</font></font> output. You can inspect the code with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
`which sort-uniq-count-rank`</font></font></p>
<p class="first-paragraph">on my system it prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!/bin/sh</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">flags=&quot;f&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">if
[ -n &quot;$*&quot; ]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">then</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 flags=`echo &quot; $*&quot; | sed 's/[^bfinrs]//g'`</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 if [ -z &quot;$flags&quot; ]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 then</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   flags=&quot;s&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 fi</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fi</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
&quot;-$flags&quot; |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq
-i -c |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">perl
-pe 's/\s*(\d+)\s(.+)/$1\t$2/' |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
-t &quot;$(printf '\t')&quot; -k 1,1nr -k &quot;2$flags&quot;</font></font></p>
<p class="first-paragraph">That&rsquo;s it. Now the above is
constitutes quite advanced shell scripting and you don&rsquo;t need
to worry if you can&rsquo;t quite understand it.</p>
<p>I am showing it here so that if you don&rsquo;t want to (or can&rsquo;t)
install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">entrez-direct</font></font>
you can also write the above into a script and run it that way.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="data-compression"></a>19 Data
compression</h1>
<p class="first-paragraph">This chapter deals with compressed files
that may be referred to as zip files, gzip files, tar files,
compressed files, archives.</p>
<h2 class="western"><a name="quick-summary"></a>19.1 Quick summary</h2>
<p class="first-paragraph">This at the beginning is just a reminder
of how the commands work if you forget them. We recommend that you
consult the full chapter at least once if you are unfamiliar with the
concepts.</p>
<h3 class="western"><a name="how-do-i-uncompress-a-tar.gz-archive"></a>
19.1.1 How do I uncompress a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar.gz</font></font>
archive?</h3>
<p class="first-paragraph">You e<font face="Consolas, serif"><font size="2" style="font-size: 11pt">x</font></font>tract,
a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">z</font></font>ipped,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">f</font></font>ile
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">v</font></font>erbose
output that shows you what is going on:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
xzfv archive.tar.gz</font></font></p>
<p class="first-paragraph">unpacks the file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">archive.tar.gz</font></font>
and you&rsquo;ll have both the original file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">archive.tar.gz</font></font>
and what was inside it. If there were directories inside the archive
these will be created.</p>
<h3 class="western"><a name="how-do-i-create-a-compressed-archive"></a>
19.1.2 How do I create a compressed archive?</h3>
<p class="first-paragraph">You <font face="Consolas, serif"><font size="2" style="font-size: 11pt">c</font></font>reate,
a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">z</font></font>ipped,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">f</font></font>ile
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">v</font></font>erbose
output that shows you what is going on:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
czfv archive.tar.gz path/to/file1 path/to/file2</font></font></p>
<p class="first-paragraph">produces the file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">archive.tar.gz</font></font></p>
<h2 class="western"><a name="what-is-a-compressed-format"></a>19.2
What is a compressed format?</h2>
<p class="first-paragraph">Compressed formats reduce file disk space
requirements. A compressed file contains the same information as an
uncompressed but has a smaller size. On the downside, it needs to be
decompressed to access its content.</p>
<p>Compression (creating a compressed file) requires substantially
more computational resources than decompression. It is always much
faster (may even be an order of magnitude faster) to restore a file
from a compressed form than to create that condensed form.</p>
<p>Some compressed formats (like the .gz format) can be concatenated
(or appended to) without being decompressed. For example</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
file1.gz file2.g file3.gz &gt; bigfile.gz</font></font></p>
<p class="first-paragraph">Will create a valid compressed file that
contains all three files. When the files are large, the time to
compress may be hours (or more) thus concatenating files as done
above can quite useful as it avoids the most time-consuming step.</p>
<h2 class="western"><a name="what-objects-may-be-compressed"></a>19.3
What objects may be compressed?</h2>
<p class="first-paragraph">We can compress single files or entire
directories. Compressed directories are often called &ldquo;archives&rdquo;.</p>
<ul>
	<li/>
<p class="compact">A &ldquo;compressed file&rdquo; is a single
	file reduced in size. The name may end with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.gz</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bz</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bz2</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.zip</font></font>
	or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xz</font></font></p>
	<li/>
<p class="compact">A &ldquo;compressed archive&rdquo; is a
	folder containing multiple files combined, and then compressed. The
	name may end with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">zip</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.tar.gz</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.tar.bz2</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar.xz</font></font>.</p>
</ul>
<h2 class="western"><a name="what-are-some-common-compression-formats"></a>
19.4 What are some common compression formats?</h2>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ZIP</font></font>,
	extension <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.zip</font></font>,
	program names are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">zip/unzip</font></font>.
	Available on most platforms.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GZIP</font></font>
	extension <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.gz</font></font>,
	program names are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip/gunzip</font></font>.
	The most common compression format on Unix-like systems.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BZIP2</font></font>
	extension <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bz/.bz2</font></font>,
	program names are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bzip2/bunzip2</font></font>.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XZ</font></font>
	extension <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.xz</font></font>.
	A more recent invention. Technically <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bzip2</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xz</font></font>
	are improvements over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip</font></font>.
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip</font></font>
	is still quite prevalent for historical reasons and because it
	requires less memory.</p>
</ul>
<p class="first-paragraph">In the majority of cases, bioinformatics
data is distributed in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gz</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar.gz</font></font>
formats.</p>
<h2 class="western"><a name="is-there-a-bioinformatics-specific-compression-format"></a>
19.5 Is there a bioinformatics-specific compression format?</h2>
<p class="first-paragraph">Yes. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
(Blocked GNU Zip Format) is a variant of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GZIP</font></font>
format that allows random access to its content. BCF, BAM, and TABIX
formats are in BGZIP compressed formats.</p>
<p>A <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
file can be decompressed with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip</font></font>
but only the specialized <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
command can create a file in the Blocked GNU Zip Format file. The
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
utility is installed when you install the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">htslib</font></font>
package.</p>
<p>Most bgzipped files carry the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gz</font></font>
extension, thus you can&rsquo;t always tell which compression was
used. You would need to create <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
files yourself when compressing VCF and TABIX files. For all other
cases <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip</font></font>
will suffice.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get a sequence file.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-db nuccore -format fasta -id AF086833 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Create bgzipped file.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa</font></font></p>
<p class="first-paragraph">More about <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>:
<font color="#4f81bd"><a href="http://www.htslib.org/doc/tabix.html">http://www.htslib.org/doc/tabix.html</a></font></p>
<h2 class="western"><a name="how-do-i-compress-or-uncompress-a-file"></a>
19.6 How do I compress or uncompress a file?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get a sequence file.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-db nuccore -format fasta -id AF086833 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Compress with gzip.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Creates the file AF086833.fa.gz.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Some tools can operate on GZ files without unpacking them. </i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Note: the tool is called gzcat on macOS</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzcat</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa.gz  </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Use zcat on Linux</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
zcat AF086833.fa.gz  | head</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Uncompress the file. Creates  AF086833.fa.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa.gz</font></font></p>
<h2 class="western"><a name="how-do-i-compress-or-uncompress-multiple-files"></a>
19.7 How do I compress or uncompress multiple files?</h2>
<p class="first-paragraph">Zip-files may contain one or multiple
files. Not knowing beforehand what an archive contains files or
directories can be confusing and annoying in that you can&rsquo;t
tell from the file name alone whether it is a single file or many
files.</p>
<p>In the Unix world the most commonly used approach is to create a
so-called &ldquo;tape archive&rdquo; even though it does not
typically involve tape anymore. The resulting file is called a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font>
file (Tape Archive) file. The extension for gzip-compressed tap
archive files is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.tar.gz</font></font>.</p>
<p>Suppose you had two files and wanted to create an archive from
them:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get two files</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-db nuccore -format fasta -id AF086833 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-db nuccore -format gb -id AF086833 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.gb</font></font></p>
<p class="first-paragraph">The archive creation command works by
subcommands that labels by the actions that it performs: create will
be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">c</font></font>
, a file will be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">f</font></font>
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
cf  archive-name  files-to-be-compressed</font></font></p>
<p class="first-paragraph">Thus if we want to create <font face="Consolas, serif"><font size="2" style="font-size: 11pt">c</font></font>,
a compressed <font face="Consolas, serif"><font size="2" style="font-size: 11pt">z</font></font>,
file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">f</font></font>,
in verbose <font face="Consolas, serif"><font size="2" style="font-size: 11pt">v</font></font>
mode and name the archive <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sequences.tar.gz</font></font>
we would use the following construct:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
czfv sequences.tar.gz AF086833.fa AF086833.gb</font></font></p>
<p class="first-paragraph">The first name listed on the command
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sequences.tar.gz</font></font>
is the file to be created, the other two are the files that should be
added, equivalent to:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
czvf sequences.tar.gz AF086833.*</font></font></p>
<h4 class="western"><a name="warning-a-super-annoying-behavior"></a>19.7.0.1
Warning, a super annoying behavior!</h4>
<p class="first-paragraph">If you accidentally put the file that you
wanted to compress as the first to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font>,
then it will be destroyed as tar will try to create that file!</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
SUPER ANNOYING ERROR! Destroys the file you wanted to archive!</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
czvf  AF086833.fa sequences.tar.gz</font></font></p>
<p class="first-paragraph">Unless we have very few files, adding them
into an archive is not a good way to package files &ndash; upon
unpacking, we may not remember which files were in the archive to
begin with.</p>
<p>The recommended practice to create archives is to put all files
into a subdirectory then compress that entire directory:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sequences</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
AF086833.* sequences/</font></font></p>
<p class="first-paragraph">then compress that entire directory like
so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
czvf sequences.tar.gz sequences/*</font></font></p>
<p class="first-paragraph">The added advantage is that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font>
will not overwrite a directory as a file even if you list it in the
wrong order, thus the super annoying behavior is not an issue, you&rsquo;ll
just get an error that it cannot create the archive.</p>
<h2 class="western"><a name="what-is-a-tarbomb"></a>19.8 What is a
tarbomb?</h2>
<p class="first-paragraph">A tarbomb is a tar file, usually created
by an unwitting user, that contains a a vast number of files.</p>
<p>Here is an example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p tarbomb</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
tarbomb</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O http://data.biostarhandbook.com/data/tarbomb.tar.gz</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
xzf tarbomb.tar.gz</font></font></p>
<p class="first-paragraph">Upon unpacking it &ldquo;explodes&rdquo;
all over your directory, perhaps instead of a bomb we should call it
a &ldquo;flashbang&rdquo; grenade. It is quite disorienting once you
look at the directory and you can&rsquo;t recognize what is was yours
and what was unpacked.</p>
<p>Disarming a tarbomb:</p>
<ol>
	<li/>
<p class="compact">First, view the content of an archive:</p>
</ol>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
tzvf tarbomb.tar.gz</font></font></p>
<ol>
	<li/>
<p class="compact">If it is a tarbomb move it to a separate
	directory and unpack it there.</p>
</ol>
<p class="first-paragraph">Usually it is a good habit to always make
a new directory for tar files you don&rsquo;t know about and unpack
them there. If you make that a habit, you&rsquo;ll never be
surprised, it is like defensive driving, always expect the worst.</p>
<h2 class="western"><a name="how-do-we-use-tar-again"></a>19.9 How do
we use tar again?</h2>
<p class="first-paragraph">You may forget which <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar</font></font>
flags you need. Come back to this chapter or Google it quickly.</p>
<p>Take solace that it is not just you. See <font color="#4f81bd"><a href="http://xkcd.com/1168/">http://xkcd.com/1168/</a></font>
for a different kind of tarbomb:</p>
<p><img src="0000_html_5c0325d96f525f73.png" name="Image44" align="bottom" hspace="12" width="560" height="180" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-write-better-scripts"></a>20 How
to write better scripts</h1>
<p class="first-paragraph">The content of this chapter has been moved
into a separate, standalone book:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/scripting/index.html">The
	Art of Bioinformatics Scripting</a></font></p>
</ul>
<p class="first-paragraph">If you are new to the field and primarily
interested in learning more about bioinformatics then skip this
content for now. It covers more advanced concepts that will make
sense later, once you better understand what needs to be done.</p>
<p><a href="https://www.biostarhandbook.com/books/scripting/index.html"></a><img src="0000_html_87c2d01e4d0c837c.png" name="Image45" align="bottom" hspace="12" width="560" height="145" border="0"/>
<a href="https://www.biostarhandbook.com/books/scripting/index.html"></a></p>
<p>We only list the content here as it belongs to the same topic of
Unix programming. Later, you will be able to locate this chapter more
easily.</p>
<p>For now continue on with the next chapter: <font color="#4f81bd"><a href="#ch021.xhtml#survival-guide">Statistics
survival guide</a></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="statistics-survival-guide"></a>21
Statistics survival guide</h1>
<p class="first-paragraph">In this short chapter, we&rsquo;ll be
providing summaries of various statistical concepts. Remember that
each concept that we cover here in a single paragraph may have entire
books dedicated to them.</p>
<p>In the <font color="#4f81bd"><a href="#ch022.xhtml#bio-statistics">next
chapter</a></font> we&rsquo;ll talk about how &ldquo;nobody is using
the right words&rdquo; when it comes to statistics. Surprising as it
might sound, we found that even statisticians routinely misuse the
terms. Even the &ldquo;inventors&rdquo; of the various measures
continue to disagree with one another as to what each measure
&ldquo;precisely&rdquo; means. Recognizing our fallibility, we&rsquo;ll
try to make our mistakes and inconsistencies as small as possible.</p>
<p>The code in this chapter is written in and should be run in R.</p>
<h2 class="western"><a name="the-data"></a>21.1 The data</h2>
<p class="first-paragraph">Suppose we have data with measurements for
two conditions <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">P</font></font>,
and each state is measured three times. A section of the table might
look like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Name
        C1        C2        C3        P1       P2      P3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53
    1887.7       8.1    1799.6     800.2    785.3   800.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEZ6L
     9.9        8.9       7.7     400.1    389.3   403.9</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MICALL1
  236.4       8.4     220.9     127.8    156.3   140.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The data has many more rows (assume there
are 10,000 rows) , here we show the first few lines.</p>
<p>Take the first row above that corresponds to gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font>,
condition <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
has the numbers <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1887.7,
8.1, 1799.6</font></font>, whereas condition <font face="Consolas, serif"><font size="2" style="font-size: 11pt">P</font></font>
the values are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">800.2,
 785.3,  800.4</font></font>.</p>
<p>We will use the data above to demonstrate specific concepts.</p>
<h2 class="western"><a name="pairwise-comparisons"></a>21.2 Pairwise
comparisons</h2>
<p class="first-paragraph">The most commonly used statistical tests
in Biology are the so-called pairwise comparisons. In the pairwise
tests, we compare not just single numbers but repeated measurements
of the same condition, a process called <i>replication</i>.
Typically, the two groups of numbers are compared by having their
difference formed; then, a formula is computed to express the
likelihood that the observed difference due to random chance. That
likelihood is <a href="https://thewire.in/the-sciences/why-the-way-we-use-statistical-significance-has-created-a-crisis-in-science"><font color="#4f81bd">the
infamous </font><font color="#4f81bd"><b>p-value</b></font></a>.</p>
<p>Typically the formulas that we apply aim to compare the variation
between replicates to the variation across replicates (across
conditions). Pairwise comparisons are deemed statistically
significant when the variation <i>across</i> conditions is
sufficiently larger than the variation <i>between</i> the replicates.</p>
<p>The challenge is that selecting the appropriate formula and
deciding &ldquo;sufficiently large&rdquo; is both more subjective
than we care to admit.</p>
<h2 class="western"><a name="normalization"></a>21.3 Normalization</h2>
<p class="first-paragraph">To perform any comparison, the values that
we measure need to reflect the same magnitudes. In many biological
experiments, however, the sample preparation cannot be controlled at
the levels required for an accurate comparison. For example, having
more cells of one kind vs.&nbsp;another, having a more successful
reagent reaction could lead to differences even when the biological
samples are identical.</p>
<p>Normalization is the mathematical process of rescaling our data to
make the numbers follow the same scale. For example, if we produce
measures for three samples as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
  2  10</font></font></p>
<p class="first-paragraph">and have the means to detect that the
second and third sample had been affected by a scale factor of 2;
then, after normalizing (rescaling) we would get:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
 1  5</font></font></p>
<p class="first-paragraph">Think about as if you were rescaling an
image taken from different vantage points so that objects in the
picture have the same size in all images.</p>
<p><i>Normalization methods</i> represent strategies of figuring out
what the rescaling factor should be, and how they should be applied
to each measure.</p>
<h2 class="western"><a name="effect-size"></a>21.4 Effect size</h2>
<p class="first-paragraph">Effect size is the quantitative measure of
the magnitude of a phenomenon. The values corresponding to the effect
size may be standardized with a well-defined meaning, for example,
correlation, percent, odd ratios, or non-standardized computed as the
difference between group means.</p>
<p>If we were to compare the expression levels for gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
= c(1887.7, 8.1, 1799.6)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
= c(800.2,  785.3,  800.4)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">print(
mean(x) - mean(y))</font></font></p>
<p class="first-paragraph">produces the effect size:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">453</font></font></p>
<p class="first-paragraph">See also the blog post: <font color="#4f81bd"><a href="http://www.argmin.net/2021/09/13/effect-size/">Effect
size is significantly more important than statistical significance.</a></font></p>
<h2 class="western"><a name="p-values"></a>21.5 P-Values</h2>
<p class="first-paragraph">A p-value represents the <i>probability</i>
for the observed <i>effect size</i> to be a product of random chance.</p>
<p>There are many formulas to compute p-values, each with various
requirements and tradeoffs. If we wanted to use a t-test it would be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
= c(1887.7, 8.1, 1799.6)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
= c(800.2,  785.3,  800.4)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">t
= t.test(x,y)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">print
(t$p.value)</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.5499189</font></font></p>
<p class="first-paragraph">Thus the test above tells us that the
likelihood of observing an effect size <font face="Consolas, serif"><font size="2" style="font-size: 11pt">453</font></font>
by chance when using the t-test is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.54</font></font>
(54%)</p>
<h2 class="western"><a name="statistical-models"></a>21.6 Statistical
models</h2>
<p class="first-paragraph">The requirements of the statistical
formula (model) need to match the properties of the data. Figuring
out which test is appropriate for a problem at hand can be a
challenge. But you don&rsquo;t need to be a statistician, nor should
you enroll in a statistics course right away.</p>
<p>Read related, published papers, identify down authors&rsquo;
rationale for choosing a specific test. Do your research on what you
learn. Running the test is usually a lot easier.</p>
<p>Here we run the Mann-Whitney-Wilcoxon test on the same data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">w
= wilcox.test(x, y)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">print
(w$p.value)</font></font></p>
<p class="first-paragraph">Prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.7</font></font></p>
<p class="first-paragraph">Thus the test above tells us that the
likelihood of observing an effect size <font face="Consolas, serif"><font size="2" style="font-size: 11pt">453</font></font>
by chance when using the Wilcoxon-test <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.7</font></font>
(70%)</p>
<p>The same data will usually produce substantially different results
with different tests.</p>
<h2 class="western"><a name="statistical-power"></a>21.7 Statistical
power</h2>
<p class="first-paragraph">Statistical power reflects the ability of
a test to produce the &ldquo;correct&rdquo; prediction. Statistical
power typically increases when</p>
<ul>
	<li/>
<p class="compact">There is more data.</p>
	<li/>
<p class="compact">The model is more restrictive.</p>
	<li/>
<p class="compact">The effect size is larger.</p>
</ul>
<p class="first-paragraph">Different formulas make different
assumptions on the data. A Wilcoxon is &ldquo;weaker&rdquo; than a
t-test because it is more generic. A Wilcoxon test works even on data
that a t-test does not. But there is no need to use Wilcoxon when a
t-test is applicable. Note that it would not be &ldquo;incorrect&rdquo;
to use a Wilcoxon test; it is just that it may miss cases that the
t-test does not.</p>
<h2 class="western"><a name="significance-threshold"></a>21.8
Significance threshold</h2>
<p class="first-paragraph">A predetermined value <font face="Consolas, serif"><font size="2" style="font-size: 11pt">x</font></font>
chosen by the scientist so that p-values smaller than that value are
called *statistically significant&rdquo;.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">p
&lt; x</font></font></p>
<p class="first-paragraph">Typical values are: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01</font></font></p>
<p>Do note, however, just how arbitrary it all is. Yet the word
&ldquo;statistical significance&rdquo; feels so much more meaningful
than that.</p>
<h2 class="western"><a name="adjusted-p-values"></a>21.9 Adjusted
p-values</h2>
<p class="first-paragraph">An <i>adjusted p-value</i> is a correction
applied to a p-value to account for the repeated chances we might
have given it to become smaller than what it is.</p>
<p>Suppose we repeat the experiment, this time the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font>
gene produces the following values:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
= c(1887.7, 1298.1, 1799.6)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
= c(800.2,  785.3,  800.4)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">t
= t.test(x,y)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">print
(t$p.value)</font></font></p>
<p class="first-paragraph">The p-value now is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.04199959</font></font></p>
<p class="first-paragraph">Will this pass a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font>
significance threshold? If it were the only experiment, then yes. But
together with the other experiment no. To account for both attempts,
we need to adjust the value like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">p.adjust(t$p.value,
n=2)</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.0839</font></font></p>
<p class="first-paragraph">The adjusted p-value is roughly <font face="Consolas, serif"><font size="2" style="font-size: 11pt">p
* n</font></font> (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">n</font></font>
the number of comparisons). There are different
methods/formulas/methods to adjust a p-value.</p>
<p>We can adjust p-values individually, just by knowing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">n</font></font>
because each p-value is computed independently of all other p-values.</p>
<p>Note: Even though tools typically report an adjusted (corrected)
p-value, conceptually, what should be adjusted is the significance
threshold. For <font face="Consolas, serif"><font size="2" style="font-size: 11pt">n=10</font></font>,
we ought to reduce a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font>
significance level to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.005</font></font>
when filtering for statistically significant differences. Due to the
way p-values are typically used and for clarity&rsquo;s sake, in most
cases, tools report the p-value multiplied by the correction factor
as a new column so that the same significance threshold of say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font>
can be used for both the adjusted and the unadjusted p-values.</p>
<h2 class="western"><a name="multiple-comparisons"></a>21.10 Multiple
comparisons</h2>
<p class="first-paragraph">Similarly to when we did an experiment
twice and had to correct for a p-value across the two tries when we
instruct a method produce all p-values then, we need to account for
all possible chances that could lead to smaller values by accident.</p>
<p>In this case, the number of tries <font face="Consolas, serif"><font size="2" style="font-size: 11pt">n</font></font>
will be the number of total comparisons, which, if you have say
10,000 genes, the number of comparisons is 10,000.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">p.adjust(t$p.value,
n=10000)</font></font></p>
<p class="first-paragraph">Note: typically, you would not need to
list the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">n</font></font>
in the function call. Passing a vector of p-values will compute the
adjustments from the size of the list.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">p
= c (0.1, 0.2, 0.3, 0.4, 0.5)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">p.adjust(p)</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.5
 0.8   0.9   0.9   0.9</font></font></p>
<p class="first-paragraph">A common correction method is the
Bonferroni correction:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Bonferroni_correction">https://en.wikipedia.org/wiki/Bonferroni_correction</a></font></p>
</ul>
<h2 class="western"><a name="false-discovery-rate"></a>21.11 False
Discovery Rate</h2>
<p class="first-paragraph">As it turns out, p-values are a bit too
strict. The p-value reflects the probability that the measurement is
due to random chance. Since each p-value is computed independently of
the others, then when we adjust, we equally penalize for all
comparisons, even those that were never close to make it into
statistical significance.</p>
<p>Scientists prefer a more &ldquo;generous/sane&rdquo; estimate
where they quantify how many errors are allowed in a result list.</p>
<p>The false discovery rate is also an adjustment of p-values, but
one where the order of p-values (smallest to largest) must be
maintained. Whereas in the original definition, a p-value on a given
row is independent of all other p-values, when FDR corrected p-values
only make sense when listed in decreasing order. The FDR reflects
what percent of p-values up to that point (going down on the list)
are false positives.</p>
<p>Since FDR is, in a way, computing the probability of
probabilities, it only applies when there are many p-values. You
can&rsquo;t FDR correct one or even ten p-values.</p>
<p>Note in the image below the interplay between p-value, p-adjusted,
FDR, and false positives.</p>
<p><img src="0000_html_7be7a5bcda7d7966.png" name="Image46" align="bottom" hspace="12" width="490" height="1032" border="0"/>
</p>
<h2 class="western"><a name="confidence-intervals"></a>21.12
Confidence intervals</h2>
<p class="first-paragraph">A confidence interval is a range of
plausible values say, for example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[1,
10]</font></font> with an associated probability that the true value
is within the range. A typical confidence level is 95%. The
interpretation is that if we were to repeat the same measurement many
times, 95% of the time the measurement would fall within the range.</p>
<p>Common misconceptions:</p>
<ul>
	<li/>
<p class="compact">Confidence intervals cannot be used to
	state anything about values outside of the intervals.</p>
	<li/>
<p class="compact">Confidence intervals are not error bars, do
	not represent the accuracy nor the precision of the measurements.</p>
	<li/>
<p class="compact">Values towards the middle of the interval
	are not more reliable than those toward the limits.</p>
</ul>
<h2 class="western"><a name="how-to-choose-the-right-statistical-test"></a>
21.13 How to choose the right statistical test?</h2>
<p class="first-paragraph">Unless you work as a statistician, closely
following the field is onerously challenging. Statisticians
continuously work on figuring out various tradeoffs, where imposing
certain limitations will increase the test&rsquo;s power.</p>
<p>Unfortunately for a life scientist, understanding what the
tradeoffs consist of and whether a method applies to a given dataset
is challenging.</p>
<p>The most crucial tool in your arsenal is common sense. Let&rsquo;s
go back to our original data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Name
        C1        C2        C3        P1       P2      P3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53
    1887.7       8.1    1799.6     800.2    501.3   800.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEZ6L
     9.9        8.9       7.7     400.1    389.3   403.9</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MICALL1
  236.4       8.4     220.9     127.8    156.3   140.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Do you see anything odd with it? Do you
see how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
seems always to be around <font face="Consolas, serif"><font size="2" style="font-size: 11pt">8</font></font>.
That is a little odd. Will our tests be correct when we have such an
outlier? Not so clear at all. And there is no statistical test that
will direcly indicate this oddity of our data; you need to learn and
recognize it yourself.</p>
<p>Of course we would need to look at more than three rows to make
that determination, but it is a mindset that we need to maintain. It
is like defensive driving, always be on the lookout for unexpected
and odd behaviors. Here sample <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
seems to be always around 8, in addition when compared to samples <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C1</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
it is the outlier for both <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MICALL1</font></font>.
At this point, it is a little suspicious, it could be invalid data.</p>
<h2 class="western"><a name="statistical-robustness"></a>21.14
Statistical robustness</h2>
<p class="first-paragraph">The robustness of a statistical test
attempts to capture how well a statistical test operates even when
the data does not follow the assumptions that the mathematical model
requires.</p>
<p>It is a tricky concept that needs very close supervision. Any
tests that you apply can go completely wrong, depending on the
characteristics of the data. Above, if the column for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
were an artifact, none of the statistical tests would be valid even -
except, of course, when the gene expression was close to 8. Note how
a broken test can still produce &ldquo;correct&rdquo; predictions.</p>
<h2 class="western"><a name="outliers-data-salvage"></a>21.15
Outliers, data salvage</h2>
<p class="first-paragraph">If there is one universal lesson I have
learned over the years is that major, unexpected problems are
surprisingly common. You think it only happens to you because you are
unlucky. In reality, odd and weird stuff happens all the time.
Samples are mislabeled, instruments go bad, reagents fail, processes
are inconsistent, etc. Many times it impossible to redo an
experiment; samples are irreplaceable, costs are exorbitant.</p>
<p>The law of big numbers is conspiring against you. While each
problem has a low probability of occurring, there are so many ways
things can go sideways that, in the end, when nothing goes wrong is
the exception.</p>
<p>Learn to recognize problems early on. Above <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
looks odd. If more rows show the same behavior, if the distribution
of the values in column <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font>
does not match the other columns, it may be invalid data. Perhaps the
entire column would need to be dropped. If you choose to do so,
document and state the rationale in your writeups. Note what happens
once we drop column <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C2</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
= c(1887.7, 1799.6)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
= c(800.2,  785.3,  800.4)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">t
= t.test(x,y)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">print
(t$p.value)</font></font></p>
<p class="first-paragraph">Look, Ma!, my p-values are nice and small:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02492168</font></font></p>
<h2 class="western"><a name="p-hacking-and-harking"></a>21.16
P-hacking and HARKing</h2>
<p class="first-paragraph">Above we dropped a column and made our
p-values more in line with our expectations. Is that an acceptable
scientific practice? Should we delete another row that does not look
right? Where do we draw the line? We could have also kept trying
different hypotheses until one &ldquo;worked out&rdquo;. Repeatedly
trying to validate different hypotheses is called &ldquo;HARKing&rdquo;:
<i>Hypothesising After Results are Known</i>.</p>
<p>The most recent high profile HARKing case was that of <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Brian_Wansink">Brian
Wansink</a></font>, for an overview of what happened see <font color="#4f81bd"><a href="https://www.vox.com/science-and-health/2018/9/19/17879102/brian-wansink-cornell-food-brand-lab-retractions-jama">A
top Cornell food researcher has had 15 studies retracted</a></font>.
You can draw your own conclusions. From our point of view, the case
of Dr.&nbsp;Wansink shines a light into uncomfortably dark corners of
science. We note a lack of a basic understanding of proper analysis
practices by a seemingly successful scientist, full professor, and a
&ldquo;John S. Dyson Endowed Chair in the Applied Economics and
Management Department&rdquo; at an Ivy League institution. At the
same time, the scientific community at large is also to blame, the
eagerness to accept papers dealing with click-baity and catchy
titles, and a lack of professionalism when reviewing manuscripts. Had
it not been for <font color="#4f81bd"><a href="http://www.timvanderzee.com/the-wansink-dossier-an-overview/">a
arrogant blog post</a></font> that riled up several data sleuths,
Dr.&nbsp;Wansink would still be active, endlessly churning out
headlines-worthy publications.</p>
<p>In my observation, everyone HARKs all the time. We (scientists)
all do it. We try to make sense with incomplete information trying to
capture previously unknown complex phenomena. We have to revise and
continuously re-evaluate what we see; the results never fully support
nor contradict what we seek. The outlook is always a bit hazy.
Cutting through all that requires going beyond what is &ldquo;right&rdquo;
or &ldquo;correct&rdquo;. Strictly sticking to mathematics does not
work, because life itself is not following statistics.</p>
<p>What separates &ldquo;misconduct&rdquo; level HARKing from valid
science is what you do next, once you HARKed:</p>
<ol>
	<li/>
<p class="compact">Are you transparent about what happened?</p>
	<li/>
<p class="compact">Are you seeking/exploring alternative
	explanations?</p>
	<li/>
<p class="compact">Are you validating what you came up with
	using another, independent method?</p>
</ol>
<p class="first-paragraph">HARKing is terrible if it is the only
evidence. If you p-hack your way to an explanation, then you redo the
experiment and find the same answer; the p-hack worked! It helped you
understand what is happening. Another way to say it: p-hacking is OK,
as long as the only person you might be fooling is yourself.</p>
<h2 class="western"><a name="meta-analysis"></a>21.17 Meta-analysis</h2>
<p class="first-paragraph">Meta-analysis is the process of combining
previous results, where individually, some results may not have
passed the required statistical significance, but all together may
provide additional information. For example, while studying gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font>
under two conditions, the changes that five different scientists
might find could be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">difference
      p-value</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  0.5            0.23</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  0.3            0.89</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  1.1            0.12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  0.7            0.66</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  0.4            0.34</font></font></p>
<p class="first-paragraph">None of the p-values for effect sizes
above passes a typical statistical threshold. Could we still conclude
that there is a small positive effect? That is what meta-analysis
does.</p>
<p>We could say that, if the change is due to random chance, then
sometimes it should be positive, sometimes negative. The fact that we
saw positive change five times is unusual, it is the same chance as
seeing five tails in a row. The outcome is uncommon, 2^5, about 1 in
32, a probability of 0.03.</p>
<p>The challenge is ensuring that those numbers above are indeed
comparable.</p>
<h2 class="western"><a name="how-to-make-big-numbers-work-for-you"></a>
21.18 How to make big numbers work for you</h2>
<p class="first-paragraph">The rule is that checking for the validity
of an expected outcome has much larger statistical power than asking
a method to produce a result.</p>
<p>For example, if you wanted to verify a starting hypothesis, that a
particular gene is differentially expressed under some conditions, no
adjustments of p-value are needed. The big numbers work for you. The
more other candidates there are, the less likely it is that your gene
of interest is the one that is presented by chance.</p>
<p>On the other hand, if you ask which genes differentially express,
letting the method come up with answers, the statistical power is
reduced proportionally with the number of possible candidates.</p>
<h2 class="western"><a name="should-i-take-a-graduate-course-in-statistics"></a>
21.19 Should I take a graduate course in statistics</h2>
<p class="first-paragraph">In general, the answer is no - unless the
course directly covers the type of data you need to analyze or that
you are passionate about the field.</p>
<p>Statistics is a gigantic field of incredible complexity; taking a
&ldquo;statistics course&rdquo; does not guarantee that you&rsquo;ll
be better at analyzing your data. It is possible, even likely, that
you&rsquo;ll learn very little about what you need.</p>
<p>I am not saying that you should not seek training.</p>
<p>There is a wealth of outstanding quality information on statistics
online, books, training materials, overview papers, etc. You can
become well versed, confident, and knowledgeable by just using these.
What you need is not that complicated in the end.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="statistics-in-biology"></a>22 Statistics
in Biology</h1>
<p class="first-paragraph">There is no shortage of books and guides
on statistical analysis, as you read them you&rsquo;ll see
beautifully crafted plans, sitting on strong mathematical
foundations. Yet as soon as you start to apply the same techniques on
your data, you&rsquo;ll find that the original assumptions and
promises never quite fit as you thought would. No matter what you
try, you&rsquo;ll grow increasingly uncomfortable, uncertain whether
you are about to toss the proverbial baby with the bathwater.</p>
<p>In this section we expand on some concepts discussed in <font color="#4f81bd"><a href="#ch021.xhtml#survival-guide">Statistics
survival guide</a></font></p>
<h2 class="western"><a name="what-is-statistics-anyway"></a>22.1 What
is statistics anyway?</h2>
<p class="first-paragraph">Our favorite summary for statistics comes
from the blog post <font color="#4f81bd"><a href="http://madhadron.com/posts/2016-01-15-aspects_of_statistics.html">The
four aspects of statistics</a></font> where Frederick J. Ross writes:</p>
<p class="block-text-western">Statistics resembles the apocryphal
elephant being examined by blind men. Each person uses, and often
only knows, a particular set of statistical tools, and, when passing
on their knowledge, does not have a picture to impart of the general
structure of statistics. Yet that structure consists of only four
pieces: planning experiments and trials; exploring patterns and
structures in the resulting data; making reproducible inferences from
that data; and designing the experience of interacting with the
results of an analysis. These parts are known in the field as</p>
<ul>
	<li/>
<p class="block-text-western">design of experiments</p>
	<li/>
<p class="block-text-western">exploratory data analysis</p>
	<li/>
<p class="block-text-western">inference</p>
	<li/>
<p class="block-text-western">visualization</p>
</ul>
<p class="block-text-western">Sadly, this basic structure of
statistics doesn&rsquo;t seem to be written down anywhere,
particularly not in books accessible to the beginner.</p>
<p class="first-paragraph">read more on each point <font color="#4f81bd"><a href="http://madhadron.com/posts/2016-01-15-aspects_of_statistics.html">on
the blog</a></font>.</p>
<h2 class="western"><a name="why-does-statistics-play-a-role-bioinformatics"></a>
22.2 Why does statistics play a role bioinformatics?</h2>
<p class="first-paragraph">Statistics gives us tools that can help us
assign a level of confidence or a degree of uncertainty to each
estimate that we make.</p>
<p>Unfortunately most statistical measures are inherently tied to
effect size, that is large effects are more likely to be found than
smaller effects. Yet there is no reason to believe that large
phenotipic differences would also need to large causative factors.
Thus many results are biased towards large effects aka &ldquo;low
hanging fruit&rdquo;.</p>
<h2 class="western"><a name="when-do-i-need-to-make-use-of-statistics"></a>
22.3 When do I need to make use of statistics?</h2>
<p class="first-paragraph">You will need to think about statistics
first when designing the experiment. In this stage you have to
enumerate the goals and parameters of the experiment. Consulting with
a statistican, if that option is available, is obviously a good
choice.</p>
<p>The most important advice I would give is to not be too ambitious
(greedy?) with the goals of the experiment. In my experience projects
that try too cover too many ideas at once, test multiple genotypes,
multiple conditions, various interactions, multiple time points and
so on end up with far less reliable results than more focused
experiments.</p>
<p>It is akin to the joke: <i>If you have one clock you know what the
time is, if you have ten clocks you never know which one is right.</i></p>
<p>The second stage for statistics comes into play when you collect
and process the data into a matrix. Then, in most cases, interpreting
the information in either a row, a column or a cell needs to be done
in the context of all other numbers in the table.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">  <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID
      C1          C2        C3       P1        P2      P3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEPT3
    1887.7      81.1    2399.6     800.2     81.3   600.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEZ6L
    1053.9     530.9     211.7     400.1    389.3   403.9</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MICALL1
   136.4     197.4     120.9     127.8    156.3   180.2</font></font></p>
<p class="first-paragraph">A statistical test is a process by which
you make quantitative or qualitative decisions about the numbers.</p>
<p class="block-text-western">Note: performing a statistical test is
never proof that a hypothesis is true. Gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEPT3</font></font>
being highly expressed in a condition does not mean gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEPT3</font></font>
is necessary or causative of that condition.</p>
<h2 class="western"><a name="what-kind-of-questions-can-we-answer-with-a-statistical-test"></a>
22.4 What kind of questions can we answer with a statistical test?</h2>
<p class="first-paragraph">Here is a selection:</p>
<ul>
	<li/>
<p class="compact">How accurate (close to reality) are these
	results?</p>
	<li/>
<p class="compact">How precise (how many digits are
	meaningful) are the values?</p>
	<li/>
<p class="compact">For which observations do values change
	between conditions?</p>
	<li/>
<p class="compact">For which observation is there at least one
	changed condition?</p>
	<li/>
<p class="compact">Are there genes for which there is a trend
	to the data?</p>
	<li/>
<p class="compact">What kinds of patterns are there?</p>
	<li/>
<p class="compact">Which observations vary the same way?</p>
</ul>
<p class="first-paragraph">Statistical tests operate with principles
such as margins of error, probabilities of observing outcomes and
other indirect measures. These measures can easily be misinterpreted
and scientists routinely and commonly summarize and reformulate
incorrectly the statements derived from statistical tests. See the
section on p-values later on this page.</p>
<h2 class="western"><a name="what-types-of-statistical-tests-are-common"></a>
22.5 What types of statistical tests are common?</h2>
<p class="first-paragraph">The pairwise comparison is one of the most
common and conceptually most straightforward tests. For example, a
pairwise comparison would compare the expressions of a gene between
two conditions. A gene that is found to have changed its expression
is called differentially expressed. The set of all genes with
modified expression forms what is called the differential expression
(DE).</p>
<p>Here, it is essential to understand the type of results that
pairwise comparisons usually produce. Almost always we want to answer
the question of whether a gene&rsquo;s expression level has changed.
But instead, what statisitcal test will tell us is the probability
that there was no difference between conditions (i.e., the
probability of the null hypothesis).</p>
<p>When this probability is low (small) we can reject the null
hypothesis, and we conclude that there is a change. The expression
&ldquo;reject the null hypothesis&rdquo; may seem like mincing words
or trying to sound clever. But when further investigations are
necessary it is essential to use these results in their proper
context. It is exceedingly common to formulate conclusions in a
manner that imply more than what the tests support.</p>
<h2 class="western"><a name="do-i-need-to-involve-a-statistician"></a>
22.6 Do I need to involve a statistician?</h2>
<p class="first-paragraph">Ideally, of course, the answer is yes.</p>
<p>But we&rsquo;re in the practical advice business here and, in
reality, it is not always all that easy to find a collaborator well
versed in statics. Besides, it would be a mistake to oversimplify
statistics into a purely procedural skill: <i>&ldquo;any statistician
can do it.&rdquo;</i>. Not really. You would need to find a
collaborator who understands the characteristics of biological data
as well as the challenges of working with it.</p>
<p>We believe that understanding and performing simple statistical
tests like pairwise comparisons, making acceptable statistical
decisions are well within anyone&rsquo;s reach and in this book, we
will provide you with alternative ways for doing it.</p>
<p>For more sophisticated data modeling, for example, time course
analysis or comparing several samples at once, you would need a
statistical collaborator, or you will need to spend some time and
effort understanding the statistics behind it.</p>
<p>Statistics is not as complicated as it looks &ndash; so don&rsquo;t
be discouraged. There is a wealth of information available on more
advanced statistical modeling.</p>
<h2 class="western"><a name="what-does-a-p-value-mean"></a>22.7 What
does a p-value mean?</h2>
<p class="first-paragraph">You probably don&rsquo;t know what a
p-value means especially when it comes to interpreting biological
data. Don&rsquo;t sweat it. We have yet to meet someone that does.</p>
<p>We have of course met overconfident people who think they
understand p-values and most of them are quite convinced that they
got it right. In our personal experience and observation, even
statisticians giving a talk on p-values, and even statistical
textbooks routinely misuse the term.</p>
<p class="block-text-western">Postulate: In our opinion nobody fully
understands p-values and their proper use in biology. There are only
people that misunderstand and misuse the term less egregiously.</p>
<p class="first-paragraph">For every precise definition there is an
expert opinion that can pull that apart and prove that it does always
mean what it is supposed to.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://theconversation.com/give-p-a-chance-significance-testing-is-misunderstood-20207">Give
	p a chance: significance testing is misunderstood</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.nature.com/articles/d41586-019-00857-9">Scientists
	rise up against statistical significance</a></font> in Nature, 2019</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.nature.com/news/statisticians-issue-warning-over-misuse-of-p-values-1.19503">Statisticians
	issue warning over misuse of P values</a></font>in Nature, 2016</p>
	<li/>
<p class="compact">&hellip;and so on&hellip;</p>
</ul>
<p class="first-paragraph">Note how there is an ever increasing
number of publications describing the misuse of p-values;
publications that are akin to old men ranting when trying to get kids
get off their lawn. I know what you&rsquo;re thinking, I am going to
ask a super-duper-expert statistician. Well, here are excerpts from
an email that one of our biologists collaborators received when they
asked one of the leading statistical experts (with multiple Nature
level &ldquo;rant&rdquo; papers) on advice and recommendation on
choosing the &ldquo;proper&rdquo; RNA-Seq method. We include it here
as we found it quite representative of what you might expect:</p>
<p class="block-text-western">[&hellip;]</p>
<p class="block-text-western">&hellip;<i>I need to caution you that
high throughput data analysis is somewhat fragile. Almost any change
in the analysis such as different options for the mappings or the
normalization or the optimization algorithm in the statistical
routines will change the gene list - sometimes dramatically. This is
one reason that biologists use supplemental methods - from wet lab
experiments to high throughput statistical methods such as clustering
and and gene set analysis - to assist in the interpretation of the
results.</i></p>
<p class="block-text-western"><i>Another idea is to use somewhat
different options in the differential expression analysis and then
take the genes in the intersection or union of the significant genes
from both analyses. Using the intersection is likely to be quite
conservative, selecting the genes that have the highest signal to
noise ratios, while the union is likely to be anti-conservative,
having a higher FDR than the adjusted P-values or q-values suggest.</i></p>
<p class="block-text-western"><i>Omitting some low-expressing genes
from the analysis has a very simple effect on the significant gene
list - it will get bigger, but no significant genes should be
dropped. The reason for this is that the unadjusted P-values for the
remaining genes usually do not change, and the adjusted P-values are
a simple function of the unadjusted values and the number of tests,
which decreases when the number of tests is smaller. I say &ldquo;usually
do not change&rdquo; because the estimated dispersion is a function
of all the genes, and will change a bit, not usually not much, when
some genes are omitted from the analysis&hellip;</i></p>
<p class="block-text-western">[&hellip;]</p>
<p class="first-paragraph">A short summary of the above: <b>P-values!
Sixty percent of the time they work every time.</b></p>
<p>A noteworthy feature of the letter is its evasiveness and
non-committal nature of it. The author does not wish to take any
responsibility for making the hard decisions - all the responsibility
is punted back to the biologist: a.k.a. do it many ways and see what
works.</p>
<p>In addition we can&rsquo;t help but point out that the
recommendation that starts with: <i>Omitting some low-expressing
genes from the analysis&hellip;</i> feels very much like <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Data_dredging">p-hacking
and data dredging</a></font>, a misuse of statistics, where, after
knowing a partial answer (the genes that are lowly expressed) we
tweak the input data (omit genes) for the sole purpose of getting
more results presented as statistically significant.</p>
<p>Let&rsquo;s clarify something here. Perhaps our explanation will
feel like splitting hairs but it is not - it cuts to the very essence
of p-hacking. Filtering data before an analysis is an acceptable
practice. You may remove lowly expressed genes, or highly expressed
genes, genes that rhyme with foo, or genes that from far away look
like flies, whatever you wish to do as long as you have an
scientifically acceptable rationale for doing so. State and explain
that rationale, then document it - all fine. What is <i>not
acceptable</i> is the reasoning that the letter above recommends: to
filter data for the <i>sole purpose</i> of making more results pass a
threshold. That is the definition of p-hacking.</p>
<p>Do you now see the origins of our postulate: <i>Nobody fully
understands p-values and their proper use in biology</i>? When under
the pressure to deliver results, and, when faced with the complexity
and messiness of real biological data, statistics was, is and will be
misused and misinterpreted even by very guardians and the experts
that later chide us for not using the concepts in the proper context.
S</p>
<h2 class="western"><a name="so-how-do-i-deal-with-p-values"></a>22.8
So how do I deal with p-values?</h2>
<p class="first-paragraph">Your primary responsibility is to avoid
outrageous, preposterous and egregious errors. Use statistics to
avoid radical mistakes instead of relying it to be the mechanism that
leads you to truth and meaningful discoveries. The bar seems awfully
low, but dont&rsquo; let that trick you into a false sense of
security</p>
<p>A common misuse of a p-value is formulating a stronger statement
than what it was created for. Perhaps the most common misuse of a
p-value is expressed in the following way: <i>&ldquo;our small
p-values show that the our results are not due to random chance&rdquo;</i>
or <i>&ldquo;our small p-values show that the value increased
two-fold&rdquo;</i> As it turns out that is not at all what p-values
indicate.</p>
<p>Another very common misuse of a p-value is to consider a smaller
p-value to be a stronger indication that an effect exists. Sorting by
p-value is not the right mechanism to put the &ldquo;better&rdquo;
results first. We sort by p-value to have &ldquo;some&rdquo; ordering
in our data and to apply a cutoff more easily.</p>
<p>To avoid misleading yourself and others here is what we recommend
on p-values:</p>
<ol>
	<li/>
<p class="block-text-western">Think of the p-value as the
	probability of obtaining an effect of the size that you observe due
	to random chance. Note how the &ldquo;size&rdquo; itself is not a
	factor here, the size could be small or large. The pvalue does not
	capture the actual size. Just that chance of observing that
	particular size (or larger).</p>
	<li/>
<p class="block-text-western">Again remember the p-value does
	not care about how &ldquo;big&rdquo; the effect is.</p>
	<li/>
<p class="block-text-western">Think of p-values as selection
	cutoffs. Use them to reduce a list for further study. But not
	because 0.05 (or whatever the cutoff) is good and 0.06 is not, it is
	because you have cut the list somewhere. There is nothing inherently
	right about 0.05 - or any other cutoff. The cutoff is arbitrary -
	beggars cannot be choosers - when you got noisy data with few
	results, you&rsquo;ll have to be more generous with the cutoff. Then
	with noisy data expect the burden of proof to be higher, you will
	need additional evidence to back up your findings.</p>
	<li/>
<p class="block-text-western">Do NOT use p-values as an
	indicator for more reliable or less reliable results nor as
	indicators of stronger or weaker effects.</p>
</ol>
<p class="first-paragraph">Now, according to my theorem the sentences
above most certainly contain inconsistencies and allow for invalid
interpretations of p-values. My apologies.</p>
<p>The problems caused by misusing p-values are well documented,
unfortunately the more papers you read, the less certain you&rsquo;ll
become that you understand the concept or that it is even worth using
p-values <i>at all</i>:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.nature.com/articles/nmeth.4210">Interpreting
	P values</a></font>, Nature Methods 2017</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.nature.com/articles/nmeth.4120">P
	values and the search for significance</a></font></p>
	<li/>
<p class="compact">&hellip; and so on &hellip;</p>
</ul>
<h2 class="western"><a name="do-i-need-to-compute-and-discuss-p-values"></a>
22.9 Do I need to compute and discuss p-values?</h2>
<p class="first-paragraph">Yes, we use them because there is no
better alternative.</p>
<p><b>Theorem:</b> The most cynical interpretation of p-values is
that their serve as the mechanism to filter out studies created by
people that were so clueless when it comes to large scale analysis
that they couldn&rsquo;t even produce small p-values. From that point
of view p-values do indeed correlate with the quality and validity of
the research.</p>
<p><b>Corollary:</b> An expert scientist can unwittingly publish a
Nature publication with tiny p-values, impressive and compelling
visualizations <font color="#4f81bd"><a href="https://www.nature.com/articles/nature12535">even
from data is not different from random noise</a></font>.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="what-is-data"></a>23 What is data?</h1>
<p class="first-paragraph">Intuitively people believe that they have
a good practical understanding of what data is, but we found that it
can be unexpectedly difficult to articulate even what the word <i>data</i>
means.</p>
<p class="block-text-western">Reflect for a moment, what is data?</p>
<p class="first-paragraph">We went around and asked a bunch of people
the hard question <b>What is data?</b>.</p>
<p>Note the most common definition:</p>
<p class="block-text-western">DATA is, uhhhh &hellip; DATA</p>
<h2 class="western"><a name="so-what-is-data"></a>23.1 So what is
data?</h2>
<p class="first-paragraph">The simplest definition relies on
clarifying that &ldquo;data&rdquo; and &ldquo;data format&rdquo; are
closely related concepts:</p>
<ol>
	<li/>
<p class="compact">A symbolic representation of <b>information</b>.</p>
	<li/>
<p class="compact">A <b>design</b>, and <b>optimization</b> of
	that information.</p>
</ol>
<p class="first-paragraph">The second point is, in my opinion, the
essential property of data that most people can&rsquo;t quite
articulate.</p>
<p>Depending on the format, information that is present in the data
may be readily accessible or difficult to obtain. Some information
may be impossible to extract, even though the information is there!
Many scientists are unaware that information organization is just as
important as the information itself.</p>
<h2 class="western"><a name="essential-properties-of-data"></a>23.2
Essential properties of data</h2>
<ul>
	<li/>
<p class="compact">The same information may be represented in
	different formats (optimized differently).</p>
	<li/>
<p class="compact">The same format (optimization) may be used
	to describe different types of information.</p>
	<li/>
<p class="compact">There may be information in data that is
	not readily accessible (optimized).</p>
</ul>
<p class="first-paragraph">Often, it can be surprisingly difficult to
predict the types of new information that we can derive from a
dataset. Pushing the limits of what we can extract from data is an
essential part of science.</p>
<p class="block-text-western">Understanding data formats, what
information is encoded in each, and when it is appropriate to use one
format over another is an essential skill of a bioinformatician.</p>
<h2 class="western"><a name="how-life-scientists-think-about-bioinformatics"></a>
23.3 How life scientists think about bioinformatics</h2>
<p class="first-paragraph">For a typical life scientist
bioinformatics means a software transforms &ldquo;data&rdquo; into an
&ldquo;answer&rdquo;.</p>
<p><img src="0000_html_c731492640345707.png" name="Image47" align="bottom" hspace="12" width="560" height="99" border="0"/>
</p>
<p>We call this &ldquo;a tool-centric&rdquo; world view, where the
software is a defining component. Running a software is expected to
immediately produce the insight and explanation that the scientist
needs.</p>
<h2 class="western"><a name="what-bioinformatics-is-in-reality"></a>23.4
What bioinformatics is in reality</h2>
<p class="first-paragraph">In reality, as a bioinformatician you need
to understand that there is <b>only data</b>. Software tools merely
transform data from one format to another. The new, resulting data
will usually have different &ldquo;optimization&rdquo;. It may be
richer in that it may contain a combination of the information from
other data. But it is still data - and the biological insight comes
from understanding and investigating this data.</p>
<p><img src="0000_html_633aa58e5573e42f.png" name="Image48" align="bottom" hspace="12" width="560" height="309" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-state-of-data-in-bioinformatics"></a>
23.5 What is the state of data in bioinformatics?</h2>
<p class="first-paragraph">It is kind of a mess, but fear not:</p>
<p class="block-text-western"><i>Everything&rsquo;s gonna be
alright</i><br/>
<i>Everything&rsquo;s gonna be okay</i><br/>
<i>It&rsquo;s
gonna be a good, good, life</i><br/>
<i>That&rsquo;s what my
therapists say</i><br/>
<br/>
<br/>

</p>
<p class="first-paragraph">Here is what is happening. Scientists that
control, influence and shape the field, as well as society as a whole
has not yet fully realized the complexity, the challenges and the
importance of data within this field of science. The current
organizational structure, the way we choose to represent, store and
disseminate data are all woefully inadequate. The amount of newly
generated information is continually overflowing the &ldquo;containers&rdquo;
and &ldquo;storage&rdquo; that were designated to hold and organize
them.</p>
<p>Your skill as a bioinformatician will be measured in part by your
ability to cut a path through the increasingly diverging information
and subjectivity to identify and isolate the relevant information.</p>
<h2 class="western"><a name="what-kind-of-problems-does-bioinformatics-data-have"></a>
23.6 What kind of problems does bioinformatics data have?</h2>
<p class="first-paragraph">Let us demonstrate the ever growing
magnitude of the problem. A little background first.</p>
<p>Have you heard about the Human Genome Project? The one <font color="#4f81bd"><a href="https://www.nature.com/articles/35057062">&ldquo;completed&rdquo;
in the year 2000</a></font>? Perhaps you heard that <font color="#4f81bd"><a href="https://www.nature.com/collections/vqfbhrpqcd">ten
years later</a></font> we realized that the job was not quite done,
but that the <font color="#4f81bd"><a href="https://www.nature.com/articles/470140a">best
is yet to come</a></font>.</p>
<p>That was about ten years ago. So what is the status now? How well
do we understand the Human Genome? Let us reproduce here a blog post
made by Heng Li on Oct 13, 2017. It merely tries to answer the
question <font color="#4f81bd"><a href="http://lh3.github.io/2017/11/13/which-human-reference-genome-to-use">Which
human reference genome to use?</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"></p>
<p class="first-paragraph"><b>Which human reference genome to use</b></p>
<p>TL;DR: If you map reads to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh37</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font>,
use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hs37-1kg</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/human_g1k_v37.fasta.gz</font></font></p>
<p class="first-paragraph">If you map to GRCh37 and believe decoy
sequences help with better variant calling, use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hs37d5</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz</font></font></p>
<p class="first-paragraph">If you map reads to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg38</font></font>,
use the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz</font></font></p>
<p class="first-paragraph">There are several other versions of
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh37/GRCh38</font></font>.
What&rsquo;s wrong with them? Here are a collection of potential
issues:</p>
<ol>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Inclusion of
	ALT contigs. ALT contigs are large variations with very long
	flanking sequences nearly identical to the primary human assembly.
	Most read mappers will give mapping quality zero to reads mapped in
	the flanking sequences. This will reduce the sensitivity of variant
	calling and many other analyses. You can resolve this issue with an
	ALT-aware mapper, but no mainstream variant callers or other tools
	can take the advantage of ALT-aware mapping.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Padding ALT
	contigs with long &ldquo;N&rdquo;s. This has the same problem with 1
	and also increases the size of genome unnecessarily. It is worse.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Inclusion of
	multi-placed sequences. In both GRCh37 and GRCh38, the
	pseudo-autosomal regions (PARs) of chrX are also placed on to chrY.
	If you use a reference genome that contains both copies, you will
	not be able to call any variants in PARs with a standard pipeline.
	In GRCh38, some alpha satellites are placed multiple times, too. The
	right solution is to hard mask PARs on chrY and those extra copies
	of alpha repeats.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Not using
	the rCRS mitochondrial sequence. rCRS is widely used in population
	genetics. However, the official <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh37</font></font>
	comes with a mitochondrial sequence 2bp longer than rCRS. If you
	want to analyze mitochondrial phylogeny, this 2bp insertion will
	cause troubles. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>
	uses rCRS.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Converting
	semi-ambiguous IUB codes to &ldquo;N&rdquo;. This is a very minor
	issue, though. Human chromosomal sequences contain few
	semi-ambiguous bases.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Using
	accession numbers instead of chromosome names. Do you know
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CM000663.2</font></font>
	corresponds to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1</font></font>
	in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>?</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Not
	including unplaced and unlocalized contigs. This will force reads
	originated from these contigs to be mapped to the chromosomal
	assembly and lead to false variant calls.</p>
</ol>
<p class="first-paragraph">Now we can explain what is wrong with
other versions of human reference genomes:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19/chromFa.tar.gz</font></font>
	from UCSC: 1, 3, 4 and 5.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg38/hg38.fa.gz</font></font>
	from UCSC: 1, 3 and 5.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCA_000001405.15_GRCh38_genomic.fna.gz</font></font>
	from NCBI: 1, 3, 5 and 6.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</font></font>
	from EnsEMBL: 3.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.dna.toplevel.fa.gz</font></font>
	from EnsEMBL: 1, 2 and 3.</p>
</ul>
<p class="first-paragraph">Using an inappropriate human reference
genome is usually not a big deal unless you study regions affected by
the issues. However,</p>
<ol>
	<li/>
<p class="compact">other researchers may be studying in these
	biologically interesting regions and will need to redo alignment;</p>
	<li/>
<p class="compact">aggregating data mapped to different
	versions of the genome will amplify the problems. It is still
	preferable to choose the right genome version if possible.</p>
</ol>
<p class="first-paragraph">Well, welcome to bioinformatics!</p>
<p>(end of blog post)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"></p>
<p class="first-paragraph">Reflect on the above a little bit. Almost
twenty years after &ldquo;completing&rdquo; the human genome project
we have a hard time deciding &ldquo;which&rdquo; human genome should
be used to begin with. There are several alternatives. Many of the
published and widely used genomics builds contain inconsistencies
that make comparing results across them more challenging and less
reliable. All these &ldquo;builds&rdquo; are being simultanously used
to generate increasing amount of information, that all gets &ldquo;mixed&rdquo;
together propagating errors and inconsistencies. But fear not:</p>
<h2 class="western"><a name="how-complete-is-the-data-that-will-i-obtain"></a>
23.7 How complete is the data that will I obtain?</h2>
<p class="first-paragraph">Let&rsquo;s take the sequence data for
chromosome 22 of the human genome, which is distributed from the UCSC
data site at
<font color="#4f81bd"><a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/">http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download and unzip the file on the fly.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz
| gunzip -c &gt; chr22.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Look at the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | head -4</font></font></p>
<p class="first-paragraph">it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;chr22</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font></p>
<p class="first-paragraph">Ok, first let&rsquo;s deal with a little
surprise. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
stands for &ldquo;unknown base&rdquo;. It looks like chromosome 22
starts with a series of unknown bases in it. Whatever we were about
to investigate got derailed, what&rsquo;s with all the Ns? Thankfully
we have Unix tools at our disposal to poke and prod at this data.
Let&rsquo;s see just how many unknown bases there are for this
chromosome? We can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
with the flags <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-o</font></font>
(only matching) on the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
pattern:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | grep -o N  | head</font></font></p>
<p class="first-paragraph">Now count the lines:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | grep -o N  | wc -l</font></font></p>
<p class="first-paragraph">The result comes at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11,658,691</font></font>,
over 11 million bases! Is that a lot? Ok, so how big is chromosome
22? We have installed many tools that could be used to print that
information</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat chr22.fa</font></font></p>
<p class="first-paragraph">or:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">infoseq
chr22.fa</font></font></p>
<p class="first-paragraph">or, why not do it the hard way with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioawk</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | bioawk -c fastx '{ print length($seq) }'</font></font></p>
<p class="first-paragraph">The above code produces <font face="Consolas, serif"><font size="2" style="font-size: 11pt">50,818,468</font></font>.
In summary, 11 million out of 50 million bases (~ 23%) of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr22</font></font>
are currently unknown.</p>
<p>Let&rsquo;s think about that a bit. Two decades after the
announcement of the successful assembly of the entire human genome,
23% of chromosome 22 is listed as unknown. Ask a biologist to
estimate how much of chromosome 22 is labeled as Ns see what they
say. You will most likely experience a disconnect between their
belief and reality.</p>
<p>But is it truly unknown? Well, not exactly. There are pieces of
DNA that have been sequenced but have not yet been localized. Often
these regions will fall into centromeres and telomeres. These
un-localized pieces may be distributed separately. Yet in the vast
majority of publications, when authors state that they are using the
human genome, they will use only the chromosome 22 as indicated
above. The one were 22% of it will be considered as &ldquo;unknown&rdquo;
which is just a scientifically palatable word for &ldquo;missing.&rdquo;</p>
<p>You have two choices: you can go out there and shout into the
void: <i>&ldquo;somebody should be doing something!&rdquo;</i> or
perhaps As Heng Li puts it, keep on the sunny side and say: <i>&ldquo;Welcome
to bioinformatics!&rdquo;</i></p>
<h2 class="western"><a name="final-thoughts-on-data"></a>23.8 Final
thoughts on data</h2>
<p class="first-paragraph">Knowing how to solve these challenges is
what makes bioinformatics and bioinformaticians valuable. That&rsquo;s
why you are here, no? The Biostar Handbook will guide you in the next
chapters so that you can look like the genius you actually are.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="biological-data-sources"></a>24
Biological data sources</h1>
<h2 class="western"><a name="where-is-biomedical-data-stored"></a>24.1
Where is biomedical data stored?</h2>
<p class="block-text-western">Walter Goad of the Theoretical Biology
and Biophysics Group at Los Alamos National Laboratory and others
established the Los Alamos Sequence Database in 1979, which
culminated in 1982 with the creation of the public GenBank. (source:
Wikipedia)</p>
<p class="first-paragraph">GenBank can be considered as the first
large-scale, public repository of biomedical data. Today, it is
integrated with the services provided by the <i>National Center for
Biotechnology Information (NCBI)</i>.</p>
<p>Today, just about all biomedical data is distributed via various
web-based services. You may already be familiar with one or more such
services; for example, <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/">NCBI
Home Page</a></font> is an important entry point to a wealth of
information.</p>
<p><img src="0000_html_c2dbf93093c2346a.png" name="Image49" align="bottom" hspace="12" width="560" height="367" border="0"/>
</p>
<p>Early warning: when visiting bioinformatics data sites, you are
likely to be subjected to an information overload coupled with severe
usability problems. Typically, it is surprisingly difficult to find
the relevant, up-to-date, and correct information on any entity of
interest. With the growth of data, we anticipate the situation only
to worsen.</p>
<p>It is not your fault: it is theirs. Persevere and keep at it. You
will eventually find what you need.</p>
<h2 class="western"><a name="what-are-the-major-dna-data-repositories"></a>
24.2 What are the major DNA data repositories?</h2>
<p class="first-paragraph">DNA sequences collected by scientists are
deposited in databases and made available to the public via the
Internet. The <font color="#4f81bd"><a href="http://www.insdc.org/">INSDC:
International Nucleotide Sequence Database Collaboration</a></font>
has assumed stewardship for maintaining copies of the so-called
&ldquo;primary DNA data.&rdquo; The member organizations of this
collaboration are:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/">NCBI:
	National Center for Biotechnology Information</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.embl.org/">EMBL:
	European Molecular Biology Laboratory</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ddbj.nig.ac.jp/">DDBJ:
	DNA Data Bank of Japan</a></font></p>
</ul>
<p class="first-paragraph">The INSDC has set up rules on the types of
data that will be mirrored. The most important of these from a
bioinformatician&rsquo;s perspective are:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/genbank/">GenBank</a></font>
	contains all annotated and identified DNA sequence information</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/sra">SRA:
	Short Read Archive</a></font> contains measurements from high
	throughput sequencing experiments</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.uniprot.org/">UniProt:
	Universal Protein Resource</a></font> is the most authoritative
	repository of protein sequence data.</p>
	<li/>
<p class="compact">[Protein Data Bank (PDB)] <font color="#4f81bd"><a href="http://www.rcsb.org/pdb/home/home.do">pdb</a></font>
	is the major repository of 3D structural information about
	biological macromolecules (proteins and nucleic acids). PDB contains
	structures for a spectrum of biomolecules - from small bits of
	proteins/nucleic acids all the way to complex molecular structures
	like ribosomes.</p>
</ul>
<h2 class="western"><a name="what-kind-of-other-data-sources-are-there"></a>
24.3 What kind of other data sources are there?</h2>
<p class="first-paragraph">Beyond the primary DNA data, various
institutes may also hosts a variety of services and vast numbers of
annotated datasets that may be supporting the work performed in each
organization.</p>
<p>Also, there are different tiers of other organizations that have
taken it upon themselves to reorganize, rename, and process existing
data and provide additional functionalities and visualizations.
Scientific interest groups formed around model organisms may also
maintain their resources:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://hgdownload.soe.ucsc.edu/downloads.html">UCSC
	Genome Browser</a></font> invented the graphical browser
	visualization of genomes. Today it offers comprehensive comparative
	genomics data across vertebrate genomes.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://flybase.org/">FlyBase</a></font>
	is the database of Drosophila (fruit fly) genes and genomes.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.wormbase.org/">WormBase</a></font>
	is the primary resource for nematode biology.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.yeastgenome.org/">SGD:
	Saccharomyces Genome Database</a></font> provides comprehensive
	integrated biological information for the budding yeast
	<i>Saccharomyces cerevisiae</i> along with search and analysis tools
	to explore these data.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://rnacentral.org/">RNA-Central</a></font>
	is a meta-database that integrates information from several other
	resources.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.arabidopsis.org/">TAIR</a></font>
	The Arabidopsis Information Resource is the primary resource for
	genetic and molecular data about <i>Arabidopsis thaliana</i>, a
	model higher plant.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://ecocyc.org/">EcoCyc</a></font>
	(Encyclopedia of <i>E. coli</i> Genes and Metabolic Pathways) is a
	scientific database for the bacterium <i>Escherichia coli</i> K-12
	MG1655.</p>
</ul>
<p class="first-paragraph">Beyond the second tier of comprehensive
databases, there are a large number of specialized data resources,
often started and/or managed by single individuals with little-to-no
support.</p>
<p>Describing each resource in detail is beyond the scope of this
book. Most claim to provide &ldquo;self-discoverable&rdquo; interface
elements, though their usability is typically far from optimal.</p>
<p>It is also hard to shake the feeling that each data provider&rsquo;s
ultimate goal is to be &ldquo;too big to fail&rdquo;; hence, there is
an eagerness to integrate more data for the sole purpose of growing
larger. It does not help that interface design is often supervised by
scientists and this typically means jamming way too much information
into way too little space. As a rule, using bioinformatics data
resources is frustrating, requires patience, and often requires the
suspension of disbelief.</p>
<p class="block-text-western">Among the most significant challenges
for a newcomer, however, is where to look for authoritative and
up-to-date information. No database will ever admit that their data
is not quite up to date or incomplete.</p>
<p class="block-text-western">In addition to using different naming
schemes, the data formats and data content will vary from resource to
resource, adding no small confusion when trying to combine and
reconcile information gleaned from different sources.</p>
<h2 class="western"><a name="is-there-a-list-of-all-resources"></a>24.4
Is there a list of &ldquo;all&rdquo; resources?</h2>
<p class="first-paragraph">Once a year the journal Nucleic Acids
Research publishes its so-called &ldquo;database issue&rdquo;. Each
article of this issue of the journal will provide and overview and
updates about a specific database written by the maintainers of that
resource.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://academic.oup.com/nar/issue/47/D1">View
	the NAR: 2019 Database Issue</a></font>.</p>
</ul>
<h2 class="western"><a name="whats-in-a-name"></a>24.5 What&rsquo;s
in a name?</h2>
<p class="first-paragraph">Two submitters using the same systematic
names to describe two different genes would lead to very confusing
situation. This makes it necessary to associate unique ids with each
biological entity.</p>
<p>On the other hand, making gene names mnemonics that reflect their
function, such as calling a <i>pigment dispersing factor</i> a <i>PDF</i>,
or <i>Protein Kinase A</i> as <i>PKA</i> helps immensely during the
biological interpretation process. The problem, of course, is that
there could be many different variants of <i>PKA</i> in different
genomes; moreover, potentially different terms could be shortened to
similar or even identical mnemonics.</p>
<p>In general, all biological terms are defined by multiple names:
there is a common name, a systematic name, a database-specific name,
an accession number and sometimes a so-called locus. The common names
typically carry a biological meaning, whereas systematic names are
built with rules that make them unique. The accession and loci values
may also follow various rules, though it is common that the original
design of the rules turned out to be insufficient to cover all use
cases. Some identifiers fall out of use. Once widely used, GenBank <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi</font></font>
numbers are now considered obsolete and were replaced by accession
numbers. Other identifiers such as &ldquo;locus&rdquo; lack a proper
definitions and should be avoided. The word locus is more appropriate
in identifying a region in a genome, rather than a certain biological
entity.</p>
<p>Most community projects formed around model organisms have adopted
custom gene naming conventions that may only apply to data produced
by that community.</p>
<h2 class="western"><a name="project-systematic-names"></a>24.6
Project systematic names</h2>
<p class="first-paragraph">Different organzations may come up with
their own naming schemes that are unique for an organism. For
example, the Saccharomyces Genome database project has detailed
documentation on <font color="#4f81bd"><a href="http://www.yeastgenome.org/help/community/nomenclature-conventions">naming
conventions</a></font> for every element. For example:</p>
<ul>
	<li/>
<p class="compact">Gene naming <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ADE12</font></font>:
	The gene name should consist of three letters (the gene symbol)
	followed by an integer</p>
	<li/>
<p class="compact">Open reading frame naming <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YGR116W</font></font>:
	First letter &lsquo;Y&rsquo; (&lsquo;Yeast&rsquo;); the second
	letter denotes the chromosome number (&lsquo;A&rsquo; is chr I,
	etc.); the third letter is either &lsquo;L&rsquo; or &lsquo;R&rsquo;
	for left or right chromosome arm; next is a three digit number
	indicating the order of the ORFs on that arm of a chromosome
	starting from the centromere (like a highway exit), irrespective of
	strand; finally, there is an additional letter indicating the
	strand, either &lsquo;W&rsquo; for Watson (forward) or &lsquo;C&rsquo;
	for Crick (reverse). Thus, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YGR116W</font></font>
	is the 116th ORF right of the centromere on chromosome VII on the
	forward strand.</p>
</ul>
<p class="first-paragraph">You may stop for a second and think about
whether this ORF naming convention is an optimal long term solution
(hint: it is not).</p>
<p>For example, naming strands of DNA to be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">W/C</font></font>
after <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/James_Watson">Watson</a></font>
and <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Francis_Crick">Crick</a></font>
is inappropriate (and not just because <font color="#4f81bd"><a href="https://liorpachter.wordpress.com/2018/05/18/james-watson-in-his-own-words/">James
Watson turned out to be a jerk</a></font>). The DNA has orientation,
that may be forward and reverse, these words make sense, they express
something. There is no reason to introduce a new terminology, that
adds no new information, in addition to the existing one.</p>
<p>Then why call chromosome <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>,
what is the benefit of introducing another layer of remapping, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
&hellip; what is the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">13</font></font>th
letter? It does not quite help. It even used to be that chromosomes
were labeled with roman numerals <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI</font></font>.
Thankfully the practice using roman numeral has been changed, though
note how the instructions above that we copied from the YGD website
still indicate roman numerals.</p>
<p>Then what about the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y</font></font>
there, what problem does that solve? Do scientists often confuse
human ORFs with yeast ones? I have a hard time imagining that.
Telling apart ORFs from two yeast strains is a far more common need
for a scientist working in this field. The present naming scheme does
not help there.</p>
<p>Do you see how the <font color="#4f81bd"><a href="http://www.nytimes.com/2014/08/18/technology/for-big-data-scientists-hurdle-to-insights-is-janitor-work.html?_r=0">For
Big-Data Scientists, &lsquo;Janitor Work&rsquo; Is Key Hurdle to
Insights</a></font> rings true?</p>
<p>We wrote this all up not to criticize (well maybe a little bit)
but to demonstrate to readers that they will often have to deal with
peculiarities that appear to make little sense. It helps to remember
that many of these problems originate in local optimizations, ad-hoc
conventions or even regrettable historical rationale, rather than
well designed and well thought out decisions.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="common-data-types"></a>25 Common data
types</h1>
<p class="first-paragraph">A typical biological data analysis will
require you to obtain and manipulate data that may be categorized as:</p>
<ol>
	<li/>
<p class="compact">Data that captures prior knowledge (aka
	reference: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BED</font></font>)</p>
	<li/>
<p class="compact">Experimentally obtained data (aka
	sequencing reads: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>)</p>
	<li/>
<p class="compact">Data generated by the analysis (aka
	results: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>,
	formats from point 1 above, and many nonstandard formats)</p>
</ol>
<p class="first-paragraph">Below we list some of the most common
bioinformatics formats. We will discuss each of these formats in more
detail; here we&rsquo;ll first give you a quick overview of their
purpose. Remember that each of these formats may contain information
that overlaps with other forms. The optimization of each format makes
it suited to specific goals. With time you will immediately be able
to recognize the format from its extension or</p>
<h2 class="western"><a name="a-quick-look-at-the-genbank-format."></a>
25.1 A Quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GENBANK</font></font>
format.</h2>
<p class="first-paragraph">The &ldquo;first&rdquo; and most ancient,
generic biological data format. Extensions <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.gb</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.genbank</font></font>.
The GenBank format is optimized for reading by humans. Not suited for
large scale data processing.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id  NM_000020 -format gb | head</font></font></p>
<p class="first-paragraph">will produce output in GenBank:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LOCUS
      NM_000020               4263 bp    mRNA    linear   PRI
23-DEC-2018</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DEFINITION
 Homo sapiens activin A receptor like type 1 (ACVRL1), transcript</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           variant 1, mRNA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCESSION
  NM_000020</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VERSION
    NM_000020.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KEYWORDS
   RefSeq.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOURCE
     Homo sapiens (human)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ORGANISM  Homo sapiens</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Eukaryota; Metazoa; Chordata; Craniata; Vertebrata;
Euteleostomi;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Mammalia; Eutheria; Euarchontoglires; Primates;
Haplorrhini;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    regulatory      4241..4246</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    /regulatory_class=&quot;polyA_signal_sequence&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    /gene=&quot;ACVRL1&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    /gene_synonym=&quot;ACVRLK1; ALK-1; ALK1; HHT;
HHT2; ORW2;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    SKR3; TSR-I&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    polyA_site      4263</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    /gene=&quot;ACVRL1&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    /gene_synonym=&quot;ACVRLK1; ALK-1; ALK1; HHT;
HHT2; ORW2;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    SKR3; TSR-I&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORIGIN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       1 aggaaacggt ttattaggag ggagtggtgg agctgggcca ggcaggaaga
cgctggaata</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      61 agaaacattt ttgctccagc ccccatccca gtcccgggag gctgccgcgc
cagctgcgcc</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     121 gagcgagccc ctccccggct ccagcccggt ccggggccgc gcccggaccc
cagcccgccg</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     181 tccagcgctg gcggtgcaac tgcggccgcg cggtggaggg gaggtggccc
cggtccgccg</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">As you can see above it is a fairly
complex format where information is distributed throughout the file.
For that reason, very few tools can operate on GenBank files.</p>
<h3 class="western"><a name="quick-look-at-the-fasta-format"></a>25.1.1
Quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
format</h3>
<p class="first-paragraph">The same GenBank record above may be
formatted in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
format. Typical extension <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fa</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fasta</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fna</font></font>,
sometimes <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.seq</font></font>.
This format retains only some of the sequence information of a
GenBank record.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id  NM_000020 -format fasta | head</font></font></p>
<p class="first-paragraph">will produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NM_000020.2
Homo sapiens activin A receptor like type 1 (ACVRL1), transcript
variant 1,
mRNA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGGAAACGGTTTATTAGGAGGGAGTGGTGGAGCTGGGCCAGGCAGGAAGACGCTGGAATAAGAAACATTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTGCTCCAGCCCCCATCCCAGTCCCGGGAGGCTGCCGCGCCAGCTGCGCCGAGCGAGCCCCTCCCCGGCT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAGCCCGGTCCGGGGCCGCGCCCGGACCCCAGCCCGCCGTCCAGCGCTGGCGGTGCAACTGCGGCCGCG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGTGGAGGGGAGGTGGCCCCGGTCCGCCGAAGGCTAGCGCCCCGCCACCCGCAGAGCGGGCCCAGAGGG</font></font></p>
<p class="first-paragraph">Note how the header seems to have
context-related information as free text.</p>
<h2 class="western"><a name="a-quick-look-at-the-fastq-format"></a>25.2
A quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
format</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
format is meant to store experimental sequence measurements produced
by a sequencing instrument. It can be thought of as FASTA with
qualities.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 100 -Z SRR1553607 | head</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SRR1553607.1
1
length=202</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTTAGCGTTGTTGATCGCGACGCAACAACTGGTAAAGAATCTGGAAGAAGGATATCAGTTCAAACGCTCAAG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+SRR1553607.1
1
length=202</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BB@FFFFFHHHHHJJJJJJJJJJJJJJJJJJJGHIJJJJJJJJJHHHHHFFFFFEEEEEEEEEDDDDDDDDD</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SRR1553607.2
2
length=202</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGTGTAAGCACAGTACTCGGCCCACATCGCCTTTGTGTTAATGAAGTTTGGGTATCAACTTTCATCCCCAAT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+SRR1553607.2
2
length=202</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">?@;BDDDDFHFHFFFGIIGHIIJJGJIGIJIIIIGDGGGHEIGJIIIGIIHJ5@FGHJJIEGGEEHHFFFFF</font></font></p>
<p class="first-paragraph">(the above input was truncated a little
bit for legibility). The header lines may contain extra information
for example, instead of</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SRR1553607.1
1 length=202</font></font></p>
<p class="first-paragraph">the instrument will typically specify more
information on the context of how this measurement was made.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@EAS139:136:FC706VJ:2:2104:15343:197393
1:N:18:1</font></font></p>
<h2 class="western"><a name="a-quick-look-at-the-gffgtfbed-formats"></a>
25.3 A quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF/GTF/BED</font></font>
formats</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF/GTF/BED</font></font>
formats are the so-called &ldquo;interval&rdquo; formats that retain
only the coordinate positions for a region in a genome. Each is tab
delimited and will contain information on the chromosomal coordinate,
start, end, strand, value and other attributes, though the order of
the columns will depend on the feature.</p>
<p>Three column BED has the columns for chromosome, start, end</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127471196  127472363</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127472363  127473530</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127473530  127474697</font></font></p>
<p class="first-paragraph">Six column BED add name, value and strand:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127471196  127472363  Pos1  0  +</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127472363  127473530  Pos2  0  +</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr7
   127473530  127474697  Pos3  0  +</font></font></p>
<p class="first-paragraph">Up to 12 columns may be specified for BED
formats. The coordinates start at 0 (that is the first valid
coordinate).</p>
<p>The <font color="#4f81bd"><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF/GTF
formats</a></font> are 9 column tab-delimited formats. The
coordinates start at value 1.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1
. mRNA            1300  9000  .  +  . 
ID=mrna0001;Name=sonichedgehog</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1
. exon            1300  1500  .  +  . 
ID=exon00001;Parent=mrna0001</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1
. exon            1050  1500  .  +  .  ID=exon00002;Parent=mrna0001</font></font></p>
<p class="first-paragraph">See the details on the <font color="#4f81bd"><a href="https://genome.ucsc.edu/FAQ/FAQformat.html">UCSC
formats</a></font> page. In the GTF format, the layout of the 9th
(last column) is different. Also, GTF formats are valid only if they
have values for &ldquo;gene_id&rdquo; and &ldquo;transcript_id&rdquo;
specified on the 9th column.</p>
<p>BED and GFF are the formats used to provide context to a genome.
We describe the formats in substantially more detail in the chapter
titled <font color="#4f81bd"><a href="#ch038.xhtml#visualize-interval-data">Visualizing
interval data</a></font></p>
<h2 class="western"><a name="a-quick-look-at-the-sambam-formats"></a>25.4
A quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAM/BAM</font></font>
formats</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAM/BAM</font></font>
formats are so-called Sequence Alignment Maps. These files typically
represent the results of aligning a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
file to a reference <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
file and describe the individual, pairwise alignments that were
found. Different algorithms may create different alignments (and
hence BAM files).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view http://data.biostarhandbook.com/bam/demo.bam | head -5</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553425.13617
   163 AF086833.2  46  60  101M    =   541 596
GAATAACTATGAGGAAGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTTGTTTCAGAGC
 
@&lt;@FFFDDHHGHHJIJJIJJJJJJJJJJJJJJIIJJJJGIJJJJJJEIEIGIIJJJJJIIIIGIIIGGEGIIIGIIJJJGGGHHFHGFDDFDEEDDEDCDCNM:i:1
MD:Z:81C19  MC:Z:101M   AS:i:96 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553425.13755
   99  AF086833.2  46  60  101M    =   46  101
GAATAACTATGAGGAAGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTTGTTTCAGAGC
 
?:@DBDDDFHGDHGIGEGAFGGHDHIIIIIIIGGGGIGEBHHGHFGHIDCHIIIIII&lt;G@FGHDFGGHH@FGGEGGGEHIE@HHHGHF;?BC@CECFC:@&gt;NM:i:1
MD:Z:81C19  MC:Z:101M   AS:i:96 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553425.13755
   147 AF086833.2  46  60  101M    =   46  -101   
GAATAACTATGAGGAAGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTTGTTTCAGAGC
 
DDDDCDDDDDECEEDEDFFFEFFGHHHHIGGHG@=@@=GIIJHIIHFDIIIIIIHHGJIFAJJJIGIJIIIJHHGGIGEGGHFHHJIIHHHGHFFFFFCCCNM:i:1
MD:Z:81C19  MC:Z:101M   AS:i:96 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553425.11219
   2227    AF086833.2  47  60  71H30M  =   146 171
AATAACTATGAGGAAGATTAATAATTTTCC  IGHF@GHEHIIHE&gt;IIGHFDHDDDFFF@@@ 
NM:i:0  MD:Z:30 MC:Z:101M   AS:i:30 XS:i:0 
SA:Z:AF086833.2,68,+,30S71M,60,1;</font></font></p>
<h3 class="western"><a name="quick-look-at-the-vcf-format"></a>25.4.1
Quick look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
format</h3>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
(Variant Call Format) describes the variation of alignments relative
to a reference. Extensions: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.vcf</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">vcf.gz</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bcf</font></font>.
If the extension ends with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.gz</font></font>
it is a block gzipped file (see the chapter on Data Compression). A
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
file is typically created from a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
file, whereas the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
file was created from a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
and a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
file.</p>
<p>Thus think of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
file as a file that captures the differences of for each of the
sequences in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQ</font></font>
file relative to the genome in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -H http://data.biostarhandbook.com/variant/subset_hg19.vcf.gz |
head -5</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400410  rs540061190 CA  C   100 PASS   
AC=0;AF=0.00179712;AN=12;NS=2504;DP=7773;EAS_AF=0.002;AMR_AF=0;AFR_AF=0.0015;EUR_AF=0;SAS_AF=0.0051;AA=?|A|-|unsure;VT=INDEL
   GT  0|0 0|0 0|0 0|0 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400666  rs11670588  G   C   100 PASS   
AC=5;AF=0.343251;AN=12;NS=2504;DP=8445;EAS_AF=0.3988;AMR_AF=0.3228;AFR_AF=0.3396;EUR_AF=0.3101;SAS_AF=0.3395;AA=G|||;VT=SNP
GT  1|0 0|1 0|1 0|0 0|0 1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400742  rs568501257 C   T   100 PASS   
AC=0;AF=0.000199681;AN=12;NS=2504;DP=15699;EAS_AF=0.001;AMR_AF=0;AFR_AF=0;EUR_AF=0;SAS_AF=0;AA=C|||;VT=SNP
 GT  0|0 0|0 0|0 0|0 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400819  rs71335241  C   G   100 PASS   
AC=0;AF=0.225839;AN=12;NS=2504;DP=10365;EAS_AF=0.2897;AMR_AF=0.2349;AFR_AF=0.2088;EUR_AF=0.161;SAS_AF=0.2434;AA=N|||;VT=SNP
GT  0|0 0|0 0|0 0|0 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400908  rs183189417 G   T   100 PASS   
AC=1;AF=0.0632987;AN=12;NS=2504;DP=13162;EAS_AF=0.002;AMR_AF=0.1153;AFR_AF=0.0726;EUR_AF=0.0885;SAS_AF=0.0511;AA=-|||;VT=SNP
   GT  0|0 0|0 0|0 0|0 0|0 0|1</font></font></p>
<h2 class="western"><a name="can-i-convert-between-formats."></a>25.5
Can I convert between formats.</h2>
<p class="first-paragraph">If on format contains information that
another format stores then you can typically convert it. For example
you can make a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF</font></font>
file from a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GENBANK</font></font>
format. Reversing the process would create an incomplete GENBANK
file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a genbank file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nucleotide -format gb -id AF086833 &gt; AF086833.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
We can convert the GenBank file into GFF3.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -feature -osformat gff3 &gt;
AF086833.gff</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
We can convert the GenBank file into FASTA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -feature -osformat fasta &gt;
AF086833.fa</font></font></p>
<h2 class="western"><a name="what-is-reference-data"></a>25.6 What is
reference data?</h2>
<p class="first-paragraph">Reference data represents the snapshot of
the accumulated knowledge at a given point in time.</p>
<p>Here is worth to note that the information about the human genome,
because of its importance to society, is treated quite differently
than information for just about all other genomes. The human genome
related data has most &ldquo;standardization&rdquo; applied to it -
though mostly by necessity.</p>
<p>Even so painful silly disagreements remain, for example, some
organization may label human chromosomes by word + number (for
example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr2</font></font>,
&hellip;) other organizations name chromosomes by numbers alone (for
example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>,
&hellip;). The reasons for these and the many other discrepancies are
numerous, in a nutshell, it can be condensed down to the following:
it takes a lot of work to standardize, yet there too few incentives
to reward those that choose to do so. From your perspective as a
bioinformatician, it means that you will need to navigate these
troubled waters by learning to recognize and navigate between the
many shoals.</p>
<p>For other organisms that have a large following, various
organizations stepped in to standardize the representation of this
accumulated knowledge. The &ldquo;local&rdquo; standards are
occasionally better enforced, often that only means that the amount
of accumulated data is too low, it is easier to &ldquo;police&rdquo;
it.</p>
<h2 class="western"><a name="what-are-genomic-builds"></a>25.7 What
are genomic builds?</h2>
<p class="first-paragraph">As more information is discovered previous
representations of the genome may need to be corrected and
rearranged. Genomic builds represent an &ldquo;edition,&rdquo; a
snapshot of the information in time. Especially when obtaining data
from disparate sources, it is essential to ensure that all evidence
refers to the same baseline information.</p>
<p>When a genomic build has changed the information associated with
the genome must be changed as well. For example, just adding a single
base at the beginning of a genome means that the coordinates of all
subsequent elements must be changed, shifted by one.</p>
<p>Because genomes get rearranged in various and complex ways:
inversions, insertions, deletions, relocations, sometimes in an
overlapping manner, remapping a coordinate to a new location turns
out to be a surprisingly challenging operation. Note that even though
if the sequence were not to change substantially the coordinates may
end up altered in a manner that could be challenging or even
impossible to reconcile with prior data. Some locations in the
previous version of a genome may &ldquo;not exist&rdquo; in the new
genome.</p>
<h2 class="western"><a name="is-there-a-list-of-all-resources-1"></a>25.8
Is there a list of &ldquo;all&rdquo; resources?</h2>
<p class="first-paragraph">Once a year the journal Nucleic Acids
Research publishes its so-called &ldquo;database issue&rdquo;. Each
article of this issue of the journal will provide an overview and
updates about a specific database written by the maintainers of that
resource.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://academic.oup.com/nar/issue/47/D1">View
	the NAR: 2019 Database Issue</a></font>.</p>
</ul>
<h2 class="western"><a name="should-download-data-first-or-get-data-on-demand"></a>
25.9 Should download data first or get data on-demand?</h2>
<p class="first-paragraph">The size and frequency of the need for a
data will determine your use case. Typically we only perform
on-demand download for small (&lt;10M) datasets and during
exploratory data analyses. We recommend creating a directory for your
reference data that you would store the references for access across
all your projects).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a directory of a human reference build.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/refs/hg38</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get chromosome 22 and store it locally.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz
| gunzip -c &gt; ~/refs/hg38/chr22.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Look at the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
~/refs/hg38/chr22.fa | head -4</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Print statistics on the sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">infoseq
~/refs/hg38/chr22.fa</font></font></p>
<p class="first-paragraph">As always remember that rule &ldquo;Make
it fast.&rdquo; Always test your processes on a subset of the genome
like chromosome 22 to have a fast turnaround.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="human-and-mouse-genomes"></a>26 Human
and mouse genomes</h1>
<p class="first-paragraph">As we mentioned before the information
about the human and mouse genomes are treated in a more standardized
fashion than the data from just about all other genomes. Even if you
are not working with the human and mouse genomes, it helps if you
familiarize yourself with the design/distribution choices made for
this data.</p>
<p>The mouse genome has a special designation as it serves as the
most studied experimental animal model for the human genome. For that
reason, many of the standards of the human genome have been adopted
for the mouse genome as well.</p>
<h2 class="western"><a name="how-many-genomic-builds-does-the-human-genome-have"></a>
26.1 How many genomic builds does the human genome have?</h2>
<p class="first-paragraph">38 as of today. Alas not only do different
genomic builds exist, but various data distribution resources may
also disagree on how to label the same data. For example, here are
the last four genomic builds as marked at NCBI vs.&nbsp;UCS Genome
Browser</p>
<ul>
	<li/>
<p class="compact">Build 35 released in 2004 is labeled as
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NCBI35</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg17</font></font></p>
	<li/>
<p class="compact">Build 36 released in 2006 is labeled as
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NCBI36</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg18</font></font></p>
	<li/>
<p class="compact">Build 37 released in 2009 is labeled as
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh37</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font></p>
	<li/>
<p class="compact">Build 38 released in 2013 is labeled as
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg38</font></font></p>
</ul>
<p class="first-paragraph">Note the painful inconsistency of earlier
years; it took until 2013 to get to the point where at least the
number matches across. Then note just how long ago was build 37
released. It is about a decade old. You&rsquo;d be surprised to know
that this genome build is still in active use even though a better
build has been available.</p>
<p>Within each year <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/grc/help/patches/">subsequent
corrections (patches)</a></font> may be applied to the genome that
does not alter the length of the genome or the coordinates of
elements.</p>
<h2 class="western"><a name="why-is-the-genomic-build-hg19-still-in-use"></a>
26.2 Why is the genomic build <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font>
still in use?</h2>
<p class="first-paragraph">Scientists that want to build upon results
that integrate with information obtained older genome build may use
that old genome as a reference, rather than using a new genome then
&ldquo;transferring&rdquo; the results to the new genome.</p>
<h2 class="western"><a name="should-we-use-the-old-version-of-a-genome"></a>
26.3 Should we use the old version of a genome?</h2>
<p class="first-paragraph">No.&nbsp;We would actively discourage the
practice. While it is conceivable that there are legitimate uses of
relying on obsolete genomic build such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font>,
we can&rsquo;t help but wonder whether it is a lack of awareness and
ignorance that allows publications still using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font>
to clear the review process</p>
<p>Scientists keep following the practice; hence you will need to
know how to remap information between genomes.</p>
<h2 class="western"><a name="how-do-we-transfer-genomic-coordinates-between-builds"></a>
26.4 How do we transfer genomic coordinates between builds?</h2>
<p class="first-paragraph">The term used to describe this process is
called &ldquo;lifting over&rdquo; coordinates. There are web and
command line applications that you may use:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genome/tools/remap">liftOver</a></font>
	from UCSC (web and command line)</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genome/tools/remap">remap</a></font>
	from NCBI (web)</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://crossmap.sourceforge.net/">crossmap</a></font>
	(command line)</p>
</ul>
<p class="first-paragraph">The &ldquo;lifting over&rdquo; process
requires the presence of a precomputed &ldquo;chain&rdquo; file
(chain data) that describes how the new genome is different from the
old one.</p>
<p>While the web interfaces may suffice for some tasks, we recommend
command line tools that allow you to generate reproducible analyses.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install crossmap -y</font></font></p>
<p class="first-paragraph">Run the tool:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CrossMap.py</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Program:
CrossMap (v0.3.3)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Description:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 CrossMap is a program for convenient conversion of genome
coordinates and genome annotation files between assemblies (e.g.,
lift from human hg18 to hg19 or vice versa). It supports file in BAM,
SAM, BED, Wiggle, BigWig, GFF, GTF and VCF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 format.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">To perform a remapping [get a
chainfile][chains]</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the chain file that maps from hg19 to hg38.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://hgdownload.soe.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a test data file that will be remapped.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/data/ucsc/test.hg19.bed</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run the remapping process.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CrossMap.py
bed hg19ToHg38.over.chain.gz test.hg19.bed  test.hg38.bed</font></font></p>
<p class="first-paragraph">The commands above will remap
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">test.hg19.bed</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">test.hg38.bed</font></font>.
As an example the coordinates:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1
   65886334    66103176</font></font></p>
<p class="first-paragraph">were changed to:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1
   65420651    65637493</font></font></p>
<h2 class="western"><a name="human-gene-naming"></a>26.5 Human gene
naming</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://www.genenames.org/">HUGO
Gene Nomenclature Committee</a></font> is the only worldwide
authority that assigns standardized nomenclature to human genes. The
names assigned by HGNC are typically formed from a so-called stem (or
root) symbol that is used as the basis for a series of approved
symbols that are defined as members of either a functional or
structural gene family. For example <i>CYP</i>: cytochrome P450; <i>HOX</i>:
homeobox; <i>DUSP</i>: dual specificity phosphatase; <i>SCN2A</i>:
sodium channel voltage-gated type II alpha 2 polypeptides etc.</p>
<p>Since the mouse and rat genomes are the closest human models and
are widely studied, making human and mouse gene names follow similar
rules has always been a priority of the <font color="#4f81bd"><a href="http://www.informatics.jax.org/mgihome/nomen/">Mouse
Gene Nomenclature Committee (MGNC)</a></font></p>
<p>The [Vertebrate Gene Nomenclature Committee (VGNC)] <font color="#4f81bd"><a href="http://vertebrate.genenames.org/">vgnc</a></font>
is an extension of the established HGNC (HUGO Gene Nomenclature
Committee) project that names human genes. VGNC is responsible for
assigning standardized names to genes in vertebrate species that
currently lack a nomenclature committee. The current prototype VGNC
naming species is the chimpanzee.</p>
<h2 class="western"><a name="which-human-data-source-should-i-trust"></a>
26.6 Which human data source should I trust?</h2>
<p class="first-paragraph">First, recognize that while some data is
nearly identical (for example the sequence data differs mainly in its
completeness, information derived from processing results may be
notably different. Also, different groups may rearrange data to serve
their needs better, further complicating the picture.</p>
<p>When it comes to the human genome scientists typically use one of
the following resources:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.gencodegenes.org/">GENCODE</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/refseq/">NCBI:
	RefSeq</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://hgdownload.soe.ucsc.edu/downloads.html">UCSC</a></font></p>
</ul>
<p class="first-paragraph"><font color="#4f81bd"><a href="https://www.ensembl.org/">Ensembl</a></font>,
the joint project between EMBL-EBI and the Wellcome Trust Sanger
Institute to publish data and software distributes the <font color="#4f81bd"><a href="https://www.gencodegenes.org/">GENCODE</a></font>
data as their human and mouse annotations. The data in Ensembl is
published via release versions, currently at version 95:</p>
<ul>
	<li/>
<p class="compact">Ensembl FTP:
	<font color="#4f81bd"><a href="ftp://ftp.ensembl.org/pub/release-95/">ftp://ftp.ensembl.org/pub/release-95/</a></font></p>
</ul>
<h2 class="western"><a name="which-humanmouse-genome-data-should-i-be-using"></a>
26.7 Which human/mouse genome data should I be using?</h2>
<p class="first-paragraph">Most scientists are selecting a data
source to integrate with other data that is present in that data
source. The choice relates to the domain of application and the
subculture that a scientist is a part of.</p>
<p>For example projects with medical and clinical applications tend
to use <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a></font>
as typically being the most conservative yet best integrated with all
available knowledge. Scientists interested in comparative genomics
may favor <font color="#4f81bd"><a href="http://hgdownload.soe.ucsc.edu/downloads.html">UCSC</a></font>.
<font color="#4f81bd"><a href="https://www.gencodegenes.org/">GENCODE</a></font>
is the choice of those that work on so &ldquo;model&rdquo; organism
or those that wish to use a more frequently updated and more
comprehensive database.</p>
<h2 class="western"><a name="is-there-a-better-resource-for-human-annotations"></a>
26.8 Is there a better resource for human annotations?</h2>
<p class="first-paragraph">The unfortunately named <font color="#4f81bd"><a href="http://ccb.jhu.edu/chess/">CHESS</a></font>
project, a name that is an acronym for <b>Comprehensive Human
Expressed SequenceS</b> was published as <font color="#4f81bd"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1590-2">CHESS:
a new human gene catalog curated from thousands of large-scale RNA
sequencing experiments reveals extensive transcriptional noise
(Genome Biology, 2018)</a></font> claims to have created a new list
of human genes and transcripts.</p>
<p>Pet peeve: Naming this resource as CHESS was incredibly
shortsighted and unfortunate as it will make searching and finding
relevant hits for it online virtually impossible.</p>
<p>This paper and data repository are worth studying in more detail.
It is well written article that provides a detailed background on the
current challenges for identifying and annotating the human genome.
Besides, it offers what we believe to be the best current annotations
for the human genome. We do foresee other resources integrating the
information that this paper has published.</p>
<h2 class="western"><a name="how-do-i-access-ncbi-refseq-and-genbank"></a>
26.9 How do I access NCBI RefSeq and GenBank?</h2>
<p class="first-paragraph">Entrez is the interface into the data
stored at NCBI.</p>
<ul>
	<li/>
<p class="compact">NCBI web: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/">https://www.ncbi.nlm.nih.gov/</a></font></p>
	<li/>
<p class="compact">NCBI FTP: <font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/">ftp://ftp.ncbi.nlm.nih.gov/</a></font></p>
	<li/>
<p class="compact">See the <font color="#4f81bd"><a href="https://academic.oup.com/nar/article/44/D1/D733/2502674">RefSeq
	publication</a></font> in the database issue of NAR</p>
</ul>
<p class="first-paragraph">NCBI and Entrez (the interface into
GenBank) are the best resources to obtain any sequence that humanity
has ever found and stored. But in the cost of handling that immense
variety is that you can forget about expecting any naming scheme that
&ldquo;makes sense.&rdquo; Chromosome 1 of the human genome will not
be called as such; instead it will be called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000001405.38</font></font>
- suffice to say you&rsquo;ll have a hard time working with data from
outside of NCBI.</p>
<p>Entrez is excellent at getting you any version of any gene that
was ever known - even if this means getting thousands of hits, most
of which may be unreliable.</p>
<h2 class="western"><a name="what-can-i-get-from-ensembl"></a>26.10
What can I get from ENSEMBL?</h2>
<p class="first-paragraph">Ensembl is the interface into the data
store at EBI.</p>
<ul>
	<li/>
<p class="compact">Ensembl web:
	<font color="#4f81bd"><a href="http://useast.ensembl.org/index.html">http://useast.ensembl.org/index.html</a></font></p>
	<li/>
<p class="compact">Ensembl FTP: <font color="#4f81bd"><a href="ftp://ftp.ensembl.org/pub/">ftp://ftp.ensembl.org/pub/</a></font></p>
</ul>
<p class="first-paragraph">Ensembl has the most detailed annotations
of the genomes. Ensembl operates <font color="#4f81bd"><a href="http://useast.ensembl.org/info/website/news.html">via
releases</a></font>. Compared to GenBank and RefSeq Ensemble stores
only a subset of the most studied organisms.</p>
<h2 class="western"><a name="a-working-strategy-for-finding-reference-information"></a>
26.11 A working strategy for finding reference information</h2>
<ol>
	<li/>
<p class="compact">Locate the resource</p>
	<li/>
<p class="compact">Understand the file structure</p>
	<li/>
<p class="compact">Write the commands to get the file you need</p>
</ol>
<p class="first-paragraph">Example: Visit
<font color="#4f81bd"><a href="https://www.gencodegenes.org/human/">https://www.gencodegenes.org/human/</a></font>.
Say you want the transcript sequences in FASTA format. Find the URL
that you need. Make a note and store these commands for the future:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.transcripts.fa.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
gencode.v29.transcripts.fa.gz</font></font></p>
<p class="first-paragraph">Sometimes/often the descriptions are
insufficient and you need to download and investigate the data to
fully understand what it contains.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="downloading-complete-genomes"></a>27
Downloading complete genomes</h1>
<p class="first-paragraph">In this chapter, we&rsquo;ll go over our
recommendations for getting complete genomic data.</p>
<p>As described before, there are many data sources. Frequently the
same data is offered for download from multiple locations. You might
even ask yourself, do we even get the same data from different
sources? Short answer is usually no - while the main information may
be the same (for example the sequence itself is the same) usually
data obtained from different sources include additional attributes.</p>
<h2 class="western"><a name="model-organisms-have-dedicated-sources"></a>
27.1 Model organisms have dedicated sources</h2>
<p class="first-paragraph">If you need data for a model organism
(human/mouse/rat/arabidopsis/drosophila etc.) then first explore
sites that other scientists working specifically in your field use.
As we mentioned before the chromosome naming and other labels may be
different across different data sources. The classic example of a
<i>little bump that upsets the applecart</i> is the choice of
chromosome naming in the human genome. For example chromosome 22 will
be named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrom22</font></font>
when getting data from UCSC versus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">22</font></font>
when getting the same data from Ensembl. This seemingly minor
difference, may, unfortunately, can have surprisingly annoying
consequences when running various tools that are unable to
automatically reconcile the names.</p>
<p>In addition different data sources may also offer information that
has been integrated with their naming scheme. Whenever you need build
upon results produced by others we recommend using the same naming
scheme or fight in the tar pit of inconsistencies that will slowly
drag you under.</p>
<h2 class="western"><a name="a-word-of-warning"></a>27.2 A word of
warning</h2>
<p class="first-paragraph">As it turns out, selecting the &ldquo;correct&rdquo;
genome can be more complicated than one might think; see the post
titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use">Which
	human reference genome to use?</a></font></p>
</ul>
<p class="first-paragraph">for a summary of the potential pitfalls.
We note that the blog post was written in 2017; thus the
recommendation made then is probably outdated.</p>
<h2 class="western"><a name="ucsc-genomes"></a>27.3 UCSC genomes</h2>
<p class="first-paragraph">USCS is an abbreviation for University of
California Santa Cruz. Historical context:</p>
<p class="block-text-western">On June 22, 2000, UCSC and the other
members of the International Human Genome Project consortium
completed the first working draft of the human genome assembly,
forever ensuring free public access to the genome and the information
it contains. A few weeks later, on July 7, 2000, the newly assembled
genome was released on the web at <font color="#4f81bd"><a href="http://genome.ucsc.edu/">http://genome.ucsc.edu</a></font>,
along with the initial prototype of a graphical viewing tool, the
UCSC Genome Browser. In the ensuing years, the website has grown to
include a broad collection of vertebrate and model organism
assemblies and annotations, along with a large suite of tools for
viewing, analyzing and downloading data.</p>
<p class="first-paragraph">The UCSC genome download sites offer
various data types, whole genomes where all chromosomes are together
or data broken down by chromosomes. There is a fair bit of clicking
around to figure out what is located where, but the naming is clear;
the readme files document all the information.</p>
<p>Visit the website below to familiarize yourself with its layout.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://hgdownload.soe.ucsc.edu/downloads.html">https://hgdownload.soe.ucsc.edu/downloads.html</a></font></p>
</ul>
<p class="first-paragraph">Watch the URL bar for changes, on any file
you may right click to copy link address. To fet chromosome 22 of the
human genome you may write:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get chromosome 22</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz</font></font></p>
<p class="first-paragraph">Many (but not all) tools can handle
compressed references, you may not need to decompress the file. I
usually always decompress files when harddrive space is not a
constraint, that way I avoid future complications:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Uncompress the file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
chr22.fa.gz</font></font></p>
<p class="first-paragraph">You may also decompress a file &ldquo;on
the fly&rdquo;, while keeping the file in its original format:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
-c  chr22.fa.gz | head -1</font></font></p>
<p class="first-paragraph">Print statistics on the downloaded FASTA
file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats chr22.fa.gz</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
        format  type  num_seqs     sum_len     min_len     avg_len   
 max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr22.fa.gz
 FASTA   DNA          1  50,818,468  50,818,468  50,818,468 
50,818,468</font></font></p>
<p class="first-paragraph">Get the GTF file for known genes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download a GTF file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.knownGene.gtf.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Uncompress the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
hg38.knownGene.gtf.gz</font></font></p>
<p class="first-paragraph">print some statistics on the file</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
How many lines in the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
hg38.knownGene.gtf | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">3091269</font></font></p>
<p class="first-paragraph">How many features per chromosome (and also
can find how many chromosomes it has):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
hg38.knownGene.gtf | cut -f 1 | sort-uniq-count-rank</font></font></p>
<p class="first-paragraph">the output summarized the number of
features per chromosome</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">260686
 chr1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">211594
 chr2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">185122
 chr3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">172067
 chr17</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">170613
 chr19</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">168949
 chr11</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">164447
 chr12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">and we can see that it has unlocalized
chromosome segments listed separately. How many feature types:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
hg38.knownGene.gtf | cut -f 3 | sort-uniq-count-rank</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1486754
exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">827935
 CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">247541
 transcript</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">165544
 3UTR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">164643
 5UTR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">110310
 start_codon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">88542
  stop_codon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<h2 class="western"><a name="understand-the-data-you-download"></a>27.4
Understand the data you download</h2>
<p class="first-paragraph">Always investigate and understand the data
you obtain. Run simple tools like the ones above to familiarize
yourself with properties that you may not be aware of.</p>
<p>Use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats</font></font> for FASTA file, and simple column oriented Unix
tools (cut/grep/sort/uniq) for column based files.</p>
<h2 class="western"><a name="ensemble-genomes"></a>27.5 Ensemble
genomes</h2>
<p class="first-paragraph">Historical context:</p>
<p class="block-text-western">Ensembl genome database project is a
scientific project at the European Bioinformatics Institute, which
was launched in 1999 in response to the imminent completion of the
Human Genome Project.</p>
<p class="block-text-western">Ensembl creates, integrates, and
distributes reference datasets and analysis tools that enable
genomics</p>
<p class="first-paragraph">Ensemble gene/transcripts models are
typically better curated and more detailed than those from other
sources. Most RNA-Seq analyses use Ensemble gene models.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ensembl.org/index.html">https://www.ensembl.org/index.html</a></font></p>
</ul>
<p class="first-paragraph">The main FTP entry point is:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://ftp.ensembl.org/pub/">http://ftp.ensembl.org/pub/</a></font></p>
</ul>
<p class="first-paragraph">Ensemble operates on numbered releases.
When downloading data for an organism all of it should from the same
release. For example, release 104 was published on March 30, 2021:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://ftp.ensembl.org/pub/release-104/">http://ftp.ensembl.org/pub/release-104/</a></font></p>
</ul>
<p class="first-paragraph">Navigate the link above and familiarize
yourself with the structure. You&rsquo;ll immediately note the the
Ensembl layout is arguably, quite counterintuitive. Instead of
grouping data by the organism, where all information on an organism
would be in one directory, they are grouping data by type. To collect
all the data for an organism, you&rsquo;ll have to click around and
visit many different pages.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the FASTA file for chromosome 22 of the Human genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://ftp.ensembl.org/pub/release-104/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.22.fa.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get GFF file for chromosome 22 of Human genome for release 104</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://ftp.ensembl.org/pub/release-104/gff3/homo_sapiens/Homo_sapiens.GRCh38.104.chromosome.22.gff3.gz</font></font></p>
<h2 class="western"><a name="ncbi-genomes"></a>27.6 NCBI genomes</h2>
<p class="first-paragraph">NCBI: The National Center for
Biotechnology Information is part of the United States National
Library of Medicine, a branch of the National Institutes of Health.
It is approved and funded by the government of the United States.</p>
<p>Both UCSC and Ensemble focus on the main model organsisms. The
NCBI genome download page is a handy resource when accessing the
largest number genomes.</p>
<p class="block-text-western">The genomes FTP site offers a
consistent core set of files for the genome sequence and annotation
products of all organisms and assemblies in scope. It supports
download needs such as:</p>
<ul>
	<li/>
<p class="block-text-western">Retrieve the unmasked or
	soft-masked genome sequence for a specific genome assembly</p>
	<li/>
<p class="block-text-western">Retrieve GenBank or RefSeq Gene,
	RNA and protein annotation for a specific organism and a specific
	assembly, or a specific RefSeq annotation release</p>
	<li/>
<p class="block-text-western">Retrieve annotation in GenBank
	flat-file, GFF or GTF format</p>
	<li/>
<p class="block-text-western">Matching sequence identifiers in
	FASTA &amp; GFF or GTF files to facilitate RNA-Seq and other
	analyses</p>
</ul>
<ul>
	<li/>
<p class="compact">Genome download help:
	<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/">https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/</a></font></p>
</ul>
<p class="first-paragraph">Main entry point:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://ftp.ncbi.nlm.nih.gov/genomes/">https://ftp.ncbi.nlm.nih.gov/genomes/</a></font></p>
</ul>
<p class="first-paragraph">There is a &ldquo;main&rdquo; file, that
summarizes all genomes at NCBI. The file is called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">assembly_summary_refseq.txt</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the assembly report</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt</font></font></p>
<p class="first-paragraph">The report is a tabular file with 22
columns. There is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ASSEMBLY_REPORTS/README_assembly_summary.txt</font></font>
file on the same site that contains information on what each column
means, but you can also figure it out yourself by poking and prodding
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat/cut/grep</font></font>.</p>
<p>As an example, let&rsquo;s get data fro Bacillus Cereus strain
ATCC 10987:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
How many Bacciuls</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
assembly_summary_refseq.txt | grep cereus | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1168</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find the strain 10987</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
assembly_summary_refseq.txt | grep cereus | grep 10987</font></font></p>
<p class="first-paragraph">While the above is straithforward do note
that grep will match <i>anywhere</i> in the text, thus always double
check that it is matching what you think it should be. Our command
now prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1
   PRJNA224116    SAMN02603979      na 222523 1396   Bacillus cereus
ATCC 10987 strain=ATCC 10987     latest Complete Genome    Major 
Full   2004/02/23 ASM800v1   TIGR   GCA_000008005.1    identical 
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/005/GCF_000008005.1_ASM800v1</font></font></p>
<p class="first-paragraph">The last column of the output above is the
link to the data.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
assembly_summary_refseq.txt | grep cereus | grep 10987 | cut -f 20</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/005/GCF_000008005.1_ASM800v1</font></font></p>
<p class="first-paragraph">Visit the link in the last column. In the
url you may need to replace <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ftp://</font></font>
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">https://</font></font>
when accessing it with your your browser:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/005/GCF_000008005.1_ASM800v1">http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/005/GCF_000008005.1_ASM800v1</a></font></p>
</ul>
<p class="first-paragraph">You will then see the data as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Parent
Directory                                                     
-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_assembly_report.txt
    2019-12-28 16:59 
1.2K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_assembly_stats.txt
     2019-12-28 16:59 
4.4K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_cds_from_genomic.fna.gz
2020-10-28 18:06  1.6M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_feature_count.txt.gz
   2020-10-28 18:06 
280</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_feature_table.txt.gz
   2020-10-28 18:06  264K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_genomic.fna.gz
         2020-02-04 18:17 
1.5M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_genomic.gbff.gz
        2020-10-28 18:06 
3.7M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_genomic.gff.gz
         2020-10-28 18:06 
401K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_genomic.gtf.gz
         2020-10-28 18:06 
435K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_protein.faa.gz
         2020-10-28 18:06 
971K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_protein.gpff.gz
        2020-10-28 18:06 
2.4M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_rna_from_genomic.fna.gz
2020-10-28 18:06  7.2K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000008005.1_ASM800v1_translated_cds.faa.gz
  2020-10-28 18:06  1.1M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">README.txt
                                      2020-09-02 16:26  
43K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">annotation_hashes.txt
                           2020-10-28 18:06  410</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">assembly_status.txt
                             2021-08-25 02:51   14</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">md5checksums.txt
                                2020-10-28 18:06  1.1K</font></font></p>
<p class="first-paragraph">Get the GFF file with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/008/005/GCF_000008005.1_ASM800v1/GCF_000008005.1_ASM800v1_genomic.gff.gz</font></font></p>
<h2 class="western"><a name="download-ncbi-genomes-via-rsync"></a>27.7
Download NCBI Genomes via <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rsync</font></font></h2>
<p class="first-paragraph">Impressively GenBank offers <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rsync</font></font>
downloads that would be used like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rsync
--copy-links --recursive --times --verbose
rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/696/305/GCF_001696305.1_UCN72.1
genome/</font></font></p>
<p class="first-paragraph">the command will creeate the directory:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome/GCF_001696305.1_UCN72.1</font></font></p>
<p class="first-paragraph">that contains all the files for the
genomic build.</p>
<h2 class="western"><a name="download-ncbi-genomes-by-taxonomic-id"></a>
27.8 Download NCBI Genomes by taxonomic id</h2>
<p class="first-paragraph">The need to ob There is a &ldquo;cottage
industry&rdquo; for doing the job that NCBI should be doing, trying
to bring some logic and sanity to data download automation. A proud
entrant in the field, is a script, aptly called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ncbi-genome-download</font></font>.
Its main purpose is to connect taxonomical information to sequence
information and to allow downloading data by taxonomical
ranks/subtrees (clades).</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/kblin/ncbi-genome-download">https://github.com/kblin/ncbi-genome-download</a></font></p>
</ul>
<p class="first-paragraph">Refer to the guide above for usage
details; in a nutshell, the script requires a taxonomical rank and it
downloads genomes that are belong to that rank:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install the software</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install NCBI-genome-download</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Show how many sequences would be downloaded</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NCBI-genome-download
--genera Streptomyces bacteria --dry-run | head</font></font></p>
<p class="first-paragraph">prints that 2305 assemblies match the
taxonomy id:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Considering
the following 2305 assemblies for download:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_003208035.1
   Streptomyces actuosus  ATCC 25421</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_016921115.1
   Streptomyces actuosus  VRA1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000156475.1
   Streptomyces albidoflavus  J1074</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000359525.1
   Streptomyces albidoflavus  J1074</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000717375.1
   Streptomyces albidoflavus  NRRL B-2307</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000719435.1
   Streptomyces albidoflavus  NRRL F-5618</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000719495.1
   Streptomyces albidoflavus  NRRL F-5621</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000719955.1
   Streptomyces albidoflavus  NRRL B-1271</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_001509565.1
   Streptomyces albidoflavus  NRRL WC-3066</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">if we were to run it as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ncbi-genome-download
--genera Streptomyces bacteria</font></font></p>
<p class="first-paragraph">will get a directory structure with 2305
files, of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">find
. | head</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria/GCF_014207445.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria/GCF_014207445.1/MD5SUMS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria/GCF_015601545.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria/GCF_015601545.1/MD5SUMS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./refseq/bacteria/GCF_001715295.1</font></font></p>
<p class="first-paragraph">where each file is a bacterial genome from
the genera of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptomyces</font></font></p>
<h2 class="western"><a name="refgenie---a-reference-genome-resource-manager"></a>
27.9 RefGenie - a reference genome resource manager</h2>
<p class="first-paragraph">A recent publication <font color="#4f81bd"><a href="https://academic.oup.com/gigascience/article/9/2/giz149/5717403">Refgenie:
a reference genome resource manager, GigaScience 2020</a></font>
introduces a software tool for managing reference genomes.</p>
<p class="block-text-western">Refgenie manages storage, access, and
transfer of reference genome resources. It provides command-line and
Python interfaces to download pre-built reference genome &ldquo;assets&rdquo;,
like indexes used by bioinformatics tools. It can also build assets
for custom genome assemblies. Refgenie provides programmatic access
to a standard genome folder structure, so software can swap from one
genome to another.</p>
<p class="first-paragraph">One reason for using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie</font></font>
is that you can also download precomputed alignment index files (bwa,
salmon) etc. These are files that may take a long time for you to
produce yourself.</p>
<p>Documentation</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://refgenie.databio.org/en/latest/">http://refgenie.databio.org/en/latest/</a></font></p>
</ul>
<p class="first-paragraph">The effort is notable and timely as it
attempts to bring some sanity to the process. That being said, I am
not entirely convinced that it has &ldquo;legs&rdquo; and a future.
Maintaining services like these should not be the responsibility of
individual scientists and usually get abandoned due to a lack of
resources. But let&rsquo;s keep an open mind and attempt to use
refgenie; it is a step in the right direction:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install refgenie</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install refgenie</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the directory to store data in</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/refs/</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This should be added to your .bashrc</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">export
REFGENIE=~/refs/genomes.yaml</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Initialize the files</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
init -c $REFGENIE</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
List the genomes</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
listr</font></font></p>
<p class="first-paragraph">the last command prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">                   
                                             <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Remote
refgenie assets</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                                        Server URL:
http://refgenomes.databio.org</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9487;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9523;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9491;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9475;
genome           &#9475; assets                                      
                                                                     
                   
&#9475;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9505;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9543;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9513;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
rCRSd            &#9474; fasta, bowtie2_index, bwa_index, star_index,
hisat2_index, bismark_bt2_index                                      
                    &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
hg38             &#9474; fasta, gencode_gtf, refgene_anno, dbsnp,
ensembl_gtf, ensembl_rb, suffixerator_index, bwa_index,
bowtie2_index, dbnsfp, star_index,    &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
                 &#9474; fasta_txome, hisat2_index,
cellranger_reference, bismark_bt2_index, salmon_partial_sa_index,
tgMap, salmon_sa_index                    &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
human_repeats    &#9474; fasta, suffixerator_index, bowtie2_index,
bwa_index, star_index, hisat2_index, bismark_bt2_index               
                       &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
mouse_chrM2x     &#9474; fasta, suffixerator_index, bowtie2_index,
bwa_index, star_index, hisat2_index, bismark_bt2_index               
                       &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
hg18_cdna        &#9474; fasta, kallisto_index                       
                                                                     
                    &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
hs38d1           &#9474; fasta, suffixerator_index, bowtie2_index,
bwa_index, star_index, hisat2_index, bismark_bt2_index               
                       &#9474;</font></font></p>
<p class="first-paragraph">let&rsquo;s make use of the information:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Pull a test file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
pull hg38_chr22/fasta</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Pull the GenCode gene annotations</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
pull hg38/gencode_gtf</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Pull the Ensembl annotations</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
pull hg38/ensembl_gtf</font></font></p>
<p class="first-paragraph">What files do I have?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
list</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                      Local refgenie assets</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       Server subscriptions:
http://refgenomes.databio.org</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9487;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9523;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9491;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9475;
genome               &#9475; assets                                  
  
&#9475;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9505;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9543;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9473;&#9513;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
hg38                 &#9474; gencode_gtf, ensembl_gtf                
   &#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9474;
hg38_chr22           &#9474; fasta                                   
  
&#9474;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&#9492;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9524;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9496;</font></font></p>
<p class="first-paragraph">Where are the files (what is the path)?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">refgenie
seek hg38_chr22/fasta</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/refs/alias/hg38_chr22/fasta/default/hg38_chr22.fa</font></font></p>
<p class="first-paragraph">So what is the statistics on that:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats `refgenie seek hg38_chr22/fasta`</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
          format  type  num_seqs     sum_len  min_len     avg_len    
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg38_chr22.fa
 FASTA   DNA         60  42,746,242      340  712,437.4  31,264,301</font></font></p>
<p class="first-paragraph">wait what? that looks different from the
data I got from UCSC:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
        format  type  num_seqs     sum_len     min_len     avg_len   
 max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr22.fa.gz
 FASTA   DNA          1  50,818,468  50,818,468  50,818,468 
50,818,468</font></font></p>
<p class="first-paragraph">Ok, see, the &ldquo;same&rdquo; data from
different sources produce different results. It appears that the data
from refgenie is not scaffolded, whereas the &ldquo;official&rdquo;
data is. But the official data places `N&rsquo;s in the genome. We
saw this before,</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;chr22</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font></p>
<p class="first-paragraph">A good percentage of the genome are Ns &hellip;
so which data is better? Not so clear. Frankly, I&rsquo;d prefer the
one without Ns.</p>
<h2 class="western"><a name="using-pyensembl-to-access-ensembl"></a>27.10
Using PyEnsembl to access Ensembl</h2>
<p class="first-paragraph">We include this tool last, for
completeness sake. While the tool offers command line interfaces, it
is probably best suited for those that also program in Python as the
tool is also a library that provides additional functionality to
operate on Ensembl data from Python programs.</p>
<p>PyEnsembl is a Python interface to Ensembl reference genome
metadata such as exons and transcripts. PyEnsembl downloads GTF and
FASTA files from the Ensembl FTP server and loads them into a local
database. PyEnsembl can also work with custom reference data
specified using user-supplied GTF and FASTA files.</p>
<p>We should note that the tool above is not an official software
created by Ensembl - and such tools may go out of date.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://pypi.org/project/pyensembl/">https://pypi.org/project/pyensembl/</a></font></p>
</ul>
<p class="first-paragraph">Often it is a good idea to create a new
environment to test out software. It is more for precaution and is
not always necessary but be advised that some software may have
unwanted effects on other software in the same environment. Thus some
caution is recommended:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a test environment to see how pyensemble works</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create -y --name test python=3.8</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Activate the test environment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate test</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install pyensembl</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install pyensembl</font></font></p>
<p class="first-paragraph">To install files for a release use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pyensembl
install --release 104 --species human</font></font></p>
<p class="first-paragraph">Moving on, to list installed files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pyensembl
list</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">EnsemblRelease(release=102,
species='homo_sapiens'):
/Users/ialbert/Library/Caches/pyensembl/GRCh38/ensembl102</font></font></p>
<p class="first-paragraph">list the contents of the directory with
the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
/Users/ialbert/Library/Caches/pyensembl/GRCh38/ensembl102</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.102.gtf.db
          
Homo_sapiens.GRCh38.ncrna.fa.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.102.gtf.gz
          
Homo_sapiens.GRCh38.ncrna.fa.gz.pickle</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.cdna.all.fa.gz
      
Homo_sapiens.GRCh38.pep.all.fa.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homo_sapiens.GRCh38.cdna.all.fa.gz.pickle
Homo_sapiens.GRCh38.pep.all.fa.gz.pickle</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pyensembl</font></font>
also creates database files (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.db</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.pickle</font></font>
that are to be used from Python programs). We can still read the
standard bioinformatics files (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.gtf</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fa</font></font>).
What is in the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
-c
/Users/ialbert/Library/Caches/pyensembl/GRCh38/ensembl102/Homo_sapiens.GRCh38.102.gtf.gz
 | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!genome-build
GRCh38.p13</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!genome-version
GRCh38</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!genome-date
2013-12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!genome-build-accession
NCBI:GCA_000001405.28</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#!genebuild-last-updated
2020-09</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
  havana  gene    11869   14409   .   +   .   gene_id
&quot;ENSG00000223972&quot;; gene_version &quot;5&quot;; gene_name
&quot;DDX11L1&quot;; gene_source &quot;havana&quot;; gene_biotype
&quot;transcribed_unprocessed_pseudogene&quot;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
  havana  transcript  11869   14409   .   +   .   gene_id
&quot;ENSG00000223972&quot;; gene_version &quot;5&quot;;
transcript_id &quot;ENST00000456328&quot;; transcript_version &quot;2&quot;;
gene_name &quot;DDX11L1&quot;; gene_source &quot;havana&quot;;
gene_biotype &quot;transcribed_unprocessed_pseudogene&quot;;
transcript_name &quot;DDX11L1-202&quot;; transcript_source &quot;havana&quot;;
transcript_biotype &quot;processed_transcript&quot;; tag &quot;basic&quot;;
transcript_support_level &quot;1&quot;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
  havana  exon    11869   12227   .   +   .   gene_id
&quot;ENSG00000223972&quot;; gene_version &quot;5&quot;;
transcript_id &quot;ENST00000456328&quot;; transcript_version &quot;2&quot;;
exon_number &quot;1&quot;; gene_name &quot;DDX11L1&quot;; gene_source
&quot;havana&quot;; gene_biotype
&quot;transcribed_unprocessed_pseudogene&quot;; transcript_name
&quot;DDX11L1-202&quot;; transcript_source &quot;havana&quot;;
transcript_biotype &quot;processed_transcript&quot;; exon_id
&quot;ENSE00002234944&quot;; exon_version &quot;1&quot;; tag &quot;basic&quot;;
transcript_support_level &quot;1&quot;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
  havana  exon    12613   12721</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pyensemble</font></font>
tool has a number of limitations though, for example only a few
genomes have human readable species selectors: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chicken</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">dog</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rat</font></font>.
For all other cases we would need to explicitly import the files like
so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pyensembl
install  --reference-name &quot;FooRef&quot;  --gtf URL_OR_PATH 
--transcript-fasta URL_OR_PATH  --protein-fasta URL_OR_PATH</font></font></p>
<h2 class="western"><a name="how-to-summarize-differences-across-annotations"></a>
27.11 How to summarize differences across annotations?</h2>
<p class="first-paragraph">Perhaps you are wondering, what is the
difference between different GTF files across releases. A tool named
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gffcompare</font></font>
can assist with the comparison:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Uncompress the GFT reference files</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
-c
/Users/ialbert/Library/Caches/pyensembl/GRCh38/ensembl104/Homo_sapiens.GRCh38.104.gtf.gz
&gt; 104.gtf</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
-c
/Users/ialbert/Library/Caches/pyensembl/GRCh38/ensembl103/Homo_sapiens.GRCh38.103.gtf.gz
&gt; 103.gtf</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install gffcompare</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install gffcompare -y</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run a GFF comparison</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gffcompare
-r 104.gtf 103.gtf</font></font></p>
<p class="first-paragraph">Several files with names such as
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gffcompare*</font></font>
will be created. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gffcompare.stats</font></font>
file will contain information such as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#=
Summary for dataset: 104.gtf</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
    Query mRNAs :  236920 in   56614 loci  (211848 multi-exon
transcripts)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
           (22093 multi-transcript loci, ~4.2 transcripts per
locus)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reference mRNAs :  232645 in   56635 loci  (207560 multi-exon)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Super-loci w/ reference transcripts:    56596</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#-----------------|
Sensitivity | Precision  |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       Base level:    99.9     |    99.1    |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       Exon level:    99.8     |    99.3    |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     Intron level:   100.0     |    99.5    |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Intron
chain level:    99.7     |    97.6    |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 Transcript level:    99.6     |    97.8    |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      Locus level:    99.9     |    99.9    |</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    Matching intron chains:  206859</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      Matching transcripts:  231819</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
             Matching loci:   56593</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         Missed exons:      64/638389  (  0.0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          Novel exons:     808/644397  (  0.1%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       Missed introns:      45/388538  (  0.0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        Novel introns:     213/390330  (  0.1%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          Missed loci:       6/56635   (  0.0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Novel loci:      11/56614   (  0.0%)</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="automating-access-to-ncbi"></a>28
Automating access to NCBI</h1>
<h2 class="western"><a name="what-is-entrez"></a>28.1 What is Entrez?</h2>
<p class="first-paragraph">The NCBI data store is a repository of
gargantuan proportions that stores data far beyond the primary DNA
data that the NCBI&rsquo;s original mission called for.</p>
<p>Currently, NCBI provides storage, categorization, and context on
just about all life-sciences-oriented information, from raw datasets
to curated results. It even stores the content of most of the
scientific publications themselves.</p>
<p class="block-text-western">Entrez is NCBI&rsquo;s primary text
search and retrieval system that integrates the PubMed database of
biomedical literature with 39 other literature and molecular
databases including DNA and protein sequence, structure, gene,
genome, genetic variation, and gene expression.</p>
<p class="first-paragraph">With the large size come the challenges -
the interfaces can be cumbersome and complicated; it is typically not
easy to find the information that one is looking for and is easy to
get lost across the many pages. The challenges of data access make
command line access even more important as it makes all actions
explicit and reproducible.</p>
<h2 class="western"><a name="how-is-entrez-pronounced"></a>28.2 How
is Entrez pronounced?</h2>
<p class="first-paragraph">Technically, it is a French word, and the
French pronunciation would sound similar to <i>&ldquo;on tray.&rdquo;</i>
But if you say <i>&ldquo;let&rsquo;s search on-tray&rdquo;</i> no
will know what you mean. Thus we recommend pronouncing it as it
written: <i>&ldquo;en-trez&rdquo;</i></p>
<h2 class="western"><a name="how-do-we-automate-access-to-entrez"></a>
28.3 How do we automate access to Entrez?</h2>
<p class="first-paragraph">From early on NCBI offered a web API
interface called <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/books/NBK25500/">Entrez
E-utils</a></font> then later released a toolset called <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/books/NBK179288">Entrez
Direct</a></font>.</p>
<p class="block-text-western">Biostar News of the Day</p>
<p class="block-text-western"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Ncbi
Releases Entrez Direct, The Entrez Utilities On The Unix Command Line</a></font></p>
<p class="block-text-western">NCBI has just released Entrez Direct, a
new software suite that enables users to use the UNIX command line to
access NCBI databases directly, as well as to parse and format the
data to create customized downloads.</p>
<h2 class="western"><a name="how-is-data-organized-in-ncbi"></a>28.4
How is data organized in NCBI?</h2>
<p class="first-paragraph">The diversity of data sources and the need
to keep up with an evolving body of knowledge poses challenges when
trying to identify current and past information unambiguously.</p>
<p>As an example, search the NCBI nucleotide database for the words
&ldquo;<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Ebola+virus+1976">Ebola
virus 1976</a></font>&rdquo;. Among the first hits will be data that
starts with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LOCUS
      AF086833               18959 bp    cRNA    linear   VRL
13-FEB-2012</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DEFINITION
 Ebola virus - Mayinga, Zaire, 1976, complete genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCESSION
  AF086833</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VERSION
    AF086833.2  GI:10141003</font></font></p>
<p class="first-paragraph">Note the accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>
and the version number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2</font></font>.
Because the number does not contain an underscore we know that this
data is in GenBank but is not in RefSeq.</p>
<p class="block-text-western">An accession number, for example,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>,
applies to the complete database record and remains stable even if
updates/revisions are made to the record.</p>
<p class="block-text-western">The version number is formed by adding
a dotted number such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.2</font></font>
to the accession number to form: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2</font></font>.
This number is a unique identifier for the sequence data within its
record.</p>
<p class="block-text-western">If any change occurs to the sequence
data, no matter how large or small, the version number for that
sequence is incremented by one decimal.</p>
<p class="first-paragraph">Starting in 2015, NCBI has decided to
phase out an older standard called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GI</font></font>
numbers that were integer numbers associated with a record. An older
document may still refer to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GI</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi</font></font>
numbers.</p>
<h2 class="western"><a name="how-do-i-use-entrez-e-utils-web-api"></a>
28.5 How do I use Entrez E-utils web API?</h2>
<p class="first-paragraph">Entrez web API allows us to query NCBI
data sources via a specially constructed URL. A query URL will be of
the form
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">https://service.nih.gov?param1=value1&amp;param2=value2&amp;param3=value3</font></font></p>
<p>Since the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&amp;</font></font>
character has a special meaning for the shell, we need to either put
the URL within single quotes, or we need to &ldquo;escape/protect&rdquo;
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&amp;</font></font>
character by placing a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">\</font></font>
in front of it like so <font face="Consolas, serif"><font size="2" style="font-size: 11pt">\&amp;</font></font></p>
<p>Returns the the data for accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2</font></font>
in the FASTA format:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-s
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2\&amp;db=nuccore\&amp;rettype=fasta
| head</font></font></p>
<p class="first-paragraph">or equivalently:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-s
'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?id=AF086833.2&amp;db=nuccore&amp;rettype=fasta'
| head</font></font></p>
<p class="first-paragraph">Read the <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/books/NBK25500/">Entrez
E-utils</a></font> documentation for information on all parameters.</p>
<h2 class="western"><a name="what-is-entrez-direct"></a>28.6 What is
Entrez Direct?</h2>
<p class="first-paragraph">The web and URL based Entrez E-utils can
be unwieldy and are not well suited for exploratory analysis. The
tool suite called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">entrez-direct</font></font>
(EDirect) simplifies this web access via a series of tools, including
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>
and others:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format gb -id AF086833 | head</font></font></p>
<p class="first-paragraph">We can retrieve the same information in
different formats:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Accession number AF086833 in Genbank format.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format gb -id AF086833 &gt; AF086833.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Accession number AF086833 in Fasta format.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; AF086833.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
efetch can take additional parameters and select a section of the
sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 -seq_start 1 -seq_stop 3</font></font></p>
<p class="first-paragraph">It can even produce the sequence from
reverse strands:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 -seq_start 1 -seq_stop 5
-strand 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 -seq_start 1 -seq_stop 5
-strand 2</font></font></p>
<p class="first-paragraph">But just what exactly did <font face="Consolas, serif"><font size="2" style="font-size: 11pt">strand=2</font></font>
do? Is it the complement or reverse complement? Always ensure you
understand what parameters do. Don&rsquo;t assume you understand
their function. Below are the actual outputs of the last two
commands. We can see that the second command produces the reverse
complement as indicated by the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">c5-1</font></font>
tag in the FASTA file header.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;gb|AF086833.2|:1-5
Ebola virus - Mayinga, Zaire, 1976, complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGAC</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;gb|AF086833.2|:c5-1
Ebola virus - Mayinga, Zaire, 1976, complete genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTCCG</font></font></p>
<h2 class="western"><a name="how-do-we-search-with-entrez-direct"></a>
28.7 How do we search with Entrez Direct?</h2>
<p class="first-paragraph">Once we have a project accession number,
say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>,
we can use it to search for the data that comes with it. We can
locate project accession numbers in published papers, or we find them
in the supporting information.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-help</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nucleotide -query PRJNA257197</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db protein -query PRJNA257197</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch</font></font>
command produces a so-called &ldquo;environment&rdquo; that will be
passed into other Entrez direct programs. You can get a sense of what
the results are from looking at the results of the search:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;ENTREZ_DIRECT&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Db&gt;nucleotide&lt;/Db&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">

&lt;WebEnv&gt;NCID_1_17858181_130.14.18.34_9001_1472654178_1691087027_0MetA0_S_MegaStore_F_1&lt;/WebEnv&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;QueryKey&gt;1&lt;/QueryKey&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Count&gt;249&lt;/Count&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Step&gt;1&lt;/Step&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;/ENTREZ_DIRECT&gt;</font></font></p>
<p class="first-paragraph">To fetch the data for a search, pass it
into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nucleotide -query PRJNA257197 | efetch -format fasta &gt;
genomes.fa</font></font></p>
<p class="first-paragraph">To get all proteins:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db protein -query PRJNA257197 | efetch -format fasta &gt;
proteins.fa</font></font></p>
<h2 class="western"><a name="how-do-i-get-run-information-on-a-project"></a>
28.8 How do I get run information on a project?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
 -db sra -query PRJNA257197 | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">Now we have a file with many columns, here
are a few:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | cut -d , -f 1,2,16 | head -3</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run,ReleaseDate,LibraryLayout</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917,2015-04-14
13:59:24,PAIRED</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918,2015-04-14
13:58:26,PAIRED</font></font></p>
<h2 class="western"><a name="how-do-i-get-even-more-information-on-a-project"></a>
28.9 How do I get even more information on a project?</h2>
<p class="first-paragraph">Another way (and why are there more ways
to do the same thing?) to get run information is via the &ldquo;summary&rdquo;
format:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
 -db sra -query PRJNA257197 | efetch -format summary &gt; summary.xml</font></font></p>
<p class="first-paragraph">This creates a file that technically
speaking is an invalid XML file (gee thanks NCBI, sarcasm) so you
cannot visualize it a browser, but can still be processed with xtract</p>
<p>cat summary.xml | xtract -pattern RUN_SET -element @accession |
head</p>
<p>will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972976</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972975</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972974</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972973</font></font></p>
<p class="first-paragraph">This XML file contains more information
than runinfo.</p>
<h2 class="western"><a name="how-do-i-extract-taxonomy-information"></a>
28.10 How do I extract taxonomy information?</h2>
<p class="first-paragraph">Match taxonomy ids to more readable,
common names:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db taxonomy -id 9606,7227,10090 -format xml &gt; output.xml</font></font></p>
<p class="first-paragraph">Open up the file output.xml in a web
browser to understand its content. It will look like</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;TaxaSet&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   &lt;Taxon&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;TaxId&gt;9606&lt;/TaxId&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;ScientificName&gt;Homo sapiens&lt;/ScientificName&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;OtherNames&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;GenbankCommonName&gt;human&lt;/GenbankCommonName&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;CommonName&gt;man&lt;/CommonName&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;Name&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;ClassCDE&gt;authority&lt;/ClassCDE&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;DispName&gt;Homo sapiens Linnaeus, 1758&lt;/DispName&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;/Name&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;Name&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;ClassCDE&gt;misspelling&lt;/ClassCDE&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &lt;DispName&gt;Home sapiens&lt;/DispName&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &lt;/Name&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">It is a full XML document with lots of
content in it. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
tool matches and extracts elements from this document:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract
-Pattern Taxon -first TaxId ScientificName GenbankCommonName Division</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">9606
   Homo sapiens    human   Primates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7227
   Drosophila melanogaster fruit fly   Invertebrates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10090
  Mus musculus    house mouse Rodents</font></font></p>
<h2 class="western"><a name="how-to-do-more-work-with-entrez-direct"></a>
28.11 How to do more work with Entrez Direct?</h2>
<p class="first-paragraph">Entrez Direct is a sophisticated tool with
surprising powers - though to fully use it, one needs to also have a
good understanding of UNIX.</p>
<p>More documentation can be found in the NCBI manuals titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Entrez
	Direct: E-utilities on the UNIX Command Line</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/">The
	E-utilities In-Depth: Parameters, Syntax and More</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="bio-fetch-download-data"></a>29 bio
fetch: download data</h1>
<p class="first-paragraph">We have implemented the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> command to facilitate data download from GenBank,
Ensembl and other sources. If you need additional functionality
beyond of what <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> offers see the end of the page for a list of
alternatives. Install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph">The full documentation for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
is maintained at <font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help/</a></font>.</p>
<h2 class="western"><a name="rationale"></a>29.1 Rationale</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> was written to provide simpler data access. As in
most cases the accession numbers uniquely define the data type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> can automatically resolve the correct destination
and will download data in the most appropriate format. Examples:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains nucleotidedata from GenBank</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_045512</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains protein data from GenBank</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch YP_009724390</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains SRR run information from the Short Read Archive</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch SRR1972976</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains FASTA sequence files from Ensembl</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENSG00000157764</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the coding sequence segments of a transcript</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENST00000288602 --type cds | head</font></font></p>
<p class="first-paragraph">Read on for more details.</p>
<h2 class="western"><a name="fetch-data-from-genbank"></a>29.2 Fetch
data from GenBank</h2>
<p class="first-paragraph">To get a Genbank nucleotides by accession
number run:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains nucleotidedata from GenBank</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_045512 | head</font></font></p>
<p class="first-paragraph">The default format is GenBank. You may
also list multiple accession numbers at once:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtains mulitple entries from GenBank</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_045512 MN996532 &gt; genomes.gb</font></font></p>
<p class="first-paragraph">you may also fetch FASTA and GFF data
directly from GenBank. The command</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_045512 --format fasta  | head -3</font></font></p>
<p class="first-paragraph">returns a FASTA file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_045512.2
Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1,
complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAAC</font></font></p>
<p class="first-paragraph">whereas:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_045512 --format gff  | head -3</font></font></p>
<p class="first-paragraph">returns a GFF file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##sequence-region
NC_045512.2 1 29903</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##species
https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=2697049</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
RefSeq  region  1   29903   .   +   .  
ID=NC_045512.2:1..29903;Dbxref=taxon:2697049;collection-date=Dec-2019;country=China;gb-acronym=SARS-CoV-2;gbkey=Src;genome=genomic;isolate=Wuhan-Hu-1;mol_type=genomic
RNA;nat-host=Homo sapiens;old-name=Wuhan seafood market pneumonia
virus</font></font></p>
<h2 class="western"><a name="fetching-genome-assemblies"></a>29.3
Fetching genome assemblies</h2>
<p class="first-paragraph">Genome assemblies at NCBI are also
accessible with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Produces a GenBank file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search GCA_000006155</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Produces a FASTA file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search GCA_000006155 --format fasta</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Produces a GFF file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search GCA_000006155 --format gff</font></font></p>
<h2 class="western"><a name="fetch-data-from-ensembl"></a>29.4 Fetch
data from Ensembl</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> recognizes Ensemble accessions (ENSG, ENST) and
will automatically connect to the Ensembl REST API:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Ensenble gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENSG00000157764 | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Transcript data in genomic context</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENST00000288602  | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Transcript data as CDNA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENST00000288602 --type cdna | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Transcript data as CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENST00000288602 --type cds | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Transcript data as protein</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ENST00000288602 --type protein | head</font></font></p>
<p class="first-paragraph">These commands fetch data in FASTA
formats. For more information see the Ensembl REST API:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://rest.ensembl.org/">https://rest.ensembl.org/</a></font></p>
</ul>
<p class="first-paragraph">And the <font color="#4f81bd"><a href="http://www.ebi.ac.uk/ena">enaBrowserTools:
interface with the ENA web services to download data from ENA</a></font></p>
<h2 class="western"><a name="tools-with-similar-utility"></a>29.5
Tools with similar utility</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch</font></font> is primarily a convenience function that
simplifies the use of Entrez and ENA in certain simple and well
defined cases.</p>
<p>See also the related tools that may have expanded functionality:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">Entrez
	Direct: E-utilities on the Unix Command Line</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ebi.ac.uk/ena">enaBrowserTools:
	interface with the ENA web services to download data from ENA</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/pachterlab/ffq">ffq:
	Fetch run information from the European Nucleotide Archive (ENA)</a></font></p>
</ul>
<p class="first-paragraph">From
<font color="#4f81bd"><a href="https://www.biostars.org/p/9492531/#9492581">https://www.biostars.org/p/9492531/#9492581</a></font>
use the <font color="#4f81bd"><a href="https://www.ebi.ac.uk/ena/portal/api/">ENA
portal API</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get information</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-X GET
&quot;https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJNA374918&amp;fields=run_accession,library_name,library_layout,fastq_ftp&amp;result=read_run&quot;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
What fields are acceptable</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-X GET
&quot;https://www.ebi.ac.uk/ena/portal/api/returnFields?dataPortal=ena&amp;result=read_run&quot;</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="bio-search-find-information"></a>30 bio
search: find information</h1>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search</font></font> provides command line interface to SRA and
MyGene. Install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install the package</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Update the database</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
--download</font></font></p>
<p class="first-paragraph">The full documentation for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
is maintained at <font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help/</a></font>.</p>
<h2 class="western"><a name="how-does-this-tool-work"></a>30.1 How
does this tool work?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search</font></font> attempts to recognize the search words with
bioinformatics signficance (like accession numbers) and attempts to
look them up in the most appropriate database. If the search term
does not match any known format, it searches the GenBank assemblies
for regular expression matches. First install the downloadable
database withL</p>
<p>Then start searching for information:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search AF086833</font></font></p>
<p class="first-paragraph">will produce the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;accessionversion&quot;: &quot;AF086833.2&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;assemblyacc&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;assemblygi&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;biomol&quot;: &quot;cRNA&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;biosample&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;caption&quot;: &quot;AF086833&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;completeness&quot;: &quot;complete&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;createdate&quot;: &quot;1999/02/10&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;extra&quot;: &quot;gi|10141003|gb|AF086833.2|&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;flags&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;geneticcode&quot;: &quot;1&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;genome&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;gi&quot;: 10141003,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;moltype&quot;: &quot;rna&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;organism&quot;: &quot;Ebola virus - Mayinga, Zaire,
1976&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;projectid&quot;: &quot;0&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;segsetsize&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;slen&quot;: 18959,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;sourcedb&quot;: &quot;insd&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;strain&quot;: &quot;Mayinga&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;strand&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;subname&quot;: &quot;Mayinga|EBOV-May&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;subtype&quot;: &quot;strain|gb_acronym&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxid&quot;: 128952,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;tech&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;term&quot;: &quot;10141003&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;title&quot;: &quot;Ebola virus - Mayinga, Zaire, 1976,
complete genome&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;topology&quot;: &quot;linear&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;uid&quot;: &quot;10141003&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;updatedate&quot;: &quot;2012/02/13&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font></p>
<p class="first-paragraph">You may also turn the output into comma
separated or tab delimited formats:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search AF086833 --csv</font></font></p>
<p class="first-paragraph">produces CSV files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2,,,cRNA,,AF086833,complete,1999/02/10,gi|10141003|gb|AF086833.2|,,1,,10141003,rna,&quot;Ebola
virus - Mayinga, Zaire,
1976&quot;,0,,18959,insd,Mayinga,,Mayinga|EBOV-May,strain|gb_acronym,128952,,10141003,&quot;Ebola
virus - Mayinga, Zaire, 1976, complete
genome&quot;,linear,10141003,2012/02/13</font></font></p>
<p class="first-paragraph">whereas to produce tab delimited formats
use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search AF086833 --tab</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
 cRNA  AF086833  complete  1999/02/10  gi|10141003|gb|AF086833.2|  1 
10141003  rna  Ebola  virus  -  Mayinga,  Zaire,  1976  0  18959 
insd  Mayinga  Mayinga|EBOV-May  strain|gb_acronym  128952  10141003 
Ebola  virus  -  Mayinga,  Zaire,  1976,  complete  genome  linear 
10141003  2012/02/13</font></font></p>
<h2 class="western"><a name="search-of-srr-numbers"></a>30.2 Search
of SRR numbers</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search SRR14575325</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;run_accession&quot;: &quot;SRR14575325&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;sample_accession&quot;: &quot;SAMN19241174&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;first_public&quot;: &quot;2021-05-20&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;country&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;sample_alias&quot;: &quot;GSM5320434&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;fastq_bytes&quot;: &quot;612817621&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;read_count&quot;: &quot;19439295&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;library_name&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;library_strategy&quot;: &quot;miRNA-Seq&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;library_source&quot;: &quot;TRANSCRIPTOMIC&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;library_layout&quot;: &quot;SINGLE&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;instrument_platform&quot;: &quot;ILLUMINA&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;instrument_model&quot;: &quot;Illumina HiSeq 2000&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;study_title&quot;: &quot;Identification of 5'isomiR in
HCC patients.&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;fastq_ftp&quot;:
&quot;ftp.sra.ebi.ac.uk/vol1/fastq/SRR145/025/SRR14575325/SRR14575325.fastq.gz&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font></p>
<p class="first-paragraph">To search the MyGene interface add a type
to the word:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search symbol:HBB --limit 1</font></font></p>
<p class="first-paragraph">it will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;name&quot;: &quot;hemoglobin subunit beta&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;refseq&quot;: {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;genomic&quot;: [</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;NC_000011.10&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;NG_000007.3&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;NG_059281.1&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;protein&quot;: &quot;NP_000509.1&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;rna&quot;: &quot;NM_000518.5&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;translation&quot;: {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;protein&quot;: &quot;NP_000509.1&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;rna&quot;: &quot;NM_000518.5&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;symbol&quot;: &quot;HBB&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxid&quot;: 9606,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxname&quot;: &quot;Homo sapiens&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
 showing 1 out of 30 results.</font></font></p>
<p class="first-paragraph">If the query word does not seem to match
any known format a regular expression search will applied on the
GenBank assembly reports:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search plasmodium</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;assembly_accession&quot;: &quot;GCA_000002415.2&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;bioproject&quot;: &quot;PRJNA150&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;biosample&quot;: &quot;SAMN02953638&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;wgs_master&quot;: &quot;AAKM00000000.1&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;refseq_category&quot;: &quot;representative genome&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxid&quot;: &quot;5855&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;species_taxid&quot;: &quot;5855&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;organism_name&quot;: &quot;Plasmodium vivax&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;infraspecific_name&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;isolate&quot;: &quot;Salvador I&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;version_status&quot;: &quot;latest&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;assembly_level&quot;: &quot;Chromosome&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;release_type&quot;: &quot;Minor&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;genome_rep&quot;: &quot;Full&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;seq_rel_date&quot;: &quot;2009/05/06&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;asm_name&quot;: &quot;ASM241v2&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;submitter&quot;: &quot;TIGR&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;gbrs_paired_asm&quot;: &quot;GCF_000002415.2&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;paired_asm_comp&quot;: &quot;different&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;ftp_path&quot;:
&quot;https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/002/415/GCA_000002415.2_ASM241v2&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;excluded_from_refseq&quot;: &quot;&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;relation_to_type_materialasm_not_live_date&quot;: &quot;&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   ...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font></p>
<h2 class="western"><a name="other-outputs"></a>30.3 Other outputs</h2>
<p class="first-paragraph">The output may be set to comma separated
values or tab delimited values</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search plasmodium --csv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search plasmodium --tab</font></font></p>
<h2 class="western"><a name="how-to-transform-json-into-other-formats"></a>
30.4 How to transform JSON into other formats</h2>
<p class="first-paragraph">Our tabular formats may be insufficient
when to content is nested. In those cases you will need to transform
the JSON into other formats.</p>
<p>The output of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search</font></font> command is produced the so called JSON
(JavaScript Object Notation) format. JSON is not biology specific,
instead, you can think of it as a generic, lightweight, human
readable approach to represent diverse and hierarchical information.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/JSON">https://en.wikipedia.org/wiki/JSON</a></font></p>
</ul>
<p class="first-paragraph">A tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
may be used to extract information from JSON data in various ways.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
install jq</font></font></p>
<p class="first-paragraph">We use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
by forming extraction patterns with the following rules:</p>
<ul>
	<li/>
<p class="compact">key names are specified with a period <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font></p>
	<li/>
<p class="compact">lists with square brackets <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[]</font></font>.</p>
</ul>
<p class="first-paragraph">We build the extraction command left to
right, matching various aspects of it. Since the output is always a
list, our first pattern will usually be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.[]</font></font>
that visits all elements in the list.</p>
<p>For example, the pattern <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.[].run_accession</font></font>
instructs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
to fetch the list for the root key <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.[]</font></font>,
then visit each element and extract the value for the key named
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">run_accession</font></font>.
Use the pattern it like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Produce a JSON output.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search PRJNA257197 &gt; data.json</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract information from the JSON.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
data.json | jq -r .[].run_accession | head -5</font></font></p>
<p class="first-paragraph">prints the accession numbers:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553416</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553417</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553418</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553419</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553420</font></font></p>
<p class="first-paragraph">To extract two colums <font face="Consolas, serif"><font size="2" style="font-size: 11pt">run_accession</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">read_count</font></font>
into a table:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
data.json | jq -r '.[]|[.run_accession,.read_count]|@tsv' | head -5</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553416
 1014800</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553417
 1865865</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553418
 18323</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553419
 281432</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553420
 223206</font></font></p>
<p class="first-paragraph">When I don&rsquo;t use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
for a while, I forget how it works. I found that the example above
helps remind me and is a good starting point to build other patterns.</p>
<h2 class="western"><a name="how-to-learn-jq"></a>30.5 How to learn
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font></h2>
<p class="first-paragraph">The way <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
works may not be everyone&rsquo;s cup of tea. On the other hand you
can do wondrous things with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
if you put in some effort to learn it. Stick with it! Once you get
it, making patterns becomes that puzzle-solving kind of fun.</p>
<p>There are numerous resources on learning <font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>.
Search and find one that suits you. I found the page below quite
useful:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://earthly.dev/blog/jq-select/">https://earthly.dev/blog/jq-select/</a></font></p>
</ul>
<h2 class="western"><a name="the-mygene-interface"></a>30.6 The
MyGene interface</h2>
<p class="first-paragraph">If the search terms do not match the SRA
number format the search proceeds with querying the MyGene interface.
The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mygene</font></font>
service was published in <font color="#4f81bd"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0953-9">High-performance
web services for querying gene and variant annotation, Genome
Research, 2016</a></font> and is an amazingly potent query interface
that focuses on returning data rather than web pages.</p>
<p>Our search is a command line convenience function that stands on
the shoulders of the giants that created the <font color="#4f81bd"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0953-9">mygene
service</a></font> in the first place.</p>
<h2 class="western"><a name="usage-example"></a>30.7 Usage example</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search HBB --limit 1</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;name&quot;: &quot;hemoglobin subunit beta&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;symbol&quot;: &quot;HBB&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxid&quot;: 9606,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxname&quot;: &quot;Homo sapiens&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
 showing 1 out of 148 results.</font></font></p>
<p class="first-paragraph">The same search for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HBB</font></font>
will return a lot more data as we add fields to it. Lets focus the
search on symbol only <font face="Consolas, serif"><font size="2" style="font-size: 11pt">symbol:HBB</font></font>
human as species, and request Ensembl annotations:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search symbol:HBB --species human --fields ensembl</font></font></p>
<p class="first-paragraph">the query will now produce a much larger
dataset:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;ensembl&quot;: {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;gene&quot;: &quot;ENSG00000244734&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;protein&quot;: [</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENSP00000333994&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENSP00000369671&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENSP00000488004&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENSP00000494175&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENSP00000496200&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;transcript&quot;: [</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000335295&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000380315&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000475226&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000485743&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000633227&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               &quot;ENST00000647020&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;translation&quot;: [</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;protein&quot;: &quot;ENSP00000333994&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;rna&quot;: &quot;ENST00000335295&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;protein&quot;: &quot;ENSP00000494175&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;rna&quot;: &quot;ENST00000647020&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;protein&quot;: &quot;ENSP00000488004&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;rna&quot;: &quot;ENST00000633227&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;protein&quot;: &quot;ENSP00000496200&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;rna&quot;: &quot;ENST00000485743&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               {</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;protein&quot;: &quot;ENSP00000369671&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   &quot;rna&quot;: &quot;ENST00000380315&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           &quot;type_of_gene&quot;: &quot;protein_coding&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       },</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;name&quot;: &quot;hemoglobin subunit beta&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;symbol&quot;: &quot;HBB&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxid&quot;: 9606,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       &quot;taxname&quot;: &quot;Homo sapiens&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">]</font></font></p>
<p class="first-paragraph">Below is a more complicated construct with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">jq</font></font>
that pairs up fields into a tab-delimited table.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
data.json | jq -r '.[].ensembl.translation[]|[.protein,.rna]|
@tsv'</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENSP00000333994
   ENST00000335295</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENSP00000494175
   ENST00000647020</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENSP00000488004
   ENST00000633227</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENSP00000496200
   ENST00000485743</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENSP00000369671
   ENST00000380315</font></font></p>
<h2 class="western"><a name="additional-examples"></a>30.8 Additional
examples</h2>
<p class="first-paragraph">Often examples are more useful than words</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Limiting searches</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search HBB --limit 10</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Adding summary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search HBB -f summary --limit 1 --species human</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Adding additional fields.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search symbol:HBB --species human --fields Ensembl,RefSeq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Gene products with the function in human</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search go:0000307 --species human</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Text search for a concept.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search insulin</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Genes with insulin in the summary then also print the summary
field</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search summary:insulin -f summary</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find aliases, produces the same output</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search symbol:XRCC2 -f alias --species human</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
search alias:SPGF50 -f alias --species human</font></font></p>
<h2 class="western"><a name="field-documention"></a>30.9 Field
documention</h2>
<p class="first-paragraph">See the fields listed at API documentation
site:</p>
<ul>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://docs.mygene.info/en/latest/index.html">https://docs.mygene.info/en/latest/index.html</a></font></p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://docs.mygene.info/en/latest/doc/query_service.html#available_fields">https://docs.mygene.info/en/latest/doc/query_service.html#available_fields</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-genbank-format"></a>31 The GenBank
format</h1>
<h2 class="western"><a name="what-is-the-genbank-format"></a>31.1
What is the GenBank format?</h2>
<p class="first-paragraph">GenBank format is one of the oldest
bioinformatics data formats, originally invented to bridge the gap
between a human-readable representation and one that can be
efficiently processed by the computer. The format (<font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html">defined
here</a></font>) has a so-called fixed-width format, where the first
ten characters form a column that serves as an identifier and the
rest of the lines are information corresponding to that identifier.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LOCUS
      AF086833               18959 bp    cRNA    linear   VRL
13-FEB-2012</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DEFINITION
 Ebola virus - Mayinga, Zaire, 1976, complete genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCESSION
  AF086833</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VERSION
    AF086833.2  GI:10141003</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KEYWORDS
   .</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOURCE
     Ebola virus - Mayinga, Zaire, 1976 (EBOV-May)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ORGANISM  Ebola virus - Mayinga, Zaire, 1976</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Viruses; ssRNA viruses; ssRNA negative-strand viruses;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Mononegavirales; Filoviridae; Ebolavirus.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REFERENCE
  1  (bases 1 to 18959)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 AUTHORS   Bukreyev,A.A., Volchkov,V.E., Blinov,V.M. and
Netesov,S.V.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 TITLE     The VP35 and VP40 proteins of filoviruses. Homology
between Marburg</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           and Ebola viruses</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 JOURNAL   FEBS Lett. 322 (1), 41-46 (1993)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  PUBMED   8482365</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REFERENCE
  2  (bases 1 to 18959)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 AUTHORS   Volchkov,V.E., Becker,S., Volchkova,V.A., Ternovoj,V.A.,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Kotov,A.N., Netesov,S.V. and Klenk,H.D.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 TITLE     GP mRNA of Ebola virus is edited by the Ebola virus
polymerase and</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           by T7 and vaccinia virus polymerases</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           </font></font>
</p>
<p class="first-paragraph">There were times (up to 1987) when the
GenBank database was printed and published as a book.</p>
<p class="compact"><img src="0000_html_f428dab1f3711f08.jpg" name="Image50" align="bottom" hspace="12" width="560" height="671" border="0"/>
</p>
<p class="compact">GenBank as a book</p>
<p>Today, if that stack of books were to be printed it would a few
miles tall. According to the <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/GenBank">Wikipedia
entry on GenBank</a></font>:</p>
<p class="block-text-western">As of 15 June 2018, GenBank release
226.0 has 209,775,348 loci, 263,957,884,539 bases, from 209,775,348
reported sequences.</p>
<p class="first-paragraph"><img src="0000_html_9791d94dcea3d47c.png" name="Image51" align="bottom" hspace="12" width="560" height="407" border="0"/>
</p>
<h2 class="western"><a name="when-do-we-use-the-genbank-format"></a>31.2
When do we use the GenBank format?</h2>
<p class="first-paragraph">The GenBank format has the advantage of
being a generic format that can represent a wide variety of
information while still keeping this information human-readable. And
that was its intended purpose.</p>
<p>For this same reason, however, it is not optimized for any
particular data analysis task. Hence, it is used mainly to exchange
information and is almost never suited as input to an analysis
protocol.</p>
<p>We typically convert a GenBank file to some other, simpler format
to work with it. <font color="#4f81bd"><a href="https://sourceforge.net/projects/readseq/files/readseq/2.1.19/">ReadSeq</a></font>
is a useful conversion tool to use for this purpose.</p>
<h2 class="western"><a name="what-is-refseq"></a>31.3 What is RefSeq?</h2>
<p class="first-paragraph">The NCBI Reference Sequence (RefSeq)
project provides sequence records and related information for
numerous organisms and provides a baseline for medical, functional,
and comparative studies. The RefSeq database is a non-redundant set
of reference standards derived from all the data deposited in GenBank
that includes chromosomes, complete genomic molecules (organelle
genomes, viruses, plasmids), intermediate assembled genomic contigs,
curated genomic regions, mRNAs, RNAs, and proteins.</p>
<h2 class="western"><a name="how-are-refseq-sequences-named"></a>31.4
How are RefSeq sequences named?</h2>
<p class="first-paragraph">The most distinguishing feature of a
RefSeq record is the distinct accession number format that begins
with two characters followed by an underscore (e.g., NP_).</p>
<p><img src="0000_html_63878c2352575791.png" name="Image52" align="bottom" hspace="12" width="560" height="455" border="0"/>
</p>
<p>Thus if a sequence name has an underscore <font face="Consolas, serif"><font size="2" style="font-size: 11pt">_</font></font>
then the sequence is part of RefSeq.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-fasta-format"></a>32 The FASTA
format</h1>
<h2 class="western"><a name="what-is-the-fasta-format"></a>32.1 What
is the FASTA format?</h2>
<p class="first-paragraph">The FASTA format is the &ldquo;workhorse&rdquo;
of bioinformatics. It is used to represent sequence information. The
format is deceivingly simple:</p>
<ul>
	<li/>
<p class="compact">A <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>
	symbol on the FASTA header line indicates a <i>fasta record</i>
	start.</p>
	<li/>
<p class="compact">A string of letters called the <i>sequence
	id</i> may follow the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>
	symbol.</p>
	<li/>
<p class="compact">The header line may contain an arbitrary
	amount of text (including spaces) on the same line.</p>
	<li/>
<p class="compact">Subsequent lines contain the sequence.</p>
</ul>
<p class="first-paragraph">The sequence is assumed to come from an
<i>alphabet</i> that corresponds to a biological entity. For example,
the standard alphabet for nucleotides would contain: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font>.
An extended alphabet may also contain the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
symbol to indicate a base that could be any of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCN</font></font>.
A different extension of a nucleotide alphabet may allow extended
symbols such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">W</font></font>
that corresponds to nucleotide that is either an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
etc. Gaps may be represented via <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>
symbols. Search the web for &ldquo;IUPAC nucleotides&rdquo; to get a
list of all such symbols that may be included in an alphabet. The
same rationale needs to be followed by protein alphabets. The
following is a FASTA file with two FASTA records:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;foo
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;bar
other optional text could go here</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCGTA</font></font></p>
<p class="first-paragraph">As it happens, the FASTA format is not
&ldquo;officially&rdquo; defined - even though it carries the
majority of data information on living systems. Its origins go back
to a <font color="#4f81bd"><a href="http://fasta.bioch.virginia.edu/">software
tool called Fasta</a></font> written by <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/David_J._Lipman">David
Lipman</a></font> (a scientist that later became, and still is, the
director of NCBI) and <font color="#4f81bd"><a href="http://www.people.virginia.edu/~wrp/">William
R. Pearson</a></font> of the University of Virginia. The tool itself
has (to some extent) been superceded by the BLAST suite but the
format (named after the tool itself) not only survived, but has
become the de facto standard.</p>
<h2 class="western"><a name="are-there-problems-with-this-format"></a>
32.2 Are there problems with this format?</h2>
<p class="first-paragraph">The lack of a definition of the FASTA
format and its apparent simplicity can be a source of some of the
most confounding errors in bioinformatics. Since the format appears
so exceedingly straightforward, software developers have been tacitly
assuming that the properties they are accustomed to are required by
some standard - whereas no such thing exists. In general, when
creating FASTA files, the following rules should be observed:</p>
<ul>
	<li/>
<p class="compact">Sequence lines should not be too long.
	While, technically, a Fasta file that contains the sequence of the
	entire human chromosome 1 on a single line (330 million bases) is a
	valid FASTA file, most tools that run on such a file would fail in
	various, occasionally spectacular ways (like bringing your entire
	computer to a screeching halt). Making a FASTA file like that would
	be useful solely as a practical joke on April 1st.</p>
	<li/>
<p class="compact">Some tools may tacitly(!) accept data
	containing alphabets beyond those that they know how to deal with.
	When your data contains alphabets beyond the 4 nucleotides or 20
	aminoacid letters, you need to ensure that whatever tool you use is
	able to recognize the extended values.</p>
	<li/>
<p class="compact">Use upper-case letters. Whereas both
	lower-case and upper-case letters are allowed by the specification,
	the different capitalization may carry additional meaning and some
	tools and methods will (tacitly again) operate differently when
	encountering upper- or lower-case letters. Curiously enough &ndash;
	and curious is a word we use to avoid mentioning how we feel about
	it &ndash; some communities chose to designate the lower-case
	letters as the non-repetitive and the upper-case letters to indicate
	repetitive regions. Needless to say, confusion can rule the game.</p>
	<li/>
<p class="compact">The sequence lines should always wrap at
	the same width (with the exception of the last line). Some tools
	will fail to operate correctly and may not even warn the users if
	this condition is not satisfied. The following is technically a
	valid FASTA but it may cause various subtle problems.</p>
</ul>
<p class="first-paragraph">For example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;foo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCATGCATGCATGCATGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCATGCA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGATGCATGCATGCATGCA</font></font></p>
<p class="first-paragraph">should be reformatted to:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;foo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCATGCATGCATGCATGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCATGCATGATGCATGCA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGCATGCA</font></font></p>
<h2 class="western"><a name="is-there-more-information-in-fasta-headers"></a>
32.3 Is there more information in FASTA headers?</h2>
<p class="first-paragraph">Some data repositories will format FASTA
headers to include structured information. Tools may operate
differently when this information is present in the FASTA header. For
example, NCBI may include both gi and gb accession numbers
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;gi|10141003|gb|AF086833.2|</font></font>.</p>
<p>Specifically the Blast tool from NCBI is able to query data in
more granular fashion when this extra information embedded via header
formatting is available. Below is a list of the recognized FASTA
header formats.</p>
<p><img src="0000_html_2ec48ef56d5db8ea.png" name="Image53" align="bottom" hspace="12" width="560" height="398" border="0"/>
</p>
<h2 class="western"><a name="is-there-more-information-in-the-fasta-sequences"></a>
32.4 Is there more information in the FASTA sequences?</h2>
<p class="first-paragraph">Lower-case letters may be used to indicate
repetitive regions for the genome for example
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCATGCagctagctATGTATGC</font></font>.
In this case <font face="Consolas, serif"><font size="2" style="font-size: 11pt">agctagct</font></font>
is indicated as being present in multiple location of the larger
genome. That being said, it is typically not easy to find out what
exactly the &ldquo;repetitiveness&rdquo; consists of and how it was
determined. Usually, it is the case that the &ldquo;repetitiveness&rdquo;
in a sequence has been determined by a tool run with certain
parameters and that has marked the sequence as such.</p>
<p><b>Importantly</b>, some tools like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">lastz</font></font>
will, by default, skip over regions that are lower-cased - this can
cause lots of confusion if not recognized right away.</p>
<h2 class="western"><a name="where-do-i-get-a-fasta-file"></a>32.5
Where do I get a fasta file?</h2>
<p class="first-paragraph">We will cover and example data access in
more detail, for now if you knew the accession number of a sequence
deposited at NCBI you can obtain it with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id  NM_000020 -format fasta | head</font></font></p>
<p class="first-paragraph">will produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NM_000020.2
Homo sapiens activin A receptor like type 1 (ACVRL1), transcript
variant 1,
mRNA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGGAAACGGTTTATTAGGAGGGAGTGGTGGAGCTGGGCCAGGCAGGAAGACGCTGGAATAAGAAACATTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTGCTCCAGCCCCCATCCCAGTCCCGGGAGGCTGCCGCGCCAGCTGCGCCGAGCGAGCCCCTCCCCGGCT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAGCCCGGTCCGGGGCCGCGCCCGGACCCCAGCCCGCCGTCCAGCGCTGGCGGTGCAACTGCGGCCGCG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGTGGAGGGGAGGTGGCCCCGGTCCGCCGAAGGCTAGCGCCCCGCCACCCGCAGAGCGGGCCCAGAGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCATGACCTTGGGCTCCCCCAGGAAAGGCCTTCTGATGCTGCTGATGGCCTTGGTGACCCAGGGAGACC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CTGTGAAGCCGTCTCGGGGCCCGCTGGTGACCTGCACGTGTGAGAGCCCACATTGCAAGGGGCCTACCTG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCGGGGGGCCTGGTGCACAGTAGTGCTGGTGCGGGAGGAGGGGAGGCACCCCCAGGAACATCGGGGCTGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGGAACTTGCACAGGGAGCTCTGCAGGGGGCGCCCCACCGAGTTCGTCAACCACTACTGCTGCGACAGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCTCTGCAACCACAACGTGTCCCTGGTGCTGGAGGCCACCCAACCTCCTTCGGAGCAGCCGGGAACAGA</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-fastq-format"></a>33 The FASTQ
format</h1>
<h2 class="western"><a name="what-is-the-fastq-format"></a>33.1 What
is the FASTQ format?</h2>
<p class="first-paragraph">The FASTQ format is the de facto standard
by which all sequencing instruments represent data. It may be thought
of as a variant of the FASTA format that allows it to associate a
quality measure to each sequence base, FASTA with QUALITIES.</p>
<p>In simpler terms, it is a format where for every base, we
associate a reliability measure: base is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
and the probability that we are wrong is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1000</font></font>.
Conceptually, the format is very similar to FASTA but suffers from
even more flaws than the FASTA format.</p>
<p>The <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ
format</a></font> consists of 4 sections (and these are usually
produced a single line each):</p>
<ol>
	<li/>
<p class="compact">A FASTA-like header, but instead of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font>
	symbol it uses the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@</font></font>
	symbol. This is followed by an ID and more optional text, similar to
	the FASTA headers.</p>
	<li/>
<p class="compact">The second section contains the measured
	sequence (typically on a single line), but it may be wrapped until
	the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font>
	sign starts the next section.</p>
	<li/>
<p class="compact">The third section is marked by the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font>
	sign and may be optionally followed by the same sequence id and
	header as the first section</p>
	<li/>
<p class="compact">The last line encodes the quality values
	for the sequence in section 2, and must be of the same length as
	section 2. It should (must?) also be wrapped the same way as the
	section 2.</p>
</ol>
<p class="first-paragraph">An example of a single FASTQ record as
seen in the Wikipedia entry:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SEQ_ID</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</font></font></p>
<p class="first-paragraph">The weird characters in the 4th section
(line) are the so called &ldquo;encoded&rdquo; numerical values. In a
nutshell, each character<font face="Consolas, serif"><font size="2" style="font-size: 11pt">!''*((((</font></font>
represents a numerical value: a so-called <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred
score</a></font>, encoded via a single letter encoding. It is,
frankly, a overly convoluted way of doing something that should be
very simple. Perhaps a cautionary tale of what happens when
scientists start to design data formats.</p>
<p>The idea behind the Phred score is to map two digit numbers that
represent error likelihoods to single characters so that the length
of the quality string stays the same as the length of the sequence.</p>
<p>Each character has a numerical value, say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>
will be mapped to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font>
will be remapped to mean <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>
will be mapped to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">30</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">I</font></font>
will be mapped to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">40</font></font>.
In most situations we don&rsquo;t need to remember the mapping, and
we just take the remappings &ldquo;as they are&rdquo; below. The
scale of characters to value mappings looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">!&quot;#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
   |    |    |    |    |    |    |    |
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0....5...10...15...20...25...30...35...40</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
   |    |    |    |    |    |    |   
|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">worst................................best</font></font></p>
<p class="first-paragraph">The quality values of the FASTQ files are
on top. The numbers in the middle of the scale from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">40</font></font>
are called Phred scores. The numbers represent the error
probabilities in a somewhat convoluted manner via the formula
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Error=10^(-P/10)</font></font>
basically summarized as:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=0</font></font>
	means 1/1 (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">100%</font></font>
	error)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=10</font></font>
	means 1/10 (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10%</font></font>
	error)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=20</font></font>
	means 1/100 (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1%</font></font>
	error)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=30</font></font>
	means 1/1000 (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.1%</font></font>
	error)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=40</font></font>
	means 1/10000 (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01%</font></font>
	error(</p>
</ul>
<p class="first-paragraph">This if <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
has the quality of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>
we can see that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>
corresponds to the number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">9</font></font>
that in turn via the formula <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10^(-9/10)</font></font>
is around <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.12</font></font>
that is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">12%</font></font>.
If the quality were <font face="Consolas, serif"><font size="2" style="font-size: 11pt">E</font></font>,
the value is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">36</font></font>
and then we would have <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10^(-36/10)=0.0002</font></font>
that is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02%</font></font></p>
<h2 class="western"><a name="how-to-recognize-fastq-qualities-by-eye"></a>
33.2 How to recognize FASTQ qualities by eye</h2>
<p class="first-paragraph">When qualities look like &ldquo;swearing&rdquo;
in a comic book <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!&quot;#$%&amp;'()*+,-.</font></font>
it means that the data is of low quality. Imagine the sequencing
instrument telling you: <i>Hey, look at all this #</i><img src="0000_html_5abaa21e8c7d32bb.gif" align="absmiddle"/>
<i>!#
data!</i></p>
<p>When the quality scores are jumbled but readable letters your data
has great quality. Imagine the sequencing instrument telling you:
<i>Hey, look at all this ABCDE data!</i></p>
<h2 class="western"><a name="are-there-different-versions-of-the-fastq-encoding"></a>
33.3 Are there different versions of the FASTQ encoding?</h2>
<p class="first-paragraph">Unfortunately, yes. There was a time when
instrumentation makers could not decide at what character to start
the scale. The current standard shown above is the so-called Sanger
(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+33</font></font>)
format where the ASCII codes are shifted by 33. There is the
so-called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">+64</font></font>
format that starts close to where the other scale ends. In this
scaling <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@</font></font>
is mapped to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">i</font></font>
is mapped to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">40</font></font>
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghi</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
   |    |    |    |    |    |    |    |
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0....5...10...15...20...25...30...35...40</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
   |    |    |    |    |    |    |   
|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">worst................................best</font></font></p>
<p class="first-paragraph">Most confusingly, in this encoding the
characters that indicate the highest base quality in Sanger will
indicate one of the lowest qualities. It is easy to recognize this,
though. When our FASTQ quality contains lower case letters <font face="Consolas, serif"><font size="2" style="font-size: 11pt">abcdefghi</font></font>
it means that your data is in this older format. Some tools will
allow you to set parameters to account for this different encoding.
For others, we will need to convert the data to the correct encoding.
You can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqtk</font></font>
for example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqtk
seq -Q64 input.fq &gt; output.fa</font></font></p>
<h2 class="western"><a name="is-there-more-information-in-fastq-headers"></a>
33.4 Is there more information in FASTQ headers?</h2>
<p class="first-paragraph">Just as with FASTA headers, information is
often encoded in the &ldquo;free&rdquo; text section of a FASTQ file.
For example (see the same Wikipedia entry on FASTQ formats):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@EAS139:136:FC706VJ:2:2104:15343:197393
1:Y:18:ATCACG</font></font></p>
<p class="first-paragraph">Contains the following information:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">EAS139</font></font>
	the unique instrument name</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">136</font></font>
	the run id</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FC706VJ</font></font>
	the flowcell id</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
	flowcell lane</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2104</font></font>
	tile number within the flowcell lane</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">15343</font></font>
	&lsquo;x&rsquo;-coordinate of the cluster within the tile</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">197393</font></font>
	&lsquo;y&rsquo;-coordinate of the cluster within the tile</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
	the member of a pair, 1 or 2 (paired-end or mate-pair reads only)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y</font></font>
	Y if the read is filtered, N otherwise</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">18</font></font>
	0 when none of the control bits are on, otherwise it is an even
	number</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATCACG</font></font>
	index sequence</p>
</ul>
<p class="first-paragraph">This information is specific to a
particular instrument/vendor and may change with different versions
or releases of that instrument.</p>
<p>This information can be useful, however, to identify quality
control problems or systematic technical issues that might affect the
instruments.</p>
<h2 class="western"><a name="what-is-a-phred-score"></a>33.5 What is
a Phred score?</h2>
<p class="first-paragraph">A <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Phred_quality_score">Phred
score</a></font> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Q</font></font>
is used to compute the probability of a base call being incorrect by
the formula<font face="Consolas, serif"><font size="2" style="font-size: 11pt">P=10^(-Q/10)</font></font>.</p>
<p>You don&rsquo;t need to remember the formula - there is a much
simpler way to remember it; like so (where the multiplier to 10 tells
you the order of magnitude):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Q
    Error        Accuracy</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
0    1 in 1            0%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10
   1 in 10          90%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20
   1 in 100         99%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">30
   1 in 1000      99.9%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">40
   1 in 10,000   99.99%</font></font></p>
<h2 class="western"><a name="how-do-i-convert-fastq-quality-codes-at-the-command-line"></a>
33.6 How do I convert FASTQ quality codes at the command line?</h2>
<p class="first-paragraph">When it comes to visually evaluating and
inspecting data we rarely need to find out precisely what a value
actually decodes to. The simplest is to remember that:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">!&quot;#$%&amp;'()*+,-.</font></font>
	means low quality <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/10</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-./0123456789</font></font>
	means medium quality <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/100</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABCDEFGHI</font></font>
	means high quality <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1000</font></font></p>
</ul>
<p class="first-paragraph">If you need to convert exactly, you may
use the following command line shortcuts:</p>
<p>Find the integer value that corresponds to character <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">65</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'print (ord(&quot;A&quot;))'</font></font></p>
<p class="first-paragraph">Find character that corresponds to value
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">65</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'print (chr(65))'</font></font></p>
<p class="first-paragraph">Find Phred quality score by the Sanger
encoding of letter <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">32</font></font>)&rdquo;</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'print (ord(&quot;A&quot;)-33)'</font></font></p>
<p class="first-paragraph">Find Sanger encoded letter that
corresponds to Phred quality score <font face="Consolas, serif"><font size="2" style="font-size: 11pt">32</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'print (chr(32+33))'</font></font></p>
<p class="first-paragraph">Compute the error probability represented
by the Pred quality score <font face="Consolas, serif"><font size="2" style="font-size: 11pt">32</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00063095734448</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'from math import*; print (10**-(32/10.0))'</font></font></p>
<p class="first-paragraph">Compute the error probability represented
by letter <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00063095734448</font></font>):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'from math import*; print (10**-((ord(&quot;A&quot;)-33)/10.0))'</font></font></p>
<p class="first-paragraph">Compute the Phred quality score that
corresponds to probability <font face="Consolas, serif"><font size="2" style="font-size: 11pt">p=0.00063095734448</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">32</font></font>)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'from math import*; print
(int(round(-10*log(0.00063095734448)/log(10))))'</font></font></p>
<p class="first-paragraph">Compute the Sanger encoding that
corresponds to probability <font face="Consolas, serif"><font size="2" style="font-size: 11pt">p=0.00063095734448</font></font>
(prints <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c 'from math import*; print
(chr(int(round(-10*log(0.00063095734448)/log(10))+33)))'</font></font></p>
<p class="first-paragraph">Due to the imprecision of the floating
point representation of numbers, a rounding of the results is
required before converting to integer. The extra call to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">int</font></font>
is required only when using Python 2. When using Python 3 <font face="Consolas, serif"><font size="2" style="font-size: 11pt">round</font></font>
alone will do.</p>
<h2 class="western"><a name="closing-thoughts-on-fastq"></a>33.7
Closing thoughts on FASTQ</h2>
<p class="first-paragraph">We have to recognize that most
bioinformatics data formats are adopted via a mixture of past
consensus and historical precedent. Most formats may appear to be
good enough when first proposed, yet the fast advancements may render
them quickly inadequate. The lack of central authority means that
even inefficient formats endure past their prime.</p>
<p>Notably the FASTQ format was originally designed as a multi-line
format just as the FASTA format. In the early days of high-throughput
sequencing, instruments always produced the entire FASTQ sequence on
a single line (on account of the sequences being so short 35-50bp).</p>
<p>A surprising number of bioinformaticians still operate under the
assumption that the FASTQ format is line-oriented and that the entire
sequence must always be on a single line, and that the FASTQ record
consists of 4 lines. In addition, a large number of &ldquo;guides&rdquo;
and &ldquo;tutorials&rdquo; will tacitly assume this and will show
you code that operates on a FASTQ file in a line by line basis.</p>
<p>While having sequences on a single line greatly simplifies
operations, do remember that FASTQ format is not required to be
line-oriented! As sequencing technology evolves and the sequencing
reads become longer, stuffing the entire sequence on a single line
will start to become more inappropriate. For now this problem is yet
acute, but it may get there.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="advanced-fastq-processing"></a>34
Advanced FASTQ processing</h1>
<p class="block-text-western"><b>Note:</b> This chapter presents
advanced topics that may not be immediately useful for beginners.
Feel free to skip it if you are new to bioinformatics. We have added
this chapter right after the FASTQ format to make it locating it
easier later on, when you will need the information in it.
Eventually, most bioinformaticians run into challenges that will
require the techniques that are presented below.</p>
<p class="first-paragraph">Author: <font color="#4f81bd"><a href="http://shenwei.me/"><b>Wei
Shen</b></a></font></p>
<p>This page illustrates advanced FASTA/Q manipulations using <font color="#4f81bd"><a href="http://bioinf.shenwei.me/seqkit/">SeqKit</a></font>.
For more complete documentation visit:</p>
<ul>
	<li/>
<p class="compact">Code: <font color="#4f81bd"><a href="https://github.com/shenwei356/seqkit">https://github.com/shenwei356/seqkit</a></font></p>
	<li/>
<p class="compact">Manual:
	<font color="#4f81bd"><a href="https://bioinf.shenwei.me/seqkit/usage/">https://bioinf.shenwei.me/seqkit/usage/</a></font></p>
</ul>
<p class="first-paragraph">Some other utilities, including <font color="#4f81bd"><a href="http://bioinf.shenwei.me/csvtk/">csvtk</a></font>
(CSV/TSV toolkit) and shell commands were also used.</p>
<p>Note: SeqKit seamlessly supports FASTA and FASTQ formats both in
their original form as well as gzipped compressed format. We list
FASTA or FASTQ depending on the more common usage but you can always
use it on the other type as well.</p>
<p>If you do not have the tools installed yet you can do so by
visiting the release page</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/shenwei356/csvtk/releases">https://github.com/shenwei356/csvtk/releases</a></font></p>
</ul>
<p class="first-paragraph">then downloading and copying the
executable to your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin</font></font>
directory. We do not recommend installing with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
as at this time the conda dependencies are not properly set and may
affect the use of other tools.</p>
<p>For example on MacOS you could do a:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL1=https://github.com/shenwei356/seqkit/releases/download/v0.10.0/seqkit_darwin_amd64.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(cd
~/bin &amp;&amp; curl -L $URL1 | tar zxv -)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chmod
+x
~/bin/seqkit</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL2=https://github.com/shenwei356/csvtk/releases/download/v0.17.0/csvtk_darwin_amd64.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(cd
~/bin &amp;&amp; curl -L $URL2 | tar zxv -)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chmod
+x ~/bin/csvtk</font></font></p>
<h2 class="western"><a name="example-data"></a>34.1 Example data</h2>
<p class="first-paragraph">One FASTQ file (sample reads, 1M) and two
FASTA files (Virus DNA and protein sequences from NCBI RefSeq
database, 60+40M) are used.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc http://data.biostarhandbook.com/reads/duplicated-reads.fq.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc
ftp://ftp.ncbi.nih.gov/refseq/release/viral/viral.2.1.genomic.fna.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc
ftp://ftp.ncbi.nih.gov/refseq/release/viral/viral.2.protein.faa.gz</font></font></p>
<p class="first-paragraph">As the data may change in time, the output
of the commands may change accordingly.</p>
<h2 class="western"><a name="how-to-produce-an-overview-of-fastq-files"></a>
34.2 How to produce an overview of FASTQ files?</h2>
<p class="first-paragraph">Sequence format and type are automatically
detected.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat *.gz</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
                     format  type     num_seqs      sum_len  min_len 
 avg_len    max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">viral.2.1.genomic.fna.gz
 FASTA   DNA         7,942  212,999,370      200  26,819.4 
2,243,109</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">viral.2.protein.faa.gz
   FASTA   Protein   256,686   64,293,716        8     250.5     
8,573</font></font></p>
<h2 class="western"><a name="how-do-i-get-the-gc-content"></a>34.3
How do I get the GC content?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab</font></font> coverts FASTA/Q to 3-column tabular format (1st:
name/ID, 2nd: sequence, 3rd: quality), and can also provide various
information in new columns, including sequence length, GC content/GC
skew, alphabet.</p>
<p>GC content:</p>
<p>seqkit fx2tab &ndash;name &ndash;only-id &ndash;gc viral*.fna.gz |
head</p>
<p>generates:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029548.1
        45.44</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029549.1
        31.66</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029550.1
        30.12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029551.1
        31.09</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029552.1
        30.73</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<h2 class="western"><a name="how-to-i-get-the-percentage-for-custom-bases"></a>
34.4 How to I get the percentage for custom bases?</h2>
<p class="first-paragraph">Suppose we wish to find the percentage for
A, C and A+C:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab -H -n -i -B a -B c -B ac viral.2.1.genomic.fna.gz | head -5</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#name
  seq qual    a   c   ac</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029548.1
        25.65   13.71   39.35</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029549.1
        31.75   16.24   47.99</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029550.1
        30.12   16.93   47.05</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_029551.1
        27.05   17.00   44.05</font></font></p>
<h2 class="western"><a name="how-to-extract-a-subset-of-sequences-with-nameid-list-file"></a>
34.5 How to extract a subset of sequences with name/ID list file?</h2>
<p class="first-paragraph">This is a frequently used manipulation.
Let&rsquo;s create a sample ID list file, which may also come from
some other method like a mapping result.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sample --proportion 0.001  duplicated-reads.fq.gz | seqkit seq --name
--only-id &gt; id.txt</font></font></p>
<p class="first-paragraph">ID list file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head
id.txt</font></font></p>
<p class="first-paragraph">shows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.2996</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.3044</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.3562</font></font></p>
<p class="first-paragraph">Searching by ID list file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file id.txt duplicated-reads.fq.gz &gt;
duplicated-reads.subset.fq.gz</font></font></p>
<h2 class="western"><a name="how-do-i-find-fastaq-sequences-containing-degenerate-bases-and-locate-them"></a>
34.6 How do I find FASTA/Q sequences containing degenerate bases and
locate them?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab</font></font> coverts FASTA/Q to tabular format and can output
the sequence alphabet in a new column. Text searching tools can then
be used to filter the table.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab -n -i -a viral.2.1.genomic.fna.gz | csvtk -H -t grep -f 4 -r
-i -p &quot;[^ACGT]&quot;</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|446730228|ref|NC_019782.1|
                  ACGNT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|557940284|ref|NC_022800.1|
                  ACGKT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|564292828|ref|NC_023009.1|
                  ACGNT</font></font></p>
<p class="first-paragraph">Long-option version of the command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab --name --only-id --alphabet  viral.1.1.genomic.fna.gz | csvtk
--no-header-row --tabs grep --fields 4 --use-regexp --ignore-case
--pattern &quot;[^ACGT]&quot;</font></font></p>
<p class="first-paragraph">You can then exclude these sequences with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
save the sequence IDs.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab -n -i -a viral.2.1.genomic.fna.gz | csvtk -H -t grep -f 4 -r
-i -p &quot;[^ACGT]&quot; | csvtk -H -t cut -f 1 &gt; id2.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
search and exclude.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file id2.txt --invert-match viral.1.1.genomic.fna.gz &gt;
clean.fa</font></font></p>
<p class="first-paragraph">Or locate the degenerate bases, e.g, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">K</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file id2.txt viral.2.1.genomic.fna.gz | seqkit locate
--ignore-case --only-positive-strand --pattern K+ --pattern N+</font></font></p>
<p class="first-paragraph">shows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqID
  patternName     pattern strand  start   end     matched            
                       </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|564292828|ref|NC_023009.1|
  N+      N+      +       87972   87972  
N</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|564292828|ref|NC_023009.1|
  N+      N+      +       100983  100983 
N</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|557307918|ref|NC_022755.1|
  K+      K+      +       1788    1788   
K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|557307918|ref|NC_022755.1|
  K+      K+      +       4044    4044   
K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|589287065|ref|NC_023585.1|
  K+      K+      +       28296   28296  
K</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|590911929|ref|NC_023639.1|
  N+      N+      +       741654  741753  NNNNNNNNNNNNNNNNNNNNNNNNNNN</font></font></p>
<h2 class="western"><a name="how-do-i-remove-fastaq-records-with-duplicated-sequences"></a>
34.7 How do I remove FASTA/Q records with duplicated sequences?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
rmdup --by-seq --ignore-case duplicated-reads.fq.gz &gt;
duplicated-reads.uniq.fq.gz</font></font></p>
<p class="first-paragraph">If the FASTA/Q file is very large, you can
use the flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-m/--md5</font></font>,
which uses MD5 instead of the original sequences to reduce memory
usage during sequence comparison.</p>
<p>You can also deduplicate according to sequence ID (default) or
full name (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--by-name</font></font>).</p>
<h2 class="western"><a name="how-do-i-locate-motifsubsequenceenzyme-digest-sites-in-fastaq-sequence"></a>
34.8 How do I locate motif/subsequence/enzyme digest sites in FASTA/Q
sequence?</h2>
<p class="first-paragraph">Related posts: <font color="#4f81bd"><a href="https://www.biostars.org/p/204658/">Question:
Count and location of strings in fastq file reads</a></font>,
<font color="#4f81bd"><a href="https://www.biostars.org/p/221325/">Question:
Finding TATAWAA in sequence</a></font>.</p>
<p>Assuming you have a list of motifs (enzyme digest sites) in FASTA
format that you would like to locate in your sequences:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
enzymes.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;EcoRI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAATTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;MmeI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TCCRAC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;SacI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAGCTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;XcmI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCANNNNNNNNNTGG</font></font></p>
<p class="first-paragraph">Flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--degenerate</font></font>
is on because our patterns of interest contain degenerate bases.
Command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
locate --degenerate --ignore-case --pattern-file enzymes.fa
viral.2.1.genomic.fna.gz</font></font></p>
<p class="first-paragraph">Sample output (simplified and reformated
by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk
-t uniq -f 3 | csvtk -t pretty</font></font>)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqID
                          patternName   pattern           strand  
start   end     matched</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245010|ref|NC_021865.1|
  MmeI          TCCRAC            +        1816    1821   
TCCGAC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245010|ref|NC_021865.1|
  SacI          GAGCTC            +        19506   19511  
GAGCTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245010|ref|NC_021865.1|
  XcmI          CCANNNNNNNNNTGG   +        2221    2235   
CCATATTTAGTGTGG</font></font></p>
<h2 class="western"><a name="how-do-i-sort-a-huge-number-of-fasta-sequences-by-length"></a>
34.9 How do I sort a huge number of FASTA sequences by length?</h2>
<p class="first-paragraph">Sorting a FASTA file in order of sequence
size (small to large).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sort --by-length viral.2.1.genomic.fna.gz &gt;
viral.1.1.genomic.sorted.fa</font></font></p>
<p class="first-paragraph">If the files are too big, use flag
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--two-pass</font></font>
which will consume less memory.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sort --by-length --two-pass viral.2.1.genomic.fna.gz &gt;
viral.2.1.genomic.sorted.fa</font></font></p>
<p class="first-paragraph">You can also sort by sequence ID
(default), full header (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--by-name</font></font>)
or sequence content (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--by-seq</font></font>).</p>
<h2 class="western"><a name="how-do-i-split-fasta-sequences-according-to-information-in-the-header"></a>
34.10 How do I split FASTA sequences according to information in the
header?</h2>
<p class="first-paragraph">Related posts: <font color="#4f81bd"><a href="https://www.biostars.org/p/223937/">Question:
extract same all similar sequences in FASTA based on the header</a></font>.</p>
<p>For example, the FASTA header line of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">viral.2.protein.faa.gz</font></font>
contains the species name in square brackets.</p>
<p>Overview of FASTA Headers:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
head -n 3 viral.2.protein.faa.gz | seqkit seq --name</font></font></p>
<p class="first-paragraph">generates:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245011|ref|YP_008320337.1|
terminase small subunit [Paenibacillus phage
phiIBB_Pl23]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245012|ref|YP_008320338.1|
terminase large subunit [Paenibacillus phage
phiIBB_Pl23]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245013|ref|YP_008320339.1|
portal protein [Paenibacillus phage phiIBB_Pl23]</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
split</font></font> can split FASTA/Q files according to ID, number
of parts, size of each part, or sequence region. In this case, we&rsquo;ll
split according to sequence ID (species name) which can be specified
by flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--id-regexp</font></font>.</p>
<p>Default ID:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
head -n 3 viral.2.protein.faa.gz | seqkit seq --name --only-id</font></font></p>
<p class="first-paragraph">outputs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245011|ref|YP_008320337.1|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245012|ref|YP_008320338.1|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245013|ref|YP_008320339.1|</font></font></p>
<p class="first-paragraph">New ID:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
head -n 3 viral.2.protein.faa.gz | seqkit seq --name --only-id
--id-regexp &quot;\[(.+)\]&quot;</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Paenibacillus
phage phiIBB_Pl23</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Paenibacillus
phage phiIBB_Pl23</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Paenibacillus
phage phiIBB_Pl23</font></font></p>
<p class="first-paragraph">Split:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
split --by-id --id-regexp &quot;\[(.+)\]&quot; viral.1.protein.faa.gz</font></font></p>
<h2 class="western"><a name="how-do-i-search-and-replace-within-a-fasta-header-using-character-strings-from-a-text-file"></a>
34.11 How do I search and replace within a FASTA header using
character strings from a text file?</h2>
<p class="first-paragraph">Related posts: <font color="#4f81bd"><a href="https://www.biostars.org/p/221962/">Question:
Replace names in FASTA file with a known character string from a text
file</a></font>, <font color="#4f81bd"><a href="https://www.biostars.org/p/205044/">Question:
Fasta header, search and replace&hellip;?</a></font>.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqKit
replace</font></font> can find substrings in FASTA/Q headers using
regular expressions and replace them with strings or corresponding
values of found substrings provided by a tab-delimited key-value
file.</p>
<p>For example, to unify names of protein with unknown functions, we
want to rename &ldquo;hypothetical&rdquo; to &ldquo;putative&rdquo;
in lower case. The replacing rules are listed below in tab-delimited
file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
changes.tsv</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Hypothetical
   putative</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">hypothetical
   putative</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Putative
       putative</font></font></p>
<p class="first-paragraph">Overview the FASTA headers containing
&ldquo;hypothetical&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --by-name --use-regexp --ignore-case --pattern hypothetical
viral.1.protein.faa.gz | seqkit head -n 3 | seqkit seq
--name</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245016|ref|YP_008320342.1|
hypothetical protein IBBPl23_06 [Paenibacillus phage
phiIBB_Pl23]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245019|ref|YP_008320345.1|
hypothetical protein IBBPl23_09 [Paenibacillus phage
phiIBB_Pl23]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi|526245020|ref|YP_008320346.1|
hypothetical protein IBBPl23_10 [Paenibacillus phage phiIBB_Pl23]</font></font></p>
<p class="first-paragraph">A regular expression, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">^([^
]+ )(\w+)</font></font>, was used to specify the key to be replaced,
which is the first word after the sequence ID in this case. Note that
we also capture the ID (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">^([^
]+ )</font></font>) so we can restore it in &ldquo;replacement&rdquo;
with the capture variable <font face="Consolas, serif"><font size="2" style="font-size: 11pt">${1}</font></font>
(more robust than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$1</font></font>).
And flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-I/--key-capt-idx</font></font>
(default: 1) is set to 2 because the key <font face="Consolas, serif"><font size="2" style="font-size: 11pt">(\w+)</font></font>
is the second captured match. Command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
replace --kv-file changes.tsv --pattern &quot;^([^ ]+ )(\w+) &quot;
--replacement &quot;\${1}{kv} &quot; --key-capt-idx 2 --keep-key
viral.1.protein.faa.gz &gt; renamed.fa</font></font></p>
<h2 class="western"><a name="how-do-i-extract-paired-reads-from-two-paired-end-reads-files"></a>
34.12 How do I extract paired reads from two paired-end reads files?</h2>
<p class="first-paragraph">Let&rsquo;s create two unbalanced PE reads
files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
rmdup duplicated-reads.fq.gz | seqkit replace --pattern &quot; .+&quot;
--replacement &quot; 1&quot; | seqkit sample --proportion 0.9
--rand-seed 1 --out-file read_1.fq.gz    </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
rmdup duplicated-reads.fq.gz | seqkit replace --pattern &quot; .+&quot;
--replacement &quot; 2&quot; | seqkit sample --proportion 0.9
--rand-seed 2 --out-file read_2.fq.gz</font></font></p>
<p class="first-paragraph">Overview:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
number of records</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat read_1.fq.gz read_2.fq.gz </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
         format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">read_1.fq.gz
 FASTQ   DNA      9,033  912,333      101      101     
101</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">read_2.fq.gz
 FASTQ   DNA      8,965  905,465      101      101      101</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
sequence headers</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
head -n 3 read_1.fq.gz | seqkit seq --name </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1
1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.3
1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.4
1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
head -n 3 read_2.fq.gz | seqkit seq --name </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1
2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.2
2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.3
2</font></font></p>
<p class="first-paragraph">Firstly, extract sequence IDs of the two
file and compute the intersection:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
seq --name --only-id read_1.fq.gz read_2.fq.gz | sort | uniq -d &gt;
id.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
number of IDs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc
-l id.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">8090
id.txt</font></font></p>
<p class="first-paragraph">Then extract reads using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">id.txt</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file id.txt read_1.fq.gz -o read_1.f.fq.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file id.txt read_2.fq.gz -o read_2.f.fq.gz</font></font></p>
<p class="first-paragraph">Check if the IDs in two files are the same
by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">md5sum</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
seq --name --only-id read_1.f.fq.gz &gt; read_1.f.fq.gz.id.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
seq --name --only-id read_2.f.fq.gz &gt;
read_2.f.fq.gz.id.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">md5sum
read_*.f.fq.gz.id.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">537c57cfdc3923bb94a3dc31a0c3b02a
 read_1.f.fq.gz.id.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">537c57cfdc3923bb94a3dc31a0c3b02a
 read_2.f.fq.gz.id.txt</font></font></p>
<p class="first-paragraph">Note that this example assumes that the
IDs in the two reads file have same order. If not, you can sort them
after previous steps. GNU <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font>
can sort large file using disk, so temporary directory is set as
current directory by option <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-T
.</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip
-d -c read_1.f.fq.gz | seqkit fx2tab | sort -k1,1 -T . | seqkit
tab2fx | gzip -c &gt; read_1.f.sorted.fq.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzip
-d -c read_2.f.fq.gz | seqkit fx2tab | sort -k1,1 -T . | seqkit
tab2fx | gzip -c &gt; read_2.f.sorted.fq.gz</font></font></p>
<h2 class="western"><a name="how-to-concatenate-two-fasta-sequences-in-to-one"></a>
34.13 How to concatenate two FASTA sequences in to one?</h2>
<p class="first-paragraph">Related posts: <font color="#4f81bd"><a href="https://www.biostars.org/p/231806/">Combining
two fasta sequences into one</a></font></p>
<p>Data (not in same order):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
1.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">aaaaa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ccccc</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ggggg</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
2.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCCCC</font></font></p>
<p class="first-paragraph"><b>Step 1</b>. Convert FASTA to
tab-delimited (3 columns, the 3rd column is blank (no quality for
FASTA)) file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab 1.fa &gt; 1.fa.tsv</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
fx2tab 2.fa &gt; 2.fa.tsv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
-A 1.fa.tsv</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq1^Iaaaaa^I$</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq2^Iccccc^I$</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq3^Iggggg^I$</font></font></p>
<p class="first-paragraph"><b>Step 2</b>. Merge two table files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk
join -H -t 1.fa.tsv 2.fa.tsv | cat
-A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq1^Iaaaaa^I^ICCCCC^I$</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq2^Iccccc^I^IGGGGG^I$</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq3^Iggggg^I^ITTTTT^I$</font></font></p>
<p class="first-paragraph"><b>Step 3</b>. Note that there are two TAB
between the two sequences, so we can remove them to join the
sequences</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk
join -H -t 1.fa.tsv 2.fa.tsv | sed 's/\t\t//'</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq1
   aaaaaCCCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq2
   cccccGGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq3
   gggggTTTTT</font></font></p>
<p class="first-paragraph"><b>Step 4</b>. Convert tab-delimited file
back to FASTA file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk
join -H -t 1.fa.tsv 2.fa.tsv | sed 's/\t\t//' | seqkit
tab2fx</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">aaaaaCCCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cccccGGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;seq3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gggggTTTTT</font></font></p>
<p class="first-paragraph">All in one command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk
join -H -t &lt;(seqkit fx2tab 1.fa) &lt;(seqkit fx2tab 2.fa) | sed
's/\t\t//' | seqkit tab2fx</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="reformatting-data"></a>35 Reformatting
data</h1>
<p class="first-paragraph">As we stated before data is defined by:</p>
<ol>
	<li/>
<p class="compact">The <b>information</b> it contains</p>
	<li/>
<p class="compact">The <b>design</b>, and <b>optimization</b>
	of that information.</p>
</ol>
<h2 class="western"><a name="should-i-re-format-transform-my-data"></a>
35.1 Should I re-format (transform) my data?</h2>
<p class="first-paragraph">A question that you will frequently face:</p>
<ol>
	<li/>
<p class="compact">Is it better to reformat existing data into
	the format that you need?</p>
	<li/>
<p class="compact">Is it better to obtain data in different
	forms from the data source?</p>
</ol>
<p class="first-paragraph">The answers is a bit complicated:</p>
<p>In general, we found that if the data source offers data in
different formats, the recommended approach is to obtain each data
format individually and avoid converting between them. Each
conversion is fraught with potential problems.</p>
<p>When the data source does not offer data in different formats, or
when you are not entirely sure that the data that you have represents
the same genomic build, you should do the re-formatting work
yourself. Double check the process and pratice &ldquo;defensive data
analysis&rdquo; where you expect that the conversion is out get you.
Nothing is more dangerous in data analysis than slightly &ldquo;wrong&rdquo;
data.</p>
<h2 class="western"><a name="how-do-i-get-the-same-data-in-different-format"></a>
35.2 How do I get the same data in different format</h2>
<p class="first-paragraph">Here is an example. Let&rsquo;s get the
Murine Leukemia Virus GenBank file under accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_001501</font></font>.
Fetch the sequence from NCBI with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>
(entrez-direct)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_001501 -format gb &gt; genome.gb</font></font></p>
<p class="first-paragraph">We may also use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
tool developed for this book (see <font color="#4f81bd"><a href="#ch131.xhtml#bio-intro">What
is bio?</a></font> ) to do the same:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_001501 &gt; genome.gb</font></font></p>
<p class="first-paragraph">View the sequence with</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.gb | head</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LOCUS
      NC_001501               8332 bp ss-RNA     linear   VRL
13-AUG-2018</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DEFINITION
 Moloney murine leukemia virus, complete genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCESSION
  NC_001501</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VERSION
    NC_001501.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DBLINK
     BioProject: PRJNA485481</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KEYWORDS
   RefSeq.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOURCE
     Moloney murine leukemia virus (MoMLV)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ORGANISM  Moloney murine leukemia virus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Viruses; Ortervirales; Retroviridae; Orthoretrovirinae;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           Gammaretrovirus; Murine leukemia virus.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Suppose you wanted the the information
from the file above in FASTA format. There are different methods to
get this information. We could go back to the data source and fetch
the FASTA file directly with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_001501 -format fasta &gt; genome1.fa</font></font></p>
<p class="first-paragraph">or we could transform our existing GenBank
file with various tools. Lets try <font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqret</font></font>
first:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.gb | seqret -filter -osformat fasta &gt; genome2.fa</font></font></p>
<p class="first-paragraph">or we could transform the GenBank with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.gb | bio fasta &gt; genome3.fa</font></font></p>
<p class="first-paragraph">Are the thefiles the same? Let&rsquo;s
look at the first two lines of each with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome1.fa | head -2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome2.fa | head -2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome3.fa | head -2</font></font></p>
<p class="first-paragraph">The three outputs will:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001501.1
Moloney murine leukemia virus, complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TATGCGCCTGCGTCGGTACTAGTTAGCTAACTAGCTCTGTATCTGGCGGACCCGTGGTGGAACTGACGAG</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001501
NC_001501 Moloney murine leukemia virus, complete
genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tatgcgcctgcgtcggtactagttagctaactagctctgtatctggcggacccgtggtgg</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001501.1
Moloney murine leukemia virus, complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TATGCGCCTGCGTCGGTACTAGTTAGCTAACTAGCTCTGTATCTGGCGGACCCGTGGTGG</font></font></p>
<p class="first-paragraph">Note how the second transformation via
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqret</font></font>
altered the output and does not match the lines from the first. The
first and third (bio) are identical.</p>
<p>Now the sequence itself may be the same, though you can&rsquo;t
tell that by eye. Let&rsquo;s first see if the sequences are
identical. We can run a pairwise alignment to verify this. You could
go about it in many ways, and here I will run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
(covered in subsequent lectures) in a manner that it is only
producing certain attributes of interest that can help us determine
whether the sequences fully match:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-query genome1.fa -subject genome2.fa -outfmt &quot;7 qacc sacc qlen
slen scov length pident&quot; | head -6</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
BLASTN 2.7.1+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Query: NC_001501.1 Moloney murine leukemia virus, complete genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Database: User specified sequence set (Input:
NC_001501-version2.fa)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fields: query acc., subject acc., query length, subject length,
alignment length, % identity</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
5 hits found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_001501.1
NC_001501   8332    8332    8332    100.000</font></font></p>
<p class="first-paragraph">Above we only display the first alignment.
It tells me that both sequences are 8332 bases long, the alignment is
over 8332 bases and the percent identity is 100%. So I can see that
the sequences are identical.</p>
<p>The only yet substantial difference is that the first file the
sequence is named by accession plus version number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_001501.1</font></font>,
whereas in the second file the sequence is named by accession number
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_001501</font></font>
alone. This difference is very annoying as the naming of sequences
may cause problems later on. We also note that another difference is,
of course, the capitalization of the sequence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TATGCG</font></font>
vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tatgcg</font></font>
&hellip;? why &hellip; that may matter later on as well.</p>
<p>The situation above captures typical scenarios that you will face
all the time. You are out there, looking for data, collecting
information, then out of the blue you need to quickly solve a problem
you did expect to have, you need data in a slightly different format,
and you have to verify that two sequences identical</p>
<p>Then of course there may be multiple options to do
transformations; you could be using yet another different tool, let&rsquo;s
use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">readseq</font></font>.
Our transformation then would work like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.gb | readseq -p -f FASTA &gt; genome4.fa</font></font></p>
<p class="first-paragraph">let&rsquo;s see what we get:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome4.fa | head -2</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001501
Moloney murine leukemia virus, complete genome. 8332
bp</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tatgcgcctgcgtcggtactagttagctaactagctctgtatctggcggacccgtggtgg</font></font></p>
<p class="first-paragraph">Guess what, we have slightly different
output yet again. There is extra information at the end of the FASTA
title. I bet you did not see that coming ;-):</p>
<p>But if we think about it, the GenBank to FASTA transformation is
one of the most straightforward changes you might want to do. The
sequence is fully embedded into the GenBank file it just needs to be
extracted, yet the opinions already differ on how to name the
extracted sequences.</p>
<h2 class="western"><a name="who-is-right"></a>35.3 Who is right?</h2>
<p class="first-paragraph">You might ask yourself, all things
considered which tool works correctly?</p>
<p>Ahem &hellip; as it happens, we think only the the tool we wrote,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
does the sensible thing that also matches <font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>.
Both <font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqret</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">readseq</font></font>
are needlessly opinionated. Often bioinformatics tools try to be
helpful in this case remove the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.version</font></font>
from each accession, but with they are tacitly altering data we are
losing information.</p>
<p>Needless to say, once transformations are more complicated, you
can expect even more differences.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="visualizing-biological-data"></a>36
Visualizing biological data</h1>
<p class="first-paragraph">For all the algorithmic power at our
disposal, visually inspecting data has been and will always be an
essential ingredient to scientific discovery. The human eye has
pattern recognition abilities that are unmatched by any computer. Our
minds are the most potent analytic instrument that we have access to.
We are just not fast enough, and we&rsquo;re not physically capable
of processing all the data at our disposal. The limitation is imposed
by the quantity of data.</p>
<p>Trust your own abilities, your eye and mind are able to analyze
data far better than any software. You will see that eventually,
you&rsquo;ll disagree with a computed result &ndash; when that
happens, almost always it means that you are correct and the
computer/software is wrong.</p>
<h2 class="western"><a name="what-are-the-challenges-of-visualization"></a>
36.1 What are the challenges of visualization?</h2>
<p class="first-paragraph">It is essential to recognize that
visualizing data so that it displays the attributes that are
interesting to us at any given moment is a far more complicated task
than we may realize. A meaningful visualization is one that shows us
<i>only</i> what we are interested in at any given moment &ndash; but
that need changes over time.</p>
<p>For this reason, building useful visualization software is far
more challenging than it may seem at first &ndash; and it is more
difficult than building tools that have objective measures
(speed/accuracy etc.) of quality.</p>
<p>No wonder that biological visualization tools have a &ldquo;graveyard&rdquo;
that is larger than that for most other software. Many, <font color="#4f81bd"><a href="https://academic.oup.com/bioinformatics/article/24/10/1305/178363">including
yours truly</a></font>, have tried to build a better, simpler and
nicer visualization approach, and many, including yours truly, have
failed to deliver a long term solution.</p>
<h2 class="western"><a name="what-is-a-genome-browser"></a>36.2 What
is a genome browser?</h2>
<p class="first-paragraph">A genome browser is a graphical interface
to display information from a database of genomic data. Most genome
browsers draw linear tracks that represent the forward strand of the
genome from its 5&rsquo; (left) towards the 3&rsquo; (right)
direction. The genomic features drawn over this linear track are
called <b>glyphs</b>.</p>
<p>A glyph is a pictogram that corresponds to a particular genomic
feature.</p>
<p>No universal standard specifies all visualization tools.
Interpreting glyphs comes with experience and a level of familiarity
with each tool. For example, here is a default screenshot of the UCSC
genome browser:</p>
<p><img src="0000_html_6888ae0d9b2fbfde.png" name="Image54" align="bottom" hspace="12" width="560" height="279" border="0"/>
</p>
<h2 class="western"><a name="why-are-default-browser-screens-so-complicated"></a>
36.3 Why are default browser screens so complicated?</h2>
<p class="first-paragraph">Visit any genome online genome browser
site, and you may be overwhelmed by the information that it shows by
default. As an example click here to see <font color="#4f81bd"><a href="https://genome.ucsc.edu/cgi-bin/hgTracks">a
default UCSC Genome Browser view</a></font>.</p>
<p>Note how most of this information will most likely not be relevant
to you. Unfortunately, scientists are not always that great at
creating practical and straightforward services, and many operate
under the mistaken assumption that &ldquo;more&rdquo; must be
&ldquo;better.&rdquo;</p>
<p>Typically, there is nothing special about the data that a browser
shows you by default. In the tracks above some features of the
elephant genome are shown, together with a so-called &ldquo;Layered
H2K27AC.&rdquo; track without us having ever asked to see it. I mean
no disrespect to you, elephant, I am not interested in seeing your
track, nor am I interested in layered H2K27AC, whatever that is -
some histone with modifications. Perhaps it is the lead developer&rsquo;s
(or their boss&rsquo; boss`) personal data, so they want to show it
off. Their fallacy is to operate under the assumption that everyone
must want to see the same.</p>
<p>Needless to say &ndash; this arbitrary, initial complexity of most
browsers makes their use more a lot more difficult than necessary. It
often takes inordinate amounts of time to find and visualize what you
want.</p>
<h2 class="western"><a name="what-types-of-data-may-be-visualized-in-a-genome-browser"></a>
36.4 What types of data may be visualized in a genome browser?</h2>
<p class="first-paragraph">In most cases, only data in simple
line-oriented formats work: FASTA, BED, GFF, SAM/BAM.</p>
<p>Data formats that contain more complex information such as GenBank
or EMBL typically need to be transformed and simplified into a
line-oriented format such as BED or GFF.</p>
<p>Most visualization tools come with some data pre-loaded, and with
different genomic builds for widely used organisms: human, mouse,
fruit fly, etc. That means that data for many genomic features will
likely already be present. For other organisms, we may need to build
our custom genome and add our custom annotations to the browser.</p>
<h2 class="western"><a name="how-do-i-interpret-glyphs"></a>36.5 How
do I interpret glyphs?</h2>
<p class="first-paragraph">Glyphs are a visual representation of some
genomic features:</p>
<ul>
	<li/>
<p class="compact">Horizontal intervals: directions, genes,
	alignments</p>
	<li/>
<p class="compact">Values over intervals: coverages,
	probabilities</p>
	<li/>
<p class="compact">Attributes at locations: mutations,
	deletions, junctions</p>
</ul>
<p class="first-paragraph"><img src="0000_html_5d89aba0a216ceb5.png" name="Image55" align="bottom" hspace="12" width="560" height="99" border="0"/>
</p>
<p>What you will find is that often it is not possible to locate a
precise definition for the reasons that a feature is drawn a specific
way.</p>
<p>For example, why is one linear feature drawn thicker than the
other? Probably because it is labeled differently. Perhaps it is a
coding sequence, and this tool chose to draw those differently. Alas,
this information may not always be present or easy to find in the
tool&rsquo;s documentation. Often you need to rely on intuition, some
detective work, and your interpretation.</p>
<p>The take-home message here is that behind every visual feature
that you see, no matter how small, it is the visualizer trying to
tell you something more. For example, is an alignment outline of an
ever-so-slightly different color? Perhaps it is a chimeric read,
perhaps it is a multi-mapped read, maybe it is duplicated - and this
is how this visualizer shows these - try to find out what the
visualizer is trying to tell you</p>
<h2 class="western"><a name="which-standalone-genome-browsers-should-i-use"></a>
36.6 Which standalone genome browsers should I use?</h2>
<p class="first-paragraph">A surprising number of standalone genome
visualization tools become abandoned within a few years of their
announcement.</p>
<p>We list those standalone applications that have withstood the test
of time and appear to have decent institutional support:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://software.broadinstitute.org/software/igv/">IGVe,Integrative
	Genomics Viewer</a></font> by the Broad Institute.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://bioviz.org/igb/index.html">IGB,
	Integrated Genome Browser</a></font> from the University of North
	Caroline, Charlotte</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.sanger.ac.uk/science/tools/artemis">Artemis</a></font>
	by the Wellcome Trust Sanger Institute. One of the few browsers that
	can handle GenBank and EMBL formats!</p>
</ul>
<h2 class="western"><a name="what-about-online-genome-browsers"></a>36.7
What about online genome browsers?</h2>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://genome.ucsc.edu/">UCSC
	Genome/Table Browser</a></font> from the University of California at
	Santa Cruz,</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://useast.ensembl.org/index.html">Ensembl
	Genome Browser</a></font> from the partnership of the European
	Bioinformatics Institute and the Wellcome Trust Sanger Institute</p>
</ul>
<p class="first-paragraph">For each of these services, it is possible
to display your data within the context of the data that is stored
within that service. See next chapters for more detail.</p>
<p>It should be noted here, while NCBI offers many facilities to
visualize data that is already deposited into the service, it offers
few options to visualize our data within the same context.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-integrative-genomics-viewer"></a>
37 Using the Integrative Genomics Viewer</h1>
<p class="block-text-western"><b>NOTE:</b> The usage instructions for
IGV and all graphical programs are kept short intentionally.</p>
<p class="block-text-western">In our opinion, when presenting
graphical user interfaces, describing steps in their annoying
minutiae (click here, then click there) have questionable
effectiveness. Step by step instructions that need to be followed
blindly take away your ability to <i>understand</i> what is going on.</p>
<p class="first-paragraph">Different people may begin from different
starting points, and they may get stuck at various stages. To succeed
you will need a conceptual understanding of what can be accomplished
and what outcomes will look like, rather than following a
click-by-click &ldquo;handholding.&rdquo;</p>
<p>What we will state in this book are possible actions, and we will
discuss many of potential pitfalls. But you, yourself will have to
explore and use the interface to learn and successfully perform the
operations.</p>
<p>Take an active approach:</p>
<ul>
	<li/>
<p class="compact">Explore the menu</p>
	<li/>
<p class="compact">Right-click on glyphs, windows, and tabs.</p>
	<li/>
<p class="compact">Try to drag and drop tracks or features.</p>
</ul>
<h2 class="western"><a name="what-is-igv"></a>37.1 What is IGV?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://software.broadinstitute.org/software/igv/">IGV,
Integrative Genomics Viewer</a></font> by the Broad Institute is both
a visualizer and a data integration platform where your data can be
displayed in the context of other information.</p>
<h2 class="western"><a name="how-to-run-igv-on-mac-and-linux"></a>37.2
How to run IGV on Mac and Linux?</h2>
<p class="first-paragraph">You can download and double click on IGV &ndash;
it is simple to run it that way and initially you might use it that
way.</p>
<p>In our experience though, this is not the more optimal way to run
it. When working with large datasets, IGV may get bogged down. The
software may hang, and its error recovery is impacted when run in the
&ldquo;double-click&rdquo; option. We have had much better luck
downloading and running the source code distribution of IGV. For
example after unpacking it into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">src</font></font>
we can run the shell script from the command line:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
~/src/IGV_2.3.85/igv.sh</font></font></p>
<p class="first-paragraph">Then minimize this window and come back to
it only if there is trouble. You will be able to see in this window
when trouble is brewing, and you&rsquo;ll be able to exit IGV more
readily if it hangs/breaks.</p>
<h2 class="western"><a name="how-to-run-igv-on-windows-bash"></a>37.3
How to run IGV on Windows Bash?</h2>
<p class="first-paragraph">Under windows Bash, you will need to run
IGV under as a Windows application! Download the Windows version,
double click it to start. When you need to load data navigate via
Windows to the Unix folder of interest. IGV was one reason why,
during setup, we wanted you to locate where you Unix files are when
accessing them from Windows.</p>
<h2 class="western"><a name="what-does-the-igv-interface-look-like"></a>
37.4 What does the IGV interface look like?</h2>
<p class="first-paragraph">Most visualization tools (and IGV is no
different) come with data preloaded, with different genomic builds of
widely used organisms: human, mouse, fruit-fly, etc. That means that
some information on your organism may already be present. For all
other organisms, we typically build our custom genome and add our
custom annotations.</p>
<p><img src="0000_html_25937557396a4cb5.png" name="Image56" align="bottom" hspace="12" width="560" height="206" border="0"/>
</p>
<p>Note the colors, the orientations; some reads have different
outlines; some regions have different colors. All of these are IGV
specific <b>glyphs</b> by which IGV is trying to display information
about the data.</p>
<h2 class="western"><a name="what-data-does-igv-come-with"></a>37.5
What data does IGV come with?</h2>
<p class="first-paragraph">The Broad Institute distributes data for
IGV from a vast number of organisms and data sources. Do make a note
&ndash; to be able to display their data, your nomenclature for
chromosomes has to match theirs (more on this in the <i>What is in a
name</i> (todo) chapter):</p>
<p><img src="0000_html_52e101300bcd6214.png" name="Image57" align="bottom" hspace="12" width="560" height="242" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-create-a-custom-genome-in-igv"></a>
37.6 How do I create a custom genome in IGV?</h2>
<p class="first-paragraph">To operate on a genome file that is not
preloaded with IGV, we need to build it ourselves. Making a custom
genome will typically require the following:</p>
<ol>
	<li/>
<p class="compact">Identify the genome file and optionally a
	feature file you want to use.</p>
	<li/>
<p class="compact">Create an index for the genome</p>
	<li/>
<p class="compact">Load the new genome into IGV</p>
</ol>
<p class="first-paragraph">How to make a gene file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a genbank file. Unfortunately we cannot load a Genbank file into
IGV.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nucleotide -format gb -id AF086833 &gt; AF086833.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
We must convert the GenBank file into GFF3.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -feature -osformat gff3 &gt;
AF086833.gff</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
We must also convert the GenBank file into FASTA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -feature -osformat fasta &gt;
AF086833.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create an index of the FASTA file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx AF086833.fa</font></font></p>
<p class="first-paragraph">You can now select the menu &ldquo;Genomes
-&gt; Load Genome from File&rdquo; to load your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTA</font></font>
file. Select the &ldquo;File -&gt; Load from File&rdquo; option (or
drag and drop) to select the GFF3 file onto your view. Success looks
like this:</p>
<p><img src="0000_html_5b9bd1babad45565.png" name="Image58" align="bottom" hspace="12" width="560" height="198" border="0"/>
</p>
<p>Keep at it, and you&rsquo;ll eventually succeed. Read the manual,
explore the options, click on menus and importantly right click on
fields and surfaces to see what type of contextual options are does
the software offer.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="visualizing-interval-data"></a>38
Visualizing interval data</h1>
<p class="first-paragraph">Perhaps the most commonly needed
functionality of genome browsers is that of displaying genomic
annotations in the context of other information. Understanding how
information in a file gets represented and displayed in a genome
browser is an essential skill for a bioinformatician.</p>
<p class="block-text-western">Note: you can open images in a new tab
to get a better, high-resolution look at them.</p>
<p class="first-paragraph">For example, take the following IGV view.
What does it mean? Why does it look the way it does? What is the data
structure that lies under it?</p>
<p><img src="0000_html_d3abd9f115f1c984.png" name="Image59" align="bottom" hspace="12" width="560" height="192" border="0"/>
</p>
<h2 class="western"><a name="data-formats-for-interval-data"></a>38.1
Data formats for interval data</h2>
<p class="first-paragraph">The two &ldquo;workhorses&rdquo; of
interval representation are the BED and GFF formats. Just like the
SAM/BAM format designed by <font color="#4f81bd"><a href="http://lh3lh3.users.sourceforge.net/">Heng
Li</a></font>, the BED format is the brainchild of a programmer
extraordinaire <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Jim_Kent">Jim
Kent</a></font>, the bioinformatician credited with &ldquo;saving&rdquo;
the human genome project in 2000. As an excellent computer scientist,
he then created the <font color="#4f81bd"><a href="https://genome.ucsc.edu/">UCSC
genome browser</a></font>, one of the first platforms that allowed
the visualization of genomic data, therein setting the direction the
field has evolved.</p>
<p>Alas, geniuses also cast long shadows. Just as with the SAM/BAM
formats, the BED format introduced and promoted by the UCSC genome
browser initially seemed to work well, later turned out to be grossly
inappropriate for future needs. Moreover, since it adopts an
unnatural model that only makes sense for computer scientists, the
BED format has caused many avoidable errors. For example: in the BED
format, the first element starts at zero; thus, x<font color="#4f81bd"><a href="https://medium.com/@OmnesRes/rickrolls-and-russian-dolls-the-world-of-mindless-citing-fda5d0b0a51e">1</a></font>
indicates the 2nd element. Being &ldquo;off-by-one&rdquo; often
because of the BED format is a pervasive error.</p>
<p>We believe that bioinformatics would be better off if the BED
format were obsoleted. For that reason, we actively discourage the
use of BED as a data format. In reality, it is unavoidable that you
have to process it; thus, we cover it in detail. If you must know it
is a bit like Fahrenheit vs Celsius</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BED</font></font>
	is &ldquo;California style&rdquo;, imperial system for
	bioinformatics data.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF</font></font>
	is &ldquo;European style&rdquo;, metric system for bioinformatics
	data.</p>
</ul>
<h2 class="western"><a name="bed-and-gff-standards"></a>38.2 BED and
GFF standards</h2>
<p class="first-paragraph">BED is shortened from Browser Extensible
Data, whereas GFF is an acronym for Generic Feature Format.</p>
<p>Both BED and GFF are tab-delimited column-oriented formats where a
specific column indicates a specific type of data. For example the
start coordinates for a feature must be in column 2 of a BED file but
a column 4 of a GFF file.</p>
<ul>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">BED
	coordinates start at 0 (zero-based), closed on start, open at the
	end, in mathematical notation: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[start,
	end)</font></font>, where the start is included, but the end is not.
	The first five bases would be represented as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[0,
	5)</font></font> and would contain <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[0,
	1, 2, 3, 4]</font></font></p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">GFF
	coordinates start at 1 (one-based), closed on both start and end, in
	mathematical notation: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[start,
	end]</font></font>, both the start and the end coordinates are part
	of the interval. The first five bases would be represented as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[1,
	5]</font></font> and would contain <font face="Consolas, serif"><font size="2" style="font-size: 11pt">[1,
	2, 3, 4, 5]</font></font></p>
</ul>
<p class="first-paragraph">BED formats may have 3 column, 6 column,
12 column variants where only the first 3, 6, or all 12 columns are
included.</p>
<p>GFF formats must always have nine columns, but the last column may
have different dialects: GFF2, GTF or GFF3</p>
<h2 class="western"><a name="where-do-we-obtain-bedgff-data"></a>38.3
Where do we obtain BED/GFF data?</h2>
<p class="first-paragraph">The data source you are using may already
provide data formatted as GFF. For example, on the NCBI genomic
assembly data distribution page for the Salvia hispanica RNA virus
below:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/viral/Salvia_hispanica_RNA_virus_1/all_assembly_versions/GCF_004134685.1_ASM413468v1/">https://ftp.ncbi.nlm.nih.gov/genomes/refseq/viral/Salvia_hispanica_RNA_virus_1/all_assembly_versions/GCF_004134685.1_ASM413468v1/</a></font></p>
</ul>
<p class="first-paragraph">the file with extension .gff is a GFF
file. For on the UCSC genome browser download page the fruit fly you
can find BED files:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/">http://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/</a></font></p>
</ul>
<p class="first-paragraph">You could also convert a GenBank file to
GFF though conversions are never without additional complexities:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the Ebola Virus Genbank file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nucleotide -format gb -id AF086833 &gt; AF086833.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert the GenBank file into FASTA with the seqret tool.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -osformat fasta &gt; AF086833.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert the GenBank file into GFF3 with the seqret tool.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | seqret -filter -feature -osformat gff3 &gt;
method1.gff</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert the GenBank file into GFF3 with the readseq tool.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | readseq -p -format GFF &gt; method2.gff</font></font></p>
<p class="first-paragraph">Alas the different methods will capture
different information into the GFF file.</p>
<h2 class="western"><a name="what-is-in-a-gffbed-file"></a>38.4 What
is in a GFF/BED file?</h2>
<p class="first-paragraph">The only information that is guaranteed to
be present in every GFF and BED file are the start/end coordinates.</p>
<p>Beyond that, GFF files obtained from two different sources may
have widely different features listed, moreover may have different
amounts of information describing even those features that are common
in both.</p>
<h2 class="western"><a name="how-to-convert-between-bed-and-gff"></a>38.5
How to convert between BED and GFF?</h2>
<p class="first-paragraph">If all you needed to convert are the
coordinates, then:</p>
<ul>
	<li/>
<p class="compact">To convert from BED to GFF, increment the
	start by one and leave the end as is.</p>
	<li/>
<p class="compact">To convert from GFF to BED, decrement the
	start by one and leave the end as is.</p>
</ul>
<p class="first-paragraph">We don&rsquo;t need to &ldquo;touch&rdquo;
the end since the interval&rsquo;s openness/closedness already
accounts for the shift in coordinates.</p>
<p>But if you wanted to build a complete representation of a BED in
GFF or vice versa, then I am afraid that is a lot more complicated
and best to avoid and find the data in the right representation, to
begin with.</p>
<h2 class="western"><a name="bed-and-gff-specification"></a>38.6 BED
and GFF specification</h2>
<p class="first-paragraph">We will cover some functionality but to
understand all details refer to the specification for each file
format:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://genome.ucsc.edu/FAQ/FAQformat.html">BED
	format specification</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF
	format specification</a></font></p>
</ul>
<h2 class="western"><a name="bed-format-structure"></a>38.7 BED
format structure</h2>
<p class="first-paragraph">The most straightforward BED format has
three columns: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sequence
name</font></font>, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">start</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">end</font></font>.
For example, the interval spanning from the 2000th position to the
3000th can be written as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  2000    3000</font></font></p>
<p class="first-paragraph">Except we made a mistake. We are
off-by-one. Since BED uses a zero-based coordinate system to cover
the bases that start on the 2000th base and extend 1000 bases upward,
what we need to write:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  1999    3000</font></font></p>
<h2 class="western"><a name="welp-i-see-nothing"></a>38.8 WELP! I see
nothing!</h2>
<p class="first-paragraph">Sometimes you load your data and you see
nothing. Like this:</p>
<p><img src="0000_html_b32abf844c8a12fa.png" name="Image60" align="bottom" hspace="12" width="560" height="158" border="0"/>
</p>
<p>It is not really &ldquo;nothing&rdquo;, the little track name
shows the filename <font face="Consolas, serif"><font size="2" style="font-size: 11pt">demo.bed</font></font>
so the track is there; something else is incorrect. If you do not see
anything when loading an interval data, it usually has one of the
following reasons:</p>
<ol>
	<li/>
<p class="compact">The sequence names do not match.</p>
	<li/>
<p class="compact">The columns are not tab-separated.</p>
	<li/>
<p class="compact">Columns are missing or in the wrong order</p>
</ol>
<p class="first-paragraph">Note that in general, you can&rsquo;t copy
paste data that contains tabs from the web. Your browser will turn
the tabs into spaces. Solutions:</p>
<ol>
	<li/>
<p class="compact">Make sure the sequence name is the same in
	the file as in the genome, in this case, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2</font></font></p>
	<li/>
<p class="compact">View your data with whitespaces turned on</p>
	<li/>
<p class="compact">Verify that the columns are in the right
	order</p>
</ol>
<p class="first-paragraph"><img src="0000_html_574a8d81b22eb449.png" name="Image61" align="bottom" hspace="12" width="560" height="42" border="0"/>
</p>
<h2 class="western"><a name="the-3-column-bed-format"></a>38.9 The 3
column BED format</h2>
<p class="first-paragraph">The primary use for this format is to
specify a quick-and-easy interval. It just has three columns (it is
zero-based!, so much for quick and easy :-) )</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  1999    3000</font></font></p>
<p class="first-paragraph">When viewed it will look like:</p>
<p><img src="0000_html_1252d97e8db1e397.png" name="Image62" align="bottom" hspace="12" width="560" height="110" border="0"/>
</p>
<p>Note how the coordinates in IGV, seen upon hovering over the
feature, will show <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2000</font></font>
and not <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1999</font></font>.
That is because IGV has already converted it to a more sane one based
system.</p>
<h2 class="western"><a name="the-6-column-bed-format"></a>38.10 The 6
column BED format</h2>
<p class="first-paragraph">We make a 6 column BED format when we need
a name and strand and in addition to what the 3 column BED format
offers. Let&rsquo;s add three more columns, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">name</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">score</font></font>,
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">strand</font></font>.
When data is not specified, it is customary to add a dot <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
as a placeholder:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  1999    3000   FOO    .     +</font></font></p>
<p class="first-paragraph">The view now gains a name <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FOO</font></font>
and the &ldquo;fishbones&rdquo;, the visual markers that indicate the
feature&rsquo;s directionality.</p>
<p><img src="0000_html_71c737e1daa6c2e6.png" name="Image63" align="bottom" hspace="12" width="560" height="82" border="0"/>
</p>
<h2 class="western"><a name="the-9-and-12-column-bed-formats"></a>38.11
The 9 and 12 column BED formats</h2>
<p class="first-paragraph">The format now gets way too complicated
and not suited for &ldquo;human consumption&rdquo;. Read the <font color="#4f81bd"><a href="https://genome.ucsc.edu/FAQ/FAQformat.html">BED
format specification</a></font> for more details. In a nutshell, the
BED format attempts to describe the &ldquo;block&rdquo; structure of
the feature to something like:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
 1999  2000  FOO  0   +  1999   2000  .  3   100,100,100   0,450,900</font></font></p>
<p class="first-paragraph">Hard to read or generate by hand, there
are 3 blocks, the second bloc starts at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1999+450</font></font>
and stretches over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">100</font></font>
bases. IGV will draw it as:</p>
<p><img src="0000_html_a5ad20df111f90b4.png" name="Image64" align="bottom" hspace="12" width="560" height="80" border="0"/>
</p>
<p>The most notable feature here is that the entire structure is
represented on a single line.</p>
<h2 class="western"><a name="the-gff-format"></a>38.12 The GFF format</h2>
<p class="first-paragraph">The GFF format has nine columns, all
required. Write a dot <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
for values that are missing. See the meaning of each column in the
<font color="#4f81bd"><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF
specification</a></font>. A straightforward GFF file would look like
so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
.   CDS 2000    3000    .   .   .   .</font></font></p>
<p class="first-paragraph">You don&rsquo;t strictly need the leading
&ldquo;pragma&rdquo;: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font> but IGV will be more helpful in rendering elements if
that information is there. The output looks like so:</p>
<p><img src="0000_html_3bd052da5a7a23f2.png" name="Image65" align="bottom" hspace="12" width="560" height="74" border="0"/>
</p>
<p>Let&rsquo;s build out our GFF file a little more. Here are two
overlapping intervals on different strands.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
.   CDS 2000    3000    .   +   .   Name=FOO</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
.   CDS 2200    2800    .   -   .   Name=BAR</font></font></p>
<p class="first-paragraph">it will show the following (but you need
to right-click on the track and select &ldquo;Expanded&rdquo; view):</p>
<p><img src="0000_html_28b8223ca7384bd5.png" name="Image66" align="bottom" hspace="12" width="560" height="102" border="0"/>
</p>
<h2 class="western"><a name="representing-hierarchical-data"></a>38.13
Representing hierarchical data</h2>
<p class="first-paragraph">Whereas BED crammed all information into
blocks that belong together into one file, GFF lists each feature
separately but uses the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID</font></font>,
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Parent</font></font>
attributes to connect the features that belong together.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   mRNA  2000   3000   .   +   .  
ID=T1;Name=Transcript1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon   2000   2100   .   +   .   Parent=T1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon   2450   2550   .   +   .   Parent=T1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon   2900   3000   .   +   .   Parent=T1</font></font></p>
<p class="first-paragraph">We specify one transcript with id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T1</font></font>
and three exons with the parent of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T1</font></font>.
The resulting view will be:</p>
<p><img src="0000_html_25b31ea121ac2a8a.png" name="Image67" align="bottom" hspace="12" width="560" height="87" border="0"/>
</p>
<p>The GFF format allows us to specify alternative transcripts just
by listing each &ldquo;child&rdquo; with more &ldquo;parents&rdquo;.
For example, we may add another transcript <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T2</font></font>
and list <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T2</font></font>
as an additional parent for the exons that they share with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T1</font></font>.
Think of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mRNA</font></font>
rows as the backbone over which the exons are layed over to form the
transcript. IGV won&rsquo;t even display the mRNA but that is just a
choice it makes.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   mRNA   2000   3000   .   +   .  
ID=T1;Name=Transcript1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   mRNA   2000   3000   .   +   .  
ID=T2;Name=Transcript2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   mRNA   2000   3000   .   +   .  
ID=T3;Name=Transcript3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon    2000   2100   .   +   .   Parent=T1,T2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon    2400   2500   .   +   .   Parent=T1,T2,T3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512.2
  .   exon    2900   3000   .   +   .   Parent=T1,T3</font></font></p>
<p class="first-paragraph">the gene model will be visualized as:</p>
<p><img src="0000_html_721555ac3b79b4d0.png" name="Image68" align="bottom" hspace="12" width="560" height="121" border="0"/>
</p>
<h2 class="western"><a name="real-gff-file-are-hard-to-read"></a>38.14
Real GFF file are hard to read</h2>
<p class="first-paragraph">Unfortunately the need to cram a lot of
information into the last column in the form of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Name=Value</font></font>
pairs makes most GFF files rather clumsy and hard to read.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_040690.1
RefSeq  region  1   3406    .   +   .  
ID=NC_040690.1:1..3406;Dbxref=taxon:2419805;collection-date=2013;country=India;gbkey=Src;genome=genomic;isolate=Chia-AV1;mol_type=genomic
RNA;nat-host=Salvia hispanica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_040690.1
RefSeq  gene    148 3310    .   +   .  
ID=gene-EXK44_gp1;Dbxref=GeneID:41702919;Name=EXK44_gp1;gbkey=Gene;gene_biotype=protein_coding;locus_tag=EXK44_gp1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_040690.1
RefSeq  CDS 148 981 .   +   0  
ID=cds-YP_009553030.1;Parent=gene-EXK44_gp1;Dbxref=Genbank:YP_009553030.1,GeneID:41702919;Name=YP_009553030.1;exception=ribosomal
slippage;gbkey=CDS;locus_tag=EXK44_gp1;product=fusion
protein;protein_id=YP_009553030.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_040690.1
RefSeq  CDS 983 3310    .   +   0  
ID=cds-YP_009553030.1;Parent=gene-EXK44_gp1;Dbxref=Genbank:YP_009553030.1,GeneID:41702919;Name=YP_009553030.1;exception=ribosomal
slippage;gbkey=CDS;locus_tag=EXK44_gp1;product=fusion
protein;protein_id=YP_009553030.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="what-do-the-words-mean"></a>39 What do
the words mean?</h1>
<p class="first-paragraph">Computational concepts and algorithms
require exact definitions. It is exceedingly important that all
scientists use the same words to describe the same information.</p>
<p>An <i>ontology</i> is a controlled vocabulary of <i>terms</i> or
<i>concepts</i> and a restricted set of <i>relationships</i> between
those terms. The most commonly used ontologies for bioinformatics
data are:</p>
<ul>
	<li/>
<p class="compact">The Sequence Ontology (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SO</font></font>):
	a vocabulary for information related to sequence features.</p>
	<li/>
<p class="compact">The Gene Ontology (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO</font></font>),
	a vocabulary for information related to gene functions.</p>
</ul>
<p class="first-paragraph"><img src="0000_html_13ec2eaf5e36e3bc.png" name="Image69" align="bottom" hspace="12" width="560" height="266" border="0"/>
</p>
<h2 class="western"><a name="why-is-the-ontology-necessary"></a>39.1
Why is the ontology necessary?</h2>
<p class="first-paragraph">In the paper titled <font color="#4f81bd"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2005-6-5-r44">The
Sequence Ontology: a tool for the unification of genome annotations</a></font>,
the authors state:</p>
<p class="block-text-western">Unfortunately, biological terminology
is notoriously ambiguous; the same word is often used to describe
more than one thing, and there are many dialects. For example, does a
coding sequence (CDS) contain the stop codon or is the stop codon
part of the 3&rsquo;-untranslated region (3&rsquo; UTR)?</p>
<p class="block-text-western">There really is no right or wrong
answer to such questions, but consistency is crucial when attempting
to compare annotations from different sources, or even when comparing
annotations performed by the same group over an extended period of
time.</p>
<p class="first-paragraph">It is essential to recognize that
consistency matters more than &ldquo;correctness&rdquo; - there is no
universally &ldquo;correct&rdquo; answer in biology.</p>
<h2 class="western"><a name="are-there-other-ontologies"></a>39.2 Are
there other ontologies?</h2>
<p class="first-paragraph">Yes. Possibly too many. See a list at the
<font color="#4f81bd"><a href="http://www.obofoundry.org/">The Open
Bioinformatic Ontologies (OBO) Foundry</a></font>.</p>
<p>Creating an ontology is a difficult, thankless job. Scientists
often underestimate the difficulty of coming up with logically
consistent definitions. For that reason, they often embark on overly
ambitious ontology building adventures. Hence, there are many
biomedical ontologies left in various states of abandon and
disrepair. See the sorry situation of the <font color="#4f81bd"><a href="http://roc.bgsu.edu/">RNA
Ontology</a></font> for example.</p>
<h2 class="western"><a name="who-names-the-genes"></a>39.3 Who names
the genes?</h2>
<p class="first-paragraph">A few different organizations have taken
upon themselves the obligation to maintain a consistent naming
scheme.</p>
<ul>
	<li/>
<p class="compact">For example <font color="#4f81bd"><a href="http://www.genenames.org/">HUGO
	Gene Nomenclature Committee (HGNC)</a></font> available via
	<font color="#4f81bd"><a href="http://www.genenames.org/">http://www.genenames.org/</a></font>
	is the only worldwide authority that assigns standardized
	nomenclature to human genes.</p>
</ul>
<p class="first-paragraph">Other model and non-model organisms might
have their consortia that may follow similar or very different naming
conventions to label the genes for that organism.</p>
<p>Whereas the gene names are standardized other names, such
transcript names, regulatory elements, etc. may have competing
labeling. Converting between these across different organizations and
their data releases continue to be one of the most annoying and
unsolved problems of bioinformatics.</p>
<h2 class="western"><a name="what-will-our-data-tell-us"></a>39.4
What will our data tell us?</h2>
<p class="first-paragraph">In general, most bioinformatics-oriented
analysis results fall into two categories (and combinations thereof):</p>
<ol>
	<li/>
<p class="compact"><b>What a segment of DNA is</b>, described
	with the Sequence Ontology -&gt; annotation, classification</p>
	<li/>
<p class="compact"><b>What a segment of DNA does</b>,
	expressed with the Gene Ontology -&gt; functional analyses, pathway
	analysis</p>
</ol>
<p class="first-paragraph">For example, a <b>de-novo genome assembly</b>
attempts to reconstruct a genome that was measured after being broken
into many small fragments. Once assembled some part of the genome may
be annotated with terms such as the capacity to encode for proteins
or the ability to regulate RNA expression. The ideal result of this
analysis would be a genome where every base carries annotations that
describe its role.</p>
<p>In contrast a typical <b>RNA-Seq</b> study, which aims to explain
phenotypes by identifying the transcripts that show variations in
their expression levels. Since transcripts have functional roles
associated with them, the change in the abundance of a transcript is
assumed to affect its functional role, and hence to influence the
phenotype of the organism. The ideal outcome of an RNA-Seq experiment
is the identification of the mechanisms by which the DNA functions
and those by which it produces an observed phenotype.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequence-ontology"></a>40 Sequence
ontology</h1>
<h2 class="western"><a name="what-is-the-sequence-ontology-so"></a>40.1
What is the Sequence Ontology (SO)?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://www.sequenceontology.org/"><b>Sequence
Ontology (abbreviated as SO)</b></a></font> located at</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.sequenceontology.org/">http://www.sequenceontology.org/</a></font></p>
</ul>
<p class="first-paragraph">is a collaborative ontology project for
the definition of sequence features used in biological sequence
annotation.</p>
<h2 class="western"><a name="where-do-i-see-sequence-ontology-so-terms-used"></a>
40.2 Where do I see Sequence Ontology (SO) terms used?</h2>
<p class="first-paragraph">Most data sources will tag information
with words that are defined in an ontology. For sequence data the
ontology will be that of Sequence ontology (SO) Let&rsquo;s revisit
the yeast feature file from the Unix tutorial:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/data/SGD_features.tab</font></font></p>
<p class="first-paragraph">the second column shows what each feature
&ldquo;is&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ARS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">telomere</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">telomeric_repeat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X_element</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X_element_combinatorial_repeat</font></font></p>
<p class="first-paragraph">or to summarize the most common ones:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SGD_features.tab | cut -f 2 | sort | uniq -c | sort -rn | head</font></font></p>
<p class="first-paragraph">we get:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">7074
CDS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">6604
ORF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
484 noncoding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
383 long_terminal_repeat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
377 intron</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
352 ARS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
299 tRNA_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
196 ARS_consensus_sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 91 transposable_element_gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 77 snoRNA_gene</font></font></p>
<p class="first-paragraph">but what do these words mean? What is a
&ldquo;transposable_element_gene&rdquo;? That is what the Sequence
ontology is for.</p>
<h2 class="western"><a name="how-do-i-access-the-sequence-ontology-browser"></a>
40.3 How do I access the Sequence Ontology browser?</h2>
<p class="first-paragraph">Search the <font color="#4f81bd"><a href="http://www.sequenceontology.org/browser/obob.cgi">Sequence
Ontology Browser</a></font> for the words that need to be defined.</p>
<p>For example, to find out what the term
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X_element_combinatorial_repeat</font></font>
means within a sequencing data context, we can visit the <font color="#4f81bd"><a href="http://www.sequenceontology.org/browser/obob.cgi">Sequence
Ontology Browser</a></font> and find <font color="#4f81bd"><a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001484">its
definition</a></font> to be:</p>
<p class="block-text-western"><b>X element combinatorial repeat</b></p>
<p class="block-text-western">An X element combinatorial repeat is a
repeat region located between the X element and the telomere or
adjacent Y&rsquo; element.</p>
<p class="first-paragraph">A graphical display shows the relationship
of the terms to its parents:</p>
<p><img src="0000_html_d5013202bcda4c8d.png" name="Image70" align="bottom" hspace="12" width="516" height="768" border="0"/>
</p>
<p>Note how the description above makes use of the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
Element</font></font>, that is a new term, that in turn can also be
found via another search in the browser. What is an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
Element</font></font>:</p>
<p class="block-text-western">** X element**</p>
<p class="block-text-western">The X element is a conserved region, of
the telomere, of ~475 bp that contains an ARS sequence and in most
cases an Abf1p binding site.</p>
<p class="first-paragraph">The definition has two new words <font face="Consolas, serif"><font size="2" style="font-size: 11pt">telomere</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ARS</font></font>.
It may feel like a wild goose chase. It won&rsquo;t really be like
that. As you go higher in the tree of words, there will be fewer new
terms. Keep searching until you find all term definitions.</p>
<p>Tip: click the words on the image to jump to another term - this
makes navigating much more manageable.</p>
<h2 class="western"><a name="does-all-sequencing-data-obey-the-rules-of-so"></a>
40.4 Does all sequencing data obey the rules of SO?</h2>
<p class="first-paragraph">It should, but sometimes it does not. You
may some words used in a data are not &ldquo;officially&rdquo;
defined. You see, the SO is more of a recommendation rather than a
mandate. There is SO Police cracking down on misfits.</p>
<p>Adopt a defensive mindset and verify your data. Other odd and
unexpected situations may occur. For example, we noticed that often
data distribution sources will not include the sequence for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">stop
codon</font></font> in a sequence labeled as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CDS</font></font>
although based on the definition they should:</p>
<p class="block-text-western"><b>CDS</b></p>
<p class="block-text-western">A contiguous sequence which begins
with, and includes a start codon and ends with, and includes a stop
codon.</p>
<p class="first-paragraph">Even data from Ensembl used to make this
mistake - though I believe this has been corrected in the meantime.
As a rule, spot check your data and verify that it is what you think
it is.</p>
<h2 class="western"><a name="how-are-the-so-relationships-defined"></a>
40.5 How are the SO relationships defined?</h2>
<p class="first-paragraph">Whereas SO terminology is precise, the
rationale for the SO relationships between terms are more complicated
to understand - and they feel more subject to interpretation. For
that and other reasons, the majority of life scientists do not make
use of the SO relationships and use the SO terms only.</p>
<p>For details we recommend consulting the publication:</p>
<p><font color="#4f81bd"><a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2005-6-5-r44">The
Sequence Ontology: A tool for the unification of genome annotations
in Genome Biology, 2005</a></font></p>
<h2 class="western"><a name="will-i-need-to-access-the-so-data-directly"></a>
40.6 Will I need to access the SO data directly?</h2>
<p class="first-paragraph">While during regular use bioinformaticians
typically don&rsquo;t need to access the original representation of
the SO data but it is worth spending a bit a bit of time to
understand how it is organized and what it contains.</p>
<p>Ontologies capture the state in which the science has reached in a
moment of time.</p>
<p>Other, more complicated ontologies work the same way.</p>
<h2 class="western"><a name="how-can-i-investigate-the-so-data"></a>40.7
How can I investigate the SO data?</h2>
<p class="first-paragraph">Finding the SO data is somewhat
circuitous. There is a GitHub repository at
<font color="#4f81bd"><a href="https://github.com/The-Sequence-Ontology/SO-Ontologies">https://github.com/The-Sequence-Ontology/SO-Ontologies</a></font>.
But even on that site, there is little information on how this data
is formatted.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://raw.githubusercontent.com/The-Sequence-Ontology/SO-Ontologies/master/Ontology_Files/so-simple.obo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
$URL</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">obo</font></font>
format is plain text and so-called &ldquo;record oriented&rdquo;
rather than being &ldquo;column oriented&rdquo;. What this means is
that information is distibute over many lines, and so called record
separators divide the information into units. Here is an example for
one record:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[Term]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
SO:0000004</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
interior_coding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">subset:
SOFA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">synonym:
&quot;interior coding exon&quot; EXACT []</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">is_a:
SO:0000195 ! coding_exon</font></font></p>
<h2 class="western"><a name="how-many-sequence-ontology-terms-are-there"></a>
40.8 How many Sequence Ontology terms are there?</h2>
<p class="first-paragraph">We can easily count how many <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SO</font></font>
terms are there:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
so-simple.obo </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'Term'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font></p>
<p class="first-paragraph">The entire Sequence Ontology is just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2359</font></font>
terms. Is that a lot? Is that little? Are we close to be finished or
not with naming things? It is hard to tell. Let&rsquo;s say that
based on our current understanding of the genome we have <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2359</font></font>
terms to associate with sequences.</p>
<h2 class="western"><a name="how-can-i-quickly-search-the-sequence-ontology"></a>
40.9 How can I quickly search the Sequence Ontology?</h2>
<p class="first-paragraph">Grepping the raw data is probably the
fastest way to find what you are looking for.</p>
<p>How many record definitions match the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
so-simple.obo </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'name:
gene'</font></font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_sensu_your_favorite_organism</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_class</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_part</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_by_transcript_attribute</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_by_polyadenylation_attribute</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_to_gene_feature</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene_array_member</font></font></p>
<p class="first-paragraph">What we want is the name that matches <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font>
only. We can specify that as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font>
followed by the end of the line. In pattern matching (and you will
see details of this later) the <i>end of line</i> character may be
specified as the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font>
character.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
so-simple.obo </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'name:
gene$'</font></font></font></p>
<p class="first-paragraph">prints only one entry:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-B</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-A</font></font>
flags make <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
print the lines before and after a match. These two flags can be very
handy.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
so-simple.obo </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'name:
gene$'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-B 1 -A 6</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
SO:0000704</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">def:
&quot;A region (or regions) that includes all of the sequence
elements necessary to encode a functional transcript. A gene may
include regulatory regions, transcribed regions and/or other
functional sequence regions.&quot; [SO:immuno_workshop]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comment:
This term is mapped to MGED. Do not obsolete without consulting MGED
ontology. A gene may be considered as a unit of inheritance.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">subset:
SOFA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">xref:
http://en.wikipedia.org/wiki/Gene &quot;wiki&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">is_a:
SO:0001411 ! biological_region</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">relationship:
member_of SO:0005855 ! gene_group</font></font></p>
<h2 class="western"><a name="how-to-search-for-other-information"></a>
40.10 How to search for other information?</h2>
<p class="first-paragraph">While the website offers a visual
representations, you can mine the Sequence ontology at the command
line various pieces of information that otherwise would be more
challenging to obtain. Which records match the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PCR</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
so-simple.obo </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'PCR'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-B 2 -A 2</font></font></p>
<p class="first-paragraph">We get information of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
SO:0000345</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
EST</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">def:
&quot;A tag produced from a single sequencing read from a cDNA clone
or PCR product; typically a few hundred base pairs long.&quot;
[SO:ke]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comment:
This term is mapped to MGED. Do not obsolete without consulting MGED
ontology.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">subset:
SOFA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
SO:0001481</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
RAPD</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">def:
&quot;RAPD is a 'PCR product' where a sequence variant is identified
through the use of PCR with random primers.&quot; [ZFIN:mh]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">synonym:
&quot;Random Amplification Polymorphic DNA&quot; EXACT []</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">is_a:
SO:0000006 ! PCR_product</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">created_by:
kareneilbeck</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">creation_date:
2009-09-09T05:26:10Z</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="gene-ontology"></a>41 Gene ontology</h1>
<h2 class="western"><a name="what-is-the-gene-ontology-go"></a>41.1
What is the Gene Ontology (GO)?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://geneontology.org/">Gene
Ontology (GO)</a></font> is a controlled vocabulary that connects
each gene to one or more functions.</p>
<p>As it happens, the name &ldquo;Gene Ontology&rdquo; is misleading.
The ontology is intended to categorize <b>gene products</b> rather
than the genes themselves. Different products of the same gene may
play very different roles, labeling and summarizing all of these
functions under the same gene name may (and often does) lead to
uncertainty and confusion.</p>
<h2 class="western"><a name="how-is-the-go-designed"></a>41.2 How is
the GO designed?</h2>
<p class="first-paragraph">The GO ontology is a directed, acyclic
graph structure where nodes are the GO terms and the edges are the
relationships between these terms. Each child node is more &ldquo;specific&rdquo;
than its parent, and function is &ldquo;inherited&rdquo; down the
line.</p>
<p>To further separate the concepts the GO project has developed no
one but three structured, independent sub-ontologies that describe
gene products in a species-independent manner. The sub-ontologies
characterize the gene products as follows:</p>
<ul>
	<li/>
<p class="compact"><b>Where does the product exhibit its
	effect?</b> -&gt; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Cellular
	Component (CC)</font></font> This ontology relates the gene product
	to a component of a cell, that is, part of a larger object, for
	example, <i>cell</i>, <i>nucleus</i>, <i>Golgi membrane</i>, <i>SAGA
	complex</i></p>
	<li/>
<p class="compact"><b>How does it work?</b> -&gt; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Molecular
	Function (MF)</font></font> Which biochemical mechanism
	characterizes the product? These terms describe activities that
	occur at the molecular level: <i>lactase activity</i>, <i>actin
	binding</i></p>
	<li/>
<p class="compact"><b>What is the purpose of the gene product?</b>
	&ndash;&gt; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Biological
	Process (BP)</font></font> The general rule to assist in
	distinguishing between a biological process and a molecular function
	is that a process must have <b>more than one</b> definite step:
	<i>cholesterol efflux</i>, <i>transport</i>, <i>mitotic prophase</i></p>
</ul>
<p class="first-paragraph">A graphical representation for GO is the
following:</p>
<p><img src="0000_html_f0dea7732bb5fc10.png" name="Image71" align="bottom" hspace="12" width="560" height="334" border="0"/>
</p>
<p>Each sub-graph above is independent of the others. When we perform
functional annotation we need treat each sub-graph separately. In a
sense we perform functional annotation three times, once for each
category. The same gene product may have fewer or more annotations in
different categories.</p>
<h2 class="western"><a name="what-kind-of-properties-do-annotated-gene-products-have"></a>
41.3 What kind of properties do annotated gene products have ?</h2>
<ul>
	<li/>
<p class="compact">Each gene product may have entries in each
	of the different categories.</p>
	<li/>
<p class="compact">Each gene product may have multiple entries
	in each of the categories</p>
	<li/>
<p class="compact">Leaf nodes (nodes with no children) are the
	most specific representation of knowledge.</p>
</ul>
<p class="first-paragraph">For example a product may be annotated as:</p>
<ul>
	<li/>
<p class="compact">Cellular Component <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0032580</font></font>:
	Golgi cisterna membrane</p>
	<li/>
<p class="compact">Molecular Function <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0008107</font></font>:
	galactoside 2-alpha-L-fucosyltransferase activity</p>
	<li/>
<p class="compact">Biological Process <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0005975</font></font>:
	carbohydrate metabolic process</p>
	<li/>
<p class="compact">Biological Process: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0006486</font></font>:
	protein glycosylation</p>
</ul>
<p class="first-paragraph">There is typically no reason to believe
that the above characterization is complete or that it captures all
the representive functions for that product. When you first start to
grasp what the GO data is, what it can and cannot tell us, is where
you start to realize just how pioneering our work is, how early in
the discovery process we all are. We are still trying to invent the
wheel, and we have come up with a triangular shape so far.</p>
<h2 class="western"><a name="where-can-access-the-go-online"></a>41.4
Where can access the GO online?</h2>
<p class="first-paragraph">Various organization have taken up the
task of visualizing GO terms, with various success when it comes to
usability and informativeness:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.geneontology.org/">GeneOntology.org</a></font>
	is the &ldquo;official&rdquo; data originator.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ebi.ac.uk/QuickGO/">Quick
	Go</a></font> by EMBL</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://amigo.geneontology.org/amigo">AmiGO</a></font>
	Gene Ontology browser</p>
</ul>
<h2 class="western"><a name="how-are-go-terms-organized"></a>41.5 How
are GO terms organized?</h2>
<p class="first-paragraph">The GO ontology is structured as a
directed, acyclic graph where each term has defined relationships to
one or more other terms in the same domain, and sometimes to other
domains. The GO vocabulary is designed to be species-agnostic and
includes terms applicable to prokaryotes and eukaryotes, and to
single and multicellular organisms.</p>
<p>Here is the definition of the Golgi Cisterna (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0031985</font></font>)
visualized with <font color="#4f81bd"><a href="https://github.com/tanghaibao/goatools">goatools</a></font>
(see the chapter for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
usage for installation instructions):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">plot_go_term.py
--term GO:0031985 go-basic.obo</font></font></p>
<p class="first-paragraph">produces the following depiction for the
terms:</p>
<p><img src="0000_html_83a3f708bb629049.gif" name="Image72" align="bottom" hspace="12" width="560" height="891" border="0"/>
</p>
<h2 class="western"><a name="where-can-the-visualize-go-terms-online"></a>
41.6 Where can the visualize GO terms online?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://geneontology.org/">Gene
Ontology</a></font> website is the authoritative source for
definitions but is not particularly well suited for data
interpretation.</p>
<p>The <font color="#4f81bd"><a href="https://www.ebi.ac.uk/QuickGO/">Quick
GO</a></font> service from the European Bioinformatics Institute
offers a web interface with more user-friendly functionality.</p>
<p>The same GO term Golgi Cisterna (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0031985</font></font>)
viewed in <font color="#4f81bd"><a href="https://www.ebi.ac.uk/QuickGO/">QuickGO</a></font>
shows</p>
<p><img src="0000_html_dd7c18bf17868c3f.gif" name="Image73" align="bottom" hspace="12" width="560" height="785" border="0"/>
</p>
<p>When compared to the other visualization, the images are similar,
but they do not represent identical information. Observing
differences when visualizing ontology data is a common theme when
working with Gene Ontology. It is often not easy to identify how an
annotation is being generated/filtered or expanded.</p>
<p>There are so-called reduced (slim) vocabularies that simplify the
data to broader terms at the cost of losing some of its lower level
granularity.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="understanding-the-go-data"></a>42
Understanding the GO data</h1>
<p class="first-paragraph">While during regular use we may not need
to access the original GO data it is worth spending just a bit a time
to understand what it contains and how it is laid out. It will give
you a different perspective, and a deeper understanding of what is
possible to infer in biology based on the way you chose to store the
data.</p>
<p>Remember every time you do a so-called &ldquo;functional analysis&rdquo;
you will make use of the information stored in this data! Do you see
now how <i>everything is data</i>, we take one data, produce results
using that data, then look it up against another data.</p>
<h2 class="western"><a name="what-format-is-the-go-data-in"></a>42.1
What format is the GO data in?</h2>
<p class="first-paragraph">GO data are available for each organism
from the <font color="#4f81bd"><a href="http://geneontology.org/page/download-annotations">GO
Download</a></font> page. The underlying data for the gene ontology
consists of two files:</p>
<ol>
	<li/>
<p class="compact">Gene ontology definition file.</p>
	<li/>
<p class="compact">A gene association file.</p>
</ol>
<p class="first-paragraph">These two pieces of information connect
the ontology terms to gene product names. The GO files are
surprisingly small and can be easily downloaded:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-OL http://purl.obolibrary.org/obo/go.obo</font></font></p>
<h2 class="western"><a name="what-does-a-go-term-file-contain"></a>42.2
What does a GO term file contain?</h2>
<p class="first-paragraph">Many early bio data file formats, and the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.obo</font></font>
is one of them, use unusual and seemingly arbitrary conventions. For
example comments in an `<font face="Consolas, serif"><font size="2" style="font-size: 11pt">.obo</font></font>
file will be indicated by a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>
sign whereas in UNIX in general comments are indicated by the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font>
character. You&rsquo;ll just have to learn to put up with the various
idiosyncratic features.</p>
<p>The content of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">go.obo</font></font>
file is a record oriented format, this means that each chunk of
information is distributed over multiple lines, in the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[Term]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
GO:0000002</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
mitochondrial genome maintenance</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">namespace:
biological_process</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">def:
&quot;The maintenance of the structure and integrity of</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 the mitochondrial genome; includes replication and</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 segregation of the mitochondrial chromosome.&quot; [GOC:ai,
GOC:vw]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">is_a:
GO:0007005 ! mitochondrion organization</font></font></p>
<p class="first-paragraph">What is this information?</p>
<p>You can think of this information as <i>biological knowledge</i>
distilled down to simple and brief defintions, encoded into a
computationally accessible format. The caveat here is that
information that is not yet encoded via the GO cannot be found
automatically with software tools.</p>
<h2 class="western"><a name="what-is-a-go-association-file"></a>42.3
What is a GO association file?</h2>
<p class="first-paragraph">An association file will connect a gene
product to a GO term:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGLV3-9
   ...     GO:0005886</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGLV3-9
   ...     GO:0003823</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGLV3-9
   ...     GO:0002250</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">This would mean that gene product <font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGLV3-9</font></font>
is annotated with the terms <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0005886</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0003823</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0002250</font></font>.
The definition for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0005886</font></font>
can be found in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">go.obo</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
go.obo | grep &quot;id: GO:0005886&quot;  -A 5</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">id:
GO:0005886</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">name:
plasma membrane</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">namespace:
cellular_component</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">alt_id:
GO:0005904</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">def:
&quot;The membrane surrounding a cell that separates the cell from
its external environment. It consists of a phospholipid bilayer and
associated proteins.&quot; [ISBN:0716731363]</font></font></p>
<p class="first-paragraph">You can see how the 7 digits long GO ids
in the form of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:xxxxxxx</font></font>
are defined in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">go.obo</font></font>
file. There will be a different association files for each organism
and the file will contain more columns beyond the two shown above.
Sometimes there is a date, an evidence code, other ids etc.</p>
<h2 class="western"><a name="where-can-i-find-the-association-files-for-different-organisms"></a>
42.4 Where can I find the association files for different organisms?</h2>
<p class="first-paragraph">The list of association files is mantained
at:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://geneontology.org/page/download-go-annotations">http://geneontology.org/page/download-go-annotations</a></font></p>
</ul>
<h2 class="western"><a name="how-to-get-the-human-gene-association-file"></a>
42.5 How to get the human gene association file?</h2>
<p class="first-paragraph">The gene association file for human gene
products can be obtained as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get the file.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-OL http://geneontology.org/gene-associations/goa_human.gaf.gz</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Unzip the compressed file.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf.gz</font></font></p>
<p class="first-paragraph">The file names may occasionally change
(sometimes suddenly and with little prior notice) If the link is
nonfunctional please visit the <font color="#4f81bd"><a href="http://geneontology.org/page/download-annotations">Gene
Ontology download page</a></font> and find and substitute the correct
location.</p>
<h2 class="western"><a name="what-format-does-the-go-association-file-have"></a>
42.6 What format does the GO association file have?</h2>
<p class="first-paragraph">The file is in the so-called GAF format
that is defined at
<font color="#4f81bd"><a href="http://geneontology.org/page/go-annotation-file-gaf-format-21">http://geneontology.org/page/go-annotation-file-gaf-format-21</a></font>
that states:</p>
<p><img src="0000_html_eb265b7011a91548.png" name="Image74" align="bottom" hspace="12" width="560" height="226" border="0"/>
</p>
<p>Comments are specified with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">!</font></font>
the rest are tab separated and column-oriented data of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">UniProtKB
  A0A024QZP7   CDC2  GO:0004672  GO_REF:0000002 ...</font></font></p>
<p class="first-paragraph">Here the first few columns indicate the
gene product name, then its corresponding GO association is listed.
Since GO terms have a hierarchical representation, a GO term for a
gene name implies that the gene is annotated with all the parent
nodes as well. The parent nodes typically describe the similar
characteristics but in a broader sense.</p>
<p>In the gene association file, each gene product may be associated
with one or more GO terms in each category.</p>
<h2 class="western"><a name="do-the-association-files-represent-all-of-the-accumulated-biological-knowledge"></a>
42.7 Do the association files represent all of the accumulated
biological knowledge?</h2>
<p class="first-paragraph">To some extent yes. While life sciences
knowledge is much richer and the domain expertise of individuals is
typically more nuanced, the GO associations represent the
&ldquo;automatically searchable&rdquo; dimension of this knowledge.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">go.obo</font></font>
file measures <font face="Consolas, serif"><font size="2" style="font-size: 11pt">33Mb</font></font>
whereas the compressed gene association file is a mere <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7Mb</font></font>.
It is nothing short of mind-boggling to realize that these files are
a road map to and product of an entire scientific field.</p>
<p class="block-text-western"><b>NOTE</b>: The GO files represent the
accumulated and distilled results of the work of tens of thousands of
scientists over many decades! Every automated, genome-based
diagnosis, data interpretation, functional characterization, or
genetic test relies on the content of this data. The life science
data interpretation relies on Gene Ontology being correct, well
annotated, and available to all.</p>
<h2 class="western"><a name="what-kind-of-properties-does-the-go-data-have"></a>
42.8 What kind of properties does the GO data have?</h2>
<p class="first-paragraph">We will investigate the Gene Ontology. We
have already downloaded the data but for completeness sake we show
the full listing.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get the GO data and their associations.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-OL http://purl.obolibrary.org/obo/go.obo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-OL http://geneontology.org/gene-associations/goa_human.gaf.gz</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Uncompress and remove lines starting with the comment character !</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
from the annotation file. Removing comments makes subsequent commands
simpler.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c goa_human.gaf.gz </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'^!'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
How many gene to association mappings are there?</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
474213 (this number changes when the data is updated)</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Gene names are in column 3.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Gene names are in column 3. Find the unique ones.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
How many unique genes are annotated?</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
19712 (this number changes when the data is updated)</i></font></font></font></p>
<p class="first-paragraph">Intuitively we would expect an increase of
annotations as science progresses, yet over some periods of time
often noted decreases in the numbers of annotations in this file (for
example a decrease from 19719 to 19712 from 2020 to 2021).</p>
<h2 class="western"><a name="what-are-the-most-annotated-human-genes-and-proteins"></a>
42.9 What are the most annotated human genes and proteins?</h2>
<p class="first-paragraph">The association file contains both gene
and protein ids.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
What are the ten most highly annotated proteins in the GO
dataset?</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 2 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
prot_counts.txt</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
What are the ten most highly annotated gene names in the GO
dataset?</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort-uniq-count-rank</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
gene_counts.txt</font></font></p>
<p class="first-paragraph">The structure of the file is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | head</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1102
   HTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">930
TP53</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">858
EGFR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">839
GRB2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">680
GOLGA2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">674
APP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">607
CTNNB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">606
UBC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">584
TRAF2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">548
RPS27A</font></font></p>
<p class="first-paragraph">A neat little tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">datamash</font></font>
lets us do data analytics at the command line.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">datamash
--help</font></font></p>
<p class="first-paragraph">We can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">datamash</font></font>
to answer a series of questions.</p>
<p>What is the average number of annotations per gene?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Compute the mean for column 1.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | datamash  mean 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">30.785308441558</font></font></p>
<p class="first-paragraph">Looks like on average there are 24
annotation per gene. Now, what are the mean, minimum, maximum and
standard deviation for annotations per gene:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Compute the mean, min, max and standard deviations for column 1.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | datamash mean 1 min 1 max 1 sstdev
1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">30.785308441558
1   1097    43.553511886089</font></font></p>
<p class="first-paragraph">This indicate that the number of
annotations is highly variable.</p>
<h2 class="western"><a name="what-are-the-ten-most-highly-annotated-genes-in-the-go-dataset"></a>
42.10 What are the ten most highly annotated genes in the GO dataset?</h2>
<p class="first-paragraph">The commands</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | head</font></font></p>
<p class="first-paragraph">On September 1st, 2021 the file contained:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1102</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   HTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">930</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
TP53</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">858</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
EGFR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">839</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
GRB2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">680</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
GOLGA2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">674</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
APP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">607</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
CTNNB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">606</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
UBC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">584</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
TRAF2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">548</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
RPS27A</font></font></p>
<p class="first-paragraph">Showing us that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HTT</font></font>
has the most annotation. The same command when executed in 2019
produced the following outcome:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">724
TP53</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">669
GRB2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">637
EGFR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">637
UBC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">580
RPS27A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">570
UBB</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">565
UBA52</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">511
CTNNB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">422
SRC</font></font></p>
<p class="first-paragraph">Note how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font>
had the most annotations and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HTT</font></font>
wasn&rsquo;t even in the top ten! Going back in time, for reference
in December 2016, when we ran this command for the first time the
list was the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1337
HLA-B</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
918 HLA-A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
819 HLA-DRB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
789 TP53</font></font></p>
<p class="first-paragraph">Today the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-B</font></font>
gene produces radically different output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | grep HLA-B</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">75
HLA-B</font></font></p>
<p class="first-paragraph">Whoa! What is going on? How did HLA-B drop
from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1337</font></font>
annotations to just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">48</font></font>?</p>
<h2 class="western"><a name="do-the-go-annotations-change"></a>42.11
Do the GO annotations change?</h2>
<p class="first-paragraph">As we mentioned the latest GO data
download produces a starkly different output than it originally did.
Notably, HLA genes are not in the top ten anymore. It reinforces the
transient nature of annotations - as new information or different
filtering parameters are applied the information content can
drastically change.</p>
<p>Why did the HLA genes disappear? Was their presence an error.
Perhaps &hellip; but if so what does it mean for the publications
that we produced during that time that the error was present? Was
anyone ever publicly notified about these change? The previous was a
rhetorical question, but if you wanted to know then short answer id
&ldquo;pfff, of course not, are you kidding?&rdquo;. We discovered
this discrepancy by accident as we were rerunning the same command a
year late. There it goes to show how not everything is all and well
in the GO world&hellip;</p>
<h2 class="western"><a name="how-complete-is-the-go"></a>42.12 How
complete is the GO?</h2>
<p class="first-paragraph">We may wonder: how complete are these GO
annotations? Would it be possible to estimate what percent of all
functional annotations have been discovered so far?</p>
<p>First we could tabulate the dates assigned to each evidence and
see how many pieces of evidence are produced per year.</p>
<p>It is &ldquo;reasonably easy&rdquo; to extract and tabulate the
date information from the two files that we downloaded to get the
rate of information growth in the GO:</p>
<p>Don&rsquo;t worry about this command if you don&rsquo;t get it
yet. We&rsquo;ll cover <font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font>
programming later, but we want to demonstrate some results. What we
do is extract the year from the evidence codes listed in the 14th
column. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">\</font></font>
is the so called line continuation command, it says to ignore the
line end and imagine the whole command is on one line. It is used to
wrap commands that would stretch to far out and make them more
readable:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf \</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 14 \</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'{
print substr($1, 1, 4) }'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
\</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
\</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c \</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-k 2,2 -nr \</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'{
print  $2 &quot;\t&quot; $1 }'</font></font></font></p>
<p class="first-paragraph">The code below produces the number of
annotations verified in a given year in a reverse chronological
order. Here are the outputs when run in 2021:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2021</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   277681</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2020</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   42252</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2019</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   17857</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2018</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   20934</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2017</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   34328</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2016</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   21792</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2015</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   17677</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2014</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   22617</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2013</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   57967</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2012</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   13701</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2011</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   20021</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2010</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   12234</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2009</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   11904</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2008</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   8298</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2007</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   4470</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2006</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   7192</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2005</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   4060</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2004</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   3017</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2003</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   8467</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2002</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   977</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2001</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   1748</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2000</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   5</font></font></p>
<p class="first-paragraph">And when run at the beginning of 2019:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2018
   190351</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2017
   56695</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2016
   24109</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2015
   19656</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2014
   23636</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2013
   57417</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2012
   14693</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2011
   21186</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2010
   13011</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2009
   12512</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2008
   8694</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">This result is quite surprising, and we
don&rsquo;t quite know what to make of it.</p>
<p>It seems that most evidence is assigned to the latest year. Or
maybe there is some feature to the system that attaches the newest
date to every evidence that is being re-observed? We don&rsquo;t
quite know.</p>
<p>There is also a surprising dip in the number of annotations in
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2012</font></font>,
then a subsequent bump in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2013</font></font>.</p>
<p>Finally we can note that the number of annotations assigned to
previous years always decreases.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="bio-explain-show-definitions"></a>43 bio
explain: show definitions</h1>
<p class="first-paragraph">We implemented the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain</font></font> command to facilitate the quick explorations of
the Gene and Sequence Ontologies.</p>
<p>For more information on ontologies, consult <font color="#4f81bd"><a href="https://www.biostarhandbook.com/">The
Biostar Handbook</a></font> chapter <font color="#4f81bd"><a href="https://www.biostarhandbook.com/what-do-the-words-mean.html">Sequence
and Gene Ontology: What do the words mean?</a></font>. To install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
--download</font></font></p>
<p class="first-paragraph">The full documentation for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
is maintained at <font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help/</a></font>.</p>
<h2 class="western"><a name="database-information"></a>43.1 Database
information</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain</font></font></p>
<p class="first-paragraph">Prints the contents of the database:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Content: 43,878 gene ontology terms; 2,354 sequence ontology terms</font></font></p>
<p class="first-paragraph">As we can see the database currently
contains <font face="Consolas, serif"><font size="2" style="font-size: 11pt">43,878</font></font>
gene ontology and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2,354</font></font>
sequence ontology terms.</p>
<h2 class="western"><a name="explain-a-term"></a>43.2 Explain a term</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain exon</font></font></p>
<p class="first-paragraph">prints that the term is part of the
Sequence Ontology (SO) with number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SO:0000147</font></font>
and is defined as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
exon (SO:0000147)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
region of the transcript sequence within a gene which is not removed
from the</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">primary
RNA transcript by RNA splicing.</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Parents:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
transcript_region</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Children:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
coding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
noncoding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
interior_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
decayed_exon (non_functional_homolog_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
pseudogenic_exon (non_functional_homolog_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
exon_region (part_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
exon_of_single_exon_gene</font></font></p>
<p class="first-paragraph">You can now query the parent or child
relationships:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain transcript_region</font></font></p>
<p class="first-paragraph">or</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain coding_exon</font></font></p>
<h2 class="western"><a name="explain-term-by-identifier"></a>43.3
Explain term by identifier</h2>
<p class="first-paragraph">If you know the identifier then you may
use it directly <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain SO:0000147</font></font> or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain GO:2000147</font></font></p>
<h2 class="western"><a name="explain-more-complex-terms"></a>43.4
Explain more complex terms</h2>
<p class="first-paragraph">All words have to match exactly; thus, try
simpler terms before going full length.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain regulation</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain positive regulation</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain positive regulation of x</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain positive regulation of xanthophore differentiation</font></font></p>
<p class="first-paragraph">What is xantophore?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain xanthophore</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
xanthophore (GO:0031633)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
chromatophore containing yellow pigment.</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Parents:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-
plasma membrane-derived chromatophore</font></font></p>
<h2 class="western"><a name="showing-the-term-lineage"></a>43.5
Showing the term lineage</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain exon --lineage</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SO:0000110
 sequence_feature</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 SO:0000001  region</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   SO:0001411  biological_region</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     SO:0000833  transcript_region</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       ## exon (SO:0000147)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       A region of the transcript sequence within a gene which is not
removed from the</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       primary RNA transcript by RNA splicing.</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       Children:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - coding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - noncoding_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - interior_exon</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - decayed_exon (non_functional_homolog_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - pseudogenic_exon (non_functional_homolog_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - exon_region (part_of)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       - exon_of_single_exon_gene</font></font></p>
<h2 class="western"><a name="searching-the-database"></a>43.6
Searching the database</h2>
<p class="first-paragraph">Any query that is not matched will be
searched for. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-go</font></font>
flag restricts tje search for gene ontology while <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-so</font></font>
restricts the search for sequence ontology. To search for both
sequence and gene ontology:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Searces both ontologies</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain histone | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Search gene ontology only</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain histone --go | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Search gene ontology only</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain histone --so |head</font></font></p>
<h2 class="western"><a name="build-the-newest-version-locally"></a>43.7
Build the newest version locally</h2>
<p class="first-paragraph">You may build a database with the newest
data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
explain --build</font></font></p>
<p class="first-paragraph">The command will download the most
up-to-date ontology data and build a new database. The process might
take about half an hour - through the completion speed depends on the
hard drive write speed, solid-state drives finish much faster.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="functional-analysis"></a>44 Functional
analysis</h1>
<p class="first-paragraph">For the majority of people, the reason for
studying bioinformatics is to make sense of the data. After all, we
all want to gain insights into biological phenomena.</p>
<h2 class="western"><a name="the-sorry-state-of-data-categorization"></a>
44.1 The sorry state of data categorization</h2>
<p class="first-paragraph">Surprisingly enough, whereas scientists
are eager to fund each other when it comes to gathering data, the
same people are unexpectedly short-sighted when it comes to funding
the organization and categorization of the information that has
already been collected. It seems that in their minds once data has
been obtained, it becomes a solved problem: &ldquo;just put it in a
database.&rdquo; In reality, organizing existing knowledge is
difficult, thankless and greatly unappreciated effort - that everyone
dearly needs, but few can understand and appreciate the magnitude and
complexity of the task. It is a manifestation of the so-called
<font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Tragedy_of_the_commons">tragedy
of commons</a></font> where: &ldquo;individual users acting
independently according to their self-interest behave contrary to the
common good.&rdquo;</p>
<p>As an example, the <font color="#4f81bd"><a href="http://www.genome.jp/kegg/">Kyoto
Encyclopedia of Genes and Genomes (KEGG)</a></font> database had to
change its licensing and started to charge thousands of dollars for
bulk data access because they were unable to fund their service
otherwise.</p>
<p>Companies have recognized the bottleneck and the commercial
potential of the inefficiencies of the process. Many offer services
that solve the increasingly painful process of summarizing data. Of
course, the services that they offer come with eye-popping pricing
and that makes them practically inaccessible to students, publicly
funded groups or citizen scientists. We want to make it clear that we
are not against the commercialization of services. The concerning
issue is the increasing challenges of data interpretation makes a
commercial services almost a necessity in some fields, rather than a
choice.</p>
<h2 class="western"><a name="what-is-a-functional-analysis"></a>44.2
What is a functional analysis?</h2>
<p class="first-paragraph">A functional analysis typically means
interpreting data in a context that explains observed phenomena,
supports or rejects a hypothesis, discovers a mechanism.</p>
<p class="block-text-western">As it happens most life science
experiments do not directly measure the attributes of interests. To
use an analogy, for example, to weigh an object, we could place it on
a scale, and we could measure the weight instantly. But what if the
object was not accessible - we would have to infer its weight from
observing, other related information. Perhaps as an astronomer does,
maybe the light bends when going around the object. Of course by
making indirect observations we have to contend with many other
facets that will greatly complicate the process. Most life sciences
observations are just like that, we always have to consider multiple
factors, and then combine our data with that collected by others to
make sense of the observation.</p>
<p class="first-paragraph">In the most common interpretation,
functional analysis means interpreting data in the context of all
known information gathered up present time.</p>
<p>For a review, see <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375">Ten
Years of Pathway Analysis: Current Approaches and Outstanding
Challenges</a></font> from 2012, Plos Computational Biology.</p>
<h2 class="western"><a name="what-is-an-over-representation-analysis-ora"></a>
44.3 What is an Over-Representation Analysis (ORA)?</h2>
<p class="first-paragraph">An Over-Representation Analysis (ORA)
attempts to find representative functions of a list of genes. It does
this by comparing the number of times a function has been observed in
an experiment to a baseline. it is one of the most frequently used
method to identify common functions in a group of genes and
transcripts.</p>
<p>In <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375">Ten
Years of Pathway Analysis: Current Approaches and Outstanding
Challenges</a></font> the authors summarize <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORA</font></font>
as:</p>
<p class="block-text-western">&hellip; over-representation analysis
(ORA), which statistically evaluates the fraction of genes in a
particular pathway found among the set of genes showing changes in
expression. It is also referred to as &ldquo;2&times;2 table method&rdquo;
in the literature. ORA uses one or more variations of the following
strategy: first, an input list is created using a certain threshold
or criteria. For example, a researcher may choose genes that are
differentially over- or under-expressed in a given condition at a
false discovery rate (FDR) of 5%. Then, for each pathway, input genes
that are part of the pathway are counted. This process is repeated
for an appropriate background list of genes (e.g., all genes measured
on a microarray).</p>
<p class="first-paragraph">In a nutshell, an ORA analysis computes
the following numbers:</p>
<ol>
	<li/>
<p class="compact">How many functions do all genes have in
	total?</p>
	<li/>
<p class="compact">How many functions do my selected (subset)
	genes have?</p>
	<li/>
<p class="compact">How many functions should we observe if we
	were to select a random subset of genes?</p>
	<li/>
<p class="compact">How many functions did we observe for my
	selected genes? Is that number larger (functional enrichment) or
	smaller (functional depletion) than the expectation?</p>
</ol>
<p class="first-paragraph">An <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ORA</font></font>
analysis will compute how likely is that the last number, that of the
functions in our subset could be obtained by random chance.</p>
<p>If the number of functions is more extensive than expected, it is
usually interpreted as the subset being enriched with respect to the
function.</p>
<h2 class="western"><a name="are-there-different-ways-to-compute-ora-analyses"></a>
44.4 Are there different ways to compute ORA analyses?</h2>
<p class="first-paragraph">If you followed the section on Gene
Ontology, you know that the GO files have a relatively simple format
to them. Moreover, the data structure is that of a (network) tree.
Building a program that assigns counts to each node and can traverse
the tree is a standard problem that requires only moderately advanced
programming skill. Yet deciding under what circumstance to call a
count over-represented is not settled <b>at all</b>! Different
methods may produce wildly different outcomes.</p>
<p>Since data interpretation is such an acute problem, a &ldquo;cottage
industry&rdquo; of ORA tools exists, perhaps even hundreds of such
tools. Alas the large number only means fragmentation and lack of
support for each, the vast majority of the tools become abandoned,
are failing or even worse give the wrong answer tacitly.</p>
<p>Moreover, the accuracy of a method will critically depend on it
integrating the most up to date information. As a tale of caution, we
note that the <font color="#4f81bd"><a href="https://david.ncifcrf.gov/home.jsp">DAVID:
Functional Annotation Tool</a></font> was not updated from 2010 to
2016! Over this time it was unclear to users whether the data that
the server operates on includes the most up to date information. New
functions and annotations are added on an almost weekly basis, over
the many years this difference increased considerably. We believe
that by the end of the 2016 DAVID operated on a mere 20% of the total
information t known in biology. Nevertheless, it had gained thousands
of citations every single year.</p>
<p><img src="0000_html_ec8105aeed71788c.png" name="Image75" align="bottom" hspace="12" width="560" height="420" border="0"/>
</p>
<p>As you can see above in 2015 alone, DAVID collected over 5000
citations, yet at that time it contained only 20% of the known
annotations. The &ldquo;good&rdquo; news is that there was an update
in 2016 when the information has caught up hence the data is not as
far behind as it used to.</p>
<p>The sorry state of DAVID has been a sore point for many. The
&ldquo;update&rdquo; might have been in response to the growing
chorus of complaints:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.nature.com/nmeth/journal/v13/n9/full/nmeth.3963.html">Impact
	of outdated gene annotations on pathway enrichment analysis</a></font>
	Nature Methods, 2016</p>
</ul>
<p class="first-paragraph">We also refer the reader to several
Biostar posts:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/100427/">Is
	the DAVID knowledgebase not updated anymore?</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/177998/">GO
	analysis: DAVID vs.&nbsp;GREAT vs.&nbsp;GOrilla</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/193803/">Finally
	the much awaited DAVID update has taken place</a></font></p>
</ul>
<p class="first-paragraph">While it is great that there was an update
in 2016, by 2019 the data has again become obsolete. As Yogi Berra
would say: &ldquo;It&rsquo;s like d&eacute;j&agrave; vu all over
again.&rdquo;</p>
<h2 class="western"><a name="what-are-problems-with-the-ora-analysis"></a>
44.5 What are problems with the ORA analysis?</h2>
<p class="first-paragraph">The shortcomings of the overlap analysis
are that:</p>
<ol>
	<li/>
<p class="compact">ORA analysis does not account for the
	magnitude of expression levels. The gene is either in the list or
	not.</p>
	<li/>
<p class="compact">ORA typically uses only a subset of genes -
	the cutoffs are &ldquo;arbitrary&rdquo; in the sense that they are
	based on convention rather than an objective measure.</p>
	<li/>
<p class="compact">Genes and functions are all considered
	independent of one another. For statistical assumptions to work,
	this is an essential requirement. If the independence constraint
	does not hold, then the mathematical basis for the test does not
	hold either. As we all know many functions in the cell are strongly
	interdependent.</p>
</ol>
<h2 class="western"><a name="why-do-we-still-use-the-ora-analysis"></a>
44.6 Why do we still use the ORA analysis?</h2>
<p class="first-paragraph">Reading the chapter makes the clear point
that ORA analyses have many pitfalls. Why do we still use them?</p>
<p>I think the reason for this is that the concept is simple, quick
to compute, and most scientists understand what takes place. For many
other techniques the internal workings are so complicated and
&ldquo;black-box&rdquo;-like that often you will be left with nothing
more than a belief that that box probably does what it claims to.</p>
<p>In our opinion, ORA analyses should be used as a hypothesis
generation and testing framework but not necessarily as a method that
provides the final answer to a problem. Especially p-values need to
be interpeted with caution, don&rsquo;t put too much stock into
neither large nor tiny p-values. So many factors are at play that
p-values are exceedingly unreliable.</p>
<h2 class="western"><a name="what-is-a-functional-class-scoring-fcs"></a>
44.7 What is a Functional Class Scoring (FCS)?</h2>
<p class="first-paragraph">In <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375">Ten
Years of Pathway Analysis: Current Approaches and Outstanding
Challenges</a></font> the authors summarize <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FCS</font></font>
as:</p>
<ol>
	<li/>
<p class="compact">&ldquo;A gene-level statistic is computed
	using the molecular measurements from an experiment.&rdquo;</p>
	<li/>
<p class="compact">&ldquo;The gene-level statistics for all
	genes in a pathway are aggregated into a single pathway-level
	statistic.&rdquo;</p>
	<li/>
<p class="compact">&ldquo;The final step in FCS is assessing
	the statistical significance of the pathway-level statistic.&rdquo;</p>
</ol>
<p class="block-text-western">FCS methods use this information to
detect coordinated changes in the expression of genes in the same
pathway. Finally, by considering the correlated changes in gene
expression, FCS methods account for dependence between genes in a
pathway, which ORA does not.</p>
<p class="first-paragraph">Now you can see another reason why ORA
analyses are still used. At least we understand what the ORA does -
whereas the text above could be applied broadly and generically to
cover a great variety of techniques - most of which could be
complicated and nearly impossible to understand for a life scientist.
When people use an FCS tool they become mere &ldquo;passengers&rdquo;
of it - they run the tool then hope for the best that it takes them
where they want to go.</p>
<p>There are many implementations of each method and approach some
more explicit than others - do refer <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375">to
the paper</a></font> for a list.</p>
<h2 class="western"><a name="should-i-trust-the-results-of-functional-analyses"></a>
44.8 Should I trust the results of functional analyses?</h2>
<p class="first-paragraph">We advise caution. The statistical methods
used to determine the values that biologists are so fond of and use
as cutoffs (p-values, FDR) are particularly ill-suited when derived
from functional annotation. It is not the methods that are flawed, it
is the data that does not conform to assumptions that the methods
make.</p>
<p>For a statistical method to work correctly, it needs access to
reasonably unbiased and representative data. To determine what &ldquo;is
enriched&rdquo;, we need to know what &ldquo;not-enriched&rdquo;
looks like. But data on gene function does not conform to either
requirement; discoveries are driven by the need to find unusual
behaviors. Genes under study are selected by scientists interested in
certain phenomena and will be published only when something
interesting is found. Thus there is &ldquo;bias&rdquo; in the
selection and a &ldquo;bias&rdquo; in what makes into a function
repository.</p>
<p>Long story short, don&rsquo;t put too much stock into p-values and
other measures. Use the statistical measures more as a suggestion or
estimation rather than precise, scientific criteria.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="gene-set-enrichment"></a>45 Gene set
enrichment</h1>
<h2 class="western"><a name="what-is-a-gene-set-enrichment-analysis"></a>
45.1 What is a gene set enrichment analysis?</h2>
<p class="first-paragraph">&ldquo;Gene set enrichment analysis&rdquo;
refers to the process of discovering the common characteristics
potentially present in a list of genes. When these characteristics
are GO terms, the process is called &ldquo;functional enrichment.&rdquo;</p>
<p>Gene set enrichment analysis can be performed in several ways. One
of the most commonly used is approach is the <i>over-representation
analysis (ORA)</i> that we have described in a previous chapter. An
ORA examines the genes in a list, summarizes the GO annotations for
each, then determines whether there any annotations are statistically
over-represented (compared to an expectation) in that list.</p>
<h2 class="western"><a name="what-tools-are-used-to-perform-enrichment-analysis"></a>
45.2 What tools are used to perform enrichment analysis?</h2>
<p class="first-paragraph">Of all bioinformatics tools developed
to-date, the GO enrichment tools are in the worst shape. Many are in
various states of abandon. Unlike other software that, once
implemented correctly, keeps working over the years, a GO tool
analysis critivally depends on using the most up-to-date GO
annotations. Hence, GO tools require ongoing maintenance and effort,
which their creators are often unable to provide long term.</p>
<p>Most enrichment tools seem to incorporate the word &ldquo;GO&rdquo;
into various &ldquo;cute&rdquo; and humorous&rdquo; tool names, such
as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GOOSE</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GOrilla</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AmiGO</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BINGO</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">QuickGO</font></font>.
Alas the joke gets old very quickly, most of these tools are
surprisingly hard to use and may produce inconsistent results. We
recommend lots of caution when using GO-based tools, as all are
notoriously difficult to evaluate for correctness.</p>
<p>Over the years, we have used and recommended (with some caveats
though) the following tools:</p>
<ul>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="http://bioinfo.cau.edu.cn/agriGO/">AgriGO</a></font>
	Web-based GO Analysis Toolkit and Database for Agricultural
	Community. We love the visualizations that this produces. Short
	review: &rdquo; &ldquo;A++, would do business again.&rdquo;</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://david.ncifcrf.gov/home.jsp">DAVID</a></font>
	This is the GO tool biologists love. It is the &ldquo;most generous&rdquo;
	of them all, as it produces copious amounts of output. DAVID was
	nearly abandoned for more than six years but was finally updated in
	2016. In the meantime (between 2009 and 2016), thousands of
	publications used it to produce GO enrichment analyses on
	increasingly outdated GO data. After 2016 the data has not been
	updated again despite the title on the main page proudly stating
	<b>Welcome to DAVID 6.8, 2003-2018</b>! Is that misleading title
	just an unfortunate interface choice? false advertising? scientific
	misconduct? You decide!</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="http://geneontology.org/">Panther</a></font>
	is offered directly from the GO website. It produces very-limited
	information and no visualization. At least it is probably not
	completely wrong. It is the equivalent of &ldquo;better than
	nothing.&rdquo; The bar is set very low.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://github.com/tanghaibao/goatools">goatools</a></font>
	a command line Python scripts. Published as <font color="#4f81bd"><a href="https://www.nature.com/articles/s41598-018-28948-z">GOATOOLS:
	A Python library for Gene Ontology analyses</a></font> in Scientific
	Reports, 2018.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="http://erminej.chibi.ubc.ca/">ermineJ</a></font>
	Standalone tool with easy to use interface. It has detailed
	documentation. Used to be our first choice.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="http://cbl-gorilla.cs.technion.ac.il/">GOrilla</a></font>
	Web-based; good visualization; downloadable results (as Excel
	files); easily see which genes contribute to which enriched terms;
	results pages indicate the date of last update GO database (often
	within the last week).</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://toppgene.cchmc.org/">ToppFun</a></font>
	- a suite of tools for human genomes that integrates a surprising
	array of data sources.</p>
</ul>
<h2 class="western"><a name="will-different-tools-produce-different-results"></a>
45.3 Will different tools produce different results?</h2>
<p class="first-paragraph">Biostar Quote of the Day: <font color="#4f81bd"><a href="https://www.biostars.org/p/122893/">Why
does each GO enrichment method give different results?</a></font></p>
<p class="block-text-western">I&rsquo;m new to GO terms. In the
beginning, it was fun, as long as I stuck to one algorithm. But then
I found that there are many out there, each with its own advantages
and caveats (the quality of graphic representation, for instance).</p>
<p class="block-text-western">[&hellip;] As a biologist, what should
I trust? Deciding on this or that algorithm may change the whole
story/article!&rdquo;</p>
<p class="first-paragraph">Figuring out how which method to trust is
a challenge with no clear solution. Unlike most other computational
approaches where there is an objective way to validate the quality of
a result, in most GO analyses we don&rsquo;t know how to tell when it
worked correctly.</p>
<p>Overall GO enrichment is surprisingly subjective and different
methods will produce different results.</p>
<p class="block-text-western">Note: When using different tools to
find enrichment, the two different results could be both correct, at
the same time. Or both could be wrong, at the same time. Usually,
they agree on about 50% of targets. If you get over 50% agreement
with different methods, we envy you. Show it off on Instagram.</p>
<h2 class="western"><a name="how-do-i-perform-a-gene-set-enrichment-analysis"></a>
45.4 How do I perform a gene set enrichment analysis?</h2>
<p class="first-paragraph">Most tools follow the same principles:</p>
<ol>
	<li/>
<p class="compact">Enter the group of gene names to be
	studied.</p>
	<li/>
<p class="compact">Enter the group of genes that are the
	&ldquo;background&rdquo; (perhaps all genes for this organism)</p>
	<li/>
<p class="compact">Select a statistical method for comparison.</p>
</ol>
<p class="first-paragraph">For example, let&rsquo;s find genes that
match <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA</font></font>
.:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get human annotation files.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://geneontology.org/gene-associations/goa_human.gaf.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Unzip the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip
goa_human.gaf.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Match the genes named HLA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
goa_human.gaf. | cut -f 3 | grep HLA |sort | uniq | head</font></font></p>
<p class="first-paragraph">it produces the following output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">HHLA1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HHLA2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HHLA3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-B</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-DMA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-DMB</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-DOA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HLA-DOB</font></font></p>
<p class="first-paragraph">Enter this list into a GO enrichment tool
shown here for <font color="#4f81bd"><a href="http://geneontology.org/">Panther</a></font>
via the main <font color="#4f81bd"><a href="http://geneontology.org/">GO</a></font>
website:</p>
<p><img src="0000_html_95e50587fd9087d5.png" name="Image76" align="bottom" hspace="12" width="560" height="307" border="0"/>
</p>
<p>Congrats, you&rsquo;ve just completed a gene set enrichment
process. Next chapters discuss other alternatives.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-agrigo-server"></a>46 Using
the AGRIGO server</h1>
<p class="first-paragraph">Authors: Aswathy Sebastian, Istvan Albert</p>
<p>AgriGO stands for - &ldquo;Go Analysis Toolkit and Database for
Agricultural Community&rdquo; but it has been expanded to support
more organisms.</p>
<ul>
	<li/>
<p class="compact">AgriGo service:
	<font color="#4f81bd"><a href="http://systemsbiology.cau.edu.cn/agriGOv2/">http://systemsbiology.cau.edu.cn/agriGOv2/</a></font></p>
	<li/>
<p class="compact">AgriGO publication: <font color="#4f81bd"><a href="https://academic.oup.com/nar/article/3796337/agriGO-v2-0-a-GO-analysis-toolkit-for-the">agriGO
	v2.0: a GO analysis toolkit for the agricultural community, NAR
	(2017)</a></font></p>
</ul>
<h2 class="western"><a name="authors-note"></a>46.1 Authors note</h2>
<p class="block-text-western">AgriGO is one of our favorite tools for
functional analysis.</p>
<p class="block-text-western">It is fairly simple to execute, the
visualizations are detailed and nicely drawn. We typically start a
functional analysis with an AgriGO visualization.</p>
<h2 class="western"><a name="how-to-use-agrigo"></a>46.2 How to use
AgriGO</h2>
<p class="first-paragraph">AgriGo is a web-based tool for gene
ontology analysis. The original goal of the service was to support
the functional analysis of agricultural species. The site and
services that it offers have been useful beyond the original purpose,
today a large number of the widely studied genome like mouse; chicken
can also be analyzed.</p>
<p><img src="0000_html_99773a673ab59c3a.jpg" name="Image77" align="bottom" hspace="12" width="560" height="135" border="0"/>
</p>
<p>Strengths of AgriGO are:</p>
<ul>
	<li/>
<p class="compact">Usable and straightforward interface.</p>
	<li/>
<p class="compact">Great graphical visualizations.</p>
</ul>
<h2 class="western"><a name="what-is-the-singular-enrichment-analysis-sea-in-agrigo"></a>
46.3 What is the Singular Enrichment Analysis (SEA) in AgriGo?</h2>
<p class="first-paragraph">Singular Enrichment Analysis (SEA)in
AgriGo can be used to find the enriched Gene Ontology terms in user&rsquo;s
list. The input consists of a list of genes. After performing
statistical tests with pre-computed or customized backgrounds, the
user will be provided with enriched GO terms. These results can be
visualized in a GO hierarchical tree structure along with its
significance level.</p>
<h2 class="western"><a name="how-to-use-sea"></a>46.4 How to use SEA?</h2>
<p class="first-paragraph">When SEA as selected as the analysis tool,
the user is required to specify the species and the gene list in the
required format. When species are chosen, allowed forms for gene list
will be displayed on the right-hand side box. For example, if <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Mus
musculus</font></font> is selected, then the gene list should be
submitted.</p>
<p><img src="0000_html_b0dda2f002043cca.png" name="Image78" align="bottom" hspace="12" width="560" height="430" border="0"/>
</p>
<p>The data file used here can be obtained using the commands below.
Here a mouse gene list is selected for AgriGO analysis by using an
expression p-value cut off of 0.05.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O
http://data.biostarhandbook.com/rnaseq/mouse-gene-expression.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
mouse-gene-expression.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'$2
&lt;0.05 {print $1}'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
genelist.txt</font></font></p>
<p class="first-paragraph">Since AgriGo requires mouse gene list to
be in &lsquo;MGI&rsquo; format, submit this to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Batch
Query</font></font> in <font color="#4f81bd"><a href="http://www.informatics.jax.org/batch">Mouse
Genome Informatics</a></font> website.</p>
<p>The MGI ids are obtained from here can be pasted in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">query
list</font></font> box. Once the reference was chosen, submit the
list.</p>
<h2 class="western"><a name="what-are-the-outputs-of-agrigo"></a>46.5
What are the outputs of AgriGO?</h2>
<p class="first-paragraph">A summary report and a table of enriched
GO terms along with their statistical significance will be displayed
once the gene list is processed. The site offers alternative options
for generating graphical visualizations.</p>
<p><img src="0000_html_5ab58f7628d5ade4.jpg" name="Image79" align="bottom" hspace="12" width="560" height="488" border="0"/>
</p>
<p>Shown below is the graphical view of enriched GO terms in
<i>Molecular Function</i> category. Different colors denote the level
of significance. Red means highly significant and orange means of
medium importance. The significance values are shown inside the
boxes.</p>
<p><img src="0000_html_370e539ed56b1726.png" name="Image80" align="bottom" hspace="12" width="560" height="307" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-prepare-a-custom-annotation-for-agrigo"></a>
46.6 How do I prepare a custom annotation for AgriGO?</h2>
<p class="first-paragraph">To use custom annotation with Agrigo, you
need to submit a tab delimited data where one column is an ID and the
other is a GO term. For example, to use human annotations, you would
need to first get the association file you get</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Obtain the human annotation files.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://geneontology.org/gene-associations/goa_human.gaf.gz</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf.gz</font></font></p>
<p class="first-paragraph">Note that you can obtain a protein or
gene-name GO pair via:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'!'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3,5 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">and that looks like:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">DNAJC25-GNG10
  GO:0003924</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DNAJC25-GNG10
  GO:0007186</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NUDT4B
         GO:0003723</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NUDT4B
         GO:0005829</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NUDT4B
         GO:0008486</font></font></p>
<p class="first-paragraph">Now sort this so that it later can be
joined with another file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
goa_human.gaf  </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'!'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3,5 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-k 1b,1 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt</font></font></p>
<p class="first-paragraph">If you had another set of differentially
expressed gene names (using the same naming scheme) then you can sort
those too then join them on this file. Take for example this
enrichment file <font color="#4f81bd"><a href="http://data.biostarhandbook.com/ontology/deseq-output.txt">deseq-output.txt</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/ontology/deseq-output.txt</font></font></p>
<p class="first-paragraph">Take the first 100 differentially
expressed genes and sort them:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
deseq-output.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-100 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-k 1,1 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
top.txt</font></font></p>
<p class="first-paragraph">You can join the two files and cut only
the first two columns; this subselects from the pairs file just those
lines that appear in the top.txt file. The action before produces the
AgriGO input file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">join</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt top.txt -t </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$'</font></font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">\t</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1, 2 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">and yes the same term can be there
multiple times since each is an association to a function.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AARS</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   GO:0000049</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AARS</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   GO:0001942</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AARS</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   GO:0002161</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AARS</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   GO:0003676</font></font></p>
<p class="first-paragraph">You can load the resulting GO annotation
file into AgriGO as the new annotation file</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">join</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 -t </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$'</font></font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">\t</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt top.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1,2 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
agrigo.txt</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-gprofiler-server"></a>47 Using
the g:Profiler server</h1>
<p class="first-paragraph">Quick access:</p>
<ul>
	<li/>
<p class="compact">g:profiler service:
	<font color="#4f81bd"><a href="http://biit.cs.ut.ee/gprofiler">http://biit.cs.ut.ee/gprofiler</a></font></p>
	<li/>
<p class="compact">g:profiler publication:
	<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/27098042">https://www.ncbi.nlm.nih.gov/pubmed/27098042</a></font></p>
</ul>
<h2 class="western"><a name="what-is-the-gprofiler"></a>47.1 What is
the g:Profiler?</h2>
<p class="first-paragraph">A useful website sporting a
counterintuitive and somewhat confusing naming scheme, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">g:Profiler</font></font>
is a suite of several tools.</p>
<p>For the full details we recommend consulting the publication
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/27098042">g:Profiler
&ndash; a web server for functional interpretation of gene lists</a></font>
Nucleic Acids Research 2016;</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">g:Profiler</font></font>
can perform: &ldquo;functional enrichment analysis and mine
additional information. These tools analyse flat or ranked gene lists
for enriched features (g:GOSt; g:Cocoa), convert gene identifiers of
different classes (g:Convert), map genes to orthologous genes in
related species (g:Orth) and find similarly expressed genes from
public microarray datasets (g:Sorter). An additional tool g:SNPense
maps human single nucleotide polymorphisms (SNP) to gene names,
chromosomal locations and variant consequence terms from Sequence
Ontology&rdquo;</p>
<h2 class="western"><a name="what-is-a-standout-feature-of-the-gprofiler"></a>
47.2 What is a standout feature of the g:Profiler?</h2>
<p class="first-paragraph">In our opinion <font face="Consolas, serif"><font size="2" style="font-size: 11pt">g:Profiler</font></font>
visualizes the evidence code for GO terms in the most transparent
way.</p>
<p>As an example load the following human <font color="#4f81bd"><a href="http://data.biostarhandbook.com/courses/2017-852/media/05/gene-list.txt">gene-list.txt</a></font>
into the [g:Profiler] <font color="#4f81bd"><a href="http://biit.cs.ut.ee/gprofiler/index.cgi">gprof-web</a></font>
and view the results.</p>
<p><img src="0000_html_cdd5cf036f23c8b8.png" name="Image81" align="bottom" hspace="12" width="560" height="288" border="0"/>
</p>
<h2 class="western"><a name="what-functionality-does-the-gprofile-have"></a>
47.3 What functionality does the g:Profile have?</h2>
<h3 class="western"><a name="ggost"></a>47.3.1 g:GOSt</h3>
<p class="first-paragraph">Performs functional enrichment analysis:</p>
<p class="block-text-western"><b>g:GOSt</b> performs pathway
enrichment analysis and is the central tool of the g:Profiler web
server. It maps a user-provided gene list to various sources of
functional information and determines significantly enriched
pathways, processes, and other annotations.</p>
<h3 class="western"><a name="gcocoa"></a>47.3.2 g:COCOA</h3>
<p class="first-paragraph">COCOA (Compact Compare of Annotations)
offers simultaneous enrichment analysis of multiple gene lists:</p>
<p class="block-text-western"><b>g:Cocoa</b> provides means to
analyze several gene lists at the same time and compare their
characteristic enriched terms. This is useful in scenarios where an
experimental design involves many comparisons of samples or
individuals, or when one wants to compare directly different clusters
of genes arising from the analysis. Each gene list is analysed for
functional enrichments similarly to g:GOSt and resulting terms are
aligned vertically into a matrix highlighting strongest findings for
every gene list.</p>
<h3 class="western"><a name="gconvert"></a>47.3.3 g:Convert</h3>
<p class="first-paragraph">Allows for automatic conversion of gene
identifiers.</p>
<p class="block-text-western"><b>g:Convert</b> provides a convenient
service to translate identifiers (IDs) of genes, proteins, microarray
probesets and many other types of namespaces. The seamless
translation process works on a mixed set of diverse identifiers and
maps these through Ensembl gene identifiers (ENSG) as reference. In
cases of multiple identifiers, all relevant combinations are
highlighted. At least 13 types of IDs are supported for all of the
213 species available in g:Profiler, and at least 40 types of IDs for
more than 50 species.</p>
<h3 class="western"><a name="gorth"></a>47.3.4 g:Orth</h3>
<p class="first-paragraph">Maps related genes across species</p>
<p class="block-text-western"><b>g:Orth</b> allows the user to map a
list of genes of interest to homologous genes in another related
organism. Many experiments are conducted in model organisms and
knowledge from such experiments is transferred to other organisms to
compare or complement previous findings. g:Orth uses g:Convert to map
gene IDs to Ensembl ENSG identifiers. Further mapping to orthologous
genes in other organisms is also based on Ensembl data. We provide
cross-references between all organisms in g:Profiler. Queries are
limited to tiers according to classes of species (animals, plants,
fungi).</p>
<h3 class="western"><a name="gsorter"></a>47.3.5 g:Sorter</h3>
<p class="first-paragraph">Findins similar genes in transcriptomics
data</p>
<p class="block-text-western"><b>g:Sorter</b> is a tool for finding
lists of co-expressed genes from public transcriptomics datasets.
Thousands of microarray and RNA-seq experiments have been conducted
in the past decades. The majority of published studies have been
accumulated in databases like ArrayExpress and Gene Expression
Omnibus. We have downloaded 7878 datasets for 18 species from
ArrayExpress and provide gene co-expression similarity searches using
six most common statistical methods. The datasets can be searched
rapidly with keywords. The input of g:Sorter is a single gene and a
dataset of interest and the result is a sorted list of genes that are
similarly expressed with the gene of interest. These lists can be
integrated into functional enrichment analysis. For comprehensive
global gene expression similarity queries, as well as support for
more species and platforms we suggest to use Multi Experiment Matrix
(MEM) tool</p>
<h3 class="western"><a name="gsnpense"></a>47.3.6 g:SNPense</h3>
<p class="first-paragraph">SNP identifier mapping</p>
<p class="block-text-western">With the rapid growth of whole genome
sequencing technology, researchers are uncovering extensive genetic
variation and large collections of known SNP are available for human
and other species. In order to easily map SNP identifiers
(e.g.&nbsp;<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rs4244285</font></font>)
to gene names, chromosomal coordinates and retrieve their functional
consequences we now provide a new service called g:SNPense.
Information about genome variants is retrieved from dbSNP and mapped
to NCBI Genes. Potential functional consequences of the variants are
retrieved from Ensembl Variation data and grouped into 35 Sequence
Ontology terms of various severity. g:SNPense is a potential entry
point to g:GOSt and its functional annotation pipeline and enrichment
analysis.</p>
<h2 class="western"><a name="how-to-use-gprofiler-at-the-command-line"></a>
47.4 How to use g:profiler at the command line</h2>
<p class="first-paragraph">An alternative to the web interface is the
command line client to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">g:profiler</font></font>.
First install the command line client:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Install the client.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
install gprofiler-official</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get help on the client parameters.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gprofiler_cli.py</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
--help</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Example run from the documentation.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gprofiler_cli.py</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-o scerevisiae </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;swi4
swi6&quot;</font></font></font></p>
<p class="first-paragraph">Let&rsquo;s run the g:profiler on the up
regulated genes from the <font color="#4f81bd"><a href="#ch054.xhtml#zikaredo">Zika
redo chapter</a></font>. Get the data</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/redo/zika/zika-up-regulated.csv</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
zika-up-regulated.csv | cut -f 1 -d , | head</font></font></p>
<p class="first-paragraph">produces the list of up regulated genes</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ELAVL3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FAM53C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SLC22A17</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">WDFY1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TBL1XR1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SON</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">EIF4G2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOGA3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SDCBP</font></font></p>
<p class="first-paragraph">The file above has too many candidate
genes. Usually this means that the output will be overly complicated,
will pull in too many alternative explanations. Biologists tend to
subfilter the results to let them isolate one particular function.
One typical filtering is that on the magnitude of change. For
example, we may want to select genes that change more than 4 fold
(this would correspond to a log2 of 2). In addtion the g:profiler
client requires a list of symbols, all on the same line so will
replace the new line with a space character.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
zika-up-regulated.csv | grep -v gene | awk -F , ' $6 &gt; 2 { print
$1 }' | tr &quot;\n&quot; &quot; &quot; &gt; query.txt</font></font></p>
<p class="first-paragraph">like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NOG
SNORA18 SESN2 JDP2 ASNS RAB3IL1 ...</font></font></p>
<p class="first-paragraph">run the profiler</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gprofiler_cli.py
-o hsapiens -q query.txt  &gt; results.txt</font></font></p>
<p class="first-paragraph">The resulting file contains a large number
of entries, the first few columns, when filtered for biological
process are:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt |  cut -f 9-12 | grep BP | head</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0034976
 BP  1   response to endoplasmic reticulum stress</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0070059
 BP  1   intrinsic apoptotic signaling pathway in response to
endoplasmic reticulum stress</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0036003
 BP  1   positive regulation of transcription from RNA polymerase II
promoter in response to stress</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0006986
 BP  1   response to unfolded protein</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0035966
 BP  1   response to topologically incorrect protein</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0030968
 BP  1   endoplasmic reticulum unfolded protein response</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0010033
 BP  1   response to organic substance</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0034620
 BP  1   cellular response to unfolded protein</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0035967
 BP  1   cellular response to topologically incorrect
protein</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO:0071310
 BP  1   cellular response to organic substance</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-david-server"></a>48 Using the
DAVID server</h1>
<p class="first-paragraph">Authors: Aswathy Sebastian, Istvan Albert</p>
<p>The word DAVID is an acronym for &ldquo;The Database for
Annotation, Visualization and Integrated Discovery&rdquo;</p>
<ul>
	<li/>
<p class="compact">DAVID service:
	<font color="#4f81bd"><a href="https://david.ncifcrf.gov/home.jsp">https://david.ncifcrf.gov/home.jsp</a></font></p>
	<li/>
<p class="compact">DAVID publication:
	<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/19131956">https://www.ncbi.nlm.nih.gov/pubmed/19131956</a></font></p>
</ul>
<h2 class="western"><a name="authors-note-1"></a>48.1 Authors note</h2>
<p class="block-text-western">DAVID is the tool we love to hate.</p>
<p class="block-text-western">Amongst all functional analysis tools
we feel DAVID is the wonkiest. It is awkward to use, has a
repetitive, redundant and circularly confusing naming scheme. Within
five clicks we lose our way and need to start a new.</p>
<p class="block-text-western">Yet biologists <b>LOVE</b> DAVID.</p>
<p class="block-text-western">There is something about the way this
tool works and how it presents information that matches the way life
scientists are trained to think. It &ldquo;speaks&rdquo; to
biologists in a way no other tool does. It is also a tool with the
most &ldquo;generous&rdquo; enrichment protocols.</p>
<p class="block-text-western">DAVID will dole out seemingly
relentless sweet suggestions of what the data is about. DAVID will
not limit your imagination with mundane statistics. As a matter of
fact we believe that DAVID does not even correct for multiple
comparisons - perhaps one of the cardinal sin of statistics. On the
other hand, it is not yet proven that the Gene Ontology should be
described with statistical models in the first place! Suggestion by
DAVID when coupled with biological intuition and other supporting
information can help scientists discover phenomena that other
approaches miss. At the same time though, it can also lead the
unwitting deep down into the tar pits of irreducibility.</p>
<p class="block-text-western">Try DAVID and see how you like it. Did
we mention that we love to hate it? When we want to &ldquo;get rid&rdquo;
of someone asking too many questions about functional analysis we
send them to DAVID. Nobody ever comes back.</p>
<h2 class="western"><a name="what-is-david"></a>48.2 What is DAVID?</h2>
<p class="first-paragraph">DAVID is a functional enrichment analysis
tool that can be used to understand the biological relevance of a
list of genes. The tool allows users to:</p>
<ul>
	<li/>
<p class="compact">Identify functionally essential genes in a
	gene list.</p>
	<li/>
<p class="compact">Get the enriched biological terms in a gene
	list.</p>
	<li/>
<p class="compact">Obtain the biological processes that a set
	of up-regulated and down-regulated genes are involved in.</p>
</ul>
<p class="first-paragraph">The tool is a valuable resource - but like
many other scientific software, it suffers from very confusing
interface elements and redundant windows that pop over others. It is
very easy to get disoriented while using it.</p>
<h2 class="western"><a name="what-does-david-do"></a>48.3 What does
DAVID do?</h2>
<p class="first-paragraph">When presented a gene list like the one
shown below</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Tmem132a</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Myl3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Myl4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Tnnt2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Hspb7</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Tnni3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Actc1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">DAVID will generate functional
interpretations:</p>
<ul>
	<li/>
<p class="compact">14 genes (My13, My14, Tnni3,&hellip;) from
	this list are involved in <i>Cardiac Muscle Contraction</i></p>
	<li/>
<p class="compact">4 genes (Actc1, Myh6, Myh7, Tnnt2) have the
	role of <i>ATPase Activity</i></p>
</ul>
<p class="first-paragraph">and so on. The value added by the tools is
that it helps find groups of genes that share specific properties
that might be interesting to a life scientist.</p>
<h2 class="western"><a name="what-are-the-different-steps-in-a-david-analysis"></a>
48.4 What are the different steps in a DAVID analysis?</h2>
<p class="first-paragraph">A typical gene enrichment and functional
annotation workflow in DAVID has the following steps:</p>
<ol>
	<li/>
<p class="compact">Load a gene list into the site.</p>
	<li/>
<p class="compact">Explore details through the annotation
	table, chart or clustering reports.</p>
	<li/>
<p class="compact">Export and save the results.</p>
</ol>
<h2 class="western"><a name="how-do-i-start-an-analysis-with-david"></a>
48.5 How do I start an analysis with DAVID?</h2>
<p class="first-paragraph">Every analysis in DAVID starts with the
submission of a gene list.</p>
<p>A gene list is a selected set of genes from an experiment that you
might be interested in learning more about. This gene list will be
evaluated against a so-called &ldquo;background&rdquo; list to detect
attributes that are only present in the gene list.</p>
<p>Often, the background is taken as all genes of an organism, but,
in some cases, it is more appropriate to use a different gene set as
background. For example, if the list of interesting genes was taken
from a microarray experiment, the background should be only those
genes queried by the microarray. Gene lists from sequencing assays
generally can use all the organism&rsquo;s genes as the background.</p>
<p>To start an analysis, users can either click on the &lsquo;Start
Analysis&rsquo; tab or choose one of the tools from &lsquo;Shortcut
to DAVID Tools&rsquo; tab. Both will take you to the upload wizard:</p>
<p><img src="0000_html_dbc8e7f8bddcc12b.png" name="Image82" align="bottom" hspace="12" width="560" height="393" border="0"/>
</p>
<p>An example gene list can be obtained using the commands below.
Here a mouse gene list is selected for DAVID analysis by using an
expression p-value cut off of 0.05.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O
http://data.biostarhandbook.com/rnaseq/mouse-gene-expression.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
mouse-gene-expression.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'$2
&lt;0.05 {print $1}'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">genelist.txt</font></font></p>
<p class="first-paragraph">Once the list is submitted, a selection of
the input species might be needed in case there are ambiguities, and
the input gene names may refer to multiple species.</p>
<p>A note on what is happening behind the scenes when a gene list is
submitted:</p>
<ul>
	<li/>
<p class="compact">Each input gene id gets converted into a
	so-called <i>DAVID Gene ID</i>, a unique identifier specific to
	DAVID. All the subsequent analysis will be applied to the DAVID IDs
	in the selected gene list.</p>
	<li/>
<p class="compact">The DAVID Gene ID connects the input genes
	across all other databases in DAVID. DAVID supports nearly 30
	different input id types and a broad set of annotation categories.</p>
</ul>
<p class="first-paragraph">Once the gene list is submitted, you will
need to explore the submitted gene list with DAVID tools and their
results are named somewhat redundantly and confusingly (at least we
have a hard time remembering which one does what):</p>
<ul>
	<li/>
<p class="compact">Functional Annotation.</p>
	<li/>
<p class="compact">Gene Functional Classification.</p>
</ul>
<h2 class="western"><a name="what-is-the-functional-annotation-tool"></a>
48.6 What is the Functional Annotation Tool?</h2>
<p class="first-paragraph">These tools assist users in the biological
interpretation of their data. It maps the genes to annotation content
from different databases DAVID computes the enriched functionality of
the gene list.</p>
<p>The results of a Functional Annotation process is a Functional
Annotation Summary where the user has options to further select
annotation categories of interest and explore them in detail.</p>
<h2 class="western"><a name="what-is-the-functional-annotation-summary"></a>
48.7 What is the Functional Annotation Summary?</h2>
<p class="first-paragraph">The selections of the Functional
Annotation Tool and Annotation Summary page are shown below:</p>
<p><img src="0000_html_64db671fc0a541d.png" name="Image83" align="bottom" hspace="12" width="560" height="401" border="0"/>
</p>
<p>Expanding on each selected category gives the annotation details
of the genes involved in it.</p>
<p>The combined view allows you to explore the combined annotation
content of all selected categories.</p>
<h2 class="western"><a name="how-do-i-explore-the-functional-annotation-summary"></a>
48.8 How do I explore the Functional Annotation Summary?</h2>
<p class="first-paragraph">The selected functional categories of
interest in the Functional Annotation Summary page can be analyzed
through any of the following reports:</p>
<ol>
	<li/>
<p class="compact"><b>Functional Annotation Chart:</b> A view
	focused on annotation terms that lists all enriched annotation terms
	and their associated genes.</p>
	<li/>
<p class="compact"><b>Functional Annotation Clustering:</b>
	Groups similar annotation terms into clusters based on their
	similarity and the genes they share in common.</p>
	<li/>
<p class="compact"><b>Functional Annotation Table:</b>
	Gene-centric view that lists all genes and their associated
	annotation terms (selected only). No statistics are provided here.</p>
</ol>
<h2 class="western"><a name="what-is-a-functional-annotation-chart"></a>
48.9 What is a Functional Annotation Chart ?</h2>
<p class="first-paragraph">The functional annotation chart is a
tabular form of all enriched annotation terms associated with a gene
list. The enriched annotation terms presented here can be redundant.
The enrichment P-value is the so-called <i>Ease Score</i>, calculated
via Fisher&rsquo;s Exact test.</p>
<p><img src="0000_html_d61dd15282399c8.png" name="Image84" align="bottom" hspace="12" width="560" height="347" border="0"/>
</p>
<p>In the example shown here, 233 annotation terms are associated
with 96 input genes. 18 of these 96 genes are involved in muscle
protein functionality. Clicking on the blue bar lists these genes.
The related terms (RT) search shows other annotation terms that these
18 genes are associated with. In other words, RT lists all annotation
terms above a certain similarity threshold that are shared by these
genes.</p>
<h2 class="western"><a name="what-is-functional-annotation-clustering"></a>
48.10 What is Functional Annotation Clustering?</h2>
<p class="first-paragraph">Annotation terms associated with genes can
be redundant. DAVID measures the degree of similarity between these
terms using kappa statistics based on the number of genes they share.
All the terms above a certain similarity threshold are then grouped
together using a clustering algorithm. Thus each annotation cluster
groups terms with similar biological meaning.</p>
<p>The clustering report below shows that 233 enriched annotation
terms associated with 96 DAVID input ids are clustered into 44
groups.</p>
<p><img src="0000_html_27bd730289570e59.png" name="Image85" align="bottom" hspace="12" width="560" height="364" border="0"/>

Cluster1 in the example shown here has six annotation terms that
share the associated genes. The relationship between the genes and
these annotation terms is shown in the image below. The green area
represents the sharing of genes by these annotation terms. The black
area denotes that there is no association between the terms and the
genes. A large green area means that more genes are shared, implying
a greater similarity between these terms. The terms are thus grouped
together in a single cluster.</p>
<p><img src="0000_html_eba4b2d3b5ca5354.png" name="Image86" align="bottom" hspace="12" width="560" height="380" border="0"/>
</p>
<h2 class="western"><a name="what-is-a-functional-annotation-table"></a>
48.11 What is a Functional Annotation Table?</h2>
<p class="first-paragraph">The Functional Annotation Table is a third
form of reporting the annotation content (selected in annotation
summary) of a gene list. It provides a gene-centric view by listing
the genes and their associated terms. No statistics are reported on
this page. A blue header row separates each gene.</p>
<p><img src="0000_html_72771d6656078499.png" name="Image87" align="bottom" hspace="12" width="560" height="245" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-gene-functional-classification-tool"></a>
48.12 What is the Gene Functional Classification Tool?</h2>
<p class="first-paragraph">DAVID has yet another tool to help the
user interpret the biological meaning of their data: the Gene
Functional Classification Tool. This tool provides a view similar to
annotation clustering, except that the genes are clustered instead of
the annotation terms.</p>
<p>The Gene Functional Classification tool reports all the
functionally related genes in the user&rsquo;s gene list. Only genes
not mapped to any of the functional categories are omitted from the
table.</p>
<p><img src="0000_html_520d6800712e4cf1.png" name="Image88" align="bottom" hspace="12" width="560" height="375" border="0"/>
</p>
<h2 class="western"><a name="what-is-an-ease-score"></a>48.13 What is
an EASE Score?</h2>
<p class="first-paragraph">In the DAVID annotation system, Fisher&rsquo;s
Exact Test is used to measure gene-enrichment in annotation terms.
The Ease score is a slightly modified Fisher&rsquo;s Exact p-value.
The smaller the value, the more enriched the associated term.
Consider,</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">a</font></font>
	= number of genes in the user&rsquo;s list that belong to a
	particular pathway, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Z</font></font>.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">b</font></font>
	= total number of genes in the user&rsquo;s list.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
	= number of genes in the background that belong to pathway <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Z</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
	= total number of genes in the background.</p>
</ul>
<p class="first-paragraph">Fisher&rsquo;s Exact Test would compare
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">a/b</font></font>
genes in the user&rsquo;s list with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A/B</font></font>
genes in the background to see if the genes in the user&rsquo;s list
belong to pathway &lsquo;Z&rsquo; by anything more than random
chance.</p>
<p>DAVID compares <font face="Consolas, serif"><font size="2" style="font-size: 11pt">(a-1)/b</font></font>
with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A/B</font></font>
in Fisher&rsquo;s Exact test to check significance. The p-value
obtained is reported as the EASE score.</p>
<h2 class="western"><a name="what-is-the-kappa-statistic"></a>48.14
What is the Kappa statistic?</h2>
<p class="first-paragraph">The clustering algorithm in DAVID is based
on the hypothesis that similar annotations should have similar gene
members. It uses the Kappa statistic to measure the degree of shared
genes between two annotations. This is followed by heuristic
clustering to group similar annotations according to Kappa values.</p>
<p><font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Cohen%27s_kappa">Cohen&rsquo;s
kappa coefficient (k)</a></font> measures the degree of agreement
between qualitative items. DAVID uses kappa values to estimate the
functional relationship between gene pairs.<br/>
A higher Kappa value
in DAVID means that genes are more similar in their annotation
content. A zero or a negative value for Kappa indicates that genes
share no functionality.</p>
<h2 class="western"><a name="what-does-the-gene-id-conversion-tool-do"></a>
48.15 What does the Gene ID Conversion Tool do?</h2>
<p class="first-paragraph">This tool enables the conversion of an
accession number from one specification to another. There are around
30 different supported id types, which makes this tool quite useful.
The gene id conversion page and the results page are shown below.</p>
<p><img src="0000_html_9306e4b106eb7a4c.png" name="Image89" align="bottom" hspace="12" width="560" height="204" border="0"/>
</p>
<h2 class="western"><a name="what-are-some-pros-and-cons-of-david"></a>
48.16 What are some pros and cons of DAVID?</h2>
<h3 class="western"><a name="pros"></a>48.16.1 Pros</h3>
<ul>
	<li/>
<p class="compact">The database in DAVID provides the user
	with &ldquo;many interesting&rdquo; results. DAVID is very well
	suited for exploratory analyses as it does not filter the results.</p>
	<li/>
<p class="compact">Since it is an online resource, no
	installation is required.</p>
	<li/>
<p class="compact">The statistical model is simple and
	straightforward.</p>
	<li/>
<p class="compact">Every tool can be used independently of
	another. In other words, you don&rsquo;t need to follow a set series
	of actions to get the results.</p>
</ul>
<h3 class="western"><a name="cons"></a>48.16.2 Cons</h3>
<ul>
	<li/>
<p class="compact">All results are in &ldquo;flat&rdquo;
	tabular form. No visualizations are provided.</p>
	<li/>
<p class="compact">It&rsquo;s hard to gain a higher level
	overview of the relationship between the terms that are enriched.</p>
	<li/>
<p class="compact">Every click in DAVID opens a new window.
	You can feel lost after just a few clicks.</p>
	<li/>
<p class="compact">DAVID presents the user with many different
	views of the same result with the aim to help better interpret the
	data. However, this can be very confusing; sometimes it feels like
	running in circles.</p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-goatools-package"></a>49 Using
the goatools package</h1>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
was originally designed as a Python package to be used from within
other Python programs. The package also supports command line usage
and offers utility in form of standalone programs that get installed
when you install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>.
The library has been published as <font color="#4f81bd"><a href="https://www.nature.com/articles/s41598-018-28948-z">GOATOOLS:
A Python library for Gene Ontology analyses</a></font></p>
<ul>
	<li/>
<p class="compact">goatools code:
	<font color="#4f81bd"><a href="https://github.com/tanghaibao/goatools">https://github.com/tanghaibao/goatools</a></font></p>
	<li/>
<p class="compact">goatools publication:
	<font color="#4f81bd"><a href="https://www.nature.com/articles/s41598-018-28948-z">https://www.nature.com/articles/s41598-018-28948-z</a></font></p>
</ul>
<h2 class="western"><a name="install-goatools"></a>49.1 Install
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font></h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">source</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
activate bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
install pygraphviz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
install goatools pydot statsmodels fisher</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Make a directory for reference data.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p refs</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Obtain the gene ontology term definitions.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Gets the file only once (-nc) and places into the prefix (-P)</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-nc -P ~/refs http://geneontology.org/ontology/go-basic.obo</font></font></p>
<h2 class="western"><a name="plot-go-terms"></a>49.2 Plot GO terms</h2>
<p class="first-paragraph">There are several levels of visualizations
that you might need, even when wanting to see a single term in
context. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
package offers various visualization capabilities. More examples can
be seen at:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/tanghaibao/goatools/blob/master/doc/md/README_go_plot.md">https://github.com/tanghaibao/goatools/blob/master/doc/md/README_go_plot.md</a></font></p>
</ul>
<p class="first-paragraph">below we reproduce a few examples from the
page above:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Plot a GO term. Creates the GO_lineage.png file.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">plot_go_term.py</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
--term GO:0031985 ~/refs/go-basic.obo</font></font></p>
<p class="first-paragraph">Will render an image like:</p>
<p><img src="0000_html_83a3f708bb629049.gif" name="Image90" align="bottom" hspace="12" width="560" height="891" border="0"/>
</p>
<p>There is a second way to plot the same information.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">go_plot.py
 --obo ~/refs/go-basic.obo -o output.png GO:0031985</font></font></p>
<p class="first-paragraph"><img src="0000_html_2f0f1b38251c7d3e.png" name="Image91" align="bottom" hspace="12" width="560" height="766" border="0"/>
</p>
<p>Add the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-r</font></font>
flag to plot all relationships:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">go_plot.py
 --obo ~/refs/go-basic.obo -r -o output.png GO:0031985</font></font></p>
<p class="first-paragraph">will now add even more context:</p>
<p><img src="0000_html_bf4986d4287db1cf.png" name="Image92" align="bottom" hspace="12" width="560" height="540" border="0"/>
</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
offer a variety of options for coloring and displaying the nodes in
your image. When it comes to publishing your results, this package
can help you produce compelling plots that show the various
functional terms in the proper context.</p>
<h2 class="western"><a name="finding-enrichment-with-goatools"></a>49.3
Finding enrichment with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font></h2>
<p class="first-paragraph">Suppose you have a list of protein or gene
ids and you wanted to know what function is representative of them.
For example, if we were to take the five most annotated genes in GO
is there something common about them? Also, ask yourself should there
be anything in common about them? We can find enrichment for a group
of genes with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
but the process is a little more tedious than you may anticipate.</p>
<p>You see the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
package suffers from what I call &ldquo;over-engineering&rdquo;. It
is a common disease that affects the best and the brightest of
programmers even more so than beginners. The design of the library is
overcomplicated, the examples in the manual lack in specificity. What
is a population file, what is an association file, what is a study
file? It is a shame, the package has been in development for almost a
decade, and clearly, the programmers that work on it are quite
talented. Still, curiously simple features and explanations are
missing.</p>
<p>For example, the enrichment tool cannot operate on the basic data
format that GO is distributed as. Think about that for a second. The
Nature Methods journal published a paper on GO data processing and
visualization, one essential component of that software cannot
process a standard GO annotation file. To use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">goatools</font></font>
on the &ldquo;official&rdquo; data format, the dear readers would
first need to transform the GAF file into a so-called association
file. We&rsquo;ll do that here:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get the human gene annotations.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-nc -P ~/refs
http://geneontology.org/gene-associations/goa_human.gaf.gz</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/refs/goa_human.gaf.gz</font></font></p>
<p class="first-paragraph">We now need to transform this <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gaf</font></font>
file into an association file where row contains all the GO terms for
an id, like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">id
 GO:1;GO2;GO10</font></font></p>
<p class="first-paragraph">the following <font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font>
script will do that:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Simplify the data into two columns first, on ID and GO term per
line.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/refs/goa_human.gaf </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-v ! </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3,5 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Linearize the file with one ID and multiple GO terms on each
line.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'{if(a[$1])a[$1]=a[$1]&quot;;&quot;$2;
else a[$1]=$2;} END {for (i in a) print i, a[i];}'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
association</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Let's consider the population (background) as all of the existing
products.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
population</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
We want to know what functions characterize the five most annotated
genes.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
This takes a bit of work, we need to squeeze the spaces and find the
right column.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pairs.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1 </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-rn </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tr</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-s </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'
'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 3 -d </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'
'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-5 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
study</font></font></p>
<p class="first-paragraph">The study file contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TP53</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">EGFR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRB2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">UBC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CTNNB1</font></font></p>
<p class="first-paragraph">We now have all the data, the association,
the background and the study file that contains the ten most
annotated gene names. Finally, let&rsquo;s get the answer, what are
common functions that these genes have? Lets run the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">find_enrichment.py</font></font>
script:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">find_enrichment.py
study population association --obo ~/refs/go-basic.obo  &gt;
results.txt</font></font></p>
<p class="first-paragraph">The file has substantial information, at
the beginning it states:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">13
GO terms found significant (&lt; 0.05=alpha) after multitest
correction: local bonferroni</font></font></p>
<p class="first-paragraph">Let&rsquo;s visualize these terms.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Subselect molecular function (MF).</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
results.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
GO: </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
MF </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
terms.txt</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Plot the terms</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">go_plot.py</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
--obo ~/refs/go-basic.obo --go_file terms.txt</font></font></p>
<p class="first-paragraph">Produces a large image (we show it
cropped, here):</p>
<p><img src="0000_html_6a49478e5c83a7ed.png" name="Image93" align="bottom" hspace="12" width="560" height="151" border="0"/>
</p>
<p>It has many functions. Note how we only had five genes under study
(though these are the most annotated genes) The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results.txt</font></font>
file has 1206 functions just for molecular function. Out of these
those that contain the word binding are:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of euchromatin binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">positive
regulation of DNA-binding transcription factor activity</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">positive
regulation of core promoter binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of core promoter binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of DNA-binding transcription factor activity</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of chromatin binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">positive
regulation of transcription regulatory region DNA
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">nucleotide-binding
oligomerization domain containing signaling
pathway</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">nucleotide-binding
domain, leucine rich repeat containing receptor signaling
pathway</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of transcription regulatory region DNA binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">positive
regulation of DNA binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulation
of DNA binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">positive
regulation of binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">protein
phosphatase binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">phosphatase
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">protein
kinase binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">kinase
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">enzyme
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">disordered
domain specific binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ubiquitin
protein ligase binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ubiquitin-like
protein ligase binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chromatin
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">signaling
receptor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">protease
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">protein
domain specific binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNA
polymerase II transcription factor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">epidermal
growth factor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">neurotrophin
TRKA receptor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">neurotrophin
TRK receptor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cadherin
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">alpha-catenin
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">neurotrophin
receptor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TFIID-class
transcription factor complex binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">I-SMAD
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">insulin
receptor substrate binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNA
polymerase II basal transcription factor binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNA
binding</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The results indicate that that the
selected genes have all have roles in regulation via DNA binding.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="using-the-erminej-tool"></a>50 Using the
ErmineJ tool</h1>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://erminej.chibi.ubc.ca/download/">ErmineJ</a></font>
is a functional analysis tool that can be used to determine the
functional categories or biological pathways that are enriched in a
list of genes generated by an experiment. The gene list will be
tested with a reference set of Gene Ontology (GO) terms to identify
relevant biological pathways.</p>
<ul>
	<li/>
<p class="compact">ErmineJ code:
	<font color="#4f81bd"><a href="http://erminej.chibi.ubc.ca/download/">http://erminej.chibi.ubc.ca/download/</a></font></p>
	<li/>
<p class="compact">ErmineJ publication:
	<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1310606/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1310606/</a></font></p>
</ul>
<h2 class="western"><a name="authors-note-2"></a>50.1 Authors note</h2>
<p class="block-text-western">There was a time when ErmineJ was the
&ldquo;GO to&rdquo; choice for GO analysis (see how easy it is to
make puns using the word &ldquo;GO&rdquo;?). S&iacute;, AmiGO!</p>
<p class="block-text-western">Those times are past, but it still a
unique program that it runs offline on your computer.</p>
<p class="block-text-western">Besides, it is one of the few tools
that account for &ldquo;multifunctionality&rdquo; (see below) and
offers a visualization that includes tiny heatmaps for each entry.
Quite the unique feature! Recently we found ourselves using ErmineJ a
lot less, we still include it here as a useful alternative.</p>
<h2 class="western"><a name="how-to-use-erminej"></a>50.2 How to use
ErmineJ</h2>
<p class="first-paragraph"><img src="0000_html_fe9d624c00992f35.jpg" name="Image94" align="bottom" hspace="12" width="560" height="133" border="0"/>
</p>
<p>In the simplest of terms, you start with a text file that contains
genes on each row and measurements in each column. ErmineJ&rsquo;s
purpose is to find the common functions across the genes that are
above a certain score.</p>
<h2 class="western"><a name="what-type-of-analyses-does-erminej-offer"></a>
50.3 What type of analyses does ErmineJ offer?</h2>
<p class="first-paragraph">ErmineJ offers different ways to analyze
the gene sets. These include,</p>
<ul>
	<li/>
<p class="compact"><b>ORA</b> - Over-Representation Analysis
	(finds enriched functions).</p>
	<li/>
<p class="compact"><b>GSR</b> - Examines the distribution of
	gene scores in each set.</p>
	<li/>
<p class="compact"><b>CORR</b> - Uses correlation of
	expression profiles.</p>
</ul>
<h2 class="western"><a name="how-is-erminej-different-from-other-options"></a>
50.4 How is ErmineJ different from other options?</h2>
<ul>
	<li/>
<p class="compact">ErmineJ runs on Java, hence it is easy to
	install and run.</p>
	<li/>
<p class="compact">It runs on a desktop ( not browser based ).</p>
	<li/>
<p class="compact">It has a simple, easy-to-use graphical user
	interface.</p>
	<li/>
<p class="compact">ErmineJ provides data visualization.</p>
	<li/>
<p class="compact">It computes a multi-functionality score.</p>
</ul>
<h2 class="western"><a name="what-are-the-input-files-in-erminej"></a>
50.5 What are the input files in ErmineJ?</h2>
<p class="first-paragraph">The main input files for ErmineJ are</p>
<ol>
	<li/>
<p class="compact">The annotation file. These specify the
	gene-to-function information.</p>
	<li/>
<p class="compact">The gene score file. The gene score file
	specific the scores assigned to genes.</p>
	<li/>
<p class="compact">The optional expression data. This file
	assigns numbers across multiple conditions for each gene.</p>
</ol>
<h2 class="western"><a name="what-are-the-annotation-files"></a>50.6
What are the annotation files?</h2>
<p class="first-paragraph">The annotations represent the names and
relationships among GO terms. ErmineJ comes with a GO XML file
containing the names and relationships among GO terms. The ErmineJ
website offers options to download annotations for many commonly
studied organisms ( see details below ).</p>
<ul>
	<li/>
<p class="compact">Download <font color="#4f81bd"><a href="http://chibi.ubc.ca/microannots/">ErmineJ
	annotation files</a></font></p>
</ul>
<p class="first-paragraph">There quite many files there. Our examples
work on with the generic human and generic mouse files, those can be
downloaded from the command line with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O http://chibi.ubc.ca/microannots/Generic_human_noParents.an.txt.gz</font></font></p>
<p class="first-paragraph">and</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O http://chibi.ubc.ca/microannots/Generic_mouse_noParents.an.txt.gz</font></font></p>
<p class="first-paragraph">When you start ErmineJ, you will need to
load one of these files into the system. Only one annotation can be
active at any given moment though.</p>
<p><img src="0000_html_eb49161e1f4b1de9.png" name="Image95" align="bottom" hspace="12" width="560" height="106" border="0"/>
</p>
<h2 class="western"><a name="what-is-a-gene-score-file"></a>50.7 What
is a gene score file?</h2>
<p class="first-paragraph">Gene scores are used to select genes, then
based on the selection the tool determines which functions are
enriched. It is a tab delimited file with unique gene identifiers as
the first column and score values as the second column.</p>
<p>For example, if you had a differential expression file (we&rsquo;ll
show you how to make these in the RNA-Seq section) <font color="#4f81bd"><a href="http://data.biostarhandbook.com/ontology/deseq-output.txt">that
looked like this</a></font> then selecting and saving the first and
last columns of would create a gene score file.</p>
<p>A score may be any value applied to a gene that represents some
measure of interest, frequently that is a p-value or fold-change.
Non-numeric values such as &lsquo;NaN&rsquo; or &lsquo;#NUM!&rsquo;
will be interpreted as zero. The first line is considered to be a
header line.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">id
     padj</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">WSB1
   5.26199960214757e-51</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TPBG
   7.77064023248874e-48</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XBP1
   1.88906929176218e-46</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ASNS
   3.89107811001364e-39</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SLC7A5
 1.68112640483516e-35</font></font></p>
<p class="first-paragraph">You can make this gene score yourself
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O http://data.biostarhandbook.com/ontology/deseq-output.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
deseq-output.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1,8 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
gene_scores.txt</font></font></p>
<h2 class="western"><a name="what-is-the-expression-data"></a>50.8
What is the expression data?</h2>
<p class="first-paragraph">Expression data contain the measurements
(e.g., read counts) associated with each gene and condition in a
tab-delimited text file format. It is not used to determine
functional enrichment. Instead, it is displayed when groups of genes
are visualized.</p>
<p>The first row is the header line. Missing values must be indicated
by blank, not by &lsquo;NA&rsquo; or any other non-numerical values.</p>
<p>An example raw data file may look like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene
      C1     C2      C3      M1      M2      M3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Myl3
   1207.16 1345.04 1247.69 2222.91 3041.39 2819.01</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Tnnt2
  2401.01 2542.16 2712.17 3648.55 5133.96 4505.65</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Mybpc3
 1689.82 2026.43 2173.50 3070.45 4439.41 3540.34</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">To get our <font color="#4f81bd"><a href="http://data.biostarhandbook.com/ontology/normalized-matrix.txt">example
expression</a></font> data from the command line do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">    <font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O http://data.biostarhandbook.com/ontology/normalized-matrix.txt</font></font></p>
<p class="first-paragraph">You will specify this expression data at
the same time that we enter the gene score file.</p>
<h2 class="western"><a name="how-do-i-start-an-analysis-in-erminej"></a>
50.9 How do I start an analysis in ErmineJ?</h2>
<p class="first-paragraph">When we start up ErmineJ, the following
page will be presented:</p>
<p><img src="0000_html_252863e231040c0b.png" name="Image96" align="bottom" hspace="12" width="560" height="308" border="0"/>
</p>
<p><b>GO XML File:</b> ErmineJ will automatically download and store
this file in the user&rsquo;s ermineJ.data directory. This file is
required even if you are not using GO.</p>
<p><b>Annotation File</b> The annotation file appropriate to the
experiment can be <i>downloaded from ErmineJ website using </i><font color="#4f81bd"><a href="http://www.chibi.ubc.ca/microannots/"><i>this</i></a></font><i>
link.</i> It is recommended to use &lsquo;No parents&rsquo; versions
of annotation file ( parent GO terms are not explicitly listed in the
file ), as it will result in a faster startup time.</p>
<p><i>Get From Gemma</i> used to be another option to get the
annotation file. At the time of writing this document, the option
appears to be non-functional.</p>
<p>Upon clicking &lsquo;start,&rsquo; these files will be loaded and
displayed in a table view. The different columns include the name and
description of a group (e.g., GO ids and description), the number of
elements in the group, and a score denoting the multi-functionality
bias of the group. More columns will be added to this as you run an
analysis. The details of the table view and tree view are explained
below ( see <i>Results of ORA analysis</i> ).</p>
<p>Once these files are loaded, choose an analysis to run from the
&lsquo;Analysis&rsquo; tab.</p>
<h2 class="western"><a name="what-is-an-over-representation-analysis-ora-1"></a>
50.10 What is an Over-Representation Analysis (ORA)?</h2>
<p class="first-paragraph">Over-representation Analysis (ORA) is a
method to identify statistically significant functional groups (e.g.,
Gene Ontology terms) among high-scoring genes in a gene list.</p>
<p>This method requires a gene-score file (see <i>input files</i>
section) of all genes in the experiment. The user needs to set a
threshold of gene scores to &lsquo;select&rsquo; the genes in the
gene list to be examined for functional enrichment. The null
hypothesis is that genes in the functional groups are randomly
distributed between &lsquo;selected&rsquo; and &lsquo;non-selected&rsquo;
genes. It uses hyper-geometric distribution to identify the
functional groups that are significant among the genes passing the
threshold.</p>
<h2 class="western"><a name="how-do-i-run-an-ora"></a>50.11 How do I
run an ORA?</h2>
<p class="first-paragraph">To start an ORA analysis, choose ORA from
Analysis -&gt;Run Analysis tab.</p>
<p><b>ORA requires a complete list of gene scores, not just the
selected gene list</b>. Once a complete list of gene scores is
uploaded, the user needs to specify a threshold to determine the
genes to be selected for analysis. Raw gene expression data file can
also be provided, though this is optional. The following parameters
may be set by the user:</p>
<ul>
	<li/>
<p class="compact"><b>Maximum and minimum gene set sizes</b> -
	This refers to the number of genes in the Gene Ontology groups to be
	considered for analysis. It is recommended not to use very small or
	huge numbers here. Having a large number of functional groups can
	worsen the multiple-testing issues.</p>
	<li/>
<p class="compact"><b>ORA-specific settings</b> - The
	ORA-specific settings dialog box is shown below. Depending on the
	gene scores in the input file, the user needs to select the
	appropriate choice. For example, if the scores are p-values, then
	the &ldquo;take the negative log of the gene scores&rdquo; box
	should be checked, and &ldquo;larger scores in your score file are
	better&rdquo; should be unchecked.<br/>
<br/>
<br/>

	</p>
	<li/>
<p class="compact"><b>Gene score threshold</b> refers to the
	score value in the input file, and this determines how genes are
	selected. For example, if the input gene list has p-values, then the
	threshold might be 0.05 or 0.001, but not 4. The threshold refers to
	the values <i>prior</i> to log-transforming.</p>
</ul>
<p class="first-paragraph"><img src="0000_html_86cb308a52d6aa48.png" name="Image97" align="bottom" hspace="12" width="560" height="194" border="0"/>
</p>
<h2 class="western"><a name="what-are-the-results-of-an-ora-analysis"></a>
50.12 What are the results of an ORA analysis?</h2>
<p class="first-paragraph">The ORA analysis displays the functional
groups that are over-represented in the gene list and the associated
p-value. The results can be viewed either in a table format (see
below) or in a tree format.</p>
<p><img src="0000_html_f3b3c20f5a4a6bc6.png" name="Image98" align="bottom" hspace="12" width="560" height="329" border="0"/>
</p>
<p>The first four columns are standard to an ErmineJ table view. As
each analysis is run, the results from each run will be added to the
standard table as additional columns.</p>
<p>The p-values associated with each functional group, as obtained
from the ORA run, are given in the 5th column. Benjamini-Hochberg
correction is used to correct the p-values for multiple testing.
Different shades of green color denote the strength of enrichment. A
green diamond next to the p-value indicates a low effect of
multi-functionality correction (low sensitivity) and a red diamond
indicates that the group is sensitive to multi-functionality
correction (see below for details).</p>
<p>Additional information on each functional group is obtained by
double-clicking on the corresponding p-value. This displays the
details of all genes in the functional group. Expression data is
optional, however, if it was loaded, the expression values will be
presented as a heatmap on the left-hand side as shown below.</p>
<p><img src="0000_html_4ba7249bd2ef06d.png" name="Image99" align="bottom" hspace="12" width="560" height="239" border="0"/>
</p>
<p>An example expression data file can be obtained using the commands
below. This file contains mouse gene expression raw data for 6
samples (3 controls - C1,C2,C3 and 3 mutants - M1,M2,M3).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-O
http://data.biostarhandbook.com/rnaseq/mouse-gene-expression.txt</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
mouse-gene-expression.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1,3,4,5,6,7,8 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene_counts.txt</font></font></p>
<h2 class="western"><a name="what-is-a-multi-functionality-score"></a>
50.13 What is a multi-functionality score?</h2>
<p class="first-paragraph">Multi-functionality refers to the fact
that a gene can have more than one function. In ErmineJ, function
applies to the annotations for a gene. Some genes will have more
functions (and hence annotations) than others, but we must recognize
that this information is incomplete. Moreover, we can&rsquo;t even
assume that the missing annotations are evenly distributed. Some
genes are more &ldquo;popular&rdquo; and better-studied than others.
Consequently, there is more information available about &ldquo;popular&rdquo;
genes than &ldquo;less popular&rdquo; genes.</p>
<p>ErmineJ calculates a multi-functionality score for each gene. The
equation takes into account the size of the functional group. In
general, however, the multi-functionality score of a gene correlates
highly with the number of annotations it has.</p>
<p>The fourth column in the ErmineJ table format is the
multi-functionality score of a functional group. It is a normalized
rank that shows the degree to which a group has multi-functional
genes. The higher the score, the more generic the functional group
is. Details on how the multi-functionality score for a functional
group is calculated can be found <font color="#4f81bd"><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0017258">here</a></font>.</p>
<h2 class="western"><a name="is-multi-functionality-good-or-bad"></a>50.14
Is multi-functionality good or bad?</h2>
<p class="first-paragraph">From solely a data interpretation
perspective, multifunctional genes are more likely to be false
positives. We can think of this as a multiple testing problem. When a
gene is multi-functional, it has more opportunities to be shown as
enriched just by chance.</p>
<p>Depending on the analysis method used, ErmineJ employs different
approaches to correct for multi-functionality. In ORA, genes are
iteratively removed in the order of their multi-functionality to test
the effect this removal has on the enriched groups. The idea here is
that if any functional group is enriched merely due to the
multi-functionality of the genes, then such groups will fall out as
you remove those genes.</p>
<h2 class="western"><a name="what-is-gene-score-resampling-gsr"></a>50.15
What is Gene Score Resampling (GSR)?</h2>
<p class="first-paragraph">Gene Score Resampling (GSR) is another
method in ErmineJ (similar to ORA) to identify significant functional
groups in a gene list. GSR is different from ORA in that it does not
require a threshold. Instead, GSR computes an aggregate score (e.g.,
a geometric mean of gene scores) for a functional group and
calculates its significance by resampling the data.</p>
<p>The results of both ORA and GSR methods applied to the same input
datasets are shown below. GSR gave fewer effects compared to ORA in
this analysis.</p>
<p><img src="0000_html_6dbffff172f0a110.png" name="Image100" align="bottom" hspace="12" width="560" height="443" border="0"/>
</p>
<h2 class="western"><a name="what-is-correlation-analysis-in-erminej"></a>
50.16 What is Correlation Analysis in ErmineJ?</h2>
<p class="first-paragraph">The correlation analysis (CORR) differs
from other ErmineJ methods by assessing how well the genes in a
functional group are clustered together. This method does not use
gene scores. Instead, it uses the expression profiles of the genes
themselves. CORR checks if the correlation among the members of a
functional group is more than expected by chance. The raw score for a
functional group is the absolute mean of the pairwise correlation of
its members.</p>
<p>Unlike other methods in ErmineJ, the CORR analysis is
computationally intensive. Hence, one needs to be careful while
setting the parameters when running this analysis.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="student-reports-on-functional-enrichment"></a>
51 Student reports on functional enrichment</h1>
<p class="first-paragraph">The following entries are submissions by
students taking the <b>BMMB 852: Applied Bioinformatics</b> class at
Penn State. I selected a entries that go beyond a typical ontology
analysis and make additional interesting observations. The assignment
stated the following:</p>
<p class="block-text-western">Take a list of genes and perform a gene
list over-representation or a gene set enrichment study with a method
of your choice.</p>
<p class="block-text-western">In a few paragraphs discuss what you
have found. What functionality appears to characterize the list? You
may use the genes in the files from here:</p>
<p class="block-text-western"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/courses/2017-852/media/05/">http://data.biostarhandbook.com/courses/2017-852/media/05/</a></font></p>
<p class="block-text-western">Alternatively, you may use any other
list of genes that you have.</p>
<p class="first-paragraph">Within each section the content is taken
verbatim from the submission by the author named in the title. Only
minor corrections were made if these were necessary. My own
assessment and reply are written in italics at the end of each
section.</p>
<h2 class="western"><a name="david-northover-is-amused"></a>51.1
David Northover is amused</h2>
<p class="first-paragraph">After examining the gene list given in
Enrichr, there were a few common enriched threads, which suggests
that there were a couple of distinct groups of genes associated with
specific functions.</p>
<p>The clearest signal was in relation to adipogenesis, the
generation of fat stores (adipose cells and tissues). Pathway
analyses showed strong enrichment for adipogenesis, transcription
factors related to adipogenesis, as well as several insulin related
pathways. This is also consistent with the identified ontologies,
matching insulin and apetite response, as well as hormone activity
and glucose transport. Cell types matched to the Pituitary (hormone)
and Liver, above adipose tissue, so the genes appear to be more
focused on the regulatory side. Disease screening was (given the
rest) unsurprisingly associated with obesity, cholesterol, and
insulin disorders (such as Hyperglycemia and Hyperinsulism).</p>
<p>A possible secondary function may be related to neural activity,
as while there was no overexpression of nerve cells, there were KEGG
hits for neuroactive ligand-receptor interaction and GO axon cell
components. This may in some way be related to sending/recieving
neural signals related to adipogenesis in some way, however.</p>
<p>As an amusing side note about the dangers of over-interpreting
limited enrichment results, and of the issues with self-reported
data, it was also enriched for the UK Biobank term &ldquo;Never Had
Laser Treatment for Glaucoma/High Eye Pressure&rdquo;, which is
likely to be a very strange coincidence. There are other UK Biobank
terms that may be leading, though - &ldquo;Relative Age Voice Broke&rdquo;,
&ldquo;Comparative Body Size (Age 10)&rdquo; and &ldquo;Age When
Periods Started&rdquo;, though less enriched, match with a
&ldquo;Precocious Puberty&rdquo; listing elsewhere. I would be
uncomfortable making too much out of these - given the number of
comparisons scanned through - but it would be worth trying to find
which genes triggered those items in specific but not making a
conclusion solely out of this kind of data.</p>
<p><i>Great points are made here, we are never quite certain of what
information makes it into the GO or other resources and what the
vetting process is.</i></p>
<h2 class="western"><a name="kristin-passero-notes-that-the-presentation-matters"></a>
51.2 Kristin Passero notes that the presentation matters</h2>
<p class="first-paragraph">I took the following gene list from the
DESeq Output file</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/ontology/deseq-output.txt">http://data.biostarhandbook.com/ontology/deseq-output.txt</a></font></p>
</ul>
<p class="first-paragraph">available on the [Using the AGRIGO
server][#agrigo] page in the BioStar Handbook. I selected the top 19
differentially expressed genes.</p>
<p>Genes: WSB1, TPBG, XBP1, ASNS, SLC7A5, SFRP2, HERPUD1, SLC3A2,
HES6, SLC7A11, CARS, VEGFA, SHMT2, TOP2A, SAFB2, MTHFD2, YARS,
DNAJB9, HYOU1</p>
<p>I used both PANTHER and AgriGO to perform gene set enrichment for
biological process. One thing I noticed is that the GO terms produced
by PANTHER were far more specific than those produced by AgriGO for
the same list of genes. PANTHER&rsquo;s terms seemed more &ldquo;event&rdquo;
oriented (response to stress, to apoptosis, etc.) while AgriGO&rsquo;s
all includes the words &ldquo;metabolic process&rdquo; preceded by a
different compound name (amino acid, cellular amine, oxyacid, etc.)
There was only one shared term among them, &ldquo;carboxylic acid
metabolic process, <font color="#4f81bd"><a href="go:0019752">GO:0019752</a></font>.&rdquo;
However, were I to have used these tools without comparison, AgriGO
would have given the impression that my subset of genes was enriched
in basic metabolic process, whereas PANTHER would have suggested the
genes were involved primarily in cellular-stress-response pathways.</p>
<p>PANTHER lists significant enrichment in processes such as
PERK-mediated unfolded protein response, negative regulation of
ER-unfolded protein response, negative regulation of endoplasmic
reticulum stress-induced intrinsic apoptotic signaling pathway,
IRE1-mediated unfolded protein response. However, it does list more
&ldquo;generic&rdquo; processes such as amino acid transport across
the plasma membrane, tryptophan transport, and branched-chain amino
acid transport. As such, knowing nothing about the actual function of
these genes, I would predict they&rsquo;re involved in collecting
free amino acids for unfolded protein repair. The AgriGO processes
include cellular amino acid metabolic process, oxoacid metabolic
process, carboxylic acid metabolic process. Based on that information
I might presume that these proteins have something to do with amino
acid modifications.</p>
<p>The PANTHER results give a picture with greater context; however
that is based on my own presumptions, knowing nothing except these
terms I have been given. As such, PANTHER has a greater ability to
lead good-bad conclusions just because it gives enough information to
make the user dangerous. On the other hand, AgriGO offers a more
general overview of processes, so the conclusions drawn from them may
be more correct due to the fact the processes listed could be
&ldquo;shared&rdquo; by difference cellular processes. However, if
one looks at all the AgriGO results, further along the list you begin
to see terms involving cell death and stress &mdash; they are just
not FDR significant. At first I thought maybe they were using
different statistical methods, but I found they are both using
Fisher&rsquo;s exact test and some variant of the FDR correction. As
such, I am not sure of the origin of the discrepancy in
significance.*</p>
<p><i>What I like the most about this entry that it notes that the
presentation of a complex information can in itself shape our
thinking and rationalization later on. The choice of words, the order
of listing information, the visualization of the data may have
substantial effects on what we come to believe about our data.</i></p>
<h2 class="western"><a name="jeremy-held-gets-somewhat-different-answers-but-all-make-sense"></a>
51.3 Jeremy Held gets somewhat different answers but all make sense</h2>
<p class="first-paragraph">I am generating my own gene list from the
model plant Arabidopsis thaliana. To generate the list, I first took
out the top 2000 genes with annotations, which equated to 20
annotations at the low end. The average for Arabidopsis is only 7.
From my list of 2000, I shuffled the gene names and took 20:</p>
<p>OCP3 AT5G11270 DMR6 AT5G24530 DELTA-OAT AT5G46180 ARAC5 AT1G75840
CBF1 AT4G25490 NRPD1B AT2G40030 DDF1 AT1G12610 KT2/3 AT4G22200 HBT
AT2G20000 RBR1 AT3G12280 PHB2 AT1G03860 ATR1 AT4G24520 EIN3 AT3G20770
MKK3 AT5G40440 CLPP6 AT1G11750 ROXY1 AT3G02000 PBP1 AT3G16420 PEPR1
AT1G73080 GAPA-2 AT1G12900 VAB1 AT1G76030</p>
<p><b>Using Panther</b> Using the TAIR IDs, PANTHER was able to
uniquely map all 20 genes. The reference list contained 27,581 genes.
I only looked at enriched biological process terms with an FDR &lt;0.05.</p>
<p>The highest enrichment was in &lsquo;regulation of cell growth&rsquo;
(over 40-fold). Only 102 genes in the reference have this annotation,
which seems low. 5 genes in my list are annotated as &lsquo;defense
response to bacterium&rsquo;, and 4 as &lsquo;innate immune
response&rsquo;, indicating my list is enriched for plant defense. I
also have enrichments for more vague terms including &lsquo;response
to acid chemical&rsquo; and &lsquo;response to hormone&rsquo;.</p>
<p><b>Using AgriGO</b> AgriGO shows many of the same terms as
PANTHER, including defense genes and chemical/hormone responses.
Calculation of FDR is different between PANTHER and AgriGO. For
example, the lowest FDR in AgriGO is &lsquo;response to
oxygen-containing compound&rsquo; (1.1E-05). In PANTHER, this term is
not the lowest, with its FDR at 1.77E-04. Also interesting, AgriGO
does not pull out &lsquo;regulation of cell growth&rsquo;, even
though it had the largest enrichment in PANTHER. Instead, new terms
pop up, albeit at high FDRs, including &lsquo;regulation of
transcription, DNA-templated&rsquo; and &lsquo;cell communication&rsquo;.</p>
<p><b>Using G:Profiler</b> This service employs the strictest
formula, as it only found 16 significant biological process terms.
However, many of the same defense and chemical response terms found
in PANTHER and AgriGO are identified here. The right side of the
table shows that the function of these genes has been verified
experimentally for most of my genes (indicated by red squares). It
also shows that some genes have broad functions, while others only
have a few. This may be because certain genes have received more
study than others.</p>
<p><b>Conclusion</b> This exercise has clearly shown the variability
in a gene enrichment analysis depending on the program you use.
However, terms relating to plant defense and responses to chemicals
and hormones were identified in each program, suggesting those
designations may be accurate. Additionally, as these genes were
randomly selected, we would not expect strong associations for
particular functions, as perhaps we would see if our gene list came
from an experiment. As shown on G:Profiler, no GO term encompasses
every gene.</p>
<p><i>This is an example where albeit different, the answers seem to
indicate similar biological mechanisms.</i></p>
<h2 class="western"><a name="kelly-gomez-campo-works-on-her-thesis"></a>
51.4 Kelly Gomez Campo works on her thesis</h2>
<p class="first-paragraph">The study case is based on an experiment
from my PhD thesis. I was testing the hypothesis that phenotypic
plasticity is associated to methylome changes in a reef-building
coral &ldquo;Acropora&rdquo;. Because coral genomes are annotated
based on human genomes, I chose 288 genes from one of my outputs</p>
<ul>
	<li/>
<p class="compact">(genes not included here as this study is
	still in progress)</p>
</ul>
<p class="first-paragraph">and use Homo sapiens as the reference
list. A large list of genes in coral genomes are still unknown or not
annotated, or are annotated as uncharacterized proteins. An
statistical overrepresentation test was performed in PANTHER.The
functionality that appears to characterize the list is cell
organization. This analysis showed a significant enrichment as high
as 114 (5850 REF)in GO biological process cellular component
organization or biogenesis (ID <font color="#4f81bd"><a href="go:0071840">GO:0071840</a></font>).
According to the source, is a process that results in the
biosynthesis of constituent macromolecules, assembly, arrangement of
constituent parts, or disassembly of a cellular component. The most
significant enrichment functions are related to this biological
process, organelle organization, regulation of cellular component
organization, regulation of organelle organization.Response to ion
calcium ion (ID <font color="#4f81bd"><a href="go:0051592">GO:0051592</a></font>)
was a differences enriched biological function from the output,
relates to any process that results in a change in state or activity
of a cell or an organism (in terms of movement, secretion, enzyme
production, gene expression, etc.) as a result of a calcium ion
stimulus. According to the ancestor chart is a biological process
that relates to response to a inorganic chemical stimulus,
specifically, to metal ion. It&acirc;&euro;&euro;s an interesting
enriched function since Calcium ions (Ca2+) are key second messengers
in a variety of eukaryotic cell signaling pathways, and they function
in the regulation of diverse cellular processes. While it has become
clear that Ca2+ stimulates many cellular processes, such as muscle
contraction, cellular proliferation, gene expression, secretion of
hormones and neurotransmitters, exocytosis, and chemotaxis, it has
also been realized that Ca2+ is very toxic (Schnellmann and
Covington, 2010). Thus, the free intracellular Ca2+ concentration
([Ca2+]) must be highly regulated to achieve a proper balance between
Ca2+-mediated cell function and Ca2+-mediated cell
death.Double-strand break repair via homologous recombination (ID GO:
0000724) is another biological function that was significantly
enriched. The function is defined as the error-free repair of a
double-strand break in DNA in which the broken DNA molecule is
repaired using homologous sequences. A strand in the broken DNA
searches for a homologous region in an intact chromosome to serve as
the template for DNA synthesis. The restoration of two intact DNA
molecules results in the exchange, reciprocal or nonreciprocal, of
genetic material between the intact DNA molecule and the broken DNA
molecule.The enriched functions go accordingly to the treatment in my
experiment and the phenotypic response I have for the organisms. The
treatment was a switch from low light conditions to a high light
conditions. According to this analysis, the high light treatment
induced a cell reorganization, DNA damage and repair, with calcium
ions as a stimulator of cellular processes.</p>
<p><i>What I love about this submission is that it takes the
concepts, methods that we cover in class and applies them right away
to a realistic problem. This is one of the paradigms that I am
building the entire book around. I hope to cover subjects in a manner
that allows readers to apply what they learn right away to
&ldquo;realistic&rdquo; situations. To go beyond toy examples and
canned demos. I am always pleased to see it work in practice</i></p>
<h2 class="western"><a name="tomas-lopez-londono-finds-that-even-randomly-selected-genes-may-be-enriched"></a>
51.5 Tomas Lopez Londono finds that even randomly selected genes may
be enriched</h2>
<p class="first-paragraph">Script (gene_counts.txt file was created
from goa_human.gaf):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
gene_counts.txt | head -n 1000 | cut -f 3 -d ' ' | shuf | head -n 20</font></font></p>
<p class="first-paragraph">Resulting gene list:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PKD2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KAT2B</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HGF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YES1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PSMA7</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">STXBP1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PSMB8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HSPB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NCBP2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SMURF2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MAD1L1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PPARGC1A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DDX58</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ZC3H12A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PPP3CA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PSMD8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATA4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PARP1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FADD</font></font></p>
<p class="first-paragraph">ORA method: Panther14.1
(<font color="#4f81bd"><a href="http://pantherdb.org/tools/compareToRefList.jsp">http://pantherdb.org/tools/compareToRefList.jsp</a></font>)
Analysis performed in terms of biological processes, with false
discovery rate of observations (FDR) set by default (FDR&lt;0.05)</p>
<p><b>Discussion:</b> What functionality appears to characterize the
list?</p>
<p>When organizing the results from highest (&lt;0.05) to lowest
false discovery rate (FDR) values, it is observed that the function
with the highest chance of FDR is &ldquo;blood vessel development&rdquo;,
whit 4 genes of the uploaded list (20 genes) out of the 487 genes in
the reference list that map to this particular function. The Fold
Enrichment for this function is 8.21, which indicates that it is
overrepresented. The definition of this function, according to AmiGO,
is &ldquo;the process whose specific outcome is the progression of a
blood vessel over time, from its formation to the mature structure&rdquo;.
The total number of annotations for this function is 7073, with a
significant portion of uncharacterized proteins for each gene/product
annotated.</p>
<p>The function with the lowest FDR is &ldquo;response to organic
substance&rdquo;, which is defined as &ldquo;any proves that results
in a change in state or activity of a cell or an organism as a result
of an organic substance stimulus&rdquo;. All genes of the uploaded
list (20) out of the 2992 in the reference list mapped this
particular function. Total number of annotations for this function is
82892, also with a significant portion of uncharacterized proteins
for each gene/product annotated. The fold Enrichment value is similar
to the previous function (6.68), indicating that this function is
also overrepresented. In fact, all functions selected with FDR&lt;0.05
were overrepresented. This may indicate that the fraction of observed
genes in the uploaded list associated to particular pathways always
exceed the number of expected genes associated to those pathways when
considering FDR values below 0.05.</p>
<p>One of the functions with highest Fold Enrichment is the
&ldquo;regulation of ATP biosynthetic process&rdquo;, with a value
&gt;100. This value indicates that this is probably one of the most
over-represented functions. 2 genes of the uploaded list out of the
19 genes in the reference list that map to this particular function.
The definition of this function is &ldquo;any process that modulates
the frequency, rate or extent of ATP biosynthetic process&rdquo;. It
has only 162 annotations, compared to the previously discussed
functions with 7073 and 82892 annotations. Interestingly, most
proteins encoded by the genes/products associated with this function
have been characterized, opposite to the other two functions with
significant portions of uncharacterized proteins. This may indicate
that this pathway is much better described and understood. The
extremely complex chart with the hierarchical classification of gene
ontology relationships can be interpreted as proof of that.</p>
<p><i>This example goes to demonstrate that even &ldquo;randomly&rdquo;
selected genes may exhibit features that distinguish them from a
background. Was Mr.&nbsp;Londono &ldquo;lucky&rdquo; to selecte genes
with common functions or was the statistical method unable to cope
with the data representation - an undecidable question in my opinion.
In general, statistical methods don&rsquo;t work all that well for
functional enrichment because the information that is being put into
context has been collected in deliberate and systematic way, aiming
for certain goals - hence violate assumptions that the statistical
methods rely upon</i></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="scientific-reproducibility"></a>52
Scientific reproducibility</h1>
<p class="first-paragraph">The reproducibility of scientific analyses
is (or rather should be) a cornerstone of science. The scientific
method itself depends on scientists&rsquo; ability to verify
statements made by others.</p>
<p>In reality, the reproducibility process is fraught with many
pitfalls, and solutions are few and far between. The problems we face
are in part due to rapid and radical transformations in different
scientific domains, especially Life Sciences.</p>
<p>At the same time the very meaning of the word &ldquo;reproducibility&rdquo;
has been distorted. By using the terms incorrectly, and focusing on
the &ldquo;outrageous&rdquo; yet rare and non-representative events,
well intentioned scientists, the very champions of &ldquo;reproducibility&rdquo;
cause more harm than good.</p>
<h2 class="western"><a name="what-does-the-word-reproducible-research-mean"></a>
52.1 What does the word &ldquo;reproducible research&rdquo; mean?</h2>
<p class="first-paragraph">In what might be the most ironic of
situations - scientists have a hard time agreeing what the words
&ldquo;reproducible research&rdquo; mean.</p>
<p>The very definition of &ldquo;<i>reproducible research</i>&rdquo;
itself seems not to be reproducible! There is a surprising amount of
uncertainty and confusion even what the word ought to mean.</p>
<h2 class="western"><a name="what-is-the-red-herring-of-reproducibility"></a>
52.2 What is the &ldquo;red herring&rdquo; of reproducibility?</h2>
<p class="first-paragraph">We invite you to perform a Google search
on the words scientific reproducibility. Here is what our first hit
is:</p>
<p><img src="0000_html_3eb1a8fe4d86ca0e.png" name="Image101" align="bottom" hspace="12" width="560" height="256" border="0"/>
</p>
<p>There is a systematic mischaracterization of the very concept of
reproducibility to imply that there is a &ldquo;crisis&rdquo; in
science and that reproducibility is about &ldquo;fixing&rdquo; broken
science. Examples abound, here are the leading scientific
publications pushing the same line of thought:</p>
<ul>
	<li/>
<p class="compact">Editorial: <font color="#4f81bd"><a href="http://science.sciencemag.org/content/343/6168/229">Reproducibility</a></font>
	in the Science</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.nature.com/collections/prbfkwmwvz">Challenges
	in irreproducible research</a></font> in Nature</p>
</ul>
<p class="first-paragraph">In the above framing, reproducibility is a
sort of &ldquo;defense agains dark arts&rdquo; a way to catch
you-know-who, plus their posse of other cheaters, liars and evil
scientists. Alas it is the painfully wrong way to frame the problem
of reproducibility that only invites fruitless solutions.</p>
<h2 class="western"><a name="is-science-really-facing-a-reproducibility-crisis"></a>
52.3 Is science really facing a reproducibility crisis?</h2>
<p class="first-paragraph">With all the alarms going off, here is a
different take:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.pnas.org/content/115/11/2628">Opinion:
	Is science really facing a reproducibility crisis, and do we need it
	to?</a></font> PNAS (2018)</p>
</ul>
<p class="first-paragraph">Here are the closing thoughts that we
fully agree with:</p>
<p class="block-text-western">Therefore, contemporary science could
be more accurately portrayed as facing &ldquo;new opportunities and
challenges&rdquo; or even a &ldquo;revolution&rdquo;. Efforts to
promote transparency and reproducibility would find complete
justification in such a narrative of transformation and empowerment,
a narrative that is not only more compelling and inspiring than that
of a crisis, but also better supported by evidence.</p>
<p class="first-paragraph">There is no crisis in the sense of
&ldquo;cheating&rdquo; and invalid science being published at
increasing rates.</p>
<p>There is an alarming trend in publishing research results that are
not properly explained. Framing reproducibility to address the
transparency of the processes would be a far more fruitful direction
to pursue.</p>
<h2 class="western"><a name="how-do-scientists-attempt-to-define-what-reproducibility-is"></a>
52.4 How do scientists attempt to define what reproducibility is?</h2>
<p class="first-paragraph">Here we present the ideas that scientists
came up with, in the order of least demands to most demands that they
put on authors:</p>
<ol>
	<li/>
<p class="compact">A brief description of the decisions that
	made during the analysis.</p>
	<li/>
<p class="compact">A detailed description of the commands that
	were used to generate the report.</p>
	<li/>
<p class="compact">A list of all instructions and a snapshot
	of the intermediate data that the commands produced.</p>
	<li/>
<p class="compact">Publishing the analysis as a single
	document that one can easily re-run.</p>
	<li/>
<p class="compact">All the above with the added information of
	the exact computer setup, operating system configuration, and
	software version that was used to generate the analysis.</p>
	<li/>
<p class="compact">All the above packaged into a single
	&ldquo;virtual&rdquo; machine that emulates the same setup the
	authors had at their disposal that can be re-run.</p>
</ol>
<h2 class="western"><a name="so-what-does-reproducibility-mean"></a>52.5
So what does reproducibility mean?</h2>
<p class="first-paragraph">First you should not conflate the
different interpretations of the word &ldquo;reproducibility&rdquo;.
In this book we focus solely on the reproducibility of an analytical
process where:</p>
<p>1. Reproducibility is not a binary classification: <i>reproducible</i>
vs <i>irreproducible</i>. Reproducibility is a <i>scale</i>.</p>
<p>Reproducibility goes from <i>easy to follow</i> to <i>practically
impossible to redo</i>; the scale should be evaluated in the context
of using the information that the scientists provided.</p>
<p>2. Reproducibility is not a measure of scientific validity. It is
a measure of <i>transparency</i>. Do we know what took place?</p>
<p>In our opinion a result is reproducible when we have been provided
with all information required to make a the same scientific
discovery. The benefit is that we can now build on this previous
discovery and take it further. Reproducibility, as used in this book,
quantifies the effort it takes to follow along the discovery process.
See also: <font color="#4f81bd"><a href="#ch055.xhtml#bestbio">Will
the best bioinformaticians on the planet produce reproducible
analyses?</a></font></p>
<h2 class="western"><a name="how-difficult-is-to-re-analyze-data"></a>
52.6 How difficult is to re-analyze data?</h2>
<p class="first-paragraph">As you will see it yourself, it is usually
far more straightforward to re-analyze data with your methods than to
reproduce and retrace the same steps as published in a scientific
paper.</p>
<p>In the majority of cases, you&rsquo;d want to know how the data
has been analyzed not to rerun the same commands, but to assess the
strengths and weaknesses of the analysis and to identify any possible
errors or inconsistencies of that process.</p>
<h2 class="western"><a name="what-is-the-best-strategy-to-reproduce-results"></a>
52.7 What is the best strategy to reproduce results?</h2>
<p class="first-paragraph">As a bioinformatician, you may often need
to reproduce published results. We must warn everyone that doing so
by following the descriptions and explanations found within research
papers is usually a tedious and frustrating process.</p>
<p>A typical scientific paper spends far more time trying to convince
readers that its findings are significant rather than explaining the
steps of how that analysis took place. The focus on &ldquo;selling
the result&rdquo; is a normal consequence of how scientific
publications are judged. Novelty is the most crucial requirement.</p>
<p>But there is another and more devious complicating factor. The
so-called <i>&ldquo;scientific narrative&rdquo;</i> is simplified and
misleading - it depicts the bioinformatics analysis as a linear chain
of decisions: <i>&ldquo;first we did this, then we did that.&rdquo;</i>
The descrptions paint a picture that does not correspond to reality.
Recall the plot from <font color="#4f81bd"><a href="#ch023.xhtml#whatisdata">What
is data</a></font>?:</p>
<p><img src="0000_html_c731492640345707.png" name="Image102" align="bottom" hspace="12" width="560" height="99" border="0"/>
</p>
<p>Alas the image above is not what the authors actually did. Most of
the time the authors explored several, perhaps dozens of
alternatives, ran into many dead ends, backtracked and tried their
analyses in several different ways. After quite a bit of work,
they&rsquo;ve finally honed in on what they believe to be the correct
approach for making the discovery. But within the scientific paper,
the long and branching chains of decisions and actions are described
as straightforward choices, and the failed attempts are never
mentioned. This is what most likely they have done:</p>
<p><img src="0000_html_633aa58e5573e42f.png" name="Image103" align="bottom" hspace="12" width="560" height="309" border="0"/>
</p>
<p>Your aim should always be to re-analyze the data in a more
straightforward way. &ldquo;Re-doing it faster and better&rdquo; is
the approach that we champion in this book and it is the one that has
served us well. With time methods improve and the algorithms change,
then all scientific observations should be discoverable in different
ways. Sticking to one protocol because that&rsquo;s what one always
used before and that is what &ldquo;my group is using&rdquo; is not
always a good choice. It may still be the optimal choice, but you
should ensure that you understand the alternatives. Once you know
what type of insights are in your data, there is almost always a
simpler and more efficient method to find them.</p>
<h2 class="western"><a name="are-the-challenges-of-data-reproducibility-recognized"></a>
52.8 Are the challenges of data reproducibility recognized?</h2>
<p class="first-paragraph">Most organizations funding science
recognize that the low level of reproduciblity of results are a
growing problem in bioinformatics and other areas of science. These
organizations try to implement what they believe is a &ldquo;carrot
and stick&rdquo; approach, policies of offering a combination of
rewards and punishment to induce behavior that should result in more
reproducibility.</p>
<p>In our opinion, these policies, at least those that we are aware
of, rely mostly on &ldquo;stick&rdquo; and have less of the &ldquo;carrot&rdquo;
component.</p>
<p>In summary, you should be mindful that the problem is widely
recognized, but that it is also widely misunderstood.The current
mitigation efforts appear to be inadequate.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="redo-genomic-surveillance-elucidates-ebola-virus-origin"></a>
53 Redo: Genomic surveillance elucidates Ebola virus origin</h1>
<p class="first-paragraph">We were interested in accessing the data
for the study titled</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.science.org/doi/full/10.1126/science.1259657">Genomic
	surveillance elucidates Ebola virus origin and transmission during
	the 2014 outbreak</a></font> published in 2014 in the Science
	research journal.</p>
</ul>
<p class="first-paragraph"><img src="0000_html_e6e4ae8bfac0e96.png" name="Image104" align="bottom" hspace="12" width="560" height="458" border="0"/>
</p>
<p>Now you could read this paper all the way through and not be able
to find the information that points to where the data is located. If
you don&rsquo;t believe us, see it for yourself. See if you could
locate the link to the sequencing data. There is even a section
called Figures &amp; Data, common sense indicates that the data
access were listed there!</p>
<p>We were only able to find the link to the data because we knew
that information deposited in the Short Read Archive is typically
designated with a code that starts with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJN</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJE</font></font>.
A text search for these patterns leads to a small note at the end of
the paper, way out of sight, past the references, in the
acknowledgments, in the section that we could call a scientific
&ldquo;<i>no-mans land</i>.&rdquo; It is there where we finally find
what many might consider one of the most important deliverables of
this project:</p>
<p class="block-text-western">Sequence data are available at NCBI
(NCBI BioGroup:PRJNA257197).</p>
<p class="first-paragraph">There is no cross linking, no other
information there. You are just supposed to know what this means. Now
make a note of this number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>.
It will be our primary means to get all the data for this project,
helps us avoid clicking around on websites and gives us a systematic,
automated and repeatable way to get the data.</p>
<h2 class="western"><a name="how-to-access-the-data-connected-to-this-publication"></a>
53.1 How to access the data connected to this publication?</h2>
<p class="first-paragraph">Now that we have the id number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>
we can search for it at the <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/">NCBI
website</a></font>, to find a Bioproject titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA257197/">Zaire
	ebolavirus sample sequencing from the 2014 outbreak in Sierra Leone,
	West Africa.</a></font></p>
</ul>
<p class="first-paragraph">At the time of writing, the project
summary included <font face="Consolas, serif"><font size="2" style="font-size: 11pt">891</font></font>
experiments that produced <font face="Consolas, serif"><font size="2" style="font-size: 11pt">249</font></font>
nucleotide and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2240</font></font>
protein sequences.</p>
<p><img src="0000_html_9a93b4b4f0bf4dc.png" name="Image105" align="bottom" hspace="12" width="560" height="275" border="0"/>
</p>
<p>Once you are on the Bioproject site, it is no less shocking just
how little information is disseminated on what exactly has been
deposited or how these results were obtained.</p>
<ul>
	<li/>
<p class="compact">What are these <font face="Consolas, serif"><font size="2" style="font-size: 11pt">249</font></font>
	nucleotide and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2240</font></font>
	protein sequences?</p>
	<li/>
<p class="compact">How reliable is the information?</p>
	<li/>
<p class="compact">How are they different from what was known
	before?</p>
	<li/>
<p class="compact">What methods and practices were used to
	derive them?</p>
</ul>
<p class="first-paragraph">The rather lengthy <font color="#4f81bd"><a href="https://www.science.org/doi/full/10.1126/science.1259657">Supplementary
Information</a></font> does not help much, either. The 29-page
document has an extensive list of tools and techniques, but it
severely lacks in specificity. It contains generic and superficial
statements such as:</p>
<p class="block-text-western">EBOV reads were extracted from the
demultiplexed Fastq files using Lastal against a custom-made database
containing all full-length EBOV genomes</p>
<p class="first-paragraph">This description is inadequate for
reproducing the results: it only poses more questions.</p>
<ul>
	<li/>
<p class="compact">What is this custom made database?</p>
	<li/>
<p class="compact">Why isn&rsquo;t this custom database
	deposited here as well?</p>
	<li/>
<p class="compact">Which full-length EBOV genomes were
	included?</p>
	<li/>
<p class="compact">What does it mean to be extracted with
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Lastal</font></font>?</p>
</ul>
<p class="first-paragraph">Note another missing element here: who do
we contact if we have a question on the specifics of the analysis?
While papers do have corresponding authors, those are almost never
the ones that know the intricate details of the analysis.</p>
<h2 class="western"><a name="how-do-we-download-additional-results-for-this-paper"></a>
53.2 How do we download additional results for this paper?</h2>
<p class="first-paragraph">We will soon learn of automated ways to
get the sequence data for this project but what about all the other
collected evidence? For example: the observed variants? Can we get
that data in a standardized format? Here again, the standards vary
widely across various publishers/authors. For this paper the approach
is borderline ridiculous. You need to visit the publisher&rsquo;s
website, then manually click and download files with obscure names
such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1259657_file_s3.zip</font></font>
without knowing precisely what these files contain.</p>
<p>Some of the files are in PDF format. One such PDF contains nothing
more than a table with four rows and five columns! Since it is
distributed as a PDF we cannot automatically process the information
contained in this file without manually transcribing it into a
machine-readable form! Other files are in Excel but follow an
arbitrary ad-hoc format unsuited for automated processing. Again
those will need to be reformatted differently to comparable to
information standard formats.</p>
<p>We think we made our point - data distribution is an ad-hoc and
superficially supervised process, a bottleneck to innovation that
demands future readers do perform extra work that should not be
necessary.</p>
<h2 class="western"><a name="what-have-we-collected"></a>53.3 What
have we collected?</h2>
<p class="first-paragraph">Once we read the paper, clicked on various
links there, scoured NCBI, clicked around for a while there as well
we identified and collected the following accession numbers:</p>
<ul>
	<li/>
<p class="compact">SRA BioProject: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA257197">PRJNA257197</a></font></p>
	<li/>
<p class="compact">All variants: <font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/ebola/iSNV-all.vcf">iSNV-all.vcf</a></font></p>
	<li/>
<p class="compact">2014 strain specific variants:
	<font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/ebola/table-S4-2014_specific_snps.csv">table-S4-2014_specific_snps.csv</a></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>
	Ebola virus - Mayinga, Zaire, 1976, complete genome.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118</font></font>
	Zaire ebolavirus isolate Ebola
	virus/H.sapiens-wt/SLE/2014/Makona-NM042.3, complete genome</p>
</ul>
<p class="first-paragraph">We can now use entrez direct as explained
in <font color="#4f81bd"><a href="#ch028.xhtml#automate-ncbi">Automating
access to NCBI</a></font> to get data for the project:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the sequencing run information</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197 | efetch -format runinfo &gt;
runinfo.csv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the assembled genome information</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nucleotide -query PRJNA257197 | efetch -format fasta &gt;
genomes.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the assembled protein information</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db protein -query PRJNA257197 | efetch -format fasta &gt;
proteins.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the strain from 1972</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id AF086833 -format fasta &gt; strain_1972.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the strain from 2014</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id KM233118 -format fasta &gt; strain_2014.fa</font></font></p>
<h2 class="western"><a name="is-there-a-way-to-automate-finding-the-project-number"></a>
53.4 Is there a way to automate finding the project number?</h2>
<p class="first-paragraph">In principle we could also search PubMed
with the title like so</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db pubmed -query &quot;Genomic surveillance elucidates Ebola virus
origin and transmission during the 2014 outbreak[title]&quot; |
efetch</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.
Science. 2014 Sep 12;345(6202):1369-72. doi: 10.1126/science.1259657.
Epub 2014 Aug 28.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DOI:
10.1126/science.1259657</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PMCID:
PMC4431643</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PMID:
25214632  [Indexed for MEDLINE]</font></font></p>
<p class="first-paragraph">Here is the paper in PubMed:</p>
<ul>
	<li/>
<p class="compact">PubMed id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">25214632</font></font>:
	<font color="#4f81bd"><a href="https://pubmed.ncbi.nlm.nih.gov/25214632/">https://pubmed.ncbi.nlm.nih.gov/25214632/</a></font></p>
</ul>
<p class="first-paragraph">then we can use the PubMed id to</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db pubmed -query PMC4431643 | elink -target sra</font></font></p>
<p class="first-paragraph">to obtain the search results that we can
then pass into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>.
The linking however is not always all that reliable.</p>
<h2 class="western"><a name="how-do-we-get-the-sequencing-data"></a>53.5
How do we get the sequencing data</h2>
<p class="first-paragraph">Previously we obtained the sequencing run
information with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197 | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">from which we can obtain the SRR run ids:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | csvcut -c Run | head</font></font></p>
<p class="first-paragraph">that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972920</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972921</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972922</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972923</font></font></p>
<p class="first-paragraph">To obtain the sequencing data for one
particular sequencing run use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
command (take 1000 reads):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 1000 --split-files SRR1972917</font></font></p>
<p class="first-paragraph">then run statistics on the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats</font></font></p>
<p class="first-paragraph">produces the output of:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
               format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917_1.fastq
 FASTQ   DNA      1,000  101,000      101      101     
101</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917_2.fastq
 FASTQ   DNA      1,000  101,000      101      101      101</font></font></p>
<p class="first-paragraph">To automate the process in a loop consult
the <font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/scripting/index.html">The
Art of Bioinformatics Scripting</a></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="redo-zika-virus-targets-human-cortical-neural-precursors"></a>
54 Redo: Zika virus targets human cortical neural precursors</h1>
<h2 class="western"><a name="quick-links"></a>54.1 Quick links:</h2>
<p class="first-paragraph">Links for easy access:</p>
<ul>
	<li/>
<p class="compact">SRA BioProject: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA313294">PRJNA313294</a></font>.</p>
	<li/>
<p class="compact">The GEO series: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE78711">GSE78711</a></font>.</p>
	<li/>
<p class="compact">Up-regulated genes: <font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/zika/zika-up-regulated.csv">zika-up-regulated.csv</a></font></p>
	<li/>
<p class="compact">Down-regulated genes:
	<font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/zika/zika-down-regulated.csv">zika-down-regulated.csv</a></font></p>
</ul>
<p class="first-paragraph">Getting the runinfo for the project:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Crosslink to sra. Need to &quot;sleep&quot; to slow things down so a
bit so NCBI does not ban us.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db gds -query GSE78711 | (sleep 1 &amp;&amp; elink -target sra) |
efetch -format runinfo &gt; runinfo.csv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Other sample information</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db gds -query GSE78711 | efetch &gt; sampleinfo.txt</font></font></p>
<p class="first-paragraph">How did we found the links listed above?
Read on.</p>
<h2 class="western"><a name="how-to-get-the-information-for-the-zika-paper"></a>
54.2 How to get the information for the Zika paper?</h2>
<p class="first-paragraph">We were interested in accessing the data
and the results for the study titled</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/26952870">Zika
	Virus Targets Human Cortical Neural Precursors and Attenuates Their
	Growth</a></font></p>
</ul>
<p class="first-paragraph">published in the journal Cell Stem Cell in
2016.</p>
<p><img src="0000_html_79e82c9da9290c63.png" name="Image106" align="bottom" hspace="12" width="560" height="279" border="0"/>
</p>
<h2 class="western"><a name="how-do-we-find-the-data-for-this-paper"></a>
54.3 How do we find the data for this paper?</h2>
<p class="first-paragraph">At the end of the paper we find the
following statement in a more visible form (though we only found it
by searching for the word GEO)</p>
<p class="block-text-western">The accession number for RNA-seq data
reported in this paper is GEO: <b>GSE78711</b></p>
<p class="first-paragraph">The GEO stands for the <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/geo/">Gene
Expression Omnibus</a></font> and is a public functional genomics
data repository that accepts array- and sequence-based data. Somewhat
confusingly only the gene expression level results are actually
stored in GEO, the sequence data is deposited into the Short Read
Archive SRA. The projects are then linked but it also means that
there are two locations for the data.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Produce run information.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db gds -query GSE78711 | efetch</font></font></p>
<p class="first-paragraph">The default output is somewhat of a hodge
podge information, if you need higher granularity use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">docsum</font></font>
format, then <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>.</p>
<p>Connect the data to the sequencing data deposited to SRA:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Crosslink to sra.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db gds -query GSE78711 | elink -target sra | efetch -format runinfo
&gt; runinfo.csv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the SRR run IDS for the data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | cut -f 1,22 -d , | head</font></font></p>
<p class="first-paragraph">produces</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run,BioProject</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3191542,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3191543,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3191544,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3191545,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3194428,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3194429,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3194430,PRJNA313294</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3194431,PRJNA313294</font></font></p>
<p class="first-paragraph">So we found the two locations for the
data:</p>
<ul>
	<li/>
<p class="compact">SRA BioProject <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA313294">PRJNA313294</a></font>.</p>
	<li/>
<p class="compact">The GEO series <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE78711">GSE78711</a></font>.</p>
</ul>
<p class="first-paragraph">Clearly we now have two numbers that we
need to keep track of: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA313294</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GSE78711</font></font>.
In addition, the scientific journal that published the results also
distributes some information from another location. This third
location, has no programmatic access to allow searches. You have to
click around the site and find the URLs with a manual process. If
you&rsquo;ve made it this far and you have the prior knowledge, you
can then save those links as command line actions like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-OkL
http://www.cell.com/cms/attachment/2055417636/2061019276/mmc2.xlsx</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-OkL
http://www.cell.com/cms/attachment/2055417636/2061019275/mmc1.pdf</font></font></p>
<p class="first-paragraph">The first file above is an Excel sheet.
You would need to download, open and export each of the pages into a
separate file. We did this for you</p>
<ul>
	<li/>
<p class="compact">List of regulated genes:
	<font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/zika/zika-up-regulated.csv">http://data.biostarhandbook.com/redo/zika/zika-up-regulated.csv</a></font></p>
	<li/>
<p class="compact">List of down regulated genes:
	<font color="#4f81bd"><a href="http://data.biostarhandbook.com/redo/zika/zika-down-regulated.csv">http://data.biostarhandbook.com/redo/zika/zika-down-regulated.csv</a></font></p>
</ul>
<p class="first-paragraph">You can now operate on the results of this
paper from command line like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Get the gene names for the study.</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/redo/zika/zika-up-regulated.csv</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
zika-up-regulated.csv </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-f 1 -d , </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font></p>
<p class="first-paragraph">produces the list of up regulated genes</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ELAVL3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FAM53C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SLC22A17</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">WDFY1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TBL1XR1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SON</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">EIF4G2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOGA3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SDCBP</font></font></p>
<p class="first-paragraph">Even identifying the location of the data
and results is quite cumbersome, and obviously not quite ready for
full automation. We think we have made our point again - currently
data distribution is an ad-hoc and superficially supervised process
that places a undue burden on the process of reproducibility.</p>
<h2 class="western"><a name="what-chapters-cover-the-zika-data"></a>54.4
What chapters cover the Zika data?</h2>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch054.xhtml#rnaseq-zika-data">Understand
	the Zika data</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch054.xhtml#rnaseq-zika-align">Alignment
	based RNA-Seq of Zika data</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch054.xhtml#rnaseq-zika-classify">Classification
	based RNA-Seq of Zika data</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="redo-a-synthetic-diploid-benchmark-for-accurate-variant-calling-evaluation"></a>
55 Redo: A synthetic-diploid benchmark for accurate variant-calling
evaluation</h1>
<h2 class="western"><a name="quick-links-1"></a>55.1 Quick links:</h2>
<ul>
	<li/>
<p class="compact">EBI BioProject: <font color="#4f81bd"><a href="https://www.ebi.ac.uk/ena/data/view/PRJEB13208">PRJEB13208</a></font></p>
	<li/>
<p class="compact">SRA BioProject: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJEB13208">PRJEB13208</a></font></p>
	<li/>
<p class="compact">Publication link: <font color="#4f81bd"><a href="https://www.nature.com/articles/s41592-018-0054-7">A
	synthetic-diploid benchmark for accurate variant-calling evaluation</a></font>
	Nature Methods (2018)</p>
</ul>
<p class="first-paragraph">Getting the runinfo for the project:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJEB13208 | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<h2 class="western"><a name="why-are-we-reproducing-this-paper"></a>55.2
Why are we reproducing this paper?</h2>
<p class="first-paragraph">We believe that some of the authors of
this paper, for example Dr Heng and Dr.&nbsp;McArthur are among the
most experienced practitioners of this field. We consider them to be
among the most skilled, most talented, and best trained computational
scientists that work in life sciences.</p>
<h2 class="western"><a name="will-the-best-bioinformaticians-on-the-planet-produce-reproducible-analyses"></a>
55.3 Will the best bioinformaticians on the planet produce
reproducible analyses?</h2>
<p class="first-paragraph">Above we take the word &ldquo;reproducible&rdquo;
in what we believe is the correct context. We believe that
reproducibility means the effort needed to redo the same analysis and
obtain the same results. Whether or not the results are
scientifically valid is completely different question that we do not
investigate here.</p>
<p>Now, having been told that these authors are the best of the best,
of the few and proud, what do you think? Can we, at least, in the
first step, obtain the same results as the authors did? Have the
authors provided sufficiently detailed descriptions to reproduce the
results?</p>
<p>Your Majesties! This humble ant wishes to follow your gigantic
footsteps. Will you permit that?</p>
<h2 class="western"><a name="what-problem-does-the-paper-attempt-to-solve"></a>
55.4 What problem does the paper attempt to solve?</h2>
<p class="first-paragraph">The accuracy of variant calling is of
utmost importance. It is typically quite challenging to determine
which method works best as we don&rsquo;t know what the correct
answer is. This paper states in the abstract:</p>
<p class="block-text-western">We derived a new benchmark dataset from
the de novo PacBio assemblies of two fully homozygous human cell
lines, which provides a relatively more accurate and less biased
estimate of small-variant-calling error rates in a realistic context</p>
<p class="first-paragraph"><img src="0000_html_6d2fe3f5a55784db.png" name="Image107" align="bottom" hspace="12" width="560" height="525" border="0"/>
</p>
<h2 class="western"><a name="where-is-the-data"></a>55.5 Where is the
data?</h2>
<p class="first-paragraph">The paper clearly states that:</p>
<p class="block-text-western">Illumina reads from this study were
deposited in the European Nucleotide Archive under accession
<font color="#4f81bd"><a href="https://www.ebi.ac.uk/ena/data/view/PRJEB13208">PRJEB13208</a></font>.</p>
<p class="first-paragraph">Listing the data explicitly is a great
step in the right direction. Not only tells us the number but
provides a link. The authors chose to upload the data into EBI;
thankfully the datasets are mirrored at NCBI sand we are able to
access the run information with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">entrez
direct</font></font> using:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJEB13208 | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<h2 class="western"><a name="where-is-the-code"></a>55.6 Where is the
code?</h2>
<p class="first-paragraph">Unlike most other publications, the
authors chose to create not just one but two GitHub repositories, and
they state that:</p>
<p class="block-text-western">Evaluation scripts are available from</p>
<ul>
	<li/>
<p class="block-text-western"><font color="#4f81bd"><a href="https://github.com/lh3/CHM-eval">https://github.com/lh3/CHM-eval</a></font></p>
</ul>
<p class="block-text-western">The variant-calling pipeline and hard
filters are implemented in</p>
<ul>
	<li/>
<p class="block-text-western"><font color="#4f81bd"><a href="https://github.com/lh3/unicall">https://github.com/lh3/unicall</a></font></p>
</ul>
<p class="first-paragraph">We have the data and code repositories.
Excellent!</p>
<h2 class="western"><a name="is-the-analysis-reproducible"></a>55.7
Is the analysis reproducible?</h2>
<p class="first-paragraph">While we commend the authors for
publishing Github repositories that act alone is insufficient to
reproduce the analysis. If you were to visit the repositories above
you will be overwhelmed with overcomplicated commands. Among the many
things we&rsquo;d like to call out is the following abomination that
demonstrates a total disconnect regarding common sense:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-qO-
ftp://ftp.sra.ebi.ac.uk/vol1/ERA596/ERA596361/bam/CHM1_CHM13_2.bam
\</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
freebayes -f hs37.fa - &gt; CHM1_CHM13_2.raw.vcf</font></font></p>
<p class="first-paragraph">There are so many things wrong with the
code above that we have a hard figuring out where to start our
critique. First, above the authors suggest that you download the BAM
file each time you might want to run this evaluation. Which is an
absurd idea, you would pay the cost of download over and over again.
So let&rsquo;s fix that first, separate the download from the variant
calling. Let&rsquo;s get the data first:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
ftp://ftp.sra.ebi.ac.uk/vol1/ERA596/ERA596361/bam/CHM1_CHM13_2.bam</font></font></p>
<p class="first-paragraph">Can you guess what happens? You&rsquo;ll
get an error:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">No
such directory &lsquo;vol1/ERA596/ERA596361/bam&rsquo;.</font></font></p>
<p class="first-paragraph">The link in the repository is incorrect.
The first command of the published code on how to reproduce the
analysis does not work! Gee thanks! After some investigation you&rsquo;d
find that the correct link is different:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam">ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</a></font></p>
</ul>
<p class="first-paragraph">Thus the correct download command is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</font></font></p>
<p class="first-paragraph">The code above shows me an estimated
download of 3 hours. Note how the original command would have you pay
the 3 hours each time you ran the command. Now, suppose that you have
downloaded the file, run the second command:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">freebayes
-f hs37.fa CHM1_CHM13_2.bam &gt; CHM1_CHM13_2.raw.vcf</font></font></p>
<p class="first-paragraph">Will it work? I am sure you know what to
expect. Now you will get the following error:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">could
not open hs37.fa</font></font></p>
<p class="first-paragraph">Neither the first nor the second commands
work as published. The file that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">freebayes</font></font>
attempts to open is the reference genome used in the variant calling
process. There is no explanation of what the author used or how to
get that data. We might guess from the name of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hs37.fa</font></font>
that it is the <i>Homo Sapiens build 37</i>, but we do not know which
particular variant of the genome it is. Also, should we really be
guessing, after all this is a top notch and major publication and we
can&rsquo;t even get past the first two commmands without
guesstimating what the author might have meant.</p>
<p>Ironically the author of the repository is the same person who, in
past has pointed out, quite emphatically, how important it is to use
the &ldquo;correct&rdquo; genomic build: <font color="#4f81bd"><a href="http://lh3.github.io/2017/11/13/which-human-reference-genome-to-use">Which
human reference genome to use?</a></font> So Dr.&nbsp;Heng which
human genome are you using? And since we are discussing this choice,
why are you using genome build that is more than ten years old? What
sense does that make to evaluate your variants against a ten-year-old
genome?</p>
<p>I hope we managed to demonstrate how even the very best
bioinformaticians fail, often <i>spectacularly</i>, at producing
reproducible analyses.</p>
<p>The pervasive lack of awareness when it comes to properly
describing an analytical process has immense negative repercussions
for life sciences, and is the problem that most urgently needs
solutions. Even for an advanced bioinformatician reproducing the
analysis that took place in this paper is mostly a mystery that they
could only uncover at significant additional effort. From that point
of view the analysis in this paper is no more reproducible than the
<font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Ebola paper
written</a></font> written five years ago by scientists with
supposedly far less awareness of computational reproducibility.</p>
<h2 class="western"><a name="the-gods-must-be-crazy"></a>55.8 The
gods must be crazy</h2>
<p class="first-paragraph">We will revisit the results from this
paper when we evaluate the variant calling accuracy. In the meantime
we&rsquo;ll leave you with a sentiment that best captures the state
of analysis reproducibility in bioinformatics:</p>
<p><img src="0000_html_5bffbd2078ffb67a.jpg" name="Image108" align="bottom" hspace="12" width="560" height="448" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="redo-explore-the-genome-of-a-paleolithic-era-archaic-human"></a>
56 Redo: Explore the genome of a paleolithic-era archaic human</h1>
<p class="first-paragraph">(Work in progress, unfinished)</p>
<p>The Denisovans are a Paleolithic-Era members of the genus Homo
that may belong to a previously unknown species. The genome sequence
of a Denisovan individual which lived about 41,000 years ago was
generated from a small fragment of a finger bone discovered in
Denisova Cave in southern Siberia in 2008.</p>
<ul>
	<li/>
<p class="compact">Website: <font color="#4f81bd"><a href="http://www.eva.mpg.de/denisova">http://www.eva.mpg.de/denisova</a></font></p>
	<li/>
<p class="compact">Data: <font color="#4f81bd"><a href="http://cdna.eva.mpg.de/denisova/">http://cdna.eva.mpg.de/denisova/</a></font></p>
</ul>
<p class="first-paragraph">Extract alignments that overlap with the 1
million bp wide interval starting on chromosome 3, at 120 million.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=http://cdna.eva.mpg.de/denisova/alignments/T_hg19_1000g.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a subset of the data that overlaps with a region.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b $URL 3:120,000,000-121,000,000 &gt; subset.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index subset.bam</font></font></p>
<p class="first-paragraph">The Aryl Hydrocarbon Receptor binds to the
location 17,375,390 on chromosome 7 of the human genome. Does the
denisovan genome show variants in the reads that overlap with this
position?</p>
<p><img src="0000_html_98867057d5cf93d7.png" name="Image109" align="bottom" hspace="12" width="560" height="312" border="0"/>
</p>
<p><b>NOTE</b>:</p>
<p class="block-text-western">URL based web access is only helpful
when you need to check a small subset of the data. If you need some
properties of the entire dataset, you are better off downloading the
data first, since you would need to stream it off the web anyway.</p>
<p class="block-text-western">Web access can be useful to answer
well-defined queries against a large number of datasets available on
the web.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequencing-instruments"></a>57
Sequencing instruments</h1>
<p class="first-paragraph">Sequencing instrumentation has evolved at
a dramatic pace.</p>
<p>In 2005 the 454 sequencing instrument jump-started the so-called
&ldquo;next-gen&rdquo; sequencing revolution. Within just ten years,
the 454 methodologies and the entire platform has been discontinued
and supplanted by more efficient instruments.</p>
<h2 class="western"><a name="is-there-a-document-that-compares-sequencing-instruments"></a>
57.1 Is there a document that compares sequencing instruments?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://onlinelibrary.wiley.com/doi/10.1111/j.1755-0998.2011.03024.x/abstract">Travis
Glenn&rsquo;s Field Guide to Next Generation DNA Sequencer</a></font>
(written in 2011) attempted to summarize the state of instrumentation
at that time. There are yearly updates to this guide that (unlike the
paper itself) are available for free.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.molecularecologist.com/2016/03/24/2016-ngs-field-guide-preview/">2016:
	Updates to the NGS Field Guide</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.molecularecologist.com/2018/03/31/sequencing-round-up-2018/">2018:
	Sequencing round-up 2018</a></font></p>
</ul>
<h2 class="western"><a name="what-is-in-a-name"></a>57.2 What is in a
name?</h2>
<p class="first-paragraph">You may have heard words such as next-gen,
high throughput sequencing, massively parallel sequencing etc. For an
eminently readable overview of the terminologies we recommend the
blog post by Keith Robison titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://omicsomics.blogspot.com/2019/02/beyond-generations-my-vocabulary-for.html">Beyond
	Generations: My Vocabulary for Sequencing Tech</a></font></p>
</ul>
<p class="first-paragraph">In the post Dr.&nbsp;Robison makes the
case that sequencing intstruments should be grouped by the
technologies that they represent:</p>
<ol>
	<li/>
<p class="compact">DNA source: Single Molecule vs.&nbsp;Clonal</p>
	<li/>
<p class="compact">Chemistry: Cyclic vs.&nbsp;Continuous</p>
	<li/>
<p class="compact">Physical surface: Beads, Wells, Surfaces,
	or Membranes</p>
	<li/>
<p class="compact">Signal type: Optical vs.&nbsp;Electrica</p>
	<li/>
<p class="compact">Measurement: Synthesis vs.&nbsp;Ligation
	vs.&nbsp;Pore Passing vs.&nbsp;Digestion vs.&nbsp;Hybridization
	vs.&nbsp;&hellip;.</p>
</ol>
<p class="first-paragraph">Overall the blog post above will give you
an quick overview of the methods of past, present and future.</p>
<h2 class="western"><a name="what-type-of-sequencing-instruments-are-in-use"></a>
57.3 What type of sequencing instruments are in use?</h2>
<p class="first-paragraph">The following instruments are currently in
use:</p>
<h3 class="western"><a name="illumina-miniseq-miseq-nextseq-hiseq"></a>
57.3.1 Illumina MiniSeq, MiSeq, NextSeq, HiSeq</h3>
<p class="first-paragraph">Illumina is the current undisputed leader
in the high-throughput sequencing market. Illumina currently offers
sequencers that cover the full range of data output. More details are
available on the <font color="#4f81bd"><a href="#ch057.xhtml##illumina">Illumina
sequencers</a></font> page.</p>
<ul>
	<li/>
<p class="compact">Up to 300 million reads (HiSeq 2500)</p>
	<li/>
<p class="compact">Up to 1500 GB per run (GB = 1 billion
	bases)</p>
</ul>
<h3 class="western"><a name="iontorrent-pgm-proton"></a>57.3.2
IonTorrent PGM, Proton</h3>
<p class="first-paragraph">The IonTorrent platform targets more
specialized clinical applications.</p>
<ul>
	<li/>
<p class="compact">Up to 400bp long reads</p>
	<li/>
<p class="compact">Up to 12 GB per run</p>
</ul>
<h3 class="western"><a name="pacbio-sequel"></a>57.3.3 PacBio Sequel</h3>
<p class="first-paragraph">This company is the leader in long read
sequencing. More details on the <font color="#4f81bd"><a href="#ch059.xhtml#pacbio">PacBio
sequencers</a></font> page.</p>
<ul>
	<li/>
<p class="compact">Up to 12,000 bp long paired-end reads</p>
	<li/>
<p class="compact">Up to 4 GB per run</p>
</ul>
<h3 class="western"><a name="minion"></a>57.3.4 MinION</h3>
<p class="first-paragraph">A portable, miniaturized device that is
not yet quite as robust and reliable as the other options. More
details on the <font color="#4f81bd"><a href="#ch060.xhtml#minionx">MinION
sequencers</a></font> page.</p>
<ul>
	<li/>
<p class="compact">Up to 10,000 long reads</p>
	<li/>
<p class="compact">Up to 240 MB per run (MB = 1 million bases)</p>
</ul>
<h2 class="western"><a name="what-are-obsolete-sequencing-instruments-i-should-know-about"></a>
57.4 What are obsolete sequencing instruments I should know about?</h2>
<p class="first-paragraph">Two significant platforms &ndash; SOLiD
and 454 &ndash; have been discontinued. Over time it is becoming
increasingly difficult to analyze data produced by these platforms,
as newer software packages do not support them. Older software
sometimes fails to run at all.</p>
<p>Specifically, the SOLiD platform uses so-called &ldquo;colorspace&rdquo;
data, a convoluted and indirect method to measure base identities
that makes data analysis even more complicated.</p>
<h2 class="western"><a name="how-accurate-are-sequencing-instruments"></a>
57.5 How accurate are sequencing instruments?</h2>
<p class="first-paragraph">Typically the longer read instruments
operate at substantially higher error rates than those producing
short reads.</p>
<ul>
	<li/>
<p class="compact">Illumina: 0.1% error rates (1/1000)</p>
	<li/>
<p class="compact">PacBio: 10% error rates(1/10)</p>
	<li/>
<p class="compact">MinION: 20% error rates (1/5)</p>
</ul>
<p class="first-paragraph">Errors in long reads are easier to correct
and account for than errors in short reads. Compare the error rates
between the Illumina (left) and PacBio (right) instruments.</p>
<p><img src="0000_html_4ebe1a96c060ddbd.png" name="Image110" align="bottom" hspace="12" width="560" height="201" border="0"/>
</p>
<p>Read lengths are radically different across platforms:</p>
<p><img src="0000_html_3be2ad9981884387.png" name="Image111" align="bottom" hspace="12" width="560" height="210" border="0"/>
</p>
<h2 class="western"><a name="how-do-sequencing-instruments-work"></a>57.6
How do sequencing instruments work?</h2>
<p class="first-paragraph">A typical sequencer measures a single
strand DNA fragment and from that fragment produces a &ldquo;sequencing
read&rdquo; of a certain length. The term &ldquo;read&rdquo; is very
widely used to define this single measurement.</p>
<p class="block-text-western"><b>Essential detail</b>: The resulting
sequencing read is typically shorter but can be longer than the
original fragment that it measures.</p>
<p class="first-paragraph">The read will end up shorter when the
sequencing protocol reaches its limit before reading the end of the
fragment. Having reads shorter than the DNA fragment is the most
common expectation for short read sequencing approaches. The read may
end up being longer than the original DNA fragment (some may also
call this &ldquo;template&rdquo;) when the read runs into artificial
constructs required by the sequencing protocols.</p>
<p>Below we will provide a sketch for the steps that go into a
sequencing library for the HiSeq sequencer.</p>
<p>Suppose that the original, double-stranded DNA fragment obtained
by isolating and fragmenting DNA from a cell is the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAACCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTGGGG</font></font></p>
<p class="first-paragraph">During library preparation, uniquely
designed DNA adapters of lengths that are typically over 30 bases are
added to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5'</font></font>
(right), and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3'</font></font>
(left) ends of each sequence.</p>
<p>Below we label these as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXX</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YYYY</font></font>.
Hence after adding the adapters the single-stranded sequences that
make it onto the instrument flowcell will be of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXAAAACCCCYYYY</font></font></p>
<p class="first-paragraph">the reverse strand&rsquo;s directionality
is reversed; hence it will form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXGGGGTTTTYYYY</font></font></p>
<p class="first-paragraph">For any given fragment, both or neither of
these strands may be sequenced.</p>
<p>The sequencer will typically recognize the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXX</font></font>
at the beginning and will not report it. We can illustrate the
process with arrows that show the direction of sequencing. Suppose
that the sequencer can generate reads of lengths 5:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAAT
  </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAATGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTAGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   AGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   &lt;----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      </font></font>
</p>
<p class="first-paragraph">It is important to note that the size
distribution of DNA fragments varies although the range typically is
within 20-50bp. What this means, however, is that the fragment may
occasionally be shorter than the measurement. When the read length is
exactly as long a given fragment, we get to the following situation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">    <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   AAAACCCCYYYY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YYYYTTTTGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   &lt;-------</font></font></p>
<p class="first-paragraph">As the read length grows longer than the
DNA fragment size, it will start running into the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3'</font></font>
adapter. Neither the sequencer nor software can immediately
differentiate artificial DNA from the natural sequence, if for
example the run-in is just one base long:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">    <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-----------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   AAAACCCCYYYY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YYYYTTTTGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;-----------</font></font></p>
<p class="first-paragraph">Hence our measurements above have
artifacts at their ends. It is a situation of a so-called
&ldquo;read-through&rdquo;, where the sequencing is longer than the
fragment. If the read-through is sufficiently long, say at least Five
bases, we may attempt to remove these computationally after
sequencing with tools that recognize the start of the adapter
sequence.</p>
<h2 class="western"><a name="can-reads-be-in-different-orientations"></a>
57.7 Can reads be in different orientations?</h2>
<p class="first-paragraph">Depending on sequencing protocols and
instrumentations, data may end up with read orientations such as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          ----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font>
</p>
<p class="first-paragraph">or even:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          ----&gt;</font></font></p>
<p class="first-paragraph">The vast majority of software tools cannot
handle data with these orientations and may treat them in various
incorrect ways.</p>
<p>It is relatively simple to convert data orientation. A simple
post-processing operation can be used to reverse complement the reads
into the most commonly used <font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;
&lt;----</font></font> orientation.</p>
<h2 class="western"><a name="what-is-paired-end-sequencing"></a>57.8
What is paired-end sequencing?</h2>
<p class="first-paragraph">Paired-end (PE) sequencing is a method to
sequence both ends of a fragment and to make the pairing information
available in the data.</p>
<p>DNA fragments are typically longer than the measured read lengths.
For many applications, it is hugely advantageous to be able to
measure (if not the entire piece) at least both ends of it. Many
biological processes start at specific locations in the genome,
knowing exactly where the fragment starts and ends can provide
critically important information.</p>
<p>For that reason, some instruments offer the option of running the
device differently to achieve two measurements from a single strand
DNA.</p>
<p>In the Illumina sequencing protocol, for example, the first round
of reads are called the &ldquo;single end&rdquo; (SE) or &ldquo;first&rdquo;
reads.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAATTTTGGGGCCCC</font></font></p>
<p class="first-paragraph">If the paired-end protocol is employed,
after producing the &ldquo;first&rdquo; reads the <i>same</i>
sequence is flipped over within the instrument, it is reverse
complemented and a second measurement is taken to produce another set
of reads. These are called the &ldquo;second&rdquo; reads.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGGGCCCCAAAATTTT</font></font></p>
<p class="first-paragraph">The net effect is that we obtain two
measurements from one single stranded fragment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;
  </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAATTTTGGGGCCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTAAAACCCCGGGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          &lt;----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      </font></font>
</p>
<p class="first-paragraph">The two reads are typically stored in
separate FASTQ files and are synchronized by name and order. Each
read in file 1 has a corresponding entry in file 2.</p>
<p>Important: It is our responsibility to maintain this order and
perform all operations on these files in such a manner that they are
kept in sync and each contains the reads in the same order.</p>
<h2 class="western"><a name="what-are-the-advantages-and-disadvantages-of-paired-end-sequencing"></a>
57.9 What are the advantages and disadvantages of paired-end
sequencing?</h2>
<p class="first-paragraph">Paired-end sequencing is more expensive
than single end sequencing but not radically so (perhaps 20% more).</p>
<p>The ability to identify the ends of a DNA fragment typically
carries a lot of importance, as biological processes always have a
directionality. Besides, two measurements of the same fragment offer
the user a better chance to correctly identify the location and
composition of the original DNA fragment. We recommend that genomic
variation and genome assembly analyses use paired-end sequencing as
much as possible.</p>
<p>On the downside, paired-end sequencing measures the same fragment
twice, hence at the same genomic coverage, it will use half as many
unique fragments. For analysis methods that use sequencing for
quantification such as RNA-Seq and ChIP-Seq, this may be a
disadvantage. Typically single end sequencing is used, as it produces
more uniform and higher resolution coverage.</p>
<h2 class="western"><a name="what-is-mate-pair-sequencing"></a>57.10
What is mate-pair sequencing?</h2>
<p class="first-paragraph">Mate-pair sequencing typically has the
same goals as the paired-end approach &ndash; it attempts to measure
two ends of a fragment. The implementation, however, is entirely
different and, unlike the paired-end sequencing, requires a better
understanding of each particular instrument.</p>
<p>Notably, mate-pair DNA fragments are much much longer than PE
methods and the read orientations are usually different, for example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAATTTTGGGGCCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          ----&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          </font></font>
</p>
<p class="first-paragraph">Paired-end data is mainstream, and most
tools support it directly. Mate-paired data is only supported by
specialized software.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="illumina-sequencers"></a>58 Illumina
sequencers</h1>
<h2 class="western"><a name="what-types-of-sequencers-does-illumina-manufacture"></a>
58.1 What types of sequencers does Illumina manufacture?</h2>
<p class="first-paragraph">MiniSeq - the smallest bench-top sequencer
Illumina sells (as of 2016). It only makes sense for those labs who
have limited budgets and want to get started with high-throughput
sequencing.</p>
<ul>
	<li/>
<p class="compact">Run time 4-24h (depending on the type of
	run)</p>
	<li/>
<p class="compact">Up to 150bp long paired-end reads</p>
	<li/>
<p class="compact">Up to 25 million reads per run</p>
	<li/>
<p class="compact">Up to 8 GB per run (GB = 1 billion bases)</p>
</ul>
<p class="first-paragraph">MiSeq - is perhaps the most popular
instrument that started the trend for truly &ldquo;bench-top&rdquo;
sequencers. MiSeq is small enough to fit on a regular lab bench and
only requires a network connection (if you wish to use BaseSpace or
off-load data from the instrument). Otherwise, it is entirely
self-contained.</p>
<ul>
	<li/>
<p class="compact">Run time 4-55h</p>
	<li/>
<p class="compact">Up to 300bp long paired-end reads</p>
	<li/>
<p class="compact">Up to 25 million reads per run</p>
	<li/>
<p class="compact">Up to 15 GB per run (GB = 1 billion bases)</p>
</ul>
<p class="first-paragraph">MiSeqDX is a particular variant of MiSeq
that is the first FDA-cleared high-throughput sequencer for <i>in
vitro</i> diagnostic (IVD) testing.</p>
<ul>
	<li/>
<p class="compact">Currently there are two Cystic Fibrosis
	assays approved for use with MiSeqDX.</p>
</ul>
<ul>
	<ul>
		<li/>
<p class="compact">A panel of 139 clinically relevant
		variants for CFTR</p>
		<li/>
<p class="compact">A test for a comprehensive view of the
		CFTR gene</p>
	</ul>
</ul>
<p class="first-paragraph">NextSeq 500/550 bench-top sequencer</p>
<ul>
	<li/>
<p class="compact">Run time 12-30h</p>
	<li/>
<p class="compact">Up to 150bp paired-end reads</p>
	<li/>
<p class="compact">Up to 400 million reads per run</p>
	<li/>
<p class="compact">Up to 120 GB per run (GB = 1 billion bases)</p>
</ul>
<p class="first-paragraph">HiSeq 2500/3000/4000 sequencers are the
workhorses of sequencing. HiSeq 2500 is meant for large-scale
genomics, and HiSeq 3000/4000 are meant for production-scale (think
core facilities).</p>
<ul>
	<li/>
<p class="compact">Run time 1-3.5d (HiSeq 3000/4000)</p>
	<li/>
<p class="compact">Run time 7h-6d (HiSeq 2500)</p>
	<li/>
<p class="compact">Up to 150bp paired end reads (HiSeq
	3000/4000)</p>
	<li/>
<p class="compact">Up to 250bp paired end reads (HiSeq 2500
	rapid mode)</p>
	<li/>
<p class="compact">Up to 5 billion reads (HiSeq 3000/4000)</p>
	<li/>
<p class="compact">Up to 300 million reads (HiSeq 2500)</p>
	<li/>
<p class="compact">Up to 1500 GB per run (GB = 1 billion
	bases)</p>
</ul>
<p class="first-paragraph">HiSeq X Five/X Ten sequencers: As the name
suggests, there are 5/10 sequencers (sold as a package) to enable
population-level genome sequencing. The HiSeq X instruments were
initially only certified for human DNA sequencing.</p>
<ul>
	<li/>
<p class="compact">Run time 3d</p>
	<li/>
<p class="compact">Up to 150bp paired end reads</p>
	<li/>
<p class="compact">Up to 6 billion reads per run</p>
	<li/>
<p class="compact">Up to 1800 GB per run (GB = 1 billion
	bases)</p>
</ul>
<p class="first-paragraph">NovaSeq 5000 : Meant for counting
applications</p>
<ul>
	<li/>
<p class="compact">New flow cells (S1, S2)</p>
	<li/>
<p class="compact">Two-color chemistry like NextSeq 500</p>
	<li/>
<p class="compact">50, 100 and 150bp paired-end</p>
	<li/>
<p class="compact">Up to 2 terabases data (1.6 billion reads)
	in 2.5 d</p>
	<li/>
<p class="compact">Two flowcells per sequencer, can be run
	independently</p>
	<li/>
<p class="compact">Same flowcells can be used for single-end
	or paired-end runs</p>
	<li/>
<p class="compact">One library pool per flow cell</p>
	<li/>
<p class="compact">Four lanes per flowcell, visually distinct
	but optically identical</p>
	<li/>
<p class="compact">On-board cluster generation</p>
</ul>
<p class="first-paragraph">NovaSeq 6000 : High sequence coverage
applications</p>
<ul>
	<li/>
<p class="compact">New flow cells (S1, S2, S3, and S4)</p>
	<li/>
<p class="compact">S3 and S4 flow cells only usable in NovaSeq
	6000</p>
	<li/>
<p class="compact">Two-color chemistry like NextSeq 500</p>
	<li/>
<p class="compact">50, 100 and 150bp paired-end</p>
	<li/>
<p class="compact">Up to 6 terabases data (10 billion reads)
	in 2 d</p>
	<li/>
<p class="compact">Two flowcells per sequencer, can be run
	independently</p>
	<li/>
<p class="compact">Same flowcells can be used for single-end
	or paired-end runs</p>
	<li/>
<p class="compact">One library pool per flow cell</p>
	<li/>
<p class="compact">Four lanes per flowcell, visually distinct
	but optically identical</p>
	<li/>
<p class="compact">On-board cluster generation</p>
	<li/>
<p class="compact">132 exomes/transcriptomes per run</p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="pacbio-sequencers"></a>59 PacBio
sequencers</h1>
<p class="first-paragraph">Pacific Biosciences is the leader in
Single Molecule Real Time (SMRT) sequencing technology. Their <i>Sequel</i>
platform is a high-throughput long read sequencer.</p>
<h2 class="western"><a name="what-resources-are-available-for-pacbio-data-analysis"></a>
59.1 What resources are available for PacBio data analysis?</h2>
<p class="first-paragraph">Since PacBio offers longer reads than
typical short read technologies, a new set of tools and pipelines are
required for data analysis.</p>
<p>Resources:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/PacificBiosciences/">https://github.com/PacificBiosciences/</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/PacificBiosciences/Bioinformatics-Training/wiki">https://github.com/PacificBiosciences/Bioinformatics-Training/wiki</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.pacb.com/products-and-services/analytical-software/devnet/">http://www.pacb.com/products-and-services/analytical-software/devnet/</a></font></p>
</ul>
<h2 class="western"><a name="what-is-an-smrt-cell"></a>59.2 What is
an SMRT cell?</h2>
<p class="first-paragraph">A SMRT cell consists of an array of
measurement devices called Zero-Mode Waveguides (ZMWs). At the time
of writing, <i>Sequel</i> system has one million ZMWs per SMRT cell.</p>
<h2 class="western"><a name="what-is-a-zero-mode-waveguide-zmw"></a>59.3
What is a Zero Mode Waveguide (ZMW)?</h2>
<p class="first-paragraph">A ZMW is a small hole with a unique
structure in a SMRT cell that enables real time sequencing
observation. It is a nanophotonic device with a diameter too small to
permit propagation of light in the wavelength range used for
detection.</p>
<h2 class="western"><a name="what-is-a-subread"></a>59.4 What is a
subread?</h2>
<p class="first-paragraph">A subread is an individual measurement
(read) that corresponds to the DNA template between the adapter
sequences. The PacBio methodology allows for the same template to be
measured multiple times, hence multiple measurements called subreads
can be produced by a single ZMW waveguide.</p>
<h2 class="western"><a name="how-many-subreads-are-produced-from-a-template"></a>
59.5 How many subreads are produced from a template?</h2>
<p class="first-paragraph">Divide the length of the DNA fragment
(template) by the length of a read. For example, if the templates are
1KB and the sequencer produces 10KB reads, 10 subreads can ideally be
produced from each template. This is a highly simplified case.
Realistically we tend to observe: (TODO)</p>
<h2 class="western"><a name="what-is-a-circular-consensus-sequence-ccs-read"></a>
59.6 What is a Circular Consensus Sequence (CCS) read?</h2>
<p class="first-paragraph">A CCS read is a consensus sequence
obtained by the alignment of all subreads. When the template is
shorter than the read length the polymerase will loop back,
sequencing the template again. A sequence from a single pass is
called a subread and multiple subreads can be used to obtain a single
consensus sequence with much higher accuracy than that of the
individual subreads. More than 2 full pass subreads are required to
generate a CCS read.</p>
<p><img src="0000_html_2c82c6d6a9e703d1.png" name="Image112" align="bottom" hspace="12" width="560" height="246" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-output-of-a-pacbio-run"></a>59.7
What is the output of a PacBio run?</h2>
<p class="first-paragraph">A PacBio run folder contains unaligned bam
files, adapter fasta file and metadata in xml format.</p>
<p>The primary output of a PacBio run is an unaligned BAM file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">subreads.bam</font></font>.
The data in this file is to be used for downstream analysis. All
rejected subreads, excised barcodes, and adapters are combined in a
file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">scraps.bam</font></font>.</p>
<h2 class="western"><a name="what-other-information-is-encoded-in-the-bam-file"></a>
59.8 What other information is encoded in the BAM file?</h2>
<p class="first-paragraph">PacBio-produced BAM files also encode
instrument-specific information in the header and alignment sections.</p>
<h2 class="western"><a name="what-sam-headers-does-the-instrument-set"></a>
59.9 What SAM headers does the instrument set?</h2>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">RG</font></font>
	(read group) - 8 character string read group identifier for PacBio
	data.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL</font></font>
	(platform) - contains &ldquo;PacBio&rdquo;.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PM</font></font>
	(platform model) - contains PacBio instrument series eg: SEQUEL.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PU</font></font>
	(platform unit) - contains PacBio movie name.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">DS</font></font>
	(description) - contains some semantic information about the reads
	in the group, encoded as a semicolon-delimited list of &ldquo;Key=Value&rdquo;
	strings.</p>
</ul>
<h2 class="western"><a name="how-is-the-bam-file-formatted"></a>59.10
How is the BAM file formatted?</h2>
<p class="first-paragraph">The query template name convention for a
subread is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{movieName}/{zmw-holeNumber}/{qStart}_{qEnd}</font></font></p>
<p class="first-paragraph">The query template name convention for a
ccs read is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{movieName}/{holeNumber}/ccs</font></font></p>
<h2 class="western"><a name="what-sam-tags-are-added-to-the-reads"></a>
59.11 What SAM tags are added to the reads?</h2>
<ul>
	<li/>
<p class="compact"><b>qs</b> - 0-based start of the query in
	the ZMW read (absent in CCS).</p>
	<li/>
<p class="compact"><b>qe</b> - 0-based end of the query in the
	ZMW read (absent in CCS).</p>
	<li/>
<p class="compact"><b>zm</b> - ZMW hole number.</p>
	<li/>
<p class="compact"><b>np</b> - NumPasses (1 for subreads, for
	CCS, it encodes the number of complete passes of the insert).</p>
	<li/>
<p class="compact"><b>rq</b> - expected accuracy of the read.</p>
	<li/>
<p class="compact"><b>sn</b> - average signal-to-noise ratio
	of A, C, G, and T (in that order) over the HQRegion</p>
</ul>
<h2 class="western"><a name="how-do-i-get-subreads-from-a-single-zmw"></a>
59.12 How do I get subreads from a single ZMW?</h2>
<p class="first-paragraph">We can use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">zm</font></font>
tag in the bam file to extract all subreads from a single zmw. This
can be done with the <i>bamtools filter</i> command.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
filter -in subreads.bam -out zmw.bam -tag 'zm:10027159'</font></font></p>
<p class="first-paragraph">The command will read the entire BAM file
to extract the matching entries. This approach is reasonable only if
we need to do this for one or two ZMWs. To split the BAM file into
parts just by ZMWs, a custom program needs to be written.</p>
<h2 class="western"><a name="why-would-i-want-to-split-a-bam-file-by-zmws"></a>
59.13 Why would I want to split a BAM file by ZMWs?</h2>
<p class="first-paragraph">The consensus caller (see below) operates
on data for a single ZMW at a time. You can massively speed up the
process by running the consensus caller in parallel on data that
corresponds to each ZMW separately.</p>
<h2 class="western"><a name="how-do-i-get-the-consensus-sequence-from-the-subreads"></a>
59.14 How do I get the consensus sequence from the subreads?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ccs</font></font>
command from the Unanimity C++ library can be used to generate CCS
reads. This command requires a minimum of 3 full pass subreads to a
generate a CCS read.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
ccs [options] INPUT OUTPUT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ccs
subreads.bam ccs.bam</font></font></p>
<h2 class="western"><a name="what-is-the-per-base-quality-of-pacbio-reads"></a>
59.15 What is the per-base quality of PacBio reads?</h2>
<p class="first-paragraph">Whereas the quality of the original reads
is low, the quality of consensus corrected reads improves
dramatically.</p>
<p>The FastQC report plot below shows the per-base quality of some
CCS reads.</p>
<p><img src="0000_html_ad4f4aa21edebc8.png" name="Image113" align="bottom" hspace="12" width="560" height="395" border="0"/>
</p>
<h2 class="western"><a name="how-good-are-the-consensus-corrected-reads"></a>
59.16 How good are the consensus corrected reads?</h2>
<p class="first-paragraph">On the lower left is the alignment of
subreads to the genome and on the right is the alignment of CCS
reads.</p>
<p><img src="0000_html_dbf58ae13fe4b1ef.png" name="Image114" align="bottom" hspace="12" width="560" height="200" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="minion-sequencers"></a>60 Minion
sequencers</h1>
<p class="block-text-western">Note: The MinION platform is undergoing
rapid changes that we are unable to keep up with in a timely fashion.</p>
<p class="block-text-western">The information in this chapter may be
obsolete.</p>
<h2 class="western"><a name="what-is-minion"></a>60.1 What is MinION?</h2>
<p class="first-paragraph">The MinION is a pocket-sized long read
sequencing instrument from Oxford Nanopore technologies powered by a
USB cable. It uses [nanopore sequencing] technology. Sequencing
occurs by measuring changes in the ionic current as DNA translocates
through protein nanopores in an insulated membrane on the flow cell.
Below each nanopore channel is a very sensitive electrical sensor. A
characteristic electric signal is generated by the flow of ions when
no molecule is in the pore. When a strand of DNA or RNA passes
through the pore, partially blocking the flow of ions, the current
pattern will be disturbed in a manner characteristic to the molecule
and dependent on the nucleotide sequence. The technology has
essentially no size limitations &ndash; if you manage to keep a
molecule intact during the library prep, reads up to hundreds of
kilobases can be obtained. A standard library prep will routinely
result in the majority of reads being greater than 10kb. Over the
past few years, Nanopore sequencing technology has seen a rapid
development with drastic improvements in throughput and accuracy.
[nanopore sequencing]:
<font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Nanopore_sequencing">https://en.wikipedia.org/wiki/Nanopore_sequencing</a></font></p>
<h2 class="western"><a name="what-is-minknow"></a>60.2 What is
MinKNOW?</h2>
<p class="first-paragraph">MinKNOW is the software that controls the
MinION sequencer. The scripts in MinKNOW software carry out several
tasks including device configuration, platform QC and calibration,
data acquisition, etc. During the platform QC, the quality of the
pores for sequencing is assessed and the most optimal pores will be
selected for starting the run. A sequencing script starts the
sequencing process. As the run progresses, it produces one FAST5 file
per DNA molecule and stores it in the specified run directory on the
SSD disk of your sequencing computer. It is possible to modify the
(python) sequencing scripts to tweak the run and obtain higher
yields, but this is something that is best left to those with
considerable experience with the technology.</p>
<h2 class="western"><a name="what-is-the-metrichor-agent"></a>60.3
What is the Metrichor Agent?</h2>
<p class="first-paragraph">The Metrichor Agent manages the connection
to the base-calling service in the cloud (hosted by Metrichor). When
Metrichor Agent is started on the sequencing laptop, it moves the
pre-base-called FAST5 files to the specified uploads folder and then
transfers a copy to the base-calling service. The FAST5 file with
original information and base calls is then returned and stored in
the specified download folder on the sequencing computer. In addition
to cloud base-calling, local alternatives are available. It&rsquo;s
expected that cloud base-calling will be discontinued in the near
future. Other applications are available in the cloud &ndash; such as
the What&rsquo;s In My Pot (WIMP) workflow, in which a metagenomic
sample is classified based on the obtained sequence to identify
species contained therein.</p>
<h2 class="western"><a name="what-is-the-output-of-a-minion-run"></a>60.4
What is the output of a MinION run?</h2>
<p class="first-paragraph">Each read sequenced on the MinION will
generate a FAST5 file, which is a variant of the [HDF5] file format.
This format is essentially a container for storing a variety of data
types (eg. integers, floats, strings, arrays) in a single file. The
format has a hierarchical structure with &lsquo;groups&rsquo; for
organizing data objects and &lsquo;datasets&rsquo; which contain a
multidimensional array of data elements. These files contain voltage
and time information, metadata identifying the pore and flow cell
used for sequencing, among other things. This data allows for
base-calling of the reads, either in the cloud or on your local
system. In addition, when new base-calling software is available, the
analysis can be repeated. Third-party base-callers for nucleotide
modifications such as methyl-cytosine are also available. [HDF5]:
<font color="#4f81bd"><a href="https://support.hdfgroup.org/HDF5/">https://support.hdfgroup.org/HDF5/</a></font></p>
<h2 class="western"><a name="what-do-pass-or-fail-subfolders-contain"></a>
60.5 What do &lsquo;pass&rsquo; or &lsquo;fail&rsquo; subfolders
contain?</h2>
<p class="first-paragraph">If you activate the quality filter at the
start of a run, the base-called data will get stored into &lsquo;pass&rsquo;
or &lsquo;fail&rsquo; subfolders in the specified download folder.
The pass/fail classification system is intended to help the user
distinguish higher and lower quality data. Lower quality data isn&rsquo;t
necessarily useless &ndash; depending on your application this data
may still be valuable. The classification criteria may change between
different versions of Metrichor.</p>
<h2 class="western"><a name="what-is-different-between-1d-and-2d-reads"></a>
60.6 What is different between 1D and 2D reads?</h2>
<p class="first-paragraph">In general two library preparations are
possible, termed 1D and 2D sequencing. The highest accuracy is
obtained by ligating a hairpin adapter on one side of the molecule,
hence sequencing the double-stranded DNA in both directions (2D). The
first strand is called the template read, the second strand the
complement. Since a consensus read can be created based on both
strands, the accuracy of a 2D read is higher. For a 1D library prep,
no hairpins are ligated and sequencing is performed for each strand
separately, effectively increasing the throughput but sacrificing
some accuracy. Since May 2016, the accuracy of 1D reads has reached
acceptable levels.</p>
<h2 class="western"><a name="how-do-i-extract-data-from-a-fast5-file"></a>
60.7 How do I extract data from a FAST5 file?</h2>
<p class="first-paragraph">Different commands and tools are available
to extract data from FAST5 file.</p>
<ul>
	<li/>
<p class="compact">HDFView</p>
	<li/>
<p class="compact">ont_fast5_api</p>
	<li/>
<p class="compact">poRe</p>
</ul>
<h2 class="western"><a name="what-is-hdfview"></a>60.8 What is
HDFView?</h2>
<p class="first-paragraph">HDFView provides a Graphical User
Interface (GUI) to browse the contents of a FAST5 file. Shown below
is an HDFView display of a 2D read sequence.</p>
<p><img src="0000_html_9585996ff0af366b.png" name="Image115" align="bottom" hspace="12" width="560" height="352" border="0"/>
</p>
<h2 class="western"><a name="how-can-i-extract-data-from-a-fast5-file"></a>
60.9 How can I extract data from a FAST5 file?</h2>
<p class="first-paragraph">Oxford Nanopore provides code to access
the files they produce.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ont_fast5_api</font></font>
is a simple interface to HDF5 files of the Oxford Nanopore .fast5
file format.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/nanoporetech/ont_fast5_api">https://github.com/nanoporetech/ont_fast5_api</a></font></p>
</ul>
<h2 class="western"><a name="where-to-find-most-up-to-date-tools-for-nanopore"></a>
60.10 Where to find most up to date tools for Nanopore?</h2>
<p class="first-paragraph">There is a Galaxy instance that collects
Nanopore specific tools that can provide you with a good starting
point:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://nanopore.usegalaxy.eu/">https://nanopore.usegalaxy.eu/</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequencing-data-preparation"></a>61
Sequencing data preparation</h1>
<h2 class="western"><a name="from-sample-submission-to-receiving-sequence-data"></a>
61.1 From sample submission to receiving sequence data</h2>
<p class="first-paragraph">Author: <b>Hemant Kelkar</b></p>
<p>Note: Discussion below draws upon personal observations over the
past few years. Your experience may vary due to practices prevalent
at the sequencing facility you use.</p>
<p>Getting samples ready for high-throughput sequencing is a result
of a days/weeks/months-long effort on your part but suffice it to say
that a significant amount of money and labor has already been
expended on that process.</p>
<h2 class="western"><a name="a-recap-of-high-throughput-sequencing"></a>
61.2 A recap of high-throughput sequencing</h2>
<p class="first-paragraph">High-throughput sequencing data broadly
fall into two categories:</p>
<ul>
	<li/>
<p class="compact">Short reads (300 bp or smaller).</p>
	<li/>
<p class="compact">Long reads (&gt;350 bp to several
	kilobases).</p>
</ul>
<p class="first-paragraph">What goes into a high-throughput
sequencing experiment may be familiar to many wet-lab scientists.</p>
<p>Here is a quick summary:</p>
<ul>
	<li/>
<p class="compact">Sequencers sequence sheared and purified
	DNA fragments.</p>
	<li/>
<p class="compact">Addition of technology-specific adapter
	sequences (oligos) create &ldquo;sequencing-ready libraries&rdquo;.</p>
	<li/>
<p class="compact">Controlled amplification (PCR) is required
	with most prep methods (PCR-free methods are available for some
	applications).</p>
	<li/>
<p class="compact">Sequencing is done by DNA synthesis (except
	for Oxford Nanopore and some optical mapping technologies).</p>
	<li/>
<p class="compact">Detection of synthesis is done using light
	(Illumina, PacBio) or non-light (Oxford Nanopore, Ion) methods.</p>
	<li/>
<p class="compact">Post-processing of &ldquo;raw&rdquo; data
	is required to generate the actual sequence.</p>
</ul>
<p class="first-paragraph">Note: Direct RNA sequencing has been
demonstrated with some technologies, but no commercial applications
are currently available.</p>
<p>Illumina sequencing is by far the most common high-throughput
sequencing technology currently in use.</p>
<ul>
	<li/>
<p class="compact">It has been commercially available for
	almost a decade.</p>
	<li/>
<p class="compact">Sequencers and sequencing kits have evolved
	to cover the entire spectrum of data yields.</p>
	<li/>
<p class="compact">Error profiles are reasonably well
	understood.</p>
	<li/>
<p class="compact">Sequencing is relatively cost-effective.</p>
</ul>
<p class="first-paragraph">Ion technology is used for specific
applications and when fast data turnaround is a high priority.</p>
<p>Pacific Biosciences implements a mature long read technology that
has been on the market for some time. Oxford Nanopore (another long
read method) is gaining interest in academic labs at a rapid pace.</p>
<h2 class="western"><a name="where-can-i-find-illustrated-summaries-of-sequencing-technologies"></a>
61.3 Where can I find illustrated summaries of sequencing
technologies?</h2>
<p class="first-paragraph">The following videos illustrate currently
popular sequencing technologies:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.youtube.com/watch?v=HMyCqWhwB8E">Illumina,
	Intro to Sequencing by Synthesis (HiSeq 2500)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.youtube.com/watch?v=pfZp5Vgsbw0">Illumina,
	Patterned Flow Cell Technology (HiSeq 3000,4000,X)</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.youtube.com/watch?v=WYBzbxIfuKs">Ion
	Torrent next-gen sequencing technology</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.youtube.com/watch?v=v8p4ph2MAvI">Pacific
	Biosciences (SMRT Technology)</a></font> and <font color="#4f81bd"><a href="http://www.youtube.com/watch?v=NHCJ8PtYCFc">Introduction
	to SMRT Sequencing</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.youtube.com/watch?v=3UHw22hBpAk">Oxford
	Nanopore Sequencing Technologies</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-select-proper-sample-identifiers"></a>
61.4 How do I select proper sample identifiers?</h2>
<p class="first-paragraph">Do NOT use spaces or the following
characters (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">?
( ) [ ] / \ = + &lt; &gt; : ; &quot; ' , * ^ | &amp; .</font></font>)
in sample identifiers. These characters can cause problems with
Illumina (and/or other technology) data pre-processing software and
should be avoided.</p>
<p>DO use sample names or identifiers that make immediate sense to
you. It is best to avoid personal identifiers of any kind to prevent
regulatory issues. If you use &ldquo;Sample1/Sample2&rdquo; for
multiple submissions, the data files that you receive will have those
exact names. While facilities will keep the data segregated, you may
have a hard time keeping track of &lsquo;what is what&rsquo; down the
road.</p>
<p>Sequencing cores may track samples using internal IDs (as part of
a LIMS or otherwise), but the sample name you submit will be carried
forward through the process. In the case of Illumina technology, this
name may get incorporated into the sequence file itself.</p>
<h2 class="western"><a name="how-to-choose-the-right-sequencing-method"></a>
61.5 How to choose the right sequencing method?</h2>
<p class="first-paragraph">The number of cycles of sequencing and
read type (single-end/paired-end) determines the kind of data you
will receive. The kind of library you made is also important. All of
these choices have a direct impact on the cost of sequencing. In
general, longer and paired-end reads (Illumina) and very long reads
(PacBio) are comparatively more expensive.</p>
<p>It is not possible to cover the entire breadth of sequencing
choices in this section.</p>
<p>Here is a representative example. If you are only interested in
read counts and differential gene expression (e.g., RNAseq), then
single-end reads of a reasonable size may be enough. You will not get
spatial information about fragments unless you choose to do
paired-end sequencing. If you were interested in identifying
full-length alternately spliced transcripts, then Pacific Biosciences
IsoSeq would be the method of choice (assuming there are no budget
limitations).</p>
<h2 class="western"><a name="why-is-a-submission-of-sample-metadata-critical"></a>
61.6 Why is a submission of sample metadata critical?</h2>
<p class="first-paragraph">Sample metadata (the type of sample,
specific sample characteristics, experiment type) that you include
with your sample submission may play a crucial role in the success of
sequencing.</p>
<p>Illumina sequencing technology assumes a reasonably uniform
distribution of nucleotides (A/C/T/G, ~25% each) in your samples. If
your sample is known to deviate from this assumption (e.g.,
amplicons, metagenomics, high GC/AT samples) then make a note of this
during sample submission. A &ldquo;spike-in&rdquo; may need to be
added to such samples (generally phiX is used, but any &ldquo;normal&rdquo;
genomic DNA may be used) to address &ldquo;low nucleotide diversity&rdquo;
(for sample examples noted above).</p>
<p>Ion/Oxford Nanopore technologies are not very robust when
homopolymers (e.g., AAAAA) of significant length are present in
samples. They would not be appropriate for such samples.</p>
<h2 class="western"><a name="why-is-samplelibrary-qc-essential"></a>61.7
Why is sample/library QC essential?</h2>
<p class="first-paragraph">Qubit and Bioanalyzer analysis are
preferred options (qPCR may also be used) for analyzing the quality
of starting material and finished libraries. Not all labs have access
to these technologies, and hence they may use alternate means for QC
(e.g., nanodrop). The amount of material you think you are submitting
invariably turns out to be less (sometimes 50% or more), which can
directly affect data yields.</p>
<p>Final sequencing data quality depends on the quality of samples
(garbage in, garbage out). It may be worth asking the sequencing
facility to do QC via their methods &ndash; some facilities may do
this as a part of their standard sequencing protocol. There may be a
charge for this QC, but it sure beats having sequencing fail because
of samples with lousy quality.</p>
<p>Making good sequencing libraries is an art. If you intend to keep
making libraries for many samples, then it would be cost effective to
acquire this expertise in house. Otherwise, it is best to get a
professional (at a core or elsewhere) to take care of this crucial
step.</p>
<h2 class="western"><a name="what-can-large-scale-projects-do-to-ensure-sample-data-integrity"></a>
61.8 What can large-scale projects do to ensure sample data
integrity?</h2>
<p class="first-paragraph">If you are submitting a large number of
samples (e.g., samples in multi-well plates), then it is useful to
think about having an independent means of identifying samples from
sequence data by correlation with an independent parameter (e.g.,
genotype). While sequencing facilities do take precautions in
handling plates/large batches of samples, a human is involved in some
steps and errors will sometimes occur.</p>
<h2 class="western"><a name="what-happens-during-a-sequencing-run"></a>
61.9 What happens during a sequencing run?</h2>
<p class="first-paragraph">Once your sample libraries clear these
initial checks, they are ready to run on the sequencer. As noted
elsewhere the machine run times may vary from a few hours to a week
depending on the type of run and model of the sequencer.</p>
<p>Illumina sequencers use an elegant combination of precision
engineered components, motors, manifolds, lasers, optics, and
proprietary onboard software to capture sequence data. It is possible
to monitor the sequencing run both locally and remotely (with the
right set up) in real time.</p>
<p>Illumina flowcells are imaged and processed in real time on the
workstation (or internal computer) attached to the sequencer.
Per-cycle base calls generated are saved in an elaborate folder
structure that contains data for each flowcell. The naming convention
for the folders is in this format:
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YYDDMM_SequencerName_RunNo_FlowcellBarcode</font></font>
(FlowcellBarcode is tagged in front with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
in case of sequencers that can run two flowcells at one time).</p>
<h2 class="western"><a name="what-does-an-original-illumina-sequence-data-folder-contain"></a>
61.10 What does an original Illumina sequence data folder contain?</h2>
<p class="first-paragraph">Depending on the run type (length,
flowcell, sequencer type) a raw data folder can contain between 25 to
~550 GB of data. There can be hundreds of thousands of files, in
multiple, nested folders, that follow an identical layout for every
flowcell. This folder structure allows for scripting and workflow
based processing of data. Individual file sizes range from a few
bytes to tens of megabytes. Some sequencing instruments (e.g., MiSeq)
may store some imaging data that are useful for diagnostic purposes
in case of run failure.</p>
<h2 class="western"><a name="how-does-one-get-finished-data-from-raw-sequence-data"></a>
61.11 How does one get &ldquo;finished&rdquo; data from &ldquo;raw&rdquo;
sequence data?</h2>
<p class="first-paragraph">Smaller sequencing cores may use the
onboard data analysis capabilities of sequencers
(MiniSeq/MiSeq/NextSeq) and BaseSpace, Illumina&rsquo;s cloud
computing solution for management of sequence data. If an institution
can&rsquo;t use third-party cloud computing solutions, a locally
installed variant of BaseSpace called &ldquo;BaseSpace Onsite&rdquo;
is also available as an option. BaseSpace uses &ldquo;apps&rdquo; to
provide added analytics functionality beyond the processing of raw
sequence data.</p>
<p>Larger cores generally use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>,
which is Illumina&rsquo;s software for converting the per-cycle base
calls (stored in BCL files) into the fastq format. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
is currently available in two versions (v. 1.8.4 for older
sequencers, v. 2.18.x for newer sequencers), but is backward
compatible with all current Illumina sequencers.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
does per-cycle BCL file to per-read FASTQ format conversion and
simultaneous demultiplexing of samples. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
can optionally trim adapters and remove Unique Molecular Identifiers
(UMI) from reads.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
conversion requires access to a cluster (or a high-performance
multicore computer). These conversion runs may take between 30 min to
a few hours (depending on the type of run). The result of a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
run is a folder of sequence files and several report/results
files/folders. These are used by sequencing facilities to do internal
quality control of sequence data. Once a sample passes these internal
QC measures, it is generally ready for release to the submitter.</p>
<h2 class="western"><a name="what-should-i-expect-to-get-from-a-sequencing-provider"></a>
61.12 What should I expect to get from a sequencing provider?</h2>
<p class="first-paragraph">The primary deliverable for high
throughput sequencing is a gzip compressed FASTQ formatted sequence
data file(s) for each sample. You will get as many files as the
number of samples you had submitted.</p>
<p>Filename formats may vary depending on how the local pipeline is
run. The file may be named something like:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SampleID_ACGGCTGAC_L001_R1_001.fastq.gz</font></font></p>
<p class="first-paragraph">If you requested paired-end sequencing,
then you should expect to get two files per sample. The order of the
reads within the files indicate the read pairings.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SampleID_ACGGCTGAC_L001_R1_001.fastq.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SampleID_ACGGCTGAC_L001_R2_001.fastq.gz</font></font></p>
<p class="first-paragraph">Note: In a rare instance, you may receive
R1/R2 reads in an interleaved format in a single file (R1_r1, R2_r1,
R1_r2, R2_r2, etc).</p>
<p>In case your samples are not multiplexed &ldquo;Barcode/Tag&rdquo;
(ACGGCTGAC above) in the filename may be replaced by the word
&ldquo;NoIndex&rdquo;. Newer versions of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
do not put the &ldquo;Barcode&rdquo; in the file name by default, so
it may not always be present.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">L001</font></font>
indicates the lane your sample ran in. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">001</font></font>
indicates a single data file per sample. Illumina software allows
sequence data files to be split into defined chunks (2 Million reads
per file is default). If that is the case, you will get a series of
files (with the same SampleID) where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">001</font></font>
in the file name will be incremented for each file chunk for that
particular sample (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">001</font></font>
through <font face="Consolas, serif"><font size="2" style="font-size: 11pt">nnn</font></font>).</p>
<h2 class="western"><a name="how-is-the-sequencing-data-delivered"></a>
61.13 How is the sequencing data delivered?</h2>
<p class="first-paragraph">This will largely depend on your
sequencing provider. Sequencing data files (even when compressed) can
be fairly large (tens of GB per sample). You will need adequate
internet bandwidth to download the files directly. Ideally, your
provider will create MD5 hashes for your files so that you can verify
the file integrity after download.</p>
<p>Commonly used data delivery options are :</p>
<ul>
	<li/>
<p class="compact">Delivery via a storage partition mounted on
	a local computer cluster (if you are using an internal sequencing
	facility).</p>
	<li/>
<p class="compact">Via the Internet (web links)</p>
	<li/>
<p class="compact">Illumina BaseSpace (if it is in use at the
	sequencing facility). You will need a free BaseSpace account to
	access data.</p>
	<li/>
<p class="compact">Amazon/Google/Microsoft/Other cloud storage
	solution. Preferred, if data needs to be shared with many people.</p>
	<li/>
<p class="compact">Shipping a copy of the data on external
	hard drives may still be an (or only) option for some who are
	security conscious, don&rsquo;t want their data via the internet, or
	do not have adequate bandwidth</p>
</ul>
<h2 class="western"><a name="what-should-i-do-with-the-raw-data"></a>61.14
What should I do with the raw data?</h2>
<p class="first-paragraph">You should ensure there is a backup copy
for internal needs. External drives are fine for small lab groups
(keep two copies on separate drives, for added security). Consider
submitting the raw data to NCBI SRA as soon as is practical. You can
request an embargo period (before the data becomes publicly
available) to allow you to complete the analysis and write a paper.</p>
<p>Many facilities keep internal backups of raw sequence and
processed data files for varying periods (ask about your facilities&rsquo;
data policy). Bear in mind that retrieval of data may incur an
additional charge, beyond a specified period.</p>
<h2 class="western"><a name="where-can-i-download-illumina-software"></a>
61.15 Where can I download Illumina software?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcl2fastq</font></font>
(and many other software packages) can be obtained from Illumina&rsquo;s
portal (<font color="#4f81bd"><a href="https://accounts.illumina.com/?ReturnUrl=http://icom.illumina.com/">iCom</a></font>)
by creating a free account. Most Illumina software is available at no
cost. Software obtained from iCom is NOT open source. The only
software hosted on iCom is software supported by Illumina technical
support.</p>
<p>Illumina maintains a GitHub repository (<font color="#4f81bd"><a href="https://github.com/Illumina">Illumina
Repo</a></font>) specifically for open source software. Open source
software on GitHub repo is NOT supported by Illumina technical
support.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequencing-coverage"></a>62 Sequencing
coverage</h1>
<h2 class="western"><a name="how-much-data-do-i-need"></a>62.1 How
much data do I need?</h2>
<p class="first-paragraph">The question of how much data one needs is
one of the first questions that everyone asks. The answers to this
question are always a bit complicated and depend on the goals of the
experiment.</p>
<p>The amount of data is typically expressed as &ldquo;coverage&rdquo;
and represents how many measurements (on average) will be taken from
each base of the genome. The coverage is typically expressed as &ldquo;x&rdquo;
(times) symbol. 10x would mean each base is, on average sequenced ten
times.</p>
<h2 class="western"><a name="what-are-typical-coverages-for-dna-sequencing"></a>
62.2 What are typical coverages for DNA sequencing?</h2>
<ul>
	<li/>
<p class="compact">Genome assembly 50x and above</p>
	<li/>
<p class="compact">Variation calling (resequencing) 20x and
	above.</p>
</ul>
<p class="first-paragraph">Always consider how many measurements you
might need to support a statement. For example, at a 20x resequencing
project, in a diploid organism, a heterozygous mutation would be, on
average covered, by ten reads.</p>
<h2 class="western"><a name="what-are-typical-coverages-for-rna-sequencing"></a>
62.3 What are typical coverages for RNA sequencing?</h2>
<p class="first-paragraph">In RNA-Seq experiments different
transcripts are expressed at different levels; hence a formula based
coverage is not directly applicable. Consult the RNA-Seq data
analysis chapter for more details.</p>
<p>In a nutshell, and at the risk of oversimplifying the problem, for
the same sequencing throughput, the number of replicates is more
important than the number of reads (coverage) per sample. For
human/mouse genomes 10 million reads per replicate (and 3 or more
replicates) usually suffice. For different transcriptome sizes scale
this number accordingly. Ten times smaller transcriptome will usually
need ten times fewer reads per sample.</p>
<p>The most reliable way to asses the required coverage for a
specific and perhaps novel analysis is to download published datasets
on the same organism and tissue and evaluate the results. Thankfully
the process of obtaining published data lends itself to automation as
demonstrated in <font color="#4f81bd"><a href="#ch064.xhtml#automate-sra">Automating
access to SRA</a></font>.</p>
<h2 class="western"><a name="how-is-genome-coverage-computed"></a>62.4
How is genome coverage computed?</h2>
<p class="first-paragraph">The simple definition of coverage is</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
= number of sequenced bases / total genome size</font></font></p>
<p class="first-paragraph">The only slightly confounding factor is
that the number of sequenced bases come in units of &ldquo;read
lengths&rdquo; rather than one base at a time.</p>
<p>Typically the coverage is expressed with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">x</font></font>
symbol. For example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10x</font></font>
indicates that on average, each base of the genome will be covered by
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>
measurements.</p>
<p>For a sequencing instrument that produces variable reads lengths
we have:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
= SUM(Li) / G</font></font></p>
<p class="first-paragraph">Where the sum goes over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
the read count:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Li
= length of read i</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">G
= size of the genome</font></font></p>
<p class="first-paragraph">For an instrument with fixed read length
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">L</font></font>,
that produces <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
reads the formula simplifies to:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
= N * L / G</font></font></p>
<p class="first-paragraph"><b>Example</b></p>
<p>What is the coverage of a 20KB virus if we measured 1000 reads of
50bp each?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
= 50 * 1000 / 20000 = 2.5</font></font></p>
<p class="first-paragraph">The coverage is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2.5x</font></font>
that is on average each base is sequenced <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2.5</font></font>
times. If we needed to raise that coverage to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10x</font></font>
we could do the following:</p>
<ol>
	<li/>
<p class="compact">Increase the read lengths.</p>
	<li/>
<p class="compact">Produce more reads.</p>
	<li/>
<p class="compact">A combination of both of the above.</p>
</ol>
<p class="first-paragraph">We can invert the formula and compute it,
but usually, it is straightforward to see by eye what needs to be
done. To get to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10x</font></font>
we need to increase the coverage by a factor of 4. We could choose
to:</p>
<ol>
	<li/>
<p class="compact">Increase <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L=200</font></font>
	and keep <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N=1000</font></font></p>
	<li/>
<p class="compact">Keep <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L=50</font></font>
	and increase <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N=4000</font></font></p>
</ol>
<p class="first-paragraph">Usually, there are various real-life
constraints on how the instrument operates and a combination ends up
being chosen such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L=100</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N=2000</font></font></p>
<h2 class="western"><a name="how-many-bases-have-not-been-sequenced"></a>
62.5 How many bases have not been sequenced?</h2>
<p class="first-paragraph">The coverage is an average metric and has
variability in it. At a given coverage some bases will likely not get
covered. There are formulas to compute it based on a theoretical
model called the Lander/Waterman model (random fragmentation).</p>
<p>At a coverage <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
the probability of a base not being sequenced is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P
= exp(-C)</font></font></p>
<p class="first-paragraph"><b>Example</b></p>
<p>How many bases will not be sequenced when we cover a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">20KB</font></font>
virus at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10x</font></font>
coverage.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">P
= exp(-10)</font></font></p>
<p class="first-paragraph">From command line:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c &quot;import math; print (math.exp(-10))&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Prints: 4.53999297625e-05</font></font></p>
<p class="first-paragraph">To figure out how many bases will not be
sequenced, we need to multiply this result by the genome length:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c &quot;import math; print (math.exp(-10) * 20000)&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Prints: 0.90799859525</font></font></p>
<p class="first-paragraph">So at 10x coverage, we will miss about one
base. When we cover the human genome at the same coverage:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">python
-c &quot;import math; print (math.exp(-10) * 3E9)&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">136199.789287</font></font></p>
<p class="first-paragraph">We will miss <font face="Consolas, serif"><font size="2" style="font-size: 11pt">136,199</font></font>
bases. Note, though, that these gaps will be theoretically dispersed
across the entire genome.</p>
<h2 class="western"><a name="do-theoretical-coverages-describe-reality"></a>
62.6 Do theoretical coverages describe reality?</h2>
<p class="first-paragraph">Not really. Think of theoretical coverages
as the ideal, absolute best case outcomes. Reality is typically a lot
more complicated. DNA does not fragment randomly and may exhibit
various non-random patterns. We don&rsquo;t sequence the genome
directly; instead, a series of complex material extraction and
preparation protocols need to be performed that will show preferences
toward some parts of the DNA. Notably:</p>
<ol>
	<li/>
<p class="compact">Increase your coverages well over the
	theoretical limits.</p>
	<li/>
<p class="compact">Sometimes some parts of the genome do not
	show up in the data. These are called <i>hard to sequence</i>
	regions, but it is not always clear why they are hard to sequence.</p>
	<li/>
<p class="compact">What part of the genome is uniquely
	coverable with a given read size? Repeat-rich regions of the genome
	require longer reads to resolve.</p>
</ol>
<p class="first-paragraph">Scientists often use terms such as
&ldquo;accessible,&rdquo; &ldquo;mappable,&rdquo; &ldquo;effective&rdquo;
genome to indicate that only some parts of the genome can be easily
studied.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="accessing-the-short-read-archive-sra"></a>
63 Accessing the Short Read Archive (SRA)</h1>
<h2 class="western"><a name="what-is-the-short-read-archive-sra"></a>63.1
What is the Short Read Archive (SRA)?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra">Short
Read Archive (SRA)</a></font> is a data storage service provided by
NCBI. The history of the SRA&rsquo;s development is relatively
convoluted and includes a widely announced, promoted, then retracted
cancellation of the service. In parallel, the data design and
interface were standardized prematurely, before the community fully
understood how scientists might need to access the data. The
incredible rate of growth of the deposited data locked and froze
those standards in place. Today it seems that it would be hard to
change them without inconveniencing a large number of people. No
wonder there is an entire &ldquo;cottage industry&rdquo; of tools
trying to bypass and automate access to SRA.</p>
<h2 class="western"><a name="what-are-the-sra-naming-schemes"></a>63.2
What are the SRA naming schemes?</h2>
<p class="first-paragraph">Data in the SRA is organized under a
hierarchical structure where each top category may contain one or
more subcategories:</p>
<ul>
	<li/>
<p class="compact">NCBI BioProject: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJN****</font></font>
	(example: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>)
	contains the overall description of a single research initiative; a
	project will typically relate to multiple samples and datasets.</p>
	<li/>
<p class="compact">NCBI BioSample: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAMN****</font></font>
	or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRS****</font></font>
	(example: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAMN03254300</font></font>)
	describe biological source material; each physically unique specimen
	is registered as a single BioSample with a unique set of attributes.</p>
	<li/>
<p class="compact">SRA Experiment: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRX****</font></font>
	a unique sequencing library for a specific sample</p>
	<li/>
<p class="compact">SRA Run: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR****</font></font>
	or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ERR****</font></font>
	(example: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553610</font></font>)
	is a manifest of data file(s) linked to a given sequencing library
	(experiment).</p>
</ul>
<h2 class="western"><a name="how-do-we-download-data-from-sra"></a>63.3
How do we download data from SRA?</h2>
<p class="first-paragraph">You can download data via a web browser
(after installing a web-plugin) or from the command line via the
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/docs/toolkitsoft/">sratoolkit</a></font>
package that can be installed via <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>.</p>
<h2 class="western"><a name="are-there-alternatives-to-the-sra"></a>63.4
Are there alternatives to the SRA?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://www.ebi.ac.uk/ena">European
Nucleotide Archive (ENA)</a></font> provides an alternative
repository to store the world&rsquo;s nucleotide sequencing
information, covering raw sequencing data, sequence assembly
information, and functional annotation. The naming schemes and the
organization of the data mirror those of the SRA, though the files
are stored directly as FASTQ and BAM formats and do not need
conversion via the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ssratoolkit</font></font>.</p>
<h2 class="western"><a name="where-is-the-sra-documentation"></a>63.5
Where is the SRA documentation?</h2>
<ul>
	<li/>
<p class="compact">The <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/books/NBK47528/">SRA
	Handbook</a></font> contains a detailed description of the Short
	Read Archive.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc">The
	SRA Toolkit Documentation</a></font> describes the command line
	tools that may be used to access data.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/ncbi/sra-tools">SRA
	Toolkit GitHub</a></font> may have more up-to-date information on
	troubleshooting.</p>
</ul>
<h2 class="western"><a name="how-does-the-sratoolkit-work"></a>63.6
How does the sratoolkit work?</h2>
<p class="first-paragraph">There are several tools (see the tool
directory). The most commonly used ones are:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>,
	downloads data in FASTQ format</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sam-dump</font></font>,
	downloads data in SAM alignment format</p>
</ul>
<p class="first-paragraph">A typical use case is</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
SRR1553607  </font></font>
</p>
<p class="first-paragraph">This creates the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553607.fastq</font></font></p>
<p class="first-paragraph">By default, the SRA data will concatenate
the paired reads. For paired-end reads the data needs to be separated
into different files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files  SRR1553607 </font></font>
</p>
<p class="first-paragraph">This creates the paired end files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553607_1.fastq
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553607_2.fastq</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
command does two things in sequence. It first downloads the data in
the so-called SRA format; then it converts it to FASTQ format. We can
ask the tool to convert just a subset of data; for example, the first
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10,000</font></font>
reads:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files -X 10000 SRR1553607 </font></font>
</p>
<p class="first-paragraph">But we note that even in that case,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
needs first to download the full data file.</p>
<h2 class="western"><a name="is-there-a-simpler-way-to-query-sra-numbers"></a>
63.7 Is there a simpler way to query SRA numbers?</h2>
<p class="first-paragraph">Philip Ewels created a service called SRA
Explorer that provides incomparably better interface than NCBI</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://sra-explorer.info/">https://sra-explorer.info/</a></font></p>
</ul>
<p class="first-paragraph">Since it is a service provided by an
independent scientists it may not be continuously updated/supported.</p>
<h2 class="western"><a name="where-is-the-sra-file"></a>63.8 Where is
the SRA file?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">srapath</font></font>
command can list the location of the file whether it has already been
downloaded locally or is still on the web:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">srapath
SRR1553607</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-run-5/SRR1553607/SRR1553607.1</font></font></p>
<p class="first-paragraph">but after performing a:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">prefetch
SRR1553607</font></font></p>
<p class="first-paragraph">the same <font face="Consolas, serif"><font size="2" style="font-size: 11pt">srapath
SRR1553607</font></font> will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/ncbi/public/sra/SRR1553607.sra</font></font></p>
<h2 class="western"><a name="where-do-downloads-go"></a>63.9 Where do
downloads go?</h2>
<p class="first-paragraph">When any download via <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
is initiated, the ** complete** data for that run id will be
downloaded to a location on your computer. You can view that path
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">srapath
SRR1553607</font></font></p>
<p class="first-paragraph">By default this location is
at<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/ncbi/public/sra/</font></font>.
There may be other sequences related to an SRA file, in which case
those too will be downloaded. To investigate the downloaded SRA data
do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l ~/ncbi/public/sra/</font></font></p>
<p class="first-paragraph">After the download is completed, a
conversion takes place according to the parameters.</p>
<p class="block-text-western">Note: Even if we only asked for the
first ten reads of a FASTQ file, the entire FASTQ content would need
to be downloaded to the &ldquo;hidden&rdquo; location then out of
that the first ten reads will be shown. For large datasets, this may
take a considerable amount of time and may use up a substantial
amount of disk space.</p>
<p class="first-paragraph">Now, if we were to access the same SRA run
the second time, the data would already be there, and conversion
would be much faster.</p>
<h2 class="western"><a name="what-is-a-spot-in-sra"></a>63.10 What is
a &ldquo;spot&rdquo; in SRA?</h2>
<p class="first-paragraph">SRA tools use the word &ldquo;spots&rdquo;
and not as &ldquo;reads&rdquo; (measurements), though the two almost
always seem to mean the same thing. Answers to the Biostar question:
<font color="#4f81bd"><a href="https://www.biostars.org/p/12047/">What
Is A &ldquo;Spot&rdquo; In Sra Format?</a></font> shed some light:</p>
<p>&ldquo;The spot model is Illumina GA centric. [&hellip;] All of
the bases for a single location constitute the spot.&rdquo;</p>
<p>As it turns out, the terminology is both confusing and may not
apply to other instruments.</p>
<p>In the Illumina sequencers, a single location on the flow cell
generates the first read, the barcode (if that exists), and the
second (paired) read (if the run was a paired-end sequencing run).</p>
<p>We almost always split the spots (via the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--split-files</font></font>
option) into originating files.</p>
<h2 class="western"><a name="how-do-we-get-information-on-the-run"></a>
63.11 How do we get information on the run?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra-stat</font></font>
program can generate an XML report on the data.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra-stat
--xml --quick SRR1553610</font></font></p>
<p class="first-paragraph">Get read length statistics on a PacBio
dataset:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra-stat
--xml --statistics SRR4237168</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="automating-access-to-sra"></a>64
Automating access to SRA</h1>
<p class="first-paragraph">In this chapter, we demonstrate how to
automate access to the Short Read Archive for the purpose of
obtaining data deposited with scientific publications.</p>
<p>Around the end of the chapter we show how to get information on
all the data stored in SRA and compile various statistics on them!
The first version of the book was written in the Fall of 2016. We
revisited this same chapter and updated the numbers in the Spring of
2019. How much did the numbers change? A lot more than I expected.
The data floodgates are open my friends!</p>
<h2 class="western"><a name="how-to-automate-downloading-multiple-sra-runs"></a>
64.1 How to automate downloading multiple SRA runs?</h2>
<p class="first-paragraph">The primary strategy is to first obtain
the project <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJN</font></font>
number, then from the project numbers get the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR</font></font>
run numbers, then run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
on each <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR</font></font>
number.</p>
<p>So start with the bio-project number found in the publication. For
example, in <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font> we
noted the that project number is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197</font></font></p>
<p class="first-paragraph">The command above returns a so called
&ldquo;search result&rdquo; that tells us that there are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">891</font></font>
sequencing runs for this project id. The search result is an XML file
that you can later pipe into other programs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;ENTREZ_DIRECT&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Db&gt;sra&lt;/Db&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">

&lt;WebEnv&gt;NCID_1_74789875_130.14.18.34_9001_1474294072_501009688_0MetA0_S_MegaStore_F_1&lt;/WebEnv&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;QueryKey&gt;1&lt;/QueryKey&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Count&gt;891&lt;/Count&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 &lt;Step&gt;1&lt;/Step&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;/ENTREZ_DIRECT&gt;</font></font></p>
<p class="first-paragraph">We can now <font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch</font></font>
and format the search result as so-called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo</font></font>
data type. The runinfo type is in CSV, comma separated value, format:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197 | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">The resulting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo.csv</font></font>
file is quite rich in metadata and contains a series of attributes
for the run. You may open the file in Excel for an easier
visualization though with time you&rsquo;ll be just as comfortable
viewing it at the command line:</p>
<p><img src="0000_html_9b21ec1cbf5a6576.png" name="Image116" align="bottom" hspace="12" width="560" height="248" border="0"/>
</p>
<p>Cutting out the first column of this file</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | cut -f 1 -d ',' | head</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">To filter for lines that match <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR</font></font>
and to store the first ten run ids in a file we could write:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | cut -f 1 -d ',' | grep SRR | head &gt; runids.txt</font></font></p>
<p class="first-paragraph">Our <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runids.txt</font></font>
file contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">We now have access to the SRR run ids for
the experiment. In the next step we wish to generate a series of
commands of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Note how the commands are mostly the same,
we only substitute some parts of each. In the UNIX world, different
techniques may be used to generate the commands above. Our choice for
automation uses the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font>
tool that executes commands in &ldquo;parallel,&rdquo; one for each
CPU core that your system has. It is an extremely versatile command
that has a large number of unique features, well suited to most tasks
you will face. It is worth spending some time to understand how
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font>
works to fully appreciate the productivity gains you can make in your
work.</p>
<ul>
	<li/>
<p class="compact">See: <font color="#4f81bd"><a href="https://www.biostars.org/p/63816/">Tool:
	Gnu Parallel - Parallelize Serial Command Line Programs Without
	Changing Them</a></font></p>
</ul>
<p class="first-paragraph">For example, here we run the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>
command on each element of the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel &quot;echo This is number {}&quot;</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">This
is number SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">This
is number SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">This
is number SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">This
is number SRR1972920</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">We can plug the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
command in the place of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel fastq-dump -X 10000 --split-files {}</font></font></p>
<p class="first-paragraph">The command above will generate twenty
files, (two files for each of the ten runs).</p>
<h2 class="western"><a name="how-to-filter-for-various-metadata"></a>64.2
How to filter for various metadata?</h2>
<p class="first-paragraph">We also have options for filtering for
various metadata in this file. We can, for example, filter for runs
with the release data of August 19, 2014:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
info.csv | grep '2014-08-19'  | cut -f 1 -d ',' | grep SRR | head -10
&gt; ids.txt</font></font></p>
<h2 class="western"><a name="is-there-even-more-metadata"></a>64.3 Is
there even more metadata?</h2>
<p class="first-paragraph">Yes there is. The runinfo format is only a
subset of all the information SRA maintains on the run, sometimes
does not include essential information that you might need. When that
happens you will need to process the metadata downloaded in another
format, called a document summary (docsum). What? Another format for
the same information? Thanks, NCBI!</p>
<p>Let&rsquo;s get the docsum XML then:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197 | efetch -format docsum &gt; docsum.xml</font></font></p>
<p class="first-paragraph">The docsum data is an awkwardly
overcomplicated format that will make you sweat for every piece of
information that you want to wrangle out of it. View the docsum and
brace yourself.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
docsum.xml | more</font></font></p>
<h2 class="western"><a name="how-do-i-process-the-docsum-format"></a>64.4
How do I process the docsum format?</h2>
<p class="first-paragraph">The strategy for processing docsum files
is to transform the XML with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
command. If you knew what tags the information is stored under you
could do a</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
docsum.xml | xtract -pattern DocumentSummary -element
Bioproject,Biosample,Run@acc | head</font></font></p>
<p class="first-paragraph">The command produces the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197
SAMN03253746    SRR1972976</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197
SAMN03253745    SRR1972975</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197
SAMN03253744    SRR1972974</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197
SAMN03254300    SRR1972973</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197
SAMN03254299    SRR1972972</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Finding the right tags is not an easy
task, especially since no effort was made to make the docsum
formatted in a humanly readable way. You can &ldquo;prettify&rdquo;
the format by setting up the following alias:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias
pretty=&quot;python -c 'import sys;import
xml.dom.minidom;s=sys.stdin.read();print(xml.dom.minidom.parseString(s).toprettyxml());'&quot;</font></font></p>
<p class="first-paragraph">using the</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
docsum.xml | pretty | more</font></font></p>
<p class="first-paragraph">allows you to investigate the XML file in
a somewhat more manageable rearrangement.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-much-data-in-the-sra"></a>65 How
much data in the SRA?</h1>
<p class="first-paragraph">As you gain a better understanding of SRA
is you may ask yourself, how much data is there? What technologies
are in use, etc. We will go through a few of these questions below.
Notably, we have run the same analyses at the end of 2016 as well as
at the start of 2019 (roughly two years later); hence it will be
interesting to see what changed within just two years.</p>
<p>While theoretically, you could get the runinfo on all SRA runs
with a single <font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch</font></font>
query, in reality, that approach will not work. The service will fail
in various ways. It was just not designed for this use case. In
general, when downloading large amounts of data, the best approach,
whenever possible, is to divide the request into smaller queries. For
example, we can limit the query to a single month with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query '&quot;2015/05&quot;[Publication Date]' | efetch
-format runinfo &gt; 2015-04.csv</font></font></p>
<p class="first-paragraph">To use for the entire period, we will need
to automate this line of code to iterate over each year and month
pair. For that purpose. We wrote a script called <font color="#4f81bd"><a href="http://data.biostarhandbook.com/sra/sra-runinfo.sh">sra-runinfo.sh</a></font>
that downloads a single month&rsquo;s worth of data and ran it for
every range. You don&rsquo;t need to run the script yourself though;
you may just as well obtain the final data from the handbook data
site. We include the command for educational reasons, as a potential
starting point for other scripts of similar functionality:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate the input sequences</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq
2007 2019 &gt; years.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seq
1 12 &gt; months.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the script that downloads one months' worth of data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc http://data.biostarhandbook.com/sra/sra-runinfo.sh</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the script executable</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chmod
+x sra-runinfo.sh</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run on each pair, but just request in parallel at a time so the NCBI
does not ban you.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel
-j 1 --eta --progress --verbose ./sra-runinfo.sh {1} {2} ::::
years.txt :::: months.txt</font></font></p>
<p class="first-paragraph">As mentioned before you can also directly
download our results of running the script with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the runinfo data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc http://data.biostarhandbook.com/sra/sra-runinfo-2019-01.tar.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Unpack the runinfo data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
xzvf sra-runinfo-2019-01.tar.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
How big is the unpacked directory?</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">du
-hs sra</font></font></p>
<p class="first-paragraph">It is a reasonably large file (512MB
compressed, 2.7GB unpacked) that doubled in size since 2016!</p>
<p>We can now answer a series of interesting questions regarding how
much data is in the SRA.</p>
<h2 class="western"><a name="how-many-sequencing-runs-are-in-the-sra"></a>
65.1 How many sequencing runs are in the SRA?</h2>
<p class="first-paragraph">Combine all files into a single file; keep
only lines that contain SRR numbers:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
sra/*.csv | grep SRR &gt; allruns.csv</font></font></p>
<p class="first-paragraph">How many runs are there:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
allruns.csv | wc -l</font></font></p>
<p class="first-paragraph">Prints the number of sequencing runs
stored in SRA, in 2016 it was almost 1.5 million, today it clocks in
at almost 4.5 million sequencing runs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4,373,757</font></font></p>
<p class="first-paragraph">In just two years the total number of
sequencing runs tripled! Note that these are data only for the
published sequencing runs deposited at NCBI!</p>
<h2 class="western"><a name="how-do-we-extract-columns-from-a-comma-separated-file"></a>
65.2 How do we extract columns from a comma separated file?</h2>
<p class="first-paragraph">Cutting columns on tabs is usually fine,
as tabs are not part of a standard text. Commas, on the other hand,
are commonly used in different contexts as well. When splitting
directly on commas with a tool like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font>,
we may end up lining up the wrong columns. Hence, we need a more
sophisticated approach that cuts the file correctly: at the &ldquo;column
delimiter&rdquo; commas and not at the commas that are within a text
field (quotes).</p>
<p>In other words, we need a tool that operates on <font face="Consolas, serif"><font size="2" style="font-size: 11pt">csv</font></font>
files, for example:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvkit</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">miller</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvtk</font></font></p>
</ul>
<p class="first-paragraph">and other similar tools.</p>
<p>In the following commands, we will replace <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cut</font></font>
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvcut</font></font>
command that comes with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">csvkit</font></font>
package. You can install csvkit with</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install csvkit</font></font></p>
<p class="first-paragraph">we now run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">datamash</font></font>
to summarize the number of sequenced bases across all runs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
allruns.csv |  csvcut -c 5 | datamash sum 1</font></font></p>
<p class="first-paragraph">in 2016 this number was <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4.97e+15</font></font>,
today is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2.25e+16</font></font></p>
<p class="first-paragraph">Note how in two years the amount of data
grew by order of magnitude! from 5 peta bases to 2.2 exa bases.</p>
<p>As you have seen before we often make use of the pattern:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq -c | sort -rn</font></font></p>
<p class="first-paragraph">to count unique entries in a file. We can
create a shortcut to this construct to run it more readily. In the
Unix world a shortcut is a called an alias:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias
tabulate='sort | uniq -c | sort -rn'</font></font></p>
<p class="first-paragraph">Now we can use the shortcut named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tabulate</font></font>
to make our the commands look a little simpler. The downside of
shortcuts like this is that they make commands more abstract and a
little less explicit, there is &ldquo;hidden&rdquo; information on
what <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tabulate</font></font>
does. Other subtle &ldquo;gotchas&rdquo; may also trip you up,
aliases for example won&rsquo;t be visible in a subshell (a shell
opened from a shell) etc. We use the aliases here as a demonstration.</p>
<h2 class="western"><a name="how-many-sequencing-runs-per-organism"></a>
65.3 How many sequencing runs per organism?</h2>
<p class="first-paragraph">Column 29 of the run info file contains
the organism:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
allruns.csv | csvcut -c 29 | tabulate | head</font></font></p>
<p class="first-paragraph">in 2016 the list was:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">480,622
Homo sapiens</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">141,135
Mus musculus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
47,929 human metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
37,644 soil metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
36,740 human gut metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
26,440 Salmonella enterica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
24,641 Drosophila melanogaster</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
21,396 Arabidopsis thaliana</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
19,577 Salmonella enterica subsp. enterica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
19,370 Plasmodium falciparum</font></font></p>
<p class="first-paragraph">in 2019 the list is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1,201,332
Homo sapiens</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 569,555 Mus musculus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 129,382 soil metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 119,229 human gut metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 117,301 Salmonella enterica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 102,580 human metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  62,858 Escherichia coli</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  52,308 gut metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  51,124 Salmonella enterica subsp. enterica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  51,078 Drosophila melanogaster</font></font></p>
<p class="first-paragraph">I have added commas into the integers for
readability, the command itself won&rsquo;t do that.</p>
<h2 class="western"><a name="how-many-sequencing-runs-for-each-sequencing-platform"></a>
65.4 How many sequencing runs for each sequencing platform?</h2>
<p class="first-paragraph">Column 19 of the run info file contains
the sequencing platform name:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
allruns.csv | csvcut -c 19  | tabulate  | head</font></font></p>
<p class="first-paragraph">in 2016 the list was:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1,231,420
ILLUMINA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 182,407 LS454</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  31,865 ABI_SOLID</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  31,223 PACBIO_SMRT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  14,550 ION_TORRENT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   3,736 COMPLETE_GENOMICS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     869 HELICOS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     398 CAPILLARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      96 OXFORD_NANOPORE</font></font></p>
<p class="first-paragraph">in 2019 the list is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3,960,154
ILLUMINA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 248,304 LS454</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  57,243 PACBIO_SMRT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  52,161 ION_TORRENT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  40,204 ABI_SOLID</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   54,83 BGISEQ</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   45,89 COMPLETE_GENOMICS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   2,493 CAPILLARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   2,119 OXFORD_NANOPORE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   1,005 HELICOS</font></font></p>
<h2 class="western"><a name="how-many-runs-per-sequencing-instrument-model"></a>
65.5 How many runs per sequencing instrument model?</h2>
<p class="first-paragraph">Column 20 of the run info file contains
the instrument model name:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
allruns.csv | csvcut -c 20  | tabulate  | head</font></font></p>
<p class="first-paragraph">in 2016 the list was:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">633,701
Illumina HiSeq 2000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">203,828
Illumina MiSeq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">174,695
Illumina HiSeq 2500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">120,715
454 GS FLX Titanium</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
99,268 Illumina Genome Analyzer II</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
47,066 Illumina Genome Analyzer IIx</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
33,187 454 GS FLX</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
26,527 NextSeq 500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
25,243 Illumina Genome Analyzer</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
17,558 PacBio RS II</font></font></p>
<p class="first-paragraph">in 2019 the list is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1,086,037
Illumina HiSeq 2000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1,013,347
Illumina MiSeq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 915,073 Illumina HiSeq 2500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 286,384 NextSeq 500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 192,972 Illumina HiSeq 4000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 158,024 454 GS FLX Titanium</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 142,459 HiSeq X Ten</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 108,684 Illumina Genome Analyzer II</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  61,321 Illumina Genome Analyzer IIx</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  44,219 454 GS FLX</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="visualizing-sequencing-data-quality"></a>
66 Visualizing sequencing data quality</h1>
<p class="first-paragraph">The first step after receiving the
sequencing data is to evaluate its quality. It is how you &ldquo;get
to know&rdquo; your data.</p>
<h2 class="western"><a name="what-part-of-the-fastq-file-gets-visualized"></a>
66.1 What part of the FASTQ file gets visualized?</h2>
<p class="first-paragraph">The FASTQ format contains sequenced bases
and a quality value associated with each. For example, consider a
record that looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@id</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">05:I</font></font></p>
<p class="first-paragraph">The characters <font face="Consolas, serif"><font size="2" style="font-size: 11pt">05:I</font></font>,
when decoded, corresponds to quality measures of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15
20 25 40</font></font>. Basically it contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@id</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10
20 25 40</font></font></p>
<p class="first-paragraph">Were each of the numbers corresponds to an
estimate generated by the instrument as to how reliable the basecall
is. The formula is to divide the number by 10 and raise 10 to that
power,</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20
-&gt; 10^(20/10) -&gt; 10^-2 -&gt; 1/100 -&gt; 0.01 -&gt; 1%</font></font></p>
<p class="first-paragraph">thus a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">20</font></font>
means that the instrument thinks that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1%</font></font>
of time the basecall would be incorrect. Thus the FASTQ file contains
the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@id</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10%
1% 0.3% 0.01%</font></font></p>
<p class="first-paragraph">when transformed into errors. Data quality
visualizations generate various plots out of these numbers, for
example, distribution of values by each base or by the average for
each sequence.</p>
<p>Also, the composition of sequences may also be used to identify
characteristics that could indicate errors.</p>
<h2 class="western"><a name="are-fastq-errors-values-accurate"></a>66.2
Are FASTQ errors values accurate?</h2>
<p class="first-paragraph">Not really. In our observation the numbers
are quite unreliable - treat them as an advisory rather than accurate
measurements.</p>
<h2 class="western"><a name="what-is-the-fastqc-tool"></a>66.3 What
is the FastQC tool?</h2>
<p class="first-paragraph">The undisputed champion of quality control
visualization is a tool named <font color="#4f81bd"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></font>
developed by <font color="#4f81bd"><a href="http://babraham.ac.uk/">Babraham
Institute</a></font>, an independent, charitable life sciences
institute involved in biomedical research.</p>
<p>Even though it is a de-facto standard of visualization, its
results are not always the simplest to interpret. On the positive
side, the tool is easy to run (requires only Java), simple,
reasonably efficient (though it is tuned for the Illumina platform
and may be unstable on other types of data) and produces
aesthetically pleasing plots. On the downside, some of its beautiful
charts are uninformative and occasionally confusing. For example, in
some cases, plots will switch from non-binned to binned columns
within the same panel, easily misleading casual observers.</p>
<p>Then there are plots, such as the K-MER plot and the
Overrepresented Sequences plot, that don&rsquo;t show what most
people assume they do. There is also the difficulty of having to open
HTML files one by one for result interpretation. In the current state
of the field, where dozens of samples are being investigated in any
given sequencing run, even just opening the results of the QC
analysis becomes exceedingly tedious. The seemingly simple task to
check the data quality requires clicking and scrolling through each
file individually.</p>
<p>To some extent, even the tool&rsquo;s name is confusing. FASTQC
does not perform quality control: it only visualizes the quality of
the data.</p>
<p>But even with its flaws, FastQCis still by far the best FASTQ
quality visualization tool. It is one of the tools that scientific
funding agencies should seek to fund even if the author were not to
apply for funding.</p>
<h2 class="western"><a name="how-does-fastqc-work"></a>66.4 How does
FastQC work?</h2>
<p class="first-paragraph">FastQC generates its reports by evaluating
a small subset of the data and extrapolating those findings to the
entirety of the dataset. Many of the metrics are only computed on the
first 200,000 measurements then are being tracked through the rest of
the data.</p>
<p>The <font color="#4f81bd"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">tool
help</a></font> contains detailed (although not entirely
satisfactory) descriptions of what each data analysis plot is
supposed to show.</p>
<h2 class="western"><a name="how-do-we-run-fastqc"></a>66.5 How do we
run FastQC?</h2>
<p class="first-paragraph">Obtain and unpack the data for different
sequencing platforms:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/data/sequencing-platform-data.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
xzvf sequencing-platform-data.tar.gz</font></font></p>
<p class="first-paragraph">Run the FastQC tool:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
illumina.fq</font></font></p>
<p class="first-paragraph">This action generates an HTML file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">illumina_fastqc.html</font></font>
that contains the results of the run. There are options you can pass
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
illumina.fq --nogroup</font></font></p>
<p class="first-paragraph">that will change the plots in some way,
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--nogroup</font></font>
option, for example, turns off the binning on columns.</p>
<p>There is data from several platforms in the file you downloaded
above. As of 2016, one will fail to run with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
minion.fq </font></font>
</p>
<p class="first-paragraph">Will raise a memory error. This is just an
example of how some tools are too closely tuned for one specific
sequencing instrument. Even our quality control tool can fail on a
&lsquo;standard&rsquo; dataset that happens to be from a different
platform.</p>
<h2 class="western"><a name="should-i-be-worried-about-the-stoplight-symbols"></a>
66.6 Should I be worried about the &ldquo;stoplight&rdquo; symbols?</h2>
<p class="first-paragraph">Usually not.</p>
<p>When FastQC runs, it generates &ldquo;stoplight&rdquo; icons for
each analysis having &ldquo;pass,&rdquo; &ldquo;warning,&rdquo; and
&ldquo;error&rdquo; symbols. Most of the time, these symbols are not
meaningful. They were developed for only a particular class of
samples and library preparation methods and just for certain types of
instruments.</p>
<p><img src="0000_html_ba240437f7acb2f6.png" name="Image117" align="bottom" hspace="12" width="560" height="432" border="0"/>
</p>
<p>Though it is fair to say if most or all your stop light icons are
red, then you probably have a data problem.</p>
<h2 class="western"><a name="what-does-the-sequence-quality-visualization-tell-us"></a>
66.7 What does the sequence quality visualization tell us?</h2>
<p class="first-paragraph">The simplest way to snapshot the quality
of a sequencing run is via a chart that plots the error likelihood at
each position averaged over all measurements.</p>
<p>The vertical axis are the FASTQ scores that represent error
probabilities:</p>
<ul>
	<li/>
<p class="compact">10 corresponds to 10% error (1/10),</p>
	<li/>
<p class="compact">20 corresponds to 1% error (1/100),</p>
	<li/>
<p class="compact">30 corresponds to 0.1% error (1/1,000) and</p>
	<li/>
<p class="compact">40 corresponds to one error every 10,000
	measurements (1/10,000) that is an error rate of 0.01%.</p>
</ul>
<p class="first-paragraph">The three-colored bands illustrate the
typical labels that we assign to these measures: reliable (30-40,
green), less reliable (20-30, yellow) and error prone (1-20, red).
The yellow boxes contain 50% of the data, the whiskers indicate the
75% outliers.</p>
<p><img src="0000_html_8b00a5ebe4d4e6b2.png" name="Image118" align="bottom" hspace="12" width="560" height="409" border="0"/>
</p>
<h2 class="western"><a name="what-does-the-sequence-length-histogram-show"></a>
66.8 What does the sequence length histogram show?</h2>
<p class="first-paragraph">The sequence length distribution shows how
many sequences of each length the data contains. For fixed read
length instruments, like the Illumina sequencer, all read lengths are
the same. For long read technologies like the PacBio and MinION, the
distribution can be a lot more varied.</p>
<p><img src="0000_html_3045823860bf625e.png" name="Image119" align="bottom" hspace="12" width="560" height="401" border="0"/>
</p>
<h2 class="western"><a name="what-does-the-sequence-quality-histogram-tell-us"></a>
66.9 What does the sequence quality histogram tell us?</h2>
<p class="first-paragraph">Another way to visualize data quality is
to generate the histograms of the average qualities. The horizontal
scales are the quality scores; the vertical axis indicates the number
of reads of that quality.</p>
<p><img src="0000_html_a3d552650505b6ce.png" name="Image120" align="bottom" hspace="12" width="560" height="407" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-next-step-after-visualizing-the-quality"></a>
66.10 What is the next step after visualizing the quality?</h2>
<p class="first-paragraph">After visualization, you need to decide
whether the data should be corrected. This process is called Quality
Control (QC) and typically follows this process:</p>
<ol>
	<li/>
<p class="compact">Evaluate (visualize) data quality.</p>
	<li/>
<p class="compact">Stop QC if the quality appears to be
	satisfactory.</p>
	<li/>
<p class="compact">If the quality is still inadequate, execute
	one or more data altering steps then go to step 1.</p>
</ol>
<p class="first-paragraph">There are many tools that you can use to
alter the data. We will cover a few in the subsequent chapters. Each
tool has benefits but also may require various tradeoffs.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequencing-quality-control"></a>67
Sequencing quality control</h1>
<h2 class="western"><a name="what-is-quality-control"></a>67.1 What
is quality control?</h2>
<p class="first-paragraph">Quality control (abbreviated as QC from
now on) is the process of improving data by removing identifiable
errors from it. It is typically the first step performed after data
acquisition.</p>
<p>Since it is a process that alters the data, we must be <b>extremely
cautious</b> not to introduce new features into it inadvertently.
Ideally, we want the same data (information) only better (more
accurate). But please do note:</p>
<p class="block-text-western">It is common to expect too much from
quality control. Remember that QC cannot turn bad data into useful
data and we can never salvage what appears to be a total loss.</p>
<p class="first-paragraph">When the sequencing data looks terrible
from the beginning, it is best to move on and collect new data. No
amount of QC will save it. Only wasted work and frustrations lie that
way.</p>
<h2 class="western"><a name="how-critical-is-quality-control"></a>67.2
How critical is quality control?</h2>
<p class="first-paragraph">The more unknowns about the genome under
study, the more critical it is to correct any errors.</p>
<p>When aligning against well-studied and known genomes, we can
recognize and identify errors by their alignments. When assembling a
de-novo genome, mistakes can derail the process; hence, it is more
important to apply a higher stringency for filtering.</p>
<h2 class="western"><a name="what-can-go-wrong-with-the-data"></a>67.3
What can go wrong with the data?</h2>
<p class="first-paragraph">Probably the greatest challenge of QC is
our inability to foresee all the possible ways that data can be
affected. Good data are all alike, but once things go awry, the
potential problems that may need correcting can run a whole gamut:
from puzzling to fascinating to outright disturbing.</p>
<p>For some people, performing quality control is also somewhat of a
psychological barrier. Many feel that they should only move to the
next step after they have done everything possible to improve the
data. They may run, rerun, alter the data, and spend weeks and months
on it. Don&rsquo;t be one of these people:</p>
<p class="block-text-western">Data improvement via quality control is
an incremental process with ever-diminishing returns. Once simple
corrections have been made, no amount of tweaking will produce
radically better outcomes.</p>
<p class="first-paragraph">Besides, a common pitfall is to keep
tweaking QC in a manner that seems to &ldquo;improve&rdquo; the final
results; the danger with this is <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Overfitting">overfitting</a></font>
- making the data match the desired outcome.</p>
<h2 class="western"><a name="when-do-we-perform-quality-control"></a>67.4
When do we perform quality control?</h2>
<p class="first-paragraph">Quality control may be performed at
different stages</p>
<ul>
	<li/>
<p class="compact">Pre-alignment: &ldquo;raw data&rdquo; - the
	protocols are the same regardless of what analysis will follow</p>
	<li/>
<p class="compact">Post-alignment: &ldquo;data filtering&rdquo;
	- the protocols are specific to the analysis that is performed.</p>
</ul>
<h2 class="western"><a name="how-do-we-perform-quality-control"></a>67.5
How do we perform quality control?</h2>
<p class="first-paragraph">QC typically follows this process.</p>
<ol>
	<li/>
<p class="compact">Evaluate (visualize) data quality.</p>
	<li/>
<p class="compact">Stop and move to the next analysis step if
	the quality appears to be satisfactory.</p>
	<li/>
<p class="compact">If not, execute one or more data altering
	steps then go to step 1.</p>
</ol>
<h2 class="western"><a name="how-reliable-are-qc-tools"></a>67.6 How
reliable are QC tools?</h2>
<p class="first-paragraph">First, let&rsquo;s me start with an
observation that might blow your mind. Are you ready? Here it comes.
Most quality control tools are of very low software quality
themselves.</p>
<p>There was an era (one that may not even be over yet) when writing
a high-throughput sequencing quality control tool seemed to be &ldquo;in
fashion&rdquo; because such tools were considered &ldquo;easy&rdquo;
to publish. A literature search may turn up dozens of such software
publications. Unfortunately, in our experience, a majority of these
tools are na&iuml;ve solutions that perform surprisingly poorly in
real-world scenarios.</p>
<p>It also doesn&rsquo;t help that there is a surprising amount of
inconsistency in the way QC tools work. Even though the concepts that
the tools implement appear to be well defined, we have never been
able to produce the same outcomes when employing different tools with
seemingly identical algorithms.</p>
<p>Finally, the truth of the matter is that an objective assessment
of the data quality is inherently tricky. What you will often see and
hear are subjective, almost hand-waving assessments of what good data
&ldquo;looks like&rdquo;. Beyond that, there are few objective
measures, standards, and recommendations that we can test against.</p>
<h2 class="western"><a name="can-quality-control-introduce-new-problems"></a>
67.7 Can quality control introduce new problems?</h2>
<p class="first-paragraph">This might be the second mind-blowing
observation we make in this same chaper - one that just about every
bioinformatician seems to conveniently forget.</p>
<p>Remember - every quality control process introduce errors. We
dearly hope that the impact of the new errors that the QC creates to
be much smaller than the impact of errors that the QC corrects. But
see how at some point it is almost a philosophical issue - would you
rather be dealing with errors that an instrument caused or mistakes
that you introduced when correcting for the machine errors?</p>
<p>Do not error correct data that does not need it.</p>
<h2 class="western"><a name="is-there-a-list-of-qc-tools"></a>67.8 Is
there a list of QC tools?</h2>
<p class="first-paragraph">A considerable number of QC tools have
been published. We have only tested a few of these and out of those
we recommend <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Trimmomatic</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BBDuk</font></font>
,<font face="Consolas, serif"><font size="2" style="font-size: 11pt">flexbar</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cutadapt</font></font>.
While all tools implement basic quality control methods, each often
includes unique functionality specific to that tool.</p>
<p>Quality control tools often provide more utility, some are full
sequence manipulation suites. Here is a list (in alphabetical order)
of tools that we have had some experiences with:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/">BBDuk</a></font>
	part of the <font color="#4f81bd"><a href="https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/">BBMap</a></font>
	package</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://code.google.com/p/biopieces/">BioPieces</a></font>
	a suite of programs for sequence preprocessing</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://code.google.com/p/cutadapt/">CutAdapt</a></font>
	application note in <font color="#4f81bd"><a href="http://journal.embnet.org/index.php/embnetjournal/article/view/200">Embnet
	Journal, 2011</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://code.google.com/p/ea-utils/">fastq-mcf</a></font>
	published in <font color="#4f81bd"><a href="http://benthamscience.com/open/openaccess.php?tobioij/articles/V007/1TOBIOIJ.htm">The
	Open Bioinformatics Journal, 2013</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://hannonlab.cshl.edu/fastx_toolkit/">Fastx
	Toolkit: collection of command line tools for Short-Reads
	FASTA/FASTQ files preprocessing</a></font> - one of the first tools</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://sourceforge.net/projects/flexbar/">FlexBar,
	Flexible barcode and adapter removal</a></font> published in
	<font color="#4f81bd"><a href="http://www.mdpi.com/2079-7737/1/3/895">Biology,
	2012</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.nipgr.res.in/ngsqctoolkit.html">NGS
	Toolkit</a></font> published in <font color="#4f81bd"><a href="http://www.plosone.org/article/info:doi/10.1371/journal.pone.0030619">Plos
	One, 2012</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://prinseq.sourceforge.net/">PrinSeq</a></font>
	application note in <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/pubmed/21278185">Bioinformatics,
	2011</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/vsbuffalo/scythe">Scythe</a></font>
	a bayesian adaptor trimmer</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/jstjohn/SeqPrep">SeqPrep</a></font>
	- a tool for stripping adaptors and/or merging paired reads with
	overlap into single reads.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/104320/">Skewer:
	a fast and accurate adapter trimmer for next-generation sequencing
	paired-end reads.</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://tagcleaner.sourceforge.net/">TagCleaner</a></font>
	published in <font color="#4f81bd"><a href="http://www.biomedcentral.com/1471-2105/11/341">BMC
	Bioinformatics, 2010</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://genome.gsc.riken.jp/osc/english/software/">TagDust</a></font>
	published in <font color="#4f81bd"><a href="http://www.ncbi.nlm.nih.gov/pubmed/19737799">Bioinformatics,
	2009</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">Trim
	Galore</a></font> - a wrapper tool around Cutadapt and FastQC to
	consistently apply quality and adapter trimming to FastQ files, with
	some extra functionality for MspI-digested RRBS-type (Reduced
	Representation Bisufite-Seq) libraries</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.usadellab.org/cms/index.php?page=trimmomatic">Trimmomatic</a></font>
	application note in <font color="#4f81bd"><a href="http://nar.oxfordjournals.org/content/40/W1.toc">Nucleic
	Acid Research, 2012, web server issue</a></font></p>
</ul>
<p class="first-paragraph">There are also many libraries via R
(Bioconductor) for QC such as: <font color="#4f81bd"><a href="http://www.bioinfo.uh.edu/PIQA/">PIQA</a></font>,
<font color="#4f81bd"><a href="http://www.bioconductor.org/packages/2.12/bioc/html/ShortRead.html">ShortRead</a></font></p>
<h2 class="western"><a name="how-does-read-quality-trimming-work"></a>
67.9 How does read quality trimming work?</h2>
<p class="first-paragraph">Historically, for early instruments, the
reliability of sequencing decreased sharply along the read. A typical
correction was to work backward from the end of each read and remove
low-quality measurements from it. The read shortening process is
called &ldquo;trimming&rdquo;.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files -X 10000 SRR1553607 </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run trimmomatic in single end mode</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic
SE SRR1553607_2.fastq trimmed_2.fq SLIDINGWINDOW:4:30</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate fastqc reports on both datasets.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
SRR1553607_1.fastq trimmed_2.fq </font></font>
</p>
<p class="first-paragraph">The resulting plot shows the original file
on the left and the quality-trimmed version on the right.</p>
<p><img src="0000_html_e0e8e89a3a4214dc.png" name="Image121" align="bottom" hspace="12" width="560" height="202" border="0"/>
</p>
<p>A similar (but not identical) result would be found using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbduk</font></font>
to trim adapters:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbduk.sh
in=SRR1553607_2.fastq out=bbduk.fq qtrim=r overwrite=true qtrim=30</font></font></p>
<p class="first-paragraph">To perform the same process on both files
at the same time, we need a more complicated command.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic
PE SRR1553607_1.fastq SRR1553607_2.fastq  trimmed_1.fq unpaired_1.fq
trimmed_2.fq unpaired_2.fq SLIDINGWINDOW:4:30</font></font></p>
<p class="first-paragraph">Or with BBDuk:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbduk.sh
in1=SRR1553607_1.fastq in2=SRR1553607_2.fastq outm1=bbduk_1.fq
out1=unpaired_bb_1.fq outm2=bbduk_2.fq out2=unpaired_bb_2.fq qtrim=r
overwrite=true qtrim=30</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequencing-adapter-trimming"></a>68
Sequencing adapter trimming</h1>
<h2 class="western"><a name="what-are-sequencing-adapters"></a>68.1
What are sequencing adapters?</h2>
<p class="first-paragraph">During library preparation, uniquely
designed DNA adapters of lengths that are typically over 30 bases are
added to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5'</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3'</font></font>
ends of each sequence.</p>
<p>Hence, after adding the adapters, the single-stranded sequences
that make it into the instrument will be of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXAAAATTTTGGGGCCCCYYYY</font></font></p>
<p class="first-paragraph">where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXX</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YYYY</font></font>
are the sequencing adapters. The instrument can recognize the
starting adapters but usually won&rsquo;t detect the end adapters if
these make it into the measurement. When the read length exceeds the
length of the DNA fragment, the adapter sequence may appear in the
data.</p>
<h2 class="western"><a name="how-do-i-visualize-the-sequencing-adapters"></a>
68.2 How do I visualize the sequencing adapters?</h2>
<p class="first-paragraph">FastQC will detect adapters based on
Illumina sequences and will display them in the following manner:</p>
<p><img src="0000_html_ee4104cbc17cbced.png" name="Image122" align="bottom" hspace="12" width="560" height="294" border="0"/>
</p>
<p>The rising line indicates the number of adapter sequences
contained at that base.</p>
<h2 class="western"><a name="can-we-customize-the-adapter-detection"></a>
68.3 Can we customize the adapter detection?</h2>
<p class="first-paragraph">Yes. FASTQC has files that specify the
adapters to search for.</p>
<p>The location of the adapter file depends on the installation (and
version number) but would be somewhere like
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/miniconda3/envs/bioinfo/opt/fastqc-0.11.8/Configuration/adapter_list.txt</font></font></p>
<p>The default content of the adapter file is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Illumina
Universal Adapter                              AGATCGGAAGAG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Illumina
Small RNA 3' Adapter                           TGGAATTCTCGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Illumina
Small RNA 5' Adapter                           GATCGTCGGACT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Nextera
Transposase Sequence                            CTGTCTCTTATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SOLID
Small RNA Adapter                                 CGCCTTGGCCGT</font></font></p>
<p class="first-paragraph">Adding your adapters into that list will
allow you to have them displayed in the FASTQC plot.</p>
<h2 class="western"><a name="why-do-we-need-to-trim-adapters"></a>68.4
Why do we need to trim adapters?</h2>
<p class="first-paragraph">As you recall, within the sequencing
instrument, each read has artificial sequences attached to its end.
If the sequence is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAAAAAAAAA</font></font>
and the adapter is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTTT</font></font>
then depending on the library size and read length the sequencing may
run into the adapter.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAAAAAAAAAATTTTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">----------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">---------------&gt;</font></font></p>
<p class="first-paragraph">The last read will have the adapter
sequence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTT</font></font>
at the end.</p>
<p>If the overlap of bases into the adapter is sufficiently long
(5-6), we can detect the adapter itself and cut the sequence at the
start of the adapter.</p>
<p>To recognize the adapters, we need to know what their sequence is.
This information is surprisingly difficult to find out from those
that put these adapters on. Most companies consider these as &ldquo;trade
secrets&rdquo; and you would be hard pressed to find clear
documentation of it. When found, it is usually buried deep in other
information and needs to be fished out from it. For example, the
Illumina TruSeq index adapters will have the following sequence:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
TruSeq Indexed
Adapter</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTG</font></font></p>
<p class="first-paragraph">To complicate matters even more, an
additional <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
is ligated to the end of the sequences before the adapter is
attached. Hence, the artificial sequence present at the end of some
reads will typically be some fraction of the sequence:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC</font></font></p>
<p class="first-paragraph">To see other adapters that for example the
FastQC tool detects see:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
See configuration files (your version number may be different)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l ~/miniconda3/envs/bioinfo/opt/fastqc-0.11.5/Configuration/</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The adapter configuration file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
~/miniconda3/envs/bioinfo/opt/fastqc-0.11.5/Configuration/adapter_list.txt</font></font></p>
<p class="first-paragraph">A full list of adapters may be found on
the data webpage at:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/data/sequencing-adapters.fa">http://data.biostarhandbook.com/data/sequencing-adapters.fa</a></font></p>
</ul>
<h2 class="western"><a name="how-do-we-trim-adapters"></a>68.5 How do
we trim adapters?</h2>
<p class="first-paragraph">You can create your adapter or use the
ones that come with Trimmomatic. Let&rsquo;s create an adapter format
for Illumina instrument</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;&gt;illumina&quot; &gt; adapter.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC&quot; &gt;&gt; adapter.fa</font></font></p>
<p class="first-paragraph">Use this to filter another SRA file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR519926</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
SRR519926_1.fastq </font></font>
</p>
<p class="first-paragraph">Trim adapters with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic
SE SRR519926_1.fastq output.fq ILLUMINACLIP:adapter.fa:2:30:5</font></font></p>
<p class="first-paragraph">The left plot shows the adapter content of
the data before removing the adapters. The right-hand side shows the
effect of the command above.</p>
<p><img src="0000_html_2deaa545b27e5d0c.png" name="Image123" align="bottom" hspace="12" width="560" height="200" border="0"/>
</p>
<p>With <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbduk</font></font>,
the trimming would be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbduk.sh
in=SRR519926_1.fastq  ref=adapter.fa out=bbduk.fq</font></font></p>
<h2 class="western"><a name="how-do-we-perform-multiple-steps-at-once"></a>
68.6 How do we perform multiple steps at once?</h2>
<p class="first-paragraph">Most tools allow us to chain together
multiple actions, though usually the resulting command line is quite
an eyeful.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic
PE SRR519926_1.fastq SRR519926_2.fastq trimmed_1.fq unpaired_1.fq
trimmed_2.fq unpaired_2.fq SLIDINGWINDOW:4:30 TRAILING:30
ILLUMINACLIP:adapter.fa:2:30:5 </font></font>
</p>
<p class="first-paragraph">It is important to recognize that the
order of action matters and results may be different depending on
which action is performed first:</p>
<ol>
	<li/>
<p class="compact">Do we first trim for quality then cut the
	adapters?</p>
	<li/>
<p class="compact">Do we cut the adapters first and then trim
	for quality?</p>
</ol>
<p class="first-paragraph">Since the adapters have to overlap if we
trim by quality first, we may run the risk of removing a base that
will not allow us to detect the adapter. On the other hand, if the
base is of low quality, it may not align to the adapter; hence, the
adapter won&rsquo;t be detected. Often the order of operation does
not make a notable difference, but on problematic data it&rsquo;s
best to experiment.</p>
<h2 class="western"><a name="do-we-have-to-remove-adapters-from-data"></a>
68.7 Do we have to remove adapters from data?</h2>
<p class="first-paragraph">Adapter read-through problems may not need
to be dealt with explicitly if the genomes that the data originated
from are known.</p>
<p>Many high-throughput aligners can automatically clip the
alignments and trim off adapters during the alignment phase.</p>
<p>On the other hand, the presence of adapter sequences may cause
substantial problems when assembling new genomes or transcriptomes;
hence they should be removed before starting that process.</p>
<p>There is no full consensus on the matter, although Brian Bushnell
makes a strong case that adapter trimming dramatically improves the
quality of data.</p>
<p>Read on the Biostar question of the day <font color="#4f81bd"><a href="https://www.biostars.org/p/212136/">Trimming
adapter sequences - is it necessary?</a></font></p>
<h2 class="western"><a name="how-do-i-cut-a-different-adapter"></a>68.8
How do I cut a different adapter?</h2>
<p class="first-paragraph">Depending on the library preparation and
experiment design other artificial sequences may be attached at the
end of a DNA fragment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 SRR1553606 --split-files </font></font>
</p>
<p class="first-paragraph">The FastQC report will show Nextera
adapter contamination:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
SRR1553606-1.fastq</font></font></p>
<p class="first-paragraph">Verify what FastQC can detect to begin
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
~/miniconda3/envs/bioinfo/opt/fastqc-0.11.5/Configuration/adapter_list.txt</font></font></p>
<p class="first-paragraph">Create the adapter that you need to cut
(this may need searching for information ) for convenience here is
the Nextera adapter:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;&gt;nextera&quot; &gt; nextera.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;CTGTCTCTTATACACATCTCCGAGCCCACGAGAC&quot; &gt;&gt; nextera.fa</font></font></p>
<p class="first-paragraph">Trim the sequences with this new adapter.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequence-duplication"></a>69 Sequence
duplication</h1>
<h2 class="western"><a name="what-is-sequence-duplication"></a>69.1
What is sequence duplication?</h2>
<p class="first-paragraph">Duplication usually means having the
identical sequences in the data. Depending on the expected coverages
duplicates may be correct measurements or errors.</p>
<p>Right off the bat the word &ldquo;duplicate&rdquo; is misleading;
it seems to imply that there is &ldquo;two&rdquo; of something, what
it really means is &ldquo;more than one&rdquo;. No wonder there is a
surprising amount of confusion of what duplication is and most
importantly what the appropriate action is when you observe it.</p>
<p>Duplicates fall into two classes:</p>
<ol>
	<li/>
<p class="compact">Natural duplicates - these were identical
	fragments present in the sample. These should not be removed.</p>
	<li/>
<p class="compact">Artificial copies - these are produced
	artificially during the sequencing process, PCR amplification,
	detection errors. These should be removed.</p>
</ol>
<p class="first-paragraph">But how can we distinguish between the two
cases? As it turns out the process is fraught with challenges.</p>
<h2 class="western"><a name="how-do-we-detect-sequence-duplication"></a>
69.2 How do we detect sequence duplication?</h2>
<p class="first-paragraph">There are two main approaches:</p>
<ol>
	<li/>
<p class="compact">Sequence identity - where we find and
	remove sequences that have the exact sequence.</p>
	<li/>
<p class="compact">Alignment identity - where we find and
	remove sequences that align the same way.</p>
</ol>
<p class="first-paragraph">As counterintuitive as that might sound,
by filtering on sequence identity, we are running the risk of
rewarding reads with errors in them. We are removing identical reads,
yet keep those reads that were supposed to be equal but were
mismeasured. There is a danger of actually enriching our data with
incorrect measurements.</p>
<p>On the other hand, alignment-based identity is only feasible if
there is a known genome to align the data against.</p>
<p>There is a new class of methods that detect subsequence
composition (called k-mers) that can be used to correct errors.</p>
<h2 class="western"><a name="what-is-the-primary-concern-with-duplicates"></a>
69.3 What is the primary concern with duplicates?</h2>
<p class="first-paragraph">The adverse effect of read duplication
manifests itself mostly during variation detection. This process
assigns a reliability score to each variant based on the number of
times it has been observed. If a single read with a variant happens
to be duplicated artificially, it runs the risk of producing a
seemingly more reliable result than in reality.</p>
<h2 class="western"><a name="should-we-remove-duplicates"></a>69.4
Should we remove duplicates?</h2>
<p class="first-paragraph">For SNP calling and genome variation
detection, the answer is usually yes. For other processes, the answer
is often no.</p>
<p>Duplicate removal is a substantial alteration of the data - one
should have additional evidence that none of the duplicates are of
natural origin before doing it.</p>
<p>As always the severity of the problem and whether it impacts the
study should be evaluated and weighted.</p>
<h2 class="western"><a name="what-does-the-fastqc-duplicate-plot-mean"></a>
69.5 What does the FastQC duplicate plot mean?</h2>
<p class="first-paragraph">It is not all that simple to interpret the
FastQC duplication report. We have struggled with it.</p>
<p>First of all, that small number circled at the top is one of the
most important values. It shows what percent of the data is distinct.</p>
<p><img src="0000_html_2e8eaefa397c702e.png" name="Image124" align="bottom" hspace="12" width="560" height="413" border="0"/>
</p>
<ul>
	<li/>
<p class="compact">Biostar question of the day: <font color="#4f81bd"><a href="https://www.biostars.org/p/107402/">Revisiting
	the FastQC read duplication report</a></font></p>
</ul>
<p class="first-paragraph">There are two kinds of duplicate counts:
one for all sequences, and another for distinct sequences.</p>
<p>As an example, take a situation where two sets would have to be
built as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
A A A B B B B</font></font></p>
<p class="first-paragraph">and</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
A A A A A A B</font></font></p>
<p class="first-paragraph">Both examples have only two distinct
sequences: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
and the duplication rate would be the same. We have two distinct
options and eight observations: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2/8</font></font>,
But obviously, the sets&rsquo; internal structure is entirely
different. We should not use the same measure to quantify the
duplication of the first set as the second.</p>
<p>This set structure is what the blue and red lines try to capture
and visualize for us.</p>
<ul>
	<li/>
<p class="compact">The blue lines show the percent (Y-axis) of
	<b>all sequences</b> that are duplicated at a given rate (X-axis).</p>
	<li/>
<p class="compact">The red line indicates the number of
	<b>distinct sequences</b> that are replicated at a given frequency.</p>
</ul>
<p class="first-paragraph">To work this out explicitly. In the first
example, we have two sequences each duplicated at a rate of 4,
whereas in the second case we have one unique sequence and one
sequence reproduced at the rate of 7.</p>
<h2 class="western"><a name="how-do-i-remove-duplicates"></a>69.6 How
do I remove duplicates?</h2>
<p class="first-paragraph">In general, avoid removing duplicated
sequences by sequence identify alone unless you have good reasons to
do so. Once you have an alignment file (BAM that we will cover
later), you may use the <font color="#4f81bd"><a href="https://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates">picard
MarkDuplicates</a></font> tool to mark (label) duplicated alignments.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="advanced-quality-control"></a>70
Advanced quality control</h1>
<h2 class="western"><a name="are-there-different-types-of-quality-control"></a>
70.1 Are there different types of quality control?</h2>
<p class="first-paragraph">Yes. Up to this point, we only covered the
QC of the so-called &ldquo;raw&rdquo; (original) data that is
produced directly by the instrument.</p>
<p>But quality control may be performed later in the process as well.
Often though at that point it typically becomes a different type of
tasks, more of a quality assessment methods to answer questions of
&ldquo;How well does the data follow expected characteristics?&rdquo;.
For example, does the RNA-Seq data align to transcripts as we expect
it to?</p>
<p>RNA-Seq analysis for that matter has two (quite unfortunately
named) software packages, which you will, like us, permanently
confuse and conflate:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://rseqc.sourceforge.net/">RSeQC</a></font><br/>
<br/>
<br/>

	</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://archive.broadinstitute.org/cancer/cga/rna-seqc">RNA-SeqC</a></font></p>
</ul>
<p class="first-paragraph">Author&rsquo;s note: By the way, if you
ever end up writing a QC software tool, please do us all a service
and find a memorable and simple name for it. Call it <font face="Consolas, serif"><font size="2" style="font-size: 11pt">speedyQC</font></font>,
call it <font face="Consolas, serif"><font size="2" style="font-size: 11pt">monsterQC</font></font>
call it <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sevenQC</font></font>,
but please, please don&rsquo;t make it an awkward variation of an
existing, already awkward name. While we are on this topic, here are
more suggestions on to avoid:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/67874/">Crac:
	Funny And Weird Names For Bioinformatics Tools</a></font></p>
</ul>
<h2 class="western"><a name="how-to-combine-the-results-into-a-single-report."></a>
70.2 How to combine the results into a single report.</h2>
<p class="first-paragraph">An inconvenience of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>
(and many other tools) is that each report goes into a separate,
standalone file. For more extensive studies that may involve dozens
or even hundreds of samples, consulting each of these reports becomes
tedious to impossible.</p>
<p>The <font color="#4f81bd"><a href="http://multiqc.info/">MultiQC</a></font>
tool is designed to facilitate combining different reports into a
single one.</p>
<p class="block-text-western">Aggregate results from bioinformatics
analyses across many samples into a single report</p>
<p class="block-text-western">MultiQC searches a given directory for
analysis logs and compiles an HTML report. It&rsquo;s a general use
tool, perfect for summarising the output from numerous bioinformatics
tools.</p>
<p class="first-paragraph">At the time of writing this document
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
supported the integration of reports generated by 51 different tools.
The documentation of <font color="#4f81bd"><a href="http://multiqc.info/">MultiQC</a></font>
reads almost like a table of contents describing the tools and
techniques you could use to assess the quality of your data. Our
original installation did not include this package since it has many
requirements thus you will have to install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
yourself. Thankfully there is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
package for it.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install multiqc</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font>
</p>
<p class="first-paragraph">Get and unpack the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/data/sequencing-platform-data.tar.gz
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
zxvf sequencing-platform-data.tar.gz </font></font>
</p>
<p class="first-paragraph">Run fastqc on two datasets to compare the
Illumina platform to that of Iontorrent&rsquo;s. The PacBio data is
so radically different that is probably not worth putting on the same
plot. The command below will create two <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>
reports and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--extract</font></font>
flags instructs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>
to keep the data directories for each report:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc
--extract illumina.fq iontorrent.fq </font></font>
</p>
<p class="first-paragraph">Now <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
can combine the data from the three data report directories into one:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc
illumina_fastqc iontorrent_fastqc </font></font>
</p>
<p class="first-paragraph">The resulting plot shows us a clear
comparison of the two runs:</p>
<p><img src="0000_html_54c9b7ec4b6da17c.png" name="Image125" align="bottom" hspace="12" width="560" height="371" border="0"/>
</p>
<p>As you can imagine when you have many sequencin runs to summarize
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
is quite the handy tool.</p>
<h2 class="western"><a name="why-did-my-multiqc-fail-with-an-error"></a>
70.3 Why did my <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
fail with an error?</h2>
<p class="first-paragraph">Here is a tale of caution when it comes to
bioinformatics tools.</p>
<p>The very next day after this guide was published readers started
reporting that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
did not work at all. Even just by invoking the tool <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
they got errors such as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TypeError:
add_edge() takes 3 positional arguments but 4 were given</font></font></p>
<p class="first-paragraph">What happened is that a new version of the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>
software was released, one that changed the way the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">add_edge</font></font>
method call works. Now <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
does not use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>
directly. But is uses a library called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">spectra</font></font>
that in turn uses a library called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">colormath</font></font>
that in turn uses <font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>.
So there is a &ldquo;network&rdquo; (pun intended) of dependencies in
play.</p>
<p>Then there is a flaw in the way the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">colormath</font></font>
library is installed. It does not &ldquo;pin&rdquo; the library to a
version of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>.
The standard and proper software engineering practice is to specify
not just a dependency but also the version of it that is proven to
work correctly. Now in the meantime the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">colormath</font></font>
library also added code to support both versions of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>
the library. Alas the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">colormath</font></font>
library installed with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font>
is a prior version and not the latest that includes the fixes.</p>
<p>Hence the two compounding problems lead to a complete failure to
run the program. Eventually, once sufficient number of people
complain the problem will be fixed. By the time you read this the
problem might not exist. But it exhibits an acute problem that will
re-occur under different circumstances. We, scientists, usually don&rsquo;t
recognize the importance of proper software maitenance practices.</p>
<p>In the meantime if you experience this problem you may downgrade
to a previous version of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">networkx</font></font>
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">source
activate multiqc</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install networkx=1.11</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font>
</p>
<h2 class="western"><a name="what-else-is-multiqc-capable-of"></a>70.4
What else is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">multiqc</font></font>
capable of?</h2>
<p class="first-paragraph">If you need to process more than a few
(2-3 datasets) we recommend that you consult the <font color="#4f81bd"><a href="http://multiqc.info/">multqc
documentation</a></font> and look for ways it can help you interpret
the results.</p>
<h2 class="western"><a name="would-we-ever-want-to-toss-out-some-of-the-good-quality-data"></a>
70.5 Would we ever want to toss out some of the good quality data?</h2>
<p class="first-paragraph">Surprising as it sounds too much data can
make processes operate worse. A saturation of information can take
place, after which the errors get magnified.</p>
<p>In those cases, you could, perhaps, choose to select a random
subset of the entire data and that may work well if your data happens
to cover the DNA evenly. However, the typical situation can be more
complicated, where some regions may be extremely highly covered,
while and others still be under shallow coverage. For some studies,
such as genome assembly, you would want to remove data from the high
coverage regions while keeping the low coverages intact.</p>
<p>So you&rsquo;d have to apply some downsampling but without knowing
beforehand where each measurement belongs. This seemingly impossible
task can be accomplished by an algorithm that keeps track of the
frequency of small patterns in each sequencing read. In a nutshell,
reads built from short patterns that have already been observed many
times over can be ignored.</p>
<p>Some tools internally implement such techniques, for example, the
Trinity transcriptome assembler will downsample the data. In other
cases, external tools such as <font color="#4f81bd"><a href="http://khmer.readthedocs.io/">khmer</a></font>
may be used for this purpose.</p>
<h2 class="western"><a name="can-i-combine-reads-into-a-single-sequence"></a>
70.6 Can I combine reads into a single sequence?</h2>
<p class="first-paragraph">Under certain circumstances, for paired
end sequencing, when the DNA fragment is shorter than the sum of the
read lengths the reads may overlap:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">------read1------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">========
 DNA  =======</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     &lt;----read2------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font>
</p>
<p class="first-paragraph">The two opposing reads that have an
overlapping region may be combined using the common regions:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">------read1------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     |||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     &lt;----read2------</font></font></p>
<p class="first-paragraph">To produce a single, longer and more
accurate read:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">---------------------&gt;</font></font></p>
<p class="first-paragraph">This process is called &ldquo;merging.&rdquo;</p>
<h2 class="western"><a name="when-should-i-merge-the-reads"></a>70.7
When should I merge the reads?</h2>
<p class="first-paragraph">Combining reads only makes sense when the
majority of the reads overlap. Otherwise, you&rsquo;ll end up with a
mixed data of overlapped single-end reads, and non-overlapping
paired-end reads - only adding to the complexity of the subsequent
tasks.</p>
<h2 class="western"><a name="how-do-i-merge-reads"></a>70.8 How do I
merge reads?</h2>
<p class="first-paragraph">In Biostar Post of the Day: <font color="#4f81bd"><a href="https://www.biostars.org/p/225683/">Tools
to merge overlapping paired-end reads</a></font>, Rafay Khan writes:</p>
<p class="block-text-western">In very simple terms, current
sequencing technology begins by breaking up long pieces of DNA into
lots more short pieces of DNA. The resultant set of DNA is called a
&ldquo;library&rdquo; and the short pieces are called &ldquo;fragments&rdquo;.
Each of the fragments in the library are then sequenced individually
and in parallel. There are two ways of sequencing a fragment - either
just from one end, or from both ends of a fragment. If only one end
is sequenced, you get a single read. If your technology can sequence
both ends, you get a &ldquo;pair&rdquo; of reads for each fragment.
These &ldquo;paired-end&rdquo; reads are standard practice on
Illumina instruments like the GAIIx, HiSeq, and MiSeq.</p>
<p class="block-text-western">Now, for single-end reads, you need to
make sure your read length (L) is shorter than your fragment length
(F) or otherwise the sequence will run out of DNA to read! Typical
Illumina fragment libraries would use F ~ 450bp, but this is
variable. For paired-end reads, you want to make sure that F is long
enough to fit two reads. This means you need F to be at least 2L. As
L=100 or 150bp these days for most people, using F~450bp is fine,
there is a still a safety margin in the middle.</p>
<p class="block-text-western">However, some things have changed in
the Illumina ecosystem this year. Firstly, read lengths are now
moving to &gt;150bp on the HiSeq (and have already been on the GAIIx)
and to &gt;250bp on the MiSeq, with possibilities of longer ones
coming soon! This means that the standard library size F~450bp has
become too small, and paired-end reads will overlap. Secondly, the
new enzymatic Nextera library preparation system produces a wide
spread of F sizes compared to the previous TruSeq system. With
Nextera, we see F ranging from 100bp to 900bp in the same library. So
some reads will overlap, and others won&rsquo;t. It&rsquo;s starting
to get messy.</p>
<p class="block-text-western">The whole point of paired-end reads is
to get the benefit of longer reads without actually being able to
sequence reads that long. A paired-end read (two reads of length L)
from a fragment of length F, is a bit like a single-read of length F,
except a bunch of bases in the middle of it are unknown, and how many
of them there are is only roughly known (as libraries are only
nominally of length F, each read will vary). This gives the reads a
more extended context, and this mainly helps in de novo assembly and
in aligning more reads unambiguously to a reference genome. However,
many software tools will get confused if you give them overlapping
pairs, and if we could overlap them and turn them into longer
single-end reads, many tools will produce better results, and faster.</p>
<p class="first-paragraph">Multiple tools are listed in the post
above. One of the most popular choices <font color="#4f81bd"><a href="http://ccb.jhu.edu/software/FLASH/">FLASH
(Fast Length Adjustment of SHort reads)</a></font> is a tool to merge
paired-end reads that were generated from DNA fragments whose lengths
are shorter than twice the length of reads. Merged read pairs result
in unpaired longer reads. Longer reads are more desired in genome
assembly and genome analysis processes.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate bioinfo</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install flash -y</font></font></p>
<p class="first-paragraph">Get the 1% error test data from the tool
website (there is 1%, 2%, 3%, and 5% error test data as well, give
those a go!):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get an example dataset.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
https://ccb.jhu.edu/software/FLASH/error1.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Unpack the data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
zxvf error1.tar.gz</font></font></p>
<p class="first-paragraph">This dataset now contains <font face="Consolas, serif"><font size="2" style="font-size: 11pt">frag1.fq</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">frag2.fq</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">flash
frag_1.fastq  frag_2.fastq</font></font></p>
<p class="first-paragraph">The tools create some ancillary files,
while reporting:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[FLASH]
Read combination statistics:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[FLASH]
    Total pairs:      1000000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[FLASH]
    Combined pairs:   701191</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[FLASH]
    Uncombined pairs: 298809</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[FLASH]
    Percent combined: 70.12%</font></font></p>
<p class="first-paragraph">Note how at 1% error rate only 70% of the
data could be merged.</p>
<p>You may also use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmerge.sh</font></font>
to perform a merge:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmerge.sh
in1=frag_1.fastq  in2=frag_2.fastq out=merged.fq</font></font></p>
<p class="first-paragraph">that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Pairs:
         1000000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Joined:
         310150          31.015%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Ambiguous:
      689850          68.985%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">No
Solution:          0           0.000%</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Too
Short:            0           0.000%</font></font></p>
<h2 class="western"><a name="how-well-do-mergers-work"></a>70.9 How
well do mergers work?</h2>
<p class="first-paragraph">A detailed <font color="#4f81bd"><a href="http://seqanswers.com/forums/showthread.php?t=43906">guide
to validating merged data</a></font> was written by <b>B</b>rian
<b>B</b>ushnell, the author of <font color="#4f81bd"><a href="https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/">BBTools</a></font>
one of a fascinating bioinformatics software suites (and now you also
know where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bb</font></font>
letters come from in: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmap</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbrename</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmerge</font></font>
etc).</p>
<p>We reproduce that guide below in a shortened format.</p>
<p>Get the E.Coli genome.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id MF521836 -format fasta &gt; ecoli.fa</font></font></p>
<p class="first-paragraph">Generate random sequences from the E.Coli
genome with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">randomreads.sh</font></font>
tool:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">randomreads.sh
ref=ecoli.fa reads=100000 paired interleaved out=reads.fq.gz len=150
mininsert=180 maxinsert=360 </font></font>
</p>
<p class="first-paragraph">Rename each read by its length. The names
are used later as the validator will checks that the merged fragment
matches this size. To use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbrename.sh</font></font>
program, you will have to fully install the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbtools</font></font>
package manually.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbrename.sh
in=reads.fq.gz int renamebyinsert out1=r1.fq out2=r2.fq</font></font></p>
<p class="first-paragraph">Merge reads with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">flash</font></font>.
Creates the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">out.extendedFrags.fastq</font></font>
file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">flash
-M 150 -x 0.25 r1.fq r2.fq </font></font>
</p>
<p class="first-paragraph">Validate the output. How many wrong
merges?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grademerge.sh
in=out.extendedFrags.fastq</font></font></p>
<p class="first-paragraph">Prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Correct:
       99.85591%       74153 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Incorrect:
      0.14409%         107 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Too
Short:       0.00943%           7 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Too
Long:        0.13466%         100 reads</font></font></p>
<p class="first-paragraph">Merge the reads with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmerge.sh</font></font>.
Creates the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">merged.fq</font></font>
file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmerge.sh
in1=r1.fq in2=r2.fq out=merged.fq</font></font></p>
<p class="first-paragraph">Validate the output. How many wrong
merges?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grademerge.sh
in=bb_merged.fq</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Correct:
        100.00000%      62346 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Incorrect:
        0.00000%         0 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Too
Short:         0.00000%         0 reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Too
Long:          0.00000%         0 reads</font></font></p>
<p class="first-paragraph">Note how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">flash</font></font>
merged <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11897</font></font>
more reads than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmap</font></font>
but it has also made more mistakes when doing so. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">214</font></font>
reads were merged incorrectly. Which one is better?</p>
<p>Above you can see an interesting example of the tradeoffs and
decisions that you may need to make along the way.</p>
<h2 class="western"><a name="is-there-anything-newer-than-fastqc"></a>
70.10 Is there anything newer than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font>?</h2>
<p class="first-paragraph">For the sake of revisiting QC yearly, we
looked for a new tool for 2017 and found <font color="#4f81bd"><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1469-3">AfterQC:
automatic filtering, trimming, error removing and quality control for
fastq data.</a></font> BMC Bioinformatics 2017 18(Suppl 3):80</p>
<p>Let&rsquo;s remind you what we said before,</p>
<p class="block-text-western">First, let&rsquo;s start with a
mind-blowing observation: Many quality control tools are of meager
software quality.</p>
<p class="first-paragraph">Sure enough, the AfterQC tool requires the
installation of Python 2.7 even though Python 3.0 has been released
nine years ago! Reviewers still accept tools written in a language
version that is now considered obsolete.</p>
<p>Lucky for us, that&rsquo;s why we have environments at our
disposal. While we cannot install afterqc into the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bioinfo</font></font>
environment, we can make a new, Python 2 based environment. Let&rsquo;s
get to it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create --name py2 python=2.7 -y</font></font></p>
<p class="first-paragraph">Activate the new environment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
activate py2</font></font></p>
<p class="first-paragraph">Install AfterQC into this environment</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install afterqc</font></font></p>
<p class="first-paragraph">Run the program on</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">after.py
-1 illumina.fq</font></font></p>
<p class="first-paragraph">Do consult this plot and the documentation
for details.</p>
<h2 class="western"><a name="what-is-error-correction"></a>70.11 What
is error correction?</h2>
<p class="first-paragraph">Data produced by sequencing instruments
may contain errors. The quality value in a FASTQ process is meant to
indicate the reliability of a given measurement.</p>
<p>Since the errors are random, it is unlikely that they will occur
in the same locations. As a consequence, if the data has high
coverage, it is possible to recognize errors by identifying regions
that have many measurements in consensus and the very few
disagreements are likely to be errors.</p>
<p>The error correction thus can be achieved even without knowing the
reference genome by computing the so-called k-mer density of the
data. In simple terms, a k-mer is a short subsequence of size <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>
from the data. k-mers that come from real data will be represented in
a considerable number of times whereas k-mers that contain errors
will be rare.</p>
<p>A FASTQ error correction program will attempt to correct or remove
reads that look to have errors in them. Many firmly believe that
error correction can make some analysis (especially genome assembly)
more efficient.</p>
<p>Also when data is of high coverage, error correction might have a
more pronounced beneficial effect.</p>
<h2 class="western"><a name="how-do-we-correct-errors"></a>70.12 How
do we correct errors?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bbmap</font></font>
package includes the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tadpole.sh</font></font>
error corrector:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR519926</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Tadpole is part of the bbmap package</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tadpole.sh
in=SRR519926_1.fastq out=tadpole.fq mode=correct  out=r1.fq
out2=r2.fq overwrite=true</font></font></p>
<p class="first-paragraph">Error correction with bfc
<font color="#4f81bd"><a href="https://github.com/lh3/bfc">https://github.com/lh3/bfc</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bfc
SRR519926_1.fastq &gt; bfc.fq</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequence-patterns"></a>71 Sequence
patterns</h1>
<p class="first-paragraph">(This section is incomplete, we are
planning to significantly expand it in the future)</p>
<h2 class="western"><a name="what-is-a-sequence-pattern"></a>71.1
What is a sequence pattern?</h2>
<p class="first-paragraph">A sequence pattern is a sequence of bases
described by certain rules. These rules can be as simple as a direct
match (or partial matches) of a sequence, like:</p>
<ul>
	<li/>
<p class="compact">Find locations where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font>
	matches with no more than one error</p>
</ul>
<p class="first-paragraph">Patterns may have more complicated
descriptions such as:</p>
<ul>
	<li/>
<p class="compact">All locations where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATA</font></font>
	is followed by one or more <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font>s
	and no more than five <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>s
	ending with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTA</font></font></p>
</ul>
<p class="first-paragraph">Patterns can be summarized
probabilistically into so called <b>motifs</b> such as:</p>
<ul>
	<li/>
<p class="compact">Locations where a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font>
	is followed by an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
	80% of time, or by a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
	20% of the time, then is followed by another <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font></p>
</ul>
<h2 class="western"><a name="can-adapters-be-identified-by-a-pattern"></a>
71.2 Can adapters be identified by a pattern?</h2>
<p class="first-paragraph">Yes. Adapters in the data are among the
simplest types of sequence pattern that may be present in a data set.</p>
<p>When the our sequences are &ldquo;line oriented&rdquo;, that is
the entire sequence is on a single line, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--color=always</font></font>
parameter will allow us to extract and color the matches:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 SRR1553606 --split-files </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR1553606_1.fastq | grep --color=always CTGTCTCTTATACA | head -2</font></font></p>
<p class="first-paragraph">will produce:</p>
<p><img src="0000_html_339d65679ed7e6a6.png" name="Image126" align="bottom" hspace="12" width="560" height="200" border="0"/>
</p>
<p>Most FASTQ files can be processed with line oriented tools.</p>
<p class="block-text-western">Do not use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--color=always</font></font>
when saving data in a new file for other downstream processing. When
you use coloring, additional color-related data is added to the file,
and that will affect downstream processing.</p>
<h2 class="western"><a name="how-can-i-search-genomic-sequences-for-patterns"></a>
71.3 How can I search genomic sequences for patterns?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>
and extended grep<font face="Consolas, serif"><font size="2" style="font-size: 11pt">egrep</font></font>
are tools that operate on one line at a time and can be used to
identify lines with patterns. When sequences wrap over many lines we
need dedicated tools like <font color="#4f81bd"><a href="http://emboss.sourceforge.net/apps/cvs/emboss/apps/dreg.html">dreg</a></font>
or <font color="#4f81bd"><a href="http://emboss.sourceforge.net/apps/cvs/emboss/apps/fuzznuc.html">fuzznuc</a></font>
(Emboss Tools). See the following chapter [Regular
expressions][regexp.md]</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This will miss patterns that wrap around new lines.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
KU182908.fa | grep  AAAAAA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The dreg tool matches and reports the locations.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
KU182908.fa | dreg -filter -pattern  AAAAAA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Search a pattern with ambiguous N bases.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
KU182908.fa | fuzznuc -filter -pattern 'AANAA'</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="regular-expressions"></a>72 Regular
expressions</h1>
<p class="first-paragraph">(This section is incomplete, we are
planning to significantly expand it in the future)</p>
<p><font color="#4f81bd"><a href="http://emboss.sourceforge.net/apps/cvs/emboss/apps/dreg.html">dreg</a></font>
or <font color="#4f81bd"><a href="http://emboss.sourceforge.net/apps/cvs/emboss/apps/fuzznuc.html">fuzznuc</a></font>
(Emboss Tools) are tools that allow you to search the genome for
complex patterns.</p>
<h2 class="western"><a name="what-are-regular-expressions"></a>72.1
What are regular expressions?</h2>
<p class="first-paragraph">Regular expressions are special sequences
of characters that define a possibly complex search pattern.</p>
<p>For example</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">^</font></font>
	matches the beginning of the line</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
	matches any character</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{m,n}</font></font>
	matches the preceeding elements at least <font face="Consolas, serif"><font size="2" style="font-size: 11pt">m</font></font>
	but not more than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">n</font></font>
	times.</p>
</ul>
<p class="first-paragraph">The Regexp pattern language is unlike most
other computing languages. It can be best learned via an interactive
service like <font color="#4f81bd"><a href="https://regexone.com/">https://regexone.com/</a></font>
or many others where the pattern and its effect are instantly
visualized.</p>
<p>Example regular expression searches:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a FASTQ dataset.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files SRR519926</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find an ATG anchored at the start of the line</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;^ATG&quot; --color=always | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find an ATG anchored at the end of the line</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;ATG\$&quot; --color=always | head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find TAATA or TATTA patterns, this is a range of characters</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;TA[A,T]TA&quot; --color=always |
head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find TAAATA or TACCTA, these are groups of words</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;TA(AA|CC)TA&quot; --color=always |
head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Quantify matches with metacharacters</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find TA followed by zero or or more A followed by TA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;TA(A*)TA&quot; --color=always |
head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find TA followed by one or or more A followed by TA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;TA(A+)TA&quot; --color=always |
head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find TA followed by two to five As followed by TA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;TAA{2,5}TA&quot; --color=always |
head</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Practice RegExp matches on online regexp testers</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
http://regexpal.com/</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Match Ilumina adaptors at the end of the reads</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Match AGATCGG anywhere followed by any number of bases</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR519926_1.fastq | egrep &quot;AGATCGG.*&quot; --color=always | head</font></font></p>
<h2 class="western"><a name="what-are-some-challenges-i-may-face-when-using-regular-expressions"></a>
72.2 What are some challenges I may face when using regular
expressions?</h2>
<p class="first-paragraph">Think of a regular expression as a
computing engine. There are different implementations of it, and each
may operate with slightly different syntax, and at different speeds
or efficiencies.</p>
<p>Get chromosome 22 of the Human Genome</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://hgdownload.cse.ucsc.edu/goldenPath/hg38/chromosomes/chr22.fa.gz
| gunzip &gt; chr22.fa</font></font></p>
<p class="first-paragraph">Let&rsquo;s find the telomeric repeats
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTAGGG</font></font>
in the human genome. In principle this is easy: build and match the
pattern. In practice, when we search at a genomic scale, things don&rsquo;t
always go as planned.</p>
<p>Let&rsquo;s first check that this pattern is present. Use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-i</font></font>
flag to make the match case insensitive since the genome may
lowercase regions (for example the repeating regions are marked as
such in the human genome):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | egrep -i '(TTAGGG)' --color=always</font></font></p>
<p class="first-paragraph">The above won&rsquo;t work perfectly since
this is line oriented matching and the genome wraps over many lines.
But it is a good start. We can refine it more</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | egrep -i '(TTAGGG){3,10}' --color=always</font></font></p>
<p class="first-paragraph">Then we can &ldquo;linearize&rdquo; the
genome by removing the new line wrapping characters from it. That way
we can look for the actual pattern:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | tr -d '\n' | egrep -o -i '(TTAGGG){20,30}' --color=always</font></font></p>
<p class="first-paragraph">To produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttagggttaggg</font></font></p>
<p class="first-paragraph">The same can be done with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dreg</font></font>,
though note that its regular expression engine is less powerful and
may not be able to process all patterns:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
chr22.fa | dreg -filter -pattern '(TTAGGG){20,30}' </font></font>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequence-k-mers"></a>73 Sequence k-mers</h1>
<p class="first-paragraph">(This section is incomplete, we are
planning to significantly expand it in the future)</p>
<h2 class="western"><a name="what-is-a-k-mer"></a>73.1 What is a
k-mer?</h2>
<p class="first-paragraph">A <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mer</font></font>
typically refers to all the possible substrings of length <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>
that are contained in a string. For example if the sequence is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCA</font></font>
then</p>
<ul>
	<li/>
<p class="compact">The 2 base long <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mers</font></font>
	(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2-mers</font></font>)
	are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AT</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TG</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CA</font></font></p>
	<li/>
<p class="compact">The 3 base long <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mers</font></font>
	(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">3-mers</font></font>)
	are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATG</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGC</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCA</font></font></p>
	<li/>
<p class="compact">The 4 base long <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mers</font></font>
	(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">4-mers</font></font>)
	are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGCA</font></font></p>
	<li/>
<p class="compact">The 5 base long <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mer</font></font>
	(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5-mer</font></font>)
	is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCA</font></font></p>
</ul>
<h2 class="western"><a name="what-are-k-mers-good-for"></a>73.2 What
are k-mers good for?</h2>
<p class="first-paragraph">The potential information gained from
k-mers evolving quickly. Interesting applications using k-mers are:</p>
<ul>
	<li/>
<p class="compact">Error correction: rare k-mers are more
	likely to be caused by sequence errors.</p>
	<li/>
<p class="compact">Classification: certain k-mers may uniquely
	identify genomes.</p>
	<li/>
<p class="compact">Pseudo-alignment: new pseudo-aligners can
	match reads to locations based solely on the presence of common
	k-mers.</p>
</ul>
<p class="first-paragraph">Computing k-mers is much faster than
producing alignments &ndash; hence they can massively speed up data
interpretation.</p>
<p>Using the <font color="#4f81bd"><a href="http://www.cbcb.umd.edu/software/jellyfish/">jellyfish</a></font>
k-mer counter:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get some sequence data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-id KU182908 -db nucleotide -format fasta &gt; KU182908.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Count the k-mers up to size 10.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">jellyfish
count -C -m 10 -s10M KU182908.fa </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Show a histogram of k-mers.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">jellyfish
histo mer_counts.jf </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The k-mers present at least 7 times.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">jellyfish
dump -L 7  mer_counts.jf</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Pick one k-mer, say TTAAGAAAAAA - is it present 7 times?</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
KU182908.fa | dreg -filter -pattern TTAAGAAAAA</font></font></p>
<h2 class="western"><a name="should-i-use-the-k-mer-report-produced-by-fastqc"></a>
73.3 Should I use the k-mer report produced by FastQC?</h2>
<p class="first-paragraph">The k-mer plot that FastQC produces should
be interpreted with caution as it can be misleading. Both the scaling
of the plot and the ratios of observed versus expected counts are
flawed. Longer k-mers will be reported many times, but shifted by one
base.</p>
<p class="compact"><br/>
<br/>

</p>
<p>To run the examples install the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package developed for this book.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph"><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-alignments"></a>74
Introduction to alignments</h1>
<h2 class="western"><a name="what-is-a-sequence-alignment"></a>74.1
What is a sequence alignment?</h2>
<p class="first-paragraph">Sequence alignment (specifically pairwise
alignment) means arranging two sequences in a way so that regions
with similarities line up. For example alignment of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA</font></font>
could look like:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||
|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA--</font></font></p>
<p class="first-paragraph">When judged by eye it does not take much
for alignments to make sense intuitively, but be warned that our
intuition and the apparent clarity is grossly misleading.</p>
<p>Our brains are hardwired to see patterns and similarities whenever
possible, and once a similarity is shown, we are often unable to
recognize other alternatives that may be just as good (if not better)
as the one we are currently observing. It is hard to &ldquo;un-see&rdquo;
a pattern. For example, we could have arranged our example words like
this</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||
 ||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAT--CA</font></font></p>
<p class="first-paragraph">or we could have shown them like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||
| ||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GA-T-CA</font></font></p>
<p class="first-paragraph">Now we have two more alignments for the
same words, that, when displayed look to be matching up just as well.
Once you look at one pattern, it may take quite a bit of effort to
notice that you could have also arranged the letters in other ways
that seem just as appropriate.</p>
<h2 class="western"><a name="what-are-alignments-used-for"></a>74.2
What are alignments used for?</h2>
<p class="first-paragraph">Alignments have two primary use cases:</p>
<ol>
	<li/>
<p class="compact">Finding similar regions between two
	sequences - probing the relatedness of sequences.</p>
	<li/>
<p class="compact">Finding which sequence (from many
	alternatives) is most similar to a query (input) sequence</p>
</ol>
<h2 class="western"><a name="what-governs-an-alignment"></a>74.3 What
governs an alignment?</h2>
<p class="first-paragraph">The following two factors determine every
alignment:</p>
<ol>
	<li/>
<p class="compact">Alignment algorithm: global, local,
	semi-global</p>
	<li/>
<p class="compact">Alignment scoring (parameters): the
	numerical values that tune the arrangement</p>
</ol>
<p class="first-paragraph">Thus always remember that:</p>
<ul>
	<li/>
<p class="compact">Different algorithms will typically produce
	different alignments for the same sequences.</p>
	<li/>
<p class="compact">Different scoring will usually generate
	different alignments for the same sequences</p>
</ul>
<h2 class="western"><a name="how-are-alignments-generated"></a>74.4
How are alignments generated?</h2>
<p class="first-paragraph">Suppose you had the three alignments
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA</font></font>
vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA
     GATTACA      GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||.|
       |||  ||      || | ||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA--
     GAT--CA      GA-T-CA</font></font></p>
<p class="first-paragraph">Which one is the &ldquo;best&rdquo;,
&ldquo;right&rdquo;, &ldquo;correct&rdquo;, &ldquo;proper&rdquo;,
&ldquo;meaningful&rdquo; alignment?</p>
<p>What would you consider to be the <i>best</i> alignment?</p>
<p>You see the best alignment will depend on how you &ldquo;value&rdquo;
the way bases line up. Do you consider one mismatch to be less
disruptive than gaps? The <i>value</i> you associate with a match,
mismatch or a gap is called <i>scoring</i>.</p>
<p>The most important concept of alignments is the following:</p>
<p class="block-text-western">There is no <i>universally best</i>
alignment. There is only the <i>best alignment</i> relative to a
<i>scoring choice</i>. Changing the score will usually change what is
selected as the best alignment. Alignment algorithms find the
arrangement that produces the maximal score over the entire
alignment.</p>
<p class="first-paragraph">Sequence alignment algorithms speed up the
process of selecting the alignment with the best score. You can think
of aligners as methods for quickly evaluating all possible
combinations then, out of those picking the alignments with the best
score.</p>
<h2 class="western"><a name="how-does-alignment-scoring-work"></a>74.5
How does alignment scoring work?</h2>
<p class="first-paragraph">The scores are the values, both positive
and negative, that an algorithm will assign to various bases when
they line up a certain way. An aligner attempts to produce an
arrangement that <b>maximizes</b> the score.</p>
<p>For example, we could choose the following scores:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
	point for a match.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-1</font></font>
	point for a mismatch.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-2</font></font>
	points for opening a gap.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-1</font></font>
	point for extending an already opened gap.</p>
</ul>
<p class="first-paragraph">We then instruct the aligner to find the
best alignments using the scores above. With this scoring system, the
alignments would be scored as (scores shown at the bottom):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA
     GATTACA      GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||.|
       |||  ||      || | ||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA--
     GAT--CA      GA-T-CA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0             2            1</font></font></p>
<p class="first-paragraph">With this scoring, the second alignment
has the highest score (2) and will be considered the &ldquo;best&rdquo;
(highest scoring) alignment.</p>
<p>As it happens, when sequences have different lengths we often want
to find the longest run that covers the shortest sequence. To account
for this situation a scoring &lsquo;adjustment&rsquo; is applied
where we do not penalize gaps at the end of either sequence. If we
were to apply that correction the scores would become:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATTACA
     GATTACA      GATTACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||.|
       |||  ||      || | ||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATCA--
     GAT--CA      GA-T-CA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 3             2            1</font></font></p>
<p class="first-paragraph">Now the first alignment ended up with the
maximal score and with that, it became the &ldquo;best alignment&rdquo;.
As you can imagine other changes to scoring could also affect what
becomes the best alignment.</p>
<h2 class="western"><a name="how-do-i-choose-the-scores"></a>74.6 How
do I choose the scores?</h2>
<p class="first-paragraph">In most cases, we start with known scores,
computed by observing the substitution rates across evolutionary
history. Since there are different ways scientists can consolidate
their observations - there will be various alternative scoring
matrices. Appropriate starting points already exist. See for example
<font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/matrices">ftp://ftp.ncbi.nlm.nih.gov/blast/matrices</a></font></p>
<p>For example, the <font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/NUC.4.4">EDNAFULL</a></font>
scoring matrix is the default nucleotide scoring choice some
aligners. Get that scoring matrix:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O  ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/NUC.4.4</font></font></p>
<p class="first-paragraph">then print it with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
NUC.4.4</font></font></p>
<p class="first-paragraph">The actual matrix is more extensive than
what we show below, as it includes all the ambiguous bases as well
(refer to the biology intro for information on those); the following
section is relevant:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">    <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
  T   G   C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
  5  -4  -4  -4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">T
 -4   5  -4  -4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">G
 -4  -4   5  -4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
 -4  -4  -4   5</font></font></p>
<p class="first-paragraph">The file shows that the score contribution
from matching an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
is 5 points, whereas the score contribution from matching an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
with a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
is -4 points (i.e., a score penalty of 4 points).</p>
<p>Just by visually inspecting the scoring matrix above we can
already tell that it will &ldquo;favor&rdquo; mismatches. For example
it will &ldquo;happily&rdquo; produce the alignmen that has
mismatches like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align ATG ACG --local -match 5 -mismatch 4</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|.|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACG</font></font></p>
<p class="first-paragraph">The second match <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G/G</font></font>
will add 5 points and that contribution would overcome the penalty of
-4 introduced by the mismatch. If instead we wrote:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align ATG ACG --local -match 4 -mismatch 5</font></font></p>
<p class="first-paragraph">the local alignment would be just:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font></p>
<p class="first-paragraph">the alignment cannot extend to the left
anymore as it is &ldquo;not worth&rdquo; adding a penalty of -5 when
the reward is only 4</p>
<p>Choosing the proper matrix, especially for protein sequences is
even more complicated. Refer to the following publication for more
details:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3848038/">Selecting
	the Right Similarity-Scoring Matrix</a></font> by William Pearson,
	the author of the FASTA program.</p>
</ul>
<h2 class="western"><a name="what-kind-of-scoring-matrices-are-there"></a>
74.7 What kind of scoring matrices are there?</h2>
<p class="first-paragraph">There are two types: nucleotide scoring
matrices and protein scoring matrices. The protein scoring matrices
come in many variants computed with different assumptions on what
similarity means. Also, (if that previous wasn&rsquo;t enough of
complexity) the scoring matrices can be normalized and
non-normalized. Care must be taken when comparing alignment scores
computed with different scoring matrices, even if they seem to have
the same name! One scoring matrix could be normalized and the other
not.</p>
<p>Visit: <font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/matrices">ftp://ftp.ncbi.nlm.nih.gov/blast/matrices</a></font>
for a list of matrices.</p>
<h2 class="western"><a name="what-other-properties-do-scoring-matrices-have"></a>
74.8 What other properties do scoring matrices have?</h2>
<p class="first-paragraph">Observe that the scoring matrix does not
include the information on the gap penalties. It is also worth noting
that the gap opening and extension penalties are typically different.</p>
<p>Specifically, the gap extension penalty is usually much smaller
than the gap opening one. This scoring method (also called <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Gap_penalty">affine
gap penalty</a></font>) has a biologically relevant rationale that
you may want to read up on your own.</p>
<p>It is essential to recognize just how impactful the choice of
scoring is and how profound its effects are because scoring governs
how many gaps versus mismatches the aligner produces.</p>
<p>A final note: a custom scoring matrix can only work in a
biologically meaningful way if it sums up to negative values along
both of its rows and columns. It is also worth keeping in mind that
the scores are typically on a logarithmic scale.</p>
<h2 class="western"><a name="the-inconvenient-truth-about-alignments"></a>
74.9 The &ldquo;inconvenient truth&rdquo; about alignments</h2>
<ul>
	<li/>
<p class="compact">When the sequences are very similar (nearly
	identical) the choice of scoring or even algorithms may not matter
	at all. The results will be <i>robust</i> - producing identical
	alignments across different algorithms and parameter settings.</p>
	<li/>
<p class="compact">When the sequences are dissimilar the
	choice of algorithm and scoring will may have radical impacts on the
	results. In general, the more different the sequences, the more
	sensitive the alignment becomes to parameter choices.</p>
	<li/>
<p class="compact">The longer the sequence the more likely is
	to allow different alignments with identical scores</p>
</ul>
<p class="first-paragraph">In our opinion, the lack of understanding
of the above constraints is the root cause of quite numerous invalid
scientific reports.</p>
<p>Scientists, having explored situations where the sequences were
similar, and the results were robust, develop a false sense of
understanding and &ldquo;comfort&rdquo; with the process and will
start applying the same strategies for situations where the alignment
processes are very sensitive to parameter choices</p>
<h2 class="western"><a name="how-are-alignments-displayed"></a>74.10
How are alignments displayed?</h2>
<p class="first-paragraph">There are several ways that alignments can
be reported, and there is no simple, universal format that can
present all the information encoded in an alignment.</p>
<p>Often, we use a visual display that employs various extra
characters to help us &ldquo;interpret&rdquo; the lineup. For
example:</p>
<ul>
	<li/>
<p class="compact">the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font>
	character may indicate a &ldquo;gap&rdquo; (space),</p>
	<li/>
<p class="compact">the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">|</font></font>
	character is used to display a match,</p>
	<li/>
<p class="compact">the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
	character may be used to display a mismatch.</p>
</ul>
<p class="first-paragraph">For example the following is an alignment
between two sequences <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCAAATGACAAATCGA</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCTGATAACTGCGA</font></font>
:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCAAATGACAAAT-CGA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||
  |||.||.| |||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC---TGATAACTGCGA</font></font></p>
<p class="first-paragraph">Above 13 bases are the same (13
identities), 5 bases are missing (5 gaps), 2 bases are different (2
mismatches).</p>
<h2 class="western"><a name="does-it-matter-which-sequence-is-on-top"></a>
74.11 Does it matter which sequence is on top?</h2>
<p class="first-paragraph">Usually, if not stated explicitly
otherwise, we read and interpret the alignment as if we were
comparing the bottom sequence against the top one.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCAAATGACAAAT-CGA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||
  |||.||.| |||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC---TGATAACTGCGA</font></font></p>
<p class="first-paragraph">Whereas the word gap is generic and refers
to either sequence, once we want to be more specific we would say
that the above is an alignment with three deletions of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
and one insertion of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>.
The word &ldquo;deletion&rdquo; means that the second sequence has
missing bases relative to the first.</p>
<p>We could generate and display this same alignment the other way
around:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC---TGATAACTGCGA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||
  |||.||.| |||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCAAATGACAAAT-CGA</font></font></p>
<p class="first-paragraph">The alignment would now be described as
one that contains three insertions of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>s
followed later by a deletion a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
relative to the top sequence. One sequences&rsquo; deletion is the
other sequences&rsquo; insertion - it all depends on what the study
is about.</p>
<h2 class="western"><a name="are-there-other-ways-to-quantify-alignments"></a>
74.12 Are there other ways to quantify alignments?</h2>
<p class="first-paragraph">Beyond the score of an alignment several
other characteristics of an alignment may be of interest. For example</p>
<ul>
	<li/>
<p class="compact">Percent Identity: What percentage of the
	two sequences are the same?</p>
	<li/>
<p class="compact">Percent Similarity: What proportion of the
	two sequences have similar bases/amino acids</p>
	<li/>
<p class="compact">E-Values: Number of observed alignments by
	chance</p>
	<li/>
<p class="compact">Mapping quality: The likelihood of
	incorrect alignment</p>
</ul>
<p class="first-paragraph">In addition additional details of the
alignments may be relevant</p>
<ul>
	<li/>
<p class="compact">How long is the aligned region?</p>
	<li/>
<p class="compact">How many bases does the aligned region
	cover in the query or the target?</p>
	<li/>
<p class="compact">How do the coordinates correspond to one
	another? The fifth base of query is aligned to which base of the
	target?</p>
	<li/>
<p class="compact">How many changes does it take to transform
	one sequence into another?</p>
</ul>
<h2 class="western"><a name="what-is-percent-identity"></a>74.13 What
is percent identity?</h2>
<p class="first-paragraph">But then it turns out that even a
seemingly well defined concept, such as percent identity may have
various definitions, each of which may produce radically different
values in some situations. Heng Li summarizes the problem in the blog
post titled:</p>
<ul>
	<li/>
<p class="compact">(On the definition of sequence
	identity](<font color="#4f81bd"><a href="https://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity">https://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity</a></font>)</p>
</ul>
<p class="first-paragraph">where enumerates the following three
measures</p>
<ol>
	<li/>
<p class="compact">Gap-excluded identity: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pident
	= matches / (matches + mismatches)</font></font></p>
	<li/>
<p class="compact">BLAST identity: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pident
	= matches / (match + mismatch + deletion)</font></font></p>
	<li/>
<p class="compact">Gap compressed identity: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pident
	= matches / (match + mismatch + gapopen)</font></font></p>
</ol>
<p class="first-paragraph">Each of the measures has advantages and
disadvantages, but, evidently when someone reports percent
identities, we ought to know which type of percent identity they
used.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font>package implements the BLAT type of percent
identity:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align CCAGTGTGGCCGATACCCCAGGTTGGCACGCATCGTTGCCTTGGTAAGC  \</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       CCAGTGTGGCCGATGCCCGTGCTACGCATCGTTGCCTTGGTAAGC --open 2 -extend
1</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (49) vs seq2 (45)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=86.0% len=50 ident=43 mis=1 del=5 ins=1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
semiglobal: score=32.0 match=1 mismatch=2 gapopen=2
gapextend=1.0</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAGTGTGGCCGATACCCCAGGTTGGC-ACGCATCGTTGCCTTGGTAAGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||||||||.|||---||--||-||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAGTGTGGCCGATGCCC---GT--GCTACGCATCGTTGCCTTGGTAAGC</font></font></p>
<h2 class="western"><a name="what-is-a-cigar-string"></a>74.14 What
is a CIGAR string?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CIGAR</font></font>
string (Compact Idiosyncratic Gapped Alignment Report) is an
alignment format used within the Sequence Alignment Map (SAM) files
that form the backbone of most bioinformatics high-throughput
analyses. For the same alignment from above:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGCAAATGACAAATAC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||
  |||.||.|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGC---TGATAACT--</font></font></p>
<p class="first-paragraph">the reported <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CIGAR</font></font>
format would be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4M3D3M1X2M1X1M2D</font></font></p>
<p class="first-paragraph">We read out this &ldquo;idiosyncratic&rdquo;
construct like so <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4M
+ 3D + 3M + 1X + 2M + 1X + 1M + 2D</font></font>:</p>
<ul>
	<li/>
<p class="compact">4 matches followed by</p>
	<li/>
<p class="compact">3 deletions,</p>
	<li/>
<p class="compact">3 matches,</p>
	<li/>
<p class="compact">1 mismatch,</p>
	<li/>
<p class="compact">2 matches,</p>
	<li/>
<p class="compact">1 mismatch,</p>
	<li/>
<p class="compact">1 match,</p>
	<li/>
<p class="compact">2 deletions.</p>
</ul>
<p class="first-paragraph">The format above is in a so-called
&ldquo;Extended CIGAR,&rdquo; meaning that it employs the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>
symbol for mismatches.</p>
<p>Another variant where <i>both</i> matches and mismatches are
designated by an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>
(as odd as that might sound) is the format used within the Sequence
Alignment Map (SAM). In that format the extended CIGAR string of</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4M3D3M1X2M1X1M2D</font></font></p>
<p class="first-paragraph">would be written as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4M3D7M2D</font></font></p>
<p class="first-paragraph">Note how the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3M
+ 1X + 2M + 1X</font></font> will be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7M</font></font>
forming the the sum of consecutive matches or mismatches. The value
of this latter CIGAR form is its compactness and an emphasis on
insertions and deletions - though in our opinion, in most cases it
sacrifices too much information to be practical.</p>
<p>Finally, there is yet another idiosyncratic way to write CIGARs
(don&rsquo;t you love it when a funny, self-deprecating name ends up
being quite deservedly so?). In that format for a single base long
change the number is not shown like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10M1D10M</font></font></p>
<p class="first-paragraph">would be written as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10MD10M</font></font></p>
<p class="first-paragraph">You may see CIGAR notations in any of the
variants above.</p>
<h2 class="western"><a name="where-can-i-learn-more-about-alignments"></a>
74.15 Where can I learn more about alignments?</h2>
<p class="first-paragraph">There is a large body of work and
information on alignments and scoring that are beyond the scope of
this book. There are good starting points in Wikipedia, as well as in
many other easy-to-find web resources.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Sequence_alignment">Wikipedia:
	Sequence Alignment</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Gap_penalty">Wikipedia:
	Gap Penalty</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="global-and-local-alignments"></a>75
Global and local alignments</h1>
<p class="first-paragraph">To run the examples install the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package developed for this book.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph">Read more about how the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> method works here at
<font color="#4f81bd"><a href="https://www.bioinfo.help/bio-align.html">https://www.bioinfo.help/bio-align.html</a></font></p>
<h2 class="western"><a name="important-note-about-alignments"></a>75.1
Important note about alignments</h2>
<p class="first-paragraph">The alignment algorithms in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> use the BioPython implementation that is best
suited for interactive and exploratory use when aligning relatively
short (up to ~30Kb long) sequences.</p>
<p>Specialized genomics alignment software will typically operate
many orders of magnitude faster while trading various features for
that speed boost. Depending on your needs you may want to use one of
the many different aligners: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blast</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blat</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fasta36</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mummer</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">minimap2</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">lastz</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">lastal</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">exonerate</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">vsearch</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">diamond</font></font>
&hellip; etc</p>
<p>The wealth of choices for selecting an aligner, together with the
immense number of options and parameters that each aligner can be
customized with indicate that <i>alignments</i> are complex and
multifaceted tasks.</p>
<h2 class="western"><a name="how-to-perform-an-alignment-with-bio"></a>
75.2 How to perform an alignment with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font></h2>
<p class="first-paragraph">Verify that the script works with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED</font></font></p>
<p class="first-paragraph">will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (8) vs seq2 (9)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=36.4% len=11 ident=4 mis=2 del=2 ins=3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
semiglobal: score=8.0 matrix=BLOSUM62 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">THISLI--NE-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--||..--||-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--ISALIGNED</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> tool is written in such a way that we may pass
either text strings or FASTA file names to it. Let&rsquo;s obtain a
different scoring matrix</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/BLOSUM30</font></font></p>
<p class="first-paragraph">This is how we can use the custom scoring
matrix:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED --matrix BLOSUM30</font></font></p>
<p class="first-paragraph">and that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (8) vs seq2 (9)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=36.4% len=11 ident=4 mis=2 del=2 ins=3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
semiglobal: score=13.0 matrix=BLOSUM30 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">THIS--LINE-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--||--..||-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--ISALIGNED</font></font></p>
<p class="first-paragraph">In the examples below, we will use
hypothetical protein sequences <font face="Consolas, serif"><font size="2" style="font-size: 11pt">THISLINE</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ISALIGNED</font></font>
real words that also happend to be valid peptide sequences. These
sequences were first used for a similar purpose, albeit in a
different context in the book <i>Understanding Bioinformatics</i> by
<i>Marketa Zvelebil</i> and <i>Jeremy Baum</i></p>
<h2 class="western"><a name="what-is-a-global-alignment"></a>75.3
What is a global alignment?</h2>
<p class="first-paragraph">By default the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> tool performs a semi-global alignment, where the
end gaps of the query do not have penalties. When performing global
alignments, the bases of both sequences are arranged next to one
another over their entire length and all . Each base of the first
sequence is matched to another base or a &ldquo;gap&rdquo; of the
second sequence.</p>
<p>We use global alignments when we are looking for an arrangment
that maximizes the similarities over the entire length of both
sequences:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED --global</font></font></p>
<p class="first-paragraph">that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (8) vs seq2 (9)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=22.2% len=9 ident=2 mis=6 del=0 ins=1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
global: score=-7.0 matrix=BLOSUM62 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">THISLINE-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">......||-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ISALIGNED</font></font></p>
<p class="first-paragraph">We may also override the gap open and gap
extension penalty:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED --global --open 7</font></font></p>
<p class="first-paragraph">now the alignment looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (8) vs seq2 (9)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=54.5% len=11 ident=6 mis=0 del=2 ins=3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
global: score=6.0 matrix=BLOSUM62 gapopen=5
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">THIS-LI-NE-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--||-||-||-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--ISALIGNED</font></font></p>
<p class="first-paragraph">Note how radically different the second
alignment is from the first one. All we did is reduce the penalty of
opening a gap to from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>.
The alignment is longer but has more gaps. The tradeoff is readily
apparent.</p>
<p>Remember, an alignment finds the arrangement that maximizes the
score of rewards and penalties across the sequences.</p>
<h2 class="western"><a name="what-is-a-local-alignment"></a>75.4 What
is a local alignment?</h2>
<p class="first-paragraph">Local alignments are used when we need to
find the region of maximal similarity between two sequences. When
performing local alignments, the algorithms look for the highest
scoring (partial) interval between the two sequences :</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED --local</font></font></p>
<p class="first-paragraph">When run as above the local alignment
generated with the default parameters will be surprisingly short:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (2) vs seq2 (2)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=100.0% len=2 ident=2 mis=0 del=0 ins=0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
local: score=11.0 matrix=BLOSUM62 gapopen=11 gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NE</font></font></p>
<p class="first-paragraph">The algorithm is telling us that these two
matching amino acids produce the highest possible score (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">11</font></font>
in this case) and any other local alignment between the two sequences
will produce a score that is worse than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11</font></font>.</p>
<p>We can use other scoring matrices as shown in
<font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/">ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/</a></font>,
many of these are included with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
DNA matrices</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This matrix is the &quot;standard&quot; EDNAFULL substitution
matrix.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/NUC.4.4</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Protein matrices</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the BLOSUM30, BLOSUM62, and BLOSUM90 matrices.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc ftp://ftp.ncbi.nlm.nih.gov/blast/matrices/BLOSUM30</font></font></p>
<p class="first-paragraph">Note how the local alignment is affected
by the scoring scheme.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align THISLINE ISALIGNED -matrix BLOSUM30 --local</font></font></p>
<p class="first-paragraph">Using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BLOSUM90</font></font>
scoring scheme produces a much longer alignment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (4) vs seq2 (4)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=50.0% len=4 ident=2 mis=2 del=0 ins=0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
local: score=15.0 matrix=BLOSUM30 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">LINE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">..||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGNE</font></font></p>
<h2 class="western"><a name="how-do-we-pick-the-right-matrix"></a>75.5
How do we pick the right matrix?</h2>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3848038/">Selecting
	the Right Similarity-Scoring Matrix</a></font> by William Pearson,
	the author of the FASTA program.</p>
</ul>
<p class="first-paragraph">Here are a few lines from the abstract:</p>
<p class="block-text-western">While &ldquo;deep&rdquo; matrices
provide very sensitive similarity searches, they also require longer
sequence alignments and can sometimes produce alignment overextension
into non-homologous regions. Shallower scoring matrices are more
effective when searching for short protein domains, or when the goal
is to limit the scope of the search to sequences that are likely to
be orthologous between recently diverged organisms.</p>
<p class="block-text-western">Likewise, in DNA searches, the match
and mismatch parameters set evolutionary look-back times and domain
boundaries.</p>
<h2 class="western"><a name="how-do-i-see-the-matrix"></a>75.6 How do
I &ldquo;see&rdquo; the matrix?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> will print the matrix (when using builtin
values):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align -matrix PAM250 | head -15</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This matrix was produced by &quot;pam&quot; Version 1.0.6
[28-Jul-93]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
PAM 250 substitution matrix, scale = ln(2)/3 = 0.231049</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Expected score = -0.844, Entropy = 0.354 bits</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Lowest score = -8, Highest score = 17</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  A  R  N  D  C  Q  E  G  H  I  L  K  M  F  P  S  T  W  Y  V  B  Z  X
 *</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A
 2 -2  0  0 -2  0  0  1 -1 -1 -2 -1 -1 -3  1  1  1 -6 -3  0  0  0  0
-8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R
-2  6  0 -1 -4  1 -1 -3  2 -2 -3  3  0 -4  0  0 -1  2 -4 -2 -1  0 -1
-8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N
 0  0  2  2 -4  1  1  0  2 -2 -3  1 -2 -3  0  1  0 -4 -2 -2  2  1  0
-8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">D
 0 -1  2  4 -5  2  3  1  1 -2 -4  0 -3 -6 -1  0  0 -7 -4 -2  3  3 -1
-8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
-2 -4 -4 -5 12 -5 -5 -3 -3 -2 -6 -5 -5 -4 -3  0 -2 -8  0 -2 -4 -5 -3
-8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<h2 class="western"><a name="why-are-scoring-values-integer-numbers"></a>
75.7 Why are scoring values integer numbers?</h2>
<p class="first-paragraph">You could wonder, rightfully so, how come
the scores are all integers? And also what does <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font>
vs.&nbsp;<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>
mean in a scoring matrix?</p>
<p>In a nutshell, the score reflects a probability. Besides, the
scores are represented as log 2 odds. A substitution score of 3 means
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2^3=8</font></font>
whereas a substitution score of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>
means <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2^5=32</font></font>,
thus a four-fold increase of probability relative to one another.</p>
<p>Thus a substitution with a score of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font>
is four times more likely to occur than a change with a score of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>.
For simplicity and to keep our sanity the probabilities log2 odds
were all rounded to nearest integers.</p>
<p>Note how the score is not absolute; they are relative to one
another. Any other number pars with the same ratios would have the
same effect.</p>
<h2 class="western"><a name="what-is-a-semi-global-alignment"></a>75.8
What is a semi-global alignment?</h2>
<p class="first-paragraph">Semi-global alignments (also known as
global-local, glocal) are cross between the global and local
alignments.</p>
<p>Semi-global alignments attempt to completely align a shorter
sequence against a longer (reference) one. The goal is achieved by
setting the end gap penalites to zero.</p>
<p>Semi-global aligners are used when matching sequencing reads
produced by sequencing instruments against reference genomes. Most
high throughput data analysis protocols that we cover in this book
rely on tools that use this type of alignment.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="misleading-alignments"></a>76 Misleading
alignments</h1>
<h2 class="western"><a name="will-alignments-always-indicate-the-correct-variation"></a>
76.1 Will alignments always indicate the &ldquo;correct&rdquo;
variation?</h2>
<p class="first-paragraph">We want to show a situation that you may
encounter to demonstrate the limitations of using mathematical
concepts to identify biological phenomena.</p>
<p>Imagine that the sequence below is subjected to two insertions of
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>s
at the locations indicated with carets:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCTCCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    ^       ^</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    |       |</font></font></p>
<p class="first-paragraph">The two sequences, when placed next to one
another, would look like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCTCCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCCTCCCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font>
</p>
<p class="first-paragraph">A better way to visualize what&rsquo;s
happening in this example is shown in the expected alignment that
would reflect the changes that we have introduced:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAA-CCCCCCCT-CCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||
|||||||| ||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCCTCCCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<p class="first-paragraph">Now suppose we did not know what the
changes were. We want to recover the alignment above to identify the
two insertions from the sequences alone. Can we discover the
variation that we introduced by using a global aligner? Let&rsquo;s
see:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC</font></font></p>
<p class="first-paragraph">Here is what we obtain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (23) vs seq2 (25)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=88.0% len=25 ident=22 mis=1 del=0 ins=2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
semiglobal: score=8.0 match=1 mismatch=2 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCC--TCCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||||||--.||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCCTCCCCCCGCTTC</font></font></p>
<p class="first-paragraph">Whoa! What just happened?</p>
<p>Our aligner reports our two separate and distant insertions of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>s
as a single insertion of a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CT</font></font>
followed by a mismatch of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>.</p>
<p>To add insult to injury, the change is shown to be in a completely
<i>different location</i>, not even where we have modified the
sequence! That&rsquo;s just crazy! How did that happen? What can we
do about it?</p>
<p>In a nutshell, there is a &ldquo;simpler&rdquo; explanation to
reconcile the differences between the two sequences - and the
aligner, using mathematical reasoning produces the simplest
explanation - even though this explanation does not reflect what
happened. Some people use the word &ldquo;misalignment&rdquo;, but
that would be incorrect - the aligner did its job correctly, but
there was no way for it to know that a more complicated explanation
was the correct one.</p>
<p>The main reason for the existence of this &ldquo;simpler
explanation&rdquo; is that the region between the two insertions has
low information content; the same base is repeated: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCCCCCC</font></font>.
This repetition throws off the neatly-crafted mathematical alignment
concept of rewarding matching bases. When shifted by one base, we
still get the same type of bases lined up, hence producing a positive
score even though it is the &ldquo;newly inserted&rdquo; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
that matches the &ldquo;original&rdquo; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>.
Let&rsquo;s compute the scores.</p>
<p>In our &ldquo;correct&rdquo; alignment, and when using the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">EDNAFULL</font></font>
matrix, the score will be formed from 23 matches and 2 gap openings:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(23
* 5) - (2 * 10) = 95</font></font></p>
<p class="first-paragraph">In the &ldquo;incorrect&rdquo; second
alignment we have only 22 matches with 1 gap open, 1 gap extension
and 1 mismatch leading to a score of</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(22
* 5) - 10 - 0.5 - 4 = 95.5</font></font></p>
<p class="first-paragraph">An optimal aligner finds the alignment
that maximizes the score. We now see that the second, &ldquo;incorrect&rdquo;
alignment produces a slightly better score; hence, it will be
reported as the most likely match between the two sequences.</p>
<p>We could recover the &ldquo;correct&rdquo; alignment by modifying
the parameters to reduce the gap open penalty and applying a
different scoring matrix:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC -open 9
-match 4 -mismatch 5</font></font></p>
<p class="first-paragraph">Now it will produce the expected output</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
seq1 (23) vs seq2 (25)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=92.0% len=25 ident=23 mis=0 del=0 ins=2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
semiglobal: score=74.0 match=4 mismatch=5 gapopen=9
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAA-CCCCCCCT-CCCCCGCTTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||-||||||||-||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCAAACCCCCCCCTCCCCCCGCTTC</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
48 identically scoring alignments! Pass the flag -A to see them all</font></font></p>
<p class="first-paragraph">note the warning, there are 48 identically
scoring alignments, basically each one of of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
may be mismatching to another <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
and then there are all possible combinations of them.</p>
<p>Does this mean that from now on we should run all alignments with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gapopen=9</font></font>?</p>
<p>Not really!</p>
<p>As we mentioned before, tweaking alignment parameters can have
far-reaching consequences on all other alignments that are typically
correct. Using this setting means that two matches <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5
+ 5 = 10</font></font> will overcome the penalty of a gap open <font face="Consolas, serif"><font size="2" style="font-size: 11pt">9</font></font>;
hence, the aligner will open gaps any time it can find two matches
later on. Sometimes that is a good idea, but that is most likely not
what we intended to do.</p>
<p>The problem above was caused not by the incorrect alignment
parameters but by the reduced information content of the series of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>s
(aka homopolymeric) region.</p>
<p>Situations such as the above will generally produce incorrect
variation calls and are an ongoing, profound, and ubiquitous
challenge when identifying genomic variations. As we shall see, the
latest SNP calling tools have features that allow them to recognize
and correct &ldquo;misalignments&rdquo; - though as you can imagine
the problems are far from being solved.</p>
<p>A universal take-home message:</p>
<p class="block-text-western">Alignment reliability also depends on
the information content of the aligned sequence itself.</p>
<p class="block-text-western">Alignments that operate on low
complexity regions are generally produce less reliable variation
calls.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="blast-basic-local-alignment-search-tool"></a>
77 BLAST: Basic Local Alignment Search Tool</h1>
<h2 class="western"><a name="installation"></a>77.1 Installation</h2>
<p class="first-paragraph">This section has been rewritten to use the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package. Learn more about <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
here:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help/</a></font>.</p>
</ul>
<p class="first-paragraph">To install and update <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph">The package changes often; make sure to
upgrade it regularly.</p>
<h2 class="western"><a name="what-is-blast"></a>77.2 What is BLAST?</h2>
<p class="first-paragraph">BLAST is an acronym for <b>Basic Local
Alignment Search Tool</b>. BLAST represents both an algorithm and a
suite of tools that implement said algorithm. The primary purpose of
BLAST is to search a collection of target sequences for alignments
that match a query sequence.</p>
<p>Using BLAST is a bit like using Google when searching for
information. When we perform a Google query, we are looking to find
documents that contain the words that we are seeking. Similarly,
BLAST search results are sequences that have similarities to the
query sequence. As the name says, the results of a BLAST search are
local alignments; thus, a blast result is generally a partial match
of the query sequence to a target sequence in the database.</p>
<p>BLAST can be run both as a <font color="#4f81bd"><a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">web
interface from NCBI</a></font> and as standalone downloadable tools.
NCBI also maintains a dull to read <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">Blast
Handbook</a></font> that contains further information on how BLAST
works.</p>
<h2 class="western"><a name="what-are-the-blast-tools"></a>77.3 What
are the BLAST tools?</h2>
<p class="first-paragraph">BLAST consists of a suite of tools</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastx</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tblastn</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tblastx</font></font>
	and others</p>
</ul>
<p class="first-paragraph">Users can then customize each tool via
so-called tasks specified with the flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-task</font></font>.
Think of these tasks as critical tuning parameters: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">megablast</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp-short</font></font>
that will radically change the search&rsquo;s operation. Generally
speaking, when communicating research findings, it is insufficient to
state which blast program one runs; one should also mention the task.
Remember that even if you don&rsquo;t explicitly choose a task, the
software tool will select one for you by default. For example, the
&lsquo;blastn&rsquo; tool will run with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">megablast</font></font>
task by default.</p>
<p>BLAST is not an optimal aligner. This means that it may not find
all alignments and there are limits for how short and how long
sequences may be for BLAST to operate correctly. Also, BLAST is tuned
for performance and efficiency; its primary purpose is to search a
large body of known information for similarities (hits). To quote the
World&rsquo;s Most Interesting Man: <i>I don&rsquo;t generally misuse
BLAST, but when I do I make huge mistakes!</i>. What this means is
that BLAST will not typically miss a hit or produce a wrong
alignment. What happens more frequently is that it is used in a
manner and context that is entirely wrong. Thankfully &ldquo;big&rdquo;
mistakes are easier to catch.</p>
<h2 class="western"><a name="what-are-the-fundamental-concepts-of-blast"></a>
77.4 What are the fundamental concepts of BLAST?</h2>
<ul>
	<li/>
<p class="compact">A search may occur in nucleotide space,
	protein space or translated spaces where nucleotides are translated
	into proteins.</p>
	<li/>
<p class="compact">Searches may implement search &ldquo;strategies&rdquo;:
	optimizations to a specific task. Different search strategies will
	produce different alignments.</p>
	<li/>
<p class="compact">Searches use alignments that rely on
	scoring matrices</p>
	<li/>
<p class="compact">Searches may be customized with many
	additional parameters.</p>
</ul>
<p class="first-paragraph">BLAST has many parameters that most users
may never need to use or set. Knowing BLAST well can be a &ldquo;full-time
job&rdquo; - there are job positions that mainly consist of using
BLAST in various contexts. There are books and courses on just how to
tune and operate BLAST searches.</p>
<p class="block-text-western">NOTE: Biologist love BLAST! It is easy
to see why. It feels like hitting the jackpot when it works, like
when you are stuck and Googling a word takes you to the answer you
were looking for all along. When we search with BLAST the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NT</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR</font></font>
databases, we are navigating through the accumulated knowledge that
humankind has collected. It is incredibly empowering.</p>
<p class="first-paragraph">But there is a downside to all this power
- as our knowledge grows, the possible rate of false discovery
increases faster than most scientists assume. From a purely
evolutionary point of view, every sequence should be &ldquo;similar&rdquo;
to any other sequence as they share common ancestry (distant as it
may be). As we add more sequences onto the same pile, the picture
gets fuzzier and unreliable for making precise statements. Knowing
where to draw the line, which similarities are extreme events of
random chance, and which are meaningful is a far more complicated
task than most scientists care to admit.</p>
<h2 class="western"><a name="can-blast-be-used-for-pairwise-alignments"></a>
77.5 Can BLAST be used for pairwise alignments?</h2>
<p class="first-paragraph">While pairwise alignment is not the
primary use case of BLAST, the tools in the suite do support the
function. Here is an example of running BLAST to align the coding
sequence for the nucleoprotein named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NP</font></font>
across the 1976 and 2018 Ebola strains.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The Ebola nucleoprotein in the 1997 strain</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch AAD14590 --format fasta &gt; AAD14590.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The Ebola nucleoprotein in the 2018 strain</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch ARG44037 --format fasta &gt; ARG44037.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run a pairwise BLAST alignment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-query AAD14590.fa -subject ARG44037.fa</font></font></p>
<p class="first-paragraph">that will produce the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Score
= 1512 bits (3915),  Expect = 0.0, Method: Compositional matrix
adjust.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Identities
= 728/739 (99%), Positives = 730/739 (99%), Gaps = 0/739 (0%)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 1    MDSRPQKIWMAPSLTESDMDYHKILTAGLSVQQGIVRQRVIPVYQVNNLEEICQLIIQAF 
60</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           MDSRPQK+WM
PSLTESDMDYHKILTAGLSVQQGIVRQRVIPVYQVNNLEEICQLIIQAF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 1    MDSRPQKVWMTPSLTESDMDYHKILTAGLSVQQGIVRQRVIPVYQVNNLEEICQLIIQAF 
60</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 61   EAGVDFQESADSFLLMLCLHHAYQGDYKLFLESGAVKYLEGHGFRFEVKKRDGVKRLEEL 
120</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          
EAGVDFQESADSFLLMLCLHHAYQGD+KLFLESGAVKYLEGHGFRFEVKKRDGVKRLEEL</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 61   EAGVDFQESADSFLLMLCLHHAYQGDHKLFLESGAVKYLEGHGFRFEVKKRDGVKRLEEL 
120</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 121  LPAVSSGKNIKRTLAAMPEEETTEANAGQFLSFASLFLPKLVVGEKACLEKVQRQIQVHA 
180</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          
LPAVSSGKNIKRTLAAMPEEETTEANAGQFLSFASLFLPKLVVGEKACLEKVQRQIQVHA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 121  LPAVSSGKNIKRTLAAMPEEETTEANAGQFLSFASLFLPKLVVGEKACLEKVQRQIQVHA 
180</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The report states that the sequences match
at the 99% level, and 728 out of 739 bases are identical. Now reading
default BLAST outputs is also a skill that develops over time.
Thankfully there are various ways we can reformat the output. If all
we wanted was to know the percent identities, we could have run
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp</font></font>
using a column-based output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-query AAD14590.fa -subject ARG44037.fa -outfmt '6 pident'</font></font></p>
<p class="first-paragraph">and that would have produced just one
value:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">98.512</font></font></p>
<h2 class="western"><a name="how-do-i-use-blast"></a>77.6 How do I
use BLAST?</h2>
<p class="first-paragraph">The most common use of blast requires the
following three steps:</p>
<ol>
	<li/>
<p class="compact">Prepare a BLAST database with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb</font></font>.
	This task only needs to be done once per sequence data.</p>
	<li/>
<p class="compact">Select the appropriate BLAST tool: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp</font></font>.
	Depending on your needs you may need to tune the parameters.</p>
	<li/>
<p class="compact">Run the tool and format the output as
	necessary.</p>
</ol>
<p class="first-paragraph">For example, suppose we want to
investigate how the sequence of the polymerase gene VP35 changed
since 1976. To do so, we will align the 1976 version of VP35 to all
protein sequences obtained during the 2014 outbreak. We also know
that the data for the 1976 strain was deposited under the accession
number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>.</p>
<p>First, we need to find the accession number for the VP35
polymerase gene from 1976. We could, or we could see it inside the
GenBank file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the Ebola genome as genbank.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch AF086833 &gt; 1976.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Find the genes listed in the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
1976.gb | bio table -type CDS -fields id,gene,size</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14590.1
NP 2220</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
VP35   1023</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14583.1
VP40   981</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14585.1
GP 2031</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14584.1
GP 1095</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14586.1
GP 894</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14587.1
VP30   867</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14588.1
VP24   756</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14589.1
L  6639</font></font></p>
<p class="first-paragraph">Fetch the sequence for the VP35 protein
from 1976:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch AAD14582 --format fasta  &gt; VP35_1976.fa</font></font></p>
<p class="first-paragraph">We will now get all proteins deposited for
project <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>
described in <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a directory to hold the sequences</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ref</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fetch the sequences for the project.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db protein -query PRJNA257197 | efetch -format fasta &gt;
ref/prots_2014.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run a sanity check to verify it all went well.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats ref/prots_2014.fa</font></font></p>
<p class="first-paragraph">It produces showing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2240</font></font>
sequences with over a million bases in total. This will be our search
space:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats ref/prots_2014.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
              format  type     num_seqs    sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ref/prots_2014.fa
 FASTA   Protein     2,240  1,367,025      217    610.3    2,212</font></font></p>
<p class="first-paragraph">In summary, our BLAST query will be
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VP35_1976.fa</font></font>
and we want to search the sequences in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ref/prots_2014.fa</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the blast database</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-dbtype prot -in ref/prots_2014.fa -out ref/prots_2014</font></font></p>
<p class="first-paragraph">Our query is a protein sequence, and the
target database is a protein sequence; thus the appropriate BLAST
tool is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run the blastp tool</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-db ref/prots_2014 -query VP35_1976.fa &gt; results.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
prints 8021</font></font></p>
<p class="first-paragraph">the command will save the results into the
file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results.txt</font></font>
- as you will see it is a file with 8000 lines, lots of alignments,
overall not so easy to make sense of - there are just too many hits.
It is best to reformat the output to list only the columns you might
be interested in. Let&rsquo;s reformat to tabular form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-db ref/prots_2014 -query VP35_1976.fa -outfmt 7 | head</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
BLASTP 2.7.1+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Query: AAD14582.1 VP35 [Ebola virus - Mayinga, Zaire, 1976]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Database: db/proteins</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fields: query acc.ver, subject acc.ver, % identity, alignment length,
mismatches, gap opens, q. start, q. end, s. start, s. end, evalue,
bit score</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
249 hits found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIG95885.1 98.824 340    4  0  1  340    1  340    0.0   
702</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIG95894.1 98.824 340    4  0  1  340    1  340    0.0   
702</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIG95903.1 98.824 340    4  0  1  340    1  340    0.0   
702</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIG95912.1 98.824 340    4  0  1  340    1  340    0.0   
702</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIG95921.1 98.824 340    4  0  1  340    1  340    0.0    702</font></font></p>
<p class="first-paragraph">Now the task of making sense of this file
is where the challenges begin. How do we identify what is essential,
what is not, which columns should we display, which are relevant, and
so on. At this point, running BLAST was the &ldquo;easy&rdquo; part -
making sense of data? That is always the actual research question,
framing the research problem in terms that blast can answer.</p>
<p>For example the three best alignments</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-db ref/prots_2014 -query VP35_1976.fa -outfmt &quot;6 qseqid sseqid
pident&quot; | sort -k3 -rn | head -5</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AKC37262.1 98.824</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AKC37253.1 98.824</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AKC37244.1 98.824</font></font></p>
<p class="first-paragraph">The worst alignments.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp
-db ref/prots_2014 -query VP35_1976.fa -outfmt &quot;6 qseqid sseqid
pident&quot; | sort -k3 -rn | tail -5</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AIE11834.1 98.529</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AKC36408.1 97.059</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14582.1
AKC35949.1 96.765</font></font></p>
<h2 class="western"><a name="what-are-the-blast-tools-1"></a>77.7
What are the blast tools?</h2>
<p class="first-paragraph">Blast is an entire suite of somewhat
confusingly named programs.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastp</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tblastn</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastx</font></font>
	etc. Each tool performs a certain type of query (nucleotide or
	peptide) versus a certain type of target (nucleotide or peptide)
	using a certain type of translation (nucleotide or peptide). It is
	often difficult to remember which blast tool does what as the naming
	is not all that logical.</p>
</ul>
<p class="first-paragraph">As strange as that might sound when
figuring out which blast tool is more appropriate, it makes more
sense to try to remember what each tool should have been named (using
the strategy below) instead of what they currently named.</p>
<ol>
	<li/>
<p class="compact">A blast search tool where both the query
	and target are in <i>nucleotide</i> space should have been called
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastNN</font></font>
	(but instead it is called just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>).</p>
	<li/>
<p class="compact">A blast tool that searches a <i>peptide</i>
	sequence against a <i>nucleotide</i> sequence in the translated
	space should have been called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastPN</font></font>
	but instead, it is called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tblastn</font></font></p>
</ol>
<p class="first-paragraph">Here&rsquo;s a helpful table that
clarifies the blast tools:</p>
<p><img src="0000_html_f54bcdcc764b4cd6.png" name="Image127" align="bottom" hspace="12" width="560" height="335" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-blast-terminology"></a>77.8
What is the Blast terminology?</h2>
<ul>
	<li/>
<p class="compact"><b>Query</b>: the sequence that we are
	looking to match.</p>
	<li/>
<p class="compact"><b>Target</b>: the database (collection of
	sequences) that we are searching for matches.</p>
	<li/>
<p class="compact"><b>Subject</b>: the sequence entry in the
	target that produced an alignment.</p>
	<li/>
<p class="compact"><b>Score</b>: the alignment score.</p>
	<li/>
<p class="compact"><b>E-Value</b>: the Expect value (E)
	describes the number of hits better than the current hit that one
	can &ldquo;expect&rdquo; to see just by chance when searching a
	database of a particular size.</p>
</ul>
<h2 class="western"><a name="what-is-an-e-value"></a>77.9 What is an
E-Value</h2>
<p class="first-paragraph">The e-values are among the most
misunderstood and misused concepts of BLAST. E-values were designed
as the means of bringing some level of confidence to the search
results, and are defined as the number of hits one can &ldquo;expect&rdquo;
to see by chance when searching a database of a particular size. A
smaller the e-value is &ldquo;better&rdquo;.</p>
<p>The e-value was intended to be used as a filtering and threshold
parameter, where, supposedly by using them we could identify more
&ldquo;trustworthy&rdquo; hits. Fundamentally the problem with
e-values is that they depend on the database size and content, hence
are not a measure of validity, reliability or correctness. Detailed
definitions for e-values come in many shapes and forms, from
[overcomplicated mathematics and statistics][evaluestats] that few
people could possibly follow to hand-waving rationalizations like the
ones seen for example <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK1734/">Chapter
9: BLAST QuickStart</a></font></p>
<p class="block-text-western"><i>The alignments found by BLAST during
a search are scored, as previously described, and assigned a
statistical value, called the &ldquo;Expect Value.&rdquo; The &ldquo;Expect
Value&rdquo; is the number of times that an alignment as good or
better than that found by BLAST would be expected to occur by chance,
given the size of the database searched. An &ldquo;Expect Value&rdquo;
threshold, set by the user, determines which alignments will be
reported. A higher &ldquo;Expect Value&rdquo; threshold is less
stringent and the BLAST default of &ldquo;10&rdquo; is designed to
ensure that no biologically significant alignment is missed. However,
&ldquo;Expect Values&rdquo; in the range of 0.001 to 0.0000001 are
commonly used to restrict the alignments shown to those of high
quality.</i></p>
<p class="first-paragraph">After reading the above, you may ask
yourself what a &ldquo;high quality&rdquo; alignment is? And why
would an alignment with an e-value of 0.0001 be of &ldquo;lower
quality&rdquo; than one with an e-value of 0.0000001? It very much
sounds like &ldquo;quality&rdquo; is just a synonym for surprise!,
E-values! But then why would one ever set an e-value to 0.0000001
when searching for a single hit?</p>
<p>You see E-values in their current form are designed for tasks that
attempt to brute-force a problem by aligning even the kitchen sink to
everything else, with the hopes that something interesting eventually
pops up. For any other project with a realistic basis, e-values are
just a hindrance and little more than a chance to derail the
information you get..</p>
<p>Ranking blast search results by e-values is pervasive and
prevalent. Since both the alignment lengths and the alignment scores
are ingredients to the E-value formula, ranking by E-values typically
shows a strong correlation to a ranking by score and lengths.
E-values end up as a single &ldquo;convenient&rdquo; number that
empirically seems to work well.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="blast-use-cases"></a>78 BLAST use cases</h1>
<p class="first-paragraph">This section demonstrates a few use cases
for BLAST, and we will explore common problems and errors that you
might encounter.</p>
<h2 class="western"><a name="objective"></a>78.1 Objective</h2>
<p class="first-paragraph">Our objective is to compare sequences from
two strains of the Ebola virus, one from 1976 (accession <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>),
the other from 2014 (accession <font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118</font></font>).
We want to characterize the ways that the virus has changed over
time.</p>
<h2 class="western"><a name="installation-1"></a>78.2 Installation</h2>
<p class="first-paragraph">This section has been rewritten to use the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package. Learn more about <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
here:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help/</a></font>.</p>
</ul>
<p class="first-paragraph">To install and update <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph">The package changes often; make sure to
upgrade it regularly.</p>
<h2 class="western"><a name="get-the-reference-data"></a>78.3 Get the
reference data</h2>
<p class="first-paragraph">Let&rsquo;s build a database out of all
sequences of the 1976 Ebola genome deposited under accession number
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>.
We will store all reference data in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">db</font></font>
folder.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Database directory.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ref</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the 1976 Ebola genome as GenBank file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch AF086833 &gt; ref/1976.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract all sequences as a fasta file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ref/1976.gb | bio fasta &gt; ref/1976.fa</font></font></p>
<p class="first-paragraph">I often run statistics on files to ensure
that they are what I expect them to be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stats ref/*.fa</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
        format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ref/1976.fa
 FASTA   DNA         49   72,925        4  1,488.3   18,959</font></font></p>
<p class="first-paragraph">What does our file contain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ref/1976.gb | bio table --fields id,type,size,gene | column -t</font></font></p>
<p class="first-paragraph">the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">column
-t</font></font> pads our output to be more readable and prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
        source           18959</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">five_prime_UTR-1
  five_prime_UTR   55</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NP
                gene             2971   NP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
        mRNA             2971   NP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulatory-1
      regulatory       12     NP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14590.1
        CDS              2220   NP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">regulatory-2
      regulatory       12     NP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">misc_feature-1
    misc_feature     5</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">VP35
              gene             1376   VP35</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
        mRNA             1376   VP35</font></font></p>
<h2 class="western"><a name="how-to-construct-a-blast-database"></a>78.4
How to construct a blast database</h2>
<p class="first-paragraph">To see short and extended help for the
blast database builder, run the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-h</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-help</font></font></p>
<p class="first-paragraph">Construct the blast database with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run the blast database builder.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-dbtype nucl -in ref/1976.fa  -out ref/1976</font></font></p>
<p class="first-paragraph">We have now created a database containing
all the features of the 1976 genome. We can directly query this
database for various information.</p>
<h2 class="western"><a name="how-to-run-blast-queries"></a>78.5 How
to run BLAST queries?</h2>
<p class="first-paragraph">Let us now obtain data from the 2014
strain deposited under accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch KM233118 &gt; 2014.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
What types of sequences are present in the data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
2014.gb | bio table -type gene</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract the sequence for gene L (take the first 120 bases only)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
2014.gb | bio fasta --gene L -end 120 &gt; L.fa</font></font></p>
<p class="first-paragraph">Above, we take the first 120 bases only so
that the output of the alignment is more manageable. Run the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
tool</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/1976 -query L.fa</font></font></p>
<p class="first-paragraph">It generates a quite lengthy output. At
the top of the header, we see the name of the sequences that produce
alignments:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sequences
producing significant alignments:                          (Bits) 
Value</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAD14589.1
{&quot;type&quot;: &quot;CDS&quot;, &quot;gene&quot;: &quot;L&quot;,
&quot;product&quot;: &quot;polymerase&quot;, ...  206    
7e-56</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
{&quot;type&quot;: &quot;mRNA&quot;, &quot;gene&quot;: &quot;L&quot;,
&quot;product&quot;: &quot;polymerase&quot;,...  206     7e-56</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">L
{&quot;type&quot;: &quot;gene&quot;, &quot;gene&quot;: &quot;L&quot;,
&quot;product&quot;: &quot;&quot;, &quot;locus&quot;: &quot;&quot;}  
        206     7e-56</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833.2
{&quot;title&quot;: &quot;Ebola virus - Mayinga, Zaire, 1976,
complet...  206     7e-56</font></font></p>
<p class="first-paragraph">We can see the exact sequence may be
present in the database under different types. It appears that we
have the same sequence for a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CDS</font></font>,
an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mRNA</font></font>,
and a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene</font></font>.
Since we are studying a virus, the three sequences are identical. For
other organisms (for example, higher-order eukaryotes, the sequence
for each type would be different).</p>
<p>The alignment specific section is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;AAD14589.1
{&quot;type&quot;: &quot;CDS&quot;, &quot;gene&quot;: &quot;L&quot;,
&quot;product&quot;: &quot;polymerase&quot;,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;locus&quot;:
&quot;&quot;}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Length=6639</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Score = 206 bits (111),  Expect = 7e-56</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Identities = 117/120 (98%), Gaps = 0/120 (0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Strand=Plus/Plus</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 1    ATGGCTACACAACATACCCAATACCCAGACGCCAGGTTATCATCACCAATTGTATTGGAC 
60</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ||||||||||||||||||||||||||||||||
|||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 1    ATGGCTACACAACATACCCAATACCCAGACGCTAGGTTATCATCACCAATTGTATTGGAC 
60</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 61   CAATGTGACCTTGTCACTAGAGCTTGCGGGTTGTATTCATCATACTCCCTTAATCCGCAA 
120</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           ||||||||||| ||||||||||||||||||||
|||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 61   CAATGTGACCTAGTCACTAGAGCTTGCGGGTTATATTCATCATACTCCCTTAATCCGCAA 
120</font></font></p>
<h2 class="western"><a name="how-do-i-format-the-output-differently"></a>
78.6 How do I format the output differently?</h2>
<p class="first-paragraph">The need to extract information on
alignments is widespread. BLAST allows us to format the outputs into
tabular and other formats. Change it to output format <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>
(tabular form with or without comments and headers):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/1976 -query L.fa  -outfmt 7</font></font></p>
<p class="first-paragraph">it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
BLASTN 2.12.0+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Query: AIG96639.1 {&quot;type&quot;: &quot;CDS&quot;, &quot;gene&quot;:
&quot;L&quot;, &quot;product&quot;: &quot;polymerase&quot;, &quot;locus&quot;:
&quot;&quot;}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Database: ref/features</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fields: query acc.ver, subject acc.ver, % identity, alignment length,
mismatches, gap opens, q. start, q. end, s. start, s. end, evalue,
bit score</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
4 hits found</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AAD14589.1  97.500  120  3  0  1  120  1      120    6.88e-56 
206</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.500  120  3  0  1  120  81     200    6.88e-56 
206</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 L           97.500  120  3  0  1  120  81     200    6.88e-56 
206</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.500  120  3  0  1  120  11581  11700  6.88e-56  206</font></font></p>
<p class="first-paragraph">We can even add custom fields to the
output. Run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-help</font></font>, then scroll through the output to find more
details:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qseqid
means Query Seq-id</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qgi
means Query GI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qacc
means Query accession</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qaccver
means Query accesion.version</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qlen
means Query sequence length</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sseqid
means Subject Seq-id</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sallseqid
means All subject Seq-id(s), separated by a ';'</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sgi
means Subject GI</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sallgi
means All subject GIs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The command now reads:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/1976 -query L.fa  -outfmt &quot;6 qseqid sseqid pident&quot;</font></font></p>
<p class="first-paragraph">That produces reduced information just the
query id, subject id, percent identity:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AAD14589.1  97.500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 L           97.500</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.500</font></font></p>
<p class="first-paragraph">We can see that the sequence for gene <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L</font></font>
is conserved at 97.5% (over the region we have selected).</p>
<h2 class="western"><a name="analyzing-the-full-gene"></a>78.7
Analyzing the full gene</h2>
<p class="first-paragraph">Now let&rsquo;s redo the same analysis but
now taking the full sequence of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L</font></font>
to produce the number of mismatches</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
2014.gb | bio fasta --gene L &gt; L.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/1976 -query L.fa  -outfmt &quot;6 qseqid sseqid pident
mismatch gapopen gaps&quot;</font></font></p>
<p class="first-paragraph">we note that from 1976 to 2014 this
particular gene has not seemed to have acquired any gaps (in the
aligned region):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
AAD14589.1 97.454 169    0  0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
AF086833.2 97.454 169    0  0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
L  97.454 169    0  0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
AF086833.2 97.454 169    0  0</font></font></p>
<p class="first-paragraph">Is our query fully covered? Again we refer
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-he and the</font></font>outfmt` parameters displayed there, which
show us the following fields:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">qcovs
means Query Coverage Per Subject</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qcovhsp
means Query Coverage Per HSP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">qcovus
means Query Coverage Per Unique Subject (blastn only)</font></font></p>
<p class="first-paragraph">here we want the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">qcovhsp</font></font>
that describes each separate alignment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/features -query L.fa  -outfmt &quot;6 qseqid sseqid pident
qcovhsp&quot;</font></font></p>
<p class="first-paragraph">that finally prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AAD14589.1  97.454  100</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.454  100</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 L           97.454  100</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AIG96639.1
 AF086833.2  97.454  100</font></font></p>
<p class="first-paragraph">we can see that the entire query is
present in the targets.</p>
<h2 class="western"><a name="what-are-blast-tasks"></a>78.8 What are
blast tasks?</h2>
<p class="first-paragraph">Tasks are an algorithmic modification of
the method (or its parameters) that make a specific blast tool better
suited for finding certain types of alignments.</p>
<p>Confusingly, the task itself may be named the same as the tool.
For example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
can have the following tasks:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
	- finds more divergent sequences</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">megablast</font></font>
	- finds less divergent sequences</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn-short</font></font>
	- short queries</p>
</ul>
<p class="first-paragraph">Even more confusingly, the default task
for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
tool is the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">megablast</font></font>
task.</p>
<p>Let&rsquo;s make a shorter sequence than before, again; take only
12 bases at the beginning of the genome:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Take the first 12 bases only)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ref/2014.gb | bio fasta -end 12 &gt; short.fa</font></font></p>
<p class="first-paragraph">where the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">short.fa</font></font>
file is now:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;KM233118.1
{&quot;title&quot;: &quot;Zaire ebolavirus isolate Ebola
virus/H.sapiens-wt/SLE/2014/Makona-NM042.3, complete genome&quot;,
&quot;type&quot;: &quot;source&quot;}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AATCATACCTGG</font></font></p>
<p class="first-paragraph">The search now is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ref/1976 -query short.fa</font></font></p>
<p class="first-paragraph">and it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*****
No hits found *****</font></font></p>
<p class="first-paragraph">We have to run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>
with a different search strategy in this case <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn-short</font></font>
to produce a hit:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-task blastn-short -db ref/1976 -query short.fa -outfmt 6</font></font></p>
<p class="first-paragraph">but as you can see, it generates many hits
as short sequences may be present in multiple locations in the
genome.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  9  0  0  2  10  1260   1252   0.66 
18.3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 VP30        100.000  9  0  0  2  10  1260   1252   0.66 
18.3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  9  0  0  2  10  9547   9539   0.66 
18.3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  1  8   10897  10904  2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  2  9   12042  12035  2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  5  12  14724  14731  2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  1  8   14988  14981  2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AAD14589.1  100.000  8  0  0  2  9   462    455    2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AAD14589.1  100.000  8  0  0  5  12  3144   3151   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AAD14589.1  100.000  8  0  0  1  8   3408   3401   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  2  9   542    535    2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  5  12  3224   3231   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  1  8   3488   3481   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 L           100.000  8  0  0  2  9   542    535    2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 L           100.000  8  0  0  5  12  3224   3231   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 L           100.000  8  0  0  1  8   3488   3481   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AAD14588.1  100.000  8  0  0  1  8   553    560    2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 AF086833.2  100.000  8  0  0  1  8   1013   1020   2.6  
16.4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KM233118.1
 VP24        100.000  8  0  0  1  8   1013   1020   2.6   16.4</font></font></p>
<h2 class="western"><a name="will-blast-find-all-alignments"></a>78.9
Will blast find all alignments?</h2>
<p class="first-paragraph">You should know that the default operation
of BLAST was tuned to the most common use cases. These common use
cases can produce unexpected results that sometimes seem to defy
common sense. Here is an example, let&rsquo;s take chromosome 1 of
the yeast genome.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The directory to store blast databases in.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ref</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fetch only the chromosomal sequence for chromosome 1 of the yeast
genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch NC_001133 --format fasta &gt; ref/yeast.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a blast database out of chromosome 1 of the yeast
genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-dbtype nucl -in ref/yeast.fa  -out ref/yeast</font></font></p>
<p class="first-paragraph">Get the first 60 base pairs a query from
the first two lines of chromosome 1 of the yeast genome (70 bases).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ref/yeast.fa | bio fasta -e 60 &gt; start.fa</font></font></p>
<p class="first-paragraph">the file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">start.fa</font></font>
contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001133.9
Saccharomyces cerevisiae S288C chromosome I, complete
sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACACATCCTAACA</font></font></p>
<p class="first-paragraph">Let&rsquo;s query the yeast genome for
this sequence:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
 -db ref/yeast -query start.fa</font></font></p>
<p class="first-paragraph">you get:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*****
No hits found *****</font></font></p>
<p class="first-paragraph">Do a pairwise alignment with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
 -subject ref/yeast.fa -query start.fa</font></font></p>
<p class="first-paragraph">it returns:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*****
No hits found *****</font></font></p>
<p class="first-paragraph">Now it is the time that desperation sets
in! We know that the sequence is in the genome, we took it from the
genome! Look:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ref/yeast.fa | head -2</font></font></p>
<p class="first-paragraph">The start of the sequence is the same
sequence as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">start.fa</font></font>.
Why won&rsquo;t BLAST find it???????</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;NC_001133.9
Saccharomyces cerevisiae S288C chromosome I, complete
sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACACATCCTAACA</font></font></p>
<p class="first-paragraph"><img src="0000_html_9b7ce1aafc375082.png" name="Image128" align="bottom" hspace="12" width="560" height="175" border="0"/>
</p>
<p>Let&rsquo;s do a local alignment with an optimal aligner as in
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">local-align.sh</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align ref/yeast.fa start.fa  --local</font></font></p>
<p class="first-paragraph">again it works, it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
NC_001133.9 (60) vs NC_001133.9 (60)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
pident=100.0% len=60 ident=60 mis=0 del=0 ins=0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
local: score=60.0 match=1 mismatch=2 gapopen=11
gapextend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACA</font></font></p>
<p class="first-paragraph">The above is the match we expected to see.
So BLAST, what&rsquo;s up?</p>
<p>Here is what happens, the query sequence has repetitive elements
(also called low complexity regions), and BLAST automatically and
silently filters out hits to these. Look at the sequence, see how
slight variation is there within the bases?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACACATCCTAACA</font></font></p>
<p class="first-paragraph">Since BLAST is tuned to help scientists
with little computational background, the developer of BLAST turned
off presenting these. Was that a wise decision? We&rsquo;ll let you
think about that, for now, let mention that to turn off this behavior
you will need to set the flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-dust
no</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
 -db ref/yeast -query start.fa -dust no</font></font></p>
<p class="first-paragraph">it now produces the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Score
= 111 bits (60),  Expect = 2e-27</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Identities = 60/60 (100%), Gaps = 0/60 (0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
Strand=Plus/Plus</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Query
 1   CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACA 
60</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         
||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Sbjct
 1   CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACCCACACACACA  60</font></font></p>
<p class="first-paragraph">Let this be a reminder to all of how easy
it is to end up with unintended consequences.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="blast-databases"></a>79 BLAST databases</h1>
<p class="first-paragraph">For BLAST to operate successfully, it
needs to build a database of the target sequences. Different
databases are necessary for nucleotide and peptide sequences.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb</font></font>
	creates blast databases.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd</font></font>
	queries blast databases.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">update_blastdb.pl</font></font>
	updates prebuilt blast databases.</p>
</ul>
<h2 class="western"><a name="how-do-we-make-a-blast-database"></a>79.1
How do we make a BLAST database?</h2>
<p class="first-paragraph">Let create a BLAST database. Get all
proteins for the Ebola project <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>
described in <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Genomic
surveillance elucidates Ebola virus origin</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a directory to store the blast database</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p db</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the proteins deposited at NCBI for project
PRJNA257197</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db protein -query PRJNA257197 | efetch -format fasta &gt;
db/proteins.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the nucleotides deposited at NCBI for project
PRJNA257197</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nuccore -query PRJNA257197 | efetch -format fasta &gt;
db/genomes.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Build the blast index for proteins.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-in db/proteins.fa -dbtype prot -out db/ebov_prots -parse_seqids</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Build the blast index for proteins.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb
-in db/genomes.fa -dbtype nucl -out db/ebov_genomes -parse_seqids</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get information on the database</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db db/ebov_prots -info</font></font></p>
<p class="first-paragraph">We then use our custom database as any
other blast search would. Suppose we have a query like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;mystery</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGGACTCTCGTCCTCAGAAAGTCTGGATGACGCCGAGTCTCACTGAATCTGACATGGAT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TACCACAAGATCTTGACAGCAGGTCTGTCCGTTCAACAGGGGGTTGTTCGGCAAAGAGTC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATCCCAGTGTATCAAGTAAACAATCTTGAG</font></font></p>
<p class="first-paragraph">that you can create the query with your
editor or from the command line with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a mystery query</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;&gt;mystery&quot; &gt; myst.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;ATGGACTCTCGTCCTCAGAAAGTCTGGATGACGCCGAGTCTCACTGAATCTGACATGGAT&quot;
&gt;&gt; myst.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;TACCACAAGATCTTGACAGCAGGTCTGTCCGTTCAACAGGGGGTTGTTCGGCAAAGAGTC&quot;
&gt;&gt; myst.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;ATCCCAGTGTATCAAGTAAACAATCTTGAG&quot; &gt;&gt; myst.fa</font></font></p>
<p class="first-paragraph">We can un blast on the mystery sequence
using different strategies.</p>
<p>Search for matches as nucleotides:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db db/ebov_genomes  -query myst.fa -outfmt '6 qacc sacc pident qlen
length' | sort -k 3rn | head -3</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233075    99.333  150 150</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233035    98.667  150 150</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233036    98.667  150 150</font></font></p>
<p class="first-paragraph">Search for matches as nucleotides matching
proteins:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastx
-db db/ebov_prots  -query myst.fa -outfmt '6 qacc sacc pident qlen
length' | sort -k 3rn | head -3</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
AIG96244    100.000 150 50</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
AKC36407    98.000  150 50</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
AKC36227    98.000  150 50</font></font></p>
<p class="first-paragraph">Note how the protein blast produces higher
percent identity over the entire query!</p>
<p>Search for matches as translations of</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tblastx
-db db/ebov_genomes -query myst.fa -outfmt &quot;6 qacc sacc pident
qlen length frames&quot; | sort -k 3rn | head -3</font></font></p>
<p class="first-paragraph">above we ask for the translation frames to
be indicated as well:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233075    100.000 150 36  3/3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233075    100.000 150 49  -2/-3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mystery
KM233075    100.000 150 50  1/1</font></font></p>
<h2 class="western"><a name="how-to-download-prebuild-blast-databases"></a>
79.2 How to download prebuild BLAST databases?</h2>
<p class="first-paragraph">While we can always create our own blast
databases with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">makeblastdb</font></font>,
this task may take a considerable amount of time (days or even weeks)
when the total number of bases to be indexed is very large.</p>
<p>To facilitate the process, prebuilt databases can be downloaded
from the NCBI webpage.</p>
<p><font color="#4f81bd"><a href="https://ftp.ncbi.nlm.nih.gov/blast/db/">https://ftp.ncbi.nlm.nih.gov/blast/db/</a></font></p>
<h2 class="western"><a name="can-we-automate-the-download-of-blast-databases"></a>
79.3 Can we automate the download of BLAST databases?</h2>
<p class="first-paragraph">NCBI offers ready-made databases for many
organisms and even the entire non-redundant database of all known
sequences is available for download.</p>
<p>Make a dedicated storage for your blast databases:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This directory will store all downloaded databases.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Note how it opens from your home directory.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/blastdb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Switch to that folder.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~/blastdb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The blast package includes a script that can be used to download</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
ready made databases. List them all:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">update_blastdb.pl
--showall</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the 16 microbial database.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">update_blastdb.pl
16S_ribosomal_RNA --decompress</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the taxonomy database.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">update_blastdb.pl
taxdb --decompress</font></font></p>
<p class="first-paragraph">We can also tell blast where to look for
files via the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BLASTDB</font></font>
variable. When you have the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BLASTDB</font></font>
variable set then using blast tools becomes simpler, you would not
need add the full path to the BLAST database, specifying the name
would be enough.</p>
<p>In addition, this is required for the taxonomy searches to work.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">export
BLASTDB=$BLASTDB:~/blastdb</font></font></p>
<p class="first-paragraph">If the above is set then instead of having
to set the full path as in:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db ~/blastdb/16S_ribosomal_RNA -info</font></font></p>
<p class="first-paragraph">BLAST can access the databases by name
alone like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -info</font></font></p>
<p class="first-paragraph">In addition, when the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BLASTDB</font></font>
variable is set the taxonomy information can be also accessed by
BLAST.</p>
<p>You can add the setting to your bash profile so that it loads
automatically in each terminal:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;export BLASTDB=$BLASTDB:~/blastdb&quot; &gt;&gt; ~/.bashrc</font></font></p>
<h2 class="western"><a name="what-information-can-be-obtained-from-a-blast-database"></a>
79.4 What information can be obtained from a BLAST database?</h2>
<p class="first-paragraph">We can extract a wide variety of
interesting information from blast databases. Sometimes it is worth
creating a BLAST database just to extract specific information from
sequences.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get some information on the 16S Microbial database.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -info</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
What is the first sequence in the file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -entry 'all' | head -1</font></font></p>
<p class="first-paragraph">Prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118889</font></font></p>
<p class="first-paragraph">Quick diversion, to show you the power of
entrez direct: which publication links to accesion number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118889</font></font>?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nuccore -query NR_118889.1 | elink -target pubmed | efetch
-format abstract</font></font></p>
<p class="first-paragraph">or you can download the accession number
itself and investigate the genbank file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NR_118889 -format gb &gt; NR_118889.gb</font></font></p>
<p class="first-paragraph">Back to BLAST, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd</font></font>
is a command with high utility. Read the help on it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-help | more</font></font></p>
<h2 class="western"><a name="how-do-i-reformat-sequences-in-a-blast-database"></a>
79.5 How do I reformat sequences in a BLAST database?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-outfmt</font></font>
parameter of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd</font></font>
allows printing a variety of information. For example, the following
will print the sequence ids and their length:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -entry 'all' -outfmt '%a %l'</font></font></p>
<p class="first-paragraph">There are many more formatting options
(see the detailed help for the tool) that can be used to answer
potentially interesting questions:</p>
<p>We replace the &lsquo;,&rsquo; (comma) separator with tabs since
it is easier to manipulate files with tabs than with commas. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tr</font></font>
command translates each comma into a tab</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -entry 'all' -outfmt '%a,%l,%T,%L' | tr ','
'\t' &gt; 16S.txt</font></font></p>
<p class="first-paragraph">This file lists the genes, lengths, and
common taxonomical ranks (this last part will only work if your
taxonomy file has been downloaded as above and if the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BLASTDB</font></font>
variable has been set properly).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118889.1
1300    36819   Amycolatopsis azurea</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118899.1
1367    1658    Actinomyces bovis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_074334.1
1492    224325  Archaeoglobus fulgidus DSM 4304</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118873.1
1492    224325  Archaeoglobus fulgidus DSM 4304</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_119237.1
1492    224325  Archaeoglobus fulgidus DSM 4304</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118890.1
1331    1816    Actinokineospora fastidiosa</font></font></p>
<p class="first-paragraph">What organism has the longest 16S gene?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
16S.txt | sort -k2,2 -rn | head</font></font></p>
<p class="first-paragraph">This sorts the file above by the second
column:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_046384.1
3600    1104324 Pyrobaculum ferrireducens</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_041958.1
2211    178306  Pyrobaculum aerophilum str. IM2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_041961.1
2207    121277  Pyrobaculum arsenaticum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_042068.1
2197    56636   Aeropyrum pernix</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_042067.1
2192    70771   Pyrobaculum neutrophilum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG_041957.1
2192    698757  Pyrobaculum oguniense TE7</font></font></p>
<h2 class="western"><a name="how-do-i-extract-a-specific-entry-from-a-database"></a>
79.6 How do I extract a specific entry from a database?</h2>
<p class="first-paragraph">Pass the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-entry
accession</font></font> parameter to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd</font></font>.
Here, the particular blast database construction choices become
essential, specifically whether or not the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-parse_seqids</font></font>
was used. Accession numbers are only available in the latter case.
Otherwise, the full sequence name needs to be used.</p>
<p>Get the first 20 bases of a specific 16S gene. We are asking for
one specific accession number:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -entry 'NR_118889.1' -range 1-20</font></font></p>
<p class="first-paragraph">This produces the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;gi|645322056|ref|NR_118889.1|:1-20
Amycolatopsis azurea strain NRRL 11412 16S ribosomal RNA gene,
partial sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGTCTNATACCGGATATAAC</font></font></p>
<p class="first-paragraph">We can format the output for it as well.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db ~/refs/refseq/16S_ribosomal_RNA -entry 'NR_118889.1' -range 1-20
-outfmt &quot;%s&quot;</font></font></p>
<p class="first-paragraph">This produces the following output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGTCTNATACCGGATATAAC</font></font></p>
<h2 class="western"><a name="how-do-i-process-all-entries-in-a-database"></a>
79.7 How do I process all entries in a database?</h2>
<p class="first-paragraph">Pass the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-entry
all</font></font> parameter to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd</font></font>
to get, for example, the first 50 bases of each gene.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16S_ribosomal_RNA -entry 'all' -range 1-50 -outfmt &quot;%s&quot;
 &gt; starts.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
starts.txt | head</font></font></p>
<p class="first-paragraph">This will produce a file containing the
first 50 bases of each 16S gene, like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGTCTNATACCGGATATAACAACTCATGGCATGGTTGGTAGTGGAAAGCT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGGTGAGTAACACGTGAGTAACCTGCCCCNNACTTCTGGATAACCGCTTG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATTCTGGTTGATCCTGCCAGAGGCCGCTGCTATCCGGCTGGGACTAAGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATTCTGGTTGATCCTGCCAGAGGCCGCTGCTATCCGGCTGGGACTAAGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATTCTGGTTGATCCTGCCAGAGGCCGCTGCTATCCGGCTGGGACTAAGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TACTTTGGGATAAGCCTGGGAAACTGGGTCTNATACCGGATATGACAACT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTGAACGGAGAGTTCGANCCTGGCTCAGGATGAACGCTGGCGGCGCGCCT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTGAGTGGCGAACGGGTGAGTAACACGTGGGTAACCTTCNNTGTACTTTG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTGAGTAACCTGCCCCTTCTTCTGGATAACCGCATGAAAGTGTGGCTAAT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACCGGATANGACCACTNATCGCATNTCGGTGGGTGGAAAGTTTTTTCGGT</font></font></p>
<p class="first-paragraph">How many unique 16S gene starts of length
50 are there?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
starts.txt | sort | uniq -c | sort -rn | head</font></font></p>
<p class="first-paragraph">The output of the above is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">381
AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
184 AGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCCTAATACATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
125 GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGTAAGGCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
101 GACGAACGCTGGCGGCGTGCTTAACACATGCAAGTCGAGCGGAAAGGCCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 97 ATTGAACGCTGGCGGCATGCCTTACACATGCAAGTCGAACGGCAGCACGG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 95 TAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAACACAT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 90 AGAGTTTGATCCTGGCTCAGAACGAACGCTGGCGGCAGGCTTAACACATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 90 AGAGTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 88 AGAGTTTGATCCTGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACATG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 87 ATTCCGGTTGATCCTGCCGGAGGCCATTGCTATCGGAGTCCGATTTAGCC</font></font></p>
<p class="first-paragraph">The above result shows that out of all 16S
genes in the NCBI database 298 have the same 50 basepairs as starting
sequences.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="short-read-aligners"></a>80 Short read
aligners</h1>
<p class="first-paragraph">Sequencing used to be a costly
proposition. The resulting few, and relatively long, sequences (1000
bp) were very &ldquo;precious&rdquo; and were meant to be studied
extensively. For that reason, aligners were initially designed to
perform near-optimal alignments and to enumerate most if not all
alternative arrangements that all fell within a level tolerance. The
landscape changed dramatically around 2005 with the arrival of the
so-called high throughput short-read technologies.</p>
<h2 class="western"><a name="how-are-short-reads-different-from-long-reads"></a>
80.1 How are short reads different from long reads?</h2>
<p class="first-paragraph">Modern sequencing instruments brought
about two significant changes:</p>
<ol>
	<li/>
<p class="compact">The read lengths became very short
	(50-300bp)</p>
	<li/>
<p class="compact">The number of reads grew extremely high
	(hundreds of millions)</p>
</ol>
<p class="first-paragraph">Hence, whereas initial algorithms were
designed to find most (if not all) alignments, scientists needed
newer and faster methods that could very quickly select the best
location for each of the measurements at the expense of not
investigating all potential alternative alignments.</p>
<p>The rationale behind the new paradigm was to treat sequencing
reads as small, independent measurements to be subsequently matched
either against a known genome (called re-sequencing) or against one
another (called assembly). This new use case and its requirements
lead to the development of an entire subfield of alignment
algorithms, often referred to as &ldquo;short read aligners&rdquo; or
&ldquo;short read mappers.&rdquo;</p>
<h2 class="western"><a name="what-are-short-read-aligners"></a>80.2
What are short read aligners?</h2>
<p class="first-paragraph">Short read aligners software tools
designed to align a vast number of short reads (billions, or <font color="#4f81bd"><a href="https://www.nature.com/articles/s41588-018-0273-y">lately
even trillions</a></font>). The dominance of the Illumina
instrumentation means that most short-read software were designed to
work best for sequences of lengths in the range of 50 to 300 bp that
the Illumina instruments produce.</p>
<h2 class="western"><a name="how-do-short-read-aligners-work"></a>80.3
How do short read aligners work?</h2>
<p class="first-paragraph">First, let us make something abundantly
clear: short read aligners are marvels of modern scientific software
engineering. A well-optimized short read aligner can match over ten
thousand sequences per second (!) against the 3 billion bases of the
human genome. Their performance, implementation, and level of
execution is nothing short of astounding.</p>
<p>But there is another side to the story. The rational expectation
is that the published aligners would produce reasonably similar
alignments and that the difference between various implementation
would manifest primarily in their performance, hardware requirements,
and small fluctuations in accuracy and precision.</p>
<p>In reality, the results that two aligners produce can be
substantially different. Surprisingly, it seems almost impossible to
unambiguously understand, describe, and characterize the changes in
the results that different algorithm&rsquo;s implementations provide.
While there is no shortage of scientific publications that claim to
compare the accuracy and performance of various tools, most of these
papers fail to capture the essential differences between the methods.
It is not for lack of trying - accurately characterizing the
&ldquo;quality&rdquo; of alignments turned out to be a lot more
complicated than anticipated.</p>
<p>A typical bioinformatics analysis requires finding exact matches,
missing locations, and partial and alternative matches/overlaps. An
method that is efficient at one of these tasks trades that
performance gain for inefficiency in another requirement.</p>
<p>Hence, the performance of software methods depends critically on
the type of data they are used to analyze, which, in turn, is
domain-specific. Different alignment methods will, therefore, be best
suited to different domains.</p>
<h2 class="western"><a name="what-are-the-limitations-of-short-read-aligners"></a>
80.4 What are the limitations of short read aligners?</h2>
<ol>
	<li/>
<p class="compact">Most short read aligners will find only
	alignments that are reasonably similar to the target. The algorithm
	will generally &ldquo;gives up&rdquo; searching beyond a certain
	threshold.</p>
	<li/>
<p class="compact">Short read aligners are designed to find
	regions of high similarity.</p>
	<li/>
<p class="compact">Most short read aligners typically cannot
	handle long reads or become inefficient when doing so.</p>
	<li/>
<p class="compact">There is also a limit to how short a read
	can be.</p>
</ol>
<p class="first-paragraph">The minimum length for the read is
algorithm- and implementation-dependent; it is rarely (if ever)
mentioned in the documentation. Many tools stop working correctly
when the read lengths drop under 30 bases. When studying small RNAs,
for example, microRNAs, we have to look for tools that can align very
short reads; the selection is surprisingly sparse.</p>
<h2 class="western"><a name="what-is-read-mapping"></a>80.5 What is
read mapping?</h2>
<p class="first-paragraph">Conceptually, alignment and mapping appear
to mean the same thing, but there are subtle differences. The word
&ldquo;mapper&rdquo; is often used to emphasize that the optimal
alignment of a read is not guaranteed. The purpose of a mapper tool
is locating a region in a genome, not producing an optimal alignment
to that region.</p>
<p>The following definition is taken from a presentation by Heng Li:</p>
<p class="block-text-western">Mapping</p>
<ul>
	<li/>
<p class="block-text-western">A mapping is a region where a
	read sequence is placed.</p>
	<li/>
<p class="block-text-western">A mapping is regarded to be
	correct if it overlaps the true region.</p>
</ul>
<p class="block-text-western">Alignment</p>
<ul>
	<li/>
<p class="block-text-western">An alignment is the detailed
	placement of each base in a read.</p>
	<li/>
<p class="block-text-western">An alignment is regarded to be
	correct if each base is placed correctly.</p>
</ul>
<p class="first-paragraph">Interestingly, and we invite the reader to
come up with their examples, we could have a situation where a read
has a correct mapping with an incorrect alignment or vice versa - has
an incorrect mapping with a proper alignment.</p>
<p>Ideally, of course, the mapping and the alignment should coincide
&ndash;&ndash; but it&rsquo;s important to remember that this is not
always the case.</p>
<h2 class="western"><a name="how-do-i-distinguish-between-a-mapper-and-an-aligner"></a>
80.6 How do I distinguish between a mapper and an aligner?</h2>
<p class="first-paragraph">It used to be that tools could be readily
categorized into the two classes of aligner vs.&nbsp;mapper. As the
field advanced and matured, however, the read lengths became longer,
and the software tools started to make use of combined techniques
that made tools behave as both mappers and aligners.</p>
<p>It is essential, however, to remember that the distinction between
mapping and alignment does exist, and to recognize further that
different studies have different requirements in regards to the use
of these concepts.</p>
<p>For example, studies examining SNPs and variations in a genome
would be primarily alignment-oriented. By contrast, studies using
ChIP-Seq data would be essentially mapping-oriented.</p>
<h2 class="western"><a name="how-do-we-pick-the-best-short-read-aligner"></a>
80.7 How do we pick the best short read aligner?</h2>
<p class="first-paragraph">So how do we pick the best aligner? There
are good starting choices, such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>,
that will perform well in all domains of application. Later, you may
want to explore other tools based on observed performance, reviews of
scientific literature, and the experience of other scientists
studying the same type of data. The optimal choice will be specific
to the data you are analyzing.</p>
<p>In the end, the choice of the aligner is a bit &ldquo;faith-based,&rdquo;
coupled to personal observations and empirical evidence. There is
also a cultural and historical element to it. Scientists at the Broad
Institute will likely use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
because it was developed there; others who work at the Beijing
Genomics Institute will probably prefer the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">novoalign</font></font>
tool because as it was created at that institute, and so on.</p>
<p>Then there is an individualistic feel to almost every
high-throughput aligner. Typically, a sole genius is behind each, an
individual with uncommon and extraordinary programming skill, a
person that has set out to solve a problem that is important to them.
Due to the sheer complexity of the task and requirements, the
software&rsquo;s unique characteristics will reflect their developer
personality, virtues, and some of their quirks, as well.</p>
<p>You don&rsquo;t just run an aligner; you perform a Vulcan style
&ldquo;<font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Vulcan_(Star_Trek)#Mind_melds">mind-meld</a></font>&rdquo;
(see Star Trek) with the author who wrote it.</p>
<h2 class="western"><a name="what-features-do-we-look-for-in-a-short-read-aligner"></a>
80.8 What features do we look for in a short read aligner?</h2>
<p class="first-paragraph">When choosing an aligner, you will need to
perform a mental checklist of what features the aligner of your
choice should provide:</p>
<ul>
	<li/>
<p class="compact">Alignment algorithm: global, local, or
	semi-global?</p>
	<li/>
<p class="compact">Can the aligner filter alignments based on
	external parameters?</p>
	<li/>
<p class="compact">Can aligner report more than one alignment
	per query?</p>
	<li/>
<p class="compact">How will the aligner handle INDELs
	(insertions/deletions)?</p>
	<li/>
<p class="compact">Can the aligner skip (or splice) over
	intronic regions?</p>
	<li/>
<p class="compact">Will the aligner find chimeric alignments?</p>
</ul>
<p class="first-paragraph">Looking at this list above, it is probably
clear what makes comparing different tools so tricky. Various
techniques excel for different requirements; moreover, as we shall
see, we are only scratching the surface when it comes to the
information that alignment record may contain.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-bwa-aligner"></a>81 The bwa aligner</h1>
<p class="first-paragraph">To run the examples install the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package developed for this book.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<p class="first-paragraph">Read more about how the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
align</font></font> method works here at
<font color="#4f81bd"><a href="https://www.bioinfo.help/bio-align.html">https://www.bioinfo.help/bio-align.html</a></font></p>
<h2 class="western"><a name="what-is-bwa"></a>81.1 What is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
(aka Burrows-Wheelers Aligner) aligner was written by <font color="#4f81bd"><a href="http://lh3lh3.users.sourceforge.net/">Heng
Li</a></font>, a research scientist at the Broad Institute. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
aligner is possibly the most widely used short read alignment tool
and is well suited for a broad number of applications.</p>
<ul>
	<li/>
<p class="compact">Github page: <font color="#4f81bd"><a href="https://github.com/lh3/bwa">https://github.com/lh3/bwa</a></font></p>
</ul>
<p class="first-paragraph">Interestingly, the algorithm that the
latest version of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
implements has not yet been published in a peer-reviewed journal (as
of 2016). There is a <font color="#4f81bd"><a href="https://gist.github.com/ialbert/3164967c853b7fd8f44e">fascinating
story of why that is so</a></font>, and it illustrates the
shortcomings of the &ldquo;scientific method&rdquo; itself. The new
algorithm is called the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font> algorithm (where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mem</font></font>
stands for Maximally Exact Matches).</p>
<h2 class="western"><a name="how-does-bwa-work"></a>81.2 How does bwa
work?</h2>
<p class="first-paragraph">In general, all short read aligners
operate on the same principles:</p>
<ol>
	<li/>
<p class="compact">First build an &ldquo;index&rdquo; from a
	reference genome (this only needs to be done once).</p>
	<li/>
<p class="compact">The reads in FASTA/FASTQ files are then
	aligned against the index created in step 1.</p>
</ol>
<p class="first-paragraph">Index building consists of pre-processing
the reference genome so that the program can search it efficiently.
Each program will build a different type of index; sometimes it may
produce multiple files with odd-looking names or extensions. For this
reason, it is best to place the reference genome in separate
directories.</p>
<p>The time and computational resources required for building an
index will depend on the genome size. These times can be substantial
(many days or more) for large genomes. For some software it is
possible to download pre-built indices from their websites.</p>
<h2 class="western"><a name="how-do-i-use-bwa"></a>81.3 How do I use
bwa?</h2>
<p class="first-paragraph">Suppose you were interested in processing
the sequencing data from <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font>
(#ebolaredo). And that you wished to investigate the changes relative
to the 1976 outbreak. We have found that the accession number for the
1976 strain is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>
and that the project id is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>.</p>
<p>Make a new directory for the reference genomes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font></p>
<p class="first-paragraph">Get the ebola genome in GenBan format, the
extract the source (genome) only:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch AF086833 &gt; refs/AF086833.gb</font></font></p>
<p class="first-paragraph">Transform GenBank to FASTA</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
AF086833.gb | bio fasta &gt; refs/AF086833.fa</font></font></p>
<p class="first-paragraph">Build an index with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index refs/AF086833.fa</font></font></p>
<p class="first-paragraph">The indexing above will finish rather
quickly, as the genome is a mere 18K base long. Note that in scripts
you should make use of shell variables to store names. Doing so will
allow you to keep the command the same and only change the value of
the variable. For example, we could have written a generic script
that can be later reused, a script prepares a bwa index for any
accession number:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Stop on any error</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-uex</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The accession number for the reference genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the directory for reference</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The GenBank reference file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GBK=refs/$ACC.gb</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The FASTA reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/$ACC.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fetch the sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch $ACC &gt; $GBK</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Transform GenBank into FASTA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fasta $GBK &gt; $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Build an index for the genome so that we can view in IGV</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Build the index.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
program provides a neat self-discovery of its parameters. It is
generally run as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
command</font></font>, and it will perform very different tasks
depending on which command is used. To discover all of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
can run execute it on its own like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font></p>
<p class="first-paragraph">Then we can find out more on each command,
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font></p>
<p class="first-paragraph">Note the instructions that it provides. To
align reads, we need sequence data files. Let&rsquo;s obtain data
from SRA. Let&rsquo;s find out all runs for the Ebola project.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA257197  | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">Pick a run from this file, say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739</font></font>,
and we&rsquo;ll only subset the data to 10K reads to get through
quickly:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972739</font></font></p>
<p class="first-paragraph">We can run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
by listing the index first and the sequencing reads as second
parameters:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem refs/AF086833.fa SRR1972739_1.fastq &gt; output.sam</font></font></p>
<p class="first-paragraph">The command above is a bit hard to read
and when listing paired reads its in only getting more complicated.
We can create shortcuts to each of our files. This will allow us to
shorten and simplify the commands greatly, now our separate,
alignment script looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Stop on error</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-uex</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/$ACC.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=SRR1972739_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=SRR1972739_2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 &gt; output.sam</font></font></p>
<p class="first-paragraph">You can always check what the variable
stands for with,</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
$R1</font></font></p>
<p class="first-paragraph">To align one of the read pairs with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font> in single end mode, we could run the command with:</p>
<p>The resulting file is in a so-called SAM (Sequence Alignment Map)
format, one that you will undoubtedly love and hate (all at the same
time). It is one of the most recent bioinformatics data formats, one
that by today has become the standard method to store and represent
all high-throughput sequencing results. The SAM file looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SQ
SN:gi|10141003|gb|AF086833.2|   LN:18959</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@PG
ID:bwa  PN:bwa  VN:0.7.12-r1039 CL:bwa mem
/Users/ialbert/refs/ebola/1976.fa SRR1972739_1.fastq
SRR1972739_2.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1
   83  gi|10141003|gb|AF086833.2|  15684   60  69M32S  =   15600  
-153   
TTTAGATTTAACAAGATACCGAGAAAATGAATTGATTTATGACAATAATCCTCTAAAAGGAGGACTCAAATGAGATATTGCAATTGAGTCCTCCTTTTAGA
 
DDDDDEEEEEDEEFFFEDHHHHIJJJJJJJJJJJJJJJJJIGIJJJJJJJJJJJJJJJJJJJIGIGJJJJJJJJJJJJJJIJJJJJJIFHHHHFFFFFCCC
  NM:i:2  MD:Z:27C16G24   AS:i:59 XS:i:0 
SA:Z:gi|10141003|gb|AF086833.2|,15735,+,33M68S,60,0;</font></font></p>
<p class="first-paragraph">A SAM file encompasses all known
information about the sample and its alignment; typically, we never
look at the FastQ file again, since the SAM format contains all (well
almost all) information that was also present in the FastQ
measurements.</p>
<h2 class="western"><a name="how-to-get-information-on-a-sam-file"></a>
81.4 How to get information on a SAM file</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat output.sam</font></font></p>
<p class="first-paragraph">that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10351
+ 0 in total (QC-passed reads + QC-failed reads)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 primary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 secondary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">351
+ 0 supplementary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 duplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 primary duplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7571
+ 0 mapped (73.14% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7220
+ 0 primary mapped (72.20% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 paired in sequencing</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 read1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 read2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 properly paired (N/A : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with itself and mate mapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 singletons (N/A : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr (mapQ&gt;=5)</font></font></p>
<h2 class="western"><a name="aligning-paired-end-reads"></a>81.5
Aligning paired end reads</h2>
<p class="first-paragraph">Since this is a paired-end dataset, the
proper way to align the data is in paired-end mode. For that purpose,
we list both files on the command line:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 &gt; bwa.sam</font></font></p>
<p class="first-paragraph">thus our script will be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Stop on error</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-uex</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/$ACC.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=SRR1972739_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=SRR1972739_2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 &gt; output.sam</font></font></p>
<p class="first-paragraph">Note how the same data can be aligned in
either single end or paired end mode. This latter alignment will
contain more information as pair related information will be added to
the alignment file.</p>
<h2 class="western"><a name="how-do-i-find-help-on-bwa"></a>81.6 How
do I find help on bwa?</h2>
<p class="first-paragraph">To see all the ways that we could tune <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font> alignment run</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font></p>
<p class="first-paragraph">Among the many options, note those that
set the scoring matrix:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-A
INT        score for a sequence match, [...] [1]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-B
INT        penalty for a mismatch [4]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-O
INT[,INT]  gap open penalties for deletions and insertions [6,6]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-E
INT[,INT]  gap extension penalty; a gap of size k cost '{-O} +
{-E}*k' [1,1]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-L
INT[,INT]  penalty for 5'- and 3'-end clipping [5,5]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-U
INT        penalty for an unpaired read pair [17]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">When compared to the more traditional
EDNAFULL matrix discussed in the alignment chapter, we note that in
the default setting, the reward for sequence match is much smaller;
hence, the default alignment is tuned to find long stretches if bases
match.</p>
<p>Gap open penalties are also smaller (6 instead of 10) whereas gap
extension penalties are a bit larger.</p>
<p>There are other parameters where a single word will set multiple
values. For example:</p>
<ul>
	<li/>
<p class="compact">note how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-x
	ont2d</font></font> is equivalent to setting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-k14
	-W20 -r10 -A1 -B1 -O1 -E1 -L0</font></font>. These are the
	recommended settings when aligning data from an Oxford Nanopore
	MinION instrument.</p>
</ul>
<p class="first-paragraph">The way the parameters alter the final
alignment gets complicated. Here is where the choice becomes more &rdquo;
faith-based&rdquo;. We place of faith in the author of the tool. In
this case, we trust that Heng Li knows why these parameters are the
best, even if we could not fully explain their interplay.</p>
<p>What is essential to keep in mind is that decisions may have been
made on our behalf and if things don&rsquo;t go the way we expect, we
know where to look for potential solutions.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-bowtie-aligner"></a>82 The bowtie
aligner</h1>
<p class="first-paragraph">The first version of <font color="#4f81bd"><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie
aligner</a></font> was the first implementation of the
Burrows-Wheeler algorithm for short read alignment, and with that it
has opened a new era in processing high-throughput data.</p>
<p>The latest version of the algorithm <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>,
is almost always preferred. In this book when we talk about the
bowtie program we mean bowtie version 2.</p>
<h2 class="western"><a name="how-does-bowtie-work"></a>82.1 How does
bowtie work?</h2>
<p class="first-paragraph">In general, all short read aligners
operate on the same principles:</p>
<ol>
	<li/>
<p class="compact">First build an &ldquo;index&rdquo; from a
	reference genome (this only needs to be done once).</p>
	<li/>
<p class="compact">The reads in FASTA/FASTQ files are then
	aligned against the index created in step 1.</p>
</ol>
<p class="first-paragraph">Index building consists of pre-processing
the reference genome so that the program can search it efficiently.
Each program will build a different type of index; sometimes it may
produce multiple files with odd-looking names or extensions. For this
reason, it is best to place the reference genome in separate
directories.</p>
<h2 class="western"><a name="how-do-i-build-an-index-with-bowtie"></a>
82.2 How do I build an index with bowtie?</h2>
<p class="first-paragraph">We will align ebola sequencing data
against the 1976 Mayinga reference genome. We recommend keeping
reference genomes in a more general location rather than the local
folder, since you will reuse them frequently. For example, use
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/refs/ebola</font></font>
for all ebola genomes and annotations.</p>
<p>This directory will hold the reference genome and all indices:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Assign the file name to a variable</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The reference genome stored locally.   </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=~/refs/$ACC.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Fetch the sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id $ACC &gt; $REF.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Build an index for the genome so that we can view in IGV</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font></p>
<p class="first-paragraph">For bowtie2, we need to specify both the
input and output prefix name on the command line. Confusingly, these
can also be the same because many prefixes are the same. The first
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">$REF</font></font>
is the reference file; the second <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$REF</font></font>
indicates the prefix name of the index; it could be anything, but it
is a good habit to keep it the same or similar to the actual
reference.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2-build
$REF $REF </font></font>
</p>
<p class="first-paragraph">See for yourself how the ancillary files
multiply like rabbits.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
ref/*</font></font></p>
<h2 class="western"><a name="how-do-i-align-with-bowtie"></a>82.3 How
do I align with bowtie?</h2>
<p class="first-paragraph">We can pick an ebola sequencing run id,
for example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739</font></font>,
and we&rsquo;ll select just 10K reads to get through quickly:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972739</font></font></p>
<p class="first-paragraph">We can create shortcuts to each of our
files. This will allow us to shorten and simplify the commands
greatly.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=SRR1972739_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=SRR1972739_2.fastq</font></font></p>
<p class="first-paragraph">To align with bowtie,</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2
-x $REF -1 $R1 -2 $R2  &gt; bowtie.sam</font></font></p>
<p class="first-paragraph">When bowtie runs, it also gives us a
report:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
reads; of these:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 10000 (100.00%) were paired; of these:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   5086 (50.86%) aligned concordantly 0 times</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   4914 (49.14%) aligned concordantly exactly 1 time</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   0 (0.00%) aligned concordantly &gt;1 times</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   ----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   5086 pairs aligned concordantly 0 times; of these:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     827 (16.26%) aligned discordantly 1 time</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   ----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   4259 pairs aligned 0 times concordantly or discordantly; of
these:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     8518 mates make up the pairs; of these:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       7463 (87.61%) aligned 0 times</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       1055 (12.39%) aligned exactly 1 time</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       0 (0.00%) aligned &gt;1 times</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">62.69%
overall alignment rate</font></font></p>
<p class="first-paragraph">Not to be outdone in the shock and awe
department, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>,
too, can be set up in myriad ways.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font></p>
<p class="first-paragraph">Run and enjoy the view:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[
... lots of lines ... ]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Presets:
                Same as:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 For --end-to-end:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 For --local:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">[
... lots of lines ... ]</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-do-i-compare-aligners"></a>83 How do
I compare aligners?</h1>
<h2 class="western"><a name="how-can-i-tell-which-aligner-is-better"></a>
83.1 How can I tell which aligner is &ldquo;better&rdquo;?</h2>
<p class="first-paragraph">The answer to this is a little more
complicated. As you will see, we can&rsquo;t quite make it a &ldquo;fair&rdquo;
comparison to begin with. Let&rsquo;s try to evaluate the alignment
on simulated sequences:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the reference genome and build an index from it.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
mkdir -p refs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
REF=refs/AF086833.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
efetch -db nuccore -format fasta -id AF086833 &gt; $REF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# Index the reference with bwa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
bwa index $REF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# Index the reference with bowtie2.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
bowtie2-build $REF $REF </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# Simulate 100,000 reads from the genome with 1% error rate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
dwgsim -N 100000 -e 0.01 -E 0.01 $REF data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# Set up shortcuts to the reads.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
R1=data.bwa.read1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
R2=data.bwa.read2.fastq</font></font></p>
<p class="first-paragraph">Align with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem</font></font> while timing the run:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">time
bwa mem $REF $R1 $R2 &gt; bwa.sam</font></font></p>
<p class="first-paragraph">Align with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">time
bowtie2 -x $REF -1 $R1 -2 $R2  &gt; bowtie.sam</font></font></p>
<p class="first-paragraph">The result is that:</p>
<ul>
	<li/>
<p class="compact">bwa aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">95.4%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>
	seconds</p>
	<li/>
<p class="compact">bowtie2 aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">94.7%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>
	seconds.</p>
</ul>
<p class="first-paragraph">So bowtie2 was slower and a hair less
efficient. We&rsquo;ll call it a tie, leaning towards <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font></p>
<p>Now let&rsquo;s raise the error rate to 10%:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim
-N 100000 -e 0.1 -E 0.1 $REF data</font></font></p>
<p class="first-paragraph">Running the same alignments now take
longer and fewer reads align:</p>
<ul>
	<li/>
<p class="compact">bwa aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">83.3%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
	seconds</p>
	<li/>
<p class="compact">bowtie2 aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">28.9%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font>
	seconds.</p>
</ul>
<p class="first-paragraph">The discrepancy is now huge. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>
is way off the mark! The first lesson to learn here is that
properties of the input data may have major effects on both runtime
and mapping rates that you may observe. But we can tune <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>
and make it more sensitive by using the &hellip; drumroll ..
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--very-sensitive-local</font></font>
flag:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2
--very-sensitive-local -x $REF -1 $R1 -2 $R2  &gt; bowtie.sam</font></font></p>
<p class="first-paragraph">In this case</p>
<ul>
	<li/>
<p class="compact">bwa aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">83.3%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
	seconds</p>
	<li/>
<p class="compact">bowtie2 aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">63.86%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11</font></font>
	seconds.</p>
</ul>
<p class="first-paragraph">Maybe we could tune bowtie2 even more. Hmm
&hellip; how about this for tuning:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">time
bowtie2 -D 20 -R 3 -N 1 -L 20 -x $REF -1 $R1 -2 $R2  &gt; bowtie.sam</font></font></p>
<p class="first-paragraph">We are getting somewhere; bowtie2 now
aligns more reads than bwa:</p>
<ul>
	<li/>
<p class="compact">bwa aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">83.3%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
	seconds</p>
	<li/>
<p class="compact">bowtie2 aligns <font face="Consolas, serif"><font size="2" style="font-size: 11pt">87.14%</font></font>
	of reads in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>
	seconds.</p>
</ul>
<p class="first-paragraph">Thus as you can see, it is not that simple
to compare aligners. A lot can depend on how aligners are configured
- though all things considered we like to start with and recommend
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
as apparently it is a lot less finicky and gets the job done without
adding elaborate decorations.</p>
<p>But does it mean that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>
is a second class citizen? Well, let&rsquo;s not rush to judgment
yet. Let&rsquo;s see what tags are present in the bwa alignment file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
bwa.sam | cut -f 12-20 | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:0
 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:0
 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM:i:6
 MD:Z:2A12G5A14G6G23A2   AS:i:44 XS:i:0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM:i:6
 MD:Z:7G8G32A8T0T3A6     AS:i:43 XS:i:0</font></font></p>
<p class="first-paragraph">the alignments made with bowtie2</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
bowtie.sam | cut -f 12-20 | head</font></font></p>
<p class="first-paragraph">contain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:-29
       XN:i:0  XM:i:10 XO:i:0  XG:i:0  NM:i:10
MD:Z:0C0A6A17C4T7T1A6G1T13A5    YS:i:-27        YT:Z:CP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:-27
       XN:i:0  XM:i:10 XO:i:0  XG:i:0  NM:i:10
MD:Z:2C8A8C5T1T5A6T14T3C2T6     YS:i:-29        YT:Z:CP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:-18
       XN:i:0  XM:i:6  XO:i:0  XG:i:0  NM:i:6  MD:Z:2A12G5A14G6G23A2 
 YS:i:-16        YT:Z:CP</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS:i:-16
       XN:i:0  XM:i:6  XO:i:0  XG:i:0  NM:i:6  MD:Z:7G8G32A8T0T3A6   
 YS:i:-18        YT:Z:CP</font></font></p>
<p class="first-paragraph">Do you see the difference? Bowtie2 fills
in a lot more information; it also offers more ways to format the
output and filter the alignments. So it has its applications and may
be indispensable in some situations.</p>
<p>By the way, do you want to know how we figured out that it should
be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-D
20 -R 3 -N 1 -L 20</font></font> and what does that even mean?</p>
<p>The task of finding settings where for this simulation <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>
outperforms <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
was a homework assignment.</p>
<p>The help for bowtie2 lists the parameters that are set when
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">--very-sensitive-local</font></font>
is set, so students started with those and kept tweaking and
re-running the alignments. It was a brute force search, not a
particularly deep insight into what these parameters mean and how
they interact. Few people (if anyone) understands these at that level
of granularity.</p>
<h2 class="western"><a name="how-do-i-choose-the-right-aligner"></a>83.2
How do I choose the right aligner?</h2>
<p class="first-paragraph">The most important lesson is not to treat
the tools as &ldquo;better&rdquo; or &ldquo;worse,&rdquo; but instead
as more or less appropriate for a given task. Also, it is essential
to understand that not all tools operate at the same levels of
sensitivity or specificity when using them on their default setting.</p>
<p>Finally note that, depending on the data and its characteristics,
different tools may be better suited to different types of data. The
requirement to have a particular kind of information in the BAM file
may also be a significant factor. The good news is that it is easy to
benchmark different aligners as needed. As we&rsquo;ve seen
conceptually, the tools are very similar.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="multiple-sequence-alignments"></a>84
Multiple sequence alignments</h1>
<p class="first-paragraph">A multiple sequence alignment is one that
uses three or more sequences. Since you already saw that even
pairwise alignments can be substantially customized it should come at
no surprise that since multiple sequence alignment involves
reconciling differences across more than two sequences there are
several different ways to go about it.</p>
<p>Just as with pairwise alignment there are optimal and non-optimal
but much speedier methods to perform the same analysis.</p>
<h2 class="western"><a name="how-do-i-align-more-than-two-sequences"></a>
84.1 How do I align more than two sequences?</h2>
<p class="first-paragraph">Suppose we wanted to align several full
length ebola viruses.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Store the genomes in this folder.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p genomes</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get all genomes for the ebola project.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nuccore -query PRJNA257197 | efetch -format fasta &gt;
genomes/ebola.fa</font></font></p>
<p class="first-paragraph">The file contains 249 full length genomes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat genomes/ebola.fa </font></font>
</p>
<p class="first-paragraph">that produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
             format  type  num_seqs    sum_len  min_len   avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genomes/ebola.fa
 FASTA   DNA        249  4,705,429   18,613  18,897.3   18,959</font></font></p>
<p class="first-paragraph">Initially let&rsquo;s just align the first
ten full ebola sequences against one another. Find the ID of the
first ten sequences:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
seq -n genomes/ebola.fa | cut -f 1 -d  ' ' | head -10 &gt; ids.txt</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ids.txt</font></font>
file now contains ten accession numbers:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105345.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105328.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105323.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Using these accession numbers we can
extract the sequences that correspond to these ids:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
grep --pattern-file ids.txt  genomes/ebola.fa &gt; small.fa</font></font></p>
<p class="first-paragraph">Perform a multiple sequence alignment with
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mafft</font></font>
tool:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mafft
--clustalout small.fa &gt; alignment.maf</font></font></p>
<p class="first-paragraph">View the alignment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head
-20 alignment.maf</font></font></p>
<p class="first-paragraph">Displays a visual alignment of all
sequences</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105345.1
    
-------ataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105328.1
    
--gattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105323.1
    
--gattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105302.1
    
--gattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105295.1
    
---attaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105294.1
    
--gattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105282.1
    
--gattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105266.1
    
---attaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105263.1
    
aagattaataattttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105253.1
     ---attattaatyttcctctcattgaaatttatatcggaatttaaattgaaattgttact</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                       ****
***********************************************</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                   </font></font>
</p>
<p class="first-paragraph">We can see that one difference between the
genomes is how complete the assemblies are at their edges.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tail
-20 alignment.maf</font></font></p>
<p class="first-paragraph">Similarly towards the ends:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105345.1
    
ggaaaaatggtcacacacaaaaatttaaaaataaatctatttcttcttttttgtgtgt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105328.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttttttgtgtgt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105323.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttttttgtgtgt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105302.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttttttgtgtg-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105295.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctattt----------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105294.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttttttgtgtgt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105282.1
    
ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttttttgtgtg-</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105266.1
    
ggaaaaatgg------------------------------------------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105263.1
    
ggaaaaatggtcgcacac----------------------------------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">KR105253.1
     ggaaaaatggtcgcacacaaaaatttaaaaataaatctatttcttcttt---------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               **********     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font>
</p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font>
character indicates a consensus.</p>
<p>Play around and perform multiple sequence alignments on 10% of the
genomes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sample --proportion 0.1 genomes/ebola.fa  &gt; small.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mafft
--clustalout small.fa &gt; alignment.maf</font></font></p>
<h2 class="western"><a name="what-programs-can-be-used-to-align-multiple-sequences"></a>
84.2 What programs can be used to align multiple sequences?</h2>
<p class="first-paragraph">Tools include <font face="Consolas, serif"><font size="2" style="font-size: 11pt">clustal-omega</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mafft</font></font></p>
<p>(TODO: COMING-SOON)</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-sambamcram-formats"></a>85 The
SAM/BAM/CRAM formats</h1>
<h2 class="western"><a name="what-is-a-sam-file"></a>85.1 What is a
SAM file?</h2>
<p class="first-paragraph">The SAM file contains information on
alignments.</p>
<p>The SAM format is a TAB-delimited, line-oriented text format
consisting of a</p>
<ol>
	<li/>
<p class="compact"><b>Header</b> section, where each line
	contains some metadata</p>
	<li/>
<p class="compact"><b>Alignment</b> section where each line
	provides information on an alignment</p>
</ol>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
format</a></font> specification lists the required and optional
content for each of these sections.</p>
<p>In general, the quality of the information within a SAM file
determines the success of analysis. Hence, producing this file so
that it contains the information we need, and the ability to
investigate the data on our own should always be our most pressing
concern.</p>
<p><b>Note</b>: Understanding the details of SAM will require that
you occasionally consult the official documentation:</p>
<ul>
	<li/>
<p class="compact">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
	format specification</a></font> is the specification of the SAM
	format.</p>
	<li/>
<p class="compact">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMtags.pdf">SAM
	tag specification</a></font> is the specification of the SAM tags.</p>
	<li/>
<p class="compact">Published as <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2723002/">The
	Sequence Alignment/Map format and SAMtools</a></font> in
	Bioinformatics. 2009</p>
</ul>
<h2 class="western"><a name="what-is-a-bam-file"></a>85.2 What is a
BAM file?</h2>
<p class="first-paragraph">A BAM file is a binary, compressed (and
almost always sorted) representation of the SAM information.
Generally, BAM files are sorted, usually by the alignment coordinate
and more rarely by the read names.</p>
<ul>
	<li/>
<p class="compact">Sorting by coordinate allows fast query on
	the information by location.</p>
	<li/>
<p class="compact">Sorting by read name is required when the
	identically named read pairs need to be accessed quickly as read
	pairs will then be stored in adjacent lines.</p>
</ul>
<h2 class="western"><a name="what-is-the-cram-format"></a>85.3 What
is the CRAM format?</h2>
<p class="first-paragraph">The file format was designed and released
in 2014 by the European Bioinformatics Institute see <font color="#4f81bd"><a href="http://genomeinformatician.blogspot.com/2014/08/cram-goes-mainline.html">CRAM
goes mainline</a></font></p>
<p>CRAM files are conceptually similar to BAM files. CRAM files
represent a more efficient version of the information in BAM files,
where the gain is made by storing the reference separately. The
essential difference between BAM and CRAM format is that a CRAM file
can only be read if the reference sequence is present at a given
location on the target computer. In contrast, the information in the
BAM file is complete.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CRAM</font></font>
format is supported directly by the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
sort</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 | samtools sort --reference $REF -O CRAM &gt;
bwa.cram </font></font>
</p>
<p class="first-paragraph">The downside of this approach is that
losing the reference sequence means losing the data (that is, being
unable to decode it). For many use-cases this is not a problem. But
there are situations where having the alignment data dependent on the
presence and availability of the reference genome adds another layer
of complexity.</p>
<h2 class="western"><a name="will-the-cram-format-replace-bam"></a>85.4
Will the CRAM format replace BAM?</h2>
<p class="first-paragraph">Time will tell (probably not).</p>
<h2 class="western"><a name="is-it-sam-bam-or-cram-now"></a>85.5 Is
it SAM, BAM or CRAM now?</h2>
<p class="first-paragraph">The format that the data is in is called
SAM. When we store it in a compressed binary form, we call it BAM or
CRAM. Most of the time we exchange the data in a binary format
because it is much more efficient and faster to query.</p>
<p class="block-text-western">Note that when we view or query a BAM
file with a tool like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
we see the results converted into the SAM format.</p>
<p class="first-paragraph">We will usually call the format and the
data as SAM with the understanding that the representation may be
either that SAM, BAM or CRAM. Each format contains the same
information; the differences are in their storage methods and
performance characteristics.</p>
<h2 class="western"><a name="what-are-sam-files-used-for"></a>85.6
What are SAM files used for?</h2>
<p class="first-paragraph">SAM files were designed for two primary
use cases:</p>
<ol>
	<li/>
<p class="compact">To store alignments in a standardized and
	efficient manner.</p>
	<li/>
<p class="compact">To allow quick access to alignments by
	their coordinates.</p>
</ol>
<p class="first-paragraph">For example, if you had 100 million
alignments in a file and wanted those alignments that overlap with
coordinate 1,200,506 - the BAM format can return that information in
milliseconds, without having to read through the entire file.</p>
<h2 class="western"><a name="is-the-sam-an-optimal-format"></a>85.7
Is the SAM an optimal format?</h2>
<p class="first-paragraph">Unfortunately no.</p>
<p>The SAM format has all the features of a product designed by a
committee (which it was), marred by premature optimizations (the root
of all evil according to Donald Knuth), crippled by flawed conceptual
design, impaired by an incomplete specification. Other than that it
is fantastic.</p>
<p>But we are stuck with SAM. Like it or not, we have to learn how to
deal with the format.</p>
<h2 class="western"><a name="what-information-is-stored-in-a-sam-file"></a>
85.8 What information is stored in a SAM file?</h2>
<p class="first-paragraph">Even practicing bioinformaticians
frequently have misconceptions about SAM files. Among them, the most
prevalent is the assumption that BAM files produced by different
tools will have a comparable amount of information and the
differences are mostly in the <i>accuracy</i> or <i>performance</i>
characteristics of the alignments.</p>
<p>Nothing could be further from the truth.</p>
<p>The misconception perhaps caused by the <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
specification</a></font> itself. This specification appears to be the
&ldquo;standard&rdquo; requirement that seems to prescribe what
exactly should go into each SAM file:</p>
<ol>
	<li/>
<p class="compact">each SAM file has to have 11 columns</p>
	<li/>
<p class="compact">each column contains the <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">following
	information</a></font> &hellip;</p>
</ol>
<p class="first-paragraph">Yet, as it turns out the &ldquo;required&rdquo;
information is quite limited, and a file that would only fill the
required columns would be useless in many cases. Most often the
&ldquo;optional&rdquo; tags and other fields of the SAM file carry
all the information needed for analysis. Thus there may be
substantial differences between alignment information produced by
different tools</p>
<p>The take-home message is that a SAM format is still somewhat
nebulous. It prescribes the presence of very little information. Each
aligner will fill in as much as it knows how to.</p>
<h2 class="western"><a name="why-do-we-need-the-sam-format"></a>85.9
Why do we need the SAM format?</h2>
<p class="first-paragraph">The SAM format was introduced to support
use cases presented by the high throughput sequencing
instrumentation:</p>
<ul>
	<li/>
<p class="compact">Fast access to alignments that overlap with
	a coordinate. For example, select alignments that overlap with
	coordinate <font face="Consolas, serif"><font size="2" style="font-size: 11pt">323,567,334</font></font>
	on chromosome <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></p>
	<li/>
<p class="compact">Quick selection and filtering of reads
	based on attributes. For example, we want to be able to quickly
	select alignments that align on the reverse strand (though we want
	to warn you that the actual implementation of the filtering
	borderlines the absurd)</p>
	<li/>
<p class="compact">Efficient storage and distribution of the
	data. For example, have a single compressed file containing data for
	all samples each labeled in some manner.</p>
</ul>
<h2 class="western"><a name="how-important-is-to-understand-the-details-of-the-sam-format"></a>
85.10 How important is to understand the details of the SAM format?</h2>
<p class="first-paragraph">On the one hand, it is conceivable that
most of the time you will not need to manipulate SAM/BAM files
directly. There is an extensive library of programs and tools
designed to process and analyze SAM/BAM files for you.</p>
<p>On the other hand, many problems cannot be understood and solved
without understanding and investigating the SAM files in more detail.
Time an again we find ourselves looking at a few lines of a SAM file
to understand a seemingly mysterious event taking place.</p>
<h2 class="western"><a name="how-do-i-create-sambam-files"></a>85.11
How do I create SAM/BAM files?</h2>
<p class="first-paragraph">We have a separate chapter titled <font color="#4f81bd"><a href="#ch086.xhtml#make-a-bam">How
to make a BAM file</a></font>, below we point out only the relevant
lines: Typically you would generate a SAM file, then sort that file
and convert it into a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
format. Finally, you will need to index the resulting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a SAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 &gt; alignments.sam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert SAM to sorted BAM.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
sort alignments.sam &gt; alignments.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index myfile.bam</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
package has come a long way in its usability and since version 1.3
will both convert and sort a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAM</font></font>
file into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
in one step:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a sorted BAM file in one line.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 | samtools sort &gt; alignments.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index alignments.bam</font></font></p>
<h2 class="western"><a name="how-do-i-create-a-cram-file"></a>85.12
How do I create a CRAM file?</h2>
<p class="first-paragraph">You need to add the reference to the
sorting step and change the output format:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a sorterd CRAM file in one line.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 | samtools sort --reference $REF -O cram  &gt;
bwa.cram</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index CRAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index bwa.cram</font></font></p>
<h2 class="western"><a name="can-unaligned-data-be-stored-in-a-sam-file"></a>
85.13 Can &ldquo;unaligned&rdquo; data be stored in a SAM file?</h2>
<p class="first-paragraph">Because the BAM files are compact and can
store additional information they have become a de-facto standard not
only for representing and storing alignment information but also for
storing raw &ldquo;unaligned&rdquo; FASTQ files.</p>
<p>It is a somewhat idiosyncratic usage (welcome to bioinformatics
folks): a format that was initially designed to store alignments is
now used to represent &ldquo;unaligned&rdquo; data. In all honesty
that is more a testament of the shortcomings of the FASTQ files.</p>
<p>The advantage of the BAM file is that each read can be tagged with
additional, user-created information, for example, the sample name.</p>
<p>When used in this manner a single BAM file can be used to store
all samples from an experiment, hence greatly simplifying file
management. Instead of having to deal with 100 paired-end samples
distributed over 200 files we can save a single BAM file.</p>
<p>Here is an example of converting paired-end FASTQ files in an
unaligned BAM format, where each read in the pair is tagged as &ldquo;FOO&rdquo;</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 10000 --split-files SRR1972739</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Set up shortcuts.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=SRR1972739_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=SRR1972739_2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert FASTQ to BAM file. Tag each sample as &quot;FOO&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">picard
FastqToSam F1=$R1 F2=$R2 O=SRR1972739.bam SM=&quot;FOO&quot;</font></font></p>
<p class="first-paragraph">Look at the headers of the resulting file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -H SRR1972739.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@HD
VN:1.5  SO:queryname</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@RG
ID:A    SM:FOO</font></font></p>
<p class="first-paragraph">The so-called &ldquo;readgroup&rdquo; A
will be tagged as &ldquo;FOO&rdquo;.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bam | head -1</font></font></p>
<p class="first-paragraph">And the file contains (shown wrapped
below):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1
   77  *   0   0   *   *   0  
0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TCTAAAAGGAGGACTCAATTGCAATATCTCATTTGAGTCCTCCTTTTAGAGGATTATTGTCATAAATCAATTCATTTTCTCGGTATCTTGTTAAATCTAAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCCFFFFFHHHHFIJJJJJJIJJJJJJJJJJJJJJGIGIJJJJJJJJJJJJJJJJJJJIGIJJJJJJJJJJJJJJJJJIHHHHDEFFFEEDEEEEEDDDDD</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">RG:Z:A</font></font></p>
<p class="first-paragraph">Note again the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">RG:Z:A</font></font>
tag at the end of the read that maps the read to id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
that in turn is mapped to sample <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SM:FOO</font></font>
in the header.</p>
<p>To analyze the data stored in an unaligned BAM file we typically
need to unpack it into FASTQ format again like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
fastq SRR1972739.bam -1 pair1.fq -2 pair2.fq</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-make-a-bam-file"></a>86 How to
make a BAM file</h1>
<p class="first-paragraph">Several chapters will start with by
analyzing a BAM file. This section shows the code for creating a BAM
file using published data. The script aligns sequencing data from the
<font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo: Genomic
surveillance elucidates Ebola virus origin</a></font> project using
both <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>.</p>
<p>You can download this script from</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/scripts/make-bam.sh">http://data.biostarhandbook.com/scripts/make-bam.sh</a></font></p>
</ul>
<p class="first-paragraph">You may also obtain it from the command
line with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
-nc http://data.biostarhandbook.com/scripts/make-bam.sh</font></font></p>
<p class="first-paragraph">In addition check the <font color="#4f81bd"><a href="https://www.bioinformatics.recipes/recipe/list/bio-data-analysis/">Bioinformatics
Data Analysis recipes</a></font> for further code examples.</p>
<p>The script mentioned above contains the following commands:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Stop on any error.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-uex</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Accession number for the reference genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
SRA run number.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR=SRR1972739</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
How many reads to extract from the dataset.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N=10000</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The reference genome in three different formats: GenBank, FASTA and
GFF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GB=refs/$ACC.gb</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FA=refs/$ACC.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GFF=refs/$ACC.gff</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the reference directory.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a genbank file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nucleotide -format gb -id $ACC &gt; $GB</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert the GenBank file into GFF3.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$GB | seqret -filter -feature -osformat gff3 &gt; $GFF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert the GenBank file into FASTA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$GB | seqret -filter -feature -osformat fasta &gt; $FA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create an index of the FASTA file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $FA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the dataset.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X $N --split-files $SRR</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index reference with bwa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $FA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the reference with samtools</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $FA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Shortcuts to read names</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=${SRR}_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=${SRR}_2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Align with bwa mem.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $FA $R1 $R2 | samtools sort &gt; $SRR.bwa.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file generated with bwa.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index $SRR.bwa.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index reference with bowtie2.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2-build
$FA $FA</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Align the same data with bowtie2.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2
-x $FA -1 $R1 -2 $R2 | samtools sort &gt; $SRR.bowtie.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file produced with bowtie2.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index $SRR.bowtie.bam</font></font></p>
<p class="first-paragraph">Running the script above produces 25
files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">find
.  | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
prints 25</font></font></p>
<p class="first-paragraph">Out of those the following are of interest
information:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">refs/AF086833.fa</font></font>
	- the reference genome sequence.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">refs/AF086833.gff</font></font>
	- is the GFF feature file for the genome.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.bwa.bam</font></font>
	- is the alignment file produced with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.bowtie.bam</font></font>
	- is the alignment file produced with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>.</p>
</ul>
<p class="first-paragraph">The script also produces a GFF file that
we can visualize with IGV together with our alignment files:</p>
<p><img src="0000_html_77b48d3c153ad9ca.png" name="Image129" align="bottom" hspace="12" width="560" height="283" border="0"/>
</p>
<p>We can produce statistics on each BAM file with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat SRR1972739.bwa.bam</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20740
+ 0 in total (QC-passed reads + QC-failed reads)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 secondary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">740
+ 0 supplementary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 duplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">15279
+ 0 mapped (73.67% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20000
+ 0 paired in sequencing</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">14480
+ 0 properly paired (72.40% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">14528
+ 0 with itself and mate mapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">11
+ 0 singletons (0.05% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr (mapQ&gt;=5)</font></font></p>
<p class="first-paragraph">whereas</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat SRR1972739.bowtie.bam</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20000
+ 0 in total (QC-passed reads + QC-failed reads)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 secondary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 supplementary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 duplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12537
+ 0 mapped (62.69% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20000
+ 0 paired in sequencing</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">9828
+ 0 properly paired (49.14% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">11482
+ 0 with itself and mate mapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1055
+ 0 singletons (5.27% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr (mapQ&gt;=5)</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-sam-format-explained"></a>87 The SAM
format explained</h1>
<h2 class="western"><a name="what-is-a-sam-file-1"></a>87.1 What is a
SAM file?</h2>
<p class="first-paragraph">The SAM format is a TAB-delimited,
line-oriented text format consisting of a</p>
<ol>
	<li/>
<p class="compact"><b>Header</b> section, where each line
	contains some metadata</p>
	<li/>
<p class="compact"><b>Alignment</b> section where each line
	provides information on an alignment</p>
</ol>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
format</a></font> specification lists the required and optional
content for each of these sections.</p>
<h2 class="western"><a name="where-do-i-get-a-bam-file"></a>87.2
Where do I get a BAM file?</h2>
<p class="first-paragraph">The commands in this section assume that
your BAM was generated in the manner described in the <font color="#4f81bd"><a href="#ch086.xhtml#make-a-bam">How
to make a BAM file</a></font> section.</p>
<p>You may also choose a remotely hosted BAM file. For example select
data from publication discussed in <font color="#4f81bd"><a href="#ch056.xhtml#redo-denisovan">Redo:
Explore the genome of a paleolithic-era archaic human</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Location of the BAM
file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=http://cdna.eva.mpg.de/denisova/alignments/T_hg19_1000g.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download alignments for chromosome 22.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b $URL 22 &gt; denisova.chr22.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the downloaded alignments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index denisova.chr22.bam</font></font></p>
<p class="first-paragraph">Alternatively you may explore a subset of
data from <font color="#4f81bd"><a href="#ch055.xhtml#syndipredo">Redo:
A synthetic-diploid benchmark for accurate variant-calling evaluation</a></font>
(takes about 15 minutes to download)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Location of the BAM
file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download alignments for chromosome 22.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b $URL 22 &gt; syndip.chr22.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the downloaded alignments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index syndip.chr22.bam</font></font></p>
<h2 class="western"><a name="what-is-the-sam-header"></a>87.3 What is
the SAM header?</h2>
<p class="first-paragraph">While we could temporarily save BAM as SAM
files then treat them as 11 column text files, in practice, we never
do that. Almost exclusively we only &ldquo;view&rdquo; BAM files in
the SAM format. The tool that does the &ldquo;live&rdquo; conversion
of BAM to SAM is called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>.</p>
<p>To view the header of a SAM file we would type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -H SRR1972739.bwa.bam</font></font></p>
<p class="first-paragraph">The lines that start with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@</font></font>
symbols are the so-called &ldquo;SAM header&rdquo; rows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@HD
VN:1.6  SO:coordinate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SQ
SN:AF086833 LN:18959</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@PG
ID:bwa  PN:bwa  VN:0.7.17-r1188 CL:bwa mem refs/AF086833.fa
SRR1972739_1.fastq SRR1972739_2.fastq</font></font></p>
<p class="first-paragraph">The headers are described in the <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
Specification</a></font>. Various information encoded here, for
example, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SN</font></font>
is the sequence name that we aligned against (the name of the
sequence in the Fasta reference file), <font face="Consolas, serif"><font size="2" style="font-size: 11pt">LN</font></font>
is the length of this sequence, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PG</font></font>
is the program version that we ran. But if you were to look at the
alignment file produced with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -H SRR1972739.bowtie.bam</font></font></p>
<p class="first-paragraph">you would see a slightly different header:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@HD
VN:1.0  SO:coordinate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SQ
SN:AF086833 LN:18959</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@PG
ID:bowtie2  PN:bowtie2  VN:2.3.4.3 
CL:&quot;/Users/ialbert/miniconda3/envs/bioinfo/bin/bowtie2-align-s
--wrapper basic-0 -x refs/AF086833.fa -1 SRR1972739_1.fastq -2
SRR1972739_2.fastq&quot;</font></font></p>
<p class="first-paragraph">Depending on what kinds of post-processing
information went into a BAM file the headers may be quite extensive.
Besides, samtools can access remote data as well. In the <font color="#4f81bd"><a href="#ch055.xhtml#syndipredo">Redo:
A synthetic-diploid benchmark for accurate variant-calling evaluation</a></font>
(#syndipredo) project we noted a BAM file under the URL</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam">ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</a></font>.
	We can readily view its header:</p>
</ul>
<p class="first-paragraph">get the header for the remote BAM file
with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -H $URL &gt; header.txt</font></font></p>
<p class="first-paragraph">Note how we cautiously placed the header
into a separate file, why?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
header.txt | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
prints 171</font></font></p>
<p class="first-paragraph">It has 171 lines, with content that looks
like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@HD
    VN:1.4  GO:none SO:coordinate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SQ
    SN:1    LN:249250621   
UR:http://www.broadinstitute.org/ftp/pub/seq/references/Homo_sapiens_assembly19.fasta
  AS:GRCh37       M5:1b22b98cdeb4a9304cb5d48026a85128</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SP:Homo Sapiens</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@PG
    ID:GATK IndelRealigner  VN:nightly-2015-07-31-g3c929b0 
CL:knownAlleles=[(RodBinding
name=knownAlleles</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">source=/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.dbsnp138.vcf),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(RodBinding
name=knownAlleles2
source=/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.known_indels.vcf),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(RodBinding
name=knownAlleles3
source=/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.variantEvalGoldStandard.vcf)]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">targetIntervals=/seq/picard_aggregation/G94794/CHMI_CHMI3_WGS1/v3/1/CHMI_CHMI3.merged.intervals</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">LODThresholdForCleaning=5.0
out=null consensusDeterminationModel=USE_READS entropyThreshold=0.15
maxReadsInMemory=150000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">maxIsizeForMovement=3000
maxPositionalMoveAllowed=200 maxConsensuses=30
maxReadsForConsensuses=120</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">maxReadsForRealignment=20000
noOriginalAlignmentTags=false
nWayOut=/seq/picard_aggregation/G94794/CHMI_CHMI3_WGS1/v3/1/final_bams.map</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">generate_nWayOut_md5s=false
check_early=false noPGTag=false keepPGTags=false
indelsFileForDebugging=null</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">statisticsFileForDebugging=/seq/picard_aggregation/G94794/CHMI_CHMI3_WGS1/v3/1/CHMI_CHMI3.indel.stats
SNPsFileForDebugging=null</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">@PG
    ID:MarkDuplicates      
VN:5589b97c07fd801bd8afb2999f2bf66a81dede7c    
CL:picard.sam.markduplicates.MarkDuplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">READ_ONE_BARCODE_TAG=RX
INPUT=[/pod/pod2temp/pod2/tmp/flowcell/HJCMTCCXX/C1-310_2016-01-13_2016-01-16/5/Pond-487422/HJCMTCCXX.5.aligned.bam]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">OUTPUT=/pod/pod2temp/pod2/tmp/flowcell/HJCMTCCXX/C1-310_2016-01-13_2016-01-16/5/Pond-487422/HJCMTCCXX.5.aligned.duplicates_marked.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">METRICS_FILE=/pod/pod2temp/pod2/tmp/flowcell/HJCMTCCXX/C1-310_2016-01-13_2016-01-16/5/Pond-487422/HJCMTCCXX.5.duplicate_metrics</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">OPTICAL_DUPLICATE_PIXEL_DISTANCE=2500
TMP_DIR=[/local/scratch/HJCMTCCXX,
/pod/pod2temp/pod2/tmp/proctmp/HJCMTCCXX]
VALIDATION_STRINGENCY=SILENT CREATE_INDEX=true
CREATE_MD5_FILE=true</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000
MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000
SORTING_COLLECTION_SIZE_RATIO=0.25 REMOVE_DUPLICATES=false
ASSUME_SORTED=false</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES
PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates
READ_NAME_REGEX=&lt;optimized capture of last three ':' :</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">It is a monster header! The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PG</font></font>
tag lists information on what programs were run and what parameters
each had. It was intended to help us better understand what has been
done to produce the BAM file. But looking at this file does it
really? &hellip; it is convoluted and complicated, most likely
needlessly so. If there is one thing I have learned from watching
geniuses at work is that they frequently lack the ability, or perhaps
the need (?), to simplify. Look at file naming as evidence:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/pod/pod2temp/pod2/tmp/</font></font></p>
<p class="first-paragraph">the people that analyzed this data use a
computer that has a root level directory <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pod</font></font>
that contains another directory called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pod2temp</font></font>
that contains another directory called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pod2</font></font>
that contains a directory called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tmp</font></font>.
That&rsquo;s a bit silly arrangement - I am noting it here as an
example of how unrelated details of a computer &ldquo;leak&rdquo;
into metadata.</p>
<h2 class="western"><a name="what-is-stored-in-the-sam-alignment-section"></a>
87.4 What is stored in the SAM alignment section?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
Specification</a></font> contains a detailed description of each
field and is a required read for all bioinformaticians. Whenever in
doubt, consult it.</p>
<p>The simplest way to understand a SAM file is to go over it a few
columns at a time. The SAM spec has a name for each column, for
example, column 1 is also called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">QNAME,</font></font>
which corresponds to the words &ldquo;Query name&rdquo;. Let&rsquo;s
look at the</p>
<h2 class="western"><a name="column-1-query-name-qname"></a>87.5
Column 1: Query Name (QNAME)</h2>
<p class="first-paragraph">Cut the first column:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 1 | head -5</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.9152</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.4087</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1404</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.7291</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.8376</font></font></p>
<p class="first-paragraph">Whereas the original FASTQ file had the
reads labeled in order:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.3</font></font></p>
<p class="first-paragraph">the alignments in the BAM file are sorted
by position. The SAM file tells us that the read ithat was the 9152th
in the file aligned in the leftmost position.</p>
<h2 class="western"><a name="column-2-flag"></a>87.6 Column 2: FLAG</h2>
<p class="first-paragraph">As it happens a substantial amount of
information necessary to figure out which read aligns in what way is
included in the second column of the SAM format, in the &ldquo;magical&rdquo;
number called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 2 | head -5</font></font></p>
<p class="first-paragraph">that produces the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">99</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
is the field of the SAM spec where the designers of the format have
strayed the furthest from what one might call <i>common sense</i>. It
was a way to &ldquo;cram&rdquo; (no pun intended) as much information
as possible into a single number. To do so, they came up with a
contorted rationalization that no biologist will ever manage to
remember, hence ensuring the long-term employment of
bioinformaticians across the globe. &#129300; Perhaps that was the
plan all along, in which case I must admit, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
is not such a bad thing after all.</p>
<p>Here is how the rationale works: Imagine that we associate an
attribute to each power of 2 starting from zero. Let&rsquo;s say</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
	means &ldquo;red&rdquo;,<br/>
<br/>
<br/>

	</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
	means &ldquo;blue&rdquo;,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>
	means &ldquo;tall&rdquo;,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8</font></font>
	means &ldquo;short&rdquo;,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16</font></font>
	means &ldquo;fast&rdquo;,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">32</font></font>
	means &ldquo;cheap&rdquo;,</p>
</ul>
<p class="first-paragraph">At this point, it is not important what
the attributes are just that they correspond to the information we
wish to know. If we were to describe an object as a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>
we would decompose that into powers (factors) of 2 like so <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
+ 4</font></font>, in essence stating that when we label something as
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font>
what mean is that it is &ldquo;red&rdquo; and &ldquo;tall&rdquo;. If
we had a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>
that would be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
+ 8</font></font> meaning that its attributes are &ldquo;blue&rdquo;
and &ldquo;short&rdquo;. If it is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">37</font></font>
it is &ldquo;red&rdquo;, &ldquo;tall&rdquo; and &ldquo;fast&rdquo;.</p>
<p>The justification behind this encoding was that instead of having
to store two labels, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">8</font></font>,
we can just store a single number, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>,
yet still represent two attributes. Expanding one this we can now
store up to 8 factors in a byte (8 bits).</p>
<p>It is essential to remember that if we wanted to identify the
items that are &ldquo;blue&rdquo; (that is, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>)
we would need to select ALL the numbers that, when decomposed into
powers of 2 would contain <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>.
And that is where it ends up quite counter-intuitive as it includes
the numbers <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font>,<font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>,<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10</font></font>
but would NOT include <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">9</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">16</font></font>
etc.</p>
<p>See why: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
+ 2 = 3</font></font>, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
+ 4 = 6</font></font>, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
+ 2 + 4 = 7</font></font>, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
+ 8 = 10</font></font> and so on. So everything described as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3,6,7,10</font></font>
is blue (among other attributes). But of course, all these objects
will be different in one or more of their other attributes. While the
method may look elegant and fast to people familiar with binary
representations - like the computational geniuses that came up with
it - it does not match at all how humans think about these problems.</p>
<p>First, it is challenging to see by eye which flags are set and
which are not: is flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
set in the numbers <font face="Consolas, serif"><font size="2" style="font-size: 11pt">19</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">35</font></font>?
Moreover, what if some of the attributes are mutually exclusive but
others are not? For example, red and blue may be mutually exclusive
but &ldquo;fast&rdquo; and &ldquo;cheap&rdquo; are not. When
selecting for attributes and lack thereof, we end up with
surprisingly hard to parse statements.</p>
<p>One of the curses of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
system is that it is exceedingly easy to mistakenly generate a wrong
type of query as one adds more terms into it.</p>
<p>Imagine handling eight different attributes, some of which are
exclusive and others not; some traits are filled in by the aligner;
others are not. Selecting or removing alignments by these attributes
becomes unnecessarily complicated and error-prone. The flaw of the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
system is that it accentuates rather than solves an already difficult
problem. Using the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
system correctly is surprisingly challenging.</p>
<p>The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
specification</a></font> lists the meaning of the flags, and the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">flags</font></font>
command will list them at the command line:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags</font></font></p>
<p class="first-paragraph">generating:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Flags:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     0x1   PAIRED        .. paired-end (or multiple-segment)
sequencing technology</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     0x2   PROPER_PAIR   .. each segment properly aligned according
to the aligner</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     0x4   UNMAP         .. segment unmapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     0x8   MUNMAP        .. next segment in the template unmapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    0x10   REVERSE       .. SEQ is reverse complemented</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    0x20   MREVERSE      .. SEQ of the next segment in the template
is reversed</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    0x40   READ1         .. the first segment in the template</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    0x80   READ2         .. the last segment in the template</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   0x100   SECONDARY     .. secondary alignment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   0x200   QCFAIL        .. not passing quality controls</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   0x400   DUP           .. PCR or optical duplicate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   0x800   SUPPLEMENTARY .. supplementary alignment</font></font></p>
<p class="first-paragraph">Adding insult to injury, these values are
listed in hexadecimal representation, as if that would have meaning
to a life scientists. What were you thinking developers? Let&rsquo;s
see what our flags mean, our list contained <font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">99</font></font>.
We can ask for a description for each flag:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags 163</font></font></p>
<p class="first-paragraph">and that results in:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0xa3
   163 PAIRED,PROPER_PAIR,MREVERSE,READ2</font></font></p>
<p class="first-paragraph">We can read this out loud. You could say:</p>
<p class="block-text-western"><i>When I see an alignment with the
</i><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>FLAG</i></font></font><i>
set to </i><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>163</i></font></font><i>
I know right away that it came from a paired read. Both the read and
its mate map in in a way that a &ldquo;proper pair&rdquo; should. The
mate aligns on the reverse strand, and the read sequence comes from
the second file.</i></p>
<p class="first-paragraph">Whoa! A lot of information in just one
number, wouldn&rsquo;t you say?</p>
<p>We are not done yet - and in my opinion - this captures the
fundamental flaw of the FLAG system. To understand the alignment
above you have also to enumerate some of the flags that are not set.
Let&rsquo;s look at the unset flags: UNMAP, MUNMAP, REVERSE, READ1,
SECONDARY, QCFAIL, DUP, SUPPLEMENTARY. Mental gymnastics is now
necessary to list those attributes that, when not set, are
meaningful.</p>
<p>For example, READ1 is not meaningful, we already know that the
read comes from the second pair as it says READ2. We have learned
nothing new by saying, it is not READ1 we knew that already. But we
do learn something from seeing how REVERSE is not set. It means that
the current alignment is on the forward strand. Then we can also
learn something from seeing how the flags for neither SUPPLEMENTARY
and SECONDARY alignments are not present. It means that this
alignment is a primary alignment on the forward strand. So the story
of this alignment now becomes:</p>
<p class="block-text-western"><i>When I see an alignment with the
</i><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>FLAG</i></font></font><i>
set to </i><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>163</i></font></font><i>
I know right away that: it aligns (UNMAP not set), it aligns on the
forward strand (REVERSE not set), it is a primary alignment
(SECONDARY and SUPPLEMENTARY not set), it came from a paired read
(PAIRED is set), where both the read and its mate map in in a way
that a &ldquo;proper pair&rdquo; should (PROPER_PAIR is set), the
mate aligns on the reverse strand (MREVERSE is set), and the sequence
comes from the second file (READ1 is set).</i></p>
<p class="first-paragraph">How about DUP, QCFAIL? Those are not
meaningful in this case since we don&rsquo;t know if we have
attempted to detect these flaws or not. A read not being a duplicate
is only meaningful if we tried to label duplicates - which most none
of the aligners do. And since we are here what does a PROPER_PAIR
even mean? I have a surprise for you, it could mean anything. The SAM
spec only says <i>&ldquo;each segment properly aligned according to
the aligner&rdquo;</i>. So there you have it, it is a proper pair &hellip;
according to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>.</p>
<p>In the <i>wonderful world</i> of SAM you <i>know</i> something is
primary when it is neither <i>secondary</i> nor <i>supplementary</i>,
and you know something must be forward from the &ldquo;trivially
obvious&rdquo; fact that it is not labeled as reverse. Gee thanks! It
is like Yoda speak, you must use the Force to understand it:</p>
<p><img src="0000_html_5a4f2854f2602448.jpg" name="Image130" align="bottom" hspace="12" width="560" height="201" border="0"/>
</p>
<p>Now take a look at the flags <font face="Consolas, serif"><font size="2" style="font-size: 11pt">99</font></font>
it says:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0x63
   99  PAIRED,PROPER_PAIR,MREVERSE,READ1</font></font></p>
<p class="first-paragraph">To generate statistics on the flags:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat SRR1972739.bwa.bam</font></font></p>
<h2 class="western"><a name="what-are-primary-secondary-and-chimeric-supplementary-alignments"></a>
87.7 What are primary, secondary and chimeric (supplementary)
alignments</h2>
<p class="first-paragraph">The most important <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAGS</font></font>
characterizing the alignment type are:</p>
<ol>
	<li/>
<p class="compact"><b>Secondary alignment</b>: a linear
	alignment that covers different regions of a read.</p>
	<li/>
<p class="compact"><b>Supplementary alignment</b>: a read that
	cannot be represented as a single linear alignment and matches two
	different locations without overlap. These are also known as
	&ldquo;chimeric&rdquo; alignments. What gets called as chimeric
	alignment is software/implementation dependent.</p>
</ol>
<p class="first-paragraph">an alignment not labeled as secondary or
supplementary is then a:</p>
<ol>
	<li/>
<p class="compact"><b>Primary alignment</b>: the best
	alignments by the score. If multiple alignments match equally well,
	the aligner will designate one (software dependent) as the primary
	alignment.</p>
</ol>
<h2 class="western"><a name="columns-3-4-reference-name-rname-and-position-pos"></a>
87.8 Columns 3-4: Reference Name (RNAME) and Position (POS)</h2>
<p class="first-paragraph">The 3rd and 4th columns of a SAM file
indicate the reference name (RNAME) and the position (POS) of where
the query, the sequence named in column 1 (QNAME) aligns:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 1,3,4 | head -5</font></font></p>
<p class="first-paragraph">In our case, there is only one reference
sequence (chromosome):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.9152
   AF086833     46</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.4087
   AF086833     62</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.1404
   AF086833     80</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.7291
   AF086833    103</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.8376
   AF086833    108</font></font></p>
<p class="first-paragraph">We can see that each sequence aligns to
the same reference at different positions: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">46</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">62</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">80</font></font>
and so on.</p>
<p class="block-text-western">Everything in a SAM file is relative to
the forward strand!</p>
<p class="block-text-western">The essential thing to remember about
the POS field is that it reflects the <b>leftmost</b> coordinate of
the alignment. Even when the alignment matches on the reverse strand
the POS coordinate is reported relative to the forward strand.</p>
<p class="first-paragraph">It is the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
column that tells us which strand the data aligns to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG=4</font></font></p>
<h2 class="western"><a name="column-5-6-mapping-quality-mapq-and-compact-idiosyncratic-gapped-alignment-representation-cigar"></a>
87.9 Column 5-6: Mapping Quality (MAPQ) and Compact Idiosyncratic
Gapped Alignment Representation (CIGAR)</h2>
<p class="first-paragraph">These columns reflect the Mapping Quality
(MAPQ) and the so-called Compact Idiosyncratic Gapped Alignment
Representation (CIGAR).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 5,6 | head -5</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 101M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 101M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 101M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 101M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 101M</font></font></p>
<p class="first-paragraph">The values in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MAPQ</font></font>
column here are all <font face="Consolas, serif"><font size="2" style="font-size: 11pt">60</font></font>.
This column was designed to indicate the likelihood of the alignment
being placed incorrectly. It is the same Phred score that we
encountered in the FASTQ files. And we read it the same way, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">60/10
= 6</font></font> so the chance of seeing this alignment being wrong
is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10^-6</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1,000,000</font></font>
one in a million.</p>
<p>There is one caveat though. The numbers that an aligner puts into
the MAPQ field are typically estimates. It is not possible to
mathematically compute this value. What this field does is to inform
us on a guess by the aligner&rsquo;s algorithm. This guess is more of
a hunch that should not be treated as a continuous, numerical value.
Instead, it should be thought of as an ordered label, like &ldquo;not
so good&rdquo; or &ldquo;pretty good&rdquo;. Aligner developers even
generate unique MAPQ qualities to mark individual cases. For example,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
will create a MAPQ=0 if a read maps equally well to more than one
location.</p>
<p>The CIGAR string is a different beast altogether. It is meant to
represent the alignment via numbers followed by letters:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>
	match of mismatch</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">I</font></font>
	insertion</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">D</font></font>
	deletion</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">S</font></font>
	soft clip</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">H</font></font>
	hard clip</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
	skipping</p>
</ul>
<p class="first-paragraph">To find CIGAR strings other than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">101M</font></font>
you could write:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 5,6 | grep -v &quot;101&quot; | head
-5</font></font></p>
<p class="first-paragraph">to produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 18S83M</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 92M9S</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 90M11S</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 98M3S</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60
 3S98M</font></font></p>
<p class="first-paragraph">These are also meant to be &ldquo;readable&rdquo;;
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">18S83M</font></font>
says that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">18</font></font>
bases soft clipped and the next <font face="Consolas, serif"><font size="2" style="font-size: 11pt">83</font></font>
are a <i>match or mismatch</i>.</p>
<p>The CIGAR representation is a neat concept, alas it was developed
for short and well-matching reads. As soon as the reads show
substantial differences the CIGAR representations are much more
difficult to read.</p>
<p>There are also different variants of CIGAR. The &ldquo;default&rdquo;
CIGAR encoding describes both the match and mismatch the same way,
with an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>.
There are some unfortunate rationale and explanation for having
adopted this choice - one that complicates analysis even more.</p>
<p>The extended CIGAR encoding adopted by others uses the symbols of
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">=</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>
to indicate matches and mismatches.</p>
<h2 class="western"><a name="columns-7-9-rnext-pnext-tlen"></a>87.10
Columns 7-9: RNEXT, PNEXT, TLEN</h2>
<p class="first-paragraph">Columns 7,8 and 9 deal with the alignment
of the mate pair. Since both reads that from a pair originated from
the same fragment, it is very informative to know right away where
both ends align. And that is what these columns do.</p>
<p>Let us also include the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">QNAME</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>
fields and look at the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 7,8,9 | head -5</font></font></p>
<p class="first-paragraph">We see the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">=
     230      281</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">=
     224      263</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">=
     205      226</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">=
     238      236</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">=
     194      187</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=</font></font>
symbol for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNEXT</font></font>
field indicates that the query mate aligns to the same reference
anchor (chromosome) as the query. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PNEXT</font></font>
fields indicate the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>
field of the <b>primary(!)</b> alignment of the mate. (see later what
primary alignment means). Note how every <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PNEXT</font></font>
always has a corresponding <font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>
column on a different line since the mate is also in the file.
Finally, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TLEN</font></font>
column indicates the distance between the outer ends of mate
alignments, that is, how long the fragment <b>appears</b> to be!
Making this distinction is imperative. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TLEN</font></font>
is not the actual length of the fragment - it is the length that it
would be if the genome under study was identical to the reference and
the mapping was correct. The specification calls it observed template
length.</p>
<p>Working out how the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TLEN</font></font>
is computed takes some effort. You usually don&rsquo;t need to do it
manually, but it is good to know how it works. To do it you need the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CIGAR</font></font>
information:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 1,2,4,6,7,8,9 | grep SRR1972739.9152</font></font></p>
<p class="first-paragraph">We have isolated the information for the
two reads that form a pair:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.9152
  163     46    101M   =    230     281</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972739.9152
   83    230   97M4S   =     46    -281</font></font></p>
<p class="first-paragraph">The flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">163</font></font>
indicates that the alignment is on the forward strand (not reverse)
and the leftmost position of it is at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">46</font></font>
with its mate aligning at position <font face="Consolas, serif"><font size="2" style="font-size: 11pt">230</font></font>
on the reverse strand. The read has a CIGAR match of 101 bases, so
its right coordinate is at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">46+101=147</font></font>.
The mate aligns at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">230</font></font>
and has a CIGAR of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">97M4S</font></font>,
so it aligns over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">97</font></font>
bases only. This means that the rightmost coordinate of the second
read is at <font face="Consolas, serif"><font size="2" style="font-size: 11pt">230
+97 = 327</font></font> following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">46
    147       230      327</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
      |         |        |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">========&gt;
        &lt;=========</font></font></p>
<p class="first-paragraph">Note how from the above picture SAM only
reports <font face="Consolas, serif"><font size="2" style="font-size: 11pt">46</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">230</font></font>
(a major flaw in my opinion by the way) it takes a lot of effort to
find the other two coordinates. How long is the distance between the
outermost coordinates? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">327
- 46  = 281</font></font>, and that is the template length.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">46
    147       230      327</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|
      |         |        |</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">========&gt;
        &lt;=========</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|---------
 281 -----------|</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font>
</p>
<p class="first-paragraph">So the distance between the outer edges is
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">281</font></font>
that if under ideal conditions corresponds to the length of the
fragment that was sequenced. For the mate, the value is reported as
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-281</font></font>
to indicate that the mate is on the reverse strand.</p>
<p>It is exceedingly common to need the alignment length information,
and we typically have to resort to tools such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>
to convert</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
SRR1972739.bwa.bam | bedtools bamtobed  | head -1</font></font></p>
<p class="first-paragraph">That prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  45   146   SRR1972739.9152/2   60   +</font></font></p>
<p class="first-paragraph">But of course, as typical in
bioinformatics this does not precisely match what we computed above.
It is now a zero-based coordinate system - where the leftmost
coordinate is off by one <font face="Consolas, serif"><font size="2" style="font-size: 11pt">45</font></font>
instead of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">46</font></font>.
Gee thanks <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>.
Will the slightly confusing behaviors ever end? Wouldn&rsquo;t bet on
it.</p>
<h2 class="western"><a name="columns-10-11-seq-and-qual"></a>87.11
Columns 10-11: SEQ and QUAL</h2>
<p class="first-paragraph">These columns store the sequence and the
FASTQ quality measures associated with the query.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 10,11 | head -1</font></font></p>
<p class="first-paragraph">will produce the read sequence and its
FASTQ quality:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAATAACTATGAGGAAGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCATACCTGGTTTGTTTCAGAGC
 
C@CFFFFFHHHGHJJIJIJJJJJJJJJJJJIJJJJJJJJJJIIJJJJJJJJJJJJGJJJJJJJIJJJIIJIJIIJJIIJIIJHHHHHEFFFDEEEEECCDD</font></font></p>
<p class="first-paragraph">The most important detail (not quite
explained well in the SAM specification) is that these columns store
the sequences relative to the <i>forward</i> strand. What this means
is that queries that align to the reverse strand will be shown as
their reverse complement.</p>
<p>The soft and hard clip qualifiers of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CIGAR</font></font>
field describe whether the clipped and unaligned sequence is still
included (soft clip) or not (hard clip) in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEQ</font></font>
column.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">QUAL</font></font>
column contains those sequence qualities that directly correspond to
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEQ</font></font>
columns.</p>
<h2 class="western"><a name="optional-fields-columns-12-13-and-on"></a>
87.12 Optional Fields: Columns 12, 13 and on</h2>
<p class="first-paragraph">Here is where it gets interesting and
results produced by different aligners start to diverge. All optional
fields follow the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TAG:TYPE:VALUE</font></font>
format for example, where TAG is a two letter word, TYPE is one
character long (like i, Z, f) while the VALUE can be arbitrarily
long. Let us look at some tags:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam | cut -f 12,13,14,15 | head -1</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM:i:1
   MD:Z:81C19    MC:Z:97M4S    AS:i:96</font></font></p>
<p class="first-paragraph">As a rule, TAGs that start with an X or Y
are tool specific and are (should be) documented by the aligner.</p>
<h2 class="western"><a name="what-do-the-sam-tags-mean"></a>87.13
What do the SAM tags mean?</h2>
<p class="first-paragraph">The definitions for SAM tags are
distributed as a standalone document called <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMtags.pdf">SAMtags.pdf</a></font></p>
<ul>
	<li/>
<p class="compact">Tags are optional.</p>
	<li/>
<p class="compact">Only some upper cased tags have pre-defined
	meanings.</p>
	<li/>
<p class="compact">Tags that start with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>
	are aligner</p>
	<li/>
<p class="compact">Tags containing lowercase letters are
	always specific to the tool that filled them in.</p>
</ul>
<p class="first-paragraph">For example, we can see there that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM</font></font>
is defined as &ldquo;edit distance to the reference, including
ambiguous bases but excluding clipping&rdquo;. And <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MD</font></font>
is <i>string for mismatching positions</i>, yet another
representation of the alignment itself, similar to CIGAR but not
quite the same. There is a long story to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MD</font></font>
tag, one that will be covered later.</p>
<p>The take-home lesson from the above is that the SAM format
represents alignments and stores a relatively complex set of
information on the way sequences align.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="working-with-bam-files"></a>88 Working
with BAM files</h1>
<p class="first-paragraph">The SAM/BAM format is the workhorse format
of bioinformatics. The information in the SAM files (or lack thereof)
may determine the success or failure of a project.</p>
<p>Several tool sets have been created to manipulate BAM files:</p>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
	- one of the most powerful data analysis tools in all of
	bioinformatics. Complete data analytics may be performed using
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
	alone.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools</font></font>
	- originally conceived of to fill in functionality missing from
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">picard</font></font>
	- a toolset developed at the Broad Institute (named after Captain
	Picard in the Star Trek Next Generation series); Sometimes used in
	close conjunction with the GATK variant caller.</p>
</ol>
<p class="first-paragraph">Many operations on BAM files fall into one
of these two categories:</p>
<ul>
	<li/>
<p class="compact">Select alignments that match an attribute:
	strand, mate information, mapping quality</p>
	<li/>
<p class="compact">Select alignments that align in a certain
	region of the genome.</p>
</ul>
<h2 class="western"><a name="how-can-i-better-understand-the-flags"></a>
88.1 How can I better understand the FLAGS?</h2>
<p class="first-paragraph">The best resource to expore SAM flags is a
site called</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://broadinstitute.github.io/picard/explain-flags.html">Explain
	SAM Flags</a></font>.</p>
</ul>
<p class="first-paragraph">The service allows you to iteratively
build flags and to reverse their meaning.</p>
<h2 class="western"><a name="where-do-i-get-a-bam-file-1"></a>88.2
Where do I get a BAM file?</h2>
<p class="first-paragraph">The commands in this section assume that
your BAM was generated in the manner described in the <font color="#4f81bd"><a href="#ch086.xhtml#make-a-bam">How
to make a BAM file</a></font> section.</p>
<p>You may also choose a remotely hosted BAM file. For example select
data from publication discussed in <font color="#4f81bd"><a href="#ch056.xhtml#redo-denisovan">Redo:
Explore the genome of a paleolithic-era archaic human</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Location of the BAM
file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=http://cdna.eva.mpg.de/denisova/alignments/T_hg19_1000g.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download alignments for chromosome 22.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b $URL 22 &gt; denisova.chr22.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the downloaded alignments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index denisova.chr22.bam</font></font></p>
<p class="first-paragraph">Alternatively you may explore a subset of
data from <font color="#4f81bd"><a href="#ch055.xhtml#syndipredo">Redo:
A synthetic-diploid benchmark for accurate variant-calling evaluation</a></font>
(takes about 15 minutes to download)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Location of the BAM
file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=ftp://ftp.sra.ebi.ac.uk/vol1/run/ERR134/ERR1341796/CHM1_CHM13_2.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download alignments for chromosome 22.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b $URL 22 &gt; syndip.chr22.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the downloaded alignments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index syndip.chr22.bam</font></font></p>
<h2 class="western"><a name="how-can-i-extract-a-section-of-the-bam-file"></a>
88.3 How can I extract a section of the BAM file?</h2>
<p class="first-paragraph">The need to isolate a subsection of a bam
file is exceedingly common. The SAM format was designed around this
use case. The &ldquo;slice&rdquo; your SAM file over a region pass a
&ldquo;query&rdquo; string at the end of the command. The query
string should be in the format &ldquo;chrom:start-end&rdquo; in a one
based coordinate system:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view SRR1972739.bwa.bam AF086833:1000-1010 | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Prints 82</font></font></p>
<p class="first-paragraph">to produce another BAM file, that contains
only data for the &ldquo;slice&rdquo; pass the -b flag to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a new BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b SRR1972739.bwa.bam AF086833:1000-1010 &gt; selected.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the new BAM file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index selected.bam</font></font></p>
<p class="first-paragraph">since we tend to forget to re-index the
BAM file when running quick examples, it makes sense to combine the
two commands into a single one, separated by the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&amp;&amp;</font></font>
symbols that bash interprets as &ldquo;AND&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b SRR1972739.bwa.bam AF086833:1000-1010 &gt; selected.bam &amp;&amp;
samtools index selected.bam</font></font></p>
<p class="first-paragraph">In the future examples we will combine the
two commands into one. When visualized in IGV the resulting BAM file
would look like this:</p>
<p><img src="0000_html_14b977a8a237e631.png" name="Image131" align="bottom" hspace="12" width="560" height="296" border="0"/>
</p>
<p>As the above image shows, after taking the slice only those
alignments were retained that overalapped with the coordinate
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1000-1010</font></font>.</p>
<h2 class="western"><a name="how-can-i-select-from-or-filter-data-from-bam-files"></a>
88.4 How can I select from or filter data from BAM files?</h2>
<p class="first-paragraph">First, let&rsquo;s clarify some wording:</p>
<ol>
	<li/>
<p class="compact">Selecting means to keep alignments that
	match a condition.</p>
	<li/>
<p class="compact">Filtering means to remove alignments that
	match a condition.</p>
</ol>
<p class="first-paragraph">Filtering on <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAGS</font></font>
can be done via <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
by passing the parameters <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-f</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-F</font></font>
that operate the following way:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-f
	flag</font></font> includes only alignments where the bits match the
	bits in the flag</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-F
	flag</font></font> includes only alignments where the bits DO NOT
	match the bits in the flag.</p>
</ul>
<h2 class="western"><a name="how-do-i-valid-mapped-alignments"></a>88.5
How do I valid (mapped) alignments?</h2>
<p class="first-paragraph">First recall what flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>
means:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags 4</font></font></p>
<p class="first-paragraph">It states:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0x4
4   UNMAP</font></font></p>
<p class="first-paragraph">This means that when this flag is set, the
read is unmapped (unaligned). You may wonder why this alignment is
included at all in the BAM file if there is no actual alignment for
it. We often need to know which read did not produce an alignment,
thus some aligners include them by default. Others do not. Again one
of the differences between aligners. To select and view alignments
where the read did not align you can do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -f 4 SRR1972739.bwa.bam | head</font></font></p>
<p class="first-paragraph">Counting alignments with a property is so
frequently needed that samtools provides flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c</font></font>
to do that:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -f 4 SRR1972739.bwa.bam</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5461</font></font></p>
<p class="first-paragraph">Thus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5461</font></font>
reads in our file are not actually aligned. To count those that are
aligned reverse the flags:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
How many don't have the UNMAP flag?</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
15279</font></font></p>
<h2 class="western"><a name="how-do-i-select-forward-or-reverse-alignments"></a>
88.6 How do I select forward or reverse alignments?</h2>
<p class="first-paragraph">To select alignments on the forward strand
we would filter out UNMAP (4) and REVERSE (16) (4+16=20):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -F 20 -b SRR1972739.bwa.bam &gt; selected.bam &amp;&amp;
samtools index selected.bam</font></font></p>
<p class="first-paragraph"><img src="0000_html_59c734b54e0ddf92.png" name="Image132" align="bottom" hspace="12" width="560" height="296" border="0"/>
</p>
<p>To select alignments on the reverse strand we would filter out
UNMAP(4) but select for REVERSE (14):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -F 4 -f 16 -b SRR1972739.bwa.bam &gt; selected.bam &amp;&amp;
samtools index selected.bam</font></font></p>
<p class="first-paragraph"><img src="0000_html_98999c4b0b879fef.png" name="Image133" align="bottom" hspace="12" width="560" height="296" border="0"/>
</p>
<h2 class="western"><a name="how-to-separate-alignments-into-a-new-file"></a>
88.7 How to separate alignments into a new file?</h2>
<p class="first-paragraph">Removing the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c</font></font>
parameter will produce the alignments, so if you wanted to create
another BAM file that only contains the aligned reads you could do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b -F 4 SRR1972739.bwa.bam &gt; aligned.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index aligned.bam</font></font></p>
<h2 class="western"><a name="how-to-get-an-overview-of-the-alignments-in-a-bam-file"></a>
88.8 How to get an overview of the alignments in a BAM file?</h2>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
	flagstat SRR1972739.bwa.bam</font></font> - produces a report on
	flags</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
	idxstats SRR1972739.bwa.bam</font></font> - produces a report on how
	many reads align to each chromosome</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
	stats -in SRR1972739.bwa.bam</font></font> - produces a report on
	flags</p>
</ol>
<p class="first-paragraph">One observation that we made over the
years is that the numbers reported by the various statistics programs
never quite match! The numbers are similar but never all the same in
each category.</p>
<p>This is an indication that even at the basic level there is
ambiguity in what the different terms mean to different people. Note,
for example, how below <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat</font></font> reports <font face="Consolas, serif"><font size="2" style="font-size: 11pt">14480</font></font>
as &ldquo;proper pairs&rdquo; whereas <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
stats</font></font> reports <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15216</font></font>
for the same quantity.</p>
<p>The output of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20740
+ 0 in total (QC-passed reads + QC-failed reads)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 secondary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">740
+ 0 supplementary</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 duplicates</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">15279
+ 0 mapped (73.67% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20000
+ 0 paired in sequencing</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000
+ 0 read2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">14480
+ 0 properly paired (72.40% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">14528
+ 0 with itself and mate mapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">11
+ 0 singletons (0.05% : N/A)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
+ 0 with mate mapped to a different chr (mapQ&gt;=5)</font></font></p>
<p class="first-paragraph">The output of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
stats</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">**********************************************</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Stats
for BAM file(s):
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">**********************************************</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Total
reads:       20740</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Mapped
reads:      15279    (73.6692%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Forward
strand:    14393    (69.3973%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Reverse
strand:    6347 (30.6027%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Failed
QC:         0    (0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Duplicates:
       0    (0%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Paired-end
reads:  20740    (100%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">'Proper-pairs':
   15216    (73.3655%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Both
pairs mapped: 15268    (73.6162%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Read
1:            10357</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Read
2:            10383</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Singletons:
       11   (0.0530376%)</font></font></p>
<p class="first-paragraph">How could you verify which one is correct?
Take the definition of proper pair as flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -f 2 SRR1972739.bwa.bam </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
15216</font></font></p>
<p class="first-paragraph">If you count the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PROPER_PAIR</font></font>
flag the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
stats</font></font> is correct.</p>
<p>So how does <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flagstat</font></font> come up with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">14480</font></font>?
It turns out that it counts only the properly mapped AND primary
alignment reads. Basically, it is counting each read only once in the
cases that a read produces multiple alignments:</p>
<p>How do we produce the alignment the primary alignments? It is a
bit circuitous, we filter out the unmapped, secondary and
supplementary alignments</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags UNMAP,SECONDARY,SUPPLEMENTARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0x904
  2308    UNMAP,SECONDARY,SUPPLEMENTARY</font></font></p>
<p class="first-paragraph">all we need is to filter (remove
alignments) with flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2308</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -f 2 -F 2308 SRR1972739.bwa.bam </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
14480</font></font></p>
<h2 class="western"><a name="what-is-a-proper-pair"></a>88.9 What is
a proper-pair?</h2>
<p class="first-paragraph">&ldquo;Proper pair&rdquo;, called
&ldquo;concordant&rdquo; in bowtie is a fuzzy and under specified
term . The SAM specification only states that PROPER_PAIR <i>&ldquo;each
segment properly aligned according to the aligner&rdquo;</i> which is
no definition at all. The aligner, may decide to start label certain
alignments as PROPER_PAIR.</p>
<p>Typically what proper pair means is that the read pair aligns in
an expected manner. The distance between the outer edges is within
expected ranges, the orientation of the reads are towards one another
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">---&gt;
&lt;----</font></font>.</p>
<p>The word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PROPER_PAIR</font></font>
is loaded and somewhat misleading, it appears to suggest that these
alignments are more &ldquo;correct&rdquo; than others. That&rsquo;s
not what it supposed to mean or imply. Always remember that when we
align against a reference, we align real data against an imaginary,
hypothetical reference. If our real data has a genome reorganization,
it will produce sequences that when forced over the hypotetical
reference may show up as &ldquo;improper pairs&rdquo;.</p>
<p>It is also true, of course, that errors in the protocols may
produce flawed data that also aligns incorrectly. Our job is to
recognize and distinguish between the two cases.</p>
<p>Often it is interesting to see which alignments are not &ldquo;proper&rdquo;
or &ldquo;primary&rdquo;. These often carry information on the
differences relative to the reference genome. Let&rsquo;s filter out
the UNMAP (4) and PROPER_PAIR (2)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -F 6 -b SRR1972739.bwa.bam &gt; selected.bam &amp;&amp; samtools
index selected.bam</font></font></p>
<p class="first-paragraph">the command above will produce the
selection</p>
<p><img src="0000_html_e2b568c50edfa0d0.png" name="Image134" align="bottom" hspace="12" width="560" height="296" border="0"/>
</p>
<p>Few reads, distributed all across the genome. The alignments do
not seem to indicate a genomic reorganization.</p>
<h2 class="western"><a name="how-do-i-use-flags"></a>88.10 How do I
use flags?</h2>
<p class="first-paragraph">The most important rule is to ensure that
you filter on flag <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>,
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">UNMAP</font></font>.
Even though it makes little sense that a read could be simultaneously
unmapped and on a reverse strand, this can be reported as such:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignments that are not mapped to a location</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -f 4  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
5461</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Alignments
that are mapped to a location.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
15279</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reads that align to reverse strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -f 16  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
6343</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignments to forward strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -F 16  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
8936</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignments  that are unmapped and on the reverse strand!  Say
what?</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -f 4 -f 16  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></p>
<p class="first-paragraph">Various curious situations may occur:</p>
<ul>
	<li/>
<p class="compact">Biostar Post of the Day: <font color="#4f81bd"><a href="https://www.biostars.org/p/14569/#14570">How
	can an unmapped read align on the reverse strand?</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-combine-multiple-flags"></a>88.11
How do I combine multiple flags?</h2>
<p class="first-paragraph">This is where it gets even more
complicated.</p>
<p>When investigating any attribute we may need to select for and
against certain features. But as we combine the flags, the way we
order the flags is not symmetric anymore when using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-f</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-F</font></font>
together. To select all reads that align (oppose <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>)
and are mapped to the reverse strand (select for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">16</font></font>)
we do this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignments on the reverse strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Oppose unmapped (4) and select for reverse strand (16).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -f 16  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
6343</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignments on the forward strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Oppose unmapped (4) and oppose the reverse strand (16).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -F 16  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
8936</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Typically the above is combined into a single number: 4 + 16 =
20</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 20  SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
8936</font></font></p>
<p class="first-paragraph">As you select for more attributes the
overhead of the mental gymnastics adds up. Here is an example that
seems contrived yet is a common task if you wanted to separate
strand-specific data (where the second pair has the same orientation
as the transcript). In that case if we wanted to separate our
alignments to correspond to forward and reverse transcripts we would
do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Select alignments for reads that are second in pair if they map to
the forward strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b -F 20 -f 128 data.bam &gt; fwd1.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Select alignments for reads that are the first in pair if they map to
the reverse strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -b -F 4 -f 80 data.bam &gt; fwd2.bam</font></font></p>
<p class="first-paragraph">Not surprisingly, the above quite
challenging to figure out and apply correctly.</p>
<p>It gets even more complicated for paired-end strand-specific data.
We had to write a tutorial (mainly so that we also remember it) on
this:</p>
<ul>
	<li/>
<p class="compact">Biostar Post of the Day: <font color="#4f81bd"><a href="https://www.biostars.org/p/92935/">How
	To Separate Illumina Based Strand Specific Rna-Seq Alignments By
	Strand</a></font></p>
</ul>
<h2 class="western"><a name="how-should-i-interpret-seemingly-impossible-alignments"></a>
88.12 How should I interpret seemingly impossible alignments?</h2>
<p class="first-paragraph">It is possible to create selections that
seemingly make no sense. For example, let us select alignments where
the read and its mate both align to the reverse strand.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags PAIRED,PROPER_PAIR,REVERSE,MREVERSE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
0x33  51  PAIRED,PROPER_PAIR,REVERSE,MREVERSE</font></font></p>
<p class="first-paragraph">There are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">124</font></font>
of these:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Unique alignments, mapped, paired, proper pair, both reads </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
in a pair align to the reverse strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -q 1 -F 4 -f 51 SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
124</font></font></p>
<p class="first-paragraph">If our genome matched the reference and
the alignments were correct this would be impossible to observe.
Given that we do see these, there are two explanations.</p>
<ol>
	<li/>
<p class="compact">Incorrect alignments: The reads have been
	misplaced.</p>
	<li/>
<p class="compact">Genomic rearrangement: The real genome has
	a sequence inversion that makes it &ldquo;look like&rdquo; one read
	of the pair is inverted.</p>
</ol>
<p class="first-paragraph">Remember that the real genome does not
need to match the reference - real, biological changes could look
like incorrect mapping.</p>
<h2 class="western"><a name="how-do-i-filter-on-mapping-quality"></a>88.13
How do I filter on mapping quality?</h2>
<p class="first-paragraph">The mapping quality column contains a
number that the aligner fills in with a value that ought to reflect
the likelihood of errors. This number is rarely an objective
quantity, it rather reflects a subjective value designated by the
tool developer. For example, for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
a mapping quality of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font>
indicates a read that maps to multiple locations. Does this make
sense? Not really, the likelihood that a read can be in multiple
locations should be closer to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/N</font></font>
(<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
is the number of possible alignments) - but <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
reports it as zero.</p>
<p>Do other tools use the same rationale? Typically not. You have to
check the documentation for the particular software you are using to
know for sure.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-q</font></font>
flag selects alignments whose mapping quality is at or above the
value</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Select uniquely mapped reads when these were aligned with
BWA.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -q 1 SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
15279</font></font></p>
<p class="first-paragraph">Selection options can be combined:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Mapped alignments with quality over 1 and on the forward
strand.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -q 1 -F 4 -F 16 SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
8936</font></font></p>
<h2 class="western"><a name="how-can-i-find-out-the-depth-of-coverage"></a>
88.14 How can I find out the depth of coverage?</h2>
<p class="first-paragraph">There are several tools that can compute
the number of reads that overlap each base. With samtools you can do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Compute the depth of coverage.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
depth SRR1972739.bwa.bam  | head</font></font></p>
<p class="first-paragraph">This produces a file with name, coordinate
and coverage on that coordinate:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   46  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   47  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   48  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
coverage</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamtools
coverage -in SRR1972739.bwa.bam | head</font></font></p>
<p class="first-paragraph">The output seems identical</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   45  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   46  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   47  1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Except note how it is &ldquo;one off&rdquo;!
The index starts at 0 rather than 1 in the latter case. These type of
inconsistencies are extraordinarily common in bioinformatics and are
the source of endless errors.</p>
<p>You could ask for all positions to be shown with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font>
and you can verify that it starts at 1</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
depth -a SRR1972739.bwa.bam  | head</font></font></p>
<p class="first-paragraph">it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   1   0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   2   0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   3   0</font></font></p>
<p class="first-paragraph">You could sort by depth, though this could
take a long for a large genome size. To find the most covered region:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
depth SRR1972739.bwa.bam  | sort -k 3 -rn | head</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   4609    163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   4608    163</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   4611    157</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<h2 class="western"><a name="what-is-a-secondary-alignment"></a>88.15
What is a &ldquo;SECONDARY&rdquo; alignment?</h2>
<p class="first-paragraph">A secondary alignment refers to a read
that produces multiple alignments in the genome. One of these
alignments will be typically referred to as the &ldquo;primary&rdquo;
alignment.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags SECONDARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
0x100 256 SECONDARY</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -f 256 SRR1972739.bwa.bam </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
0</font></font></p>
<p class="first-paragraph">There are no reads that map in multiple
locations in this data set.</p>
<p>The SAM specification states:</p>
<p class="block-text-western">Multiple mappings are caused primarily
by repeats. They are less frequent given longer reads. If a read has
multiple mappings, and all these mappings are almost entirely
overlapping with each other; except the single-best optimal mapping,
all the other mappings get mapping quality &lt;Q3 and are ignored by
most SNP/INDEL callers.</p>
<p class="first-paragraph">To be honest we don&rsquo;t fully
understand the statement above.The definition is somewhat meandering
and fuzzy. Oh look, I even have the gall to criticise someone else&rsquo;s
writing, especially considering the quantity of typos and broken
sentences that I myself produce.</p>
<h2 class="western"><a name="what-is-a-supplementary-alignment"></a>88.16
What is a &ldquo;SUPPLEMENTARY&rdquo; alignment?</h2>
<p class="first-paragraph">A supplementary alignment (also known as a
chimeric alignment) is an alignment where the read partially matches
different regions of the genome without overlapping the same
alignment:</p>
<p>If this the original read was say 9 bp long:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">---------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">123456789</font></font></p>
<p class="first-paragraph">Then these alignments would be reported as
two alignments, one of them chimeric:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">------------------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   ---        ------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   123        456789</font></font></p>
<p class="first-paragraph">Above the entire sequence aligns in two
different locations with no overlap. Another arrangement as below
would not be reported as chimeric:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">----------------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    -----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    12345</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        ----</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        6789</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font>
</p>
<p class="first-paragraph">Typically, one of the linear alignments in
a chimeric alignment is considered the &ldquo;representative&rdquo;
or &ldquo;primary&rdquo; alignment, and the other is called
&ldquo;supplementary&rdquo; The decision regarding which linear
alignment is representative (primary) is not all that well
documented. We assume it is the alignment with higher score.</p>
<p>The SAM specification states:</p>
<p class="block-text-western">Chimeric alignment is primarily caused
by structural variations, gene fusions, misassemblies, RNA-seq or
experimental protocols. []. For a chimeric alignment, the linear
alignments consisting of the aligment are largely non-overlapping;
each linear alignment may have high mapping quality and is
informative in SNP/INDEL calling.</p>
<p class="first-paragraph">To select supplementary alignments:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags SUPPLEMENTARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
0x800 2048    SUPPLEMENTARY</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -f 2048 SRR1972739.bwa.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
740</font></font></p>
<h2 class="western"><a name="what-is-a-primary-or-representative-alignment"></a>
88.17 What is a &ldquo;PRIMARY&rdquo; or &ldquo;REPRESENTATIVE&rdquo;
alignment?</h2>
<p class="first-paragraph">Typically these represent the &ldquo;best&rdquo;
alignment although the word &ldquo;best&rdquo; may not be well
defined. Each read will only have one primary alignment and other
secondary and supplemental alignments.</p>
<p>There is no flag to select for primary alignments, the opposite of
&ldquo;SUPPLEMENTARY&rdquo; and &ldquo;SECONDARY&rdquo; will be the
primary alignment. To select primary alignments:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags SUPPLEMENTARY,SECONDARY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
0x900 2304    SECONDARY,SUPPLEMENTARY</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -c -F 4 -F 2304 SRR1972739.bwa.bam </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
14539</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-sam-reference-sheet"></a>89 The SAM
reference sheet</h1>
<p class="first-paragraph">This page collects information on SAM
files that is often needed.</p>
<ul>
	<li/>
<p class="compact">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMv1.pdf">SAM
	format specification</a></font> is the official specification of the
	SAM format.</p>
	<li/>
<p class="compact">The <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMtags.pdf">SAM
	tag specification</a></font> is the official specification of the
	SAM tags.</p>
</ul>
<h2 class="western"><a name="what-are-the-required-sam-columns"></a>89.1
What are the required SAM columns?</h2>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">QNAME</font></font>
	String [!-?A-~]{1,254} Query template NAME</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FLAG</font></font>
	Int [0,65535] bitwise FLAG</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNAME</font></font>
	String *|[!-()+-&lt;&gt;-<sub>][!-</sub>]* Reference sequence NAME</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>
	Int [0,2147483647] 1-based leftmost mapping POSition</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MAPQ</font></font>
	Int [0,255] MAPping Quality</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CIGAR</font></font>
	String *|([0-9]+[MIDNSHPX=])+ CIGAR string</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">RNEXT</font></font>
	String *|=|[!-()+-&lt;&gt;-<sub>][!-</sub>]* Ref. name of the
	mate/next read</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PNEXT</font></font>
	Int [0,2147483647] Position of the mate/next read</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TLEN</font></font>
	Int [-2147483647,2147483647] observed Template LENgth</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SEQ</font></font>
	String *|[A-Za-z=.]+ segment SEQuence</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">QUAL</font></font>
	String [!-~]+ ASCII of Phred-scaled base QUALity+33</p>
</ol>
<h2 class="western"><a name="what-do-the-flags-mean"></a>89.2 What do
the flags mean?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in">   <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Binary
      Integer   Name              Meaning </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000000001
      1     PAIRED          Read paired</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000000010
      2     PROPER_PAIR     Read mapped in proper pair</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000000100
      4     UNMAP           Read unmapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000001000
      8     MUNMAP          Mate unmapped</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000010000
     16     REVERSE         Read reverse strand</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000000100000
     32     MREVERSE        Mate reverse strand</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000001000000
     64     READ1           First in pair, file 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000010000000
    128     READ2           Second in pair, file 2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">000100000000
    256     SECONDARY       Not a primary alignment </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">001000000000
    512     QCFAIL          Read fails platform/vendor quality
checks</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">010000000000
   1024     DUP             Read is PCR or optical
duplicate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">100000000000
   2048     SUPPLEMENTARY   Supplementary alignment</font></font></p>
<h2 class="western"><a name="how-to-build-compound-flags"></a>89.3
How to build compound flags?</h2>
<p class="first-paragraph">The best resource is a site called <font color="#4f81bd"><a href="https://broadinstitute.github.io/picard/explain-flags.html">Explain
SAM Flags</a></font>. The site allows us to not only build a series a
flags but to instantly reverse their meaning, a process that can be
very error-prone otherwise.</p>
<p>It is worth reflecting for a moment that the format is so
complicated that we need a separate, standalone website just to
understand one of its columns.</p>
<p>You can also build the flag that has all the right bits at the
command line:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags PAIRED,PROPER_PAIR,REVERSE</font></font></p>
<p class="first-paragraph">will print the integers value <font face="Consolas, serif"><font size="2" style="font-size: 11pt">19</font></font>
:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0x13
   19    PAIRED,PROPER_PAIR,REVERSE</font></font></p>
<p class="first-paragraph">You can also go in &ldquo;reverse&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
flags 19</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0x13
   19  PAIRED,PROPER_PAIR,REVERSE</font></font></p>
<h2 class="western"><a name="what-are-sam-tags"></a>89.4 What are SAM
tags?</h2>
<p class="first-paragraph">There are two types of tags:</p>
<ul>
	<li/>
<p class="compact">Header tags: These are predefined in the
	header and each one is present for each alignment.</p>
	<li/>
<p class="compact">Alignment tags: These are filled in by the
	aligner and may be predefined or custom tags.</p>
</ul>
<h2 class="western"><a name="what-do-the-sam-header-tags-mean"></a>89.5
What do the SAM header tags mean?</h2>
<p class="first-paragraph">Header tags start with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@</font></font>
symbol and are at the beginning of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SAM</font></font>
file.</p>
<h3 class="western"><a name="hd-the-header-line"></a>89.5.1 @HD: The
header line</h3>
<p class="first-paragraph">The first line if present.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">VN</font></font>
	Format version</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SO</font></font>
	Sorting order of alignments</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GO</font></font>
	Grouping of alignments</p>
</ul>
<h3 class="western"><a name="pg-program"></a>89.5.2 @PG: Program</h3>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID</font></font>
	Program record identifier</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PN</font></font>
	Program name</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CL</font></font>
	Command line</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PP</font></font>
	Previous @PG-ID</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">DS</font></font>
	Description</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">VN</font></font>
	Program version</p>
</ul>
<h3 class="western"><a name="rg-read-group"></a>89.5.3 @RG: Read
group</h3>
<p class="first-paragraph">This is one of the most important headers
in a SAM file.</p>
<p>Unordered multiple <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@RG</font></font>
lines are allowed.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID</font></font>
	Read group identifier</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CN</font></font>
	Name of sequencing center producing the read</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">DS</font></font>
	Description</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">DT</font></font>
	Date the run was produced (ISO8601 date or date/time)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FO</font></font>
	Flow order</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">KS</font></font>
	The array of nucleotide bases that correspond to the key sequence of
	each read</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LB</font></font>
	Library</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PG</font></font>
	Programs used for processing the read group</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PI</font></font>
	Predicted median insert size</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL</font></font>
	Platform/technology used to produce the reads</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PM</font></font>
	Platform model</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PU</font></font>
	Platform unit, a unique identifier</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SM</font></font>
	Sample</p>
</ul>
<h3 class="western"><a name="sq-reference-sequence-dictionary"></a>89.5.4
@SQ: Reference sequence dictionary</h3>
<p class="first-paragraph">The order of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SQ</font></font>
lines defines the alignment sorting order.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SN</font></font>
	Reference sequence name</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LN</font></font>
	Reference sequence length</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS</font></font>
	Genome assembly identifier</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">M5</font></font>
	MD5 checksum of the sequence in the uppercase</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SP</font></font>
	Species</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">UR</font></font>
	URI of the sequence</p>
</ul>
<h2 class="western"><a name="what-do-the-sam-alignment-tags-mean"></a>
89.6 What do the SAM alignment tags mean?</h2>
<p class="first-paragraph">The definitions for SAM tags are
distributed as a standalone document called <font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/SAMtags.pdf">SAMtags.pdf</a></font></p>
<ul>
	<li/>
<p class="compact">Tags are optional.</p>
	<li/>
<p class="compact">Only some upper case tags have pre-defined
	meanings.</p>
	<li/>
<p class="compact">Tags that start with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>,
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y</font></font>&rsquo;
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Z</font></font>
	or are in lowercase letters are always specific to the tool that
	filled them in. You need to consult the tool documentation for what
	they mean.</p>
</ul>
<p class="first-paragraph">The current specification calls for the
following pre-defined tags but please do note that no aligner fills
in all and most aligners only fill very few of them:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AM</font></font>
	The smallest template-independent mapping quality of segments in the
	rest</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS</font></font>
	Alignment score generated by aligner</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BC</font></font>
	Barcode sequence</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BQ</font></font>
	Offset to base alignment quality (BAQ)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CC</font></font>
	Reference name of the next hit</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CM</font></font>
	Edit distance between the color sequence and the color reference
	(see also NM)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CO</font></font>
	Free-text comments</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CP</font></font>
	Leftmost coordinate of the next hit</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CQ</font></font>
	Color read base qualities</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CS</font></font>
	Color read sequence</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CT</font></font>
	Complete read annotation tag, used for consensus annotation dummy
	features.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">E2</font></font>
	The 2nd most likely base calls</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FI</font></font>
	The index of segment in the template</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FS</font></font>
	Segment suffix</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FZ</font></font>
	S Flow signal intensities</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GQ</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GS</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">H0</font></font>
	Number of perfect hits</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">H1</font></font>
	Number of 1-difference hits (see also NM)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">H2</font></font>
	Number of 2-difference hits</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">HI</font></font>
	Query hit index</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">IH</font></font>
	Number of stored alignments in SAM that contains the query in the
	current record</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">LB</font></font>
	Library</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MC</font></font>
	CIGAR string for mate/next segment</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MD</font></font>
	String for mismatching positions</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MF</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MQ</font></font>
	Mapping quality of the mate/next segment</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NH</font></font>
	Number of reported alignments that contains the query in the current
	record</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM</font></font>
	Edit distance to the reference</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">OC</font></font>
	Original CIGAR</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">OP</font></font>
	Original mapping position</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">OQ</font></font>
	Original base quality</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PG</font></font>
	Program</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PQ</font></font>
	Phred likelihood of the template</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PT</font></font>
	Read annotations for parts of the padded read sequence</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PU</font></font>
	Platform unit</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">QT</font></font>
	Barcode (BC or RT) phred-scaled base qualities</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Q2</font></font>
	Phred quality of the mate/next segment sequence in the R2 tag</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2</font></font>
	Sequence of the mate/next segment in the template</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">RG</font></font>
	Read group</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">RT</font></font>
	Barcode sequence (deprecated; use BC instead)</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SA</font></font>
	Other canonical alignments in a chimeric alignment</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SM</font></font>
	Template-independent mapping quality</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SQ</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">S2</font></font>
	Reserved for backwards compatibility reasons</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TC</font></font>
	The number of segments in the template</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">U2</font></font>
	Phred probability of the 2nd call being wrong conditional on the
	best being wrong</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">UQ</font></font>
	Phred likelihood of the segment, conditional on the mapping being
	correct</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">X?</font></font>
	Reserved for end users</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y?</font></font>
	Reserved for end users</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Z?</font></font>
	Reserved for end users</p>
</ul>
<h2 class="western"><a name="what-are-bwa-aligner-specific-tags"></a>89.7
What are BWA aligner specific tags?</h2>
<p class="first-paragraph">Out of the pre-defined tags <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
fills the following:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NM</font></font>
	Edit distance</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MD</font></font>
	Mismatching positions/bases</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AS</font></font>
	Alignment score</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BC</font></font>
	Barcode sequence</p>
</ul>
<p class="first-paragraph">Lowercase tags or those that start with a
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Z</font></font>
are application specific. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
aligner fills in the following <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
specific tags:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">X0</font></font>
	Number of best hits</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">X1</font></font>
	Number of suboptimal hits found by BWA</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XN</font></font>
	Number of ambiguous bases in the referenece</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XM</font></font>
	Number of mismatches in the alignment</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XO</font></font>
	Number of gap opens</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XG</font></font>
	Number of gap extentions</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XT</font></font>
	Type: Unique/Repeat/N/Mate-sw</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XA</font></font>
	Alternative hits; format: (chr,pos,CIGAR,NM;)*</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XS</font></font>
	Suboptimal alignment score</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XF</font></font>
	Support from forward/reverse alignment</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">XE</font></font>
	Number of supporting seed</p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="sequence-variation-in-sam-files"></a>90
Sequence variation in SAM files</h1>
<p class="first-paragraph">The section discusses how variants are
represented inside SAM files. The concept we cover are somewhat
technical but don&rsquo;t let that discourage you. Being confident
about your files means that you fully understand how information is
represented inside them.</p>
<p>Thankfully in the majority of cases, you won&rsquo;t need to
operate on your files like this. But I guarantee you there will be a
time when you do need to look inside the SAM file. At that time
knowing where to look will save you substantial effort.</p>
<p>For example, when variation calling goes awry, you will need to
investigate the alignment file and attempt to understand the
information encoded in the file.</p>
<p>Let&rsquo;s prepare some data for this chapter.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   REF=refs/AF086833.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   efetch -db nuccore -format fasta -id AF086833 &gt; $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index reference with bwa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    bwa index $REF</font></font></p>
<h2 class="western"><a name="how-to-mutate-a-sequence-from-the-command-line"></a>
90.1 How to mutate a sequence from the command line?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
EMBOSS tool was designed to generate mutations in sequences. It can
produce point mutations with the flag:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-point
Types of point mutations to perform</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      0 (None); </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      1 (Any of the following)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      2 (Insertions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      3 (Deletions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      4 (Changes)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      5 (Duplications)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      6 (Moves)</font></font></p>
<p class="first-paragraph">Or block mutations with the flags</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-block
Types of block mutations to perform</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      0 (None)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      1 (Any of the following)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      2 (Insertions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      3 (Deletions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      4 (Changes)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      5 (Duplications)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      6 (Moves)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font>
</p>
<p class="first-paragraph">Block mutations are a series of
consecutive mutations that span from ranges specified with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-minimum</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-maximum</font></font>.</p>
<p>Instead of aligning millions of reads will will use our short read
aligner to align just one with known properties. We will then look at
how the variant is reflected within the SAM file.</p>
<p>Get a 50bp query from the reference file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | seqret -filter -send 50  &gt; query.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Introduce a point deletion </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -point 3 &gt; mutated.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Visualize the output:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">global-align.sh
query.fa mutated.fa </font></font>
</p>
<p class="first-paragraph">It will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||||||||||||||||||||||||
|||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTA-GATCTTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">Perhaps you&rsquo;d want to make a
different mutation, with a 2 base block mutation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -block 3 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">that case produced on my system the
following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||||||||||||||||||||||||||||||||
 |||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGT--GAATA</font></font></p>
<p class="first-paragraph">We can introduce more than one mutation
with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-count</font></font>
parameter:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -count 2 -block 3 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">that will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||
 |||||||||||||||||| 
|||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAA--AAAGAAGAATTTTTAGGA--TTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">Since the tool selects mutations randomly,
when choosing mutations like substitutions it is possible to mutate
the sequence back to itself i.e.&nbsp;replacing an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
with another <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font></p>
<p>One flaw of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
tool is that it does not tell us what changes it has made.</p>
<h2 class="western"><a name="how-do-i-reconstruct-alignment-from-cigar-and-md-strings"></a>
90.2 How do I reconstruct alignment from CIGAR and MD strings?</h2>
<p class="first-paragraph">The encoding of the SAM file (add
cross-reference to SAM format information) makes reconstructing the
actual alignment a little more difficult than one might imagine. Let
us walk through a simple example of mutating a 50bp sequence from the
Ebola genome, then reconstructing that information from its alignment
back to reference.</p>
<p>In a nutshell we will create the mutation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||.|||||||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACATAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">and this above will be represented in the
CIGAR <font face="Consolas, serif"><font size="2" style="font-size: 11pt">50M</font></font>
and MD tag of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">8C41</font></font><br/>
You
can see advantages of the shortened representations the entire long
stretch above is stored in the short word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">8C41</font></font></p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-point
4</font></font> flag creates a substitution in a random location of
the sequence (in your case the mutation will be somewhere else and it
is also possible that the mutation generates the existing base. In
that case rerun it):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar --filter --point 4 &gt; mutated.fa</font></font></p>
<p class="first-paragraph">To see what the change is we can run our
global aligner:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">global-align.sh
query.fa mutated.fa </font></font>
</p>
<p class="first-paragraph">it will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||.|||||||||||||||||||||||||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACATAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA
</font></font>
</p>
<p class="first-paragraph">We can see how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
above replaced a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
with a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>.
So what will this look like in a SAM record? Let us generate the
alignment.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF mutated.fa &gt; align.sam </font></font>
</p>
<p class="first-paragraph">Then view the CIGAR, SEQUENCE and MD tags
(columns 6,10 and 13):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view align.sam | cut -f 6,10,13</font></font></p>
<p class="first-paragraph">This prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">50M
   CGGACACATAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA    MD:Z:8C41</font></font></p>
<p class="first-paragraph">The SAM specification states that the MD
field aims to achieve SNP/INDEL calling without looking at the
reference. What this means is that from the CIGAR, SEQUENCE and MD
columns we can reconstruct the original data and we don&rsquo;t need
to know the reference.</p>
<p>The MD tag is value is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">8C41</font></font>.
It says that 8 bases match then the 9th is a mismatch and the
reference contains a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
at that location. The 9th base in the SEQUENCE is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
hence this we now know this is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
variation.</p>
<h2 class="western"><a name="what-will-deletions-look-like-in-a-sam-format"></a>
90.3 What will deletions look like in a SAM format?</h2>
<p class="first-paragraph">Let&rsquo;s look at a block deletion:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -count 2 -block 3 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">That produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||
 |||||||||||||||||| 
|||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAA--AAAGAAGAATTTTTAGGA--TTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">When aligned with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
the resulting CIGAR and MD tags are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">13M2D18M2D15M</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">13^AG18^TC15</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">13M2D18M2D15M
  CGGACACACAAAAAAAGAAGAATTTTTAGGATTTTGTGTGCGAATA  MD:Z:13^AG18^TC15</font></font></p>
<p class="first-paragraph">Note how the CIGAR string tells us that we
have <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2D</font></font>
but it is the MD tag that tells us what has been deleted: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">^AG</font></font></p>
<h2 class="western"><a name="what-will-insertions-look-like-in-a-sam-format"></a>
90.4 What will insertions look like in a SAM format?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -count 2 -block 2 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAG--AAGAATTTTTAGGATCTTTTGTG--TGCGAATA
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||||||||
 ||||||||||||||||||||||| 
||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGATAAGAATTTTTAGGATCTTTTGTGCTTGCGAATA</font></font></p>
<p class="first-paragraph">the corresponding SAM CIGAR and MD fields
are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">19M2I23M2I8M</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">50</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19M2I23M2I8M
   CGGACACACAAAAAGAAAGATAAGAATTTTTAGGATCTTTTGTGCTTGCGAATA  MD:Z:50</font></font></p>
<h2 class="western"><a name="how-long-is-an-alignment"></a>90.5 How
long is an alignment?</h2>
<p class="first-paragraph">This question can have different
interpretations. What it typically means is the distance between the
start and end coordinates on the reference.</p>
<p>For either case, we can compute the alignment lengths from the
CIGAR string alone following these rules:</p>
<p>Dealing with matches and mismatches cigar strings such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">50M</font></font>
or in extended format <font face="Consolas, serif"><font size="2" style="font-size: 11pt">24=1X25=</font></font></p>
<p>These will produce an alignment that is the sum of the integers in
the formula: 50. We now see some of the logic behind labeling both
matches and mismatches with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">M</font></font>.
They produce the same alignment lengths. Though this labeling is
hopefully on the way out - as being a grossly misleading and frankly
dumb way of describing alignments. When using the extended cigar
string the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=</font></font>
will be used for matches and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">X</font></font>
sign for mismatches and that too would also produce a 24 + 1 + 25 =
50bp long alignments.</p>
<p>Dealing with deletions say we have <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3S28M2D15M</font></font>
the MD tag is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">28^TC15</font></font>.
Soft clipping means that this region is not aligned. Then then the
alignment length is 28 + 2 + 15 = 45 so the alignment will start at
column POS and will go to POS + 45.</p>
<p>Dealing with insertions for example: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">14M2I19M2I17M</font></font>
the MD tag is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">50</font></font>.
The insertions are relative to the query and do not alter the
alignment length corresponding to the reference thus it will be 14 +
19 + 17 = 50.</p>
<h2 class="western"><a name="how-to-get-the-alignment-lengths-with-a-tool"></a>
90.6 How to get the alignment lengths with a tool?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>
suite can do this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
bamtobed -i align.bam </font></font>
</p>
<p class="first-paragraph">though be warned that by default this
suite uses the BED format - a zero-based data representation hence
coordinates will go from zero.</p>
<h2 class="western"><a name="how-to-visualize-a-sam-file-as-pairwise-alignment"></a>
90.7 How to visualize a SAM file as pairwise alignment?</h2>
<p class="first-paragraph">Matt LaFave&rsquo;s tool <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sam2pairwise</font></font>
does this:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/mlafave/sam2pairwise">https://github.com/mlafave/sam2pairwise</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostars.org/p/110498/">https://www.biostars.org/p/110498/</a></font></p>
</ul>
<p class="first-paragraph">usage:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view align.bam | sam2pairwise</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   0   AF086833    6   60  3S28M2D15M  *   0  
0</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACAAAAAGAAAGAAGAATTTTTAGGA--TTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  |||||||||||||||||||||||||||| 
|||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NNNACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="an-introduction-to-genomic-variation"></a>
91 An introduction to genomic variation</h1>
<h2 class="western"><a name="what-is-a-variant"></a>91.1 What is a
variant?</h2>
<p class="first-paragraph">When it comes to defining genomic
variations, the nomenclature and definitions show a surprising level
imprecision. We all know what variation is <i>supposed</i> to mean of
course. It should be a difference from what we expect to see. The
problem is we rarely know reality is and our expectations may be
different in the future. At any given moment we only have imperfect
information that changes in time.</p>
<p>Moreover as our understanding of genomes evolves concepts such as
&ldquo;reference genome&rdquo; become harder to apply correctly. What
should be considered a &ldquo;reference genome&rdquo;? After all what
we call a reference genome changes too - albeit usually in a more
systematic way. Come to think of it just how different are different
reference genomes from one another? Can we visualize that? As it
turns out it is surprisingly challenging. Then will a variant
detected for the previous version of a genome still be a variant in
the updated genome?</p>
<h2 class="western"><a name="how-do-we-classify-variants"></a>91.2
How do we classify variants?</h2>
<p class="first-paragraph">Genomic variations are typically
categorized into different classes and are often denoted with a
shortened acronym:</p>
<ul>
	<li/>
<p class="compact">SNP, a single nucleotide polymorphism - A
	change of a single base.</p>
	<li/>
<p class="compact">INDEL, an insertion or a deletion - A
	single base added or removed.</p>
	<li/>
<p class="compact">SNV, a single nucleotide variant. A SNPs or
	an INDEL. The change is still single basepair long.</p>
	<li/>
<p class="compact">MNP, a multi-nucleotide polymorphism -
	Several consecutive SNPs in a block.</p>
	<li/>
<p class="compact">MNV, a multi-nucleotide variant - Multiples
	SNPs and INDELs as a block.</p>
	<li/>
<p class="compact">Short variations. Variations (MNVs) that
	are typically less than <b>50bp</b> in length.</p>
	<li/>
<p class="compact">Large-scale variations. Variations that are
	larger <b>50bp</b>.</p>
	<li/>
<p class="compact">Structural variants. Variations on the
	kilobase scale that involve (one or more) chromosomes.</p>
</ul>
<p class="first-paragraph">Radically different techniques need to be
used to study problems in each class. For example for short
variations, individual reads can capture them in entirety.
Large-scale variants are typically detected from the relative
positioning of reads about their read pairs. For structural
variations entirely new techniques (e.g.&nbsp;nano-channel arrays,
Bionanogenomics ) may be needed.</p>
<p>As a rule, SNPs are the most reliably detectable from data. In
contrast very short MNVs that include insertions and deletions are
the most difficult to resolve correctly. As you shall see in one of
the next chapters, for example, a deletion followed by another
variation can often be better aligned in alternate positions and that
that can &ldquo;fool&rdquo; variant callers.</p>
<h2 class="western"><a name="what-are-snps"></a>91.3 What are SNPs?</h2>
<p class="first-paragraph">SNPs (Single Nucleotide Polymorphisms) are
a particular case of small scale variations. Originally were defined
as:</p>
<p class="block-text-western">A single nucleotide polymorphism, also
known as simple nucleotide polymorphism, (SNP, pronounced snip;
plural snips) is a DNA sequence variation occurring commonly within a
population (e.g.&nbsp;1%) in which a single nucleotide - A, T, C or G
of a shared sequence differs between members of a biological species
(or paired chromosome of the same individual).</p>
<p class="first-paragraph">Note how a SNP is defined as only a
polymorphism (a change of a base) and also implies a frequency of
observing it in a population.</p>
<h2 class="western"><a name="is-the-snp-term-used-consistently"></a>91.4
Is the SNP term used consistently?</h2>
<p class="first-paragraph">Not at all. Previously I have mentioned
how &ldquo;gene&rdquo; was the most misused word in biology. Well
&ldquo;SNP&rdquo; is a close second. In our observations, most
bioinformaticians and most reference and training materials will
often use the word SNP to refer to SNVs or even MNVs, sometimes to
<b>all</b> short variants. For a start, the word SNPs (snip) is easy
to pronounce. It&rsquo; s fun, it just rolls off your tongue. Then
the 1% limitation is, of course, arbitrary. The first cutoff for SNPs
was at 5% a limit later reduced to 1%.</p>
<p>Another problem with the definition of SNPs is that scientists are
unable to randomly sample populations and with that, they introduce
the so-called selection bias. It could easily happen that the
frequency of a variation that initially appears to be over 1% then
later turns out to be less than that value.</p>
<p>We believe that even the scientific language is dynamic and we
should accept the reality that most scientists use the word SNP to
refer to polymorphisms where the frequency in a population is not
known. In this book, we will do the same. Also, we need to remember
that what others call SNPs may often include not just polymorphisms
but also single (or very short) nucleotide insertions and deletions.</p>
<h2 class="western"><a name="do-snps-cause-disease"></a>91.5 Do SNPs
cause disease?</h2>
<p class="first-paragraph">SNPs are in a sense the holy grail of
genomics. Their &ldquo;popularity&rdquo; perhaps is a manifestation
of our hopes and wishful thinking. After all wouldn&rsquo;t it be
great if every disease or human condition could be distilled down to
a few changes and mutations in the genome? We would need to find that
one SNP, and with that, we solve the problem, collect our reward and
move onto the next challenge.</p>
<p>And sometimes (rarely though) things do work out this way. Some
individual mutations do indeed cause disease(s). The OMIM database
(Online Mendelian Inheritance in Man) is an online catalog of human
genes and genetic disorders collect this data.</p>
<p>Visit the chapter called <font color="#4f81bd"><a href="#ch091.xhtml#omim">Online
Mendelian Inheritance in Man</a></font> for details on what is
currently known about inheritable diseases and how we can interpret
the data that describes the inheritance.</p>
<h2 class="western"><a name="what-is-a-genotype"></a>91.6 What is a
genotype?</h2>
<p class="first-paragraph">A genotype is the genetic constitution of
an individual organism.</p>
<p>In the context of genomic variations, the genotyping typically
refers to identifying one, more than one, or all mutations of the
sequence under investigation.</p>
<p>Specifically, the word &ldquo;genotyping&rdquo; seems to be used
very liberally in publications and training materials. Most of the
time, when reading these it is not clear what the authors mean when
they state that they have &ldquo;genotype&rdquo; data. What they
probably and usually mean is that they have established with some
confidence that a variant is present.</p>
<h2 class="western"><a name="what-is-a-haplotype"></a>91.7 What is a
haplotype?</h2>
<p class="first-paragraph">A haplotype is a group of genes in an
organism that is inherited together from a single parent.</p>
<p>In the context of genomic variations, a haplotype is a group of
mutations that are inherited together. But just as with the
&ldquo;genotyping,&rdquo; the word &ldquo;haplotyping&rdquo; may be
used in a broader and narrower sense.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="online-mendelian-inheritance-of-man"></a>
92 Online Mendelian Inheritance of Man</h1>
<p class="first-paragraph">The Online Mendelian Inheritance in Man
(OMIM) is a continuously updated catalog of human genes and genetic
disorders and traits, with the particular focus on the molecular
relationship between genetic variation and phenotypic expression. It
contains information about <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Mendelian_inheritance">Mendelian
disorders</a></font>.</p>
<h2 class="western"><a name="how-does-omim-work"></a>92.1 How does
OMIM work?</h2>
<p class="first-paragraph">Whereas the <font color="#4f81bd"><a href="http://www.omim.org/">http://www.omim.org/</a></font>
website offers useful search and visualization functions, a different
type of data mining can be performed from the command line. Do note
that OMIM website is primarily intended to assist physicians.</p>
<p>Here I want to make a point. Providing a service to clinicians is
a high-risk task as the responsibilities associated are much higher
than that of other services. OMIM tries to manage this risk by
producing humongous amounts of information that may even distantly
related to a search or term of interest. In essence, it organizes
information to some extent, but it also tries to transfer all
responsibility to the end user. Basically, it makes it their job to
figure out what all that means. Just about any searchable term will
produce pages and pages of results. I believe that this is by design
so - but remember the actual useful information is probably a tiny
fraction (if any) of all those results. When we investigate the data
that underlies OMIM do we fully realize how far from completion this
information is.</p>
<h2 class="western"><a name="can-the-omim-data-be-obtained"></a>92.2
Can the OMIM data be obtained?</h2>
<p class="first-paragraph">Whereas the main web site is the entry
point most people - the command line allows us to see and investigate
the data that connects it all together. Hence it may give different
insights and deeper understanding and appreciation. The data itself
has a relatively flat and straightforward structure and is amenable
to be processed via the Unix command line.</p>
<p>Do note however that only the gene-to-phenotype data is available
for download. The content of the pages that provide clinical advice
is not included.</p>
<p>Registering at <font color="#4f81bd"><a href="http://www.omim.org/">http://www.omim.org/</a></font>
offers download access to their data in files named such as</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mim2gene.txt</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mimTitles.txt</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">genemap.txt</font></font>,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">morbitmap.txt</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">genemap2.txt</font></font></p>
</ul>
<p class="first-paragraph">files that we can investigate from the
Unix command line as well. Getting the OMIM files is somewhat
ludicrous it requires a process of registering, feels a bit
discriminative where you can&rsquo;t have a gmail account etc.</p>
<h2 class="western"><a name="what-format-is-the-omim-data-in"></a>92.3
What format is the OMIM data in?</h2>
<p class="first-paragraph">The OMIM file formats (as typical in
bioinformatics) are somewhat obscure and counter-intuitive, going
back many decades well before the internet era. Identifiers carry
markers used to inform readers of the meaning of each term:</p>
<ul>
	<li/>
<p class="compact">Asterisk (*) Gene</p>
	<li/>
<p class="compact">Plus (+) Gene and phenotype, combined</p>
	<li/>
<p class="compact">Number Sign (#) Phenotype, molecular basis
	known</p>
	<li/>
<p class="compact">Percent (%) Phenotype or locus, molecular
	basis unknown</p>
	<li/>
<p class="compact">NULL () Other, main phenotypes with a
	suspected Mendelian basis</p>
	<li/>
<p class="compact">Caret (^) Entry has been removed from the
	database or moved to another entry</p>
</ul>
<p class="first-paragraph">As seen above, for example, the &ldquo;percent&rdquo;
symbol was used initially as a marker to indicate a certain mendelian
phenotype or phenotypic locus for which the underlying molecular
basis is not known. Today, in the data files the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent</font></font>
will be written out explicitly. It looks (and is) a bit ridiculous:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100070  AORTIC ANEURYSM, FAMILIAL ABDOMINAL</font></font></p>
<p class="first-paragraph">So here we are in the 21st century,
exploring the magnificent science of biology, but to do so we have to
write: &ldquo;Asterix&rdquo; when we mean &ldquo;Gene&rdquo; &hellip;
oh well.</p>
<h2 class="western"><a name="what-terms-and-concepts-does-omim-know-about"></a>
92.4 What terms and concepts does OMIM know about?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mimTitles.txt</font></font>
file list the terms in the OMIM database. Think of it as table of
contents. It connects a so called MIM number to a concept.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Copyright (c) 1966-2016 Johns Hopkins University. Use of this file
adheres to the terms specified at https://omim.org/help/agreement.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generated: 2016-11-05</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Prefix        Mim Number      Preferred Title; symbol Alternative
Title(s); symbol(s) Included Title(s); symbols</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NULL
   100050  AARSKOG SYNDROME, AUTOSOMAL DOMINANT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100070  AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1; AAA1    ANEURYSM,
ABDOMINAL AORTIC; AAA;; ABDOMINAL AORTIC ANEURYSM</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Number
Sign     100100  PRUNE BELLY SYNDROME; PBS       ABDOMINAL MUSCLES,
ABSENCE OF, WITH URINARY TRACT ABNORMALITY AND CRYPTORCHIDISM;;
EAGLE-BARRETT SYNDROME; EGBRS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NULL
   100200  ABDUCENS PALSY</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Number
Sign     100300  ADAMS-OLIVER SYNDROME 1; AOS1   AOS;; ABSENCE DEFECT
OF LIMBS, SCALP, AND SKULL;; CONGENITAL SCALP DEFECTS WITH DISTAL
LIMB REDUCTION ANOMALIES;; APLASIA CUTIS CONGENITA WITH TERMINAL
TRANSVERSE LIMB DEFECTS    APLASIA CUTIS CONGENITA, CONGENITAL HEART
DEFECT, AND FRONTONASAL CYSTS, INCLUDED</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Caret
  100500  MOVED TO 200150</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100600  ACANTHOSIS NIGRICANS</font></font></p>
<p class="first-paragraph">The number of total terms in OMIM:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
mimTitles.txt | grep -v '#' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
24970</font></font></p>
<h2 class="western"><a name="how-many-phenotypes-of-mendelian-inheritance"></a>
92.5 How many phenotypes of Mendelian inheritance?</h2>
<p class="first-paragraph">Phenotypes where the molecular origin is
known:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
mimTitles.txt | grep 'Number' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
4869</font></font></p>
<p class="first-paragraph">Phenotypes where the molecular origin is
unknown:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
mimTitles.txt | grep 'Percent' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
1615</font></font></p>
<p class="first-paragraph">What are the phenotypes of unknown
molecular origin:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
mimTitles.txt | grep 'Percent' | head</font></font></p>
<p class="first-paragraph">Producing the file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100070  AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1; AAA1    ANEURYSM,
ABDOMINAL AORTIC; AAA;; ABDOMINAL AORTIC ANEURYSM</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100600  ACANTHOSIS NIGRICANS</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
100820  ACHOO SYNDROME  AUTOSOMAL DOMINANT COMPELLING HELIOOPHTHALMIC
OUTBURST SYNDROME;; PHOTIC SNEEZE REFLEX;; SNEEZING FROM LIGHT
EXPOSURE;; PEROUTKA SNEEZE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
101850  PALMOPLANTAR KERATODERMA, PUNCTATE TYPE III; PPKP3     
ACROKERATOELASTOIDOSIS; AKE;; COLLAGENOUS PLAQUES OF HANDS AND
FEET</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102100  ACROMEGALOID CHANGES, CUTIS VERTICIS GYRATA, AND CORNEAL
LEUKOMA        ROSENTHAL-KLOEPFER SYNDROME</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102150  ACROMEGALOID FACIAL APPEARANCE SYNDROME AFA SYNDROME;; THICK
LIPS AND ORAL MUCOSA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102300  RESTLESS LEGS SYNDROME, SUSCEPTIBILITY TO, 1; RLS1     
ACROMELALGIA, HEREDITARY;; EKBOM SYNDROME</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102350  ACROMIAL DIMPLES        SUPRASPINOUS FOSSAE,
CONGENITAL</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102510  ACROPECTOROVERTEBRAL DYSPLASIA; ACRPV   F SYNDROME</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Percent
102800  ADENOSINE TRIPHOSPHATASE DEFICIENCY, ANEMIA DUE TO</font></font></p>
<p class="first-paragraph">The file can be searched (case
insensitive) for diseases:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
mimTitles.txt | egrep -i sickle</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NULL
   143020  HPA I RECOGNITION POLYMORPHISM, BETA-GLOBIN-RELATED; HPA1 
     RESTRICTION FRAGMENT LENGTH POLYMORPHISM, SICKLE CELL
ANEMIA-RELATED</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Number
Sign     603903  SICKLE CELL ANEMIA</font></font></p>
<p class="first-paragraph">The number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">603903</font></font>
is the so-called MIM number, a unique identifier within the
framework. To find the gene connected to sickle cell Anemia</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | grep 603903</font></font></p>
<p class="first-paragraph">produces a fairly lengthy (but still
single line entry with):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr11
  5225465 5227070 11p15.4         141900  HBB     Hemoglobin beta HBB
    3043    ENSG00000244734 pseudogene</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HBBP1
between HBG and HBD loci       Delta-beta thalassemia, 141749 (3),
Autosomal dominant; Erythremias, beta- (3);</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Heinz
body anemias, beta-, 140700 (3), Autosomal dominant; Hereditary
persistence of fetal hemoglobin, 141749 (3),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Autosomal
dominant; {Malaria, resistance to}, 611162 (3); Methemoglobinemias,
beta- (3); Sickle cell anemia, 603903</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(3),
Autosomal recessive; Thalassemia-beta, dominant inclusion-body,
603902 (3); Thalassemias, beta-, 613985 (3)    </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  Hbb-bt,Hbb-bs (MGI:5474850,MGI:5474852)</font></font></p>
<p class="first-paragraph">To see just a first few columns:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | grep 603903 | cut -f 1-8</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr11
  5225465 5227070 11p15.4         141900  HBB     Hemoglobin beta</font></font></p>
<h2 class="western"><a name="how-many-phenotypes-can-be-connected-to-genes"></a>
92.6 How many phenotypes can be connected to genes?</h2>
<p class="first-paragraph">The 13th column of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genemap2.txt</font></font>
files connects a gene to a phenotype. But not all those columns have
values:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | cut -f 13 |  wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
16086</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | cut -f 13 | egrep '.+' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
4996</font></font></p>
<p class="first-paragraph">Out of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">16086</font></font>
genes in this file just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4996</font></font>
have a non-zero field for phenotypes.</p>
<h2 class="western"><a name="how-many-unique-pairs-of-gene-and-phenotypes-are-there"></a>
92.7 How many unique pairs of gene and phenotypes are there?</h2>
<p class="first-paragraph">As it turns out this is unexpectedly
difficult to extract. Columns 6 and 13 contain what we need, but the
latter is a free text format from which each 6 digit MIM number needs
to be parsed out.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | grep 603903 | cut -f 13</font></font></p>
<p class="first-paragraph">The output of which is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Delta-beta
thalassemia, 141749 (3), Autosomal dominant; Erythremias, beta- (3);
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Heinz
body anemias, beta-, 140700 (3), Autosomal dominant; Hereditary
persistence of </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fetal
hemoglobin, 141749 (3), Autosomal dominant; {Malaria, resistance to},
611162 (3);</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Methemoglobinemias,
beta- (3); Sickle cell anemia, 603903 (3), Autosomal
recessive;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Thalassemia-beta,
dominant inclusion-body, 603902 (3); Thalassemias, beta-, 613985 (3)</font></font></p>
<p class="first-paragraph">What we need from this are the numbers
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">141749</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">140700</font></font>
and preferably a file that lists them like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">603903
  141749</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">603903
  140700</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">603903
  141749</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The above roadblock is one of those
typical moments when a bioinformatician needs to write their simple
code to get them past the obstacle. There are many ways to solve this
problem, for example with an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font>
program that would look like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Split input by tabs. </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Format output by tabs.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BEGIN
{ FS = OFS = &quot;\t&quot; }</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">{</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # Gene MIM number is column 6.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   genemim = $6</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   geneid = $8</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # The phenotypes are in column 13.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   split($13, elems, &quot; &quot;)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # Go over each element and print the matches to numbers.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   for (i in elems){</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       word = elems[i]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       # MIM numbers have exactly 6 digits.            </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       if (word ~ /^[0-9]{6}/ ){</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           print genemim, geneid, word</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       } </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   }</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">}</font></font></p>
<p class="first-paragraph">Running this code on the entire data and
tabulating at the end. How many disease/phenotype entries are there:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | awk -f prog.awk | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
6964</font></font></p>
<p class="first-paragraph">So the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4996</font></font>
annotated genes produce a total of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6964</font></font>
phenotypes.</p>
<h2 class="western"><a name="what-genes-have-the-most-annotated-phenotypes"></a>
92.8 What genes have the most annotated phenotypes?</h2>
<p class="first-paragraph">These are the genes where mutations appear
to cause the most phenotypes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | awk -f prog.awk | cut -f 1,2 | sort | uniq -c | sort
-rn | head</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15
120140  Collagen II, alpha-1 polypeptide</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
14 134934  Fibroblast growth factor receptor-3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
12 176943  Fibroblast growth factor receptor-2 (bacteria-expressed
kinase)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
12 150330  Lamin A/C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
11 601728  Phosphatase and tensin homolog (mutated in multiple
advanced cancers 1)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
11 191170  Tumor protein p53</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
11 171834  Phosphatidylinositol 3-kinase, catalytic, alpha
polypeptide</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
11 109270  Solute carrier family 4, anion exchanger, member 1
(erythrocyte membrane protein band 3, Diego blood group)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
10 605427  Transient receptor potential cation channel, subfamily V,
member 4 (vanilloid receptor-related osmotically activated channel)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
10 300017  Filamin A, alpha (actin-binding protein-280)</font></font></p>
<p class="first-paragraph">So there are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15</font></font>
entries in this file for gene with MIM number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">120140</font></font>
that corresponds to gene <i>Collagen II, alpha-1 polypeptide</i>.
What are these phenotypes?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genemap2.txt | awk -F '\t' ' $6 ~ /120140/ { print $13 } '</font></font></p>
<p class="first-paragraph">That produces</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Achondrogenesis,
type II or hypochondrogenesis, 200610 (3), Autosomal dominant;
Avascular necrosis of the femoral head,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">608805
(3), Autosomal dominant; Czech dysplasia, 609162 (3), Autosomal
dominant; Epiphyseal dysplasia, multiple, with</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">myopia
and deafness, 132450 (3), Autosomal dominant; Kniest dysplasia,
156550 (3), Autosomal dominant;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Legg-Calve-Perthes
disease, 150600 (3), Autosomal dominant; Osteoarthritis with mild
chondrodysplasia, 604864 (3),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Autosomal
dominant; Otospondylomegaepiphyseal dysplasia, 215150 (3), Autosomal
recessive; Platyspondylic skeletal</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">dysplasia,
Torrance type, 151210 (3), Autosomal dominant; SED congenita, 183900
(3), Autosomal dominant; SMED Strudwick</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">type,
184250 (3), Autosomal dominant; Spondyloepiphyseal dysplasia,
Stanescu type, 616583 (3), Autosomal dominant;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Spondyloperipheral
dysplasia, 271700 (3), Autosomal dominant; Stickler sydrome, type I,
nonsyndromic ocular, 609508 (3),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Autosomal
dominant; Stickler syndrome, type I, 108300 (3), Autosomal dominant;
Vitreoretinopathy with phalangeal</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">epiphyseal
dysplasia (3)</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="why-simulating-reads-is-a-good-idea"></a>
93 Why simulating reads is a good idea</h1>
<p class="first-paragraph">Simulating data allows us to evaluate the
performance of our tools and techniques. By knowing what the data
contains, we can verify that the tools do indeed operate as we expect
them to.</p>
<p>When coupled with recipes that automate an entire workflow data
simulations allows you to understand the whole process.</p>
<p>We recommend that you start exploring all bioinformatics methods
with simulated data to see the strengths and weaknesses of each.
Unfortunately, while data simulators are essential for learning data
analysis, this domain of bioinformatics is neglected with quite a few
abandoned software. We tried to collect a few practical examples
though each has shortcomings.</p>
<h2 class="western"><a name="what-types-of-data-simulators-exist"></a>
93.1 What types of data simulators exist?</h2>
<p class="first-paragraph">There are simulators available for the
whole genome, exome, RNA-seq and chip-seq data simulation. Some tools
run from command line; others have graphical user interfaces.</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim/dwgsim</font></font>
	simulates obtaining sequencing reads from whole genomes.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
	EMBOSS tool can simulate mutations in a single sequence.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">biosed</font></font>
	EMBOSS tool provides control on what mutations we introduce.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ReadSim</font></font>
	is a sequencing read simulator to target long reads such as PacBio
	or Nanopore.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Art</font></font>
	is the most sophisticated and up to date simulator that attempts to
	mimic the errors that sequencing instruments produce.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Metasim</font></font>
	simulates reads from metagenomes (from a diverse taxonomical
	composition present in different abundances).</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Polyester</font></font>
	an <font color="#4f81bd"><a href="http://bioconductor.org/packages/release/bioc/vignettes/polyester/inst/doc/polyester.html">RNA-Seq</a></font>
	measurement simulator</p>
</ul>
<h2 class="western"><a name="are-the-simulations-realistic"></a>93.2
Are the simulations realistic?</h2>
<p class="first-paragraph">The level of &ldquo;realism&rdquo; depends
on what the simulator aims to capture. Some properties of the
sequencing are easier to model, but there will always be features
that are unique to real data that no simulator can fully capture.</p>
<p>In general, data simulators are best suited for exploring how
various events may manifest themselves within sequencing data. For
example: suppose we have a &ldquo;large scale inversion&rdquo; what
will that data look like when aligned against the non-inverted
reference genome.</p>
<h2 class="western"><a name="what-is-wgsim-and-dwgsim"></a>93.3 What
is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
are tools for simulating sequence reads from a reference genome. The
tools can simulate diploid genomes with SNPs and insertion/deletion
(INDEL) polymorphisms, and affect reads with uniform sequencing
errors.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
is easier to install but it only produces the mutations in text
format that is harder to visualize. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
on the other hand will produce VCF files in addition to the text
file.</p>
<h2 class="western"><a name="how-do-i-simulate-sequencing-data-with-wgsim"></a>
93.4 How do I simulate sequencing data with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; genome.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Simulate a read pair</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim
-N 1000 genome.fa sim1.fq sim2.fq &gt; mutations.txt</font></font></p>
<p class="first-paragraph">The commands above will create the
following files:</p>
<ul>
	<li/>
<p class="compact">First in pair: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim1.fq</font></font></p>
	<li/>
<p class="compact">Second in pair: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim2.fq</font></font></p>
	<li/>
<p class="compact">Mutations as a text file: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">mutations.txt</font></font></p>
</ul>
<h2 class="western"><a name="how-do-i-simulate-sequencing-data-with-dwgsim"></a>
93.5 How do I simulate sequencing data with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>?</h2>
<p class="first-paragraph">You may install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
as instructed on the page <font color="#4f81bd"><a href="#ch132.xhtml#wgsim">Installing
wgsim and dwgsim</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; genome.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Simulate sequencing reads from the reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim
-N 1000 genome.fa sim</font></font></p>
<p class="first-paragraph">This will create a number of files:</p>
<ul>
	<li/>
<p class="compact">First in pair: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim.bwa.read1.fastq</font></font></p>
	<li/>
<p class="compact">Second in pair: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim.bwa.read2.fastq</font></font></p>
	<li/>
<p class="compact">Mutations as text file: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim.mutations.txt</font></font></p>
	<li/>
<p class="compact">Mutations as a VCF file: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sim.mutations.vcf</font></font></p>
</ul>
<p class="first-paragraph">You may align the paired-end reads against
a genome to explore aligner behavior.</p>
<h2 class="western"><a name="how-do-i-mutate-a-sequence-with-msbar"></a>
93.6 How do I mutate a sequence with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
EMBOSS tool was designed to generate mutations in sequences. It can
produce point mutations with the flag:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-point
Types of point mutations to perform</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      0 (None); </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      1 (Any of the following)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      2 (Insertions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      3 (Deletions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      4 (Changes)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      5 (Duplications)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      6 (Moves)</font></font></p>
<p class="first-paragraph">Or block mutations with the flags</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-block
Types of block mutations to perform</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      0 (None)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      1 (Any of the following)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      2 (Insertions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      3 (Deletions)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      4 (Changes)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      5 (Duplications)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      6 (Moves)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font>
</p>
<p class="first-paragraph">Block mutations are a series of
consecutive :mutations that span from ranges specified with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-minimum</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-maximum</font></font>.</p>
<p>Get a 50bp query from the reference file and introduce a point
mutation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; genome.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract the first 50 bases as a query.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.fa | seqret -filter -send 50  &gt; query.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Introduce a point deletion.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -point 3 &gt; mutated.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Visualize the output.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">global-align.sh
query.fa mutated.fa </font></font>
</p>
<p class="first-paragraph">on our system this produces the following
(the mutations are random so you will get something different):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">||||||||||||||||||||||||||||||
|||||||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTA-GATCTTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">Mutate the query to contain a 2 base block
mutation:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -block 3 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">The alignment now shows:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||||||||||||||||||||||||||||||||
 |||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGT--GAATA</font></font></p>
<p class="first-paragraph">We can introduce more than one mutation
with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-count</font></font>
parameter:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
query.fa | msbar -filter -count 2 -block 3 -minimum 2 -maximum 2 &gt;
mutated.fa</font></font></p>
<p class="first-paragraph">one result could be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||
 |||||||||||||||||| 
|||||||||||||||</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAA--AAAGAAGAATTTTTAGGA--TTTTGTGTGCGAATA</font></font></p>
<p class="first-paragraph">Since the tool selects mutations randomly,
when choosing substitutions it is possible to mutate the sequence
back to itself i.e.&nbsp;replacing an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
with another <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font></p>
<p>One (serious) flaw of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">msbar</font></font>
tool is that it does inform us what changes it has made. Hence, it is
a bit harder to use it in a context of validating a variant caller.</p>
<h2 class="western"><a name="how-do-i-simulate-mutations-with-biosed"></a>
93.7 How do I simulate mutations with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">biosed</font></font>?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 | head -3 &gt; genome.fa</font></font></p>
<p class="first-paragraph">The file contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;AF086833.2
Ebola virus - Mayinga, Zaire, 1976, complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATAACTATGAGGAAGATTAATAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTGTAATCACACCTGGTTTGTTT</font></font></p>
<p class="first-paragraph">Now replace the first ten bases with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">biosed</font></font>
as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.fa | biosed -filter -target CGGACACACA -replace GGGGGGGGGG</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;AF086833.2
Ebola virus - Mayinga, Zaire, 1976, complete
genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGGGGGGGGGAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGCGAATAACTATGAGGA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGATTAATAATTTTCCTCTCATTGAAATTTATATCGGAATTTAAATTGAAATTGTTACTG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TAATCACACCTGGTTTGTTT</font></font></p>
<p class="first-paragraph">It also will slightly reformat the fasta
file; the lines are wrapping at a shorter length.</p>
<h2 class="western"><a name="how-do-i-simulate-reads-with-readsim"></a>
93.8 How do I simulate reads with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">readsim</font></font>?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">readsim</font></font>
program has a large variety of options, it claims to be able to
produce realistic errors that correspond to instruments, it can
account for higher ploidy (to simulate variants from populations) and
so on.</p>
<p>Manual: <font color="#4f81bd"><a href="https://sourceforge.net/p/readsim/wiki/manual/">https://sourceforge.net/p/readsim/wiki/manual/</a></font></p>
<p>The program will only work with Python 2.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get a reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; genome.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Simulate pacbio technology reads into a file named
reads.fq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">readsim.py
sim fq --ref genome.fa --pre reads --tech pacbio</font></font></p>
<h2 class="western"><a name="how-do-i-simulate-reads-with-art"></a>93.9
How do I simulate reads with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">art</font></font>?</h2>
<p class="first-paragraph">Manual:
<font color="#4f81bd"><a href="http://www.niehs.nih.gov/research/resources/software/biostatistics/art/">http://www.niehs.nih.gov/research/resources/software/biostatistics/art/</a></font></p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Art</font></font>
suite is the most sophisticated simulator that we know of.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">art_illumina
-ss HS25 -sam -i genome.fa -p -l 150 -f 20 -m 200 -s 10 -o reads</font></font></p>
<p class="first-paragraph">The output contains even an &ldquo;optimal&rdquo;
SAM alignment file for the simulated reads that we can compare
against:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Quality
Profile(s)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   First Read:   HiSeq 2500 Length 150 R1 (built-in profile) </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   First Read:   HiSeq 2500 Length 150 R2 (built-in profile) </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Output
files</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 FASTQ Sequence Files:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    the 1st reads: reads1.fq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    the 2nd reads: reads2.fq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ALN Alignment Files:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    the 1st reads: reads1.aln</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    the 2nd reads: reads2.aln</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 SAM Alignment File:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   reads.sam</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="visualizing-large-scale-genomic-variations"></a>
94 Visualizing large scale genomic variations</h1>
<p class="first-paragraph">One of the most unexpectedly challenging
tasks of a bioinformatics visualization is correctly interpreting the
effects that large-scale genomic variations have on alignments.</p>
<h2 class="western"><a name="the-reference-genome-is-not-real"></a>94.1
The reference genome is not &ldquo;real&rdquo;</h2>
<p class="first-paragraph">The confusion starts right at the
terminology &ndash; we call one of our sequences the &ldquo;reference&rdquo;
and we compare our data to it. The word &ldquo;reference&rdquo;
implies that we are comparing against a &ldquo;ground truth&rdquo; &ndash;
but it is the other way around. It is our sequenced data that exists;
the reference is often not even a &ldquo;real&rdquo; sequence. For
example the human reference is built from a pooled sample from
several donors. It almost certainly the case that the resulting human
reference does not fully match neither of the donors; instead, it is
a mosaic of them all.</p>
<p>It is also very likely that the human reference genome is not even
a viable - meaning that if the sequence were &ldquo;implanted&rdquo;
into a human cell, it would not function. Now, knowing that, think
about the repercussions of this for a second. Scientists are using a
non-existing, potentially non-viable human reference to characterize
all other &ldquo;real&rdquo; human genomes. And when we compare two
real genomes, we do so by comparing them via this third, non-real one
through a process that is not &ldquo;transitive&rdquo; - knowing how
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
and how <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
vs <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
are different from one another is not sufficient to characterize how
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">B</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
are different from one another.</p>
<p><img src="0000_html_8109be487114ddd6.png" name="Image135" align="bottom" hspace="12" width="560" height="159" border="0"/>
</p>
<p>I found it useful to think of the reference genome as a &ldquo;mirror&rdquo;
in which we &ldquo;reflect&rdquo; our real data to make it visible to
us. It is an imperfect mirror, and what we see is often in some sense
the opposite of reality.</p>
<p>To run the software for this chapter prepare some datasets first:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id AF086833 &gt; refs/reference.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index refs/reference.fa</font></font></p>
<h2 class="western"><a name="which-alignments-are-informative"></a>94.2
Which alignments are informative?</h2>
<p class="first-paragraph">When it comes to structural variations the
main rule to remember:</p>
<p class="block-text-western">Alignments that <b>cross over</b> the
variation carry the most information on the nature of the change.</p>
<p class="first-paragraph">Alignments that are entirely contained
within a continuous segment do not know about what is &ldquo;outside&rdquo;
of that segment, and hence are less informative. The very fact that
paired-end sequencing does not fully cover the fragment and has
&ldquo;holes&rdquo; in the middle becomes, in this case, is quite
useful. It does two things for us:</p>
<ol>
	<li/>
<p class="compact">It allows the read to cover a longer
	distance.</p>
	<li/>
<p class="compact">The span and orientation of the pairs will
	indicate the type and direction of variation.</p>
</ol>
<p class="first-paragraph">It is also worth noting that typically the
vast majority of alignments will fall outside of these so-called
junctions &ndash; hence, only a small subset of the reads will carry
useful information.</p>
<h2 class="western"><a name="what-would-perfect-data-look-like"></a>94.3
What would perfect data look like?</h2>
<p class="first-paragraph">Ask any bioinformatician what ideal,
&ldquo;perfect&rdquo; data should look like - chances are you&rsquo;ll
get very different answers. We practitioners never see ideal data &ndash;
we don&rsquo;t quite know how even to define it. For this book, we
wrote a simple Python script called the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">perfect_coverage.py</font></font>
that simulates sequencing every 500bp long fragment of a genome
exactly once, with no errors, with an Illumina paired-end protocol.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the script</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O http://data.biostarhandbook.com/align/perfect_coverage.py</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate perfect data.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | python perfect_coverage.py</font></font></p>
<p class="first-paragraph">The code above produces the files <font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1.fq</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2.fq</font></font>.
Align these reads with.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem refs/reference.fa R1.fq R2.fq | samtools sort &gt;
perfect.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index perfect.bam</font></font></p>
<p class="first-paragraph">Visualize the output in IGV. The image
below is what we saw:</p>
<p><img src="0000_html_ae5084cddc509926.png" name="Image136" align="bottom" hspace="12" width="560" height="176" border="0"/>
</p>
<p>Note the coverages at the beginning and end of the genome. These
drops are the so-called <i>edge effects</i> and will show up even in
ideal data. In essence, even under ideal conditions there will be
less coverage at the ends of a chromosome because fewer reads can
overlap. The range where the effect will show up is as long as the
fragment size &ndash; 500bp in this case.</p>
<p>Since a chromosome has only two ends and the fragments are
typically much smaller than the sequence itself these artifacts
generally are ignored. But that for an analysis that measures lots of
short sequences (like an RNA-Seq experiment) there may be tens or
hundreds of thousands of transcripts, and each transcript will have
its ends affected. Moreover, transcripts have varying lengths whereas
the edge effect is fragment size dependent. Thus, shorter transcripts
will be more affected and will get less coverage than expected.</p>
<p>Dealing with edge effects in transcripts is more complicated,
often solved by introducing a new concept called &ldquo;effective
length normalization&rdquo; &ndash; treating all transcripts as if
they were shorter than they are in reality. We&rsquo;ll discuss this
further in the RNA-seq section. It is important to note that the
effect is present even in this idealized case.</p>
<h2 class="western"><a name="what-would-realistic-and-useful-data-look-like"></a>
94.4 What would realistic and useful data look like?</h2>
<p class="first-paragraph">The data simulation methods are explained
in more detail in <font color="#4f81bd"><a href="#ch093.xhtml#simulate">Why
simulating reads is a good idea</a></font>. For the rest of the
chapter we will repeat the following steps:</p>
<ol>
	<li/>
<p class="compact">We will make a copy of our reference file.
	We will call this the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.fa</font></font>
	and we will use it as the source of our simulations.</p>
	<li/>
<p class="compact">We will modify our genome file, generate
	reads from it, and compare it to the reference.</p>
</ol>
<p class="first-paragraph">Initially, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">reference.fa</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.fa</font></font>
files are the same. We will mimic evolution by modifying <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.fa</font></font>,
then we will align data obtained from it against <font face="Consolas, serif"><font size="2" style="font-size: 11pt">reference.fa</font></font>.
The process of simulating and aligning can be done step-by-step or
via a recipe. When using a recipe, comment out the steps that obtain
or index the reference as those don&rsquo;t change.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cp
refs/reference.fa genome.fa </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim
-N 10000 genome.fa r1.fq r2.fq </font></font>
</p>
<p class="first-paragraph">Now run the alignment and bam file
creation steps. The result will look like:</p>
<p><img src="0000_html_c6318c3dc465b973.png" name="Image137" align="bottom" hspace="12" width="560" height="194" border="0"/>
</p>
<p>The is more realistic data than our perfect alignment. We can
observe the stochasticity of the coverage; it varies across the
genome.</p>
<h2 class="western"><a name="what-would-a-deletion-from-the-genome-look-like"></a>
94.5 What would a deletion from the genome look like?</h2>
<p class="first-paragraph">Open the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.fa</font></font>
file in a text editor and delete a big chunk of it. Turning it from
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABCD</font></font>
to say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACD</font></font>.
You could also perform that from command line like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1 -send 2000 &gt;
part1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 3000 &gt; part2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
part1 part2 | union -filter  &gt; genome.fa</font></font></p>
<p class="first-paragraph">Then rerun the alignment. Here is what we
get:</p>
<p><img src="0000_html_dbfaba322d13396b.png" name="Image138" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<p>This image makes a lot of sense. The parts that we deleted are
missing. But note something else: there are read pairs that start on
one end and continue on the others. From the fragment (template size)
we can figure out how long the deletion is. Note how the fragment is
around 1500bp? Since our fragment size should be 500 on average, we
can tell that the deletion should be around 1000bp this way it adds
up to 1500.</p>
<p>From the above example, we note that the deletion made the
fragments look larger. It is because we are &ldquo;stretching&rdquo;
a shorter real genome over a longer &ldquo;reference&rdquo; genome.</p>
<h2 class="western"><a name="what-would-copy-number-variation-look-like"></a>
94.6 What would copy number variation look like?</h2>
<p class="first-paragraph">For this, add half of the beginning of the
genome back to the front of it. Basically imagine the genome is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABC</font></font>
then make it <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AABC</font></font>.
Again you can do this in an editor or from command line with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1 -send 2000 &gt;
part1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1 -send 2000 &gt;
part2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
part1 part2 refs/reference.fa | union -filter  &gt; genome.fa</font></font></p>
<p class="first-paragraph">These events are called duplications or
more generically copy number variation.</p>
<p><img src="0000_html_5eaced0f888c8426.png" name="Image139" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<h2 class="western"><a name="how-can-you-tell-when-parts-of-the-genome-are-swapped"></a>
94.7 How can you tell when parts of the genome are swapped?</h2>
<p class="first-paragraph">Now we modify our genome to go from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABCD</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CDAB</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -send 5000 &gt; part1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 5000 &gt; part2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
part2 part1 | union -filter  &gt; genome.fa</font></font></p>
<p class="first-paragraph">How will this data align? The image that
we see has more subtle changes:</p>
<p><img src="0000_html_9d498c2e1d5188a0.png" name="Image140" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<p>It is only when we fully expand the pairs that we start to
understand the complexity of it all.</p>
<p><img src="0000_html_8dd505b4a7344667.png" name="Image141" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<h2 class="western"><a name="what-if-the-genome-contains-new-regions"></a>
94.8 What if the genome contains new regions?</h2>
<p class="first-paragraph">What if would turn our genome from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABCD</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ABFCD</font></font>.
How will this show up in our data? Let&rsquo;s insert a sequence of a
different virus, in this case, HIV-1 into our genome</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -send 1000 &gt; part1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_001802.1 -format fasta -seq_stop 1000 &gt;
middle</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1000 &gt; part2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
part1 middle part2 | union -filter  &gt; genome.fa</font></font></p>
<p class="first-paragraph">The read pairs that come from the novel
genome will not align anywhere in the reference. These reads will
stay unmapped. For read pairs where one of the pairs comes from the
reference and the other from the novel genome, the matching one will
be aligned and the mate lost. These are called &ldquo;orphan&rdquo;
alignments. From the &ldquo;orphans&rdquo; we can see that a new
region must be present.</p>
<p>Also, reads falling over the junctions may be soft-clipped all at
the same coordinate. When we look at the alignment in paired-end mode
a large number of orphan reads show up:</p>
<p><img src="0000_html_4c090bfa5ce35155.png" name="Image142" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<h2 class="western"><a name="what-if-we-reverse-complement-parts-of-the-genome"></a>
94.9 What if we reverse complement parts of the genome?</h2>
<p class="first-paragraph">In this example, we reverse complement
some section of genome, but keep the rest the same:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1 -send 1000 &gt;
part1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 1000 -send 2000
-sreverse1 &gt; part2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/reference.fa | seqret --filter -sbegin 2000 &gt; part3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
part1 part2 part3 | union -filter  &gt; genome.fa</font></font></p>
<p class="first-paragraph">The resulting image is best described as a
&ldquo;quite the head-scratcher&rdquo;. Bioinformaticians look at one
another and say: &ldquo;What the heck is THAT?&rdquo;. Since we know
what we put in we can explain what we see &ndash; had this been an
unknown reorganization, figuring out what happened would have been a
lot more difficult.</p>
<p><img src="0000_html_750600a0ddf4fdea.png" name="Image143" align="bottom" hspace="12" width="560" height="196" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-variant-calling"></a>95
Introduction to variant calling</h1>
<p class="first-paragraph">Variant calling is the process of
identifying and cataloging the differences between the observed
sequencing reads and a reference genome.</p>
<h2 class="western"><a name="what-is-the-process-for-variant-calling"></a>
95.1 What is the process for variant calling?</h2>
<p class="first-paragraph">The process is relatively straightforward
as long as you recognize that the process is a series of
interconnected yet still separate and independent tasks, rather than
a single one-step analysis.</p>
<p>At the simplest level variants are usually determined (&ldquo;called&rdquo;)
from alignments (BAM) files. A typical process involves the
following:</p>
<ol>
	<li/>
<p class="compact">Align reads to the reference.</p>
	<li/>
<p class="compact">Correct and refine alignments (optional).</p>
	<li/>
<p class="compact">Determine variants from the alignments.</p>
	<li/>
<p class="compact">Filter the resulting variants for the
	desired characteristics.</p>
	<li/>
<p class="compact">Annotate filtered variants.</p>
</ol>
<p class="first-paragraph">Let us mention that each of the steps
above allows for substantial customizations that in turn may
significantly alter the outcomes in the later stages.</p>
<p>In the variant calling process you will run into terminologies
that are not always used consistently. We will attempt to define a
few of these terms while making a note that these definitions are not
&ldquo;textbook&rdquo; definitions designed to accurately capture the
proper terminology of Genetics as science. Our definitions are what
people &lsquo;call&rsquo; certain features of their data.</p>
<h2 class="western"><a name="what-is-the-ploidy"></a>95.2 What is the
ploidy?</h2>
<p class="first-paragraph">The ploidy is the number of complete sets
of chromosomes in a cell and hence represent the number of possible
alleles (variant forms) of the DNA of a <b>single</b> cell.</p>
<h2 class="western"><a name="how-many-possible-alleles-will-i-see"></a>
95.3 How many possible alleles will I see?</h2>
<p class="first-paragraph">It is essential that you identify early on
the constraints of the problem under study. If all the cells carry
the same DNA then the number of alleles will correspond to the ploidy
of the organism. If the cells represent a population of N distinct
individuals then the maximum possible alleles will be N x ploidy.</p>
<h2 class="western"><a name="what-is-a-haplotype-1"></a>95.4 What is
a haplotype?</h2>
<p class="first-paragraph">A haplotype (haploid genotype) is a group
of alleles in an organism that are (or may be) inherited together
from a single parent. Variants that occur on the same DNA molecule
form a haplotype. We also call such variants as &ldquo;phased&rdquo;
variants.</p>
<h2 class="western"><a name="what-is-a-genoytpe"></a>95.5 What is a
genoytpe?</h2>
<p class="first-paragraph">We found that the word &ldquo;genotype&rdquo;
is used quite inconsistently, often with wildly different meanings.
The simplest way to think about a genoytpe is as the collection of
all known alleles. Both genotypes and haplotypes may be complete or
incomplete (partially known).</p>
<p>The word &ldquo;genotyping&rdquo; usually means that we can assign
one or more sequence reads to a known group of variants (genotype).</p>
<h2 class="western"><a name="what-is-the-best-method-to-call-variants"></a>
95.6 What is the best method to call variants?</h2>
<p class="first-paragraph">There is no universal rule, method or
protocol that would always produce correct answers or guarantee some
range of optimality. Moreover, any method, even a simple evaluation
or even an incorrect rationale may be able to create seemingly right
calls when the signal in the data is robust and unambiguous.</p>
<p>Much ink and effort are spent generating and comparing results
across massive datasets; you can find these as being labeled as
&ldquo;concordance.&rdquo; of SNP callers. Alas, in our finding none
of these reports ever provide a clear picture or even advice on which
tool to pick or which works better under what circumstance.</p>
<p>This is not to say that we don&rsquo;t have access to reasonably
reliable variant callers. Scientists have been working feverishly in
this field, and radical improvements have been and are continually
being made.</p>
<p>What you need to recognize is that the task of describing
differences may in some cases turn out to be far more complicated
than anticipated - yet there is no way to tell - apriori - which
category a given study will fall into.</p>
<h2 class="western"><a name="what-are-the-most-commonly-used-variant-callers"></a>
95.7 What are the most commonly used variant callers?</h2>
<p class="first-paragraph">There are &ldquo;generic&rdquo; variant
callers:</p>
<ul>
	<li/>
<p class="compact">bcftools</p>
	<li/>
<p class="compact">FreeBayes</p>
	<li/>
<p class="compact">GATK</p>
	<li/>
<p class="compact">VarScan2</p>
</ul>
<p class="first-paragraph">Besides, there are specialized software
designed to account for specific properties of the data, such as
somatic mutations, germline mutations, family trios, and many others.
Specialized tools typically can make use of other features of the
data and produce more accurate variant calls.</p>
<h2 class="western"><a name="how-are-variants-represented"></a>95.8
How are variants represented?</h2>
<p class="first-paragraph">The &ldquo;standard&rdquo; representation
for variants is the so-called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
format described in a subsequent chapter: <font color="#4f81bd"><a href="#ch099.xhtml#vcf">VCF:
The Variant Call Format</a></font>. In this section we jump ahead and
show you how to create <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VCF</font></font>
files, then later we explain how the format represents the
information.</p>
<h2 class="western"><a name="how-do-i-generate-vcf-files"></a>95.9
How do I generate VCF files?</h2>
<p class="first-paragraph">Typically a VCF file is a result of
running a &ldquo;variant caller&rdquo; (or &ldquo;SNP caller&rdquo;
as some people call it) on one or more BAM alignment files. The
result of running the variant caller will be a VCF file that contains
a column for each sample. Samples may be stored in different BAM
files or may a single BAM file may include multiple samples tagged
via read-groups.</p>
<p>We have used the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
simulator before, another simulator based on <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
(install with that command <font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install dwgsim</font></font>) also produces a VCF file that lists the
mutations</p>
<p>As before there is a lot of value in generating known mutations as
it allows us to evaluate and benchmark different variant calling
approaches. Below we show how to generate a mutated genome then use a
variant caller to predict the known mutations.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reference accession numbers.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create the directory for reference file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The name of the reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/${ACC}.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The name of the BAM file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM=align.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id $ACC &gt; $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a bwa index for the reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a samtools index for the reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Simulate reads from the reference file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim
$REF simulated</font></font></p>
<p class="first-paragraph">Running <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
as above will produce a file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">simulated.mutations.vcf</font></font>.
Take a moment to investigate this file (your file will be different
since it will contain different random changes)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##fileformat=VCFv4.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##contig=&lt;ID=AF086833,length=18959&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##INFO=&lt;ID=AF,Number=A,Type=Float,Description=&quot;Allele
Frequency&quot;&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##INFO=&lt;ID=pl,Number=1,Type=Integer,Description=&quot;Phasing:
1 - HET contig 1, #2 - HET contig #2, 3 - HOM both
contigs&quot;&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##INFO=&lt;ID=mt,Number=1,Type=String,Description=&quot;Variant
Type: SUBSTITUTE/INSERT/DELETE&quot;&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#CHROM
     POS    ID    REF ALT   QUAL    FILTER        INFO</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   1093    .    T    C    100    PASS   
AF=0.5;pl=2;mt=SUBSTITUTE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   1195    .    A    T    100    PASS   
AF=0.5;pl=1;mt=SUBSTITUTE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   2566    .    G    T    100    PASS   
AF=0.5;pl=2;mt=SUBSTITUTE</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
   3345    .    A    AC   100    PASS    AF=0.5;pl=2;mt=INSERT</font></font></p>
<p class="first-paragraph">The file above is a VCF file discussed in
the section: <font color="#4f81bd"><a href="#ch099.xhtml#vcf">VCF:
The Variant Call Format</a></font>.</p>
<p>Let&rsquo;s now generate sequencing reads from the mutated genome.</p>
<p>In this example we will use the software package called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>.
This <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
suite has a very broad number of use cases beyond variant calling
that we will explore in other chapters.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This is the data naming generated by
dwgsim.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=simulated.bwa.read1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=simulated.bwa.read2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate the alignment from the data simulated above.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $R1 $R2 | samtools sort &gt; $BAM</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the BAM file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index $BAM</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Compute the genotypes from the alignment file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
mpileup -Ovu -f $REF $BAM &gt; genotypes.vcf</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Call the variants from the genotypes.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
call -vc -Ov genotypes.vcf &gt; observed-mutations.vcf</font></font></p>
<p class="first-paragraph">The resulting file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">observed-mutations.vcf</font></font>
may be visualized in IGV and compared to the file
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">simulated-mutations.vcf</font></font>:</p>
<p><img src="0000_html_de7bafac987bbfb9.png" name="Image144" align="bottom" hspace="12" width="560" height="187" border="0"/>
</p>
<p>Study these outputs, perform multiple runs, tune <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
to produce more variations, perhaps larger errors and so on. You will
develop a sense of what types of mutations are reliably called and
which ones are less so.</p>
<p>Experiment with different genomes, repetitive and more redundant
genomes will produce data that produces less accurate results.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="variant-calling-example"></a>96 Variant
calling example</h1>
<p class="first-paragraph">We demonstrate variant calling for the
data from <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font>.</p>
<h2 class="western"><a name="which-differences-should-i-be-looking-at"></a>
96.1 Which differences should I be looking at?</h2>
<p class="first-paragraph">Remember that &ldquo;variants&rdquo; are
always defined relative to &ldquo;something&rdquo;. Variants do not
exist on their own, only relative to another, similar sequence. You
decide which comparison needs to be made. Sometimes there is a
commonly used &ldquo;reference&rdquo; - other times it is up to you
to decide which comparison is the most appropriate.</p>
<p>Let us investigate the genomic changes of the 2014 outbreak
relative to the sequence of the 1976 outbreak stored under the
accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833</font></font>.
We will now have to make use of all the techniques that we&rsquo;ve
have learned so far to find out what has changed in 38 years. We will
do the following:</p>
<ol>
	<li/>
<p class="compact">Obtain and prepare the reference genomes</p>
	<li/>
<p class="compact">Obtain the sequencing data</p>
	<li/>
<p class="compact">Perform the alignments</p>
	<li/>
<p class="compact">Perform the variant calling</p>
</ol>
<h2 class="western"><a name="how-do-i-prepare-the-reference-genome-for-variant-calling"></a>
96.2 How do I prepare the reference genome for variant calling?</h2>
<p class="first-paragraph">Since there are quite many steps, the
workflow is well suited for automation and scripting. The following
code is structured as a script that you could create as a &ldquo;starter
variation calling&rdquo; script.</p>
<p>First, we set up the filenames as variables. Doing so helps reuse
the code for other data and makes the steps shorter and easier to see
what file goes where:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reference genome accession number.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
SRR run number.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR=SRR1553500</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The directory to store the reference in.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The name of the file that stores the reference
genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/$ACC.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
The resulting alignment file name.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM=$SRR.bam</font></font></p>
<p class="first-paragraph">Obtain the sequences:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the reference sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id $ACC | seqret -filter -sid $ACC &gt;
$REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index reference for the aligner.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the reference genome for IGV</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font></p>
<h2 class="western"><a name="how-do-i-obtain-the-sequencing-data"></a>
96.3 How do I obtain the sequencing data?</h2>
<p class="first-paragraph">We will use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
to access data deposited in SRA:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get data from an Ebola sequencing run.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-X 100000 --split-files $SRR</font></font></p>
<h2 class="western"><a name="how-do-i-align-sequencing-data-against-a-reference"></a>
96.4 How do I align sequencing data against a reference?</h2>
<p class="first-paragraph">We may use different aligners; we chose
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
below:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Shortcut to read names.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R1=${SRR}_1.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R2=${SRR}_2.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Tag the alignments. GATK will only work when reads are
tagged.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TAG=&quot;@RG\tID:$SRR\tSM:$SRR\tLB:$SRR&quot;</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Align and generate a BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem -R $TAG $REF $R1 $R2 | samtools sort &gt; $BAM</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# Index the BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index $BAM</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font>
</p>
<h2 class="western"><a name="how-do-i-use-bcftools-to-call-variants"></a>
96.5 How do I use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
to call variants?</h2>
<p class="first-paragraph">Samtools/BCFTools can be used to call SNPs
in the following way:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Determine the genotype likelihoods for each base.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
mpileup -Ovu -f $REF $BAM &gt; genotypes.vcf</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Call the variants with bcftools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
call --ploidy 1 -vm -O v genotypes.vcf &gt; variants1.vcf</font></font></p>
<p class="first-paragraph">You may also run both commands on one line
to avoid having to store the intermediate, often fairly large
genotype file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
mpileup -O v -f $REF $BAM | bcftools call --ploidy 1 -vm -O v &gt; 
variants1.vcf</font></font></p>
<p class="first-paragraph">The genomes have changed a lot, view the
resulting VCF file in IGV:</p>
<p><img src="0000_html_63eee3e4e432c81c.png" name="Image145" align="bottom" hspace="12" width="560" height="188" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-use-the-freebayes-variant-caller"></a>
96.6 How do I use the FreeBayes variant caller?</h2>
<p class="first-paragraph">On this same data the FreeBayes variant
caller may be invoked as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">freebayes
-f $REF $BAM &gt; variants2.vcf</font></font></p>
<h2 class="western"><a name="what-is-gatk-the-genome-analysis-toolkit"></a>
96.7 What is GATK (The Genome Analysis Toolkit)?</h2>
<p class="first-paragraph">GATK is not just a variant caller - it is
a toolset of epic complexity that offers features and functionality
that most of us won&rsquo;t ever understand (or need for that
matter).</p>
<p>The Genome Analysis Toolkit may be one of the best variant callers
and would be our first choice especially for higher order organisms
with complex genomes (human, mouse etc) had it not been for the
unnecessarily awkward usage and incredible mental overhead that comes
with it.</p>
<p>When you use GATK you have to submit to their will, jump through
hoops, bend the knee, drink the kool-aid all at the same time. Every
step ends up overcomplicated with the many small inconsistencies
adding up to create perhaps the worst user experience of all
bioinformatics.</p>
<p>In our opinion, GATK is the worst offender when it comes to
capricious and seemingly ludicrous behaviors, tedious and odious
program interfacing with people. Now, if you learn how to put up with
it all other tools will be a joy to use - you will burst into signing
every time it is not GATK that you have to use.</p>
<p>To their credit, the GATK team maintains a forum where they do
their best to address the myriad of legitimate questions that their
user-unfriendly tool generates:</p>
<ul>
	<li/>
<p class="compact">Main website:
	<font color="#4f81bd"><a href="https://software.broadinstitute.org/gatk/">https://software.broadinstitute.org/gatk/</a></font></p>
</ul>
<p class="first-paragraph">GATK also has a &ldquo;best practices&rdquo;
site that presents recommended workflows for variant calling using
GATK.</p>
<ul>
	<li/>
<p class="compact">GATK best practices:
	<font color="#4f81bd"><a href="https://software.broadinstitute.org/gatk/best-practices/">https://software.broadinstitute.org/gatk/best-practices/</a></font></p>
</ul>
<p class="first-paragraph">Good luck comrade, you are going to need
it!</p>
<h2 class="western"><a name="are-there-multiple-versions-of-gatk"></a>
96.8 Are there multiple versions of GATK?</h2>
<p class="first-paragraph">Version 3.8 has many features that version
4.0 does not support and vice versa. In addition, the way we make use
of GATK version 3.8 is only remotely similar to the way GATK version
4.0 works.</p>
<p><img src="0000_html_c1e3fe8d0eafe66d.png" name="Image146" align="bottom" hspace="12" width="560" height="127" border="0"/>
</p>
<p>Long story short, GATK is not for the faint-hearted. But if you
are a seasoned bioinformatician, you will find that GATK is a great
toolset to use.</p>
<h2 class="western"><a name="how-to-install-gatk"></a>96.9 How to
install GATK?</h2>
<p class="first-paragraph">The instructions to install GATK seemingly
change in time. Currently two gatk versions are available:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
search gatk</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mamba
search gatk4</font></font></p>
<p class="first-paragraph">The former package installs gatk3. The
latter isntalls gatk4. Each can be run as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk3</font></font></p>
<p class="first-paragraph">or</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk4</font></font></p>
<p class="first-paragraph">You may also download the latest version
of GATK from the Broad website:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://gatk.broadinstitute.org/">https://gatk.broadinstitute.org/</a></font></p>
</ul>
<p class="first-paragraph">Unpack the archive and to use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk</font></font>
located there.</p>
<h2 class="western"><a name="how-do-i-use-gatk"></a>96.10 How do I
use GATK?</h2>
<p class="first-paragraph">GATK operate via &ldquo;tools&rdquo;.
Tools are independent programs that provide specific functionality.
These are typically run as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk
SomeToolName -R $REF -I $BAM -O something -X 200</font></font></p>
<p class="first-paragraph">Here the tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SomeToolName</font></font>
will be run with the parameters listed on the right.</p>
<p>Besides, a separate program named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">picard</font></font>
(a homage to Capt.&nbsp;Picard of Star Trek Nex-Generation) is often
required to prepare and manipulate data.</p>
<p>For our example, the Genome Analysis Toolkit could be used with
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HaplotypeCaller</font></font>
tool as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index the reference with samtools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
It needs yet another so-called Dictionary index.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">picard
CreateSequenceDictionary REFERENCE=$REF  OUTPUT=refs/$ACC.dict</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate the variants.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk
HaplotypeCaller -R $REF -I $BAM -O variants3.vcf</font></font></p>
<h2 class="western"><a name="how-can-i-improve-the-quality-of-variant-calls"></a>
96.11 How can I improve the quality of variant calls?</h2>
<p class="first-paragraph">Most pipelines and protocols recommend a
series of quality filtering steps and processes that will improve the
quality of the variant calls. These are too numerous to enumerate
here but can be found under titles such as:</p>
<ul>
	<li/>
<p class="compact">GATK best practices:
	<font color="#4f81bd"><a href="https://software.broadinstitute.org/gatk/best-practices/">https://software.broadinstitute.org/gatk/best-practices/</a></font></p>
</ul>
<p class="first-paragraph">You would also need to read a few
documents, papers, and manuals that summarize other researcher&rsquo;s
experiences with the exact domain that you are interested in. For
example, the VarScan2 manual states that when investigating trios (2
parents + child):</p>
<p class="block-text-western">the ideal de novo mutation call will
have high depth (&gt;20x) in all three samples, with good support in
the child (40-50% of reads for autosomal calls) and no
variant-supporting reads in either parent.&rdquo;</p>
<p class="first-paragraph">Do understand that best-practices merely
improve on a process with diminishing returns but cannot, on their
own, solve some of the most significant problems of variant calling.</p>
<h2 class="western"><a name="can-i-customize-the-variant-calling-process"></a>
96.12 Can I customize the variant calling process?</h2>
<p class="first-paragraph">Just as with aligners the process of
calling variants can be significantly customized. Some parameters
(e.g., ploidy), reflect certain expectations of the genome and sample
collection under study and will need to be set to match those
expectations.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="multi-sample-variant-calling"></a>97
Multi-sample variant calling</h1>
<h2 class="western"><a name="what-is-multi-sample-variant-calling"></a>
97.1 What is multi-sample variant calling</h2>
<p class="first-paragraph">A multi-sample variant call is an
extension of a single sample variant call, where all the samples are
evaluated at the same time, and the presence and absence of a variant
is indicated for all samples in parallel.</p>
<p><img src="0000_html_65e9cdd8048b5be7.jpg" name="Image147" align="bottom" hspace="12" width="560" height="152" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-perform-multi-sample-variant-calling-process"></a>
97.2 How do I perform multi-sample variant calling process?</h2>
<p class="first-paragraph">There are two ways to perform multi-sample
variant calls:</p>
<ol>
	<li/>
<p class="compact">Call the variants simultaneously on all BAM
	files</p>
	<li/>
<p class="compact">Merge variants called individually in
	separate VCF files</p>
</ol>
<p class="first-paragraph">In general if possible use the first
method. Merging VCF files may be fraught by subtle problems,
especially whenever different versions of tools created the VCF files</p>
<p>Recall that we found the SRR <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runids</font></font>
for the Ebola project by searching the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA</font></font>
database with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA257197</font></font>
and formatting the output as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo</font></font>.</p>
<p>If you were to follow the variant calling as presented in <font color="#4f81bd"><a href="#ch096.xhtml#variant-calling-example">Variant
calling example</a></font> you would end up with a BAM file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1553500</font></font>.
Place the code (or consult the corresponding recipe) to create other
BAM files (one or more) for ids that you found when searching for SRR
numbers that belong to the same project. For example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972920</font></font></p>
<p class="first-paragraph">Now you have a BAM file for each of the
SRR runs.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
mpileup -O v -f $REF  bam/*.bam &gt; genotypes.vcf</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
call --ploidy 1 -vm -O v |  bcftools norm -O v -f $REF -d all - &gt;
variants.vcf</font></font></p>
<p class="first-paragraph">to produce a multisample variant call that
contains the information on all samples:</p>
<p><img src="0000_html_d93e877cf7283bfc.png" name="Image148" align="bottom" hspace="12" width="560" height="220" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-interpret-the-visualization"></a>
97.3 How do I interpret the visualization?</h2>
<p class="first-paragraph">Read more on how IGV visualizes VCF files
below:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://software.broadinstitute.org/software/igv/viewing_vcf_files">Viewing
	VCF Files</a></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="variant-normalization"></a>98 Variant
normalization</h1>
<p class="first-paragraph">Conceptually the &ldquo;word&rdquo;
variation captures unexpectedly complex information. Once we move
past simple, single step changes, the same results could be
represented in alternative ways.</p>
<h2 class="western"><a name="what-is-variant-normalization"></a>98.1
What is variant normalization?</h2>
<p class="first-paragraph">In <font color="#4f81bd"><a href="http://bioinformatics.oxfordjournals.org/content/31/13/2202">Unified
representation of genetic variants, Bioinformatics 2015</a></font>
the author state:</p>
<p class="block-text-western">A genetic variant can be represented in
the Variant Call Format (VCF) in multiple different ways.
Inconsistent representation of variants between variant callers and
analyses will magnify discrepancies between them and complicate
variant filtering and duplicate removal.</p>
<p class="first-paragraph">For example, the same variant may be
represented as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAC
--&gt; AC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGAC
--&gt; AC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GACC
--&gt; ACC</font></font></p>
<p class="first-paragraph">Variant normalization is the process of
simplifying the representation of a variant to obey the following
rules:</p>
<ul>
	<li/>
<p class="compact">Represent the variant with as few letters
	as possible.</p>
	<li/>
<p class="compact">No allele may be zero length.</p>
	<li/>
<p class="compact">Variant must be &ldquo;left aligned&rdquo;
	(this also applies to repetitive regions).</p>
</ul>
<p class="first-paragraph">A variant is &ldquo;left aligned&rdquo; if
it is no longer possible to shift its position to the left while
keeping the length of all its alleles constant. For example, an
insertion of C into</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCCC</font></font></p>
<p class="first-paragraph">would need to be reported at the leftmost
position.</p>
<h2 class="western"><a name="should-variant-callers-report-normalized-variants-by-default"></a>
98.2 Should variant callers report normalized variants by default?</h2>
<p class="first-paragraph">Yes and eventually they will do so by
default. GATK for example already produces normalized variants. The
output of other tools ought to be normalized.</p>
<h2 class="western"><a name="how-do-i-normalize-a-variant"></a>98.3
How do I normalize a variant?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">vt</font></font>
tools have actions that perform variant normalization.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
norm -f $REF samples.vcf</font></font></p>
<p class="first-paragraph">and</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">vt
normalize -r $REF samples.vcf</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-variant-call-format-vcf"></a>99 The
Variant Call Format (VCF)</h1>
<h2 class="western"><a name="what-is-the-vcf-format"></a>99.1 What is
the VCF format?</h2>
<p class="first-paragraph">The variant call format (VCF) is a data
representation format used to describe variations in the genome. A
VCF file may contain information on any number of samples, thousands
even and can be thought of as a single database that summarizes the
final results of multiple experiments in a single file.</p>
<p>As a format, VCF is composed of two sections: a <b>header</b>
section and a <b>record</b> section.</p>
<p>While it is a plain text format, its contents can go from being
easy to read and understand to seemingly impossible to visually
decipher.</p>
<p>The challenge of reading out variants by eye is (probably) is not
the format&rsquo;s &ldquo;fault&rdquo; - the variations in the genome
can have relatively complex representations.</p>
<h2 class="western"><a name="how-important-is-it-to-understand-the-vcf-format"></a>
99.2 How important is it to understand the VCF format?</h2>
<p class="first-paragraph">In many cases, it is possible to perform
sophisticated analyses without ever directly looking inside a VCF
file. Many tools take VCF files as inputs and produce other reports
out of them. At other times a cursory understanding of what a VCF
file contains and how the information is laid inside of them out is
sufficient to complete an entire analysis.</p>
<p>But just as with the SAM format - understanding the VCF format is
essential for anyone that needs to venture off the beaten path. There
will be times when crucial information is buried deep inside a VCF
file with hundreds of samples. You may have to use toolsets like
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
to filter, combine and subselect VCF files in every which way
imaginable. At those times understanding the small details of the VCF
format become indispensable.</p>
<h2 class="western"><a name="what-is-a-vcf-header"></a>99.3 What is a
VCF header?</h2>
<p class="first-paragraph">VCF header sections are located at the
beginning of the VCF files and consists of lines that start with the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##</font></font>
symbols like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##fileformat=VCFv4.1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##contig=&lt;ID=AF086833,length=18959&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##INFO=&lt;ID=AF,Number=A,Type=Float,Description=&quot;Allele
Frequency&quot;&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</font></font></p>
<p class="first-paragraph">As variant calling is a more complex
operation than alignment, additional information needs to be recorded
about the process.</p>
<p>The VCF headers contain the specifications on various terms used
throughout the file. For example, from the above header, we learn
that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF</font></font>
in the VCF file will be used to describe a variant and it means
&ldquo;Allele Frequency&rdquo; and seeing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF=0.5</font></font>
would mean that the allele frequency for a variant was measured to be
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.5</font></font>.</p>
<h2 class="western"><a name="what-is-the-difference-between-the-info-and-the-format-fields"></a>
99.4 What is the difference between the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">INFO</font></font>
and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
fields?</h2>
<p class="first-paragraph">The starting word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">INFO</font></font>
in the definition line above means that the ID will describe a
variant across <b>all samples</b>. The definition from the header
applies to (or has been derived from) all samples together and those
characterize the variant.</p>
<p>When the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
is used it would mean that, for this ID you will have a different
value for each sample.</p>
<h2 class="western"><a name="what-do-the-words-mean-1"></a>99.5 What
do the words mean?</h2>
<p class="first-paragraph">What does <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Allele
Frequency</font></font> mean precisely? And where is the actual
definition of how the value for this term was computed?</p>
<p>Some words are well-defined concepts from related scientific
fields such as population genetics and will be relatively easy to
find a meaning for. For example, if you did not know beforehand, a
Google search could tell you that &ldquo;allele frequency&rdquo; is
the frequency of a variant in a population. It can be defined as the
fraction of all chromosomes that carry the allele.</p>
<p>But there will be terms for which it will be much harder to find a
proper explanation. For example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
INFO=&lt;ID=VDB,Number=1,Type=Float,Description=&quot;Variant
Distance Bias for </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
filtering splice-site artefacts in RNA-seq data (bigger is
better)&quot;,Version=&quot;3&quot;&gt;</font></font></p>
<p class="first-paragraph">So what is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VDB</font></font>
then? It states that it is: <i>Variant Distance Bias for ## filtering
splice-site artefacts in RNA-seq data (bigger is better)</i>. Ok what
does that even mean? Turns out you&rsquo;d have to do a lot of
legwork to figure it out, perhaps read the manual or publication of
the variant caller that produced this value. And occasionally, for
some terms, you may not be able to find (or understand) the
explanation on your own. To some extent, this is the antithesis of
what science is. Some number that someone thought is relevant&hellip;</p>
<p>Note how even the creator of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VDB</font></font>
ID felt the need to mention that for this value &ldquo;bigger is
better.&rdquo; That is not all that encouraging from the perspective
of a scientist - you are given a number that you don&rsquo;t even
know which way is &ldquo;better&rdquo;. You might ask yourself the
question, do I even have to know about <font face="Consolas, serif"><font size="2" style="font-size: 11pt">VDB</font></font>
in the first place? To which the answer usually is, no, you don&rsquo;t.</p>
<p>Some (many or even most) IDs might be immaterial to the study that
you are conducting. Variant callers are very prolific and liberally
sprinkle values of all kinds of quantities and measures that the
creators of these tools felt that should always be mentioned. Only a
few are these will be relevant most of the time you neither need to
nor would want to know them.</p>
<h2 class="western"><a name="what-are-vcf-records"></a>99.6 What are
VCF records?</h2>
<p class="first-paragraph">The record section of a VCF file consists
of at least eight tab-delimited columns where the first eight columns
describe a variant and the rest of the columns describe the
properties of each sample. The 9th column is the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
and each column past the 9th will represent a sample.</p>
<p>As mentioned before, but it is worth reiterating, a<br/>
VCF file
may contain any number of sample columns, thousands even and can be
thought of as a single database that represents all variations in all
samples.</p>
<p>The first nine columns of a VCF file are:</p>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CHROM</font></font>:
	The chromosome (contig) on which the variant occurs</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font>:
	The genomic coordinates on which the variant occurs. For deletions,
	the position given here are on of the bases preceding the event.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID</font></font>:
	An identifier for the variant (if it exists). Typically a dbSNP
	database if that is known.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF</font></font>:
	The reference allele on the forward strand.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>:
	The alternate allele(s) on the forward strand. More than one may be
	present.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">QUAL</font></font>:
	A probability that the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF/ALT</font></font>
	variant exists at this site. It is in Phred scale, just as the FASTQ
	quality and the MAPQ field in the SAM file are.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FILTER</font></font>:
	The name of filters that the variant fails to pass, or the value
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">PASS</font></font>
	if the variant passed all filters. If the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FILTER</font></font>
	value is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>,
	then no filtering has been applied to the record.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">INFO</font></font>:
	Contains the site-specific annotations represented as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ID=VALUE</font></font>
	format.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>:
	Sample-level annotations as colon separated TAGS.</p>
</ol>
<p class="first-paragraph">For example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#CHROM
   POS   ID  REF    ALT   QUAL   FILTER  INFO    FORMAT    
align.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  60   .    A      T     54       .    DP=43    GT:PL    0/1:51,0,48</font></font></p>
<h2 class="western"><a name="how-to-interpret-the-format-field"></a>99.7
How to interpret the FORMAT field?</h2>
<p class="first-paragraph">This field specifies the meaning of the
numbers in the each column of the sample. The fields are colon <font face="Consolas, serif"><font size="2" style="font-size: 11pt">:</font></font>
separated and map each field of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
to each value in the sample column. Suppose that the following were
columns 9,10 and 11 of a VCF file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT
        sample1         sample2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT:PL
        0/1:51,0,48    1/1:34,32,0    </font></font>
</p>
<p class="first-paragraph">We interpret it in the following way.</p>
<ul>
	<li/>
<p class="compact">The variant observed for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sample1</font></font>
	has the values <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT=0/1</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL=51,0,48</font></font></p>
	<li/>
<p class="compact">This same variant when observed in <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sample2</font></font>
	has the values <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT=1/1</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL=34,32,0</font></font></p>
</ul>
<p class="first-paragraph">What if you wanted to know what do <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL</font></font>
mean? You&rsquo;d have to go to the header for their definition, and
there you would find</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description=&quot;List
of Phred-scaled genotype likelihoods&quot;&gt;</font></font></p>
<h2 class="western"><a name="what-is-represented-in-the-refalt-columns."></a>
99.8 What is represented in the REF/ALT columns.</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF</font></font>
column represents the reference allele at position <font face="Consolas, serif"><font size="2" style="font-size: 11pt">POS</font></font></p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>
column represent <b>all</b> variants observed at that site. When
there is only a single simple variant, say a SNP at position 60 where
an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
was replaced by a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>
the VCF will look eminently readable. The variant record would state:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#CHROM
   POS   ID  REF    ALT   QUAL   FILTER  INFO    FORMAT    
align.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  60   .    A      T     54       .    DP=43    GT:PL    0/1:51,0,48</font></font></p>
<p class="first-paragraph">It also tells us that the genotype is
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT=0/1</font></font>
and that Phred-scaled genotype likelihoods are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL=51,0,48</font></font>
(we will explain both concepts in more detail a bit later).</p>
<p>Now when there are two possible variants in the same location, it
starts to be a little more complicated. This site now has two
alternates at the same site:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#CHROM
   POS   ID  REF    ALT   QUAL  FILTER  INFO     FORMAT        
align.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  60   .    A     T,C   43.2     .    DP=95    GT:PL  
1/2:102,124,42,108,0,48</font></font></p>
<p class="first-paragraph">There are two alternates, and we see from
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
field that our sample was heterozygous for the alternates. Neither of
the alleles matched the reference.</p>
<p>Alas, the complexity does not stop there. This is where the VCF
format becomes a lot harder to interpret. Previously we only mutated
base <font face="Consolas, serif"><font size="2" style="font-size: 11pt">60</font></font>.
But what if beyond just the mutation, some members of this population
had a three-base deletion at coordinates <font face="Consolas, serif"><font size="2" style="font-size: 11pt">58</font></font>,<font face="Consolas, serif"><font size="2" style="font-size: 11pt">59</font></font>
and, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">60</font></font>.
Now the our VCF record will show:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#CHROM
   POS   ID  REF    ALT   QUAL  FILTER   INFO         FORMAT       
align.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  55   .   TGAGGA  TGA   182    .     INDEL;IDV=42   GT:PL   
0/1:215,0,255</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AF086833
  60   .   A       C,T   39.8   .     DP=126;        GT:PL   
1/2:99,87,47,86,0,49</font></font></p>
<p class="first-paragraph">What is important to note here that even
though the deletion started at coordinate <font face="Consolas, serif"><font size="2" style="font-size: 11pt">58</font></font>
is shown at coordinate <font face="Consolas, serif"><font size="2" style="font-size: 11pt">55</font></font>
in the file, but since the both the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF</font></font>
start with identical bases the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGA</font></font>
difference between them begin only after three bases. So when <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGAGGA</font></font>
is replaced by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGA</font></font>
the actual change is the three base deletion starting from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">58</font></font>
and will overlap with coordinate <font face="Consolas, serif"><font size="2" style="font-size: 11pt">60</font></font>.
But it is clear that you can&rsquo;t easily see that from the VCF
file.</p>
<p>You may even ask yourself why is the variant reported as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TGAGGA
--&gt; TGA</font></font>? Why not <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GAGGA
--&gt; GAGGA</font></font> or even <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGA
-&gt;</font></font>? Clearly, we could report these difference in
many different ways. You&rsquo;ll be surprised to know that this is
still a not fully resolved problem in variant reporting.</p>
<p>There is a method called &ldquo;variant normalization&rdquo; that
attempts to ensure that variants are reported identically. To explore
that and understand how variants are determined and represented see
the chapter [NEXT WEEK]</p>
<h2 class="western"><a name="how-are-genotypes-described"></a>99.9
How are genotypes described?</h2>
<p class="first-paragraph">Whereas the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>
columns show the change, we need to know how many of the copies of
the DNA carry the variant or variants. The variant is encoded in the
field called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT</font></font>
that indicates the genotype of this sample at this site.</p>
<p>It is constructed out of slash-separated numbers where:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font>
	indicates the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF</font></font>
	field,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
	indicates the first entry in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>
	field,</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>
	indicates the second entry in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ALT</font></font>
	field and so on.</p>
</ul>
<p class="first-paragraph">For example for a diploid organism the GT
field indicates the two alleles carried by the sample:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/0</font></font>
	- the sample is a homozygous reference</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1</font></font>
	- the sample is heterozygous, carrying one of each the REF and ALT
	alleles</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/2</font></font>
	- would indicate a heterozygous carrying one copy of each of the ALT
	alleles.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1</font></font>
	- the sample is homozygous for the first alternate</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2/2</font></font>
	- the sample is heterozygous for the second alternate</p>
</ul>
<p class="first-paragraph">and so on. For samples of higher ploidy
(multiple copies of chromosomes), there would be as many <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font>
characters as there are copies of each chromosome. For example, a
tetraploid genotype could be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1/0/2</font></font>.</p>
<h2 class="western"><a name="what-are-genotype-likelihoods"></a>99.10
What are genotype likelihoods?</h2>
<p class="first-paragraph">The genotype likelihoods are normalized
probabilities that describe the likelihood that each of the possible
genotypes occurs. The value is normalized to 0. So the most likely
genotype is 0, and the rest are computed relative to that. Since it
is a statistical measure (a p-value), it is the chance of not
observing the genotype hence the &ldquo;best&rdquo; value is 0.</p>
<p>So let&rsquo;s take the:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GT:PL
   0/1:51,0,48</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PL</font></font>
field will contain three numbers, corresponding to the three possible
genotypes (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/0</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1</font></font>,
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1</font></font>).
What is says the following:</p>
<ul>
	<li/>
<p class="compact">The support is highest for genotype <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1</font></font>
	and that is shown as zero.</p>
	<li/>
<p class="compact">The chance that this genotype is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/0</font></font>
	instead of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1</font></font>
	is Phred score 51 meaning <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.000001</font></font></p>
	<li/>
<p class="compact">The chance that this genotype is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1/1</font></font>
	instead of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0/1</font></font>
	is Phred score 48 meaning <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00001</font></font></p>
</ul>
<p class="first-paragraph">We got the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.000001</font></font>
from the formula <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10^(-51/10)</font></font>
it is the same Phred score that we grew to love (not really) where
you drop a digit from it and that&rsquo;s how many numbers you go to
the left 40 -&gt; 0.0001</p>
<h2 class="western"><a name="what-is-a-bcf-file"></a>99.11 What is a
BCF file?</h2>
<p class="first-paragraph">The BCF format is a binary representation
of the VCF format. It has a similar relationship as the SAM/BAM but
with the exception that the BCF files are not sorted and indexed and
cannot be queried the same way as typical BAM files do.</p>
<h2 class="western"><a name="what-are-additional-resources"></a>99.12
What are additional resources?</h2>
<ul>
	<li/>
<p class="compact">VCF Poster:
	<font color="#4f81bd"><a href="http://vcftools.sourceforge.net/VCF-poster.pdf">http://vcftools.sourceforge.net/VCF-poster.pdf</a></font></p>
	<li/>
<p class="compact">VCF short summary:
	<font color="#4f81bd"><a href="http://www.htslib.org/doc/vcf.html">http://www.htslib.org/doc/vcf.html</a></font></p>
	<li/>
<p class="compact">VCF Specification:
	<font color="#4f81bd"><a href="http://samtools.github.io/hts-specs/">http://samtools.github.io/hts-specs/</a></font></p>
	<li/>
<p class="compact">A detailed description of the VCF: <font color="#4f81bd"><a href="http://gatkforums.broadinstitute.org/gatk/discussion/1268/what-is-a-vcf-and-how-should-i-interpret-it">What
	is a VCF file</a></font> on the GATK forums.</p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="filtering-information-in-vcf-files"></a>100
Filtering information in VCF files</h1>
<p class="first-paragraph">If you produce VCF file, sooner or later
you will need to filter this file, to keep or to remove variants that
satisfy certain constraints.</p>
<p class="block-text-western">Note: In this section we have collected
examples that demonstrate the tools and techniques that perform
specific tasks. Use these examples as starting points to achieve what
you might need.</p>
<h2 class="western"><a name="how-can-i-filter-vcf-files"></a>100.1
How can I filter VCF files?</h2>
<p class="first-paragraph">There is a cottage industry of tools that
may be used to manipulate VCF files, from simple pattern matching
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">grep</font></font>,
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>
that can intersect files by coordinate, to tools that seem to
implement a &ldquo;mini-programming languages&rdquo; like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font></p>
<p>Recommended software that can be used for filtering and
manipulating VCF files include</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a></font>
	- utilities for variant calling and manipulating VCFs and BCFs</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://snpeff.sourceforge.net/SnpSift.html">SnpSift</a></font>
	- a toolbox that allows you to filter and manipulate annotated
	files.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></font>
	- a powerful toolset for genome arithmetic</p>
</ul>
<p class="first-paragraph">Most <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>
commands accept VCF, bgzipped VCF and BCF. However, some commands
(e.g.:<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query</font></font>) require a <font color="#4f81bd"><a href="http://www.htslib.org/doc/tabix.html">tabix</a></font>
indexed and hence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bgzip</font></font>
compressed file.</p>
<p>SnpSift filtering tool is especially useful if you have a &lsquo;SnpEff&rsquo;
annotated VCF file.</p>
<h2 class="western"><a name="how-do-i-get-the-example-data"></a>100.2
How do I get the example data?</h2>
<p class="first-paragraph">Download the VCF file and its index used
in the examples with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/variant/subset_hg19.vcf.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://data.biostarhandbook.com/variant/subset_hg19.vcf.gz.tbi</font></font></p>
<p class="first-paragraph">This file contains variants in chromosome
19:400kb-500kb.</p>
<h2 class="western"><a name="how-do-i-extract-information-from-vcf-in-a-custom-format"></a>
100.3 How do I extract information from VCF in a custom format?</h2>
<p class="first-paragraph">Bcftools <font face="Consolas, serif"><font size="2" style="font-size: 11pt">query</font></font>
command with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-f</font></font>
flag can be used to extract fields from VCF or BCF files. To extract
just the CHROM, POS, REF and ALT fields:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -f '%CHROM %POS %REF %ALT\n' subset_hg19.vcf.gz | head -3</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
400410 CA C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
400666 G C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
400742 C T</font></font></p>
<h2 class="western"><a name="how-do-i-list-all-samples-stored-in-a-vcf-file"></a>
100.4 How do I list all samples stored in a VCF file?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -l subset_hg19.vcf.gz</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00115</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00116</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00117</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00118</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00119</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00120</font></font></p>
<h2 class="western"><a name="how-do-i-extract-all-variants-from-a-particular-region"></a>
100.5 How do I extract all variants from a particular region?</h2>
<p class="first-paragraph">To specify the region, use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-r</font></font>
option with region specified in the format
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr|chr:pos|chr:from-to|chr:from-[,&hellip;]</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -r '19:400300-400800' -f '%CHROM\t%POS\t%REF\t%ALT\n'
subset_hg19.vcf.gz | head -3</font></font></p>
<p class="first-paragraph">This will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400410  CA  C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400666  G   C</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400742  C   T</font></font></p>
<p class="first-paragraph">If you want to output the results as a VCF
file, use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">filter</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">view</font></font>
command instead of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">query</font></font>.
Multiple regions can be listed together in a bed file and can be
loaded with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-R</font></font>
flag to extract variants in those intervals.</p>
<h2 class="western"><a name="how-do-i-extract-variants-excluding-a-specific-region"></a>
100.6 How do I extract variants excluding a specific region?</h2>
<p class="first-paragraph">To exclude a specific region, prefix it
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">^</font></font>.
The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-r</font></font>
option cannot be used in this case instead use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-t</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -t ^'19:400300-400800' -f '%CHROM\t%POS\t%REF\t%ALT\n'
subset_hg19.vcf.gz | head -3</font></font></p>
<p class="first-paragraph">To produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400819  C   G</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400908  G   T</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400926  C   T</font></font></p>
<p class="first-paragraph">To exclude multiple regions, compile them
into a bed file and use with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-T</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">^</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -T ^exclude.bed -f '%CHROM\t%POS\t%REF\t%ALT\n'
subset_hg19.vcf.gz | head -3</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400742  C   T</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400819  C   G</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 401013  G   A</font></font></p>
<p class="first-paragraph">where exclude.bed looks like this</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#cat
exclude.bed</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400300  400700</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400900  401000</font></font></p>
<h2 class="western"><a name="how-do-i-get-variants-present-in-all-samples"></a>
100.7 How do I get variants present in all samples?</h2>
<p class="first-paragraph">We can exclude sites with one or more
missing genotype and one or more homozygous reference call to extract
variant sites that are present in all samples. Bcftools <font face="Consolas, serif"><font size="2" style="font-size: 11pt">view</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g,--genotype</font></font>
option helps to achieve this.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -e 'GT= &quot;.&quot; | GT=&quot;0|0&quot;' subset_hg19.vcf.gz
|bcftools query -f '%POS[\t%GT\t]\n'  | head -3</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">402556
 0|1     0|1     1|1     1|0     0|1     1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">402707
 0|1     0|1     1|1     1|0     0|1     1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">402723
 0|1     0|1     1|1     1|0     0|1     1|1</font></font></p>
<p class="first-paragraph">The output of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view</font></font> command is piped into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">query</font></font>
command to print it in a user-friendly format.</p>
<h2 class="western"><a name="how-do-i-select-indels-only"></a>100.8
How do I select INDELs only?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">query</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-v,
--types</font></font> option can be used to select a variant type of
interest. Here a site is chosen if any of the ALT alleles is of the
type requested. In this case, types are determined by comparing the
REF and ALT alleles in the VCF record not INFO tags like INFO/INDEL
or INFO/SNP.</p>
<p>Another way to extract a specific variant is to use the
information given in INFO tag. For this we can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-i,
--include</font></font> option along with an expression.</p>
<p>In our example, we do not have any INDEL sites.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -v indels subset_hg19.vcf.gz | bcftools query -f '%POS\t%TYPE\n'
 |wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">141</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -i 'TYPE=&quot;indel&quot;' subset_hg19.vcf.gz | bcftools query
-f '%POS\t%TYPE\n'  |wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">140</font></font></p>
<p class="first-paragraph">The difference in the two commands is
caused by the position 485135 as marked in the INFO/TYPE field as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">485135
SNP,INDEL</font></font></p>
<p class="first-paragraph">Hence when searched with an expression for
exact match in the INFO field, this variant won&rsquo;t be included.</p>
<p>Multiple variant types can be selected with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-v</font></font>
by listing them separated by a comma. To exclude a variant type use
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-V,--exclude-types</font></font>.</p>
<h2 class="western"><a name="how-do-i-extract-per-sample-information-from-a-multi-sample-vcf-file"></a>
100.9 How do I extract per-sample information from a multi-sample VCF
file?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
which is the 9th column in VCF file specifies how to interpret the
colon separated values of each sample given in 10th column onwards.
These <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FORMAT</font></font>
tags can be extracted using the square brackets [] operator, which
loops over all samples.</p>
<h2 class="western"><a name="how-to-print-genotype-gt-of-all-samples-at-each-position."></a>
100.10 How to print genotype (GT) of all samples at each position.</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -f '%CHROM\t%POS[\t%GT\t]\n' subset_hg19.vcf.gz | head -3</font></font></p>
<p class="first-paragraph">Produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400410  0|0     0|0     0|0     0|0     0|0     0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400666  1|0     0|1     0|1     0|0     0|0     1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">19
 400742  0|0     0|0     0|0     0|0     0|0     0|0</font></font></p>
<p class="first-paragraph">If you want to print header too, use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-H</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -H -f '%CHROM\t%POS[\t%GT\t]\n' subset_hg19.vcf.gz | head -3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<h2 class="western"><a name="how-do-i-select-specific-samples"></a>100.11
How do I select specific samples?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">view</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-s,
--samples</font></font> option can be used to subset a VCF file based
on samples.</p>
<p>To select samples <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00115</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00118</font></font>
from the sample vcf file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -s HG00115,HG00118 subset_hg19.vcf.gz | bcftools query -H -f
'%POS[\t%GT]\n' | head -4</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
[1]POS    [2]HG00115:GT   [3]HG00118:GT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400410
 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400666
 1|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400742
 0|0 0|0</font></font></p>
<h2 class="western"><a name="how-do-i-exclude-specific-samples"></a>100.12
How do I exclude specific samples?</h2>
<p class="first-paragraph">To exclude samples <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00115</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HG00118</font></font>
from VCF file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -s ^HG00115,HG00118 subset_hg19.vcf.gz  | bcftools query -H -f
'%POS[\t%GT]\n' | head -4</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
[1]POS    [2]HG00116:GT   [3]HG00117:GT   [4]HG00119:GT  
[5]HG00120:GT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400410
 0|0 0|0 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400666
 0|1 0|1 0|0 1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400742
 0|0 0|0 0|0 0|0</font></font></p>
<h2 class="western"><a name="how-do-i-get-variants-for-which-allele-count-is-above-a-specific-value"></a>
100.13 How do I get variants for which allele count is above a
specific value?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">view</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-c,
--min-ac</font></font> helps to set the minimum allele count of sites
to be printed.</p>
<p>The command below removes sites where allele count (AC tag) &lt;
5.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -c 5 subset_hg19.vcf.gz | grep -v &quot;#&quot; | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
185</font></font></p>
<h2 class="western"><a name="how-do-i-select-sites-where-all-samples-have-homozygous-genotype"></a>
100.14 How do I select sites where all samples have homozygous
genotype?</h2>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">view</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g,--genotype</font></font>
helps to include or exclude one or more homozygous (hom),
heterozygous (het) or missing (miss) genotypes.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
view -g ^het subset_hg19.vcf.gz | bcftools view -g ^miss | bcftools
query -f '%POS[\t%GT\t]\n'| head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
# 3891</font></font></p>
<p class="first-paragraph">In the command above <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g
^het</font></font> and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-g
^miss</font></font> exclude sites with one or more heterozygous
variant or a missing genotype.</p>
<h2 class="western"><a name="how-do-i-select-sites-where-n-samples-have-heterozygous-variants"></a>
100.15 How do I select sites where &lsquo;n&rsquo; samples have
heterozygous variants?</h2>
<p class="first-paragraph">SnpSift <font face="Consolas, serif"><font size="2" style="font-size: 11pt">filter</font></font>
command below selects sites where 3 or more samples have heterozygous
genotypes.</p>
<p>In the command below snpsift is an alias to &ldquo;java -jar
~/bin/SnpSift.jar&rdquo;</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias
snpsift=&quot;java -jar ~/bin/SnpSift.jar&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpsift
filter  &quot;countHet() &gt;=3&quot; subset_hg19.vcf.gz |snpsift
extractFields - &quot;POS&quot; &quot;GEN[*].GT&quot; |  head -4</font></font></p>
<p class="first-paragraph">will produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#POS
   GEN[*].GT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400666
 1|0 0|1 0|1 0|0 0|0 1|1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400941
 0|1 1|0 0|1 1|0 0|0 0|0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">401235
 0|0 0|0 0|1 0|0 0|1 1|0</font></font></p>
<h2 class="western"><a name="how-do-i-select-variant-sites-based-on-quality-and-read-depth"></a>
100.16 How do I select variant sites based on quality and read depth?</h2>
<p class="first-paragraph">If we want to extract sites with QUAL &gt;50
and read depth &gt; 5000, then these conditions can be included in an
expression like this.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">QUAL&gt;50
&amp;&amp; DP&gt;5000</font></font></p>
<p>This can be used with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-i,--include</font></font>
to extract sites satisfying the expression. The command would then be</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
query -i 'QUAL&gt;50 &amp;&amp; DP&gt;5000' -f '%POS\t%QUAL\t%DP\n'
subset_hg19.vcf.gz |  head -3</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400410
 100 7773</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400666
 100 8445</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">400742
 100 15699</font></font></p>
<h2 class="western"><a name="how-do-i-merge-multiple-vcfbcf-files"></a>
100.17 How do I merge multiple VCF/BCF files?</h2>
<p class="first-paragraph">If we had multiple VCF files, then we can
merge them into a single multi-sample VCF with bcftools <font face="Consolas, serif"><font size="2" style="font-size: 11pt">merge</font></font>
command.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
merge -l samplelist &gt; multi-sample.vcf</font></font></p>
<p class="first-paragraph">where sample list is</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">#cat
samplelist</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sample1.vcf.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sample2.vcf.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sample3.vcf.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
....</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<h2 class="western"><a name="how-do-i-find-variant-sites-present-in-all-vcf-files"></a>
100.18 How do I find variant sites present in all vcf files?</h2>
<p class="first-paragraph">If we have 3 samples and a vcf file for
each of them, then to find sites that are common to all 3</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
isec -p outdir -n=3 sample1.vcf.gz sample2.vcf.gz sample3.vcf.gz</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">n=3</font></font>
in the command specifies to output sites present in 3 files. The
output files will be in a directory named &lsquo;outdir&rsquo;.</p>
<h2 class="western"><a name="how-do-i-extract-variant-sites-present-in-at-least-2-files"></a>
100.19 How do I extract variant sites present in at least 2 files?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
isec -p outdir -n+2 sample1.vcf.gz sample2.vcf.gz sample3.vcf.gz</font></font></p>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">n+2</font></font>
in the command specifies to output sites present in 2 or more files.
The output files will be in a directory named &lsquo;outdir&rsquo;.</p>
<h2 class="western"><a name="how-do-i-find-variants-unique-to-one-sample-but-absent-in-all-other-samples"></a>
100.20 How do I find variants unique to one sample but absent in all
other samples?</h2>
<p class="first-paragraph">Bcftools <font face="Consolas, serif"><font size="2" style="font-size: 11pt">isec</font></font>
command with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-C,--complent</font></font>
option output positions present only in the first file but missing in
others.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools
isec -p outdir -C sample1.vcf.gz sample2.vcf.gz sample3.vcf.gz</font></font></p>
<p class="first-paragraph">The above command will produce sites
unique to sample1.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="variant-annotation-and-effect-prediction"></a>
101 Variant annotation and effect prediction</h1>
<h2 class="western"><a name="what-is-variant-annotation"></a>101.1
What is variant annotation?</h2>
<p class="first-paragraph">Variant annotation means predicting the
effects of genetic variants (SNPs, insertions, deletions, copy number
variations (CNV) or structural variations (SV)) on the function of
genes, transcripts, and protein sequence, as well as regulatory
regions.</p>
<p>Variant effect annotators typically require additional information
from other data sources. The predicted results will depend on the
type of other data that is known about the organism or genome under
study. The starting points for a variant annotator are the variants
(SNPs, insertions, deletions, and MNPs) usually in VCF format.</p>
<p>The results of variant annotation may be reported in text, HTML or
pdf format along with an annotated VCF file that now contains
additional information on each variant. The text/HTML reports contain
substantially more information than the VCF files as this latter
format limits the type of content that may be added.</p>
<h2 class="western"><a name="are-are-consequences-of-variant-effects"></a>
101.2 Are are consequences of &ldquo;variant effects&rdquo;?</h2>
<p class="first-paragraph">Whereas the way the word is used seems to
imply a &ldquo;universal&rdquo; effect, keep in mind that all effects
are &ldquo;relative&rdquo; to the information that is currently
present in a database.</p>
<p>The same way as in an alignment an insertion in one of the
sequences could be reported as being a deletion in the other - it all
depends on what we choose to be the reference, any effect predicted
by an annotator refers to the information already known and stored in
the database.</p>
<h2 class="western"><a name="what-kind-of-effects-can-variant-effect-predictors-find"></a>
101.3 What kind of effects can variant effect predictors find?</h2>
<p class="first-paragraph">The variant effects (sometimes called
consequences) fall into multiple categories:</p>
<ul>
	<li/>
<p class="compact">Correlate the location of the variant with
	other genomic annotations (e.g., upstream of a transcript, in the
	coding sequence, in non-coding RNA, in regulatory regions)</p>
	<li/>
<p class="compact">List which genes and transcripts are
	affected by the variant.</p>
	<li/>
<p class="compact">Determine consequence of the variant on
	protein sequence (e.g., stop_gained, missense, stop_lost,
	frameshift, other property changes)</p>
	<li/>
<p class="compact">Match known variants that may have been
	found in different projects such as the 1000 Genomes Project.</p>
</ul>
<p class="first-paragraph">Several terms are used to express the
effect (also called consequence) of variants. The actual definition
of each word is part of the <font color="#4f81bd"><a href="#ch040.xhtml#sequence-ontology">Sequence
Ontology</a></font> that we covered in the first chapters. Also, each
effect is associated with a somewhat subjective qualifier (LOW,
MODERATE, HIGH) that attempts to characterize the severity and impact
of the impact on a biological process.</p>
<p>Here are a few potential &ldquo;effects&rdquo;:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">stop_gained</font></font>:
	A sequence variant whereby at least one base of a codon is changed,
	resulting in a premature stop codon, leading to a shortened
	transcript, SO:0001587, HIGH</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">inframe_insertion</font></font>:
	An inframe nonsynonymous variant that inserts bases into in the
	coding sequence, SO:0001821, MODERATE</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">missense_variant</font></font>:
	A sequence variant, that changes one or more nucleotides, resulting
	in a different amino acid sequence but where the length is
	preserved, SO:0001583, MODERATE</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">protein_altering_variant</font></font>:
	A sequence_variant which is predicted to change the protein encoded
	in the coding sequence, SO:0001818, MODERATE</p>
</ul>
<p class="first-paragraph">Variant Effect Predictor (VEP) from
Ensembl maintains a <font color="#4f81bd"><a href="http://useast.ensembl.org/info/genome/variation/predicted_data.html#consequences">list
of consequence terms</a></font> and their meaning when used in VEP.</p>
<h2 class="western"><a name="what-kind-of-variant-annotators-can-i-use"></a>
101.4 What kind of variant annotators can I use?</h2>
<p class="first-paragraph">Variant effect annotation is a burgeoning
field with several tools and techniques:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.ensembl.org/info/docs/tools/vep/index.html">VEP:
	Variant Effect Predictor</a></font> A web-based tool from Ensembl
	that integrates with the Ensembl data.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://snpeff.sourceforge.net/index.html">snpEff</a></font>:
	Command line genetic variant annotation and effect prediction
	toolbox.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://annovar.openbioinformatics.org/en/latest/">AnnoVar</a></font>:
	a software tool to functionally annotate genetic variants detected
	from diverse genomes.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.yandell-lab.org/software/vaast.html">VAAST
	2</a></font>: Variant annotation, analysis, and search tool.</p>
</ul>
<h2 class="western"><a name="how-do-i-use-snpeff"></a>101.5 How do I
use snpEff?</h2>
<p class="first-paragraph">We will compare two strains of the Ebola
virus, and we will attempt to determine which changes are likely to
impact the function of the genome.</p>
<p>For all annotation tools, we need to know which genome build the
annotation database is using. We will need to align our data against
the same genome build.</p>
<p>Find what databases does snpEff know about the present.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff
databases &gt; listing.txt</font></font></p>
<p class="first-paragraph">Find the information on ebola</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
listing.txt | grep ebola </font></font>
</p>
<p class="first-paragraph">Produces the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ebola_zaire
  Ebola Zaire Virus    KJ660346.1  
OK</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">http://downloads.sourceforge.net/project/snpeff/databases/v4_1/snpEff_v4_1_ebola_zaire.zip</font></font></p>
<p class="first-paragraph">We need to download that database to use
it</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff
download ebola_zaire</font></font></p>
<p class="first-paragraph">We can visualize the content of a database
for example to check the chromosomal naming and have an understanding
of what the database contains.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff
dump ebola_zaire | more</font></font></p>
<p class="first-paragraph">Note the version number of the accession
number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">KJ660346</font></font>.
As a rule, we have to ensure that our alignments are performed using
the same build. Using this accession number means that our effects
will be in reference to this genomic build.</p>
<p>Get the script that we used before:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O http://data.biostarhandbook.com/variant/find-variants.sh</font></font></p>
<p class="first-paragraph">The quality of a script becomes evident
when we need to modify it to run on a different data set. Fewer the
modifications we need to make, the less likely it is that we will
introduce an error. Ideally, we change just a small section or a
parameter, and the whole script should run with no other changes.
Think about how much time and effort this saves you. Within seconds
you can re-run the entire analysis on a different genome (or with
different SRR data).</p>
<p>We also need to select SRR runs that we wish to study with the
same commands that we demonstrated in the chapter <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Variant
calling Ebola data</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://data.biostarhandbook.com/sra/ebola-runinfo.csv &gt;
runinfo.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.txt | grep &quot;04-14&quot; | cut -f 1 -d ',' | grep SRR |
head -5 &gt; samples.txt</font></font></p>
<p class="first-paragraph">The file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samples.txt</font></font>
contains five sample run ids:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972920</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1972921</font></font></p>
<p class="first-paragraph">Run the script:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
find-variants.sh KJ660346 samples.txt</font></font></p>
<p class="first-paragraph">And finally to produce the annotated
results:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff
ebola_zaire combined.vcf &gt; annotated.vcf</font></font></p>
<p class="first-paragraph">Now to show these annotations we also need
to build the custom genome in IGV.</p>
<p><img src="0000_html_b3abc95a8a7327d5.png" name="Image149" align="bottom" hspace="12" width="560" height="159" border="0"/>
</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff</font></font>
also generates report files <font face="Consolas, serif"><font size="2" style="font-size: 11pt">snpEff_summary.html</font></font>
The file is an HTML report with several entries. For example &ldquo;Codon
changes&rdquo; shows:</p>
<p><img src="0000_html_643df9d452a8ae33.png" name="Image150" align="bottom" hspace="12" width="560" height="241" border="0"/>
</p>
<h2 class="western"><a name="can-i-build-custom-annotations-for-snpeff"></a>
101.6 Can I build custom annotations for snpEff?</h2>
<p class="first-paragraph">Custom annotations can be built for snpEff
from GeneBank and other files. See the <font color="#4f81bd"><a href="http://snpeff.sourceforge.net/">Install
snpEff</a></font> page for details.</p>
<h2 class="western"><a name="how-do-i-use-the-ensemble-variant-effect-predictor-vep"></a>
101.7 How do I use the Ensemble Variant Effect Predictor (VEP)</h2>
<p class="first-paragraph">Get our subset of variants obtained from
1000 genomes project called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">subset_hg19.vcf.gz</font></font>
from <font color="#4f81bd"><a href="http://data.biostarhandbook.com/variant/">http://data.biostarhandbook.com/variant/</a></font>
or from the command line as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O http://data.biostarhandbook.com/variant/subset_hg19.vcf.gz</font></font></p>
<p class="first-paragraph">Visit the VEP website at
<font color="#4f81bd"><a href="http://useast.ensembl.org/Tools/VEP">http://useast.ensembl.org/Tools/VEP</a></font>.</p>
<p>Just as with the Ebola data, it is essential to use the same
genome build for predictions as for alignment. For human genomes, the
process is more error prone since, unlike the Ebola virus, the
chromosome names stay the same across builds. Hence there is no way
to tell from the name of the sequence what genomic build it came
from.</p>
<p>Uploading the results and running VEP produces a report with quite
a bit of detail.</p>
<p><img src="0000_html_139b9ab4d9464fbe.png" name="Image151" align="bottom" hspace="12" width="560" height="187" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="why-is-variant-calling-challenging"></a>102
Why is variant calling challenging?</h1>
<p class="first-paragraph">As you will find there is no shortage of
scientific papers comparing different variant callers. What is
however never adequately explained is why is this task challenging
and how are various tools trying to handle the problems.</p>
<h2 class="western"><a name="why-is-it-so-difficult-to-call-variants"></a>
102.1 Why is it so difficult to call variants?</h2>
<p class="first-paragraph">Most genomics-related DNA sequences have
several properties that confound intuition as they require thinking
about context that human mind does not typically need to solve:</p>
<ol>
	<li/>
<p class="compact">The DNA sequences can be exceedingly long
	(millions of bases)</p>
	<li/>
<p class="compact">Just four building blocks A/T/G/C make up
	any DNA sequence. These, even when distributed randomly could
	produce a variety of patterns just by sheer chance.</p>
	<li/>
<p class="compact">The DNA is only partially affected by
	randomness - usually, most of it has a function.</p>
	<li/>
<p class="compact">Depending on the organisms various segments
	of the DNA are affected by different rates of randomness.</p>
	<li/>
<p class="compact">Experimental protocols break long pieces of
	DNA into tiny ones that we then attempt to match back to a
	reference. Ideally, we hope that it still matches the most closely
	to the original location - because only then will our match be a
	variant of the site of interest.</p>
</ol>
<p class="first-paragraph">If our fragment were to match better
somewhere else than the original location our sequence would align
there instead of its &ldquo;correct&rdquo; position.</p>
<p>We call this a &ldquo;misalignment,&rdquo; though as we mentioned
before the word is usually a misrepresentation of what happens. It is
not actually &ldquo;misaligning.&rdquo; From algorithmic perspective
the alignment is correct, and it does what we instructed it to do.</p>
<p>As you can see the deck is stacked quite a bit against us - it is
surprising that variant calling works as well as it does!</p>
<h2 class="western"><a name="how-do-variant-calls-overcome-these-challenges"></a>
102.2 How do variant calls overcome these challenges?</h2>
<p class="first-paragraph">Nothing gets scientist&rsquo;s minds going
better than the need to tackle a seemingly impossible problem.
Unexpectedly, the very fact that we chop sequences into little pieces
has an unexpected benefit. If we only had a single sequence at hand
then, once a &ldquo;misalignment&rdquo; took place there would be
little extra information to recognize and correct it.</p>
<p>When the sequences are short, then each is affected slightly
differently - and as it turns out - it is possible to attempt to
recover the &ldquo;correct&rdquo; alignment from all the different
types of errors that affect each shorter read differently. In
essence, it is feasible to find another variation that would better
explain and reconcile the differences present in each read.</p>
<p>This process has different implementations: it may be called
<i>realignment</i> or <i>probabilistic alignment</i>.</p>
<h2 class="western"><a name="how-to-explore-the-effect-of-mutations-on-variant-callers"></a>
102.3 How to explore the effect of mutations on variant callers?</h2>
<p class="first-paragraph">We have written a script called
<font color="#4f81bd"><a href="http://data.biostarhandbook.com/variant/compare-variant-callers.sh">compare-variant-callers.sh</a></font>
that allows us to generate variants with multiple methods. The script
assumes that you have installed <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bcftools</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">freebayes</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GATK</font></font>
- if you only have a subset of these installed, you will need to edit
the script and comment out the sections that run programs that your
computer does not have installed.</p>
<p>Obtain the code:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O http://data.biostarhandbook.com/variant/compare-variant-callers.sh</font></font></p>
<p class="first-paragraph">Now run it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
compare-variant-callers.sh </font></font>
</p>
<p class="first-paragraph">It should print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Obtain data for AF086833 and save it as refs/REFERENCE.fa.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Index refs/REFERENCE.fa with several tools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Generating simulated reads from: GENOME.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Aligning the reads into: align.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Genome alignment into: global.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Calling variants with samtools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Calling variants with freebayes.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Calling variants with GATK.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
Run completed successfully.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">***
VCF files: samtools.vcf freebayes.vcf gatk.vcf</font></font></p>
<p class="first-paragraph">It ran three variant callers and created
quite a few files, but the ones we are interested in are:</p>
<ol>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools.vcf</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">freebayes.vcf</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gatk.vcf</font></font></p>
</ol>
<p class="first-paragraph">Since we have not yet introduced any
mutations, none of these files should contain any mutations.</p>
<p>Let&rsquo;s change that.</p>
<p>The file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GENOME.fa</font></font>
contains the genome from which the sequencing samples are generated.
With a method of your choice modify and save this file. We will use
the command line tool <font face="Consolas, serif"><font size="2" style="font-size: 11pt">biosed</font></font>
to create a SNP replacing an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">A</font></font>
with a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>.
We have also selected this variation to be at position 2000 to help
us more quickly identify whether the SNP caller indicates it
correctly.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
A shortcut to the reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/REFERENCE.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This is the genome file that gets mutated from the
reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GENOME=GENOME.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Replace one sequence with another (the only difference is the A -&gt;
T at the end  </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | biosed -filter -target AGGGTGGACAACAGAAGAACA -replace
AGGGTGGACAACAGAAGAACT &gt; $GENOME</font></font></p>
<p class="first-paragraph">now rerun the comparison script:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
compare-variant-callers.sh</font></font></p>
<p class="first-paragraph">Here is what we get. All variant callers
can reproduce and find this SNP:</p>
<p><img src="0000_html_7fb6894d8c588517.png" name="Image152" align="bottom" hspace="12" width="560" height="163" border="0"/>
</p>
<h2 class="western"><a name="what-happens-when-the-variation-is-more-complicated"></a>
102.4 What happens when the variation is more complicated?</h2>
<p class="first-paragraph">Let us cut two bases from the end of the
sequence; this is now a two base INDEL</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | biosed -filter -target AGGGTGGACAACAGAAGAACA -replace
AGGGTGGACAACAGAAGAA &gt; $GENOME</font></font></p>
<p class="first-paragraph">now rerun the comparison script:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
compare-variant-callers.sh</font></font></p>
<p class="first-paragraph">What do we get now? Each of the variant
callers produces a DIFFERENT variant call.</p>
<p><img src="0000_html_1bdcaa64b2687a6c.png" name="Image153" align="bottom" hspace="12" width="560" height="147" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="rna-seq-by-example"></a>103 RNA-SEQ by
Example</h1>
<p class="first-paragraph">The content of this chapter has been moved
into a separate, standalone book:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html">RNA-Seq
	by example</a></font></p>
</ul>
<p class="first-paragraph"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html"></a><img src="0000_html_dd723eb74742c147.png" name="Image154" align="bottom" hspace="12" width="560" height="145" border="0"/>
<a href="https://www.biostarhandbook.com/books/rnaseq/index.html"></a></p>
<p>Please visit <font color="#4f81bd"><a href="https://www.biostarhandbook.com/books/rnaseq/index.html">the
new book</a></font> for all RNA-Seq related analysis content.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-chip-seq-analysis"></a>104
Introduction to ChIP-Seq analysis</h1>
<h2 class="western"><a name="what-is-chip-seq"></a>104.1 What is
ChIP-Seq?</h2>
<p class="first-paragraph">ChIP-Seq stands for chromatin
immunoprecipitation followed by sequencing. In a nutshell, the
process consists of a laboratory protocol (abbreviated as ChIP) by
the end of which the full DNA content of a cell is reduced to a much
smaller subset of it. This subset of DNA is then sequenced
(abbreviated as Seq) and is mapped against a known reference genome.
Since you are now obtaining data from a potentially variable subset
of the entire genome, it poses unique challenges to data
interpretation, hence a whole subfield of data analytics has grown
out around it.</p>
<p>Whereas the name appears to imply that the analysis methods
&ldquo;require&rdquo; that the data was obtained via chromatin
immunoprecipitation (ChIP) this is not necessarily so. Many other
laboratory protocols where the net effect is that only a subset of
the total DNA is selected for can be processed with ChIP-Seq methods.</p>
<h2 class="western"><a name="what-are-the-challenges-of-chip-seq"></a>
104.2 What are the challenges of ChIP-Seq?</h2>
<p class="first-paragraph">Perhaps one of the most significant
problem of this entire subfield is the insufficiently well-defined
terminology. See for example the section &ldquo;<b>What the heck is a
peak</b>&rdquo; later on. Also, more so than in other analysis
methods, there a &ldquo;subjectivity creep.&rdquo; Seemingly simple
decisions are made all the time - that in the end seem to accumulate
and slowly grow out of control as the analysis proceeds.</p>
<p>Examples of these insufficiently defined terms are for instance
the concepts of &ldquo;upstream&rdquo; and &ldquo;downstream&rdquo;
regions which typically refer to the genomic intervals that lie ahead
of and after a particular genomic coordinate.</p>
<p>When referring to these ranges, you will need to associate a
finite length to them: how far out do you still call a region to be
&ldquo;upstream.&rdquo; The choice of this distance may, in turn,
significantly alter the outcome of the analysis. You will likely
obtain different results when the upstream region is defined to be
500bp, 1000bp or 5000 kb long.</p>
<p>Further complicating the picture is that different genomes have
varying complexities; in some genomes, the genes are densely packed,
other vast tracts are seemingly nonfunctional. Hence there is no
&ldquo;objective&rdquo; measure by which a &ldquo;reliable&rdquo;
length for an upstream or downstream region could be determined from.
The size of the &ldquo;upstream&rdquo; region is little more than a
handwaving approximation.</p>
<h2 class="western"><a name="what-the-heck-is-a-peak"></a>104.3 What
the heck is a peak?</h2>
<p class="first-paragraph">You will see how ChIP-Seq analysis is all
about peaks - a term, that in our opinion is adopted widely yet lacks
any exact definition. Somewhat ironically speaking, perhaps the best
description of a peak is the following:</p>
<p class="block-text-western">A &ldquo;peak&rdquo; is what a
&ldquo;peak-finder&rdquo; finds.</p>
<p class="first-paragraph">What makes the problem complicated is that
we all think that we know what a &ldquo;peak&rdquo; is (or should
be). A peak is a &ldquo;shape&rdquo; that rises sharply to a maximum
then quickly drops back to zero. In that ideal case then the peak
consists of the <b>coordinate</b> of the maxima on the genome (an
integer number) and a <b>peak height</b>.</p>
<p>As it happens, in reality, most of your signals may not follow
such well-behaved patterns - the peaks may be a lot broader,</p>
<p>Now if the peak is wide you may need to compute the <b>peak width</b>
that will indicate the range where the signal is present. Then once
you have a broader peak representing it with its center may be
counterproductive as it is not clear that the summit will be
symmetrical around it. Instead, you&rsquo;ll have to keep track of
the <b>right</b> and <b>left</b> borders of it.</p>
<p>Just having a signal, of course, is insufficient, we may want to
ensure that the signal rises <b>above the background</b> of the noise
levels. Also, most studies are interested in how cells respond to
stimuli that in turn will alter the signal. So the changes may occur
by the peak rising in a <b>different location</b> or via <b>changed
peak height</b> often denoted with the term <b>differential peak
expression</b>.</p>
<p>You can see how a seemingly innocuous word: &ldquo;peak&rdquo; may
require managing more information than you&rsquo;d expect.</p>
<h2 class="western"><a name="how-are-rna-seq-studies-different-from-chip-seq-studies"></a>
104.4 How are RNA-Seq studies different from ChIP-Seq studies?</h2>
<p class="first-paragraph">Whereas both approaches appear to operate
under similar constraints and count reads over intervals, there are
several significant differences:</p>
<ol>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">The DNA
	fragments coming from a ChIP-Seq study are much shorter than a
	transcript studied in RNA-Seq analysis. Instead of counting
	measurements distributed over transcripts a single read may cover
	most of the fragment (or may even be longer than the original DNA
	fragment)</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">The DNA
	fragments of a ChIP-Seq experiment are less localized than
	transcripts. The cell machinery accurately locates and transcribes
	DNA always starting from the same start locations. The ChIP-Seq
	protocol is a whole lot more inaccurate. You need to apply a
	peak-calling process to establish the origin of the fragments from
	incorrect data.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">ChIP-Seq
	data produces a more significant number of false positives. In a
	typical RNA-Seq experiment, only transcripts are isolated
	identically, and we can be fairly sure that an observed sequence did
	exist as RNA. In contrast, ChIP-Seq protocols strongly depend on the
	properties of the reagents, selectivity, and specificity of the
	binding process, and these properties will vary across different
	proteins and across the genome.</p>
</ol>
<p class="first-paragraph">As a matter of fact, for ChIP seq results
we need to compare to control (baseline) to verify that the protocol
worked <b>at all</b>. Whereas in RNA-Seq we can immediately see the
success of it by having reads cover transcript, in ChIP-Seq study no
such certainty exists. Also when comparing peaks across conditions
each peak has to first pass the comparison to a background, and then
only those peaks that pass the first stage are compared to one
another in a second stage. Since each statistical test is fraught
with uncertainties, differential peak expression studies will often
exhibit compounding errors.</p>
<h2 class="western"><a name="what-will-the-data-look-like"></a>104.5
What will the data look like?</h2>
<p class="first-paragraph">For the record let&rsquo;s enumerate the
possible observations for a given genomic location.</p>
<ol>
	<li/>
<p class="compact">No signal of any kind.</p>
	<li/>
<p class="compact">A signal that shows no peak structure and
	appears to be noise.</p>
	<li/>
<p class="compact">A &ldquo;peak like&rdquo; signal that is
	reported to be under a background level.</p>
	<li/>
<p class="compact">A &ldquo;peak like&rdquo; signal that is
	reported to be over a background level.</p>
	<li/>
<p class="compact">A &ldquo;peak like&rdquo; signal that
	exhibits differential expression across samples.</p>
</ol>
<p class="first-paragraph">Many (most) peak finders will start
reporting data only from level 4 on - leaving you with many
head-scratching moments of the form: &ldquo;Why is this perfectly
looking peak not reported?&rdquo;. It all makes sense once you
recognize that &ldquo;peak finding&rdquo; as the field currently
defines it, combines finding the peaks and comparing them to a
background. In addition, the statistical methods employed to
determine these background levels have very high levels of
uncertainty.</p>
<h2 class="western"><a name="what-does-a-peak-represent"></a>104.6
What does a &ldquo;peak&rdquo; represent?</h2>
<p class="first-paragraph">Most ChIP-Seq guides start with explaining
how ChIP-seq works, and how it allows us to identify where proteins
were bound to DNA. We believe that is far more important to
understand what your data stands for in a more straightforward way
first, then extend that to the laboratory process.</p>
<p>Imagine that we have DNA from a cell that looks like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGTGATACCGGTCTAAGCCTCGCCTCTTCCACAGGGGAAACTAGGTGGCC</font></font></p>
<p class="first-paragraph">Now let&rsquo;s say some locations are
&ldquo;covered,&rdquo; say &ldquo;protected&rdquo; from an
interaction. Let&rsquo;s visualize those locations via the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=</font></font>
symbols.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">         <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=========
            
==========</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGTGATACCGGTCTAAGCCTCGCCTCTTCCACAGGGGAAACTAGGTGGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        =========              ==========              </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font>
</p>
<p class="first-paragraph">Suppose that can wash away all unprotected
regions. If we did that then sequenced the remaining DNA we&rsquo;d
be left with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
GGTCTAAGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
GGGGAAACTA</font></font></p>
<p class="first-paragraph">The first number indicates how many of
those sequences were found. So far so good. Now let&rsquo;s take
another cell from the same population. Suppose that the DNA for this
new cell is also now covered, but in only one location:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">         <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=========
              </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGTGATACCGGTCTAAGCCTCGCCTCTTCCACAGGGGAAACTAGGTGGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        =========               </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font>
</p>
<p class="first-paragraph">After washing away the uncovered regions
we add the outcome to the prior results, and we have:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
GGTCTAAGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
GGGGAAACTA</font></font></p>
<p class="first-paragraph">Finally, let&rsquo;s assume we have DNA
from a third cell, this one has a protein bound in the second
location.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">                   
            
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">==========</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGTGATACCGGTCTAAGCCTCGCCTCTTCCACAGGGGAAACTAGGTGGCC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                ========== </font></font>
</p>
<p class="first-paragraph">Thus our cumulative data that we obtain
after washing away unprotected regions from three cells would be.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
GGTCTAAGC</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
GGGGAAACTA </font></font>
</p>
<p class="first-paragraph">In a ChIP-Seq experiment, instead of
sequencing the entire genomes from three different cells we only
sequence relatively short pieces of DNA (in reality the sequences are
longer than these example sequences above but still radically shorter
than say a transcript). The alignments that we generate will pass
through a &ldquo;peak finding&rdquo; program that will attempt to
reconstruct which locations had sequences mapping to and how many
from each of the small sequences we have observed. If everything goes
well, the ideal outcome of our analysis for the data above will
produce:</p>
<ul>
	<li/>
<p class="compact">A peak of height 2 that starts at
	coordinate 10 and ends at coordinate 19</p>
	<li/>
<p class="compact">A peak of height 2 that starts at
	coordinate 34 and ends at coordinate 43</p>
</ul>
<p class="first-paragraph">Do note however that the DNA from one cell
had two regions covered, whereas the two other cells had only one
zone each. Our peaks above do not indicate this. They represent a
superposition of configurations across all cells. New configurations
leading to the same picture may be possible; there is no way to know
that the covered regions appear all together or separately in the
original DNA molecule.</p>
<p>Essential to remember:</p>
<p class="block-text-western">The peaks are cumulative, superimposed
measures of the observed configurations across millions of different
cells.</p>
<h2 class="western"><a name="do-peaks-represent-binding"></a>104.7 Do
peaks represent binding?</h2>
<p class="first-paragraph">Now, up to this point, we avoided the word
&ldquo;binding&rdquo; altogether - yet the purpose of ChIP-Seq is to
measure exactly that. A protein bound to a DNA molecule can &ldquo;protect&rdquo;
it from being washed away - then in a second &ldquo;pulldown&rdquo;
stage, only the DNA bound to a particular protein is kept. Hence
under the ideal condition, the result of a ChIP-Seq experiment
contains DNA that was protected by one specific type of protein.</p>
<p>The reality is a bit different. Remember your ChIP-Seq data does
not detect binding. It recognizes the abundance of DNA fragments in a
pool of DNA molecules. Each measure indicated that a given DNA
fragment was present when all the DNA was combined. As it turns out
in practice, there may be several other reasons for DNA to be present
in your library. Control samples are almost always required (see next
question)</p>
<h2 class="western"><a name="how-do-we-determine-peaks-from-sequencing-data"></a>
104.8 How do we determine peaks from sequencing data?</h2>
<p class="first-paragraph">There is a tendency to gloss over this
step, relying on the &ldquo;peak finder&rdquo; then using the results
that it produces. Once you think about the problem, you&rsquo;ll see
how peak finding process is more counterintuitive than you&rsquo;d
imagine.</p>
<p>Remember how sequencing instrument produces a linear
sub-measurement of DNA fragments? This measured &ldquo;read&rdquo;
has a length associated with it, but this length has nothing to do
with the original fragment length.</p>
<p>If we mark the first fragment as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=</font></font>
then we can see that depending on the DNA fragment lengths and our
measurement (read) lengths the measured reads could fall into three
different orientations.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">        <font face="Consolas, serif"><font size="2" style="font-size: 11pt">=========
DNA FRAGMENT =========</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       # Reads much shorter than fragment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       -------&gt;                 &lt;------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       # Reads as long as the fragment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       ------------------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                    &lt;------------------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       # Reads longer than the fragment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       -------------------------------------&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  &lt;-------------------------------------</font></font></p>
<p class="first-paragraph">Three different situations occur.</p>
<ol>
	<li/>
<p class="compact">The read length is much smaller than the
	fragment.</p>
	<li/>
<p class="compact">The read length is about the same size as
	the DNA fragment</p>
	<li/>
<p class="compact">The read length is larger than the DNA
	fragment and runs past the fragment ends.</p>
</ol>
<p class="first-paragraph">As a result, the analysis methods fall
into two categories: 1 and 3 as the second case occurs very rarely.
Often these are called &ldquo;narrow&rdquo; peak and &ldquo;broad&rdquo;
peak methods.</p>
<p>Be advised that a different mindset and approach is required for
each situation - methods that fit broad peaks are likely to work
sub-optimally in for narrow ones and vice versa.</p>
<h2 class="western"><a name="what-does-a-chip-seq-data-measure"></a>104.9
What does a ChIP-Seq data measure?</h2>
<p class="first-paragraph">If you refer to the sketch above only one
thing is common across all data sets. The start of the reads that
align on the forward strand indicates the left border of the DNA
fragment. The start of the reads that align on the reverse strand all
indicates the right border of the DNA fragment.</p>
<p>Hence we can state the fundamental &ldquo;law&rdquo; of ChIP-Seq
data:</p>
<p class="block-text-western">ChIP-Seq data measure the outer
borders, the &ldquo;edges&rdquo; of the DNA fragments.</p>
<p class="first-paragraph">Of course, this is not all that great of
news. It makes the picture a whole lot more complicated. Let me show
you what this means. Imagine that you see coverage that looks like
this image below. it seems so clear cut, there is one binding event,
in the middle, indicated by the highest point of the peak.</p>
<p><img src="0000_html_2c0522b244dd493d.jpg" name="Image155" align="bottom" hspace="12" width="560" height="197" border="0"/>
</p>
<p>Sadly this is misleading. Where the fragments appear to overlap to
most may be a superposition of many different, alternative positions.
In this particular case if you were to break down the data by strand
then plot only the coverages for the 5&rsquo; ends for each strand
separately you would see the following coverages:</p>
<p><img src="0000_html_cbdaa9779d03139e.jpg" name="Image156" align="bottom" hspace="12" width="560" height="211" border="0"/>
</p>
<p>It is the superposition of these two lower tracks that gives the
top coverage. But when plotted this way you can clearly see at least
two distinct peaks. Remember these now track what the instrument
measures: the ends of the DNA fragments. It appears that there are a
few distinct fragments (position) on the DNA. Let me mark these
positions.</p>
<p><img src="0000_html_ad27baa8d71c5e63.jpg" name="Image157" align="bottom" hspace="12" width="560" height="211" border="0"/>
</p>
<p>If we trust that our ChIP-Seq method measures bound locations,
then these are the locations occupied by our protein. But note how
none of them are bound in the location that the peak maxima indicate.
If location accuracy is critical, the peaks must be found by pairing
up the peaks called separately from the forward and reverse strands
to attempt to reconstitute the original DNA fragments, rather than
the center of each peak.</p>
<p><img src="0000_html_1898a1f2e1eb6a53.jpg" name="Image158" align="bottom" hspace="12" width="560" height="235" border="0"/>
</p>
<p>If peak location is not all that important, you may be able to
take the midpoint of the original peak.</p>
<p>Much algorithmic effort is spent to create software that
automatically builds the &ldquo;consensus&rdquo; location out of
data. Some work better than others - none of the tools can handle all
the situations, and much subjectivity remains. The type of the data
and its properties, whether or not a superposition of locations are
present will all substantially alter the performance of the method
you choose.</p>
<h2 class="western"><a name="what-type-of-chip-seq-studies-are-performed"></a>
104.10 What type of ChIP-Seq studies are performed?</h2>
<p class="first-paragraph">Speaking by the extremes, and your study
will likely fall somewhere in between but closer to one of these
extremes, ChIP-Seq studies fall into two major categories:</p>
<ol>
	<li/>
<p class="compact">The <b>Theory of Everything</b>: these are
	extensive scale studies, usually studying millions of locations,
	then often make sweeping, large-scale generalizations on where
	certain positions are about others. The principle &ldquo;gain-some
	lose-some drives the methods&rdquo;. The data analysis typically
	processes all peaks with the same parameters producing very high
	rates of false positives and false negatives. The goal of these
	studies is to beat random chance - which is a very low bar to pass.</p>
	<li/>
<p class="compact">The <b>Theory of One Peculiar Thing</b>:
	these are focused studies that use ChIP-Seq surveys to attempt to
	explain one particular phenomenon. The data often gets reduced to
	just a few locations where the very high accuracy of the methods is
	required.</p>
</ol>
<p class="first-paragraph">The needs of the two approaches are
radically different. As a rule, if you need to accurately determine
the &ldquo;peak&rdquo; at just a few locations the best tool that you
can use is YOUR EYE! We make use of &ldquo;peak callers&rdquo;
primarily because we can&rsquo;t afford to investigate every single
location.</p>
<h2 class="western"><a name="does-the-data-need-to-be-compared-to-control"></a>
104.11 Does the data need to be compared to control?</h2>
<p class="first-paragraph">Yes. The ChIP-Seq protocols have turned
out to be affected by many more systematic errors than initially
thought. Hence control samples are needed to identify these errors.</p>
<p>Typically two kinds of controls samples are used: <b>IgG</b>
control and <b>input</b> control. Each control can be thought of as
an incomplete ChIP-Seq protocol, where the process intentionally
skips a step of the process. These controls attempt to identify
regions that are enriched by other processes beyond the protein being
bound to the DNA.</p>
<ul>
	<li/>
<p class="compact">The <b>IgG</b> control is DNA resulting
	from a &ldquo;mock&rdquo; ChIP with Immunoglobulin G (IgG) antibody,
	which binds to the non-nuclear antigen.</p>
	<li/>
<p class="compact"><b>Input</b> control is DNA purified from
	cells that are cross-linked, fragmented, but without adding any
	antibody for enrichment.</p>
</ul>
<p class="first-paragraph">One problem with IgG control is that if
too little DNA is recovered after immunoprecipitation(IP), sequencing
library will be of low complexity (less diverse) and binding sites
identified using this control could be biased. I</p>
<p>Input DNA control is thought to be ideal in most of the cases. It
represents all the chromatin that was available for IP. More
information on choosing controls can be seen in:</p>
<ul>
	<li/>
<p class="compact">Biostars post: <font color="#4f81bd"><a href="https://www.biostars.org/p/15817/">What
	Control For Chip-Seq: Input, Igg Or Untagged Strain?</a></font> for
	more details.</p>
	<li/>
<p class="compact">Nature, 2015: <font color="#4f81bd"><a href="http://www.nature.com/ni/journal/v12/n10/abs/ni.2117.html">ChIP-Seq:
	technical considerations for obtaining high-quality data</a></font></p>
</ul>
<h2 class="western"><a name="should-my-chip-seq-aligner-be-able-to-detect-indels"></a>
104.12 Should my ChIP-Seq aligner be able to detect INDELs?</h2>
<p class="first-paragraph">Absolutely YES!</p>
<p>One of the most misleading advice - one that is reiterated even in
peer-reviewed publications like <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003326">Practical
Guidelines for the Comprehensive Analysis of ChIP-Seq Data</a></font>
PLoS Comp. Biology 2014, stems from a fundamental misunderstanding of
scientists when it comes to defining how &ldquo;alignment&rdquo; is
different from &ldquo;mapping.&rdquo;</p>
<p>We described the duality before in the alignment section stating
that &ldquo;mapping&rdquo; refers to finding the genomic location of
where a read originates from, whereas &ldquo;alignment&rdquo; refers
to finding the place and arrangement about the genome. Since ChIP-Seq
methods need to identify genomic locations many wrongly assume that
the alignment methods that cannot handle insertions and deletion are
still acceptable since we don&rsquo;t care about the exact alignment
anyway.</p>
<p>It is time to put this misconception out of its misery as it has
already caused and continues to produce substantial wasted efforts.
So let&rsquo;s make it abundantly clear:</p>
<p class="block-text-western">An aligner that cannot handle
insertions and deletions will not be able to find the correct
coordinates when a measurement contains such differences. The
&ldquo;mapping&rdquo; produced by such aligners over the reads in
question will be severely and systematically affected.</p>
<p class="first-paragraph">As a rule, aligners that cannot detect
insertions and deletions should <b>NEVER</b> be used under <b>ANY</b>
circumstances - unless no other aligner can be utilized (very rare
cases where other aligners would not work at all).</p>
<h2 class="western"><a name="what-are-the-processing-steps-for-chip-seq-data"></a>
104.13 What are the processing steps for ChIP-Seq data?</h2>
<p class="first-paragraph">The general processing steps are as
follows:</p>
<ol>
	<li/>
<p class="compact">Visualize and, in necessary, correct the
	quality of the sequencing data.</p>
	<li/>
<p class="compact">Align sequencing reads to a reference
	genome. The most popular read aligners are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font>.</p>
	<li/>
<p class="compact">Call peaks from the alignment bam files.</p>
	<li/>
<p class="compact">Visualize the resulting peak files and
	signal files.</p>
	<li/>
<p class="compact">Find the biological interpretation of the
	positions where the peaks are observed.</p>
</ol>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="aligning-chip-seq-data"></a>105 Aligning
ChIP-Seq data</h1>
<p class="first-paragraph">As we explained before chromatin
immunoprecipitation (ChIP) followed by high-throughput DNA sequencing
(ChIP-Seq) is a technique that is used to identify locations in the
genome, for example, the footprints of genome-wide transcription
factor binding sites and histone modification enriched regions.</p>
<p>We will demonstrate the steps used to process ChIP-Seq data
through a re-analysis of a published scientific work.</p>
<h2 class="western"><a name="which-paper-is-going-to-be-re-analyzed"></a>
105.1 Which paper is going to be re-analyzed?</h2>
<p class="first-paragraph">To illustrate all the analysis steps, we
are going to use data from the following publication:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/27143390">Genome-Wide
	Mapping of Binding Sites Reveals Multiple Biological Functions of
	the Transcription Factor Cst6p in Saccharomyces cerevisiae.</a></font>
	published in MBio in 2016</p>
</ul>
<p class="first-paragraph"><img src="0000_html_515b975232ca58c4.jpg" name="Image159" align="bottom" hspace="12" width="560" height="356" border="0"/>
</p>
<p>The data corresponding to this paper can be found in the SRA
Bioproject:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA306490">https://www.ncbi.nlm.nih.gov/bioproject/PRJNA306490</a></font></p>
	<li/>
<p class="compact">The script that will reproduce this paper
	is located in <font color="#4f81bd"><a href="http://data.biostarhandbook.com/chipseq/">http://data.biostarhandbook.com/chipseq/</a></font></p>
</ul>
<p class="first-paragraph">This section provides an annotated guide
to the commands in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chip-exo-redo.sh</font></font>
script. This script makes use of GNU Parallel to execute multiple
commands at a time (up to the number of CPU on the system).</p>
<h2 class="western"><a name="how-do-i-obtain-the-data-for-project-prjna306490"></a>
105.2 How do I obtain the data for project PRJNA306490?</h2>
<p class="first-paragraph">Make a few directoried to store data in:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p bam</font></font></p>
<p class="first-paragraph">Then perform the search for the run
information:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA306490  | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">The file contains information for four SRR
runs</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo.csv
| cut -d , -f 1,2,3 | head</font></font></p>
<p class="first-paragraph">will produce</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run,ReleaseDate,LoadDate</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033157,2016-04-08,2015-12-18</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033156,2016-04-08,2015-12-18</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033155,2016-04-08,2015-12-18</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154,2016-04-08,2015-12-18</font></font></p>
<p class="first-paragraph">You could run the following command on
each SRR run:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
-O data --split-files -F SRR3033154 </font></font>
</p>
<p class="first-paragraph">Or you could automate that into a single
line, by cutting the first column for SRR ids then piping the result
into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Isolate just the run ids for the run.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | cut -f 1 -d , | grep SRR &gt; runids.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the fastq files.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel --eta --verbose  &quot;fastq-dump -O data
--split-files -F {}&quot;</font></font></p>
<p class="first-paragraph">The data download may chug along for a
bit; the files are a few gigabytes each. You could use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-X</font></font>
flag to limit the dataset sizes.</p>
<h2 class="western"><a name="how-do-i-find-out-more-information-each-sample"></a>
105.3 How do I find out more information each sample?</h2>
<p class="first-paragraph">While the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo.csv</font></font>
contains quite a bit of information, maddeningly it lacks what
arguably would be the most important detail: which SRA number
corresponds to the samples mentioned in the paper. We need to run the
Entrez Direct tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">esummary</font></font>
to connect run ids to sample names:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA306490  | esummary &gt; summary.xml</font></font></p>
<p class="first-paragraph">The XML file returned by this query has a
nested structure with a wealth of information that can be difficult
to extract. Sometime the easiest is to read it by eye and manually
match the information. This might work here since there are only four
samples. But there are ways to automate this data extraction.</p>
<p>The tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
that is also part of <font color="#4f81bd"><a href="https://dataguide.nlm.nih.gov/edirect/xtract.html">EDirect:
xtract</a></font> is meant to facilitate turning XML files into
tabular files. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
tool operates by parsing an XML file, matching and extracting and
printing elements from this file.</p>
<p>Now XML as an industry standard with methods to do just this,
there is a so-called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XPath</font></font>
query language and the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XSL</font></font>
styling language that could be used to turn XML documents into any
other format. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
has nothing to do with either of these - yet it is used to achieve
the same. It is, in a way a re-imagining of what <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XPath</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XSL</font></font>
do - but this time filtered through a bioinformaticians mind.</p>
<p>For example, the following command generates a new row on each
element called named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">DocumentSummary</font></font>
then extracts the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">acc</font></font>
attribute of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Run</font></font>
element and prints that with the title. Investigate the XML file to
see where this information is present.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
summary.xml | xtract -pattern DocumentSummary -element Run@acc Title</font></font></p>
<p class="first-paragraph">will transform the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">summary.xml</font></font>
file into :</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033157
GSM1975226: Cst6p_ethanol_rep2; Saccharomyces cerevisiae;
ChIP-Seq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033156
GSM1975225: Cst6p_ethanol_rep1; Saccharomyces cerevisiae;
ChIP-Seq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033155
GSM1975224: Cst6p_glucose_rep2; Saccharomyces cerevisiae;
ChIP-Seq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154
GSM1975223: Cst6p_glucose_rep1; Saccharomyces cerevisiae; ChIP-Seq</font></font></p>
<p class="first-paragraph">From this output we can see here that each
sample has two replicates:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033157</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033156</font></font>
	correspond to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Cst6p_ethanol</font></font></p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033155</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154</font></font>
	correspond to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Cst6p_glucose</font></font></p>
</ul>
<h2 class="western"><a name="what-other-information-do-i-need-to-start-a-chip-seq-analysis"></a>
105.4 What other information do I need to start a ChIP-Seq analysis?</h2>
<p class="first-paragraph">You will now need a genome sequence and an
annotation file that lists the genes.</p>
<p>Sometimes (often?) it is surprisingly complicated to get the
latest information even from websites dedicated to the distribution
of data. As you will note for yourself, the data always seems to be
either in the wrong format or combined in ways that are inappropriate
to your current needs.</p>
<p>The data sources claim that it is a matter of &ldquo;just
downloading it&rdquo; yet soon you&rsquo;ll be knee-deep in
elbow-grease as you wrestle in the mud with uncooperative file
formats. For example, SGD claims that the latest genome releases can
be downloaded from
<font color="#4f81bd"><a href="http://downloads.yeastgenome.org/sequence/S288C_reference/genome_releases/">http://downloads.yeastgenome.org/sequence/S288C_reference/genome_releases/</a></font>.
As you happily and cheerfully download that data, you realize that
the chromosomal naming coordinates are different across the sequence
and the annotations. For example, the FASTA file will have the
sequence names listed as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;
ref|NC_001133| [org=Saccharomyces cerevisiae] [strain=S288C]
[moltype=genomic] [chromosome=I]</font></font></p>
<p class="first-paragraph">whereas the features the GFF files in the
<i>very same</i> downloaded data will list the chromosomal
coordinates as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;
chrI</font></font></p>
<p class="first-paragraph">Suffice to say no bioinformatics tool will
be able to match up these two. To use this data you would need
postprocess the file to rewrite <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chromosome=I</font></font>
to be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI</font></font>.
A data janitor job if there ever was one.</p>
<h2 class="western"><a name="how-do-i-get-the-standard-yeast-genome-data-and-annotation"></a>
105.5 How do I get the &ldquo;standard&rdquo; yeast genome data and
annotation?</h2>
<p class="first-paragraph">The paper that we are reanalyzing makes
use of the most up-to-date UCSC yeast genome release code named
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sacCer3</font></font>
from the <font color="#4f81bd"><a href="http://hgdownload.cse.ucsc.edu/goldenPath/sacCer3/">UCSC
download site</a></font> some might balk at knowing that this release
dates back to 2011. Quite a bit of knowledge could have been
accumulated since then. One advantage is though that many tools, like
IGV, already know of this genomic build and no custom genome building
is necessary.</p>
<p>Let&rsquo;s get the genome:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reference genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/saccer3.fa</font></font></p>
<p class="first-paragraph">Download the chromosomes and build a
genome. The sequences are stored by chromosome:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=http://hgdownload.cse.ucsc.edu/goldenPath/sacCer3/bigZips/chromFa.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
$URL | tar zxv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the chromosome sizes. Will be used later.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://hgdownload.cse.ucsc.edu/goldenPath/sacCer3/bigZips/sacCer3.chrom.sizes
&gt; refs/sacCer3.chrom.sizes</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Move the files</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv
*.fa refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create genome.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
refs/chr*.fa &gt; $REF</font></font></p>
<p class="first-paragraph">Remember to index the reference.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font></p>
<p class="first-paragraph">We now have the saccer3 genome on our
local filesystem.</p>
<h2 class="western"><a name="how-do-i-align-the-chip-seq-data"></a>105.6
How do I align the ChIP-Seq data?</h2>
<p class="first-paragraph">We assume that by this time your
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
programs above have finished. The alignment commands will all be of
the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF data/SRR3033154_1.fastq | samtools sort &gt; SRR3033154.bam </font></font>
</p>
<p class="first-paragraph">Note how this data is aligned in single
end mode and only file 1 is used. We too were surprised to learn this
from the authors, their rationale is that they are using a ChIP-Exo
method that sequences only a short segment of the fragment. We don&rsquo;t
quite believe this rationale to be correct - suffice to say only half
of the data was used and half of the data was tossed out.</p>
<p>We can automate the alignments as just one single line.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel --eta --verbose &quot;bwa mem -t 4 $REF
data/{}_1.fastq | samtools sort -@ 8 &gt; bam/{}.bam&quot;</font></font></p>
<p class="first-paragraph">Once the step above completes the
computationally most demanding process of ChIP-Seq has now been
completed - but the analysis itself is only starting in its earnest.</p>
<h2 class="western"><a name="do-i-need-to-trim-data-to-borders"></a>105.7
Do I need to trim data to borders?</h2>
<p class="first-paragraph">As we mentioned this before ChIP-Seq data
in general and ChIP-Exo in particular measures the left-right borders
of the binding sites rather than the center of them. Even though
tools should be able to reconstruct these borders algorithmically
from the entire alignments, sometimes (often) it is best to not even
&ldquo;show&rdquo; this spurious data to the algorithm, and to keep
only the essential sections of the data, in this case, the borders
only.</p>
<p>That is what the authors chose here, trimming 70 bp from the ends
of each alignment using the bamUtil tool:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Trim each bam file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel --eta --verbose &quot;bam trimBam bam/{}.bam
bam/temp-{}.bam -R 70 --clip&quot;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
We also need to re-sort the alignments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel --eta --verbose &quot;samtools sort -@ 8
bam/temp-{}.bam &gt; bam/trimmed-{}.bam&quot;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get rid of temporary BAM files.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm
-f bam/temp*</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Reindex trimmed bam files.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runids.txt | parallel --eta --verbose &quot;samtools index
bam/trimmed-{}.bam&quot;</font></font></p>
<p class="first-paragraph">The same process may work without this
step.</p>
<h2 class="western"><a name="how-do-i-visualize-the-alignments"></a>105.8
How do I visualize the alignments?</h2>
<p class="first-paragraph">ChIP-Seq data is all about coverages and
locations. As such the details of alignments are usually less
important. Visualizing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BAM</font></font>
files directly puts too much strain on visualizers as often many
samples need to be viewed simultaneously.</p>
<p>A <font color="#4f81bd"><a href="https://genome.ucsc.edu/goldenpath/help/bedgraph.html">bed
graph</a></font> file is a variant of a BED file that is meant to
display coverages over intervals. It is a simple format consisting of
a tab separated format of this form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrom
start end value</font></font></p>
<p class="first-paragraph">Where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">value</font></font>
is the coverage over the interval between <font face="Consolas, serif"><font size="2" style="font-size: 11pt">start</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">end</font></font>.
To create a coverage bedgraph file, we can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create an genome file that bedtools will require.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
faidx $REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a bedgraph file out of the BAM file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
genomecov -ibam bam/SRR3033154.bam  -g $REF.fai -bg &gt;
bam/SRR3033154.bedgraph </font></font>
</p>
<p class="first-paragraph">The <font color="#4f81bd"><a href="https://genome.ucsc.edu/goldenpath/help/bedgraph.html">bed
graph</a></font> file are smaller than the BAM file and may be
visualized and processed with many other tools</p>
<p>Below is a figure of visualizing bedgraph and BAM files
simultaneously. The upper track is bedgraph, the lower two panels are
the coverage and alignment sections of a BAM file.</p>
<p><img src="0000_html_1ed7c87949255f5e.jpg" name="Image160" align="bottom" hspace="12" width="560" height="257" border="0"/>
</p>
<p>As it happens the bedgraph is not quite efficient either, IGV has
to load the entire file into memory to visualize it. Yet another file
format had to be invented (and we&rsquo;ll spare your the gory
details) one that is called <font color="#4f81bd"><a href="https://genome.ucsc.edu/goldenpath/help/bigWig.html">bigwig</a></font>.
A bigwig is a binary format that contains the same information as a
bedgraph but allows for efficient access, hence allows software tools
to jump to various location in the file more quickly. This format, by
the way, was invented by Jim Kent, the &ldquo;benevolent dictator&rdquo;
of the UCSC genome browser.</p>
<p>The format itself is not documented or described properly - a
single program code, also written by Jim Kent was the sole source of
understanding how the format works. We invite the readers to reflect
on this a bit.</p>
<p>The tool called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedGraphToBigWig</font></font>
in <font color="#4f81bd"><a href="#ch133.xhtml#kent">Jim Kent&rsquo;s
utilities</a></font> can perform this transformation from bedgraph to
bigwig:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedGraphToBigWig
bam/SRR3033154.bedgraph $REF.fai bam/SRR3033154.bw</font></font></p>
<p class="first-paragraph">We can automate these steps as well:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create the coverage files for all BAM files.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
bam/*.bam | parallel --eta --verbose &quot;bedtools genomecov -ibam
{} -g $REF.fai -bg | sort -k1,1 -k2,2n &gt; {.}.bedgraph&quot;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Generate all bigwig coverages from bedgraphs.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
bam/*.bedgraph | parallel --eta --verbose &quot;bedGraphToBigWig {}
$REF.fai {.}.bw&quot;</font></font></p>
<p class="first-paragraph">Let&rsquo;s recapitulate what each file
contains and what their sizes are:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154.sam</font></font>
	(if generated) would be 5.3Gb, plain text, has alignment details, no
	random access.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154.bam</font></font>
	is 423Mb, binary compressed format, has alignment details, allows
	random access.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154.bedgraph</font></font>
	is 34Mb, plain text, coverage info only, loads fully into memory.</p>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR3033154.bw</font></font>
	is 8.1Mb, binary format, coverage info only, allows random access.</p>
</ul>
<p class="first-paragraph">We went from 5.3Gb to 8.1Mb, an
approximately 1000 fold data size reduction. Hence, in general, when
we intend to visualize these ChIP-Seq alignment files we transform
them from BAM files into a <b>bigwig</b> files and subsequently, we
only display these. Note that this same approach should be used when
visualizing RNA-Seq coverage data as well.</p>
<p><img src="0000_html_dbb0da67337a1287.jpg" name="Image161" align="bottom" hspace="12" width="560" height="257" border="0"/>
</p>
<h2 class="western"><a name="are-there-other-ways-to-generate-bigwig-files"></a>
105.9 Are there other ways to generate bigwig files?</h2>
<p class="first-paragraph">There are other tools that can generate
bigwig files directly from BAM files. The package that contains them
is called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">deeptools</font></font>
and may be installed with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a new environment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create --name bioinfo</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Activate the environment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">source
activate bioinfo</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install the deeptools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install deeptools</font></font></p>
<p class="first-paragraph">One specific program of this package
called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage</font></font>
offers a large number of functions to generate coverages. It is also
able to produce outputs with various smoothing, averaging and
normalizing parameters:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Counts within a window of 50bp.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage
-b bam/SRR3033154.bam -o bam/SRR3033154-digitized.bw</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Smooth the signal in a window.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage
-b bam/SRR3033154.bam --smoothLength 300 -o bam/SRR3033154-smooth.bw</font></font></p>
<p class="first-paragraph">We recommend a lot of caution when
applying any of these data altering, methods. Use them only when
there are good reasons and a well-defined rationale for using them.
Never &ldquo;borrow&rdquo; settings from another analysis or
publication. In general, the best bet is to use the original coverage
and resort to smoothing and averaging methods only if these appear to
be necessary.</p>
<p>As an example here is the comparison between the original
coverage, on the top and a smoothed data based on the coverage at the
bottom. As you can see for yourself, the data characteristics change
quite a bit - sometimes this is desirable but not always.</p>
<p><img src="0000_html_7354b21ebee8bb05.jpg" name="Image162" align="bottom" hspace="12" width="560" height="226" border="0"/>
</p>
<h2 class="western"><a name="what-is-the-next-step-of-analyzing-the-data"></a>
105.10 What is the next step of analyzing the data?</h2>
<p class="first-paragraph">Once you have the alignments and coverages
you can move onto the peak calling processes.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="chip-seq-peak-calling"></a>106 Chip-Seq
peak calling</h1>
<p class="first-paragraph">This chapter continues with the data
processing started in</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch105.xhtml#chip-seq-alignments">ChIP-Seq
	alignments</a></font>.</p>
</ul>
<p class="first-paragraph">We assume that you have the BAM alignment
and coverage files.</p>
<h2 class="western"><a name="how-do-i-reanalyze-a-chip-seq-experiment"></a>
106.1 How do I reanalyze a ChIP-Seq experiment?</h2>
<p class="first-paragraph">As you recall we are trying to reanalyze
data from the following publication:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/27143390">Genome-Wide
	Mapping of Binding Sites Reveals Multiple Biological Functions of
	the Transcription Factor Cst6p in Saccharomyces cerevisiae.</a></font>
	published in MBio in 2016</p>
</ul>
<p class="first-paragraph">Whereas we can download the data from SRA
the publication offers substantially less information on how the
analysis itself was performed. It briefly states that:</p>
<p class="block-text-western">To identify peaks and compare
biological duplicates, the program GEM (39) was used.</p>
<p class="first-paragraph">Now if you were to visit the <font color="#4f81bd"><a href="http://groups.csail.mit.edu/cgs/gem/">GEM
homepage</a></font> you&rsquo;ll see that the tool takes a wide
variety of parameters - each of which has consequences. Another line
states:</p>
<p class="block-text-western">The noise level was calculated from
averaged noise throughout each replicate.</p>
<p class="first-paragraph">You may not fully understand what this
means in practice- we don&rsquo;t either. Finally, you&rsquo;ll find
the following statement:</p>
<p class="block-text-western">&ldquo;Binding events were manually
curated in the Integrative Genomics Viewer o remove redundant events
representing the same binding and false-positive events with poor
peak shape or poor duplication&rdquo;</p>
<p class="first-paragraph">I think we can all agree that this is not
an objective process that would allow you to repeat the selection for
yourself. What you will find that a typical ChIP-Seq data analysis is
characterized by surprisingly subjective decision making. You will
have quite a hard time following the analysis as the processes used
to generate the results will often rely on seemingly subjective
decisions.</p>
<h2 class="western"><a name="canshould-i-contact-the-author-of-a-study"></a>
106.2 Can/should I contact the author of a study?</h2>
<p class="first-paragraph">Yes. We too have contacted the authors of
the study and they, in turn, have been very responsive to my
questions. I would like to thank both Dr.&nbsp;Jens Nielsen and David
Bergenholm for being most speedy and accommodating with their time.</p>
<p>Yet even so, as you yourself will experience it, if you ever
undertake a reanalysis of this type, the process of moving forward
with reproducing the results may still proceed at a surprisingly slow
pace.</p>
<p>Somehow every single question/response pair missed an important
detail, an essential yet perhaps tacitly assumed decision, one that
required yet another followup email.</p>
<p>After twenty (!) or so email exchanges with the author I was
finally able to simplify the process to the following script
<font color="#4f81bd"><a href="http://data.biostarhandbook.com/chipseq/chip-exo-redo.sh">chip-exo-redo.sh</a></font>
that in turn, I believe to contain everything one needs to redo the
analysis.</p>
<p>Notably, while the script contains most reproducible commands that
I believe were used in the paper, the results do not exactly match
(but are similar to) the published ones.</p>
<h2 class="western"><a name="should-i-first-summarize-my-data"></a>106.3
Should I first summarize my data ?</h2>
<p class="first-paragraph">Some like to simplify and summarize their
data by merging all replicates into a single file and comparing them.
By doing this merging, you will, of course, lose the independent
measurements. On the other hand, it may allow you to gain a higher
level understanding of all of its properties.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
merge -r bam/glucose.bam bam/SRR3033154.bam
bam/SRR3033155.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
merge -r bam/ethanol.bam bam/SRR3033156.bam
bam/SRR3033157.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index bam/glucose.bam </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index bam/ethanol.bam </font></font>
</p>
<p class="first-paragraph">Generate the coverages for each, here you
may skip making bigwig files since these files are not as large and
IGV will handle them as is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
genomecov -ibam bam/glucose.bam  -g $REF.fai -bg &gt;
bam/glucose.bedgraph</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
genomecov -ibam bam/ethanol.bam  -g $REF.fai -bg &gt;
bam/ethanol.bedgraph</font></font></p>
<p class="first-paragraph"><img src="0000_html_2b2880392bae3850.jpg" name="Image163" align="bottom" hspace="12" width="560" height="195" border="0"/>
</p>
<p>We can clearly see that in the region that we have visualized, the
glucose sample has a more uniform coverage throughout whereas the
ethanol affected sample shows a differential binding event.
Auto-scaling, in this case, may be misleading, it appears as if the
upper track has more data yet both have a similar number of reads 48
million versus 39 million.</p>
<p>The main difference is that the scale on the upper track goes to
1436 whereas the scale of the lower track goes to 5642.</p>
<h2 class="western"><a name="what-tools-can-i-use-to-predict-peaks"></a>
106.4 What tools can I use to predict peaks?</h2>
<p class="first-paragraph">There is no shortage of tools that one can
invoke to predict peaks. The most well-known is called MACS:
Model-based Analysis for ChIP-Seq. Even for this tool there are two
versions of it <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs2</font></font>
and each can be installed with conda or home brew.</p>
<p>Since these are python based tools they are best installed into a
separate environment:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Create a namespace for the tool</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
create --name macs python=2.7</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Activate the new environment.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">source
activate macs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Install the tools.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install numpy</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install macs2</font></font></p>
<p class="first-paragraph">You can also install the GEM peak
predictor designed specifically for ChIP-Exo data - this is the tool
used in the paper, and we&rsquo;ll use it as well later. Let&rsquo;s
call peaks with macs2:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Glucose
samples.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GLU1=bam/trimmed-SRR3033154.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GLU2=bam/trimmed-SRR3033155.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Ethanol
samples.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ETH1=bam/trimmed-SRR3033156.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ETH2=bam/trimmed-SRR3033157.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs2
callpeak -t $ETH1 $ETH2 -c $GLU1 $GLU2 --gsize 1E7  --name ethanol
--outdir ethanol/ </font></font>
</p>
<p class="first-paragraph">The results of this peak calling process
will go into the directory <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ethanol</font></font>
where a number of files are created. The file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ethanol_peaks.bed</font></font>
contains the regions that form peaks.</p>
<p>Visually inspecting the results show the regions that MACS
believes to form &ldquo;peaks&rdquo;. As a comparison we also plot in
a track the final results of the paper that would indicate the
binding locations. As we can see <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs2</font></font>
does reasonably good job in finding the locations, but at the same
time the peaks don&rsquo;t quite seem to be well localized, they
spread over larger regions.</p>
<h2 class="western"><a name="how-do-i-refine-the-peaks"></a>106.5 How
do I refine the peaks?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ethanol_summits.bed</font></font>
file contains the most likely maxima of each peak. These are single
base pair wide location computed at the same time as the rest of the
peaks:</p>
<p>Subsequent analysis typically consists of processing these summits
into one place (if there are multiple of them overlapping in one
region) and filtering for the most likely binding location. Alas
neither process is entirely objective, much effort and custom coding
goes into refining the final positions</p>
<p>As we mentioned before the parameter settings for filtering are
almost always &ldquo;reverse engineered.&rdquo; There are
expectations, and then in turn the parameters are tuned in the
direction that strengthens these expectations. For example, we could
keep only peaks with a score above 10 with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ethanol/ethanol_summits.bed | awk ' $5&gt; 10 { print $0 }' &gt;
highscore.bed</font></font></p>
<p class="first-paragraph">Of course, here we treading into the
dangerous territory of data manipulation - that is artificially
selecting data with some properties. Unfortunately, this is a risk we
must take in this case - as objective measures are few and far
between.</p>
<h2 class="western"><a name="what-else-can-i-do-to-refine-the-peaks"></a>
106.6 What else can I do to refine the peaks?</h2>
<p class="first-paragraph">Due to the biased nature of the sequencing
protocols ChIP-Seq data is typically affected by a very higher number
of false positives. Thus the most common necessary step is that of
removing some (or many) of the resulting peaks.</p>
<p>One such example is that of filtering out repetitive or low
complexity regions of the genome. Data from these areas almost always
indicate false signals as, if there is a measurement that matches
equally</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sdust</font></font>
program of the [minimap][<font color="#4f81bd"><a href="https://github.com/lh3/minimap">https://github.com/lh3/minimap</a></font>]can
be used to create an interval file with the regions of low
complexity:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sdust
refs/sc.fa &gt; lowcomplexity.bed</font></font></p>
<p class="first-paragraph">then you could keep the high scoring
values that do not fall on low complexity regions with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
intersect -v -a highscore.bed -b lowcomplexity.bed &gt;
highcomplexity.bed</font></font></p>
<p class="first-paragraph">You could also filter to keep only peaks
that fall upstream of gene starts. Again bedtools could help:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
flank -g refs/sc.fa.fai -l 1000 -r 0 -s -i refs/genes.gff &gt;
flanked.gff</font></font></p>
<p class="first-paragraph">Then intersect the peaks with that</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
intersect -u -a ethanol_summits.bed -b flanked.gff &gt;
upstream-peaks.bed</font></font></p>
<p class="first-paragraph">These types of tweaking and filtering are
perhaps the most time-consuming aspects of a ChIP-Seq analysis - they
are part of the quest to refine the very high false positive rates to
a more reliable subset.</p>
<p>In the end the success of the ChIP-Seq analysis gets decided right
here at peak filtering. Every single subsequent process operates on
the position and sequence information of these peaks hence these peak
filtering have the most effect on the results.</p>
<p>At the same time, and we reiterate, there is the danger that you
may create an artificial, non-representative smaller subset of data.</p>
<h2 class="western"><a name="how-did-the-paper-call-peaks"></a>106.7
How did the paper call peaks?</h2>
<p class="first-paragraph">The paper used the GEM peak caller in the
following way:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Shortcuts.
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GEM=~/src/gem/gem.jar</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">DIST=~/src/gem/Read_Distribution_ChIP-exo.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run GEM on ethanol samples.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">java
-Xmx3G -jar $GEM --d $DIST --g refs/sacCer3.chrom.sizes --genome refs
--exptCond1 $ETH1  --exptCond1 $ETH2 --f SAM --out ethanol --outBED
--k 5  --smooth 3 --mrc 20 --k_win 30</font></font></p>
<p class="first-paragraph">and these will generate the peaks that
most closely reproduce the published results.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="chip-seq-motifs"></a>107 Chip-Seq motifs</h1>
<p class="first-paragraph">This chapter continues with the data
processing starting in both</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch105.xhtml#chip-seq-alignments">ChIP-Seq
	alignments</a></font>.</p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="#ch106.xhtml#chip-peak-calling">ChIP-Seq
	peak-calling</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://data.biostarhandbook.com/chipseq/">ChIP-Seq
	data</a></font></p>
</ul>
<p class="first-paragraph">Recall the section on Sequence Patterns,
see <font color="#4f81bd"><a href="#ch071.xhtml#sequence-patterns">Pattern
matching</a></font> where a sequence pattern (a motif) was defined as
a sequence of bases described by certain rules. These rules could
vary from requiring exact matches to being more elaborate
requirements such as:</p>
<p class="block-text-western">All locations where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATA</font></font>
is followed by one or more <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GC</font></font>s
and no more than five <font face="Consolas, serif"><font size="2" style="font-size: 11pt">T</font></font>s
then ending with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTA</font></font></p>
<h2 class="western"><a name="why-do-scientists-consider-sequence-motifs-to-be-important"></a>
107.1 Why do scientists consider sequence motifs to be important?</h2>
<p class="first-paragraph">Sequence motifs are of particular interest
in ChIP-Seq data as it is believed that these motifs control (or at
least influence) many biological processes: for example the binding
of transcriptions factors to DNA.</p>
<h2 class="western"><a name="what-makes-motifs-difficult-to-evaluate"></a>
107.2 What makes motifs difficult to evaluate?</h2>
<p class="first-paragraph">The most common challenge when dealing
with motifs stems from not having a proper picture of just how often
a given motif might be present in the genome, to begin with. Finding
a motif in a binding location would be interesting only if seeing
that motif has predictive value. Since the content of genome itself
is not random the only way to properly assess the motif&rsquo;s
abundance is to match it back to the genome.</p>
<p>Then, of course, a motif might turn out to be &ldquo;statistically
significant&rdquo; in that it may be observed at a rate that is
different from random chance. That alone is a weak observation. At
the same time, you should also be aware of and take into account the
&ldquo;effect size&rdquo;: just how big this deviation from random
chance is.</p>
<h2 class="western"><a name="how-do-i-find-known-motifs"></a>107.3
How do I find known motifs?</h2>
<p class="first-paragraph">The paper that we are re-analyzing
indicates that the motif <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTGACGT</font></font>
was found in the location of the peaks. This pattern is then
represented by a so-called sequence logo and described as</p>
<p class="block-text-western">A binding site motif of 5&prime;
GTGACGT 3&prime; was enriched from the bound region sequences
(discovered in 32 of the 40 regions)</p>
<p class="first-paragraph">A sequence (consensus) logo consist of a
stack of letters where the size of each letter corresponds to the
frequency of observation of each letter.</p>
<p class="compact"><img src="0000_html_caac4577061742aa.jpg" name="Image164" align="bottom" hspace="12" width="560" height="326" border="0"/>
</p>
<p class="compact">o</p>
<p>How do you find this motif in the genome? In the chapters <font color="#4f81bd"><a href="#ch071.xhtml#sequence-patterns">Pattern
matching</a></font> we describe a few tools, below we&rsquo;ll use
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit</font></font>
as an example:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/sc.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Turn the output into a GFF file:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | seqkit locate -p GTGACGT | head</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<p class="first-paragraph">will produce the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqID
patternName pattern strand start end matched</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI
GTGACGT GTGACGT + 10705 10711 GTGACGT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI
GTGACGT GTGACGT + 19820 19826 GTGACGT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI
GTGACGT GTGACGT + 84191 84197 GTGACGT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI
GTGACGT GTGACGT + 101042 101048 GTGACGT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chrI
GTGACGT GTGACGT + 120400 120406 GTGACGT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">To visualize this result in IGV you can
turn it into a 9 column GFF file with an <font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font>
incanation (add <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
for unknown columns):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | seqkit locate -p GTGACGT | grep -v seqID | awk ' { OFS=&quot;\t&quot;;
print $1, &quot;.&quot;, &quot;.&quot;, $5, $6, &quot;.&quot;, $4,
&quot;.&quot;, &quot;.&quot; }' &gt; motifs.gff</font></font></p>
<p class="first-paragraph">Visualizing the file you can see that
there are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">716</font></font>
occurrences of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTGACGT</font></font>
motif throughout the genome. The tracks below show the peaks found in
the paper and the locations of the exact motif matches.</p>
<p><img src="0000_html_f27d653a13439f09.jpg" name="Image165" align="bottom" hspace="12" width="560" height="192" border="0"/>
</p>
<p>The published data finds <font face="Consolas, serif"><font size="2" style="font-size: 11pt">40</font></font>
binding locations. Since there are <font face="Consolas, serif"><font size="2" style="font-size: 11pt">716</font></font>
motif locations, it seems that even if all peaks fell on motifs only
5% of motif locations were bound by factors. Clearly, the presence of
motif does not indicate binding. But as it turns out even that is not
quite right. Let&rsquo;s find out how many of the published peaks
fall on motifs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O
http://data.biostarhandbook.com/chipseq/peaks-published.bed</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
intersect -u -a peaks-published.bed -b motifs.gff | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font>
</p>
<p class="first-paragraph">It turns out only 4 of the reported peaks
intersect exactly with this motif locations. So the wording of 32 of
40 regions containing the motif is probably interpreted differently -
though again we don&rsquo;t know what it means. Most likely the word
&ldquo;region&rdquo; means a wider interval, not just the reported
peak coordinate. Let&rsquo;s extend these peaks:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
slop -b 50 -g refs/saccer3.fa.fai -i peaks-published.bed &gt;
wider.bed</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
intersect -u -a wider.bed -b motifs.gff | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
11</font></font></p>
<p class="first-paragraph">Extending peaks by 50 bases left and right
produced 11 more hits. We&rsquo;re getting there but not quite there
yet.</p>
<p>Of course, as you may have already noticed the motif above does
not need to be exact - it allows for mismatches. On the other hand of
course when we allow a motif to be inexact the count for its
locations rises, often dramatically.</p>
<h2 class="western"><a name="how-many-degenerate-motifs-in-the-genome"></a>
107.4 How many &ldquo;degenerate&rdquo; motifs in the genome?</h2>
<p class="first-paragraph">The sequence logo as drawn in the paper
allows for some variations, but just how many it covers is difficult
to visually assess. For all their visual flair and seeming &ldquo;logic&rdquo;
sequence logos can often be misleading. They show the most common
letter, but we don&rsquo;t see the individual motifs - only a
superposition of all states. Let&rsquo;s run our pattern locator by
allowing two alternatives at three locations:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
$REF | seqkit locate -p 'G(T|A)GA(C|T)(G|A)T' | wc -l| grep -v seqID
| awk ' { OFS=&quot;\t&quot;; print $1, &quot;.&quot;, &quot;.&quot;,
$5, $6, &quot;.&quot;, $4, &quot;.&quot;, &quot;.&quot; }' &gt;
motifs-generic.gff</font></font></p>
<p class="first-paragraph">It now produces <font face="Consolas, serif"><font size="2" style="font-size: 11pt">11349</font></font>
possible matches! That is a tenfold increase over the exact motif!
Note just how dense the motif coverage over the genome has become.</p>
<p><img src="0000_html_204e6ad7f61bcb16.jpg" name="Image166" align="bottom" hspace="12" width="560" height="200" border="0"/>
</p>
<p>Even so, this pattern has picked up just two new peaks, while
raising the total non-relevant locations for the motif 15 fold. Not a
good tradeoff! That still does not reproduce all the motifs as
described in the paper.</p>
<h2 class="western"><a name="how-do-i-call-motifs"></a>107.5 How do I
call motifs?</h2>
<p class="first-paragraph">The GEM peak predictor that besides
calling peaks also calls motifs, though the authors appear to have
not used those results, instead chose to use the MEME suite.</p>
<p>For example, if you were to attempt to find motifs for the peaks
called with methods shown in the previous chapter you would need to
do the following:</p>
<ol>
	<li/>
<p class="compact">take the peak locations,</p>
	<li/>
<p class="compact">extend them left and right by a fixed
	amount,</p>
	<li/>
<p class="compact">extract the sequences that these intervals
	cover then</p>
	<li/>
<p class="compact">run a motif search program on these
	sequences</p>
</ol>
<p class="first-paragraph">The script might look like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">BED=ethanol/ethanol_summits.bed</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/saccer3.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N=50</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extend the peak locations by 50 basepairs left and right.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
slop -g $REF.fai -i $BED -b $N &gt; extended.bed</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract the sequences corresponding to these locations.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
getfasta -bed extended.bed -fi $REF -fo sequences.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run meme-chip on the sequences.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm
-rf results</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">meme-chip
-meme-maxw 10 -meme-minw 5 -o results sequences.fa</font></font></p>
<p class="first-paragraph">These commands will generate a directory
called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results</font></font>
that among other information will contain the logo. In this case the
logo looks like:</p>
<p><img src="0000_html_5adad54986167d7.jpg" name="Image167" align="bottom" hspace="12" width="560" height="167" border="0"/>
</p>
<p>Obviously, this motif is not similar to the published motif at
all. It appears to have a low complexity to it, indicating that our
sequences represent too many repetitive regions. A good reminder that
many/most peak prediction methods still produce a high number of
false positives and that repetitive regions can influence results.
Replace the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">BED</font></font>
file above with the different filtered files that we also suggested
before: the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">highscoring.bed</font></font>
then the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">highcomplexity.bed</font></font>
file and rerun the script to see how the motif evolves. In the latter
case we get the following:</p>
<p><img src="0000_html_5302b5c9ab756af4.jpg" name="Image168" align="bottom" hspace="12" width="560" height="85" border="0"/>
</p>
<p>This motif now matches the published result, though we can
recognize the relative frequencies of the letters is different. If
anything this motif may look a little cleaner than the published one.
In addition we have filtered the data no by eye but with an objective
measure of taking into account the genome complexity.</p>
<h2 class="western"><a name="what-is-the-conclusion-on-finding-motifs"></a>
107.6 What is the conclusion on finding motifs?</h2>
<p class="first-paragraph">When it comes to motifs the results appear
to be more subjective than the scientific field currently cares to
admit. As much as we would all like to find a simple explanation of
why a factor binds to a certain location, why a phenomenon is
localized in the genome the evidence shows that the influence of the
sequences and motifs are less well defined than we hoped for.</p>
<p>In our opinion motifs are pursued primarily because they appear to
bring objectivity and determinism as to the why a factor binds to a
given location - in reality, this effect is much weaker than many
assume.</p>
<h2 class="western"><a name="have-motifs-ever-misled-scientists"></a>107.7
Have motifs ever misled scientists?</h2>
<p class="first-paragraph">The most educational cautionary tale of
motif finding comes from the now retracted paper titled <font color="#4f81bd"><a href="http://www.nature.com/nature/journal/v502/n7469/full/nature12535.html">Genomic
organization of human transcription initiation complexes</a></font>
published in Nature in 2013 (retracted in 2014 by the authors). In it
the authors report sequence patterns of the form:</p>
<p><img src="0000_html_92f0120f97bce0b9.jpg" name="Image169" align="bottom" hspace="12" width="560" height="292" border="0"/>
</p>
<p>As it turns, it has been found that due to the degeneracy of the
patterns the resulting motifs could have been produced just by chance
in any preselected region of the genome. While the authors have
attempted to correct and verify that the motif appears at rates above
this chance they performed the comparison incorrectly and hence
believed their results as being novel.</p>
<p>It does not help that the motif itself is so &ldquo;obviously&rdquo;
clear cut. It is quite hard to believe that this pattern could be due
to random chance. The goal here not to criticize the authors but to
learn from them - you&rsquo;ll have to protect yourself from errors
like with ongoing vigilance.</p>
<p>The lesson here is to understand and remember to treat motifs with
a healthy dose of skepticism and to recognize their valuable yet also
limited information content. It is very unlikely that whatever
mechanisms you are after will be explained by the presence/absence of
motifs alone. The DNA is composed of only four bases A,T,G and C.
This low variety of choices coupled to a large number of measurements
and samples can easily produce results that seem impossible to occur
by chance.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="ming-tangs-guide-to-chip-seq-analysis"></a>
108 Ming Tang&rsquo;s guide to ChIP-Seq analysis</h1>
<p class="first-paragraph">The author of this guide is <font color="#4f81bd"><a href="https://github.com/crazyhottommy">Ming
Tang</a></font>. The material was developed for the Biostar Handbook.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/crazyhottommy">GitHub
	profile</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://crazyhottommy.blogspot.hu/">Diving
	into Genetics and Genomics</a></font></p>
</ul>
<h2 class="western"><a name="what-is-chip-seq-1"></a>108.1 What is
ChIP-seq?</h2>
<p class="first-paragraph">Chromatin immunoprecipitation (ChIP)
followed by high-throughput DNA sequencing (ChIP-seq) is a technique
to map genome-wide transcription factor binding sites and
histone-modification enriched regions.</p>
<p>Briefly, DNA bounding proteins and DNA (Chromatin) are
cross-linked by formaldehyde and the chromatin is sheared by
sonication into small fragments (typically 200 ~ 600 bp). The
protein-DNA complex is immnuoprecipitated by an antibody specific to
the protein of interest. Then the DNA is purified and made to a
library for sequencing. After aligning the sequencing reads to a
reference genome, the genomic regions with many reads enriched are
where the protein of interest bounds. ChIP-seq is a critical method
for dissecting the regulatory mechanisms of gene expression.</p>
<h2 class="western"><a name="what-are-the-processing-steps-for-chip-seq-data-1"></a>
108.2 What are the processing steps for ChIP-seq data?</h2>
<p class="first-paragraph">The general processing steps are as
following:</p>
<ol>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Quality
	control of your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq</font></font>
	read files using tools such as <font color="#4f81bd"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a></font>.
	Read our previous section: <font color="#4f81bd"><a href="https://read.biostarhandbook.com/data/fastq-quality-visualization.html">Visualizing
	sequencing data quality</a></font>.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Aligning
	fastq reads to reference genome. The most popular read aligners are
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font>.
	Read section <font color="#4f81bd"><a href="https://read.biostarhandbook.com/align/short-read-aligners.html">Short
	Read Aligners</a></font>. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font>
	has two versions: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie1</font></font>
	and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>.
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie2</font></font>
	is better for reads length greater than 50bp. It is still very
	common to use 36bp single-end sequencing library for ChIP-seq, so
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie1</font></font>
	is preferred for ChIP-seq with short reads.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Calling
	peaks from the alignment bam files.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">Visualizing
	the resulting peak files and raw signal files.</p>
</ol>
<h2 class="western"><a name="what-are-igg-control-and-input-control-for-chip-seq"></a>
108.3 What are IgG control and input control for ChIP-seq?</h2>
<p class="first-paragraph">Like any other experiments, a control is
needed for ChIP-seq. There are two kinds of controls for ChIP-seq:
IgG control and input control. IgG control is DNA resulting from a
&ldquo;mock&rdquo; ChIP with Immunoglobulin G (IgG) antibody, which
binds to non-nuclear antigen; input control is DNA purified from
cells that are cross-linked, fragmented, but without adding any
antibody for enrichment.</p>
<p>One problem with IgG control is that if too little DNA is
recovered after immunoprecipitation(IP), sequencing library will be
of low complexity and binding sites identified using this control
could be biased. Input DNA control is ideal in most of the cases. It
represents all the chromatin that was available for IP. Read <font color="#4f81bd"><a href="https://www.biostars.org/p/15817/">this
biostars post</a></font> for discussion.</p>
<h2 class="western"><a name="data-sets"></a>108.4 Data sets</h2>
<p class="first-paragraph">To illustrate all the analysis steps, we
are going to use a public data set from two papers:</p>
<ul>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/26258633">Genome-wide
	association between YAP/TAZ/TEAD and AP-1 at enhancers drives
	oncogenic growth</a></font>. Francesca Zanconato et.al. 2014. <i>Nature
	Cell Biology</i>. We will use transcription factor YAP1 ChIP-seq
	data in MDA-MB-231 breast cancer cells from this paper.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/24885402">Nucleosome
	positioning and histone modifications define relationships between
	regulatory elements and nearby gene expression in breast epithelial
	cells</a></font>. Suhn Kyong Rhie et.al. 2014. <i>BMC Genomics</i>.
	We will use H3K27ac ChIP-seq data in MDA-MB-231 cells from this
	paper.</p>
	<li/>
<p style="margin-top: 0in; margin-bottom: 0.14in">The data
	sets can be found with accession number <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE66081">GSE66081</a></font>
	and <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49651">GSE49651</a></font>.</p>
</ul>
<h2 class="western"><a name="how-to-obtain-the-data"></a>108.5 How to
obtain the data?</h2>
<p class="first-paragraph">Read our previous section: <font color="#4f81bd"><a href="#ch063.xhtml#SRA">Accessing
the Short Read Archive (SRA)</a></font>.</p>
<p>Prepare folders and download the data.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
go to your home directory </i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
make folders for this example</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p ChIP-seq/</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{data,results,scripts,software}</font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p ChIP-seq/results/</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{bams,peaks,fastqc,motifs}</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ChIP-seq/data</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
for YAP1. It takes time, alternatively download fastq directly from
EGA https://www.ebi.ac.uk/ega/home</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR1810900</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
for H3K27ac</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949140</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
for input DNA</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949142</font></font></p>
<p class="first-paragraph">In general, do not change the names of
files. This can be error prone, especially when you have many
samples. Instead, prepare a name mapping file, and interpret the
final results with the mapped names. For demonstrating purpose, I
will re-name the files to more meaningful names.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR1810900.fastq YAP1.fastq</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949140.fastq H3K27ac.fastq</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mv</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949142.fastq Input.fastq</font></font></p>
<p class="first-paragraph">Now, run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">FASTQC</font></font>
on all the samples. Read previous section on <font color="#4f81bd"><a href="https://read.biostarhandbook.com/unix/loops.html">looping
over files</a></font>. or use GNU parallel:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
always specify -j to not abuse the computing cluster</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">find</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*fastq </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-j 4 fastqc </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-o ../results/fastqc</font></font></p>
<p class="first-paragraph">Overall, the sequencing qualities are good
for all three samples, I will skip trmming low quality bases, and go
ahead with the original fastq files for aligning.</p>
<h2 class="western"><a name="how-to-align-chip-seq-reads"></a>108.6
How to align ChIP-seq reads?</h2>
<p class="first-paragraph">once you have finished QC confirming the
sequencing reads are of high quality and with no adaptor
contamination, you can do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
use 10 threads to speed up. should finish within 10 mins</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
change the path to the bowtie1 index in your own machine.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p 10 --best --chunkmbs 320 /risapps/reference/bowtie1/hg19 -q
YAP1.fastq -S ../results/bams/YAP1.sam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads processed: 24549590</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads with at least one reported alignment: 19664247 (80.10%)</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads that failed to align: 4885343 (19.90%)</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Reported</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
19664247 alignments to 1 output stream(s)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p 10 --best --chunkmbs 320 /risapps/reference/bowtie1/hg19 -q
H3K27ac.fastq -S ../results/bams/H3K27ac.sam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads processed: 29759754</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads with at least one reported alignment: 29100444 (97.78%)</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads that failed to align: 659310 (2.22%)</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Reported</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
29100444 alignments to 1 output stream(s)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p 10 --best --chunkmbs 320 /risapps/reference/bowtie1/hg19 -q
Input.fastq -S ../results/bams/Input.sam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads processed: 18811563</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads with at least one reported alignment: 18127743 (96.36%)</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
reads that failed to align: 683820 (3.64%)</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Reported</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
18127743 alignments to 1 output stream(s)</font></font></p>
<p class="first-paragraph">For demonstrating purpose, I will write
down explicitly all the commands, but the GNU parallel trick can be
used to save you some typing. In the end, I will introduce a shell
script to chain all the steps together and show you how to make
reusable codes.</p>
<p>Convert <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sam</font></font>
to sorted <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
avoid writing unsorted bam to disk</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
note that samtools sort command changed invoke pattern after version
1.3 https://www.biostars.org/p/169745/</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
samtools view and index will be multi-thread soon
https://github.com/samtools/htslib/pull/397</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
../results/bams</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
view -bS YAP1.sam </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sort -@ 4 - -T YAP1 -o YAP1.sorted.bam</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
index the sorted bam</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
index YAP1.sorted.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
view -bS H3K27ac.sam </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sort -@ 4 - -T H3K27ac -o H3K27ac.sorted.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
index H3K27ac.sorted.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
view -bS Input.sam </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sort -@ 4 - -T Input -o Input.sorted.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
index Input.sorted.bam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
remove sam files to save space , rm can be dangerous, use rm -i if
necessary</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*sam</font></font></p>
<h2 class="western"><a name="how-do-i-call-peaks"></a>108.7 How do I
call peaks?</h2>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://liulab.dfci.harvard.edu/MACS/00README.html">MACS</a></font>
is short for Model Based Analysis for ChIP-Seq data, which was
developed in Sheirly Liu&rsquo;s lab at Harvard by Tao Liu. It is one
of the most popular peak-calling algorithums used in literatures.
First, you need to install MACS.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
install through pip</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
install macs</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
peak calling for human samples, with default p value 1e-5</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-t YAP1.sorted.bam -c Input.sorted.bam -n YAP1 -g hs -p 1e-5 -o
../peaks/</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
peak calling for H3K27ac</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-t H3K27ac.sorted.bam -c Input.sorted.bam -n H3K27ac -g hs -p 1e-5 -o
../peaks/</font></font></p>
<p class="first-paragraph">In addition to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font>
can take in other input file formats such as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bed</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sam</font></font>.
type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font>
on your terminal and press <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Enter</font></font>
to see full list of helps.</p>
<p>Parameter settings can be different depending on your own data
sets. There is no one-fit-all settings. You may want to load the
called peaks in to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGV</font></font>
and visually inspect them.</p>
<p>Note: Tao Liu is developing <font color="#4f81bd"><a href="https://github.com/taoliu/MACS">MACS2</a></font>.
there is a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-broad</font></font>
option to call broad peaks for histone modification
ChIP-seq(e.g.&nbsp;H3K36me3). However, in my experience, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs14</font></font>
is still widely used and performs very well.</p>
<p>Take a look at <font color="#4f81bd"><a href="https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.3_MACS2_peak_calling_details.md">How
to set MACS2 peak calling parameters</a></font>.</p>
<h2 class="western"><a name="how-do-i-call-super-enhancers"></a>108.8
How do I call super-enhancers?</h2>
<p class="first-paragraph">The fancy &ldquo;super-enhancer&rdquo;
term was first introduced by <font color="#4f81bd"><a href="http://younglab.wi.mit.edu/publications.htm">Richard
Young</a></font>&rsquo;s lab in Whitehead Institute. Basically,
super-enhancers are clusters of enhancers (most commonly defined by
H3K27ac peaks) stitched together if they are within 12.5kb apart. The
concept of super-enhancer is NOT new. One of the most famous example
is the Locus Control Region (LCR) that controls the globin gene
expression, and this has been known for decades. If you are
interested in the controversy, please read:</p>
<p>A review in Nature Genetics <font color="#4f81bd"><a href="http://www.nature.com/ng/journal/v47/n1/full/ng.3167.html">What
are super-enhancers?</a></font><br/>
Another comment in Nature
Genetics <font color="#4f81bd"><a href="http://www.nature.com/ng/journal/v49/n1/full/ng.3759.html?WT.feed_name=subjects_molecular-biology">Is
a super-enhancer greater than the sum of its parts?</a></font></p>
<p>From the <font color="#4f81bd"><a href="http://homer.salk.edu/homer/ngs/peaks.html">HOMER
page</a></font><br/>
<b>How finding super enhancers works:</b></p>
<p class="block-text-western">Super enhancer discovery in HOMER
emulates the original strategy used by the Young lab. First, peaks
are found just like any other ChIP-Seq data set. Then, peaks found
within a given distance are &lsquo;stitched&rsquo; together into
larger regions (by default this is set at 12.5 kb). The super
enhancer signal of each of these regions is then determined by the
total normalized number reads minus the number of normalized reads in
the input. These regions are then sorted by their score, normalized
to the highest score and the number of putative enhancer regions, and
then super enhancers are identified as regions past the point where
the slope is greater than 1.</p>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://homer.salk.edu/homer/index.html">HOMER</a></font>
is a very nice tool for finding enriched peaks, doing motif analysis
and many more. It requires making a tag directory first. <font color="#4f81bd"><a href="http://younglab.wi.mit.edu/super_enhancer_code.html">ROSE</a></font>
is the tool from Richard Young&rsquo;s lab. Since it works directly
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files, we will use it for demonstration.</p>
<p>Download <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ROSE</font></font>
to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">software</font></font>
folder.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
../../software/</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
https://bitbucket.org/young_computation/rose/get/1a9bb86b5464.zip</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">unzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
1a9bb86b5464.zip</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
young_computation-rose-1a9bb86b5464/</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
there are multiple python scripts in the folder.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
one has to run ROSE inside the ROSE folder.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">python</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ROSE_main.py -g hg19 -i ../../results/peaks/H3K27ac_peaks.bed -r
../../results/bams/H3K27ac.sorted.bam -c
../../results/bams/Input.sorted.bam -o ../../results/peaks</font></font></p>
<h2 class="western"><a name="how-do-i-get-a-normalized-bigwig-track-for-visualizing-the-raw-signal"></a>
108.9 How do I get a normalized <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bigWig</font></font>
track for visualizing the raw signal?</h2>
<p class="first-paragraph">After you get the peak calls, we want to
inspect the quality of the peaks by visulizing them in a browser such
as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGV</font></font>
along with the raw signal track. We will use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bigWig</font></font>
track for this purpose.</p>
<p>From the <font color="#4f81bd"><a href="http://genome.ucsc.edu/goldenpath/help/bigWig.html">UCSC
help</a></font>:</p>
<p class="block-text-western">The bigWig format is for display of
dense, continuous data that will be displayed in the Genome Browser
as a graph. BigWig files are created initially from wiggle (wig) type
files, using the program wigToBigWig. The resulting bigWig files are
in an indexed binary format. The main advantage of the bigWig files
is that only the portions of the files needed to display a particular
region are transferred to UCSC, so for large data sets bigWig is
considerably faster than regular wiggle files</p>
<p class="first-paragraph">Make sure you understand the other two
closely related file formats: <font color="#4f81bd"><a href="http://genome.ucsc.edu/goldenpath/help/bedgraph.html">bedgraph</a></font>
and <font color="#4f81bd"><a href="http://genome.ucsc.edu/goldenpath/help/wiggle.html">wig</a></font>
file.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs14</font></font>
can generate a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedgraph</font></font>
file by specifying <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--bdg</font></font>,
but the resulting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedgraph</font></font>
is <a href="https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.2_convert_bam2_bigwig.md#how-macs12-output-wigbedgraph-files-are-produced"><font color="#4f81bd"><b>NOT</b></font><font color="#4f81bd">
normalized to sequencing depth</font></a>. Instead, we are going to
use another nice tool <font color="#4f81bd"><a href="https://github.com/fidelram/deepTools">deeptools</a></font>
for this task. It is a very versatile tool and can do many other
things.</p>
<p>Make a bigWig file:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
install deeptools</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
install -c bioconda deeptools</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
normalized bigWig with Reads Per Kilobase per Million mapped reads
(RPKM)</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-b YAP1.sorted.bam --normalizeUsingRPKM --binSize 30 --smoothLength
300 -p 10 --extendReads 200 -o YAP1.bw</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-b H3K27ac.sorted.bam --normalizeUsingRPKM --binSize 30
--smoothLength 300 -p 10 --extendReads 200 -o H3K27ac.bw</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bamCoverage</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-b Input.sorted.bam --normalizeUsingRPKM --binSize 30 --smoothLength
300 -p 10 --extendReads 200 -o Input.bw</font></font></p>
<p class="first-paragraph">I set <font face="Consolas, serif"><font size="2" style="font-size: 11pt">--extendReads</font></font>
to 200bp which is the fragment length. Why should we extend the
reads? Because in a real ChIP-seq experiment, we fragment the genome
into small fragments of ~200bp, and pull down the protein bound DNA
with antibodies. However, we only sequence the first 36bp(50bp, or
100bp depending on your library) of the pull-down DNA. To
recapitulate the real experiment, we need to extend it to the
fragment size.</p>
<p>You can read more details <font color="#4f81bd"><a href="https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.2_convert_bam2_bigwig.md#why-do-we-need-to-extend-the-reads-to-fragment-length200bp">Why
do we need to extend the reads to fragment length/200bp?</a></font></p>
<h2 class="western"><a name="how-do-i-put-all-the-steps-together"></a>
108.10 How do I put all the steps together?</h2>
<h3 class="western"><a name="shell-script-comes-to-rescue"></a>108.10.1
Shell script comes to rescue</h3>
<p class="first-paragraph">Up until now, we have finished the basic
analyzing steps for ChIP-seq. We aligned the reads to get <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files; we called peaks using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font>;
we generated <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bigWig</font></font>
tracks as well. This is great, but one of the stumbling blocks for
beginners is learning how to reuse the commands we have typed. We do
not want to type the same command time and time again changing the
file names only when new data come in.</p>
<p>The strategy is to write a shell script to chain all the steps
together. see the section <font color="#4f81bd"><a href="https://read.biostarhandbook.com/unix/writing-scripts.html">writing
scripts</a></font> in the handbook.</p>
<p>First, you need to think about how the workflow should be. We will
need to download the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq</font></font>
files and convert them to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files; next, we will call peaks for a pair of IP and Input.
Sometimes, you may get sorted <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files, which can be used directly to call peaks. It is reasonable
that we have two shell scripts for each task, respectively.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#!
/bin/bash</i></font></font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>set</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-euo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pipefail</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
the SRA number to work with</i></font></font></font><br/>
<font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA=$1</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
the reference file path, change to where you put the
reference</i></font></font></font><br/>
<font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;/risapps/reference/bowtie1/hg19&quot;</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##extract
the fastq files</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$SRA</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
step1, quality control of the fastq files</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqc</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fastq</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
step2, algin the reads with bowtie</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-p 10 --best --chunkmbs 320 </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$REF</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-q </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.fastq
-S </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.sam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
step3, convert sam to bam, and index the bam</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
view -bS </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.sam
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sort -@ 4 - -T </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-o </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.sorted.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
index </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.sorted.bam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
remove sam to save space</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${SRA}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.sam</font></font></p>
<p class="first-paragraph">For more on <font face="Consolas, serif"><font size="2" style="font-size: 11pt">set
-euo pipefail</font></font>, read this post <font color="#4f81bd"><a href="http://redsymbol.net/articles/unofficial-bash-strict-mode/">Use
the Unofficial Bash Strict Mode (Unless You Looove Debugging)</a></font>.</p>
<p>Save it to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra2bam.sh</font></font>.
Make it executable <font face="Consolas, serif"><font size="2" style="font-size: 11pt">chmod
u+x sra2bam.sh</font></font>.</p>
<p>execute:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
YAP1</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./sra2bam.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR1810900</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
H3K27ac</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./sra2bam.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949140</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Input</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./sra2bam.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949142</font></font></p>
<p class="first-paragraph">Now, with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra2bam.sh</font></font>
script, it saves you from typing different file names for aligning
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastqs</font></font>
and converting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sam</font></font>
files. Moreover, it can be used to process any number of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra</font></font>
files.</p>
<p>If you have a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra_id.txt</font></font>
file with one <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra</font></font>
id on each line, you can:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>##
run 6 jobs in parallel</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sra_ids.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">parallel</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-j 6 ./sra2bam </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">{}</font></font></font></p>
<p class="first-paragraph">to process all of them.</p>
<p>second shell script <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam2peaks.sh</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#!
/bin/bash</i></font></font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>set</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-euo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pipefail</font></font><br/>
<br/>
<font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">IP_bam=$1</font></font></font><br/>
<font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Input_bam=$2</font></font></font><br/>
<font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">oprefix=$(</font></font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">basename</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${IP_bam}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
.sorted.bam</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-t </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${IP_bam}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${Input_bam}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-n </font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-g hs -p 1e-5 -o </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">_peaks</font></font></p>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra2bam.sh</font></font>
script will generate the indexed <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files, those <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam</font></font>
files can be fed into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam2peaks.sh</font></font>
to call peaks. In our example data set, we only have two IPs and one
Input. We can call peaks by:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
call peaks for H3K27ac</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./bam2peaks.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR949140.sorted.bam SRR949142.sorted.bam</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
call peaks for YAP1</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">./bam2peaks.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
SRR1810900.sorted.bam SRR949142.sorted.bam</font></font></p>
<p class="first-paragraph">Imagine we have a lot of bam files
generated by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sra2bam.sh</font></font>,
and we want to call peaks for them, how should we stream the
workflow?</p>
<p>Because it invovles a pair of files (IP and Input control) for
calling peaks, we need to make a tab delimited <font face="Consolas, serif"><font size="2" style="font-size: 11pt">txt</font></font>
file with pairs of file names on each line.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
bam_names.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR949140.sorted.bam</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   SRR949142.sorted.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1810900.sorted.bam</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  SRR949142.sorted.bam</font></font></p>
<p class="first-paragraph">Now, we can loop over the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bam_names.txt</font></font>
file one by one and call peaks:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
bam_names.txt </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>while</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">read</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-r </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">IP</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Input</font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>do</b></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">./bam2peaks.sh</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${IP}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${Input}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>done</b></font></font></font></p>
<h3 class="western"><a name="arguments-handling-for-shell-script"></a>
108.10.2 Arguments handling for shell script</h3>
<p class="first-paragraph">What we have so far is all good, but what
if you want to make your script more flexiable? e.g.&nbsp;you want to
specify mouse genome for mapping reads. You can add arguments for
your script.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#!
/bin/bash</i></font></font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>set</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-euo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pipefail</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
show help</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">show_help()</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>{</b></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;&lt;
EOF</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 This is a wrapper to align fastq reads to bam files for ChIP-seq
experiments.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 usage: </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${0##</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*/</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-d &lt; a directory path containing the fastq.gz files &gt; -r &lt; h
or m&gt;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       -h display this help and exit</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       -f the full path name of the fastq file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       -r reference genome to be used. m for mouse; h for
human</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">EOF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">}</font></font><br/>
<br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>if</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">there</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
are no arguments provided, show help</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">if
[[ </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$#</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
== 0 ]]; then show_help; exit 1; fi</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
parsing arguments</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">while
getopts &quot;:hf:r:&quot; opt; do</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 case &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$opt</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;
in</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   h) show_help;exit 0;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   f) fq=</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$OPTARG</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   r) REF=</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$OPTARG</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   '?') echo &quot;Invalid option </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$OPTARG</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;;
show_help &gt;&amp;2; exit 1;;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 esac</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">done</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
set up some defaults</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${REF:-</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;h&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">}</font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
check if the fastq file exist</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">if
[ ! -f &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$fq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;
]; then</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 echo &quot;file </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${fq}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
does not exit&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 exit 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fi</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
reference genome path for mouse and
human</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">human_ref=&quot;/risapps/reference/bowtie1/hg19&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mouse_ref=&quot;/risapps/reference/bowtie1/mm9&quot;</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
which species? human or mouse?</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">if
[[ </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$REF</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
== &quot;m&quot; ]]; then</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ref_genome=&quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${mouse_ref}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">elif
[[ </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$REF</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
== &quot;h&quot; ]]; then</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ref_genome=&quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${human_ref}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">else</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 echo &quot;please only specify m or h for the reference genome&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 exit 1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fi</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
extract output name</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">oprefix=</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$(</font></font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">basename</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${fq}</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
.fastq</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
mapping</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie
-p 10 --best --chunkmbs 320 </font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${ref_genome}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-q &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${fq}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;
-S &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;.sam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
convert sam to sorted bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
view -bS &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;.sam
| samtools sort -@ 4 - -T &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;
-o &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;.sorted.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
index sorted bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index &quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;.sorted.bam</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">rm
&quot;</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">${oprefix}</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;.sam</font></font></p>
<p class="first-paragraph">With this script, you can map not only
ChIP-seq data for human but also for mouse. You can surely add more
arguments to set <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bowtie</font></font>
mapping thread numbers and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
sort</font></font> thread numbers (the above script uses 10 and 4
respectively). You can read more on arguments handling for shell
scripts: <font color="#4f81bd"><a href="http://wiki.bash-hackers.org/howto/getopts_tutorial">small
getopts tutorial</a></font>.</p>
<p>Congratulations! You have come this far with me. If you want to
take the next level of making your analysis reproducible and
flexible, read our <font color="#4f81bd"><a href="#ch108.xhtml#makefiles">Using
Makefiles</a></font> section.</p>
<h2 class="western"><a name="what-are-the-black-listed-regions"></a>108.11
What are the black-listed regions?</h2>
<p class="first-paragraph">For ChIP-seq, it is important to filter
artifact regions that has abnomral high signals. From the <font color="#4f81bd"><a href="https://sites.google.com/site/anshulkundaje/projects/blacklists">website
of Anshul Kundaje</a></font> in Stanford University:</p>
<p class="block-text-western">These regions are often found at
specific types of repeats such as centromeres, telomeres and
satellite repeats. It is especially important to remove these regions
that computing measures of similarity such as Pearson correlation
between genome-wide tracks that are especially affected by outliers</p>
<p class="first-paragraph">One can download the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">hg19</font></font>
blacklist:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
https://www.encodeproject.org/files/ENCFF001TDO/@@download/ENCFF001TDO.bed.gz</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
411 regions are black-listed</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">zless</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENCFF001TDO.bed.gz </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">411</font></font></p>
<p class="first-paragraph">Note that more and more sequencing
projects are moving their reference genome to the latest version
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>.
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRCh38</font></font>
blacklist was uploaded by Anshul Kundaje on 2016-10-16. The file can
be downloaded by:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
https://www.encodeproject.org/files/ENCFF419RSJ/@@download/ENCFF419RSJ.bed.gz</font></font></p>
<h2 class="western"><a name="how-do-i-visualize-the-peaks"></a>108.12
How do I visualize the peaks?</h2>
<p class="first-paragraph">We will use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGV</font></font>
to visualize the peaks and raw signal. check our <font color="#4f81bd"><a href="https://read.biostarhandbook.com/visualize/igv.html">previous
section on using IGV</a></font>.</p>
<p>Open <font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGV</font></font>,
click <font face="Consolas, serif"><font size="2" style="font-size: 11pt">File</font></font>,
then <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Load
From File</font></font>, choose the peak files end with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bed</font></font>
and the raw signal files end with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bw</font></font>.
you can change the scale of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bigWig</font></font>
tracks by right click the tracks and choose <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Set
Data Range</font></font>. Now, you can browser through the genome or
go to your favoriate genes.</p>
<p>YAP1 binds to known target gene <i>CTGF</i>: 
<img src="0000_html_59cfb86a5c3f94f6.png" name="Image170" align="bottom" hspace="12" width="560" height="138" border="0"/>
</p>
<p>one other example from the original <i>Nature Cell Biology</i>
paper at the <i>RAD18</i> locus: 
<img src="0000_html_ba9df97d881c41f3.png" name="Image171" align="bottom" hspace="12" width="560" height="136" border="0"/>
</p>
<p>We see that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs</font></font>
did a pretty good job in identifying enriched regions for histone
modifications and transcription factor (TF) binding peaks. However,
we do also see some potential YAP1 binding sites are missed. Fine
tuning the peak calling parameters may improve the results.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="chip-seq-downstream-analysis-1"></a>109
ChIP-Seq Downstream Analysis 1</h1>
<p class="first-paragraph">The author of this guide is <font color="#4f81bd"><a href="https://github.com/crazyhottommy">Ming
Tang</a></font>. The material was developed for the Biostar Handbook.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/crazyhottommy">GitHub
	profile</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://crazyhottommy.blogspot.hu/">Diving
	into Genetics and Genomics</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-compare-the-peak-sets"></a>109.1
How do I compare the peak sets?</h2>
<p class="first-paragraph">Now, we have the peaks called for H3K27ac
and YAP1 and we want to exclude the peaks that overlap with the
blacklisted regions. How should you do it? We will use a very popular
tool <font color="#4f81bd"><a href="http://bedtools.readthedocs.io/en/latest/">bedtools</a></font>
developed by <font color="#4f81bd"><a href="http://quinlanlab.org/">Aaron
Quinlan lab</a></font> for this purpose. Documentation of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>
is of top niche. Also read <font color="#4f81bd"><a href="https://read.biostarhandbook.com/tools/install.html#interval">Interval
analysis tools</a></font> section in the handbook.</p>
<p>There are many sub-commands for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>,
but the most commonly used one is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools
intersect</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
make sure you are inside the folder containing the peak files and the
black-listed region file.</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
unzip the file </i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENCFF001TDO.bed.gz</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
How many H3K27ac peaks overlap with black-listed regions?</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a H3K27ac_peaks.bed -b ENCFF001TDO.bed -wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#14</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
exclude those peaks</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a H3K27ac_peaks.bed -b ENCFF001TDO.bed -v </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
H3K27ac_filtered_peaks.bed </font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
do the same for YAP1</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a YAP1_peaks.bed -b ENCFF001TDO.bed -v </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAP1_filtered_peaks.bed</font></font></p>
<p class="first-paragraph">How many YAP1 peaks overlap with H3K27ac
peaks?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a YAP1_filtered_peaks.bed -b H3K27ac_filtered_peaks.bed
-wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#1882</i></font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a YAP1_filtered_peaks.bed -b H3K27ac_filtered_peaks.bed
-wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#1882</i></font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a H3K27ac_filtered_peaks.bed -b YAP1_filtered_peaks.bed
-wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#1882</i></font></font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a H3K27ac_filtered_peaks.bed -b YAP1_filtered_peaks.bed
-wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-l</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#1772</i></font></font></font></p>
<p class="first-paragraph">what&rsquo;s happening here? why there are
only 1772 unique H3K27ac peaks overlap with YAP1 peaks?</p>
<p>It turns out that bedtools will output the overlapping peaks of
H3K27ac whenever there is an overlapping with YAP1 peaks, and it is
possible that the same H3K27ac peak (tends to be really broad peaks)
overlaps with multiple YAP1 peaks. With that in mind, I always do
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort
| uniq</font></font> following <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font>
command.</p>
<p>You can identify those H3K27ac peaks by:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
intersect -a H3K27ac_filtered_peaks.bed -b YAP1_filtered_peaks.bed
-wa </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">uniq</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-c </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sort</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-k1,1nr </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">head</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr14   93495168    93513756    MACS_peak_8569  935.17</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr20   30276949    30312747    MACS_peak_16217 3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr10   95216260    95244053    MACS_peak_3871  3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr11   65654426    65689186    MACS_peak_5045  3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr14   23441596    23453313    MACS_peak_7876  3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr1    86040764    86052311    MACS_peak_1241  3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr1    86070388    86080738    MACS_peak_1243  2714.64</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr19   13943280    13955945    MACS_peak_13069 3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr19   13956264    13965827    MACS_peak_13070 3100.00</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
chr1    95080161    95091780    MACS_peak_1347  3038.66</font></font></p>
<p class="first-paragraph">We see some H3K27ac peaks can overlap with
up to 4 peaks of YAP1.</p>
<p>If you choose to believe your eyes, let&rsquo;s visualize it in
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">IGV</font></font>:</p>
<p><img src="0000_html_a5f2c1bfc3af14fd.png" name="Image172" align="bottom" hspace="12" width="560" height="195" border="0"/>
</p>
<p>One of the under-appreciated tools is <font color="#4f81bd"><a href="https://bedops.readthedocs.io/en/latest/">BEDOPS</a></font>.
It has many nice features as well and the documentaion is also very
good. Read this biostar post: <font color="#4f81bd"><a href="https://www.biostars.org/p/13516/">Question:
Bedtools Compare Multiple Bed Files?</a></font></p>
<h2 class="western"><a name="how-do-i-annotate-my-peaks"></a>109.2
How do I annotate my peaks?</h2>
<p class="first-paragraph">The next obvious question is where are the
peaks in terms of their genomic context. e.g.&nbsp;What genes are the
peaks associated with? Are some of the peaks located in intergenic
region? You need to <b>annotate</b> your peaks. There are many tools
you can use. You can find a list <font color="#4f81bd"><a href="https://github.com/crazyhottommy/ChIP-seq-analysis#peak-annotation">here</a></font>.</p>
<p>I will show you how to do peak annotation using an R <font color="#4f81bd"><a href="https://bioconductor.org/">bioconductor</a></font>
package <font color="#4f81bd"><a href="http://bioconductor.org/packages/release/bioc/html/ChIPseeker.html">ChIPseeker</a></font>
develeped by <font color="#4f81bd"><a href="https://github.com/GuangchuangYu">Guangchuan
Yu</a></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
load the packages, install them following the instructions in the
links above</i></font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(ChIPseeker)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(TxDb.Hsapiens.UCSC.hg19.knownGene)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(rtracklayer)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;org.Hs.eg.db&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">txdb
&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TxDb.Hsapiens.UCSC.hg19.knownGene</font></font><br/>
<br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
read in the peaks for YAP1</i></font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>import</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1_peaks.bed&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">format
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BED&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GRanges
object with </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2150</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ranges and </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
metadata columns</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">:</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqnames
                ranges strand   </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">|</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">name
    score</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
           </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Rle</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
             </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&lt;</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">IRanges</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&lt;</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Rle</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&lt;</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">character</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&lt;</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">numeric</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">[</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chr1   [  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">959961</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">960483</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">132.10</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chr1   [ </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1295901</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1296269</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">59.44</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chr1   [ </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1992556</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1993176</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">98.23</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chr1   [ </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8963802</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8964418</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">126.02</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chr1   [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16499970</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16500516</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">142.50</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    ...      ...                    ...    ... ...            ...    
  ...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2146</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chrX [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">102911475</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">102912040</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2146</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">73.47</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2147</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chrX [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">103168254</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">103168801</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2147</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">58.64</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2148</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chrX [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">115403519</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">115404015</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2148</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">50.08</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2149</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chrX [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">149252969</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">149253470</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2149</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">61.44</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2150</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
    chrX [</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">153147396</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">153148074</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">]
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font></font><font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2150</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">61.75</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-------</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqinfo</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">:</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">23</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
sequences from an unspecified genome; no seqlengths</font></font><br/>
<br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1_anno&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>annotatePeak</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">tssRegion=</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">),
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                        </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TxDb=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">txdb,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">level
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;gene&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">annoDb=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;org.Hs.eg.db&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                        </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sameStrand
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FALSE</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ignoreOverlap
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">FALSE</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                        </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">overlap
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;TSS&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
some nice visualization you can do</i></font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>plotAnnoPie</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno)</font></font><br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>upsetplot</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">vennpie=</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TRUE</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
check the annotation</i></font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>head</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno))</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 seqnames    start      end width strand        name  score</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1   </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">959961</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">960483</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">523</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">132.10</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1295901</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1296269</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">369</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">59.44</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1992556</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1993176</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">621</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">98.23</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8963802</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8964418</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">617</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">126.02</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16499970</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16500516</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">547</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">142.50</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    chr1 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">17454948</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">17455379</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">432</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">*</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MACS_peak_</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">59.60</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                 annotation geneChr geneStart 
geneEnd</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>Intron</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
(uc001ack</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.2</font></font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">375790</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
intron </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
of </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">35</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
      </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">955503</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">991499</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                          </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>Promoter</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
(</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1kb)
      </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1288071</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1297157</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>Intron</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
(uc001aiq</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.3</font></font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5590</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
intron </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
of </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">17</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
      </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1981909</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2116834</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         Distal Intergenic       </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8921059</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8939151</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         Distal Intergenic       </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16450832</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">16482582</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         Distal Intergenic       </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">17393256</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">17445948</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 geneLength geneStrand geneId distanceToTSS         ENSEMBL SYMBOL</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">35997</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">375790</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4458</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000188157   AGRN</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">9087</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">54587</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">888</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000162576  MXRA8</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
    </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">134926</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5590</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10647</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000067606  PRKCZ</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">18093</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2023</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-24651</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000074800   ENO1</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">31751</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1969</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-17388</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000142627  EPHA2</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">52693</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
         </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">11240</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
        </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-9000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
ENSG00000117115  PADI2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                             GENENAME</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                               agrin</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
     matrix</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">remodelling
associated </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">8</font></font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               protein kinase C zeta</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                  enolase </font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
(alpha)</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                     EPH receptor A2</font></font><br/>
<font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
peptidyl arginine deiminase, type II</font></font><br/>
<br/>
<font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
you can save it to a txt file to your computer</i></font></font></font><br/>
<font color="#ff0000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>&gt;</b></font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>write.table</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno),
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1_peaks_anno.txt&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">row.names
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">F,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">col.names=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">T,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sep
=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">\t</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">quote
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
F)</font></font></p>
<h2 class="western"><a name="how-do-i-do-pathway-enrichment-analysis-for-the-peaks"></a>
109.3 How do I do pathway enrichment analysis for the peaks?</h2>
<p class="first-paragraph">Now, you have the genes that are
associated with the peaks, you can do pathway enrichment analysis
using the genes. I will use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">clusterProfiler</font></font>
package developed by Guangchuan Yu as well:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(clusterProfiler)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
GO term enrichment</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ego
&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>enrichGO</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene
          =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno)</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SYMBOL,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">OrgDb
        =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
org.Hs.eg.db,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">keytype
      =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;SYMBOL&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ont
          =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BP&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pAdjustMethod
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BH&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pvalueCutoff
 =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">qvalueCutoff
 =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
visualization</i></font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>dotplot</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(ego,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">showCategory
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
Kegg pathway enrichment, need the Entrez ID</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">kk&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>enrichKEGG</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gene
   =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno)</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">geneId,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">organism
    =</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'hsa'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
          </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pvalueCutoff
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.05</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>dotplot</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(kk,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">showCategory
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">title
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1
binding pathway enrichment&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
you can write the result to a tsv file</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>write.table</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(kk</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">result,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP_KEGG_pathway_genes.txt&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sep
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">\t</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">col.names
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
T, </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">row.names
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
F, </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">quote
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">F)</font></font></p>
<p class="first-paragraph"><img src="0000_html_ee3f0e868fa093e1.png" name="Image173" align="bottom" hspace="12" width="560" height="373" border="0"/>

Note that we used all the genes that associated with a peak for the
pathway analysis. This may not be optimal, especially when you have a
lot of peaks. One alternative is that you can filter the genes/peaks
by some rules first and then do the enrichment analysis.</p>
<p>e.g.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(dplyr)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
only consider genes within 5000 bp of the
peaks</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">%&gt;%</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">dplyr</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>filter</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>abs</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(distanceToTSS)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
or you can rank the peaks by the Pvalue and choose the top 1000 peaks
(this number is arbitrary..) </font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
score column is the -10*log10 Pvalue</i></font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>as.data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_anno)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">%&gt;%</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">dplyr</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>arrange</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>desc</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(score))
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">%&gt;%</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>head</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">n
=</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font></p>
<p class="first-paragraph">The underlying mechanism is to compare the
peaks associated gene set with the vairous annotated pathway gene
sets to see whether the peak assoicated genes are overpresented in
any of the known gene sets using hypergeometric test. Of course, many
other complex algorithums have been developed for this type of
analysis. Further reading: <font color="#4f81bd"><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0079217#pone-0079217-t001">A
Comparison of Gene Set Analysis Methods in Terms of Sensitivity,
Prioritization and Specificity</a></font></p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://bejerano.stanford.edu/great/public/html/">GREAT</a></font>
	predicts functions of cis-regulatory regions</p>
</ul>
<h2 class="western"><a name="how-do-i-do-motif-analysis-with-the-peaks"></a>
109.4 How do I do motif analysis with the peaks?</h2>
<p class="first-paragraph">One of the most popular tools is
<font color="#4f81bd"><a href="http://meme-suite.org/">MEME-suite</a></font>
for motif enrichment analysis. I will demonstrate how to use
<font color="#4f81bd"><a href="http://meme-suite.org/tools/meme-chip">MEME-ChIP</a></font>
for YAP1 peaks.</p>
<p><font face="Consolas, serif"><font size="2" style="font-size: 11pt">MEME-ChIP</font></font>
needs 500bp DNA sequences around the YAP1 binding summits (with the
highest signal of binding), which is output by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">macs14</font></font>
when you call peaks.</p>
<p>Fetching the 500 bp DNA sequence around the YAP1 summits:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
get the coordinates of 500bp centered on the summit of the YAP1
peaks</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAP1_summits.bed </font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>|</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'$2=$2-249,
$3=$3+250'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
OFS=</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;\t&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAP1_500bp_summits.bed</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
you need a fasta file of the whole genome, and a bed file containing
the coordinates </i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
the whole genome fasta file can be gotten by:</i></font></font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">rsync</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
-avzP rsync://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/
.</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
*fa.gz </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
UCSC_hg19_genome.fa.gz</font></font><br/>
<font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gunzip</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
UCSC_hg19_genome.fa.gz</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Use betools get fasta
http://bedtools.readthedocs.org/en/latest/content/tools/getfasta.html
</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedtools</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
getfasta -fi UCSC_hg19_genome.fa -bed YAP1_500bp_summits.bed -fo
YAP1_500bp.fa</font></font></p>
<p class="first-paragraph">Now, upload it to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MEME-ChIP</font></font>:</p>
<p><img src="0000_html_a946be63658dbe0e.png" name="Image174" align="bottom" hspace="12" width="560" height="603" border="0"/>

I will just keep the defaults for all the options. It takes a while
to finish.</p>
<p>Output from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MEME-ChIP</font></font>:</p>
<p><img src="0000_html_ab22914206f787bc.png" name="Image175" align="bottom" hspace="12" width="560" height="177" border="0"/>
</p>
<p>YAP1 is not a DNA binding transcription factor itself, rather it
is in a complex with TEAD, which is a DNA binding transcription
factor. The first motif is a TEAD binding motif in tandem, and the
second motif is AP-1 binding motif.</p>
<p>One downside of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MEME-ChIP</font></font>
is that you can only input the DNA sequences with the same length.
Instead, you can use Homer <font color="#4f81bd"><a href="http://homer.salk.edu/homer/ngs/peakMotifs.html">findMotifsGenome.pl</a></font>
with the length of the peaks for finding motifs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">findMotifsGenome.pl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAP1_filtered_peaks.bed hg19 YAP1_motifs -size given</font></font></p>
<p class="first-paragraph"><img src="0000_html_47a843791cf57adc.png" name="Image176" align="bottom" hspace="12" width="560" height="260" border="0"/>
</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Homer</font></font>
results are in consistence with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">MEME-ChIP</font></font>
results.</p>
<h2 class="western"><a name="how-to-call-differential-peaks"></a>109.5
How to call differential peaks?</h2>
<p class="first-paragraph">One of the other frequent questions
biologists ask is how the peaks change in different conditions.
e.g.&nbsp;different treatment of the cells, different subtypes of the
same cancer. To answer this question, you&rsquo;d better have
biological replicates of each condition.</p>
<p>There is a review paper on this topic: <font color="#4f81bd"><a href="http://bib.oxfordjournals.org/content/early/2016/01/12/bib.bbv110.short?rss=1">A
comprehensive comparison of tools for differential ChIP-seq analysis</a></font>.
You can choose tools based on various rules:</p>
<p><img src="0000_html_748f841411ef4be4.png" name="Image177" align="bottom" hspace="12" width="560" height="311" border="0"/>
</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="chip-seq-downstream-analysis-2"></a>110
ChIP-Seq Downstream Analysis 2</h1>
<p class="first-paragraph">The author of this guide is <font color="#4f81bd"><a href="https://github.com/crazyhottommy">Ming
Tang</a></font>. The material was developed for the Biostar Handbook.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://github.com/crazyhottommy">GitHub
	profile</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://crazyhottommy.blogspot.hu/">Diving
	into Genetics and Genomics</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-generate-a-heatmap-with-chip-seq-data"></a>
110.1 How do I generate a heatmap with ChIP-seq data?</h2>
<p class="first-paragraph">Heatmap is of no mystery. It is a way to
visualize the data a.k.a. using colors to represent values. However,
one really needs to understand the details of heatmaps. I recommend
you to read <font color="#4f81bd"><a href="http://www.nature.com/nmeth/journal/v9/n8/full/nmeth.2134.html">Mapping
quantitative data to color</a></font> and <font color="#4f81bd"><a href="http://www.nature.com/nmeth/journal/v9/n3/full/nmeth.1902.html">Heat
maps</a></font> from a series of articles from <i>Nature Methods</i>.</p>
<p>Usually one has a matrix and then plot the matrix using R
functions such as <font color="#4f81bd"><a href="https://cran.r-project.org/web/packages/gplots/index.html">heatmap.2</a></font>,
<font color="#4f81bd"><a href="https://cran.r-project.org/web/packages/pheatmap/index.html">pheatmap</a></font>
or <font color="#4f81bd"><a href="https://github.com/jokergoo/ComplexHeatmap">Heatmap</a></font>.
With those R packages, it is very easy to make heatmaps, but you do
need to read the documentations of the packages carefully to
understand the details of the arguments. Read <font color="#4f81bd"><a href="https://rpubs.com/crazyhottommy/a-tale-of-two-heatmap-functions">a
tale of two heatmap functions</a></font>.</p>
<p>For ChIP-seq data, one wants to plot ChIP-seq signal around the
genomic features(promoters, CpG islands, enhancers, and gene bodies).
To do this, you need to first take the regions of say 5kb upstream
and downstream of all (around 30000) the transcription start sites
(TSS), and then bin each 10kb region to 100 bins (100 bp per bin).
Count the ChIP-seq signal (reads number) in each bin. Now, you have a
data matrix of 30000 (promoters) x 100 (bin), and you can plot the
matrix using a heatmap function mentioned above.</p>
<p>You can of course do it from scratch, but one of the taboos of
bioinformatics is re-inventing the wheel. Most likely, someone has
written a package for this type of analysis. Indeed, check this
<font color="#4f81bd"><a href="https://www.biostars.org/p/180314/">biostar
post</a></font> for all the available tools. Choose the right tool
for yourself, if you are not that familiar with R, you can use some
GUI tools. <font color="#4f81bd"><a href="http://easeq.net/">EaSeq</a></font>
is a pretty powerful tool for windows users.</p>
<p>I personally like <font color="#4f81bd"><a href="https://bioconductor.org/packages/release/bioc/html/EnrichedHeatmap.html">EnrichmentHeatmap</a></font>
by Zuguang Gu the most because it gives me the most flexiability to
control the looks of my heatmap. It is based on <font color="#4f81bd"><a href="https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html">ComplexHeatmap</a></font>,
which I highly recommend you to learn how to use. Below, I will walk
you through how to make a heatmap with the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">EnrichmentHeatmap</font></font>
package.</p>
<p>First, let&rsquo;s read in the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(EnrichedHeatmap)
</font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
for making heatmap</i></font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(rtracklayer)
 </font></font><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
for reading bigwig and bed files</i></font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(GenomicRanges)</font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
read in the bed peak files to GRanges object</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H3K27ac.bed&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>import</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;~/ChIP-seq/results/peaks/H3K27ac_peaks.bed&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">format
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BED&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1.bed&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>import</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;~/ChIP-seq/results/peaks/YAP1_peaks.bed&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">format
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BED&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
read in bigwig files to GRanges object, it can be slow. bigwig is
several hundred MB</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
you can use which argument to restrict the data in certain
regions.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H3K27ac.bw&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>import</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;~/ChIP-seq/results/bams/H3K27ac.bw&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">format
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BigWig&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1.bw&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>import</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;~/ChIP-seq/results/YAP1.bw&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">format
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;BigWig&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font></p>
<p class="first-paragraph">We want to plot the H3K27ac signal and
YAP1 signal 5kb flanking the center of YAP1 peaks.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.10</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kb&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>resize</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.bed,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">width
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fix
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;center&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
take only the center of the peaks</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.10</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kb.center&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>resize</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.10</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kb,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">width
=</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fix
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;center&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1.mat&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>normalizeToMatrix</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.bw,
YAP1</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.10</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kb.center,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">value_column
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;score&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                 </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mean_mode=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;w0&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">w=</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">100</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">extend
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H3K27ac.mat&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>normalizeToMatrix</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.bw,
YAP1</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">.10</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kb.center,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">value_column
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;score&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                                 </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mean_mode=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;w0&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">w=</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">100</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">extend
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">5000</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
check data range</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>quantile</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">probs
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.005</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.90</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>quantile</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">probs
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.005</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.5</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.90</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
mapping colors</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(circlize)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
from the quantile, I choose the color mapping range</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">col_fun_YAP&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">circlize</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colorRamp2</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">20</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">),
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;white&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;red&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">col_fun_H3K27ac&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">circlize</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colorRamp2</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">100</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">),
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;white&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;red&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))</font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>EnrichedHeatmap</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">axis_name_rot
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">name
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">column_title
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">use_raster
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TRUE</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">col
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
col_fun_YAP,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">top_annotation
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>HeatmapAnnotation</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">lines
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>anno_enriched</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">()),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">top_annotation_height
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>unit</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;cm&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>EnrichedHeatmap</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">axis_name_rot
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">name
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;H3K27ac&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">column_title
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;H3K27ac&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">use_raster
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TRUE</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">col
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
col_fun_H3K27ac,</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">top_annotation
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>HeatmapAnnotation</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">lines
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>anno_enriched</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">()),</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
               </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">top_annotation_height
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>unit</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;cm&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))</font></font></p>
<p class="first-paragraph"><img src="0000_html_5d8071523ad0cf3a.png" name="Image178" align="bottom" hspace="12" width="560" height="560" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-generate-a-meta-profile-plot-with-chip-seq-data"></a>
110.2 How do I generate a meta profile plot with ChIP-seq data?</h2>
<p class="first-paragraph">Although we used <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HeatmapAnnotation(lines
= anno_enriched())</font></font> to add a line plot on top of the
heatmap, it may not easy to put the two line graph together. It is
very common to compare signal strength for the same ChIPed factor in
different conditions (treatment vs control, tumor vs normal etc). How
can we do it?</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1_mat</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">H3K27ac_mat</font></font>
are just like regular matrix, you can use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">colMeans</font></font>
to get the average and plot the average in the same figure with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ggplot2</font></font>.
We can even add 95% confidence intervals onto the graph.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">YAP1_mean&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">avg
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colMeans</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat),
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CI_lower
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>apply</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
Hmisc</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">smean.cl.normal)[</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CI_upper
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>apply</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
Hmisc</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">smean.cl.normal)[</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,])
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">%&gt;%</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>mutate</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">factor
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;YAP1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pos
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colnames</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1.mat))
</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H3K27ac_mean&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>data.frame</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">avg
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colMeans</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat),
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CI_lower
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>apply</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
Hmisc</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">smean.cl.normal)[</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,],</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
                         </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CI_upper
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>apply</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat,
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
Hmisc</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">::</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">smean.cl.normal)[</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,])
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">%&gt;%</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>mutate</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">factor
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;H3K27ac&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pos
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>colnames</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(H3K27ac.mat))</font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>library</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(tidyverse)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">combine_both&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>bind_rows</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(YAP1_mean,
H3K27ac_mean)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
change position to factor and order it </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">combine_both</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pos&lt;-</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>factor</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(combine_both</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pos,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">levels=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
YAP1_mean</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pos)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
without confidence interval</font></font><br/>
<br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ggplot</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(combine_both,
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>aes</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pos,</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
avg, </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">group
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
factor)) </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>geom_line</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>aes</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">color
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
factor)) </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>theme_bw</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">base_size
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">14</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>theme</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">axis.ticks.x
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>element_blank</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">())
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>scale_x_discrete</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">breaks
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;u1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;d1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;d50&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">),
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">labels
=</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;-5kb&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;center&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;5kb&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>xlab</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NULL</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ylab</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;RPKM&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ggtitle</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;ChIP-seq
signal&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">##
take some touch up to finalize the figure</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ggplot</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(combine_both,
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>aes</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">x
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
pos,</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">y
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
avg, </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">group
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
factor)) </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>geom_line</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>aes</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">color
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
factor)) </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>geom_ribbon</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>aes</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ymin=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
CI_lower, </font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ymax=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CI_upper),
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">alpha=</font></font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.2</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">col
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;#8B7E66&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>theme_bw</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">base_size
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#40a070"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">14</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>theme</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">axis.ticks.x
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>element_blank</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">())
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>scale_x_discrete</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">breaks
=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;u1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;d1&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;d50&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">),
</font></font><font color="#902000"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">labels
=</font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>c</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;-5kb&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;center&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">,
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;5kb&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">))
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>xlab</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NULL</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ylab</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;RPKM&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">+</font></font></font><br/>
<font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
       </font></font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>ggtitle</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">(</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;ChIP-seq
signal&quot;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">)</font></font></p>
<p class="first-paragraph"><img src="0000_html_7504b9e4094493f7.png" name="Image179" align="bottom" hspace="12" width="560" height="560" border="0"/>

<img src="0000_html_d4e45b1c26e0bd92.png" name="Image180" align="bottom" hspace="12" width="560" height="560" border="0"/>
</p>
<p>As you can see, when you really know the power of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">R</font></font>
programming language, you do much more!</p>
<h2 class="western"><a name="where-can-i-get-public-available-chip-seq-data-sets"></a>
110.3 Where can I get public available ChIP-seq data sets?</h2>
<h3 class="western"><a name="encode"></a>110.3.1 ENCODE</h3>
<p class="first-paragraph">It is not complete for a ChIP-seq chapter
without mentioning the <font color="#4f81bd"><a href="https://www.encodeproject.org/">ENCODE
project</a></font>. Many ChIP-seq data sets from various cell types
can be found there. In addtion, ENCODE has many other data sets such
as ChIA-PET, Hi-C and DNase-seq.</p>
<h3 class="western"><a name="ucsc-genome-browser"></a>110.3.2 UCSC
genome browser</h3>
<p class="first-paragraph">Every biologist need to know how to use
<font color="#4f81bd"><a href="https://genome.ucsc.edu/">UCSC genome
browser</a></font>! It is one of the most popular web genome
browsers.</p>
<p>If you scroll to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Regulation</font></font>
tracks (hg19 version). There are many ChIP-seq data avaiable for you
to browse. 
<img src="0000_html_207bfec382118ffd.png" name="Image181" align="bottom" hspace="12" width="560" height="186" border="0"/>
</p>
<p>click <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ENC
TF Binding</font></font>:</p>
<p><img src="0000_html_764d284b3434ae90.png" name="Image182" align="bottom" hspace="12" width="560" height="192" border="0"/>
</p>
<p>click <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HAIB
TFBS</font></font>:</p>
<p><img src="0000_html_cd06ac455db728f2.png" name="Image183" align="bottom" hspace="12" width="560" height="347" border="0"/>
</p>
<p>Now you can check various boxes to make it show up in the browser.
quite useful! Many of the data sets are from ENCODE.</p>
<p>I recommend you to watch the tutorials from <font color="#4f81bd"><a href="http://www.openhelix.com/ucsc">openhelix</a></font>
to take full advantage of this great resource.</p>
<h3 class="western"><a name="cistrome"></a>110.3.3 Cistrome</h3>
<p class="first-paragraph"><font color="#4f81bd"><a href="http://www.cistrome.org/Cistrome/Cistrome_Project.html">Cistrome</a></font>
is a project maintained by Sheirly Liu&rsquo;s lab in Harvard. It has
a lot more data sets including ENCODE and other public data sets from
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/gds/">GEO</a></font>.
Some features are very friendly to wet biologists.</p>
<h3 class="western"><a name="geo-and-ena"></a>110.3.4 GEO and ENA</h3>
<p class="first-paragraph">Of course, if you want to adventure
yourself with the the raw data in a publication, you can always go to
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/gds/">GEO</a></font>
to get the SRA file and convert to fastq. Alternatively, you can
directly download fastqs from <font color="#4f81bd"><a href="http://www.ebi.ac.uk/ena/browse/read-download">European
Nucleotide Archive (ENA)</a></font>.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-sequence-assembly"></a>111
Introduction to sequence assembly</h1>
<h2 class="western"><a name="what-does-sequence-assembly-mean"></a>111.1
What does &ldquo;sequence assembly&rdquo; mean?</h2>
<p class="first-paragraph">Assembly is a &ldquo;catch-all&rdquo; term
used to describe methods where we combine shorter individual
measurements called &ldquo;reads&rdquo; into longer contiguous
sequences typically called &ldquo;contigs.&rdquo;</p>
<p class="block-text-western">ASSEMBLY is the process of turning
shorter READS into longer CONTIGS.</p>
<p class="first-paragraph">Because the sequencing process works by
&ldquo;breaking&rdquo; the original DNA into smaller fragments, the
assembly process is conceptually similar to putting together an image
puzzle from its many pieces.</p>
<p>The software that performs the assembly is called the &ldquo;assembler.&rdquo;</p>
<h2 class="western"><a name="what-applications-does-assembly-have"></a>
111.2 What applications does assembly have?</h2>
<p class="first-paragraph">First, there is the &ldquo;obvious&rdquo;
use of identifying the base composition of longer DNA segments. Since
the majority of organisms have not yet been sequenced, thus there is
plenty to do there.</p>
<p>Interestingly enough there is a resurgence of interest in using
assembly methods to identify variants of already known genomes.
Alignment based variation calling is inadequate whenever the length
of the genomic changes is comparable to measurement lengths (read
lengths). The reads spanning variations of this size often fail to
align altogether - or even worse; they align in another location. By
assembling these reads into longer contigs, you can identify more
substantial differences in the genomes.</p>
<h2 class="western"><a name="what-are-the-challenges-of-assembly"></a>
111.3 What are the challenges of assembly?</h2>
<p class="first-paragraph">Sequence assembly is perhaps the
application domain of bioinformatics where &ldquo;skill&rdquo; and
&ldquo;expertise&rdquo; are the most difficult to identify and
define.</p>
<p>As you will see, it is a field where the procedural method
descriptions <i>&ldquo;this is the command we used to assemble the
genome&rdquo;</i> hide the staggering complexity and challenges of
finding the particular command reported to work well. Assemblers are
quite unlike any other software tool you will ever use. Most come
with a bewildering array of parameters - the purpose of which are not
explained, yet many will have profound effects on the results that
they produce.</p>
<p>Trial and error are one of the most commonly used strategies - you
will have to keep tuning the parameters and rerun the entire process
hoping that the results improve - sometimes in vain. As it turns out,
genome assembly is the most computational demanding bioinformatics
method of them all. Assembling a large genome may take even weeks(!)
and substantial computational resources. Thus any expertise built on
trial and error will have to be accumulated over a much more extended
period.</p>
<p>Finally, even when assembly appears to work, almost always it will
contain several severe and substantial errors. That is where, in our
opinion, bioinformatics expertise matters more. The ability to
understand, visualize and correct the mistakes of an assembly has a
utility that will outlast the present and is more valuable than
knowing the exact invocation of a tool by heart.</p>
<h2 class="western"><a name="what-is-the-white-whale-of-assembly"></a>
111.4 What is the &ldquo;white whale&rdquo; of assembly?</h2>
<p class="first-paragraph">In reference to Moby Dick, the &ldquo;white
whale&rdquo; is something someone chases and chases then becomes
obsessed with, but in the end, they can never get it.</p>
<p>The white whale of assembly is obtaining &ldquo;single linear
genome.&rdquo; What it means is that you hope that even when starting
with the many millions of short reads you will end up with one,
unique, unambiguous, contiguous long sequence that corresponds to the
truth, the &ldquo;reference&rdquo; genome that other scientists will
now refer to in their work.</p>
<p>The reason that this is a white whale is that there is no such
thing as a single linear best, representative genome. The better we
understand large-scale genomic variation, the more apparent it
becomes that no single reference genome could characterize each
member of the same species. Instead of a single genome, the new
school of thought is to approach the problem in terms of a network of
genomes, and within that framework identify the path that describes
our genome in this much larger graph space. Hence our goal should be
to identify the segments and connectors in this graph space and worry
less about the concept of single linear &ldquo;reference&rdquo;. For
a higher level overview we recommend the following article that
describes the upcoming shift in the way we think about genomes:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.statnews.com/2016/10/07/dna-genome-sequencing-new-maps/">As
	DNA reveals its secrets, scientists are assembling a new picture of
	humanity</a></font></p>
</ul>
<p class="first-paragraph"><img src="0000_html_5194ca57ff0156a0.jpg" name="Image184" align="bottom" hspace="12" width="560" height="315" border="0"/>
</p>
<h2 class="western"><a name="what-can-be-assembled"></a>111.5 What
can be &ldquo;assembled&rdquo;?</h2>
<p class="first-paragraph">For individual organisms scientist
typically perform a:</p>
<ul>
	<li/>
<p class="compact">Genome assembly: establish the DNA
	composition of the organism</p>
	<li/>
<p class="compact">Transcriptome assembly: build the
	transcript (RNA) expression for the organism</p>
</ul>
<p class="first-paragraph">Applications to metagenomics:</p>
<ul>
	<li/>
<p class="compact">Meta-genome assembly.</p>
	<li/>
<p class="compact">Meta-transcriptome assembly.</p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="concepts-in-sequence-assembly"></a>112
Concepts in sequence assembly</h1>
<h2 class="western"><a name="how-do-assemblers-work"></a>112.1 How do
&ldquo;assemblers&rdquo; work?</h2>
<p class="first-paragraph">The last decade has brought substantial
innovation to this field, from so-called &ldquo;overlap layout&rdquo;
algorithms that use multiple sequence alignments to k-mer and
graph-based algorithms that attempt to find the shortest path that
connects shorter subsequences.</p>
<p>At the same time, different instrumentation requires radically
different approaches - it is unlikely that a tool designed for
Illumina data would perform well on reads produced by a PacBio
instrument.</p>
<h2 class="western"><a name="what-makes-genome-assembly-difficult"></a>
112.2 What makes genome assembly difficult?</h2>
<p class="first-paragraph">Since there are only four bases in the DNA
and because genomes contain repetitive elements some regions may
allow for many different yet seemingly equally good alternatives in
which these shorter segments could be pieced back together. In
essence, there could be multiple, similarly good solutions to the
puzzle.</p>
<p>Assembly processes are more sensitive to the read lengths and the
systematic errors of the sequencing process. Short reads may contain
fewer identifiable regions. Systematic errors may accumulate and
support spurious results.</p>
<p class="block-text-western"><b>Note</b>: Assembly as an application
domain of bioinformatics is immature. The success of an assembly
depends on (potentially hidden) attributes of the data.</p>
<p class="block-text-western">It is not uncommon for software to
produce radically different outcomes when slightly modifying their
settings and parameters. The algorithms require several levels of
decision making, often with far-reaching consequences.</p>
<h2 class="western"><a name="beyond-technical-limitations-what-else-makes-genome-assembly-difficult"></a>
112.3 Beyond technical limitations what else makes genome assembly
difficult?</h2>
<p class="first-paragraph">Sequencing measurements originating from
non-identical genomic sequences cause a variety of problems. Diploid
organism already contains two slightly different copies of DNA.
Various members of the same species may exhibit even more divergence.
These difference make the job of an assembly program even harder as
it tries to reconcile and consolidate these differences into a single
consensus sequence. Also, certain regions produce much lower
coverages than others leading to uneven coverage. Collecting more
data may not be sufficient to address the systematic problems of this
type.</p>
<h2 class="western"><a name="how-good-is-good-enough"></a>112.4 How
good is good enough?</h2>
<p class="first-paragraph">Let us make something very clear from the
beginning. No genome assembly is <i>entirely correct</i>. Even to
this day, vast tracts of the human genome remain unknown; different
genomic builds and releases are used to publish ever more accurate
genomes of even model organisms.</p>
<p>Your goal is always to define &ldquo;how good is good enough.&rdquo;
What quality of the assembly would you need to answer questions under
study? That should be your driving motivation rather than producing a
single, linear sequence that corresponds to a whole genome.</p>
<h2 class="western"><a name="is-data-quality-control-more-important"></a>
112.5 Is data quality control more important?</h2>
<p class="first-paragraph">Data consistency has a more profound
effect on assembly than for resequencing. When aligning against a
reference, your aligner has additional information that can assist in
making the right decisions. In contrast when assembling from scratch
errors and mistakes compound and may lead to substantially lower
performance.</p>
<h2 class="western"><a name="does-the-sequencing-platform-matter"></a>
112.6 Does the sequencing platform matter?</h2>
<p class="first-paragraph">Assembly as a field of science got
jump-started by the Illumina sequencers. Most of the published
results, algorithms and evaluations refer to assembling data from
Illumina instruments.</p>
<p>Long read technologies such as PacBio and MinION produce reads
that are substantially longer. In a manner of speaking, they get a
&ldquo;head start&rdquo; on assembly. At the same time since the
error rate on these platforms are considerably higher, radically
different algorithmic approaches are required to overlap and extend
these measurements. &ldquo;Classical&rdquo; methods may not be
appropriate at all.</p>
<h2 class="western"><a name="what-is-a-hybrid-assembly"></a>112.7
What is a hybrid assembly?</h2>
<p class="first-paragraph">A hybrid assembly is a process that uses
data from different platforms and tries to make use of the strengths
of each. For example, the Illumina platform provides higher coverage
data whereas the PacBio platforms produce longer reads that span over
longer intervals. Combining these data can significantly improve the
assembly.</p>
<h2 class="western"><a name="does-more-data-lead-to-better-assembly"></a>
112.8 Does more data lead to better assembly?</h2>
<p class="first-paragraph">While intuitively we may assume that more
data gives better chances to find overlaps in practice more data does
not guarantee better assembly. The coverage of the genomes will
eventually yield diminishing returns.</p>
<p>In general assembly, processes will perform better with long and
paired-end reads. The data quality (having DNA from a single genome)
and genome&rsquo; complexity have the most pronounced effects. The
less complex (more unique) a region of a genome is the more difficult
will be to assemble it. We mentioned before how &ldquo;complexity&rdquo;
is used backward in bioinformatics. Scientists in this field call
repetitive and redundant information as being more complex &hellip;
go figure &hellip; not the first and not last of the oddities.</p>
<h2 class="western"><a name="what-does-a-genome-assembly-rely-on"></a>
112.9 What does a genome assembly rely on?</h2>
<p class="first-paragraph">The current experimental procedures for
genome assembly demand that some of our measurements (reads) overlap
with each other. The assembler attempts to detect the overlaps and
will try to piece and extend the sequences by overlapping the
identical (or similar) regions.</p>
<p>For example, suppose that the following were your &ldquo;sequencing
reads&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATCGAAGAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GGAAATACATTTATTA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GTACAAACATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGAATAAGGAAAG</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">CATAGAAGGAGGAAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AGGAAAGATGGCATTA</font></font></p>
<p class="first-paragraph">Then your assembler may find what it
believes to be the most &ldquo;optimal&rdquo; way to overlap this
data and may consider the following as the best arrangement:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATCGAAGAA
                           GTACAAACATA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      AGAATAAGGAAAG                          CATAGAAGGAGGAAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
            AGGAAAGATGGCATTA                          
GGAAATACATTTATTA</font></font></p>
<p class="first-paragraph">And from the overlaps above it forms two
&ldquo;contigs&rdquo;:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATCGAAGAATAAGGAAAGATGGCATTA
        GTACAAACATAGAAGGAGGAAATACATTTATTA</font></font></p>
<p class="first-paragraph">Thus the six reads have been assembled
into two contigs. Essential to be aware of the following limitations:</p>
<ol>
	<li/>
<p class="compact">Contigs may get assembled on different
	strands.</p>
	<li/>
<p class="compact">Contigs do not indicate the order in which
	they follow each other in the genome.</p>
</ol>
<p class="first-paragraph">Then note that it instead of overlapping
reads <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>,<font face="Consolas, serif"><font size="2" style="font-size: 11pt">4</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
our assembler could have overlapped just <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">6</font></font>
like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATCGAAGAA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
      AGGAAAGATGGCATTA</font></font></p>
<p class="first-paragraph">To produce a shorter but different contig:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATATCGAAGAAAAGATGGCATTA</font></font></p>
<p class="first-paragraph">This contig was built at the cost of
discarding read <font face="Consolas, serif"><font size="2" style="font-size: 11pt">2</font></font>.
The series of choices that the &ldquo;assembler&rdquo; makes along
the process will always have substantial implications on the final
result.</p>
<h2 class="western"><a name="why-are-repetitive-regions-challenging-to-assemble"></a>
112.10 Why are repetitive regions challenging to assemble?</h2>
<p class="first-paragraph">Sequencing libraries are built by
fragmenting the original DNA into smaller pieces. When regions are
repetitive, it becomes difficult to tell if a repeated part comes
from the same or a different location. If your reads looked like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TROLLOLLOOTROLLOL</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">LOOTROLLOLLOTROLLOLL</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ROLLOLLOTROLLOLLO</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">OLLOLLOOTROLLOLLOTROLLO</font></font></p>
<p class="first-paragraph">Each of the following contigs below could
be a valid assembly:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">TROLLOLLOOTROLLOLLOOTROLO</font></font></p>
<p class="first-paragraph">or</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ROLLOLLOOTROLLOLLOTROLLOLLOOTROLLOLLOOTROLLOLLO</font></font></p>
<p class="first-paragraph">or</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">OLLOLLOOTROLLOLLOOTROLLOLLOTROLLOLLOOTROLLOLLOOTROLLOLLO</font></font></p>
<p class="first-paragraph">Plus many other valid alternatives may
exist. Note how the length of the read becomes essential in
identifying the minimal &ldquo;repeat&rdquo; structure. The read
length has to be comparable to the repeat length to be able to
resolve the repeats reliably.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="redo-gage-a-critical-evaluation-of-genome-assemblies-and-assembly-algorithms."></a>
113 Redo: GAGE A critical evaluation of genome assemblies and
assembly algorithms.</h1>
<p class="first-paragraph">We are interested in accessing the data
for the study titled:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/22147368">GAGE:
	A critical evaluation of genome assemblies and assembly algorithms</a></font></p>
</ul>
<p class="first-paragraph">The paper was published in 2012 in the
Genome Research journal. Specifically, we will discuss the results
for the sequence data deposited for the bacteria <i>Staphylococcus
Aureus</i> known for its potential to cause various skin and soft
tissue infections.</p>
<p><img src="0000_html_330021ceee7193a9.png" name="Image185" align="bottom" hspace="12" width="560" height="427" border="0"/>
</p>
<p>We do like this paper and do no doubt that the authors of it were
(and probably still are) the most skilled &ldquo;assemblers&rdquo; on
the planet. That being said we have to point out what we think are
fundamental weaknesses of the whole approach. The whole paper
suggests that assembly is a &ldquo;procedural&rdquo; task. They
publish scripts with hard-coded and parameters that are not
explained. With they imply that you could download these scripts, use
the same parameters on your data to get results of comparable
quality.</p>
<p>Alas, that is not how assembly works. The recipes that they come
up with match the &ldquo;data&rdquo; - they have most likely obtained
these parameters after a lengthy process that itself is not
adequately discussed - the rationale for picking each parameter is
not mentioned. We believe that hidden properties of their data are
&ldquo;baked into&rdquo; the parameters themselves. We believe that
it is unlikely that these same parameters would produce optimal or
even acceptable results on your data. Moreover, we want to note how
there are no scripts that would help readers properly evaluate and
correct the assemblies - those steps are left mostly unexplained.</p>
<p>That being said we consider this paper as a recommended reading to
anyone interested in deepening their understanding of this field.</p>
<h2 class="western"><a name="where-is-the-supporting-material"></a>113.1
Where is the supporting material?</h2>
<p class="first-paragraph">Commendably the authors of the publication
maintain a website to distribute their results:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://gage.cbcb.umd.edu/index.html">http://gage.cbcb.umd.edu/index.html</a></font></p>
</ul>
<h2 class="western"><a name="how-does-gage-differ-from-the-other-assembly-bake-offs"></a>
113.2 How does GAGE differ from the other assembly bake-offs?</h2>
<p class="first-paragraph">Both the question and the answers are
lifted right from their web page and are reproduced verbatim because
we find them amusingly combative and &ldquo;in-your-face&rdquo;:</p>
<ul>
	<li/>
<p class="compact">&ldquo;<i>GAGE is being run by assembly
	experts.</i> Our team has assembled hundreds of genomes and has
	written some of the leading genome assembly software. We have been
	evaluating assemblers for more than 10 years. All of the assemblies
	and the comparisons among them will be conducted by experts.&rdquo;</p>
	<li/>
<p class="compact">&ldquo;<i>All natural ingredients</i>. GAGE
	will use FOUR different whole-genome shotgun data sets, all from
	recent sequencing projects. Assemblathon and dnGASP will both use
	simulated data. Who can say how simulated data relates to real
	results? We prefer the real thing.&rdquo;</p>
	<li/>
<p class="compact">&ldquo;<i>Completely open protocols</i>. We
	will compare multiple genome assembly programs, and we will describe
	all the parameters used to run them. We will also explain all the
	steps we take in cleaning up the data (pre-processing) and scrubbing
	the output of the assembly programs (post-processing). All of our
	results will thus be reproducible by anyone with sufficient
	computing resources.&rdquo;</p>
</ul>
<h2 class="western"><a name="how-do-i-get-the-data-for-the-paper"></a>
113.3 How do I get the data for the paper?</h2>
<p class="first-paragraph">As always any analysis reproducibility
starts with obtaining the data itself. Helpfully the GAGE paper lists
the accession numbers for the data. In this case, let&rsquo;s focus
on <i>Staphylococcus aureus</i> datasets listed as shown below:</p>
<p><img src="0000_html_548191bace77be5b.png" name="Image186" align="bottom" hspace="12" width="560" height="203" border="0"/>
</p>
<p><b>Except the information above is not correct!</b> Oh yes, the
curse of reproducibility strikes again.</p>
<p>There is a typo in the paper. The number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRX016063</font></font>
was entered for two entries, for two different organisms! See it for
yourself:</p>
<p><img src="0000_html_cfd1d27778b355be.png" name="Image187" align="bottom" hspace="12" width="560" height="203" border="0"/>
</p>
<p>Typos are nothing new - what is unique here is that the paper
offers no other indication as to what the &ldquo;correct&rdquo;
accession number is. Hence from the publication alone, it is
impossible to identify what data has been analyzed!</p>
<p>Well, well, one more indication (if you needed more) on just how
complicated data reproducibility is. Even authors who are acutely
aware of its importance, who make a concerted effort to publish
everything that they do and create detailed descriptions of their
methods may commit errors that render the entire process <i>impossible
from the very beginning</i>.</p>
<p>If anything this is a flaw of the old publishing process itself.
Essential numbers should not need to be manually be embedded in a
paper. Data should be cross-referenced, labeled automatically to make
errors like this standout.</p>
<p>Did we find the correct SRR number? In this case, yes. Thanks to
the author&rsquo;s foresight to create another, standalone site to
distribute their data we were able to figure the accession number
eventually &mdash; yet another example of how bioinformatics requires
problem-solving of unexpected nature.</p>
<p>Visit the GAGE data site <font color="#4f81bd"><a href="http://gage.cbcb.umd.edu/data/">http://gage.cbcb.umd.edu/data/</a></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the short jump file: </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
http://gage.cbcb.umd.edu/data/Staphylococcus_aureus/Data.original/shortjump_1.fastq.gz`</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Look at the first read name</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzcat
shortjump_1.fastq.gz | head -1 </font></font>
</p>
<p class="first-paragraph">It will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">@SRR022865.7</font></font></p>
<p class="first-paragraph">From this SRR run number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022865</font></font>
we can find the real experiment id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRX007711</font></font>:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/SRX007711">https://www.ncbi.nlm.nih.gov/sra/SRX007711</a></font></p>
</ul>
<p class="first-paragraph">Finally, we find that:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRX007714">https://www.ncbi.nlm.nih.gov/sra/?term=SRX007714</a></font>
	that corresponds to run id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022868</font></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRX016063">https://www.ncbi.nlm.nih.gov/sra/?term=SRX016063</a></font>
	that corresponds to run id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR034528</font></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRX033397">https://www.ncbi.nlm.nih.gov/sra/?term=SRX033397</a></font>
	that corresponds to run id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR081522</font></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRX007711">https://www.ncbi.nlm.nih.gov/sra/?term=SRX007711</a></font>
	that corresponds to run id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022865</font></font></p>
</ul>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="whole-genome-assembly"></a>114 Whole
genome assembly</h1>
<h2 class="western"><a name="what-are-the-steps-of-genome-assembly"></a>
114.1 What are the steps of Genome Assembly?</h2>
<p class="first-paragraph">The typical genome assembly process is
composed of three different stages:</p>
<ol>
	<li/>
<p class="compact"><b>Assembling contigs</b>: Merging reads
	into longer contigs.</p>
	<li/>
<p class="compact"><b>Scaffolding the contigs</b>: Spatially
	orienting contigs relative to one another.</p>
	<li/>
<p class="compact"><b>Finishing the genome</b>: Bridging over
	the spaces between the spatially oriented contigs and connecting
	them.</p>
</ol>
<p class="first-paragraph">While possible it is infrequent that a
single assembly process would start with sequencing reads and would
produce a finished genome. Perhaps if your coverage is high (over
50x), the data is of very high quality and the genome does not
contain repetitive regions.</p>
<p>The steps above require increasing human intervention, especially
when finishing genomes may need another laboratory protocol where
specially targeted regions of the genome that are thought to be close
to one another are re-sequenced with a low-throughput method that
attempts to bridge the gaps.</p>
<h2 class="western"><a name="how-do-we-quantify-assembly-quality"></a>
114.2 How do we quantify assembly quality?</h2>
<p class="first-paragraph">Assembly processes are challenging to
evaluate because we lack simple terms to quantify when is one outcome
superior to another. There is a shorthand notation called the 3Cs
(CCC): contiguity, correctness, and completeness. Ideally, we expect
assemblies to be:</p>
<ol>
	<li/>
<p class="compact">Contiguity. Produce the longest possible
	contigs.</p>
	<li/>
<p class="compact">Correctness. Assemble contigs with few/no
	errors.</p>
	<li/>
<p class="compact">Completeness. Cover the entire original
	sequence and minimize missing regions.</p>
</ol>
<p class="first-paragraph">Balancing these three requires trade-offs
based on the requirement of the problem that is being solved. Expect
to have to make these tradeoffs, and you should identify your
priorities before you start the process: e.g.&nbsp;&ldquo;I would
rather be more accurate even if it means the assembly will be less
complete or contiguous.&rdquo;</p>
<p>Scientists working in the field have long been searching for
various measures that would incorporate all the three Cs above. In
practice measuring the first &lsquo;C&rsquo; &ldquo;contiguity&rdquo;
appears to be the dominant way to quantify assembly quality. And
within that realm, a measure called the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N50</font></font>
statistic is used as the primary differentiator. The N50 statistic
was adopted as one that most closely captures the quality of the
first step of assembly: extending reads to form longer contigs.</p>
<h2 class="western"><a name="what-type-of-assembly-errors-will-i-get"></a>
114.3 What type of assembly errors will I get?</h2>
<p class="first-paragraph"><b>Chaff</b> contigs that are comparable
to a read size. These are indicative of dead ends in the assembly
process.</p>
<p><b>Misjoins</b>: joining two sequences that should not be
together.</p>
<p><b>Coverage misses</b>: regions of the genome that are not
assembled at all.</p>
<p><b>Repeat compression</b>: a common error that makes consecutive
repeats appear as a single sequence. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ZZZ
ABC ABC ABC  ZZZ</font></font> may look like as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ZZZ
ABC ZZZ</font></font></p>
<h2 class="western"><a name="what-is-the-n50-statistic"></a>114.4
What is the N50 statistic?</h2>
<p class="first-paragraph">Even though the N50 statistic is the most
commonly used measure of the quality of a genome assembly, it might
come as a surprise that its meaning is not rigorously defined. As a
matter of fact, no definition for N50 that we&rsquo;ve seen appears
to be entirely correct. Take a moment to reflect on that a bit. Like
p-values, that I describe later - everyone is using p-values, but
practically nobody is using the concepts quite right.</p>
<p>The authors of <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/22147368">GAGE:
A critical evaluation of genome assemblies and assembly algorithms</a></font>
Genome Research (2012) states that:</p>
<p class="block-text-western">The N50 value is the size of the
smallest contig (or scaffold) such that 50% of the <i>genome</i> is
contained in contigs of size N50 or larger.</p>
<p class="first-paragraph">Let us note first that &ldquo;GAGE&rdquo;
stands for &ldquo;Genome Assembly Gold-Standard Evaluation.&rdquo;
Calling something &ldquo;gold standard&rdquo; from the start is a
common and annoying habit of some scientists. In our (though
admittedly limited) experience papers declaring themselves as gold
standard from the beginning are rarely that.</p>
<p>Then in the definition above you might immediately raise the
question of how would you know what the <i>genome</i> size is when
you are assembling it for the first time?. Now while there are
experimental and comparative methods to estimating genome sizes,
these are a different domain of science.</p>
<p>Many other scientists use different designations <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N50</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG50</font></font>
where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N50</font></font>
refers to the sum of all contigs lengths of the assembly whereas <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NG50</font></font>
relates to using the length of the genome (if known or estimated).
When the assembly length is reasonably close to the genome length,
the two measurements are very similar. But note that there are
legitimate reasons (not necessarily errors) that may cause an
assembly to end up much longer or much shorter than the target
genome.</p>
<p>Confusingly, several alternatives and slightly different
definitions of N50 exist. Sometimes the median of contig lengths is
used. For example in the <font color="#4f81bd"><a href="http://genome.cshlp.org/content/21/12/2224.full">Assemblathon
1: A competitive assessment of de novo short read assembly methods</a></font>
Genome Research, 2011 the authors overcomplicate the definition and
state:</p>
<p class="block-text-western">The N50 of an assembly is a weighted
median of the lengths of the sequences it contains, equal to the
length of the longest sequence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">s</font></font>,
such that the sum of the lengths of sequences greater than or equal
in length to s is greater than or equal to half the length of the
genome being assembled.</p>
<p class="first-paragraph">Frankly, I don&rsquo;t understand this
definition - especially when it comes to weighted medians. I guess
the authors agreed with this sentiment since in the <font color="#4f81bd"><a href="https://gigascience.biomedcentral.com/articles/10.1186/2047-217X-2-10">Assemblathon
2: evaluating de novo methods of genome assembly in three vertebrate
species</a></font> Genome Research, 2013 the authors write:</p>
<p class="block-text-western">N50 is calculated by summing all
sequence lengths, starting with the longest, and observing the length
that takes the sum length past 50% of the total assembly length.</p>
<p class="first-paragraph">This definition is almost right, except it
does not explicitly state that the summing should proceed by sorted
lengths. It is noteworthy to observe just how difficult &rsquo;b]r93rg2ris
to even state the definition for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N50</font></font>
perhaps foreshadowing that this domain of application is far more
complicated than bioinformaticians care to admit.</p>
<h2 class="western"><a name="so-what-is-the-n50-statistic"></a>114.5
So what is the N50 statistic?</h2>
<p class="first-paragraph">A simpler explanation of N50, (in our own
interpretation, that of course might not be quite correct), starts by
ordering contigs by length. Suppose we have 10 different contigs
(designated by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXX</font></font>
) and we ordered these by their decreasing sizes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Contig
        Length           Sum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXXXXX
      10              10</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXXXX
        9              19</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXXX
         8              27</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXX
          7              34</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXX
           6              40</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXX
            5              45</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXX
             4              49</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXX
              3              52</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XX
               2              54</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                1              55</font></font></p>
<p class="first-paragraph">The sum of these lengths starting with the
longest is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">55</font></font>.
Half of that is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">27.5</font></font>.</p>
<p>Go down on this list and add up the lengths to find the contig
where the cumulative length exceeds this half value. When we hit
contig number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>
we have <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10
+ 9 + 8 + 7 = 34</font></font>, this value is larger than <font face="Consolas, serif"><font size="2" style="font-size: 11pt">27.5</font></font>
so it is at this point at least half of the genome is stored in
contigs of size <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>
or greater.</p>
<p>Our N50 is then <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7</font></font>.
Another simple way to state this (following the wording of <font color="#4f81bd"><a href="http://jermdemo.blogspot.com/2008/11/calculating-n50-from-velvet-output.html">this
blogpost</a></font>) is that</p>
<p class="block-text-western">&ldquo;At least half of the nucleotides
in this assembly belong to contigs of size 8bp or longer.&rdquo;</p>
<p class="first-paragraph">See how we gave some many definitions for
N50. Pick the one you like :-)</p>
<h2 class="western"><a name="what-are-the-problems-with-n50"></a>114.6
What are the problems with N50?</h2>
<p class="first-paragraph">One of the biggest problems using the N50
metric as the primary means of evaluating assembly quality is that it
rewards &ldquo;misjoins.&rdquo; A &ldquo;misjoin&rdquo; is the error
of concatenating separate contigs into a single segment. Such
mistakes can happen when the evidence is insufficient, but the
algorithm is tuned to be overly generous in accepting these pieces of
evidence.</p>
<p>Also applying cutoffs can lead to odd situations when selecting
contigs would lead us to be either well under or well over the 50%
see the section below?</p>
<h2 class="western"><a name="so-what-is-your-n50"></a>114.7 So what
is your N50?</h2>
<p class="first-paragraph">While building artificial corner cases to
demonstrate a weakness of a method or measure is akin to building
straw men we could not resist it this time around. We invite you to
do the math below and with that demonstrate to yourself that you do
indeed understand how the N50 computation works.</p>
<p>Imagine that you have a genome of a size 1 million bp with and you
ran an assembler that created one large contig of the length of
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">500,000bp</font></font>
plus another 500,000 contigs of sizes of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1bp</font></font>
each.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Contig
                                     
Length</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXXXXXXXXXXXXXXX....XXXXXXXXXXXX
       500,000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...
500,000 1bp long sequences ....</font></font></p>
<p class="first-paragraph">What is your N50?</p>
<p>Now take the same example and make your largest contig just a
single basepair shorter <font face="Consolas, serif"><font size="2" style="font-size: 11pt">499,999bp</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Contig
                                     
Length</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">XXXXXXXXXXXXXXXXXXXX....XXXXXXXXXXXX
       499,999</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X
                                                1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...
500,000 1bp long sequences ....</font></font></p>
<p class="first-paragraph">So what is your N50 now?</p>
<p>Well, what do you think?</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-improve-the-assembly-process"></a>
115 How to improve the assembly process</h1>
<p class="first-paragraph">When you start out in the field of
assembly you will note with surprise how much of the advice can be
summarized as:</p>
<p class="block-text-western">Keep running your assembler with many
different parameters until you get a proper assembly.</p>
<p class="first-paragraph">There is a &ldquo;handwaving,&rdquo; and
almost &ldquo;unscientific&rdquo; feel to the recommended processes.
In general, should we ever tweak and run a tool until it gives you
what you want? Besides, you will get somewhat ad-hoc guidance as to
what a &ldquo;good&rdquo; assembly even means.</p>
<h2 class="western"><a name="why-are-assemblers-so-sensitive-to-parameters"></a>
115.1 Why are assemblers so sensitive to parameters?</h2>
<p class="first-paragraph">We believe that assemblers are sensitive
to parameters because we as scientists are not quite sure when they
work correctly - hence the rationale is to allow for extensive
customization. Then, the information content and structure of
sequences will vary to a great extent, both within a genome and more
across species. This variation poses severe challenges to any
algorithmic approach - as principles and optimizations that work well
in one region will not work for others.</p>
<p>Also, it is hard to identify with clarity how the assembler works
internally and what decision it makes - hence it hard to foresee the
effect of each parameter value on the result.Thus we are left with
trial and error.</p>
<h2 class="western"><a name="should-i-error-correct-my-data"></a>115.2
Should I error correct my data?</h2>
<p class="first-paragraph">As you recall error correction means
correcting for sequencing errors using the data itself. There are
many logical explanations of why this should work. For example, error
corrections will significantly reduce the number of k-mers in the
data - and as such will make k-mer based algorithms job easier.</p>
<p>At the same time error correction is a new application domain that
we have not had sufficient experience with. Later examples in this
book may explore the effect of error correction and it, in general,
it should be a choice that you evaluate yourself. Many authors, with
experience, believe that error correction is a requirement.</p>
<h2 class="western"><a name="what-are-the-k-mer-sizes"></a>115.3 What
are the k-mer sizes?</h2>
<p class="first-paragraph">Refer to the chapter title <font color="#4f81bd"><a href="#ch073.xhtml#sequence-kmers">Sequence
K-Mers</a></font> for a definition of k-mers. In nutshell, a k-mer
are all possible subsequences of size <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>.
The rationale for breaking our reads into even smaller pieces, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mers</font></font>
is that we want to identify &ldquo;correct&rdquo; k-mers, those that
originate from the real data. The assumption is that whatever errors
the reads may have these are distributed randomly - hence will
produce different erroneous k-mers. The correct k-mers will always be
the most abundant.</p>
<p>In general, it is true that the longer a k-mer is, the fewer
identical k-mers to it exist. At the same time it is also true the
longer a k-mer is, the more likely is that it will contain an error.
When it comes to assembly processes the rule of reasoning is that:</p>
<ul>
	<li/>
<p class="compact">A larger <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>
	value allows resolving more repetitions.</p>
	<li/>
<p class="compact">A smaller <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k</font></font>
	increases the chances of seeing a given k-mer.</p>
</ul>
<p class="first-paragraph">Hence the selection of a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">k-mer</font></font>
is a tradeoff between longer repeats and more reliable measures. In
general, you should assemble sequences using the largest k-mer size
possible, such that the k-mer coverage is sufficient.</p>
<p>You may estimate these via trial and error (as stated above) but
there are also tools like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">kmergenie</font></font>
that will assist you in determining the most likely k-mer sizes and
coverage cutoffs. These tools sometimes work well, other times no so
much.</p>
<h2 class="western"><a name="how-to-tune-the-expected-coverages"></a>115.4
How to tune the expected coverages?</h2>
<p class="first-paragraph">From experience, we found that parameters
indicating the &ldquo;expected coverage&rdquo; have a pronounce and
substantial effect on the resulting assembly. This parameter may take
different names for different tools:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-abundance-min</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-cov_cutoff
</font></font>
</p>
<p class="first-paragraph">each indicates the minimal number of
observations necessary to &ldquo;trust&rdquo; that a k-mer sequence
is correct. Again it is never clear what the value should be from the
start. Some tools allow for an automatic inference.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="the-lucky-bioinformaticians-guide-to-genome-assembly"></a>
116 The lucky bioinformatician&rsquo;s guide to genome assembly</h1>
<p class="first-paragraph">Assembling a genome is a challenging
process. Except when you get lucky. Then &ldquo;it just works&rdquo;.
But even then you need a little more than luck. You have to make your
own luck.</p>
<p>In this section, we will be using the <font color="#4f81bd"><a href="http://minia.genouest.org/">Minia
assembler</a></font> to demonstrate the typical usage patterns during
an assembly process. We like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia</font></font>
because it is easy to run, is super fast and has other tools that
integrate with it.</p>
<h2 class="western"><a name="how-do-i-run-an-assembly"></a>116.1 How
do I run an assembly?</h2>
<p class="first-paragraph">Genome assembly is an iterative process
where you will:</p>
<ol>
	<li/>
<p class="compact">Apply quality control measures</p>
	<li/>
<p class="compact">Estimate initial parameters (k-mers-sizes,
	coverages)</p>
	<li/>
<p class="compact">Run the assembly</p>
	<li/>
<p class="compact">Evaluate assembly (then go to 2 if
	necessary)</p>
</ol>
<h2 class="western"><a name="is-quality-control-necessary"></a>116.2
Is quality control necessary?</h2>
<p class="first-paragraph">Yes. Absolutely. Every single guide will
tell you so. In fact, quality control for assembly is a more
complicated process. You will need to correct not just &ldquo;visible&rdquo;
errors that the sequencing instrument marks with its quality values
but even the invisible ones that need to be fixed with the most
sophisticated methods the human mind has ever invented.</p>
<p>Unless you get lucky. In which case you don&rsquo;t need to
correct data at all. You &ldquo;just carry on&rdquo; and it will
&ldquo;just work&rdquo;.</p>
<h2 class="western"><a name="do-i-need-a-lot-of-data"></a>116.3 Do I
need a lot of data?</h2>
<p class="first-paragraph">Yes. Absolutely. Not only that, the
consensus is that you will need the longest reads the instrument can
generate, preferably 250bp or longer. Plus you&rsquo;ll need
paired-end reads that help scaffold the resulting contigs. Also, you
better collect more data than you think you need.</p>
<p>Unless of course, you are lucky. In which case you may be able to
get by using a small fraction of the sequenced data, say a hundred
thousand, single end, short reads.</p>
<h2 class="western"><a name="what-does-it-look-like-when-i-am-getting-lucky"></a>
116.4 What does it look like when I am getting lucky?</h2>
<p class="first-paragraph">Remember the study that we mentioned in
the <font color="#4f81bd"><a href="#ch053.xhtml#ebolaredo">Redo:
Genomic surveillance elucidates Ebola virus origin</a></font>&rdquo;</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.sciencemag.org/content/early/2014/08/27/science.1259657.full">Genomic
	surveillance elucidates Ebola virus origin and transmission during
	the 2014 outbreak</a></font></p>
</ul>
<p class="first-paragraph">published in 2014 in the Science research
journal. Let&rsquo;s obtain a sequencing run from it. It has hundreds
of sequencing runs, Hmmm, where to even start&hellip; well, I don&rsquo;t
know &hellip; well let&rsquo;s pick a run &hellip; how about run id
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/sra/?term=SRR1553425">SRR1553425</a></font>:</p>
<p>Get the data, it has 1.8 million paired-end reads (3.6 million
total reads) but for now, for the sake of this example let&rsquo;s
take a subset of 100,000 reads. In fact, we did not even expect it to
work, we only want to demonstrate the process:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
SRR1553425 -X 100000 --split-files</font></font></p>
<p class="first-paragraph">The data is paired-end, but for, and for
the sake of simplicity we&rsquo;ll focus on first pairs, ignoring
that we have extra information and even more data in the other pairs.
Notably, when we are taking 100K reads from a single file, hence we
are using just 3% (!) of the total sequencing data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">READS=SRR1553425_1.fastq</font></font></p>
<p class="first-paragraph">Off we go with the assembly. Let&rsquo;s
run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia</font></font>
with no other parameters set:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia
-in $READS -out genome</font></font></p>
<p class="first-paragraph">First of all, hats off to <font color="#4f81bd"><a href="http://cristal.univ-lille.fr/~chikhi/">Dr.&nbsp;Rayan
Chikhi</a></font>. The process took only 5 seconds on a iMac! It is
mind boggling that it only takes 5 seconds to build an internal graph
structure out of millions of k-mers, that were spread over 10 million
basepairs of 100,000 reads, then to churn through this network with a
graph algorithm to find the shortest paths in them. We can
unequivocally state that bioinformatics software developers are the
best the world has ever seen!</p>
<p>The results of the process are stored in a file called
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.contigs.fa</font></font>
that we can run a quick stats on:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat genome.contigs.fa</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
              format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.contigs.fa
 FASTA   DNA     190     16,812      63     88.5      277</font></font></p>
<p class="first-paragraph">Okay, so we obtained:</p>
<ul>
	<li/>
<p class="compact"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">190</font></font>
	contigs</p>
	<li/>
<p class="compact">the average length is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">88.5bp</font></font></p>
	<li/>
<p class="compact">the longest is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">270bp</font></font></p>
</ul>
<p class="first-paragraph">This result seems useless, all <i>chaff</i>.
Our read lengths were <font face="Consolas, serif"><font size="2" style="font-size: 11pt">101</font></font>
long. How does one even &ldquo;assemble&rdquo; <font face="Consolas, serif"><font size="2" style="font-size: 11pt">101</font></font>
long reads into an average of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">88.5</font></font>?
&hellip; Upon closer inspection, the numbers are more encouraging.
See the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sum_leng</font></font>
field? It says <font face="Consolas, serif"><font size="2" style="font-size: 11pt">16,812</font></font>.
We do know that the Ebola genomes are of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">18,000</font></font>bp
size.</p>
<p>Now consider that we started with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">100,000</font></font>
reads where each read was <font face="Consolas, serif"><font size="2" style="font-size: 11pt">101bp</font></font>
long. So the assembler was able to collapse <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10,000,000</font></font>
sequenced bases into a total length of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">16,000</font></font>!
This assembly tells us that the algorithm has a good &ldquo;sense&rdquo;
of what the individual pieces are, but it has a hard time connecting
these parts. Things may not be all that grim.</p>
<h2 class="western"><a name="how-do-i-get-luckier"></a>116.5 How do I
get &ldquo;luckier&rdquo;?</h2>
<p class="first-paragraph">You saw how the assembly above looked
&ldquo;wrong&rdquo; from contig length perspective but promising when
evaluating the total genome length. This information tells you it
might be worth poking around a little more, perhaps rerun the
assembly with other parameters, such as different k-mers. When the
assembly runs in seconds like the example above this is feasible. For
now, we continue, hoping to get even more &ldquo;lucky.&rdquo;</p>
<p>The default k-mer size is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">31</font></font>
and we mentioned before how usually we want to raise this as high as
possible, but still keeping sufficient coverages. Let&rsquo;s go
higher to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">100</font></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia
-in $READS -out genome -kmer-size 100</font></font></p>
<p class="first-paragraph">Now the assembly reports a total assembly
size of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">233bp</font></font>
- still well off the mark.</p>
<p>When you are searching for a parameter range you shouldn&rsquo;t
change it small steps. First find the limits, then perform what is
called a &ldquo;binary-search&rdquo; in between. Take the half
between the explored intervals and chose the direction towards the
better results. We saw how 31 seemed ok, but 100 was not. Half
between is about 75</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia
-in $READS -out genome -kmer-size 75</font></font></p>
<p class="first-paragraph">this improved our assembly quite a bit:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
              format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.contigs.fa
 FASTA   DNA     21      19,896     156      947.4   3,539</font></font></p>
<p class="first-paragraph">We got <font face="Consolas, serif"><font size="2" style="font-size: 11pt">19</font></font>
contigs, the longest is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">3.5Kb</font></font>.
Sounds good. Let&rsquo;s keep going higher with the k-mer sizes:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia
-in $READS -out genome -kmer-size 90</font></font></p>
<p class="first-paragraph">What? No way!</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">file
              format  type  num_seqs  sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.contigs.fa
 FASTA   DNA          1   18,820   18,820   18,820   18,820</font></font></p>
<p class="first-paragraph">Minia has now assembled our data into a
single sequence. Do you recall that for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-kmer-size
100</font></font> the assembly was practically useless? And for
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-kmer-size
90</font></font> we get complete genome in a single contig. Is this
is skill or luck?</p>
<h2 class="western"><a name="how-do-i-evaluate-the-assembly"></a>116.6
How do I evaluate the assembly?</h2>
<p class="first-paragraph">Let&rsquo;s align the assembled contig
against the Zaire Ebola build of 1976.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a folder for the references.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Accesion number for the 1976 outbreak.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=AF086833</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This is the reference name.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/${ACC}.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the reference sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id $ACC | seqret -filter -sid $ACC &gt;
$REF</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Index reference for the aligner.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font></p>
<h2 class="western"><a name="what-do-i-see-in-this-assembly-that-i-wouldnt-otherwise"></a>
116.7 What do I see in this assembly that I wouldn&rsquo;t otherwise?</h2>
<p class="first-paragraph">Then align the assembled contigs against
the 1976 strain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">CONTIGS=genome.contigs.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $CONTIGS | samtools sort &gt; genome.bam</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">samtools
index genome.bam</font></font></p>
<p class="first-paragraph">Visualizing the resulting <font face="Consolas, serif"><font size="2" style="font-size: 11pt">genome.bam</font></font>
file in IGV we would see:</p>
<p><img src="0000_html_2b7b8c063e19c709.png" name="Image188" align="bottom" hspace="12" width="560" height="169" border="0"/>
</p>
<p>Note how the alignment is blue and not red. This alignment
indicates that we have assembled the reverse strand! Since our data
comes from a single stranded RNA virus that was transcribed into DNA
for the purpose of sequencing you would need to reverse complement
your assembly to find the real viral sequence.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genome.contigs.fa | seqkit seq -p -r &gt; ebola.fa</font></font></p>
<p class="first-paragraph">Now align this reverse complemented
sequence. It now aligns in the proper direction. When zoomed in you
will see that genomic variants show up directly from the alignments
and no SNP calling is necessary.</p>
<p><img src="0000_html_aec63828307111a7.png" name="Image189" align="bottom" hspace="12" width="560" height="169" border="0"/>
</p>
<p>So there. You can produce and almost complete assembly and find
all variants in an Ebola genome in four tries, totaling about 20
seconds. And you can do that by using only 3% of the collected
measurements, using paired end reads in single end mode, without any
error correction. Well this is how good is to be lucky.</p>
<h2 class="western"><a name="how-do-i-get-lucky"></a>116.8 How do I
get lucky?</h2>
<p class="first-paragraph">The history of this chapter is rather
amusing. We did not cherry pick this example out of the many hundreds
that come with this publications. We randomly picked an SRR run and
geared up to demonstrate the challenges and frustrations that one
feels when attempting to assemble even a short and straightforward
genome.</p>
<p>Much to our surprise the assembly &ldquo;just worked.&rdquo; It
also helped crystallize in our mind a long-standing suspicion that
we&rsquo;ve always had:</p>
<p class="block-text-western">Assembly is all about the data and not
the method/protocol.</p>
<p class="block-text-western">Clean and well-defined data can be
easily assembled.</p>
<p class="block-text-western">&ldquo;Mosaic&rdquo; data will give
your troubles. What is &ldquo;mosaic&rdquo; data?</p>
<ul>
	<li/>
<p class="block-text-western">data obtained from the
	superposition of multiple genomes or</p>
	<li/>
<p class="block-text-western">data containing regions with
	high self-similarity</p>
</ul>
<p class="block-text-western">High quality, simple data does not need
much filtering or effort. Difficult data may be impossible to
assemble. It is very unfair!</p>
<p class="first-paragraph">You get lucky by making your own luck.
Choose the problem wisely, recognize problems early, then don&rsquo;t
blame yourself if the assembly does not work.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-to-perform-a-genome-assembly"></a>117
How to perform a genome assembly</h1>
<h2 class="western"><a name="how-do-i-run-an-assembly-1"></a>117.1
How do I run an assembly?</h2>
<p class="first-paragraph">Genome assembly is an iterative process
where you will:</p>
<ol>
	<li/>
<p class="compact">Apply quality control measures</p>
	<li/>
<p class="compact">Estimate initial parameters (k-mers-sizes,
	coverages)</p>
	<li/>
<p class="compact">Run the assembly</p>
	<li/>
<p class="compact">Evaluate assembly (then go to 2 if
	necessary)</p>
</ol>
<h2 class="western"><a name="how-to-evaluate-a-genome-assembly"></a>117.2
How to evaluate a genome assembly</h2>
<p class="first-paragraph">The term &ldquo;assembly evaluation&rdquo;
has two distinct and unrelated meanings:</p>
<ol>
	<li/>
<p class="compact">Characterize the assembled contigs by their
	observed properties.</p>
	<li/>
<p class="compact">Characterize the assembled contigs relative
	to the &ldquo;reality,&rdquo; typically a known, similar genome.</p>
</ol>
<p class="first-paragraph">The two definitions often get confounded
into a single term.</p>
<h2 class="western"><a name="what-are-the-steps-of-genome-assembly-1"></a>
117.3 What are the steps of genome assembly?</h2>
<p class="first-paragraph">This section will work through genome
assembly in the context of the assembly presented in the paper: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/pubmed/22147368">GAGE:
A critical evaluation of genome assemblies and assembly algorithms</a></font>
Genome Research (2012). We start with the sequence data deposited for
the bacteria Staphylococcus aureus known for its potential to cause
various skin and soft tissue infections.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a directory for the reference.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p refs</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Accession number for the 1976 outbreak.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ACC=NC_010079</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This is the reference name.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">REF=refs/${ACC}.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the reference sequence.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -format fasta -id $ACC | seqret -filter -sid $ACC &gt;
$REF</font></font></p>
<p class="first-paragraph">The data for a sequencing run for this
bacteria is deposited in the <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA40073">Bioproject:
PRJNA40073</a></font> The project describes 56 sequencing runs, out
of which we start with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022868</font></font>
as the GAGE paper does. This sequencing run produces reads of length
101.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA=SRR022868</font></font></p>
<p class="first-paragraph">The genome of this bacteria is known and
deposited in GenBank under the accession number <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_010079</font></font>.
Its size is reported to be 2.8 million. Now as you recall the
coverage <font face="Consolas, serif"><font size="2" style="font-size: 11pt">C</font></font>
of a genome is:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C
= N * L / G </font></font>
</p>
<p class="first-paragraph">Where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
are the number of reads, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">L</font></font>
is the length of a read, and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">G</font></font>
is the size of the genome. As we are interested in the number of
reads of length 101 that</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">N
= C * G / L </font></font>
</p>
<p class="first-paragraph">If we were aiming for say a 45x coverage,
then for an L-101 we would need around 1.2 million reads. Since the
data is paired-end, we need to select half as many reads. Let&rsquo;s
obtain those data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make a directory to hold the original data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p data</font></font></p>
<p class="first-paragraph">We could extract the reads.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA=SRR022868</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
$SRA -X 600000 --split-files -O data    </font></font>
</p>
<p class="first-paragraph">When a sequencing run is affected by many
errors (as this one is, as we shall see) it is inadvisable to build a
smaller dataset by taking the first <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
reads of the file, as we did in other examples. Errors may manifest
themselves more systematically with respect of the physical layout of
the flowcell. The order of the reads matches the flowcell; hence the
first <font face="Consolas, serif"><font size="2" style="font-size: 11pt">N</font></font>
reads are not a proper statistical sample. The proper way to randomly
sample the data would be to get all the data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRA=SRR022868</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
$SRA --split-files -O data</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font>
</p>
<p class="first-paragraph">Then randomly extract reads from each
dataset.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sample -2 -n 600000 data/SRR022868_1.fastq  &gt; data/F1.fq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
sample -2 -n 600000 data/SRR022868_2.fastq  &gt; data/F2.fq</font></font></p>
<p class="first-paragraph">Set up variables to make the commands
easier to rerun on other input:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">READ1=data/F1.fq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">READ2=data/F2.fq</font></font></p>
<p class="first-paragraph">The paper states that the dataset has an
insert size of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">180</font></font>
with a standard deviation of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">20</font></font>.
If we did not know this, we could find out this information by
aligning the sequences then computing the average template length.
Below we filter (S=see the <font color="#4f81bd"><a href="https://read.biostarhandbook.com/sam/sam-filtering.html">SAM
filtering chapter</a></font>)for paired, correctly mapped reads on
the forward strand.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
index $REF</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bwa
mem $REF $READ1 $READ2 | samtools view -F 4 -f 3 -F 16 | datamash
mean 9</font></font></p>
<p class="first-paragraph">This command gives us <font face="Consolas, serif"><font size="2" style="font-size: 11pt">193</font></font>
as the fragment (insert size).</p>
<p>The FastQC report shows data of surprisingly deficient quality
(Left is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022868</font></font>;
the right plot is for run <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR022865</font></font>).</p>
<p><img src="0000_html_c2ef5d73962ceee4.png" name="Image190" align="bottom" hspace="12" width="560" height="204" border="0"/>
</p>
<p>Looking at this data our first instinct is that the data is
useless in reality that is not the case. As scientists, we are
trained to trust measurements and instruments, and we give
credibility to quality values. The more I learn about the sequencing
instruments, the more I come to believe that most FASTQ sequencing
qualities are wildly inaccurate. Moreover, the data above cannot be
corrected in the &ldquo;traditional&rdquo; way. Try it, and you will
see why. Filtering for say the average quality of 30 would remove
just about all the data. It looks like a total loss. In fact, it is
not.</p>
<p>First, let&rsquo;s see what can we do if we took the data as is
with if we were to put it through an easy to run assembler like
Minia.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Concatenate both paired file into one.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
data/F1.fq data/F2.fq &gt; data/minia.fq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Assembly with Minia.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">minia
-in data/minia.fq -out minia</font></font></p>
<p class="first-paragraph">Minia completes in about 27 seconds but it
is not a &ldquo;pair-end aware&rdquo; assembler, and with that, we
are losing information. It does finish though and then again, just as
in the previous example to total assembled genome size is
surprisingly accurate.</p>
<h2 class="western"><a name="how-do-i-visualize-a-resulting-assembly"></a>
117.4 How do I visualize a resulting assembly?</h2>
<p class="first-paragraph">Thanks to the leadership of <font color="#4f81bd"><a href="http://cseweb.ucsd.edu/~ppevzner/">Pavel
Pevzner</a></font> <font color="#4f81bd"><a href="http://cab.spbu.ru/">The
Center for Algorithmic Biotechnology</a></font> St.&nbsp;Petersburg
State University in Russia has become a powerhouse of assembly
software. Among the many tools they have created, QUAST, Quality
Assessment Tool for Genome Assemblies stands out as one that produces
one of the most straightforward reports that characterize an
assembly.</p>
<p>Using <font face="Consolas, serif"><font size="2" style="font-size: 11pt">quast</font></font>
is quite straightforward; it requires a reference and a file with
contings:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">quast
-R $REF minia.contigs.fa </font></font>
</p>
<p class="first-paragraph">The result is an interactive web page that
visualizes the assembly that you have created.</p>
<p><img src="0000_html_2eb3505e59551aad.png" name="Image191" align="bottom" hspace="12" width="560" height="341" border="0"/>
</p>
<h2 class="western"><a name="can-i-improve-the-assembly"></a>117.5
Can I improve the assembly?</h2>
<p class="first-paragraph">Let&rsquo;s give a different assembler a
chance. A perennial favorite, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">velvet</font></font>
is run in two steps with two subcommands. First, you will need to
compute subsequences of a given k-mer (31) size and as shown below
store them in a folder, called here <font face="Consolas, serif"><font size="2" style="font-size: 11pt">velvet31</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">velveth
velvet31 31 -fastq -separate -shortPaired $READ1 $READ2</font></font></p>
<p class="first-paragraph">You can use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">velvetg</font></font>
subcommand to assemble the genome from these precomputed data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">velvetg
velvet31 -exp_cov auto -ins_length 190</font></font></p>
<p class="first-paragraph">Again the process is very speedy, taking
no more than about 2 minutes. The resulting assembly will be located
in the file out31 called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">contigs.fa</font></font>.
We can now plot both assemblies on the same QUAST report.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">quast
-R $REF minia.contigs.fa velvet31/contigs.fa</font></font></p>
<p class="first-paragraph"><img src="0000_html_c44253a552926c64.png" name="Image192" align="bottom" hspace="12" width="560" height="330" border="0"/>
</p>
<p>The resulting image shows the new assembly is shown as a blue
line. Note how it rises much quicker to 100%; hence it contains
longer segments.</p>
<h2 class="western"><a name="what-assembler-should-i-use"></a>117.6
What assembler should I use?</h2>
<p class="first-paragraph">There are many choices you can choose
from. See the GAGE paper for more options and inspiration if you need
to. We like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">velvet</font></font>
because it is easy to run and allows you to set a baseline for your
assembly. One of our other favorite assemblers is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">spades</font></font>,
another tool made at <font color="#4f81bd"><a href="http://cab.spbu.ru/">The
Center for Algorithmic Biotechnology</a></font>.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">spades.py
-1 $READ1 -2 $READ2 -o spades</font></font></p>
<p class="first-paragraph">Over the years <font face="Consolas, serif"><font size="2" style="font-size: 11pt">spades</font></font>
hs 
<img src="0000_html_f41d152362f8e97a.png" name="Image193" align="bottom" hspace="12" width="560" height="297" border="0"/>
</p>
<p>The assembly with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">spades</font></font>
appears the best by large, yet that information alone may not tell
you the whole story. QUAST comes with the so-called Icarus contig
browser, another visual aid that helps assess how well the assembly
performs:</p>
<p><img src="0000_html_3dca4dfd3d28793.png" name="Image194" align="bottom" hspace="12" width="560" height="172" border="0"/>
</p>
<p>Better than any other explanation this browser shows you the
tradeoffs. The Minia assembly is built out of many more pieces, and
none are incorrect or in the wrong orientation, and it does cover the
genome correctly. On the other end, the spades assembly contains the
largest of the pieces - but it also includes misjoins, large
misassembled regions. The velvet assembler produces output that is in
between them both.</p>
<h2 class="western"><a name="what-do-i-do-next"></a>117.7 What do I
do next?</h2>
<p class="first-paragraph">At this point is where the &ldquo;white
whale&rdquo; chase starts in earnest. It is easy to get excited and
start tuning and running different methods. Progress is initially
quick and fun, but then it can drag on. You can error correct, filter
the data, make use of extra information, continuously evaluate the
effects of each.</p>
<p>You may be able to use the reference genome to provide additional
information to the assembler. Other data, preferably from a different
sequencing platform helps even more.</p>
<p>Slowly you can build out better and better assemblies, stumble
upon parameter setting, read up on other experiences, find and tune
your results. It won&rsquo;t be pretty, and it won&rsquo;t be quick.
There won&rsquo;t be a recipe that you could make use of right away
the next time around - but you will do the second assembly much
better as you have learned what works and what does not for your
data.</p>
<p>The GAGE paper lists many other choices and strategies.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-metagenomics"></a>118
Introduction to metagenomics</h1>
<p class="first-paragraph">Microbial communities cover the entire
surface of Earth, including the inside of our bodies. Life of all
forms coexists with and rely on a vast variety of microorganisms.
Understanding the role, function and purpose of microbial species are
essential to understanding higher order organisms.</p>
<p>Through environmental sequencing, scientists have come to realize
that the level of functional and taxonomic diversity in the earth&rsquo;s
microbial communities is much greater than previously anticipated.</p>
<p>Metagenomics analyses are more challenging than other sequencing
data analysis because the skills required to succeed will swing
wildly across the different domains of science even more abruptly
than you might expect. Do you need to be a programmer, a biologist, a
statistician, an ecologist, a microbiologist to succeed? Perhaps all
of them at the same time. And while wearing many hats is not an
unexpected requirement in bioinformatics where metagenomics will
suprise you is that decision making is exceedingly subjective and
choices made along the way have a pronounced impact on the results.</p>
<h2 class="western"><a name="what-is-metagenomics"></a>118.1 What is
metagenomics?</h2>
<p class="first-paragraph">Taking the definition from Wikipedia:</p>
<p class="block-text-western">Metagenomics is the study of genetic
material recovered directly from environmental samples. The broad
field may also be referred to as environmental genomics, ecogenomics
or community genomic</p>
<p class="first-paragraph">From a bioinformatician&rsquo;s point of
view, metagenomics is the application of high-throughput genomic
techniques to the study of communities of microbial organisms.</p>
<h2 class="western"><a name="what-are-the-challenges-of-metagenomics"></a>
118.2 What are the challenges of metagenomics?</h2>
<p class="first-paragraph">Traditionally genetics has focused on
<i>vertical transfer</i> of genetic material: from parent to
offspring. There is another mechanism of evolution, the so called
<i>horizontal gene transfer</i> that allows some types of organisms
to share genetic material <i>across</i> different species. Among
single-celled organisms, it is perhaps the dominant form of genetic
transfer.</p>
<p>Consequently, the variety and diversity of the genomic content of
microorganisms is incomparably larger than that of higher order
organisms, and cannot easily be modeled in the traditional top-down,
across generation models. Also whereas in higher order organisms the
phenotypes are well defined within a taxonomic level, bacteria from
the same genera can manifest widely different functions.</p>
<h2 class="western"><a name="do-humans-host-ten-times-as-many-bacteria-as-their-cells"></a>
118.3 Do humans host ten times as many bacteria as their cells?</h2>
<p class="first-paragraph">Crack open any book on metagenomics there
is a good chance that a statement of the following nature will greet
you:</p>
<p class="block-text-western">&hellip; humans have more bacterial
cells inhabiting our body than our own cells &hellip;</p>
<p class="first-paragraph">For example as seen in <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000667">A
Primer on Metagenomics, PLoS Comp Biology, 2010</a></font>.</p>
<p><img src="0000_html_872fbe502df164da.png" name="Image195" align="bottom" hspace="12" width="560" height="87" border="0"/>
</p>
<p>This claim often surfaces as a favorite introductory quote to
presentations on metagenomics.</p>
<p>Amusingly, tracking down the origin of the numbers in this
statement (10^13 = trillion, 10^4 = ten trillion) is a mixture of
amusement and incredulity. It appears that most scientists keep
borrowing this statement from one another, repeating it without
making any effort to ensure that it is true. Each paper cites another
paper, stretching back to seemingly the beginning of modern biology.
For example, the <font color="#4f81bd"><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000667">A
Primer on Metagenomics, PLoS Comp Biology, 2010</a></font> cites a
paper from 1996 in &ldquo;Trends Microbiol&rdquo; titled &ldquo;The
indigenous gastrointestinal microflora&rdquo; as the origin of the
statement above. But this cited article is not the source of this
number. Instead, it repeats the claim and cites another source for
from 1977 titled &ldquo;Microbial ecology of the gastrointestinal
tract.&rdquo; But that paper is not the actual source either; it
refers the reader to the origin of the number to another publication
from 1972 in &ldquo;American Journal of Clinical Nutrition&rdquo;
titled &ldquo;Introduction to intestinal microecology.&rdquo;
Unsurprisingly this paper, in turn, cites a book from 1971 titled
&ldquo;Genetics of the evolutionary process&rdquo; by Theodosius
Grigorievich Dobzhansky. It is in the introduction of that book, in
the very first sentence actually where Dobzhansky states that the
human body contains ten trillion cells. There is no indication as to
where he got that number from, this estimate (an opinion really) has
been cited as a fact ever since.</p>
<p><img src="0000_html_6b58103d2ee07118.png" name="Image196" align="bottom" hspace="12" width="560" height="165" border="0"/>
</p>
<p>Jordan Anaya has series of blog posts that demonstrate the
absurdity of this scientific &ldquo;habit&rdquo; that is
unfortunately not as rare as one might hope. He calls the practice
<font color="#4f81bd"><a href="https://medium.com/@OmnesRes/rickrolls-and-russian-dolls-the-world-of-mindless-citing-fda5d0b0a51e">russian-doll
citations</a></font> that often, like above culminate in a so called
<font color="#4f81bd"><a href="https://medium.com/@OmnesRes/rickrolls-and-russian-dolls-the-world-of-mindless-citing-fda5d0b0a51e">&ldquo;rick-roll&rdquo;
citation</a></font>. The origin for the estimate of the bacterial
numbers is similarly lengthy and equally unreliable. A more recent
estimate states that <font color="#4f81bd"><a href="https://www.sciencenews.org/article/body%E2%80%99s-bacteria-don%E2%80%99t-outnumber-human-cells-so-much-after-all">Body&rsquo;s
bacteria don&rsquo;t outnumber human cells so much after all</a></font></p>
<p>Why even mention this here? It is an example and a reminder that
you will need to approach every statement and discovery that has to
do with metagenomics with a very healthy dose of skepticism. Yes, the
field is fascinating, yes the field has a lot of potential - but at
the same time it is also a domain that is rife with incorrect
assumptions, flawed analyses, hand-waving arguments, grandiose
claims, intriguingly sounding yet meaningless &ldquo;discoveries.&rdquo;</p>
<h2 class="western"><a name="what-type-of-answers-does-metagenomics-produce"></a>
118.4 What type of answers does metagenomics produce?</h2>
<p class="first-paragraph">Metagenomics is the science of answering
questions that fall into the following categories:</p>
<ol>
	<li/>
<p class="compact">Which microorganisms are present in a
	sample?</p>
	<li/>
<p class="compact">What abundance (quantity) is each
	microorganism present in?</p>
	<li/>
<p class="compact">How do the observed abundances change
	across conditions?</p>
	<li/>
<p class="compact">What are the functions of the
	microorganisms that are observed?</p>
</ol>
<p class="first-paragraph">Important: It is essential to recognize
from the very beginning that most microorganisms cannot exist
independently and their survival requires the presence of diverse
communities of other microbes. Thus we need to keep in mind that
organisms need to be treated in the context of their communities
rather than individual species</p>
<h2 class="western"><a name="what-are-the-metagenomics-approaches"></a>
118.5 What are the metagenomics approaches?</h2>
<p class="first-paragraph">Current practices in metagenomics fall
into the categories of:</p>
<ol>
	<li/>
<p class="compact">16S rRNA sequencing</p>
	<li/>
<p class="compact">Whole-metagenome sequencing</p>
</ol>
<h2 class="western"><a name="what-is-16s-sequencing"></a>118.6 What
is 16S sequencing?</h2>
<p class="first-paragraph">From the Wikipedia page on <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/16S_ribosomal_RNA">16S
Sequencing</a></font></p>
<p class="block-text-western">16S ribosomal RNA (or 16S rRNA) is the
component of the 30S small subunit of a prokaryotic ribosome []. The
genes coding for it are referred to as 16S rRNA gene and are used in
reconstructing phylogenies, due to the slow rates of evolution of
this region of the gene.</p>
<p class="first-paragraph">In a nutshell the 16S gene forms a
peculiar part of a genome that has both very highly conserved regions
(identical across just about all bacteria) as well as highly variable
regions (hypervariable) that may be specific to a taxonomical level
such as genus or species. You can use the conserved regions to
isolate the variable regions that may be used to identify the
bacteria. Since only a tiny section of the entire bacteria is genome
will be sequenced, the 16S method is very &ldquo;economical&rdquo;
requires a fraction of the coverage and produces far lower quantities
of data. The analysis methods are simpler, more efficient and more
focused.</p>
<p>The 16S rRNA gene is about 1500bp long the nine hypervariable
regions labeled as <font face="Consolas, serif"><font size="2" style="font-size: 11pt">V1-V9</font></font>
range from about 30-100 base pairs long. Due to sequencing technology
read length limitations, most 16S methods operate on just a single
one of these regions at any given time. For example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">V4</font></font>.
Picking the &ldquo;best&rdquo; region is a source of lots of
(somewhat pointless) discussions.</p>
<p><img src="0000_html_a952e9a8c51e7b36.jpg" name="Image197" align="bottom" hspace="12" width="96" height="96" border="0"/>
</p>
<h2 class="western"><a name="what-is-whole-genome-metagenomics"></a>118.7
What is whole-genome metagenomics?</h2>
<p class="first-paragraph">This method is identical to conventional
whole genome sequencing. The mixture of DNA that originates from many
different organisms is randomly sheared into smaller fragments
(100-600bp) that are then sequenced. Of course unlike the situations
where the origin of the DNA is from a single organism a metagenomic
will represent a mixture of DNA. The main complication arises from
following factors:</p>
<ol>
	<li/>
<p class="compact">Each organism may be present in wildly
	different abundances.</p>
	<li/>
<p class="compact">The sequencing represents random sample.</p>
	<li/>
<p class="compact">All measures are relative to the total
	number of DNA fragments.</p>
</ol>
<h2 class="western"><a name="how-many-bacteria-have-been-fully-sequenced"></a>
118.8 How many bacteria have been fully sequenced?</h2>
<p class="first-paragraph">In a subsequent chapter, we will see how
to get this data. For now, we&rsquo;ll say that NCBI stores 7318
fully sequenced bacterial genomes. There are much more sequences of
course that are known to correspond to bacteria but have not yet been
assembled into a single genome.</p>
<p>The NCBI bacterial taxonomy currently labels 65,231 species under
the superkingdom of bacteria.</p>
<h2 class="western"><a name="how-many-bacteria-are-unknown"></a>118.9
How many bacteria are unknown?</h2>
<p class="first-paragraph">Scientists believe that the vast majority
of bacteria are yet unknown. Depending on who you ask large numbers
potential species are considered to exist, sometimes ranging into
millions. Notably when the scientist providing the estimate is
involved in so called &ldquo;big data analysis&rdquo; their
assessment tends to be very large (billions).You may even notice an
interesting pattern,</p>
<p>The publication from 2004: <font color="#4f81bd"><a href="http://mmbr.asm.org/content/68/4/686.full">Status
of the Microbial Census, Microbiolgy and Molecular Biology Reviews,
2014</a></font> states that:</p>
<p><img src="0000_html_5eb754aecad41536.png" name="Image198" align="bottom" hspace="12" width="560" height="86" border="0"/>
</p>
<p>and it cites a paper from 1998 titled &ldquo;Prokaryotes: the
unseen majority, PNAS, 1998&rdquo; that in turn states that:</p>
<p><img src="0000_html_1586fbbd9569ae61.png" name="Image199" align="bottom" hspace="12" width="560" height="93" border="0"/>
</p>
<p>and cites a paper from 1994 titled &ldquo;The deep hot biosphere,
PNAS, 2014&rdquo; and so on. The pattern might be familiar now, as
the citations go back in time each publication is a little less
rigorous and a little more speculative. The problem here is that what
was acceptable as speculation in the early 1990s has, by the virtue
of being cited over and over, become &ldquo;a fact&rdquo;.</p>
<p>Now, we do need to recognize that bacteria are subject to more
diverse mechanisms for evolution. It makes sense to assume that there
would be more variety to them. The confounding factor is that there
is no definition of when two bacteria should be labeled as different
species. Discussing this distinction is beyond the scope of this
book. Interested readers consult chapters from: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/books/NBK8406/">Chapter
3: Classification in Medical Microbiology. 4th edition.</a></font> or
the many other resources.</p>
<p>In summary, there are many inconsistencies when trying to place
labels on organisms as diverse as bacteria.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="what-is-a-metagenomics-analysis"></a>119
What is a metagenomics analysis?</h1>
<h2 class="western"><a name="what-type-of-analyses-will-i-need-to-perform"></a>
119.1 What type of analyses will I need to perform?</h2>
<p class="first-paragraph">Metagenomics analyses span the whole gamut
from well-defined tasks of the classification category i.e.&nbsp;&ldquo;what
is in my pot&rdquo; to the highly speculative tasks where you may be
called upon to evaluate the &ldquo;metabolic potential&rdquo; of
metagenomes. Whereas the former job can be objectively evaluated, the
success of the second types of analyses will strongly depend not just
on your results but perhaps more so on your persuasion skills.</p>
<h2 class="western"><a name="what-online-tools-can-i-use"></a>119.2
What online tools can I use?</h2>
<p class="first-paragraph">Metagenomics is a unique field in that
most of the time its results need to characterize hundreds if not
thousands of species. For that reason, visualization takes a central
role. Unlike other fields of bioinformatics, there are several
alternatives for online analysis and visualization can be performed.
All these choices can also be confusing. It is almost never clear
exactly what type of studies a given site offers - they all seem to
promise &ldquo;everything&rdquo; yet the implementation details vary
considerably. Only after you upload the data do you get to see the
limitations of the tools. For that reason, we firmly prefer command
line tools though these add substantial overhead and complexity to
the analysis.</p>
<p>Another factor to consider is that most online tools are developed
to operate on a single dataset. Most scientists collect dozens if not
more samples for each experiment. Processing the multitude of samples
efficiently poses radically different challenges that online services
may not be able to handle.</p>
<p>We will mention a few online tools, then in next chapters, we will
perform operations with command line tools.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://metagenomics.anl.gov/">MGRAST:
	Metagenomics Analysis Server</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://virome.dbi.udel.edu/">VIROME</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://qiita.ucsd.edu/">QIITA</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ebi.ac.uk/metagenomics/">EBI
	Metagenomics</a></font></p>
</ul>
<h2 class="western"><a name="how-can-i-quantify-and-compare-results"></a>
119.3 How can I quantify and compare results?</h2>
<p class="first-paragraph">Terms such as alpha, beta diversity were
introduced to describe biodiversity. The alpha diversity is the
species diversity at a local scale (same site). The beta diversity
represents differences in species composition among different sites.</p>
<p>Do note however that the mathematical definitions behind these
words are not uniquely defined. There are several ways to calculate
each. You should think in terms of categories such as:</p>
<ol>
	<li/>
<p class="compact">Community richness</p>
	<li/>
<p class="compact">Community evenness</p>
	<li/>
<p class="compact">Community diversity</p>
	<li/>
<p class="compact">Shared community richness</p>
	<li/>
<p class="compact">Similarity in community membership</p>
	<li/>
<p class="compact">Similarity in community structure</p>
	<li/>
<p class="compact">&hellip;</p>
</ol>
<p class="first-paragraph">Each category, in turn, can have one or
more formulas that capture one dimension of the concept. See for
example the [mothur calculators][calculator] that the above listing
was taken from. As you can see there, for example, the <b>Berker
Parker Index</b> is calculated as the ratio of the abundance of the
dominant OTU divided by the total number of individuals in the
sample. It is a fraction that represents how much of the total does
the most abundant species correspond to.</p>
<p>The BP index is a single number and you should think of it as one,
particular characteristic of the data. There are several more that
you could (and should) compute. A Simpson index, a Jaccard
similarity, etc. Picking the right choice depends a great deal on
what your data looks like and what changes take place. Therefore you
need to pick the right &ldquo;calculator&rdquo; for the job.</p>
<h2 class="western"><a name="what-command-line-tools-may-be-used-to-analyze-metagenomics-data"></a>
119.4 What command line tools may be used to analyze metagenomics
data?</h2>
<p class="first-paragraph">The metagenomics command line analysis
communities are split into two categories:</p>
<ol>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://qiime.org/index.html">QIIME</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.mothur.org/">mothur</a></font></p>
</ol>
<p class="first-paragraph">See the Biostar question of the day <font color="#4f81bd"><a href="https://www.biostars.org/p/94869/">Qiime
Vs Mothur : Why Use One Over The Other?</a></font>. Here is what I
wrote as an answer a few years ago about them, this distinction still
holds well:</p>
<p class="block-text-western">There is a fundamental almost
philosophical difference in how the tools are developed</p>
<p class="block-text-western"><b>Mothur</b> is a single program that
re-implements a large number of very useful algorithms into a single,
high performance standalone executable program for each platform:
linux, mac and windows. In my opinion, it is one of the most amazing
feats of bioformatics software engineering especially considering
that is being developed by only two people.</p>
<p class="block-text-western"><b>QIIME</b> is a python interface
(glue) that connects a very large number of disparate programs and
what QIIME really does is transforms the input/outputs of these and
allows you to feed one program into the other. You will need to
install a very large number of dependent programs to use QIIME. In
fact it is not even correct to say that you used QIIME to compute
something, you are using the programs that QIIME runs for you:
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pynast</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">uclust</font></font>
etc.</p>
<p class="block-text-western">Now because in QIIME each program is
developed independently some by entire groups some of these may offer
higher performance than what is implemented in mothur. On the other
hand the unified interfaces of mothur make it far more consistent and
better documented not to mention easier to run.</p>
<h2 class="western"><a name="what-are-typical-steps-of-an-analysis"></a>
119.5 What are typical steps of an analysis?</h2>
<p class="first-paragraph">Tools like Mothur and QIIME offer detailed
instuctions that you can follow. For example the <font color="#4f81bd"><a href="https://www.mothur.org/wiki/MiSeq_SOP">Mothur
SOP (Standard Operating Protocol)</a></font> for MiSeq data lists and
describes in great detail the following steps:</p>
<ol>
	<li/>
<p class="compact">Reducing sequencing and PCR errors</p>
	<li/>
<p class="compact">Processing improved sequences</p>
	<li/>
<p class="compact">Assessing error rates</p>
	<li/>
<p class="compact">Processing OTUs</p>
</ol>
<ul>
	<li/>
<p class="compact">Calling OTUs</p>
	<li/>
<p class="compact">Phylotypes</p>
	<li/>
<p class="compact">Phylogenetic Analysis</p>
</ul>
<ol start="5">
	<li/>
<p class="compact">Diversity</p>
</ol>
<ul>
	<li/>
<p class="compact">Alpha diversity</p>
	<li/>
<p class="compact">Beta diversity measurements</p>
	<li/>
<p class="compact">Population-level analysis</p>
</ul>
<ol start="6">
	<li/>
<p class="compact">Phylotype-based analysis</p>
	<li/>
<p class="compact">Phylogeny-based analysis</p>
</ol>
<p class="first-paragraph">We recommend following the instructions as
described there.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="introduction-to-taxonomies"></a>120
Introduction to taxonomies</h1>
<h2 class="western"><a name="what-is-a-taxonomy"></a>120.1 What is a
taxonomy?</h2>
<p class="first-paragraph">The &ldquo;taxonomy&rdquo; is the science
of defining and naming groups of biological organisms by shared
characteristics. Swedish botanist Carl Linnaeus ushered a new era of
taxonomy in the 18th century when he proposed to characterize
organisms into a hierarchical scheme:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Domain
-&gt; Kingdom -&gt; Phylum -&gt; Class -&gt; Order -&gt; Family -&gt;
Genus -&gt; Species</font></font></p>
<p class="first-paragraph">But then what was an excellent idea in the
18th century for classifying higher order organisms is less suited to
the needs of the 21st century when we need to describe
microorganisms. The Linnean scheme is not appropriate for
characterizing bacterial species where the spectra of differences go
beyond the top-down, vertical, hierarchical representation. For that
reason, the classification of a bacteria into any given taxonomical
rank is far less informative when it comes to the function and
phenotype.</p>
<p>The flaws of the taxonomical classification have led to creating
several alternatives versions to them - for an outsider this only
adds to the confusion.</p>
<h2 class="western"><a name="what-the-heck-is-an-otu"></a>120.2 What
the heck is an OTU?</h2>
<p class="first-paragraph">As soon as you get into the field, you&rsquo;ll
soon realize that the word OTU is deeply embedded into the vocabulary
of meta-genomicists - yet many of them will struggle to explain what
an OTU is or is not. Other words are built by using the word OTU:</p>
<ul>
	<li/>
<p class="compact">OTU calling</p>
	<li/>
<p class="compact">OTU clustering</p>
</ul>
<p class="first-paragraph">etc. many suffering from the same level
ambiguity. Common misunderstanding includes believing that the term
has a more substantial meaning than what it stands for, as well as
believing that finding an OTU is the result of a well determined,
unique process.</p>
<p>The word OTU stands for &ldquo;Operational Taxonomic Unit.&rdquo;
The term itself is almost meaningless: it was originally defined as
the taxonomic unit that was being studied, &ldquo;operated&rdquo;
with. Obviously, this is a very loose definition. The meaning got
narrowed when the OTU started to represent &ldquo;things&rdquo; that
were also &ldquo;similar.&rdquo; Later the concept of &ldquo;similarity&rdquo;
grew to represent similarity over a marker gene, such as the 16S
gene.</p>
<p>Today in most parlance the OTU accounts for a grouping of
sequences based on their sequence similarity. An OTU group does imply
that the sequences have identical (or even similar) taxonomical
assignments, though there may be a representative taxonomical
classification that gets assigned to represent an OTU. Further
complications arise as well when sequences could belong equally well
in more than one OTU or many very different clades.</p>
<h2 class="western"><a name="how-many-taxonomies-are-available"></a>120.3
How many taxonomies are available?</h2>
<p class="first-paragraph">Too many. Picking and sticking to a
taxonomy appears to have an almost tribal quality to it. For a more
detailed discussion see <font color="#4f81bd"><a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-3501-4">SILVA,
RDP, Greengenes, NCBI and OTT &mdash; how do these taxonomies
compare?</a></font> in BMC Genomics 2017 where the authors conclude
that:</p>
<p class="block-text-western">While we find that SILVA, RDP and
Greengenes map well into NCBI, and all four map well into the OTT,
mapping the two larger taxonomies on to the smaller ones is
problematic.</p>
<h2 class="western"><a name="which-taxonomy-should-i-use"></a>120.4
Which taxonomy should I use?</h2>
<p class="first-paragraph">We typically use the NCBI taxonomy as it
makes it a lot easier to connect to other data sources. There are
also command line tools from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">edirect</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit</font></font>
that were designed to interact with the taxonomy files developed by
NCBI.</p>
<p>Scientists (while few admit that the problem exists) can be biased
when it comes to their pet projects. In those cases, it is best to go
with the flow. Go ahead and read a few papers and see what taxonomy
is commonly used in the field that you need to publish in and choose
that.</p>
<h2 class="western"><a name="how-to-view-the-ncbi-taxonomy"></a>120.5
How to view the NCBI taxonomy?</h2>
<p class="first-paragraph">The NCBI web site at
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/taxonomy">https://www.ncbi.nlm.nih.gov/taxonomy</a></font>
provides access to a visual display of the taxonomy.</p>
<h2 class="western"><a name="what-is-the-story-with-the-ncbi-disclaimer"></a>
120.6 What is the story with the NCBI disclaimer?</h2>
<p class="first-paragraph">Most pages on the NCBI taxonomy will show
you the following &ldquo;disclaimer&rdquo;:</p>
<p class="block-text-western">Disclaimer: The NCBI taxonomy database
is not an authoritative source for nomenclature or classification -
please consult the relevant scientific literature for the most
reliable information.</p>
<p class="first-paragraph">Thanks for clearing that up NCBI (note:
sarcasm). May we humbly ask though, if you are <i>not authoritative</i>
then why are you storing and distributing this data at all? And if
you know so well that you are not the <i>authoritative source</i>
then what is the <i>authoritative</i> resource? Why are you punting
the responsibility back to us, the readers?</p>
<p>Mocking aside the way we read this is that making taxonomies
turned out to be somewhat subjective and controversial. NCBI tries to
fend off any criticism by pointing to this disclaimer.</p>
<h2 class="western"><a name="how-to-get-the-ncbi-taxonomy-file"></a>120.7
How to get the NCBI taxonomy file?</h2>
<p class="first-paragraph">NCBI states that it has phased out the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi</font></font>
numbers even though some tools rely on this information. The
&ldquo;old-style&rdquo; tools require <font face="Consolas, serif"><font size="2" style="font-size: 11pt">gi</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxid</font></font>
type of information whereas the &ldquo;new-style&rdquo; tools work on
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">accession</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxid</font></font>
mapping.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-k $URL | gunzip -c &gt; acc2taxid.txt</font></font></p>
<p class="first-paragraph">It contains a mapping of gi number to
taxonomy id.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
acc2taxid.txt | head</font></font></p>
<p class="first-paragraph">It prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">accession
accession.version taxid gi</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A00002
A00002.1 9913 2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A00003
A00003.1 9913 3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X17276
X17276.1 9646 4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">X60065
X60065.1 9913 5</font></font></p>
<p class="first-paragraph">The &ldquo;old-style&rdquo; files are
still available and approximately 1.5Gb in size and it was last
updated:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/gi_taxid_nucl.dmp.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O $URL</font></font></p>
<p class="first-paragraph">It contains a mapping of gi number to
taxonomy id.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">gzcat
gi_taxid_nucl.dmp.gz | head</font></font></p>
<p class="first-paragraph">like so:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">3
9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">4
9646</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5
9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7
9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">The old style files may cease to be
updated at some point (we are not even sure if that has already
come). Still, as of now, there may be tools that need the information
in this old format. Thankfully we can &ldquo;convert&rdquo; a new
style file to the old-style with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
acc2taxid.txt | grep -v taxid | awk ' { print $4,&quot;\t&quot;, $3 }
' | head</font></font></p>
<p class="first-paragraph">to produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
  9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">3
  9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">4
  9646</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5
  9913</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">11
  9913</font></font></p>
<h2 class="western"><a name="how-many-taxonomical-ranks-are-there"></a>
120.8 How many taxonomical ranks are there?</h2>
<p class="first-paragraph">Whereas the Linnean taxonomy prescribed
nine ranks for the taxonomy the NCBI data has a lot many more:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
$URL</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
zxvf taxdump.tar.gz</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
nodes.dmp | cut -f 3 -d '|' | uniq -c | sort -rn </font></font>
</p>
<p class="first-paragraph">that produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1281994
  species     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">217328
   no rank     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">82878
    genus     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20883
    subspecies     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">8594
     family     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">6826
     varietas     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2775
     subfamily     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1990
     tribe     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1420
     order     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1314
     subgenus     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
807      superfamily     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
480      subtribe     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
474      forma     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
438      species group     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
331      suborder     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
309      class     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
220      phylum     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
133      species subgroup     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
132      subclass     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
104      infraorder     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 49      superorder     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 28      subphylum     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 15      infraclass     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 10      parvorder     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  5      superkingdom     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  4      superclass     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  3      kingdom     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  3      cohort     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  2      superphylum     </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  1      subkingdom</font></font></p>
<p class="first-paragraph">Quite a few many more ranks than you might
expect.</p>
<h2 class="western"><a name="how-do-i-search-the-ncbi-taxonomy-from-the-command-line"></a>
120.9 How do I search the NCBI taxonomy from the command line?</h2>
<h3 class="western"><a name="access-taxonomy-with-edirect"></a>120.9.1
Access taxonomy with EDirect</h3>
<p class="first-paragraph">For some use cases, searching NCBI with
Entrez Direct may be sufficient. For other more advanced
functionality we recommend the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit</font></font>
command line program.</p>
<p>##How do I search the taxonomies?</p>
<p>What does the taxid <font face="Consolas, serif"><font size="2" style="font-size: 11pt">9913</font></font>
stand for?</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fetch
-db taxonomy -id 9913</font></font></p>
<p class="first-paragraph">it will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.
Bos taurus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   (cattle), species, even-toed ungulates</font></font></p>
<h3 class="western"><a name="access-taxonomy-with-taxonkit"></a>120.9.2
Access taxonomy with TaxonKit</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit
list --show-rank --show-name --ids 9913</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">9913
[species] Bos taurus</font></font></p>
<h2 class="western"><a name="how-do-i-find-the-taxid-of-a-taxonomy-name"></a>
120.10 How do I find the taxid of a taxonomy name?</h2>
<h3 class="western"><a name="taxonomy-name-with-edirect"></a>120.10.1
Taxonomy name with EDirect</h3>
<p class="first-paragraph">For this, to work you&rsquo;d have to
(reasonably) correctly spell the scientific name then <font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch</font></font>
will work:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-query &quot;E. coli[txn]&quot; -db taxonomy | efetch </font></font>
</p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">562</font></font></p>
<h3 class="western"><a name="taxonomy-name-with-taxonkit"></a>120.10.2
Taxonomy name with TaxonKit</h3>
<p class="first-paragraph">You have to specify the exact name with
taxonkit:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
Escherichia coli | taxonkit name2taxid </font></font>
</p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Escherichia
coli    562</font></font></p>
<p class="first-paragraph">As you can see above <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit</font></font>
is designed to work with streams. If you had a file that contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Archaea</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Eukaryota</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Viroids</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Viruses</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">other
sequences</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">unclassified
sequences</font></font></p>
<p class="first-paragraph">you could get the taxid of each term with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
names.txt | taxonkit name2taxid</font></font></p>
<p class="first-paragraph">to obtain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Archaea
                  2157</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bacteria
                    2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Eukaryota
                2759</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Viroids
                 12884</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Viruses
                 10239</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">other
sequences          28384</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">unclassified
sequences   12908</font></font></p>
<h2 class="western"><a name="how-do-i-get-the-taxid-for-an-accession-number"></a>
120.11 How do I get the taxid for an accession number?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db nuccore -query NC_009565,NC_002570,NC_008358 | elink -target
taxonomy | efetch </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
336982</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
272558</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
228405</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db taxonomy -id 336982,272558,228405</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
1. Mycobacterium tuberculosis F11</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
  high GC Gram+</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
2. Bacillus halodurans C-125</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
   firmicutes</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
3. Hyphomonas neptunium ATCC 15444</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
   a-proteobacteria</font></font></p>
<h2 class="western"><a name="how-do-i-get-the-lineage-of-a-taxid"></a>
120.12 How do I get the lineage of a taxid?</h2>
<h3 class="western"><a name="lineage-with-edirect"></a>120.12.1
Lineage with EDirect</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db taxonomy -id 562 -format xml &gt; out.xml</font></font></p>
<p class="first-paragraph">will produce quite the output. You can
process that XML output with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
or other tools:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
output.xml | xtract -pattern LineageEx/Taxon -element
TaxId,ScientificName</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">131567
cellular organisms</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2
Bacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1224
Proteobacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1236
Gammaproteobacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">91347
Enterobacterales</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">543
Enterobacteriaceae</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">561
Escherichia</font></font></p>
<p class="first-paragraph">you can also format it in a row oriented
way with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
output.xml | xtract -pattern TaxaSet/Taxon -element
TaxId,ScientificName</font></font></p>
<p class="first-paragraph">that prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">562
   131567    2    1224    1236    91347    543    561    Escherichia
coli    cellular organisms    Bacteria    Proteobacteria   
Gammaproteobacteria    Enterobacterales    Enterobacteriaceae   
Escherichia</font></font></p>
<h3 class="western"><a name="lineage-with-taxonkit"></a>120.12.2
Lineage with TaxonKit</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
9913 | taxonkit lineage </font></font>
</p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">9913
   cellular
organisms;Eukaryota;Opisthokonta;Metazoa;Eumetazoa;Bilateria;Deuterostomia;Chordata;Craniata;Vertebrata;Gnathostomata;Teleostomi;Euteleostomi;Sarcopterygii;Dipnotetrapodomorpha;Tetrapoda;Amniota;Mammalia;Theria;Eutheria;Boreoeutheria;Laurasiatheria;Cetartiodactyla;Ruminantia;Pecora;Bovidae;Bovinae;Bos;Bos
taurus</font></font></p>
<h2 class="western"><a name="how-do-i-list-all-taxids-of-a-given-rank"></a>
120.13 How do I list all taxids of a given rank?</h2>
<p class="first-paragraph">This command lists all the taxonomies that
correspond to the viruses (taxid 10239)</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit
list --ids 10239 --indent &quot;&quot;  | head</font></font></p>
<p class="first-paragraph">Will print:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">10239</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12333</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12340</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12347</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12366</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12371</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">If you were to store the above query into
a file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">virus.taxid.txt</font></font>
you would see how many taxonomical ranks do virii have:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wc
-l virus.taxid.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">5285
virus.taxid.txt</font></font></p>
<h2 class="western"><a name="how-many-species-of-viruses-are-in-the-current-taxonomy"></a>
120.14 How many species of viruses are in the current taxonomy?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit
list --ids 10239 --indent &quot;&quot; --show-rank --show-name  |
grep species | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">12340
[species] Enterobacteria phage 933J</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12347
[species] Actinophage JHJ-1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12366
[species] Streptococcus pyogenes phage H4489A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12371
[species] Phage h30</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12374
[species] Lactococcus phage</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12375
[species] Lactococcus phage (ISOLATE 7-9)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12386
[species] Lactococcus phage mi7-9</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12388
[species] Mycobacterium phage FRAT1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12392
[species] Lactobacillus phage mv4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">12403
[species] Leuconostoc phage P32</font></font></p>
<p class="first-paragraph">or if you want to count them:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit
list --ids 10239 --indent &quot;&quot; --show-rank --show-name  |
grep species | wc -l</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">4824</font></font></p>
<p class="first-paragraph">The answer is that NCBI knows of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">4824</font></font>
virus species distributed over <font face="Consolas, serif"><font size="2" style="font-size: 11pt">5285</font></font>
ranks.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-do-i-obtain-sequences-based-on-taxonomies"></a>
121 How do I obtain sequences based on taxonomies?</h1>
<h2 class="western"><a name="how-do-i-get-all-known-bacterial-genomes"></a>
121.1 How do I get all known bacterial genomes?</h2>
<p class="first-paragraph">It used to be that obtaining all known
bacterial genomes from NCBI was as easy as downloading a prebuilt
FASTA file that contained these genomes. Alas this practice has been
discontinued and replaced with a somewhat <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genome/doc/ftpfaq/#allcomplete">convoluted
series of actions</a></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the summary as a tabular text file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O
ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/assembly_summary.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Filter for complete genomes.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk
-F &quot;\t&quot; '$12==&quot;Complete Genome&quot; &amp;&amp;
$11==&quot;latest&quot;{print $20}' assembly_summary.txt &gt;
ftpdirpaths</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Identify the FASTA files (.fna.) other files may also be downloaded
here.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">awk
'BEGIN{FS=OFS=&quot;/&quot;;filesuffix=&quot;genomic.fna.gz&quot;}{ftpdir=$0;asm=$10;file=asm&quot;_&quot;filesuffix;print
ftpdir,file}' ftpdirpaths &gt; ftpfilepaths</font></font></p>
<p class="first-paragraph">Moreover at this point the official NCBI
instructions basically punt the task right back to you the reader and
tell you to &ldquo;use a script&rdquo; to download the data for the
URLs collected into this file. Here is a way that we you can do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
all</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ftpfilepaths | parallel -j 20 --verbose --progress &quot;cd all &amp;&amp;
curl -O {}&quot;</font></font></p>
<p class="first-paragraph">The above will put all the genomes into
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">all</font></font>
directory that now contains. Do note that the process may take many
hours.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000005825.2_ASM582v2_genomic.fna.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000005845.2_ASM584v2_genomic.fna.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">GCF_000006605.1_ASM660v1_genomic.fna.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">At this point the folder contains
sufficiently large number of files that shell meta-charater expansion
won&rsquo;t work anymore and you will get errors like the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-l all/*.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-bash:
/bin/ls: Argument list too long</font></font></p>
<p class="first-paragraph">This is because the shell attempts to
expand the pattern and list all the matching files <i>before</i>
executing the command. You can still list all files:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls
-1 | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7182</font></font></p>
<p class="first-paragraph">Due to the limitation mentioend before if
you wanted to sub-select files with a certain meta-character you
would need to use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">find</font></font>
command and pipe the results into the next action:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">find
all -name '*fna.gz' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">7182</font></font></p>
<h2 class="western"><a name="are-there-other-ways-to-get-all-bacterialviral-genomes"></a>
121.2 Are there other ways to get all bacterial/viral genomes?</h2>
<p class="first-paragraph">Probably, though you should not get your
hopes too high. On this topic there are many tutorials and guides yet
it appears that most advice becomes obsoleted at a rapid pace. NCBI
appears to move around the files quite a bit while also changing the
data naming and formatting guidelines. In just 4 month the advice in
the Biostar Question of the Day: <font color="#4f81bd"><a href="https://www.biostars.org/p/236690/">Extract
all bacteria sequences from the nr database</a></font> has become
invalid.</p>
<p>NCBI offers advice and support services that may be of some use -
though we can&rsquo;t help but wonder of why the processes need to be
so complicated. Below is an &ldquo;official&rdquo; answer for the
question on how to obtain all viral genomes from NCBI:</p>
<p class="block-text-western">Dear Colleague,</p>
<p class="block-text-western">NCBI processes submitted
(GenBank/INSDC) viral genomes differently than prokaryotic/eukaryotic
genomes: - In general, we do not create Assembly records for the
primary (submitted) genomes. As you have already established there is
/genomes/refseq/viral path on the Genomes FTP site:
<font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/viral/">ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/viral/</a></font>
that is populated with assemblies, while the genomes/genbank/viral
path has a negligible content (that is coming from viral metagenomes,
an oddball of environmental data that can&rsquo;t be made into RefSeq
by RefSeq&rsquo;s standards):
<font color="#4f81bd"><a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/viral/">ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/viral/</a></font>
-NCBI generates RefSeqs only for selected GenBank records, so to get
coverage for each viral species: at least one assembly, sometimes
more. The good thing is that part of the RefSeq processing is to
calculate (find) GenBank neighbors for the RefSeq entries. The total
of the neighbors approximates the number of complete GenBank/INSDC
genomes. - You can retrieve genome neighbors on the web by using one
of the two methods that are described in this viral genomes resource
FAQ:
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genome/viruses/about/HowTo/#retrieve_neighbors">https://www.ncbi.nlm.nih.gov/genome/viruses/about/HowTo/#retrieve_neighbors</a></font>
Note that the number of displayed records will be larger than the
number of the actual genomes, because the genomes for some organisms
are multi-segmented and there is an individual record for each
segment. There is another exception that I have to mention: RefSeq
does not calculate neighbors for the flu virus. You will have to
obtain these sequences separately as described below. Since you are
after all of the GenBank genomes, the download from the web may time
out for those formats that generate larger files. In such case use
Batch Entrez or eutils as described in this article:
<font color="#4f81bd"><a href="https://support.ncbi.nlm.nih.gov/link/portal/28045/28049/Article/1928">https://support.ncbi.nlm.nih.gov/link/portal/28045/28049/Article/1928</a></font>
For the influenza sequences, go to the Influenza virus resource:
<font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi?go=database">https://www.ncbi.nlm.nih.gov/genomes/FLU/Database/nph-select.cgi?go=database</a></font>
Click on the FTP link at the top of the page, that will take you to:
<font color="#4f81bd"><a href="ftp://ftp.ncbi.nih.gov/genomes/INFLUENZA/">ftp://ftp.ncbi.nih.gov/genomes/INFLUENZA/</a></font></p>
<p class="block-text-western">Refer to the README file in the
directory.</p>
<h2 class="western"><a name="how-do-i-set-up-the-blast-for-taxonomy-operations"></a>
121.3 How do I set up the BLAST for taxonomy operations?</h2>
<p class="first-paragraph">To enable blast searches that recognize
taxonomies you have to place the contents of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxdb.tar.gz</font></font>
file in the location where blast looks for databases.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz</font></font></p>
<p class="first-paragraph">Then move this file to the blast database
directory. Once this is set up the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">-outfmt</font></font>
paramter to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">blast</font></font>
may contain taxonomy related placeholders <font face="Consolas, serif"><font size="2" style="font-size: 11pt">%T</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">%t</font></font>
etc.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db 16SMicrobial -entry all -outfmt &quot;%a %T %t&quot; | head</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118889.1
36819 Amycolatopsis azurea strain NRRL 11412 16S ribosomal RNA gene,
partial sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118899.1
1658 Actinomyces bovis strain DSM 43014 16S ribosomal RNA gene,
partial sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_074334.1
224325 Archaeoglobus fulgidus strain DSM 4304 16S ribosomal RNA gene,
complete sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118873.1
224325 Archaeoglobus fulgidus strain VC-16 16S ribosomal RNA gene,
complete sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_119237.1
224325 Archaeoglobus fulgidus strain VC-16 16S ribosomal RNA gene,
complete sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118890.1
1816 Actinokineospora fastidiosa strain ATCC 31181 16S ribosomal RNA
gene, partial sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_044838.1
1381 Atopobium minutum strain NCFB 2751 16S ribosomal RNA gene,
complete sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118908.1
1068978 Amycolatopsis methanolica strain 239 16S ribosomal RNA gene,
partial sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_118900.1
1655 Actinomyces naeslundii strain DSM 43013 16S ribosomal RNA gene,
complete sequence</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NR_044822.1
2075 Pseudonocardia nitrificans strain IFAM 379 16S ribosomal RNA
gene, partial sequence</font></font></p>
<p class="first-paragraph">Note how the results include the taxid
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">36819</font></font>
and scientific name <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Amycolatopsis
azurea</font></font>.</p>
<h2 class="western"><a name="how-do-i-extract-sequences-by-taxonomy-from-a-blast-database"></a>
121.4 How do I extract sequences by taxonomy from a BLAST database?</h2>
<p class="first-paragraph">Suppose we wanted all sequences deposited
under the taxonomy id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1392</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxonkit
list --show-name --show-rank --ids 1392</font></font></p>
<p class="first-paragraph">that corresponds to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bacillus
anthracis</font></font> species:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1392
[species] Bacillus anthracis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 191218 [no rank] Bacillus anthracis str. A2012</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 198094 [no rank] Bacillus anthracis str. Ames</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 205919 [no rank] Bacillus anthracis str. Kruger B</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 212045 [no rank] Bacillus anthracis str. Western North America
USA6153</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 260799 [no rank] Bacillus anthracis str. Sterne</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 261591 [no rank] Bacillus anthracis str. Vollum</font></font></p>
<p class="first-paragraph">The following steps would achieve that</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
$URL | gunzip -c &gt; acc2taxid.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Filter accession numbers by the known taxonomy 9606.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
acc2taxid.txt | awk ' $3 == 1392 {print $1}'  &gt; acc.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Extract sequences that belong to that taxonomy.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
acc.txt | blastdbcmd -db nt -entry_batch -  &gt; sequences.fa</font></font></p>
<h2 class="western"><a name="what-does-the-env_nt-blast-database-contain"></a>
121.5 What does the env_nt blast database contain?</h2>
<p class="first-paragraph">NCBI states:</p>
<p class="block-text-western">Sequences from environmental samples,
such as uncultured bacterial samples isolated from soil or marine
samples. The largest single source is Sagarsso (Would this be
Sargasso ?) Sea project. This does NOT overlap with nucleotide nr</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/blastdb</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(cd
~/blastdb &amp;&amp; update_blastdb.pl --decompress
env_nt)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastdbcmd
-db env_nt -entry all -outfmt &quot;%a %T %S&quot; | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139340.1
256318 metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139339.1
256318 metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139338.1
256318 metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139324.1
256318 metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139323.1
256318 metagenome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">AAFX01139322.1
256318 metagenome</font></font></p>
<p class="first-paragraph">most of this data consists of short,
random, unlocalized and unidentified sequences obtained during large
scale sequencing.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="classifying-16s-sequences-16s-classification"></a>
122 Classifying 16S sequences {#16S-classification}</h1>
<h2 class="western"><a name="what-data-will-be-analyzed"></a>122.1
What data will be analyzed?</h2>
<p class="first-paragraph">For this chapter, we will use data from
the Human Microbiome Project (HMP) Demonstration Projects that
contain 1065 sequencing runs that attempt to connect the disease of
Necrotizing enterocolitis a devastating complication of prematurity
to microbial composition of the gut.</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/46337">https://www.ncbi.nlm.nih.gov/bioproject/46337</a></font></p>
</ul>
<p class="first-paragraph">It is important to note that often data
with clinical relevance has many limitations imposed due to patient
privacy practices. To fully analyze this data in its full context you
would need permission to access phenotype (disease) related
information on each patient. In the following, we limit ourselves to
publicly available information.</p>
<p>Get all the SRR runs for the project (1065 runs in total):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Search the runs for this project.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-query PRJNA46337 -db sra | efetch -format runinfo &gt; runinfo.csv</font></font></p>
<p class="first-paragraph">Investigating the file (open in Excel as
it is comma separated) we can see that the first run id is
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899</font></font>.
Get the data for it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the data for the run.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files SRR1614899</font></font></p>
<p class="first-paragraph">This produces three files, with no
explanation of what these are. A little detective work is necessary.
Depending on the instrumentation related metadata makes it into the
repository. For example the barcode by which the instrument tagged
the samples or the primer that was used to isolate the sequences.</p>
<p>In this case the first file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899_1.fastq</font></font>
contains the sequence <font face="Consolas, serif"><font size="2" style="font-size: 11pt">TCAGACGGCTC</font></font>
and second file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899_2.fastq</font></font>
contains the 16S rRNA primer that was used to select the 16S rRNA
region <font face="Consolas, serif"><font size="2" style="font-size: 11pt">CCGTCAATTCATTTGAGT</font></font>.
Checking the <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/16S_ribosomal_RNA">16S
rRNA wiki page</a></font> we can see how this sequence corresponds to
primer <font face="Consolas, serif"><font size="2" style="font-size: 11pt">907R</font></font>.</p>
<p>Finally, the file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899_3.fastq</font></font>
contains the primary dataset and will be used for data analysis.</p>
<h2 class="western"><a name="can-i-classify-16s-sequences-via-alignment"></a>
122.2 Can I classify 16S sequences via alignment?</h2>
<p class="first-paragraph">The most &ldquo;logical&rdquo; and
&ldquo;straightforward&rdquo; approach to classification - though as
we&rsquo;ll see it is neither the most accurate nor efficient is to
align the sequences to their closest match. Since we do expect that
many of our sequences will only be partially similar but not
necessarily identical to known sequences a local alignment tool, like
BLAST might be the most appropriate:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the 16S database and place it into a newly made folder called
blastdb.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(mkdir
-p ~/blastdb &amp;&amp; cd ~/blastdb &amp;&amp; update_blastdb.pl
--decompress 16SMicrobial)</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Convert FASTQ file to FASTA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqtk
seq -A SRR1614899_3.fastq &gt;  SRR1614899.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Align converted sequence with blastn</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">blastn
-db ~/blastdb/16SMicrobial -query SRR1614899.fa &gt;
SRR1614899.blastn</font></font></p>
<p class="first-paragraph">The input file has <font face="Consolas, serif"><font size="2" style="font-size: 11pt">7280</font></font>
sequences. Guess how many lines does the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899.blastn</font></font>
contain? Done? There are probably more than you think, a lot more.
Perhaps add a few zeroes to estimate. In the case above the total
number of lines in the file was <font face="Consolas, serif"><font size="2" style="font-size: 11pt">88,073,057</font></font>,
and the data was generated in about 30 minutes. Apparently, this
process would not scale well as the amount of data rises.</p>
<p>In the resulting alignment file, each sequence may contain
hundreds if not thousands of hits to all other similar 16S genes. Of
course, since all 16S genes are similar to one another the resulting
file is humongous. Obviously, the method is extremely redundant.
You&rsquo;d be surprised however just how many publications use blast
to classify sequences.</p>
<h2 class="western"><a name="how-do-i-visualize-the-16s-classification"></a>
122.3 How do I visualize the 16S classification?</h2>
<p class="first-paragraph">The <font color="#4f81bd"><a href="https://ab.inf.uni-tuebingen.de/software/megan5">Metagenome
Analyzer (MEGAN)</a></font> is a software can be used to load up the
data. There are different versions of the tool, and as it happens the
latest version, MEGAN 6, appears to have some data loading bugs. Only
certain types of data can be visualized with it. In the following
images use MEGAN version 5. Explaining what MEGAN does and how it
works is beyond the scope of this book. The tool has a detailed
manual and several publications that describe the use cases.</p>
<p>What we will do is demonstrate a few of the visualizations. For
example, MEGAN can draw a taxonomical tree with the number of reads
assigned to a rank:</p>
<p><img src="0000_html_bda3d2705b8db140.png" name="Image200" align="bottom" hspace="12" width="560" height="265" border="0"/>
</p>
<p>Our <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614899.blastn</font></font>
classification will look like the image above. Note how the
classification stops at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Family</font></font>
rank.</p>
<h2 class="western"><a name="are-there-better-ways-to-classify-16s-sequences"></a>
122.4 Are there better ways to classify 16S sequences?</h2>
<p class="first-paragraph">Oh, indeed there are. Blast alignment
&ldquo;kind-of&rdquo; work but are an incredibly redundant and
wasteful choice for this process. For the 16S gene, in particular,
far more efficient classifiers have been built that rely on the
patterns in the sequences rather than alignments.</p>
<p>We prefer the RPD classifier as a command line choice. It used to
be that it was distributed as a simple command line tool that
required only Java to be installed. Today is a more complicated as
described in <font color="#4f81bd"><a href="https://github.com/rdpstaff/RDPTools">RDPTools
GitHub page</a></font>. Once installed you can run the classifier as:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">java
-jar ~/src/RDPTools/classifier.jar classify SRR1614899_3.fastq -o
assign.txt -h ranks.txt </font></font>
</p>
<p class="first-paragraph">This process classifies the same number of
sequences as before but does so in just 50 seconds and produces much
smaller files. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">assign.txt</font></font>
file contains a taxonomical assignment for each input read and can be
loaded up into Megan to visualize the results. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ranks.txt</font></font>
file contains the number of reads assigned to a taxonomical rank and
the number of reads (count) assigned to that rank.</p>
<p><img src="0000_html_c82820bbcafccb38.png" name="Image201" align="bottom" hspace="12" width="560" height="395" border="0"/>
</p>
<p>Note how this classification has now assigned reads up to the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">genus</font></font>
level though still not at species level that many might have
expected.</p>
<h2 class="western"><a name="how-can-i-visually-compare-classifications"></a>
122.5 How can I visually compare classifications?</h2>
<p class="first-paragraph">MEGAN allows us to compare
classifications. For example, we may compare the BLAST and RDP
results to draw the image below. Note just how much more detailed the
blue assignments are. RDP (blue) can classify the sequences faster
and at a lower rank than BLAST (light green). Unfortunately, the
limitations of BLAST for classification are not well understood and
large numbers of scientific publications still make use of it.</p>
<p><img src="0000_html_47021395b0683860.png" name="Image202" align="bottom" hspace="12" width="560" height="430" border="0"/>
</p>
<h2 class="western"><a name="are-there-other-16s-classification-methods"></a>
122.6 Are there other 16S classification methods?</h2>
<p class="first-paragraph">Yes, there is a wide variety of tools and
techniques, often integrated into automated pipelines. See the
chapter on <font color="#4f81bd"><a href="#ch119.xhtml#metagenomics-analysis">Metagenomics
Analysis</a></font></p>
<h2 class="western"><a name="how-do-i-classify-multiple-samples"></a>122.7
How do I classify multiple samples?</h2>
<p class="first-paragraph">The RPD classifier can take multiple
datasets and can classify them simultaneously. Let&rsquo;s get data
for four files and classify all four:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain four runs.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.txt | cut -f 1 -d , | grep SRR | head -4 | xargs -n 1
fastq-dump --split-files</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Classify all four datasets.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">java
-jar ~/src/RDPTools/classifier.jar classify *_3.fastq -f allrank -o
assign.txt -h ranks.txt </font></font>
</p>
<p class="first-paragraph">the resulting file contains a tabular
ranking that :</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ranks.txt | cut -f 1,3-10 | head</font></font></p>
<p class="first-paragraph">of the form:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">taxid
name                rank SRR1614899_3.fastq SRR1614900_3.fastq
SRR1614901_3.fastq SRR1614903_3.fastq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0
    Root              rootrank 7280 3590 8891 11472</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1
    Bacteria            domain 7256 3572 8850 11393</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">841
  &quot;Proteobacteria&quot;    phylum 2674 3055 933 8000</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1501
 Gammaproteobacteria  class 2653 3039 933 7993</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1616
 &quot;Enterobacteriales&quot;  order 2589 3030 933 7975</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1617
 Enterobacteriaceae  family 2589 3030 933 7975</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1649
 Plesiomonas          genus    2    2   0    3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1637
 Klebsiella           genus 2100 2450 598 6413</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1631
 Enterobacter         genus 2 1 0 1</font></font></p>
<p class="first-paragraph">You can subselect the classifiction by
different taxonomical levels:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ranks.txt | cut -f 1,3-10 | awk ' $3==&quot;class&quot; { print $0}'
| sort -rn -k 4 </font></font>
</p>
<p class="first-paragraph">that produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">2722
Negativicutes       class 3497    3   41    2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1501
Gammaproteobacteria class 2653 3039  933 7993</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2260
Bacilli             class  803  428 5831 3079</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2375
Clostridia          class  220    6  492   19</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1175
Betaproteobacteria  class   14    6    0    0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">842
 Alphaproteobacteria class    5    8    0    0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">458
 Flavobacteriia      class    3   13    0    0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">3
   Actinobacteria      class    2   18  160  143</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">790
 Fusobacteriia       class    1    0    0    0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">423
 &quot;Bacteroidia&quot;       class    0    0 1135    0</font></font></p>
<p class="first-paragraph">A lot can be learned from a classification
as simple as this. While we are at it note how widely different the
counts are across samples.</p>
<h2 class="western"><a name="how-do-i-determine-statistically-relevant-differences-between-the-classification-counts"></a>
122.8 How do I determine statistically relevant differences between
the classification counts?</h2>
<p class="first-paragraph">There is a debate on whether the count
data obtained via classification is conceptually similar to the
counts obtained when assigning reads to transcripts. If the
statistical assumptions used to model data originating from RNA-Seq
experiments still apply to metagenomics data, then we could use the
wide variety of methods designed for RNA-Seq to analyze
classification counts as well.</p>
<p>One such assumption, for example, is that the total number of
non-changing elements is substantially larger than the number of
differentially expressed entries (basically that only a small subset
changes). If this requirement is not met, then methods originally
developed for RNA-Seq should not be used unless there is a reason to
believe that this requirement may be ignored.</p>
<p>Much has been written about the subject of comparison and
normalization; a recent overview can be read in the paper
<font color="#4f81bd"><a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-017-0237-y">Normalization
and microbial differential abundance strategies depend upon data
characteristics</a></font> the authors summarize their findings in
what we interpreted as: &ldquo;Well.. it depends on a lot of factors.
Good luck my friend!&rdquo;</p>
<p>Let&rsquo;s give it a shot here. Suppose we have two groups. The
first three runs belong to one group, the other to another group. Put
these runs into a file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">groups.txt.</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614918</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614919</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614920</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614917</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614922</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1614924</font></font></p>
<p class="first-paragraph">obtain the data and classify it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p sra</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
groups.txt |  xargs -n 1 fastq-dump --outdir sra --split-files
--skip-technical -F</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Ensure to list the SRR numbers by groups. First three, second
three.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">java
-jar ~/src/RDPTools/classifier.jar classify SRR1614918_3.fastq
SRR1614919_3.fastq SRR1614920_3.fastq SRR1614917_3.fastq
SRR1614922_3.fastq SRR1614924_3.fastq -f allrank -o assign.txt -h
ranks.txt </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Filter the data for only the columns of interest.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Add a header as well.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ranks.txt | head -1 | cut -f 3,5-11 &gt; counts.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
ranks.txt | awk ' $4 == &quot;genus&quot; { print $0 }' | cut -f
3,5-11 &gt;&gt; counts.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Obtain the deseq2 script from the Biostar Handbook.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-O http://data.biostarhandbook.com/rnaseq/code/deseq2.r &gt;
deseq2.r</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Run the deseg2 </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
counts.txt | Rscript deseq2.r 3x3 &gt; results_deseq2.txt </font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Draw the heatmap</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
norm-matrix-deseq2.txt | Rscript draw-heatmap.r &gt; output.pdf</font></font></p>
<p class="first-paragraph">That will, in turn, generate a heat map at
genus level for the different patients and the genus level abundance
of the various bacteria.</p>
<p><img src="0000_html_92b129d3ffdba0b5.png" name="Image203" align="bottom" hspace="12" width="560" height="561" border="0"/>
</p>
<p>One thing is clear - there is a lot of variation across the
different groups. Every individual in every group shows radical
changes in their metagenomic composition. On the other hand, we don&rsquo;t
know if the groups were built correctly - the grouping data was not
available due to privacy reasons.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="classifying-whole-genome-data"></a>123
Classifying whole genome data</h1>
<p class="first-paragraph">Shotgun metagenomic sequencing allows
scientists to study the full DNA of organisms present in a given
&ldquo;biological sample.&rdquo; Whereas 16S sequencing allows you to
group bacteria into OTUs and different taxonomical ranks, whole
shotgun sequencing allows for a more detailed study of organisms from
classifying into subspecies to strains to investigating functional
roles as well.</p>
<p>Of course, there is the added complexity of having to deal with
massively increased amounts of data that often are of many orders of
magnitude larger than a 16S type study.</p>
<h2 class="western"><a name="how-do-i-evaluate-how-well-a-given-method-works"></a>
123.1 How do I evaluate how well a given method works?</h2>
<p class="first-paragraph">Since most bacteria are currently unknown,
it is somewhat difficult to assess methods as we have to evaluate
success relative to a small known subset of reality. Also, as we
found out, there is a bit of wild west mentality in the field, as you
will see it is surprisingly difficult to evaluate methods as we
continuously struggle with data janitorial tasks: the naming of
bacteria can vary wildly, taxonomies not used consistently, etc.</p>
<p>Several publications have attempted to objectively quantify the
quality of metagenomic techniques, among them the paper titled
<font color="#4f81bd"><a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0031386">Assessment
of Metagenomic Assembly Using Simulated Next Generation Sequencing
Data, PLoS One, 2012</a></font>. We will use a simulated data set
from this paper to get a practical feel for the strengths and
weaknesses of the various methods.</p>
<p><img src="0000_html_f15e98155053ba1f.png" name="Image204" align="bottom" hspace="12" width="560" height="320" border="0"/>
</p>
<p>The data is distributed from:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.bork.embl.de/~mende/simulated_data">http://www.bork.embl.de/~mende/simulated_data</a></font></p>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="http://www.bork.embl.de/~mende/simulated_data/README.txt">http://www.bork.embl.de/~mende/simulated_data/README.txt</a></font></p>
	<li/>
<p class="compact">Supporting Table 2:
	<font color="#4f81bd"><a href="https://doi.org/10.1371/journal.pone.0031386.s002">https://doi.org/10.1371/journal.pone.0031386.s002</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-obtain-the-data"></a>123.2 How
do I obtain the data?</h2>
<p class="first-paragraph">If your computer can handle the
computational load you could download the original data from the site
above with</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://www.bork.embl.de/~mende/simulated_data/illumina_100species.1.fq.gz
| gunzip -c &gt; read1.fq</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://www.bork.embl.de/~mende/simulated_data/illumina_100species.2.fq.gz
| gunzip -c &gt; read2.fq</font></font></p>
<p class="first-paragraph">And run your analyses on <font face="Consolas, serif"><font size="2" style="font-size: 11pt">read1.fq</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">read2.fq</font></font>
respectively. That being said these datasets are a wee bit too large
for exploratory analyses. We have down-sampled the first-pair file
into 1 million random reads with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqtk
sample -2 -s 2 read1.fq 1000000 &gt; subset.fq</font></font></p>
<p class="first-paragraph">and made it available from the Biostar
Handbook data site. You can download this data with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
http://data.biostarhandbook.com/meta/data/subset.fq.gz | gunzip -c &gt;
subset.fq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
See what the file contains.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">seqkit
stat subset</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#file
      format  type   num_seqs     sum_len  min_len  avg_len 
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#subset.fq
 FASTQ   DNA   1,000,000  75,000,000       75       75       75</font></font></p>
<p class="first-paragraph">This subsampled data can help us in
evaluating classification methods but would not be sufficient to say
assemble new genomes.</p>
<h2 class="western"><a name="what-are-the-expected-abundances-of-the-data"></a>
123.3 What are the expected abundances of the data?</h2>
<p class="first-paragraph">The paper comes with some supplementary
files among them:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This will download the supplemental info in XLS format.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl
-OJL https://doi.org/10.1371/journal.pone.0031386.s002</font></font></p>
<p class="first-paragraph">In this case before we could evaluate a
classification method we ought to understand what we expect to see -
what kind of properties do the artificially generated data have? Look
at the Excel file that we downloaded above.</p>
<p><img src="0000_html_d2f7ac70b0be3cf6.png" name="Image205" align="bottom" hspace="12" width="560" height="100" border="0"/>
</p>
<p>Take for example the first entry (table sorted by coverage) states
that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Psychrobacter
cryohalolentis K5</font></font> corresponds to accession numbers
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_007969</font></font>
and will be present at an expected coverage of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">17.84</font></font>.
Let&rsquo;s see how the well the data supports this information.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This took </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
subset.fq | egrep '^@NC_007969|^@NC_007968' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
15902</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the size of NC_007969 + NC_007968</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_007969,NC_007968 -format fasta | seqkit stat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
file  format  type  num_seqs    sum_len  min_len      avg_len   
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
-     FASTA   DNA       2  3,101,097   41,221  1,550,548.5  3,059,876</font></font></p>
<p class="first-paragraph">Above we see that a total of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15902</font></font>
reads were found as being generated from <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Psychrobacter
cryohalolentis K5</font></font>. The original data contained about 53
million reads and we used a subset of 1 million of it, we can
extrapolate (estimate) the total number of reads in the original data
as being about <font face="Consolas, serif"><font size="2" style="font-size: 11pt">15902
* 53 ~ 842,000</font></font>. Since we did download the full data it
is worth checking out how well the estimated number matches reality:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
read1.fq read2.fq | egrep '^@NC_007969|^@NC_007968' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
849442</font></font></p>
<p class="first-paragraph">The result is very close indeed. The
accuracy of our estimate gives us confidence that we&rsquo;ll be able
to estimate other quantities of interest just as well even from a
sample of 1 million reads.</p>
<h2 class="western"><a name="are-there-different-kinds-of-abundances"></a>
123.4 Are there different kinds of abundances?</h2>
<p class="first-paragraph">The word abundance can carry different
meanings. Whereas during 16S sequencing each read corresponds to a
bacterial cell, in the case of whole genome sequencing the length of
the genome also matters. Longer bacteria will produce more DNA and
will be seen in more sequencing reads. Another way to say this is
that ten read counts of one bacteria vs.&nbsp;one read count of
another bacteria does not mean that there were more of the first than
the second.</p>
<h2 class="western"><a name="what-percent-of-a-measurements-originate-from-a-given-species"></a>
123.5 What percent of a measurements originate from a given species?</h2>
<p class="first-paragraph">This measure depends on the length of the
DNA of the organism multiplied by the number of copies of that
organism. It can be found by dividing the number of reads that are
classified at that rank with the total number of reads in the sample:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">15902/1000000</font></font></p>
<p class="first-paragraph">This value does not vary when we subsample
the data. In both cases <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.5%</font></font>
of the reads in our file were quantified as belonging to
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Psychrobacter
cryohalolentis K5</font></font>. This number does not appear in the
supporting materials of this research paper. Some classifiers will
only report this value.</p>
<p>Again it is important to remember that these numbers do not
indicate the counts of the bacteria in the sample. They indicate what
percent of measurements were &ldquo;spent&rdquo; identifying one
particular member of the community.</p>
<h2 class="western"><a name="what-is-the-coverage"></a>123.6 What is
the coverage?</h2>
<p class="first-paragraph">As you recall the coverage represents the
number of times each base of genome is expected to be sequenced. The
genome size for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_007969</font></font>
can be found with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">efetch
-db nuccore -id NC_007969,NC_007968 -format fasta | seqkit stat</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#file
 format  type  num_seqs    sum_len  min_len      avg_len   
max_len</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#-
    FASTA   DNA       2  3,101,097   41,221  1,550,548.5  3,059,876</font></font></p>
<p class="first-paragraph">Since each read is <font face="Consolas, serif"><font size="2" style="font-size: 11pt">75</font></font>
bp long and you have sequenced 849442 reads the coverage of of this
bacteria will be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">849442
* 75 / 3101097 = 20.5x</font></font>. Notably, this does not match
the coverage in the Excel sheet that is reported to be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">17.4x</font></font>.
The difference is not substantial, but it is a bit discomforting that
we are unable to reproduce it from the data. It is not clear why this
discrepancy exists - unfortunately it is not uncommon to run into
such issues.</p>
<h2 class="western"><a name="how-can-i-tell-how-many-bacteria-are-in-the-sample"></a>
123.7 How can I tell how many bacteria are in the sample?</h2>
<p class="first-paragraph">First, you need to recognize that
sequencing does not produce absolute numbers. From sequencing data
alone we cannot tell if a bacteria was present in one, ten, one
million or one billion copies. What we can identify are the so called
&ldquo;relative abundances&rdquo; how many more of one bacteria are
there compared to another.</p>
<p>When a coverage tells you that an organism was covered at 10x
versus another one covered at 1x it directly means that the first
must have had ten times as many more copies present as the latter. To
figure out the relative abundance percentages sum up all coverages
then express each coverage as a percentage of the total.</p>
<p>For example, if you had bacteria A, B, and C covered at 10x, 20x
and 30x coverages (10+20+30=60) then their relative abundances will
be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">10/60=17</font></font>,
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">20/60=33</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">30/60=50</font></font>.
The fractions help to express the composition as percentages. Your
sample is then composed of 17% bacteria A, 33% bacteria B and 50%
bacteria C. These numbers represent abundances based on the number of
cells.</p>
<h2 class="western"><a name="how-do-i-match-the-results-of-a-classifier"></a>
123.8 How do I match the results of a classifier?</h2>
<p class="first-paragraph">As we prepared this chapter, we realized
that it is surprisingly difficult to verify and validate the results
produces by classifiers. Several inconsistencies may exist in both
the expected data and the results generated by our software.
Sometimes the taxonomy id is missing, in other cases, the names may
be spelled slightly differently, for each organism, or perhaps the
accession numbers are not annotated with a taxonomy etc.. This makes
it surprisingly difficult to match results to expectations.</p>
<h2 class="western"><a name="how-does-kraken-classifier-work"></a>123.9
How does Kraken classifier work?</h2>
<p class="first-paragraph">The classifier has quite high performance
and can be run with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken
-db ~/refs/kraken/minikraken_20141208 -fastq-input subset.fq &gt;
results.out</font></font></p>
<p class="first-paragraph">that produces the output:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">26667004
sequences (2000.03 Mbp) processed in 401.844s (3981.7 Kseq/m, 298.63
Mbp/m).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 16202163 sequences classified (60.76%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 10464841 sequences unclassified (39.24%)</font></font></p>
<p class="first-paragraph">The above was generated with the
&ldquo;prebuilt&rdquo; mini-kraken database distributed by the
authors. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results.out</font></font>
file tracks each input read and lists the classification for each
measurement. The information typically needs to be consolidated into
a more informative report with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken-report
-db ~/refs/kraken/minikraken_20141208 results.out &gt; results.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><br/>
<br/>

</p>
<p class="first-paragraph">Where the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results.txt</font></font>
generated above is a file that contains the sample abundances at each
taxonomical rank (check the <font color="#4f81bd"><a href="http://ccb.jhu.edu/software/kraken/MANUAL.html">Kraken
Manual</a></font> for details on what each column means):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"> <font face="Consolas, serif"><font size="2" style="font-size: 11pt">39.38
 393820  393820  U       0        unclassified</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
60.62  606180  3126    -       1        root</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
60.31  603053  9       -       131567   cellular organisms</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
59.29  592937  930     D       2        Bacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
25.12  251162  491     P       1224     Proteobacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
13.69  136877  323     C       1236     Gammaproteobacteria</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 6.69  66892   0       O       91347    Enterobacteriales</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 6.69  66892   24149   F       543      Enterobacteriaceae</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 1.00  10042   426     G       561      Escherichia</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0.96  9611    8721    S       562      Escherichia coli</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0.06  602     602     -       331111   Escherichia coli E24377A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0.02  243     243     -       405955   Escherichia coli APEC O1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0.00  21      19      -       83333    Escherichia coli K-12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 0.00  2       2       -       316407   Escherichia coli str. K-12
substr. </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 ...</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 </font></font>
</p>
<p class="first-paragraph">Each row indicates the taxonomical rank
(for example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">S</font></font>
means Species). To select the rows at species rank with at least 100
reads supporting that rank</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt | awk ' $2&gt;100 &amp;&amp; $4==&quot;S&quot; { print
$0}' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
83</font></font></p>
<p class="first-paragraph">Hence the classifier predicts the presence
of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">83</font></font>
species. It remains to be seen if these are indeed matching the 100
species that we know were present. It is clear though that we have
missed out on some 20% of the species. To compare to the expected
data reformat the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">results.txt</font></font>
file to select as species level and keep the taxid and name of the
bacteria (we do also sort it so that we can later join this file to
another):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Put the results with their taxid into a file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt | awk ' $2&gt;100 &amp;&amp; $4==&quot;S&quot; {
OFS=&quot;\t&quot;; print $5 }' | sort  &gt; kraken.taxid</font></font></p>
<p class="first-paragraph">This file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken.taxid</font></font>
now contains just the taxonomic ids:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">101534</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">106590</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1076</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1100</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">We will now attempt match this to the
expected data. Because the tax ids were not specified for the data in
the paper it is surprisingly challenging to compare the expected
results to the obtained ones.</p>
<h2 class="western"><a name="how-do-i-match-back-the-bacteria"></a>123.10
How do I match back the bacteria?</h2>
<p class="first-paragraph">As you&rsquo;ll experience yourself often
the greatest of challenges occur when trying to do something really
simple, that has seemingly nothing to do with bioinformatincs all. In
this case we want to figure out how well the classification worked,
we want to match the file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken.txt</font></font>
to the contents of the XLS sheet distributed as Supplementary
information. It is a surprisingly tough job to get right.</p>
<p>The taxid is present in one file but os not there in the second,
hence we first need to generate a taxid for each entry of the XLS
sheet. Lets give it a go. First create a <font face="Consolas, serif"><font size="2" style="font-size: 11pt">tmp</font></font>
file by pasting the bacteria names from the D column of the Excel
sheet into it. Then cut the first two words and place them into the
file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">names.txt</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
tmp | cut -f 1,2 -d ' ' </font></font>
</p>
<p class="first-paragraph">So that stream will give you this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Psychrobacter
cryohalolentis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Mycobacterium
sp.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Synechococcus
elongatus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bacillus
cereus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">Now pipe this output into the taxonkit:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
tmp | cut -f 1,2 -d ' ' | taxonkit name2taxid </font></font>
</p>
<p class="first-paragraph">that will give you:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Psychrobacter
cryohalolentis  330922</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Mycobacterium
sp.             1785</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Synechococcus
elongatus       32046</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bacillus
cereus               1396</font></font></p>
<p class="first-paragraph">We notice that some bacterial names did
not produce a taxid and there are duplicates there as well. Two
species may correspond to same taxid.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
tmp | cut -f 1,2 -d ' ' | taxonkit name2taxid | cut -f 2 | sort |
uniq | egrep &quot;[0-9]+&quot; &gt; expected.taxid</font></font></p>
<p class="first-paragraph">When all said and done the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">expected.taxids</font></font>
file contains <font face="Consolas, serif"><font size="2" style="font-size: 11pt">76</font></font>
distinct taxonomic ids. That is quite the drop from the original 100.</p>
<p>The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm</font></font>
tool allows us to select lines that are common (or different in
files).</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids in both file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-1 -2 expected.taxid kraken.taxid  | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
70</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids missed in classification.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-2 -3 expected.taxid kraken.taxid  </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
6</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids present in the kraken classification only.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-1 -3 expected.taxid kraken.taxid  | wc -l# 78 </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#13</font></font></p>
<p class="first-paragraph">So what was missed</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-2 -3 expected.taxid kraken.taxid | taxonkit lineage </font></font>
</p>
<p class="first-paragraph">Expected species not found:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1765
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium;Mycobacterium
tuberculosis complex;Mycobacterium bovis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1773
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium;Mycobacterium
tuberculosis complex;Mycobacterium tuberculosis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1785
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium;Mycobacterium
sp.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1831
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Nocardiaceae;Rhodococcus;Rhodococcus
sp.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">203804
 cellular
organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;unclassified
Enterobacteriaceae;ant, tsetse, mealybug, aphid, etc.
endosymbionts;ant endosymbionts;Candidatus Blochmannia</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">50422
  cellular
organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Alteromonadales;Shewanellaceae;Shewanella;Shewanella
sp</font></font></p>
<h2 class="western"><a name="would-the-full-kraken-database-be-more-accurate"></a>
123.11 Would the full kraken database be more accurate?</h2>
<p class="first-paragraph">If we were to build the most up-to-date
full database as instructed in the Kraken manual (though the process
is a bit tedious) the rate of classification ratio rises
substantially:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken
--db /export/refs/kraken_db/kraken_std_db/ --fastq-input subset.fq &gt;
results-full.out</font></font></p>
<p class="first-paragraph">that in turn produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1000000
sequences (75.00 Mbp) processed in 25.376s (2364.5 Kseq/m, 177.33
Mbp/m).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 826977 sequences classified (82.70%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 173023 sequences unclassified (17.30%) </font></font>
</p>
<p class="first-paragraph">So even in the &ldquo;best-case scenario&rdquo;
of a controlled experiment with 100 known species 17% of our
sequences will be unclassified even though they match known genomes.</p>
<h2 class="western"><a name="how-to-use-sourmash"></a>123.12 How to
use sourmash?</h2>
<p class="first-paragraph">According to their documentation:</p>
<p class="block-text-western">sourmash is a command-line tool and
Python library for computing MinHash sketches from DNA sequences,
comparing them to each other, and plotting the results. This allows
you to estimate sequence similarity between even very large data sets
quickly and accurately.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install
https://github.com/dib-lab/sourmash/archive/2017-ucsc-metagenome.zip</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install khmer</font></font></p>
<p class="first-paragraph">Download and install the sourmash
signature database:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p
~/refs/sourmash</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">URL=https://s3-us-west-1.amazonaws.com/spacegraphcats.ucdavis.edu/microbe-refseq-sbt-k21-2017.05.09.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">(cd
~/refs/sourmash &amp;&amp; curl -k $URL | tar xzv)</font></font></p>
<p class="first-paragraph">Compute a signature on your subset of
data:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash
compute --scaled 10000 -k 21 subset.fq </font></font>
</p>
<p class="first-paragraph">Characterize the signature:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash
sbt_gather -k 21 ~/refs/sourmash/refseq-k21.sbt.json subset.fq.sig -o
sourmash.csv</font></font></p>
<p class="first-paragraph">The tool will produce somewhat differnt
information on the screen versus the results file <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash.csv</font></font></p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">overlap
   p_query p_genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-------
   ------- --------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.4
Mbp    2.7%     24.5%      NZ_CP009464.1 Bacillus anthracis strain
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.1
Mbp    2.2%     40.1%      NZ_JOPX01000001.1 Staphylococcus aureus
</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.1
Mbp    2.1%     26.0%      NZ_CP009685.1 Escherichia coli str.
K-12</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.0
Mbp    1.9%     15.9%      NC_009077.1 Mycobacterium sp. JLS,
compl</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.8
Mbp    1.6%      9.0%      NC_010628.1 Nostoc punctiforme PCC
73102</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.8
Mbp    1.5%      8.0%      NC_008268.1 Rhodococcus jostii RHA1,
com</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">whereas the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash.csv</font></font>
file contains:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">intersect_bp,f_orig_query,f_found_genome,name</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1410000.0,0.24479166666666666,0.02744258466329311,&quot;NZ_CP009464.1
Bacillus anthracis strain SK-102, complete
genome&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1140000.0,0.40070921985815605,0.022187621642662515,&quot;NZ_JOPX01000001.1
Staphylococcus aureus LysK 2 2011 psaLysK211.contig.0, whole genome
shotgun
sequence&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1100000.0,0.26004728132387706,0.02140910860256909,&quot;NZ_CP009685.1
Escherichia coli str. K-12 substr. MG1655, complete
genome&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">970000.0,0.15901639344262294,0.018878941222265473,&quot;NC_009077.1
Mycobacterium sp. JLS, complete
genome&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">800000.0,0.08958566629339305,0.015570260801868432,&quot;NC_010628.1
Nostoc punctiforme PCC 73102, complete genome&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<p class="first-paragraph">How well did the method work? Firs
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash</font></font>
runs substantially slower than kraken, at least 10x slower. Then the
accession numbers are not useful for matching purposes since the same
sequence may be present under different accession numbers. Let us
test how well the results match. Note the command line contraptions
that we had to come up with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
sourmash.csv | cut -d ','  -f 4 | cut -d ' ' -f 2,3 | taxonkit
name2taxid | cut -f 2 | sort | uniq | egrep &quot;[0-9]+&quot; &gt;
sourmash.taxid</font></font></p>
<p class="first-paragraph">Now let&rsquo;s compare the outputs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids in both file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-1 -2 expected.taxid sourmash.taxid  | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
69</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids missed in the sourmash classification.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-2 -3 expected.taxid sourmash.taxid  | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
7</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Taxids present in the sourmash classification only.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-1 -3 expected.taxid sourmash.taxid  </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#6</font></font></p>
<p class="first-paragraph">what has been missed by sourmash:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">comm
-2 -3 expected.taxid sourmash.taxid | taxonkit lineage</font></font><br/>
<br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1100
   cellular organisms;Bacteria;FCB group;Bacteroidetes/Chlorobi
group;Chlorobi;Chlorobia;Chlorobiales;Chlorobiaceae;Chlorobium/Pelodictyon
group;Pelodictyon;Pelodictyon luteolum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1172
   cellular organisms;Bacteria;Terrabacteria
group;Cyanobacteria/Melainabacteria
group;Cyanobacteria;Nostocales;Nostocaceae;Anabaena;Anabaena
variabilis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1765
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Mycobacteriaceae;Mycobacterium;Mycobacterium
tuberculosis complex;Mycobacterium bovis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1831
   cellular organisms;Bacteria;Terrabacteria
group;Actinobacteria;Actinobacteria;Corynebacteriales;Nocardiaceae;Rhodococcus;Rhodococcus
sp.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">203804
 cellular
organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;unclassified
Enterobacteriaceae;ant, tsetse, mealybug, aphid, etc.
endosymbionts;ant endosymbionts;Candidatus Blochmannia</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">65
 cellular organisms;Bacteria;Terrabacteria
group;Chloroflexi;Chloroflexia;Herpetosiphonales;Herpetosiphonaceae;Herpetosiphon;Herpetosiphon
aurantiacus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">9
  cellular
organisms;Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Erwiniaceae;Buchnera;Buchnera
aphidicola</font></font></p>
<p class="first-paragraph">Quite a few are the same as before.</p>
<h2 class="western"><a name="what-is-binning"></a>123.13 What is
binning?</h2>
<p class="first-paragraph">Binning is the process of grouping reads
or contigs and assigning them to &ldquo;bins&rdquo; that will
represent taxonomical ranks or taxonomical units. Typically, binning
methods are based on sequence compositional features.</p>
<p>The incomplete nature of the sequencing copuled to amiguity
introduced by the method itself (breaking long pieces of DNA into
smaller fragments) makes it hard to assemble individual genes, much
less recovering the full genomes of each organism. Binning techniques
represent a &ldquo;best effort&rdquo; to identify reads or contigs
with certain groups of organism.</p>
<p>Typically reads are first assembled into longer contigs, then
these contigs are binned into categories.</p>
<h2 class="western"><a name="how-do-i-assemble-metagenomes"></a>123.14
How do I assemble metagenomes?</h2>
<p class="first-paragraph">This topic is so early in the development
of the field that we can&rsquo;t, in good faith cover it here. There
are many tools that can be run just as other assembler are run, but
assessing their results and understanding tuning their quality,
interpreting the assemblies are still in our opining no more than
just guesstimates.</p>
<p>Most practical advice follows the rule: &ldquo;Find an assembler
that others say is good, keep running it with different parameters
until the results seem to make sense&rdquo;. It is weak sauce - but
the pretty much the only advice we can share as well.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="alaska-oil-reservoir-metagenome-study"></a>
124 Alaska Oil Reservoir Metagenome study</h1>
<p class="first-paragraph">Note: this chapter is unfinished.</p>
<p>The reason we kept it is to show you what happens when you go to
remote part of Alaska and sequence samples from the ice.
Unsurprisingly only 10% of your samples can be assigned to known
species. But then you have to write a paper from that. Quite the
conundrum of course.</p>
<h2 class="western"><a name="which-data-are-we-going-to-re-analyze"></a>
124.1 Which data are we going to re-analyze?</h2>
<p class="first-paragraph">We will obtain data for a terrestrial
metagenome project with id <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRJNA278302</font></font></p>
<p><img src="0000_html_13689b2d0ffdf281.png" name="Image206" align="bottom" hspace="12" width="560" height="327" border="0"/>
</p>
<p>The URL for the bioproject is at:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA278302">https://www.ncbi.nlm.nih.gov/bioproject/PRJNA278302</a></font></p>
</ul>
<h2 class="western"><a name="how-do-i-get-more-information-on-the-project"></a>
124.2 How do I get more information on the project?</h2>
<p class="first-paragraph">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">runinfo</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">summary</font></font>
formats contain information on the runs:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA278302 | efetch -format runinfo &gt;
runinfo.txt</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-db sra -query PRJNA278302 | efetch -format summary &gt; summary.xml</font></font></p>
<p class="first-paragraph">In the latter case the result is an XML
file that you could view in a browser:</p>
<p><img src="0000_html_1bf3a9e9016b2dd2.png" name="Image207" align="bottom" hspace="12" width="560" height="209" border="0"/>
</p>
<p>Visualizing this file will allow you to understand its
(convoluted) structure. Then you can use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">xtract</font></font>
tool may be used to parse this XML file and extract various
components of it and format those as a tabular file.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
summary.xml | xtract -pattern &quot;EXPERIMENT_PACKAGE&quot; -element
PRIMARY_ID, TITLE | cut -f 8,10</font></font></p>
<p class="first-paragraph">will connect the SRR run numbers to the
title of each section.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1977365
 metagenome K3(S122) from petroleum reservoir (this well has been
treated with materials containing sulfide), Kuparuk formation, Alaska
North Slope</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1977357
 metagenome K2 (MPF) from not soured petroleum reservoir, Kuparuk
formation, Alaska North Slope</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1977304
 metagenome I2(PW) from soured petroleum reservoir, Ivishak
formation, Alaska North Slope</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1977296
 metagenome I1 (DS1215) from not soured petroleum reservoir , Ivishak
formation, Alaska North Slope</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1977249
 metagenome SB2 from not soured petroleum reservoir, Schrader bluffer
formation, Alaska North Slope</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1976948
 metagenome SB1 from not soured petroleum reservoir, Schrader bluffer
formation, Alaska North Slope</font></font></p>
<h2 class="western"><a name="how-do-i-classify-this-data"></a>124.3
How do I classify this data?</h2>
<p class="first-paragraph">We can employ several strategies discussed
in the previous chapters, though in general you should not be
surprised if these don&rsquo;t quite work at the level of accuracy of
precision that you were expecting:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the data for one run. Add the -X 100000 flag to limit the data to
a subset. </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump
--split-files -skip-technical  SRR1976948</font></font></p>
<p class="first-paragraph">Now it is worth checking out just how many
of our sequences can be classified before and after quality control:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken
-db /export/refs/kraken_db/kraken_std_db/ --threads 16 -fastq-input
SRR1976948_1.fastq SRR1976948_2.fastq &gt; results.out</font></font></p>
<p class="first-paragraph">in just 3 minutes the tool completes with
the following result:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">34456404
sequences (8648.56 Mbp) processed in 175.190s (11800.8 Kseq/m,
2962.00 Mbp/m).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 3720501 sequences classified (10.80%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 30735903 sequences unclassified (89.20%)</font></font></p>
<p class="first-paragraph">It reports that that majority of the reads
could not be classified.</p>
<h2 class="western"><a name="how-many-species-were-present"></a>124.4
How many species were present?</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken-report
-db /export/refs/kraken_db/kraken_std_db/ results.out &gt;
results.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt | awk ' $2&gt;10 &amp;&amp; $4==&quot;S&quot; { print
$0}' | wc -l</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1230</font></font></p>
<p class="first-paragraph">that starts with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">3.47
 1196351 0       S       301375   Methanosaeta harundinacea</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.06
 20863   0       S       2223     Methanosaeta concilii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01
 4390    0       S       420950   Methanolobus psychrophilus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 1136    0       S       2209     Methanosarcina mazei</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 571     0       S       2214     Methanosarcina acetivorans</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 377     0       S       2208     Methanosarcina barkeri</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 734     0       S       101192   Methanomethylovorans
hollandica</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 619     0       S       2176     Methanohalophilus mahii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 170     0       S       29291    Methanococcoides burtonii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.00
 73      0       S       2322     Methanohalobium evestigatum</font></font></p>
<h2 class="western"><a name="does-quality-control-improve-the-classification"></a>
124.5 Does quality control improve the classification?</h2>
<p class="first-paragraph">Let us apply quality control and see how
the classification performance may change with that.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Make the illumina adapter file.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;&gt;illumina&quot; &gt; adapter.fa</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
&quot;AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC&quot; &gt;&gt; adapter.fa</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Let apply adapter trimming and quality correction.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">trimmomatic
 PE -baseout trimmed.fq SRR1976948_1.fastq SRR1976948_2.fastq
ILLUMINACLIP:adapter.fa:2:30:5 SLIDINGWINDOW:4:30</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Error correction with tadpole</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tadpole.sh
in=trimmed_1P.fq in2=trimmed_2P.fq out=corrected_1P.fq
out2=corrected_2P.fq mode=correct k=50</font></font></p>
<p class="first-paragraph">Classify after error correction:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Classify with kraken</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken
-db /export/refs/kraken_db/kraken_std_db/ --threads 16 -fastq-input
corrected_1P.fq  corrected_2P.f &gt; results-corrected.out</font></font></p>
<p class="first-paragraph">the result is unexpected! The
classification produces fewer results overall only 2424 sequences
were classified instead of 7424:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">78272
sequences (14.88 Mbp) processed in 5.743s (817.7 Kseq/m, 155.45
Mbp/m).</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 2424 sequences classified (3.10%)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
 75848 sequences unclassified (96.90%)</font></font></p>
<p class="first-paragraph">In all honesty we don&rsquo;t know how to
explain this. Chalk it up to one of the many mysteries you will
undoubtebly run into in your career.The first rule of the
&ldquo;Bioinformatics Fight Club&rdquo; is that there are no rules.</p>
<p>We can format the result differently, though that alone won&rsquo;t
help with the results:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">  <font face="Consolas, serif"><font size="2" style="font-size: 11pt">kraken-report
-db ~/refs/kraken/minikraken_20141208 results.out &gt; results.txt</font></font></p>
<p class="first-paragraph">to filter at species level with at least
10 hits we can do:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
results.txt | awk ' $2&gt;10 &amp;&amp; $4==&quot;S&quot; { print
$0}' </font></font>
</p>
<p class="first-paragraph">that will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.44
   1442      0    S    28108     Alteromonas macleodii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.49
    494      0    S    85698     Achromobacter xylosoxidans</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01
     11      0    S    994695    Candidatus Kinetoplastibacterium
galatii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.07
     66      0    S    316277    Syntrophus aciditrophicus</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02
     17      0    S    181663    Desulfococcus oleovorans</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.35
    349      0    S    1184387   Mesotoga prima</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.10
    105      0    S    69499     Petrotoga mobilis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.04
     44      0    S    102134    Desulfotomaculum gibsoniae</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01
     11      0    S    58135     Desulfotomaculum kuznetsovii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.06
     58      0    S    110500    Pelotomaculum thermopropionicum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.01
     12      0    S    233055    Desulfitobacterium
dichloroeliminans</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02
     19      0    S    35701     Heliobacterium modesticaldum</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02
     21      0    S    336261    Thermovirga lienii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">1.98
   1980      0    S    301375    Methanosaeta harundinacea</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.02
     18      0    S    2223      Methanosaeta concilii</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.83
    827      0    S    2198      Methanoculleus marisnigri</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.45
    445      0    S    83986     Methanoculleus bourgensis</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.14
    140    140    S    1379702   Methanobacterium sp. MB1</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">0.41
    411    411    S    374840    Enterobacteria phage phiX174 sensu
lato</font></font></p>
<h2 class="western"><a name="are-there-other-methods-to-classify-samples"></a>
124.6 Are there other methods to classify samples?</h2>
<p class="first-paragraph">There is no shortage of approaches that
claim to do better. In reality the results are typically decidedly
less reliable.</p>
<h2 class="western"><a name="how-do-i-use-sourmash-to-understand-my-data"></a>
124.7 How do I use &ldquo;sourmash&rdquo; to understand my data?</h2>
<p class="first-paragraph">Based on the tool description</p>
<p class="block-text-western">sourmash can also be used to quickly
search large databases of genomes for matches to query genomes and
metagenomes;</p>
<p class="first-paragraph">First install the database:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/ref/sourmash</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~/ref/sourmash</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">wget
https://s3-us-west-1.amazonaws.com/spacegraphcats.ucdavis.edu/microbe-refseq-sbt-k21-2017.05.09.tar.gz</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">tar
xzvf microbe-refseq-sbt-k21-2017.05.09.tar.gz</font></font></p>
<p class="first-paragraph">You can now use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash</font></font>
to classify:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Compute a signature.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash
compute --scaled 10000 -k 21 SRR1976948_1.fastq</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Classify the signature.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sourmash
sbt_gather -k 21 ~/refs/sourmash/refseq-k21.sbt.json
SRR1976948_1.fastq.sig </font></font>
</p>
<p class="first-paragraph">it will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">overlap
   p_query p_genome</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-------
   ------- --------</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">130.0
kbp   0.9%      4.7%      NZ_LN515531.1 Methanobacterium
formicicu</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">120.0
kbp   0.8%      4.9%      NC_009051.1 Methanoculleus marisnigri
JR</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">60.0
 kbp   0.4%      2.3%      NC_017527.1 Methanosaeta harundinacea
6A</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">50.0
 kbp   0.3%      1.6%      NC_017934.1 Mesotoga prima MesG1.Ag.4.2,</font></font></p>
<p class="first-paragraph">so out of 100,000 reads it was able to
classify just 3 percent of them into 4 categories.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="analyzing-the-neonatal-microbiome"></a>125
Analyzing the neonatal microbiome</h1>
<p class="first-paragraph">Th following study contains analyis of
full datasets. You may complete the analysis even if your
computational resources by limiting the data to a smaller subset.</p>
<h2 class="western"><a name="how-to-classify-all-the-data-for-the-project-prjna46337"></a>
125.1 How to classify all the data for the project PRJNA46337?</h2>
<p class="first-paragraph">Suppose you&rsquo;d want to classify all
the data in project PRJNA46337? How much of computational resourse
would it take? The answer depends on a number of factors data
download speed as well as number of computer cores.</p>
<p>The project contains two kinds of data:</p>
<ol>
	<li/>
<p class="compact">1407 samples of 16S sequencing data
	producing on average 10000 reads per run</p>
	<li/>
<p class="compact">6 runs of whole genonome sequencing (WGS)
	on a Illumina Hiseq sequencer producing 70 million reads per sample.</p>
</ol>
<p class="first-paragraph">Here we demonstrate the analysis of the
1407 samples</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Get the runinfo. Took 40 seconds.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">esearch
-query PRJNA46337 -db sra | efetch -format runinfo &gt;
runinfo.csv</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
 Keep the runs performed on a 454 sequencer. </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
runinfo.csv | grep &quot;454 GS FLX Titanium&quot; | cut -f 1 -d , &gt;
srr-16s.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Store the 16S data here.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p data</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Download the SRR files 10 at a time. On our system this took about 60
minutes. </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
srr-16s.txt | parallel -j 10 --verbose --progress --eta fastq-dump
--outdir data --split-files --skip-technical -F</font></font></p>
<p class="first-paragraph">Annoyingly some SRR data was split into 3
pieces other into 4, hence the files that contain our biological data
are sometimes called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">..._3.fastq</font></font>
versus <font face="Consolas, serif"><font size="2" style="font-size: 11pt">..._4.fastq</font></font>.
These type of inconsistencies are very annoying and slow down the
automated processing of data. We do also note that some runs failed
to download for mysterious reasons. We have only <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1338</font></font>
datasets whereas the file lists <font face="Consolas, serif"><font size="2" style="font-size: 11pt">1397</font></font>
oh well - we just carry on.</p>
<p>We&rsquo;d like to rename each file in such a way to split off the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">_3</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">_4</font></font>
and keep just the SRA number itself turning <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1038144_3.fastq</font></font>
into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR1038144.fastq</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052332_4.fastq</font></font>
into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052332.fastq</font></font>
respectively. But guess what, it is time to get in touch with that
data janitor slice of your soul because there is another complication
that you probably miss the first time around and create the wrong
solution first (like we did). The number after the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR</font></font>
varies too, it may be 6 or 7 digits long so it renaming won&rsquo;t
be as simple as slicing of the string at a position (an easy
operation in bash). Instead we have to find and remove either
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">_3.fastq</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">_4.fastq</font></font>
from the end of a string. Thankfully there is an operator to do that,
and you find it by googling it, the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">%</font></font>
operator.</p>
<p>The sloppines of the data modeling at SRA causes all kinds of
unexpected complications right from the beginning.</p>
<p>So we need a renaming script. Call that <font face="Consolas, serif"><font size="2" style="font-size: 11pt">rename.sh</font></font>
and have it contain:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">for
file in *.fastq; do</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   name=$file</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # Chop off certin types of ends if they exists.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   name=${name%_3.fastq}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   name=${name%_4.fastq}</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # Perform the renaming.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   echo mv $file &quot;${name}.fq&quot;</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">done</font></font></p>
<p class="first-paragraph">Running this script in the data folder:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
rename.sh </font></font>
</p>
<p class="first-paragraph">will show you the operations that it wants
to perform. To have the actions perform instead of showing them to
you pipe the output again into bash:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash
rename.sh | bash</font></font></p>
<p class="first-paragraph">We now have nicely named file. Let&rsquo;s
classify them all:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
This directory will store the results of classification.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p results</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Classify in parallel 10 at a time.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
srr-16s.txt | parallel -j 10 --verbose --progress --eta java -jar
~/src/RDPTools/classifier.jar classify data/{}.fq -f allrank -o
results/{}.assign.txt -h results/{}.ranks.txt</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Classify all 1381 columns.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">time
java -jar ~/src/RDPTools/classifier.jar classify data/*.fq -f allrank
-o results/assign.txt -h results/ranks.txt</font></font></p>
<p class="first-paragraph">On our system this took about 7 hours and
you end up with a file with 0ver 1000 columns where each sample is
fully classified.</p>
<p>The datamash utility can help us probe the data to compute various
values:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
counts.txt | datamash -H sum 2-8</font></font></p>
<p class="first-paragraph">produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sum(SRR052150.fq)
  sum(SRR052151.fq)   sum(SRR052152.fq)   sum(SRR052153.fq)  
sum(SRR052154.fq)   sum(SRR052155.fq)   sum(SRR052156.fq)</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">2509
   3346    3379    3302    2988    2905    3350</font></font></p>
<p class="first-paragraph">The above shows the number of classified
sequences per sample. Pretty low actually. Let us count how many
bacteria of the Streptoccocus genus each column has</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
counts.txt | grep Streptococcus | cut -f 1-20 | head</font></font></p>
<p class="first-paragraph">it produces:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus
  191 290 27  2536    1   0   0   309 1   0   10  370 86  1   392 166
37  0</font></font></p>
<p class="first-paragraph">It is pretty clear that the variability
accross samples is gigantic. Some samples indicate no such bacteria
is present (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052155.fq</font></font>)
whereas others (example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052153.fq</font></font>)
show that over 75% of the bacteria (2536/3302) of the sequences
assigned to this genus. This genus include over 50 species of
bacteria and while many have quite the bad rep for causing pink-eye,
meningitis, pneumonia many more are not pathogenic at all and are
often present in healthy guts as well. In a manner of speaking this
makes all classifications at higher levels of taxonomy less
informative than initially assumed.</p>
<p>Now due to privacy considerations scientists were not allowed to
distribute phenotype data of clinical relevance. For example you
can&rsquo;t immediately access information that would connect samples
to a disease phenotype. You would need to apply for permission to
obtain that. The typical process here would create subgroups of our
columns then attempt to discover trends in data for these.</p>
<h2 class="western"><a name="what-is-the-most-common-misunderstanding-paradox-of-metagenomics"></a>
125.2 What is the most common misunderstanding (paradox) of
metagenomics?</h2>
<p class="first-paragraph">Look at the results above. Note how
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052155.fq</font></font>
has no reads assigned to the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus</font></font>
genus whereas <font face="Consolas, serif"><font size="2" style="font-size: 11pt">SRR052153.fq</font></font>
has 75% of its reads assigned to it. As yourself the following</p>
<p class="block-text-western">Would you conclude that the first
sample contains fewer bacteria at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus</font></font>
genus than the second?</p>
<p class="first-paragraph">As surprising as it sounds the answer is:
NO. The only thing you may conclude is that, when compared to all the
other existing bacteria, relatively speaking there are fewer
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus</font></font>
bacteria in the first sample than the other.</p>
<p>Here is a better explanation. Imagine that both samples have the
same number of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus</font></font>
bacteria in them makes that 10,000. But the first sample also has
many millions of other bacteria and our random sampling of 2000 never
had the chance to pick them up, they always hit something else. In
the second sample there were very few other bacteria, our sampling
mostly hit the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Streptococcus</font></font>
variety. Hence turned out at 75%.</p>
<p>See how the counts reflect the relative abundances and not
absolute ones. And of course that also provides us with very valuable
information. We only wish to caution against interpreting the
classification rates incorrectly.</p>
<h2 class="western"><a name="how-reliable-are-these-results"></a>125.3
How reliable are these results?</h2>
<p class="first-paragraph">You may wonder, just how reliable is the
observation that 75% of the bacteria in a certain gut sample belongs
to the Streptococcus genus? The short answer is that probably not
that reliable.</p>
<p>The entire field of metagenomics is at its infancy. It is not
entirely clear how to sample the same gut correctly after all there
are many imhomogenities in different areas of the gut. There is
variability and errors in sample extraction, handling and of the
lenghty process of analysis.</p>
<p>A reliable method would be one that replicated each measurement
multiple times and had the means to validate the accuracy and
precision of each obtained value. The field is extremely far from
doing that.</p>
<p>Instead of proper experimental design and validation scientists
perfer to rely on &ldquo;fancy&rdquo; statistical methods that few
(usually not even the people that use them) understand. There is a
long way to go yet in obtaining reliabke results.</p>
<p>Metagenomics must still be considered as an early, exploratory
science where we still bumbling around.</p>
<p class="compact"> 
</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="setting-up-macos"></a>126 Setting up
MacOS</h1>
<h2 class="western"><a name="how-do-i-get-started"></a>126.1 How do I
get started?</h2>
<p class="first-paragraph">First of all, ensure that your Mac runs
the latest version of MacOS. Go to the</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&quot;Apple
Icon&quot; -&gt; &quot;About This Mac&quot; -&gt; &quot;Software
Update&quot;.</font></font></p>
<p class="first-paragraph">Verify that your Mac is updated to the
latest version of MacOS.</p>
<h2 class="western"><a name="if-you-have-apple-silicon-m1-processor"></a>
126.2 If you have Apple Silicon (M1 processor)</h2>
<p class="first-paragraph">The following applies only for Macs with
M1 processor!</p>
<p>Apple Silicon is the designation of a family of processors
designed by Apple using the ARM architecture. To make the new CPU
architecture compatible with existing software Apple also implemented
a &ldquo;translation layer&rdquo; called Rosetta. If you own a new
Mac with the M1 processor you will need to install the Rosetta
emulator. In a nutshell, you will open a terminal and type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">softwareupdate
--install-rosetta</font></font></p>
<p class="first-paragraph">Read more here:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://osxdaily.com/2020/12/04/how-install-rosetta-2-apple-silicon-mac/">https://osxdaily.com/2020/12/04/how-install-rosetta-2-apple-silicon-mac/</a></font></p>
</ul>
<p class="first-paragraph">To ensure that the new Terminals that you
open later run in Rosetta mode you would also need to follow these
instructions:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://osxdaily.com/2020/11/18/how-run-homebrew-x86-terminal-apple-silicon-mac/">https://osxdaily.com/2020/11/18/how-run-homebrew-x86-terminal-apple-silicon-mac/</a></font></p>
</ul>
<p class="first-paragraph">The steps described in the link above are:</p>
<ol>
	<li/>
<p class="compact">Locate the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Terminal</font></font>
	application within the Utilities folder (<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Finder
	&gt; Go menu &gt; Utilities</font></font>)</p>
	<li/>
<p class="compact">Select <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Terminal.app</font></font>
	and right-click on it, then choose <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Duplicate</font></font>.</p>
	<li/>
<p class="compact">Rename the duplicated Terminal app
	something obvious and distinct, like <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bio
	Terminal</font></font></p>
	<li/>
<p class="compact">Select the freshly renamed <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bio
	Terminal</font></font> app and right-click and choose <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Get
	Info</font></font> (or hit Command+i).</p>
	<li/>
<p class="compact">Check the box for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Open
	using Rosetta</font></font>, then close the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Get
	Info</font></font> window.</p>
	<li/>
<p class="compact">Run the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bio
	Terminal</font></font> instead of the original <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Terminal.app</font></font>.</p>
</ol>
<p class="first-paragraph">The best is to pin the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Bio
Terminal</font></font> to your Dock or make an alias of it and drag
it to your Desktop so that you can easily start it in the future.</p>
<h2 class="western"><a name="change-the-shell-to-bash"></a>126.3
Change the shell to BASH!</h2>
<p class="first-paragraph">Starting with Mac OSX version 10.14 the
default shell has been changed to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">zsh</font></font></p>
<p>All our instructions use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bash</font></font>
shell, thus you will need to configure the Mac OSX to use bash by
default as well. Type the following into a terminal:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">chsh
-s /bin/bash</font></font></p>
<p class="first-paragraph">It will ask for your password then set the
default shell.</p>
<h2 class="western"><a name="what-other-app-store-software-is-needed"></a>
126.4 What other App Store software is needed?</h2>
<p class="first-paragraph">A compiler is a software that turns a
&ldquo;text&rdquo; based program into one that the computer can
execute. You will need to install the &ldquo;XCode&rdquo; compiler.
Go to the &ldquo;App Store&rdquo; and find the &ldquo;XCode&rdquo;
application. Install it, it is a free program.</p>
<p><img src="0000_html_da8c6217be38308a.png" name="Image208" align="bottom" hspace="12" width="560" height="283" border="0"/>
</p>
<p>You will also need to install the additional command line tools to
XCode by opening the &ldquo;Terminal&rdquo; application, then writing
the following into it:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">xcode-select
--install</font></font></p>
<p class="first-paragraph">This will trigger an installation
procedure.</p>
<p>You may occasionally need to re-launch the XCode application
whenever the license agreement created by Apple changes. You will get
messages of the sort:</p>
<p class="block-text-western">Agreeing to the Xcode/iOS license
requires admin privileges, please run &ldquo;sudo xcodebuild
-license&rdquo; and then retry this command.</p>
<p class="first-paragraph">You can accept a license agreement change
from the command line by typing (requires your password):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
xcodebuild -license accept</font></font></p>
<p class="first-paragraph">The automatic software update will update
XCode as well. In those cases, you may need to re-accept the license.</p>
<h2 class="western"><a name="how-do-i-view-my-home-directory-in-the-finder"></a>
126.5 How do I view my home directory in the Finder?</h2>
<p class="first-paragraph">By default, the Finder will now show the
home directory that your Terminal opens to.</p>
<p>To be able to view your home directory within the Mac Finder you
need to tick a checkbox next to it. Go to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Finder-&gt;Preferences</font></font>.
Your home directory is named as your username. My username is
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ialbert</font></font>,
yours will be different. See below:</p>
<p><img src="0000_html_e2eabc8d5d6d9b20.png" name="Image209" align="bottom" hspace="12" width="473" height="833" border="0"/>
</p>
<h2 class="western"><a name="will-the-finder-show-all-files"></a>126.6
Will the Finder show all files?</h2>
<p class="first-paragraph">Some files won&rsquo;t be visible via the
Finder. For example, the Finder will not show so-called &ldquo;hidden&rdquo;
files, these are files whose name starts with a period: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font></p>
<p>There is nothing really &ldquo;hidden&rdquo; about these files, it
is just a convention that Unix adopted. These files are typically
used for software configuration and the Finder tries to keep you from
accidentally altering/deleting them.</p>
<p>To make the Finder show the &ldquo;hidden&rdquo; files press
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Command+Shift+.</font></font>
where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
is the dot character. Note, you don&rsquo;t &ldquo;have to&rdquo;
turn on the hidden file view, the instruction here is mostly a
reminder of how to do it if needed.</p>
<h2 class="western"><a name="what-is-the-next-step-2"></a>126.7 What
is the next step?</h2>
<p class="first-paragraph">Continue of with the steps described in
<font color="#4f81bd"><a href="#ch126.xhtml#terminal-init">Initialize
the terminal</a></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="setting-up-linux"></a>127 Setting up
Linux</h1>
<h2 class="western"><a name="which-distributions-of-linux-does-this-book-support"></a>
127.1 Which distributions of Linux does this book support?</h2>
<p class="first-paragraph">There are numerous Linux distributions
with various naming schemes.</p>
<p>To set up their system users will need to install packages via
so-called package managers: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">apt-get</font></font>
or <font face="Consolas, serif"><font size="2" style="font-size: 11pt">yum</font></font>
on Fedora based systems.</p>
<p>Our commands assume an Ubuntu/Debian based systems.</p>
<h2 class="western"><a name="what-is-the-first-step-for-linux"></a>127.2
What is the first step for Linux?</h2>
<ol>
	<li/>
<p class="compact">Update your Linux system.</p>
	<li/>
<p class="compact">Install the required Linux libraries and
	Java.</p>
</ol>
<h2 class="western"><a name="how-do-i-update-a-linux-based-system"></a>
127.3 How do I update a Linux based system?</h2>
<p class="first-paragraph">On Ubuntu Linux start a &ldquo;Terminal&rdquo;
then run the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
apt-get update &amp;&amp; sudo apt-get upgrade -y</font></font></p>
<p class="first-paragraph">These commands will update your
distribution (while printing copious amounts of information on the
screen) then upgrade all installed packages.</p>
<h2 class="western"><a name="what-are-the-required-libraries-for-linux"></a>
127.4 What are the required libraries for Linux?</h2>
<p class="first-paragraph">Depending on the Linux distribution&rsquo;s
initial configuration one may need to install the package shown
below. On Ubuntu the following was necessary:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
apt-get install -y curl unzip build-essential ncurses-dev </font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
apt-get install -y byacc zlib1g-dev python-dev git cmake</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">sudo
apt-get install -y default-jdk ant</font></font></p>
<p class="first-paragraph">Wait for these to complete.</p>
<h2 class="western"><a name="what-is-the-next-step-3"></a>127.5 What
is the next step?</h2>
<p class="first-paragraph">Continue of with the steps described in
<font color="#4f81bd"><a href="#ch127.xhtml#terminal-init">Initialize
the terminal</a></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="setting-up-windows"></a>128 Setting up
Windows</h1>
<p class="first-paragraph">Starting with the August 2016 Anniversary
Edition the Windows 10 operating system supports running Linux based
software.</p>
<p><img src="0000_html_f71ee54a2b5a65b2.png" name="Image210" align="bottom" hspace="12" width="152" height="163" border="0"/>
</p>
<p>The Linux subsystem for Windows is a fully configurable Linux
environment. Setting it up takes a few extra steps, but once you
follow those you&rsquo;ll have a fully functional Linux Environment
embedded right into Windows.</p>
<p class="compact">Installation processes may change. We try our best
to keep the book up to date, but occasionally we may fail to notice
changes in the latest recommended practices. Let us know via the
<font color="#4f81bd"><a href="https://github.com/biostars/biostar-handbook-issues">Issue
Tracker</a></font> if the installation instructions below appear ouf
of date. See also <font color="#4f81bd"><a href="https://aka.ms/wslinstall">https://aka.ms/wslinstall</a></font>
for Microsoft recommended instructions.</p>
<h2 class="western"><a name="how-to-enable-linux-on-the-windows-system"></a>
128.1 How to enable Linux on the Windows System?</h2>
<p class="first-paragraph">First you need to turn on the Windows
Subsytem for Windows</p>
<p>In your Windows search bar type the following: <i>Turn Windows
features on or off</i> to find the settings named the same. Launch
the settings panel and ensure that the <b>Windows Subsystem for Linux</b>
option is selected as below.</p>
<p><img src="0000_html_d7317d9acc6a81f1.png" name="Image211" align="bottom" hspace="12" width="560" height="153" border="0"/>
</p>
<p>Restart your computer for the changes to take effect.</p>
<h2 class="western"><a name="how-do-i-install-ubuntu-linux-on-windows"></a>
128.2 How do I install Ubuntu Linux on Windows?</h2>
<p class="first-paragraph">In a nutshell:</p>
<ol>
	<li/>
<p class="compact">Use the Start menu to launch the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Microsoft
	Store</font></font> application.</p>
	<li/>
<p class="compact">Search for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Ubuntu</font></font>
	and select the first result, &lsquo;Ubuntu&rsquo;, published by
	Canonical Group Limited.</p>
	<li/>
<p class="compact">Click on the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">Install</font></font>
	button.</p>
</ol>
<h2 class="western"><a name="how-do-i-start-ubuntu-linux"></a>128.3
How do I start Ubuntu Linux?</h2>
<p class="first-paragraph">Double-click the icon for the Ubuntu
subsystem.</p>
<p>The first time Ubuntu starts it will ask you for a username and
password.</p>
<p>Make the username short and simple word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">biouser</font></font>
or something similar. You can reuse your password that you use for
Windows or pick something else. The username/password combo will be
specific to Linux and won&rsquo;t be used within Windows in any other
context.</p>
<p><img src="0000_html_b964198f212b39c9.png" name="Image212" align="bottom" hspace="12" width="560" height="324" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-see-the-unix-filesystem-from-windows"></a>
128.4 How do I see the Unix filesystem from Windows?</h2>
<p class="first-paragraph">It is extremely useful to be able to
access the files from both Linux and with Windows Explorer. You want
to be able to &ldquo;see&rdquo; your files from both the terminal and
from Windows Explorer. To launch a new Windows Explorer connected
your filesystem type the following into your terminal:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">explorer.exe
.</font></font></p>
<p class="first-paragraph">Note the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.</font></font>
at the end. This instructs the explorer program to open and display
the directory you typed the command in. No matter what directory you
are in you can pop the an explorer window right there with the
command shown above.</p>
<p><b>Note</b>: The Windows Explorer window that you start with
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">explorer.exe
.</font></font> may not always auto-refresh. If you create a file
under Linux while the Explorer Window is open you may need to right
click and select &ldquo;Refresh&rdquo; to see the file. Hopefully
this behavior will change (may have already changed by the time you
read this).</p>
<p><img src="0000_html_31fa8f4bef039f8e.png" name="Image213" align="bottom" hspace="12" width="560" height="275" border="0"/>

This first approach is quite simple and handy, though having to
manually refresh to see new files created after the windows was
opened can be confusing.</p>
<p><b>Alternative approach for accessing your Linux files</b></p>
<p>It is also possible to mount the filesystem directly, though a bit
more tedious as the Unix filesystem is tucked away in spurious
corners of your Windows machine.</p>
<p>Specifically when installed as an Ubuntu App your filesystem will
be located at (the path below is so long that you will need to scroll
to see it all:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">C:\Users\&lt;Windows
UserName&gt;\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc\LocalState\rootfs\home\&lt;Ubuntu
UserName&gt;</font></font></p>
<p class="first-paragraph">Where <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;Windows
UserName&gt;</font></font> and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">&lt;Ubuntu
UserName&gt;</font></font> are the usernames that are set for Windows
and Ubuntu respectively.</p>
<p>Here is how it may look when using the File Explorer on Windows</p>
<p><img src="0000_html_748a617c453e0a65.png" name="Image214" align="bottom" hspace="12" width="560" height="74" border="0"/>
</p>
<p>For easy access to this monstrous path:</p>
<ol>
	<li/>
<p class="compact">Locate it via Windows Explorer</p>
	<li/>
<p class="compact">Right click to create a shortcut to it on
	your Desktop</p>
</ol>
<p class="first-paragraph">When mounted in this second way, the
filesystem will automatically refresh, hence a bit more convenient.</p>
<p>A report from a reader:</p>
<p class="block-text-western">A user was unable to locate the file
path because the AppData folder is hidden in windows.</p>
<p class="block-text-western">They manually navigated to
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">C:\Users\&lt;Windows
UserName&gt;\</font></font> then got stuck because <font face="Consolas, serif"><font size="2" style="font-size: 11pt">AppData</font></font>
was not listed.</p>
<p class="block-text-western">The solution was to type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">\AppData</font></font>
into the address bar and press enter.</p>
<h2 class="western"><a name="how-to-start-a-text-editor-in-your-terminal"></a>
128.5 How to start a text editor in your terminal?</h2>
<p class="first-paragraph">If you install the Microsoft Visual Code
editor as described in the section titled <font color="#4f81bd"><a href="#ch013.xhtml#choose-editor">Choose
a Text Editor</a></font> then you can also start from Linux by
typing:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">code
.</font></font></p>
<p class="first-paragraph">The command above will start a <b>Microsoft
Visual Code</b> editor connected to the Linux Filesystem. The code
that you write in the editor when run in the editor will execute
under the Linux environment. If you were to start the editor from
Windows it would not see the Linux filesystem. For more information
consult the page:</p>
<ul>
	<li/>
<p class="compact"><font color="#4f81bd"><a href="https://code.visualstudio.com/docs/remote/wsl">Developing
	in WSL</a></font></p>
</ul>
<h2 class="western"><a name="what-does-not-work-at-the-command-line-in-windows"></a>
128.6 What does not work at the command line in Windows?</h2>
<p class="first-paragraph">Tools that generate graphical user
interfaces need to be installed in Windows, then run through Windows.</p>
<p>For example, IGV, the Integrative Genome Viewer has a graphical
interface. To run IGV, download the Windows version of it, then, to
load up files created from the terminal locate your Linux filesystem
from Windows.</p>
<h2 class="western"><a name="how-do-i-customize-the-windows-bash-terminal"></a>
128.7 How do I customize the Windows Bash terminal?</h2>
<p class="first-paragraph">The default settings for the terminal can
be hard on the eyes. Right-click the terminal bar to customize the
colors and fonts sizes.</p>
<p><img src="0000_html_55a99831798fd0dc.png" name="Image215" align="bottom" hspace="12" width="560" height="179" border="0"/>
</p>
<p>Here is a closeup to the settings panel:</p>
<p><img src="0000_html_72bbb53ff1d0f223.png" name="Image216" align="bottom" hspace="12" width="394" height="473" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-finish-setting-up-ubuntu-on-windows"></a>
128.8 How do I finish setting up Ubuntu on Windows?</h2>
<p class="first-paragraph">Once the Ubuntu subsystem is installed as
above you can treat the terminal as a standalone Linux computer.</p>
<p>Follow the <font color="#4f81bd"><a href="#ch127.xhtml#setup-linux">Setup
Linux</a></font> page an type these commands into your Ubuntu
terminal.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="setting-up-your-bash-profile"></a>129
Setting up your Bash profile</h1>
<p class="first-paragraph">You may want to apply certain settings
automatically when a terminal is opened.</p>
<h2 class="western"><a name="what-is-a-shell-1"></a>129.1 What is a
&ldquo;shell&rdquo;?</h2>
<p class="first-paragraph">As the Wikipedia page on <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Shell_(computing)">Shell(computing)</a></font>
states:</p>
<p class="block-text-western">In computing, a shell is a user
interface for access to an operating system&rsquo;s services. In
general, operating system shells use either a command-line interface
(CLI) or graphical user interface (GUI), depending on a computer&rsquo;s
role and particular operation. It is named a shell because it is a
layer around the operating system kernel.</p>
<h2 class="western"><a name="what-is-bash"></a>129.2 What is Bash?</h2>
<p class="first-paragraph">Bash is a Unix shell and command language
written for the GNU project as a free software replacement for the
Bourne shell. The name is an acronym for &ldquo;Bourne-again shell&rdquo;.</p>
<h2 class="western"><a name="what-are-shell-profiles"></a>129.3 What
are shell profiles?</h2>
<p class="first-paragraph">A file that is processed when a shell is
opened is called a <i>shell profile</i>.</p>
<p>This file has to have a special name and has to be located in the
home directory. All the commands listed in this shell profile file
will be applied when a new shell is initialized.</p>
<h2 class="western"><a name="can-we-have-multiple-shell-profiles"></a>
129.4 Can we have multiple shell profiles?</h2>
<p class="first-paragraph">Yes. Due to historical and other usage
considerations, multiple files may contain shell configurations.
There are rules regarding which of these settings are applied when
more than one shell profile file is present.</p>
<p>Josh Staiger in <font color="#4f81bd"><a href="http://www.joshstaiger.org/archives/2005/07/bash_profile_vs.html">.bash_profile
vs .bashrc</a></font> writes:</p>
<p class="block-text-western">When you login (type username and
password) via console, either sitting at the machine, or remotely via
ssh: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>
is executed to configure your shell before the initial command
prompt. But, if you&rsquo;ve already logged into your machine and
open a new terminal window (xterm) inside Gnome or KDE, then <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
is executed before the window command prompt. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
is also run when you start a new bash instance by typing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/bin/bash</font></font>
in a terminal.</p>
<p class="block-text-western">An exception to the terminal window
guidelines is Mac OS X&rsquo;s Terminal.app, which runs a login shell
by default for each new terminal window, calling <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>
instead of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>.
Other GUI terminal emulators may do the same, but most tend not to.&rdquo;</p>
<h2 class="western"><a name="whats-the-best-setup-for-multiple-shell-profiles"></a>
129.5 What&rsquo;s the best setup for multiple shell profiles?</h2>
<p class="first-paragraph">You should handle this in the following
way:</p>
<ol>
	<li/>
<p class="compact">The file called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>
	should be set up to automatically load <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>.</p>
	<li/>
<p class="compact">The main shell profile file that you use
	should be <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
	and it should contain all shell related settings.</p>
</ol>
<p class="first-paragraph">The easiest way to set this up is to add
the following lines to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>if</b></font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
[</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">-f</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
]</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">;
</font></font><font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>then</b></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  </font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">source</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font><br/>
<font color="#007020"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><b>fi</b></font></font></font></p>
<p class="first-paragraph">This command loads <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
when loading the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>.
You will never need to look at the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bash_profile</font></font>
file again. From now on you need to only change the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
file.</p>
<p>Both of these files probably already exist on your system, if not
you can create them as below.</p>
<h2 class="western"><a name="how-can-i-create-the-minimal-settings"></a>
129.6 How can I create the minimal settings?</h2>
<p class="first-paragraph">You may also apply our recommended
settings with the following (apply only once!):</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/install/bash_profile.txt </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bash_profile</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">curl</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
http://data.biostarhandbook.com/install/bashrc.txt </font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<h2 class="western"><a name="what-should-my-.bashrc-file-contain"></a>
129.7 What should my .bashrc file contain?</h2>
<p class="first-paragraph">Minimally we recommend the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
A minimal BASH profile.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
ON Mac OS uncomment the line below.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
alias ls='ls -hG'</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
On Linux use the following:</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
alias ls='ls -h --color'</i></font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Safe versions of the default commands.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Will ask permissions before overwriting files.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
rm=</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'rm
-i'</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
mv=</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'mv
-i'</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">alias</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
cp=</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'cp
-i'</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Extend the program search PATH and add the ~/bin folder.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">export</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin:</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Makes the prompt much more user friendly.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
But I do agree that the command looks a bit crazy.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">export</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PS1=</font></font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'\[\e]0;\w\a\]\n\[\e[32m\]\u@\h
\[\e[33m\]\w\[\e[0m\]\n\$ '</font></font></font><br/>
<br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
This is required for Entrez Direct to work.</i></font></font></font><br/>
<font color="#60a0b0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt"><i>#
Disables strict checking of an encryption page.</i></font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">export</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PERL_LWP_SSL_VERIFY_HOSTNAME=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">0</font></font></p>
<h2 class="western"><a name="can-i-customize-the-terminal-colors-and-fonts"></a>
129.8 Can I customize the terminal colors and fonts?</h2>
<p class="first-paragraph">Yes, and please do so. You can vary font
sizes, background colors etc. It makes a tremendous difference when
reading it for extended periods of time. Here is an easy to read
terminal:</p>
<p><img src="0000_html_2f36de05b0f3b844.png" name="Image217" align="bottom" hspace="12" width="560" height="367" border="0"/>
</p>
<h2 class="western"><a name="how-do-i-edit-my-shell-profile"></a>129.9
How do I edit my shell profile?</h2>
<p class="first-paragraph">You can use the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font>
editor</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bash_profile</font></font></p>
<p class="first-paragraph">then:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<h2 class="western"><a name="how-do-i-activate-a-shell-profile"></a>129.10
How do I activate a shell profile?</h2>
<p class="first-paragraph">Changing a shell profile file will not
automatically apply the changed settings to terminals that are
already open.</p>
<p>New terminals will use the new settings but existing terminal
windows need to be instructed to apply the new settings by sourcing
the file in question:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">source</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<h2 class="western"><a name="troubleshooting-the-path"></a>129.11
Troubleshooting the PATH</h2>
<p class="first-paragraph">It is surprisingly easy to accidentally
break everything so thoroughly that not even basic UNIX commands work
anymore.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-bash</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
ls: command not found</font></font></p>
<p class="first-paragraph">Sooner or later we all do this. Wear it as
a badge of honor.</p>
<h3 class="western"><a name="how-did-i-nuke-my-path"></a>129.11.1 How
did I nuke my path?</h3>
<p class="first-paragraph">If you mistype the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
variable while adding it into the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
shell profile you may end up with the unenviable situation of
applying an incorrect setting every time your terminal initializes.
For example, on a beautiful and sunny morning you&rsquo;ll do a:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'export
PATH=~/bin:PATH'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<p class="first-paragraph">It all looks great, but there is a fatal
typo there. The net effect of it all will be that even basic system
programs that used to work before won&rsquo;t do so anymore:&rdquo;</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
</font></font><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ls</font></font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">-bash</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">:
ls: command not found</font></font></p>
<p class="first-paragraph"><b>Congratulations!</b> You&rsquo;ve just
<b>nuked</b> your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>.
You forgot the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$</font></font>
sign on the right hand side of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
variable. Instead of using the value of variable named <font face="Consolas, serif"><font size="2" style="font-size: 11pt">$PATH</font></font>
you are setting it to be the actual word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>.
Your command should have looked like this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo</font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#4070a0"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">'export
PATH=~/bin:$PATH'</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
</font></font><font color="#666666"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;&gt;</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<p class="first-paragraph">First let&rsquo;s look at the bright side
&ndash; there is probably no command line guru out there that did not
do this at one point in their career. Obviously you are well on your
way towards &ldquo;guru-ness&rdquo;. Rock on!</p>
<p>Jules Winnfield of <font color="#4f81bd"><a href="https://en.wikipedia.org/wiki/Pulp_Fiction">Pulp
Fiction</a></font> explains it best:</p>
<p class="block-text-western">The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
to any righteous tool is beset on all sides by the inequities of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">UNIX</font></font>
and the tyranny of Environment Variables. Blessed is he who, in the
name of charity and good will, shepherds the weak through the valley
of darkness, for he is truly his brother&rsquo;s keeper and the
finder of lost programs.</p>
<h3 class="western"><a name="how-do-i-fix-my-nuked-path"></a>129.11.2
How do I fix my nuked path?</h3>
<p class="first-paragraph">Access your terminal. Oh yes, it is still
broken. But you can temporarily override the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
within this terminal with:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#19177c"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH=</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/bin:/usr/bin</font></font></p>
<p class="first-paragraph">Now edit your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">.bashrc</font></font>
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font>
and remove the lines that cause the trouble:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font color="#06287e"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font></font><font face="Consolas, serif"><font size="2" style="font-size: 11pt">
~/.bashrc</font></font></p>
<p class="first-paragraph">Now <font face="Consolas, serif"><font size="2" style="font-size: 11pt">nano</font></font>
is a bit quirky for a text editor. Take it as punishment meted out
for clobbering your path. It could be worse. At least you don&rsquo;t
have to use <font face="Consolas, serif"><font size="2" style="font-size: 11pt">vi</font></font>
(But some of us love vi!).</p>
<p>Edit the file, correct the offending line(s), then save the file
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ctrl
+ O</font></font> (that is a capital letter O), then exit the editor
with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">ctrl
+ X</font></font>.</p>
<p>Open a NEW terminal window and verify that your &ldquo;corrections&rdquo;
work. In times like these it is very easy to make another error while
&ldquo;fixing&rdquo; the first one. Rinse and repeat until everything
checks out. It will eventually.</p>
<p><b>Good luck, comrade!</b></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="how-do-i-set-up-my-path"></a>130 How do
I set up my PATH?</h1>
<p class="first-paragraph">One of the most frustrating issues faced
by newcomers to the Unix command line is how to run a program they
have just installed. This typically happens if they install a program
from source instead of via a global installer.</p>
<p>Suppose you install the program <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>
but then as you try to run it you could get this:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">$
fastq-dump</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump:
command not found</font></font></p>
<p class="first-paragraph">The main concept to remember here is that
our computers look at only a few locations when trying to find a
program. When bash cannot run a program it is because it cannot &ldquo;see&rdquo;
the program. When we install a new program ourselves, we need to tell
our computer where to look for it. There are different ways to do
this:</p>
<h2 class="western"><a name="solution-1-use-the-full-program-path"></a>
130.1 Solution 1: Use the full program path</h2>
<p class="first-paragraph">Run the program by its fully qualified
path (location). First you need to figure out where that is. Go to
the location where you installed the software. From that software
directory, type <font face="Consolas, serif"><font size="2" style="font-size: 11pt">pwd</font></font>
and that will tell you the full path that you can then use:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/src/sratoolkit.2.5.2-mac64/bin/fastq-dump</font></font></p>
<p class="first-paragraph">We may use a shorter form when the program
directory opens from our home directory:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/src/sratoolkit.2.5.2-mac64/bin/fastq-dump</font></font></p>
<p class="first-paragraph">If that works, then we can always invoke
the program by this full name. It is a bit hard on eyes. For a
simpler way, see the next solution.</p>
<h2 class="western"><a name="solution-2-extend-the-search-path-to-include-the-installation-directory."></a>
130.2 Solution 2: Extend the search PATH to include the installation
directory.</h2>
<p class="first-paragraph">If we wanted to be able to run the program
just by typing its name (say <font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font>)
we need to instruct the computer to also look for programs to run in
the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">/Users/ialbert/src/sratoolkit.2.5.2-mac64/bin/</font></font>
directory. To do so, we need add this location to the so called
<i>search path</i> of the shell.</p>
<p>The shell looks at the variable called <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
and searches all locations listed in the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
for names of programs to run. What is the current value of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>?
Type the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">echo
$PATH</font></font></p>
<p class="first-paragraph">This will produce a colon <font face="Consolas, serif"><font size="2" style="font-size: 11pt">:</font></font>
separated list of all the locations that bash will look in when
searching for a program. We can extend this by adding the new
location via the following construct:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">export
PATH=~/src/sratoolkit.2.5.2-mac64/bin:$PATH</font></font></p>
<p class="first-paragraph">The new <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
will be constructed as the new directory + <font face="Consolas, serif"><font size="2" style="font-size: 11pt">:</font></font>
+ old value of the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>.
Now we should be able to run the program as</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">fastq-dump</font></font></p>
<p class="first-paragraph">Now, the above is a solution, but typing
this (and all other) <font face="Consolas, serif"><font size="2" style="font-size: 11pt">export</font></font>
commands we might add every time we open a new terminal is too
tedious. We would be better off applying the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PATH</font></font>
and other settings automatically. To do this, we must take one more
step down the system admin rabbit hole. In this case, you need to
append this information to the so-called <i>shell profile</i> as
described in <font color="#4f81bd"><a href="#ch129.xhtml#bash-profile">How
to set the profile</a></font>.</p>
<h2 class="western"><a name="solution-3-create-shortcuts"></a>130.3
Solution 3: Create shortcuts</h2>
<p class="first-paragraph">There are many cases when we only need to
access a single program, and adding the whole directory to the path
is tedious and error prone. In this case, the solution is to first
create one directory that is being searched by our bash shell,
then<br/>
create symbolic links (shortcuts) from the programs we want
to run in this <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin</font></font>
directory. Since the directory is already on the search path, any new
programs added there will also become accessible.</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Creates the ~/bin directory</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">mkdir
-p ~/bin</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Add this directory to the search path.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">export
PATH=~/bin:$PATH</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Link fastq-dump program into the ~/bin folder</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Syntax: ln -s source destination</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ln
-s ~/src/sratoolkit.2.5.2-mac64/bin/fastq-dump ~/bin/fastq-dump</font></font></p>
<p class="first-paragraph">Of course we still need to add <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin</font></font>
to our path. But now we only have to do it once. After that, linking
the program into <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin</font></font>
will make it available everywhere. The export command should be:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">export
PATH=~/bin:$PATH</font></font></p>
<p class="first-paragraph">To load these settings automatically see
the page <font color="#4f81bd"><a href="#ch129.xhtml#bash-profile">How
to set the profile</a></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="welcome-to-bio"></a>131 Welcome to bio</h1>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
is a bioinformatics toy to play with.</p>
<p>Like LEGO pieces that match one another <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
aims to provide you with commands that naturally fit together and let
you express your intent with short, explicit and simple commands. It
is a project in an exploratory phase, we&rsquo;d welcome input and
suggestions on what it should grow up into.</p>
<p>If you&rsquo;ve ever done bioinformatics you know how even
seemingly straightforward tasks often require multiple steps,
searching the web, reading documentation, clicking around various
websites that all together can slow down your progress.</p>
<p>Time and again, I found myself not pursuing ideas because getting
to the fun part was too tedious. The <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
package was designed to solve that tedium by making bioinformatics
explorations more enjoyable. The software lets users quickly answer
questions such as:</p>
<ul>
	<li/>
<p class="compact">How do I download data by an accession
	number? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	fetch NC_045512 &gt; genome.gb</font></font></p>
	<li/>
<p class="compact">How do I view the sequence annotation of
	the data? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	gff genome.gb</font></font></p>
	<li/>
<p class="compact">How do I get the coding sequence for a
	specific gene? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	fasta --gene S genome.gb</font></font></p>
	<li/>
<p class="compact">What is the taxid for SARS-COV-2? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	taxon sars2</font></font></p>
	<li/>
<p class="compact">What is the lineage of SARS-COV-2? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	taxon 2697049 --lineage</font></font></p>
	<li/>
<p class="compact">At what date were viral samples collected?
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	data 2697049 | head</font></font></p>
	<li/>
<p class="compact">What is a microsatellites? <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	explain microsatellite</font></font></p>
	<li/>
<p class="compact">What genes match the word <font face="Consolas, serif"><font size="2" style="font-size: 11pt">HBB</font></font>?
	<font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
	search HBB --species human</font></font></p>
</ul>
<p class="first-paragraph"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
combines data from different sources: <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a></font>,<font color="#4f81bd"><a href="https://www.ensembl.org/">Ensembl</a></font>,<font color="#4f81bd"><a href="http://geneontology.org/">Gene
Ontology</a></font>,<font color="#4f81bd"><a href="http://www.sequenceontology.org/">Sequence
Ontology</a></font>, <font color="#4f81bd"><a href="https://www.ncbi.nlm.nih.gov/taxonomy">NCBI
Taxonomy</a></font> and provides an unified, consistent interface.</p>
<p>The software is also used to demonstrate and teach bioinformatics
and is the companion software to the <font color="#4f81bd"><a href="https://www.biostarhandbook.com/">Biostar
Handbook</a></font>.</p>
<h2 class="western"><a name="quick-start"></a>131.1 Quick start</h2>
<p class="first-paragraph">Install <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">pip
install bio --upgrade</font></font></p>
<h2 class="western"><a name="code-and-info"></a>131.2 Code and info</h2>
<ul>
	<li/>
<p class="compact">Source code: <font color="#4f81bd"><a href="https://github.com/ialbert/bio">https://github.com/ialbert/bio</a></font></p>
	<li/>
<p class="compact">Documentation: <font color="#4f81bd"><a href="https://www.bioinfo.help/">https://www.bioinfo.help</a></font></p>
	<li/>
<p class="compact">Report errors:
	<font color="#4f81bd"><a href="https://github.com/ialbert/bio/issues">https://github.com/ialbert/bio/issues</a></font></p>
	<li/>
<p class="compact">Discussion board (new):
	<font color="#4f81bd"><a href="https://github.com/ialbert/bio/discussions">https://github.com/ialbert/bio/discussions</a></font></p>
</ul>
<h2 class="western"><a name="usage-examples"></a>131.3 Usage examples</h2>
<p class="first-paragraph">Suppose we found the accession number to
data of interest: <font face="Consolas, serif"><font size="2" style="font-size: 11pt">NC_045512</font></font>
representing the Wuhan-Hu-1 isolate of the coronavirus and the
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532</font></font>
that stores information on the most similar bat coronavirus. We would
like to investigate the differences in the S protein betweeen the two
organisms. Here is how you could do it with <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>:</p>
<h3 class="western"><a name="fetch-data"></a>131.3.1 Fetch data</h3>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fetch MN996532 NC_045512 &gt; genomes.gb</font></font></p>
<h3 class="western"><a name="extract-sequences-of-interest"></a>131.3.2
Extract sequences of interest</h3>
<p class="first-paragraph">Look at the first 10 bases for the coding
sequence of gene S:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
fasta genomes.gb --end 10  --gene S | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;YP_009724390.1
CDS gene S, surface glycoprotein [1:10]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGTTTGTTT</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">&gt;QHU36824.1
CDS gene S, surface glycoprotein [1:10]</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ATGTTTGTTT</font></font></p>
<h3 class="western"><a name="align-sequences"></a>131.3.3 Align
sequences</h3>
<p class="first-paragraph">Let&rsquo;s align the protein sequences
for the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">S</font></font>
gene</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genomes.gb | bio fasta --gene S --protein | bio align | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
PEP: YP_009724390.1 (1,273) vs QHR63300.2 (1,269) score=6541.0</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Alignment: length=1273 ident=1240/1273(97.4%) mis=29 del=4 ins=0
gap=4</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Parameters: matrix=BLOSUM62 gap-open=11
gap-extend=1</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHVSGTNGTKRFDN</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">|||||||||||||||||||||||||||||||.|||||||||||||||||.|||||||||||||||||||||||||.|||||
81</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSSTRGVYYPDKVFRSSVLHLTQDLFLPFFSNVTWFHAIHVSGTNGIKRFDN</font></font></p>
<h3 class="western"><a name="show-alignment-as-mutations"></a>131.3.4
Show alignment as mutations</h3>
<p class="first-paragraph">Pairwise alignments can be overly verbose,
let&rsquo;s look at mutations alone:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cat
genomes.gb | bio fasta --gene S --protein | bio align --mut | tail</font></font></p>
<p class="first-paragraph">prints the type and variant at each
location:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y493Q
  SNP     493     Y       Q</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">R494S
  SNP     494     R       S</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">Y498Q
  SNP     498     Y       Q</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">D501N
  SNP     501     D       N</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">H505Y
  SNP     505     H       Y</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">N519H
  SNP     519     N       H</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">A604T
  SNP     604     A       T</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S680SPRRA
      INS     680     S       SPRRA</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">S1121N
 SNP     1121    S       N</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">I1224V
 SNP     1224    I       V</font></font></p>
<p class="first-paragraph">It shows that at position <font face="Consolas, serif"><font size="2" style="font-size: 11pt">680</font></font>
the coronavirus has an four aminoacid insertion <font face="Consolas, serif"><font size="2" style="font-size: 11pt">PRRA</font></font>,
the so called furin-cleavage.</p>
<h3 class="western"><a name="visualize-the-genome-data"></a>131.3.5
Visualize the genome data</h3>
<p class="first-paragraph">Convert to GFF</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio
gff genomes.gb | head</font></font></p>
<p class="first-paragraph">prints:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">##gff-version
3</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532.2
     .       gene    266     21552   .       +       .      
ID=1;Name=orf1ab;Parent=orf1ab;color=#cb7a77</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532.2
     .       CDS     266     13465   .       +       .      
ID=2;Name=QHR63299.2;Parent=QHR63299.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532.2
     .       CDS     13465   21552   .       +       .      
ID=3;Name=QHR63299.2;Parent=QHR63299.2</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532.2
     .       gene    21560   25369   .       +       .      
ID=4;Name=S;Parent=S;color=#cb7a77</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">MN996532.2
     .       CDS     21560   25369   .       +       .      
ID=5;Name=QHR63300.2;Parent=QHR63300.2</font></font></p>
<p class="first-paragraph">Load and view the resulting files in IGV</p>
<p><img src="0000_html_4c69db40372ce390.png" name="Image218" align="bottom" hspace="12" width="560" height="166" border="0"/>
</p>
<p>Among the many useful features, <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bio</font></font>
is also able to generate informative gene models from a GenBank file.</p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="installing-wgsim-and-dwgsim"></a>132
Installing <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font></h1>
<p class="first-paragraph"><b>Wgsim</b> (whole genome simulator) is a
tool for simulating sequence reads from a reference genome. It is
able to simulate diploid genomes with SNPs and insertion/deletion
(INDEL) polymorphisms, and simulate reads with uniform substitution
sequencing errors. It does not generate INDEL sequencing errors, but
this can be partly compensated by simulating INDEL polymorphisms.</p>
<p>Wgsim outputs the simulated polymorphisms, and writes the true
read coordinates as well as the number of polymorphisms and
sequencing errors in read names. One can evaluate the accuracy of a
mapper or a SNP caller with wgsim_eval.pl that comes with the
package.</p>
<p>Webpage: <font color="#4f81bd"><a href="https://github.com/lh3/wgsim">https://github.com/lh3/wgsim</a></font></p>
<p><b>DWGsim</b> is similarly named tool written by Nils Homer
inspired by <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>.
This tool offers more options and it can generate reads corresponding
to different sequencing platforms and produces its results in VCF
format. <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
also has a more streamlined installation instructions.</p>
<p>Webpage: <font color="#4f81bd"><a href="https://github.com/nh13/DWGSIM">https://github.com/nh13/DWGSIM</a></font></p>
<h2 class="western"><a name="installing-dwgsim"></a>132.1 Installing
DWGSim</h2>
<p class="first-paragraph">On Mac type:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">brew
install dwgsim</font></font></p>
<p class="first-paragraph">The conda version of <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
will roll back samtools to a previous version so on Linux you need to
install from source.</p>
<h2 class="western"><a name="testing"></a>132.2 Testing</h2>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font></p>
<h2 class="western"><a name="information-on-mutations"></a>132.3
Information on mutations</h2>
<p class="first-paragraph">The information on the mutations that a
read was subjected to is embedded into the read name in the form:
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">chr1_6181150_6181637_2:0:0_2:0:0_2/1</font></font>:</p>
<p class="block-text-western">By default, wgsim simulates a diploid
genome with indels and then simulates reads from the simulated genome
without indels. In the example above, the first 2:0:0 indicates that
the first read has two sequencing errors and that no substitution and
no gaps are involved in the simulated genome. The second 2:0:0 is for
the second read. As sequencing errors and substitutions are added
separately, they may overlap. You may apply &ldquo;-e 0.0&rdquo; to
disable sequencing errors.</p>
<p class="first-paragraph">Both the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgswim</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim</font></font>
distributions come with evaluation scripts <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim_eval.pl</font></font>
and <font face="Consolas, serif"><font size="2" style="font-size: 11pt">dwgsim_eval</font></font>
that can be used evaluate mapping qualities of SAM files where the
reads names are encoded as above. The scripts will parse the read
name and compare the results to the alignment and, from that
establish mapping accuracy.</p>
<h2 class="western"><a name="source-code-installation"></a>132.4
Source code installation</h2>
<p class="first-paragraph">Source code install for <font face="Consolas, serif"><font size="2" style="font-size: 11pt">wgsim</font></font>:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~/src</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">git
clone https://github.com/lh3/wgsim.git</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
wgsim</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">gcc
-g -O2 -Wall -o wgsim wgsim.c -lz -lm</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Link the program to your bin folder</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ln
-s ~/src/wgsim/wgsim ~/bin/wgsim</font></font></p>
<p class="first-paragraph">A nicely formatted manual can be accessed
via:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">more
~/src/wgsim/README</font></font></p>
<p class="first-paragraph">Source install for dwgsim:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
~/src</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">git
clone --recursive https://github.com/nh13/DWGSIM</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">cd
DWGSIM</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">make</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Link the program to your bin folder</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ln
-s ~/src/DWGSIM/dwgsim ~/bin/dwgsim</font></font></p>
<p class="compact"><br/>
<br/>

</p>
<p><br/>
<br/>

</p>
<h1 class="western"><a name="jim-kents-bioinformatics-utilities"></a>133
Jim Kent&rsquo;s bioinformatics utilities</h1>
<p class="first-paragraph">The page
<font color="#4f81bd"><a href="http://hgdownload.cse.ucsc.edu/admin/exe/">http://hgdownload.cse.ucsc.edu/admin/exe/</a></font>
location contains a number of utility programs developed by Jim Kent.</p>
<p>Some of these tools are the only &ldquo;reference&rdquo; for
formats such as bigWig, bigBed etc.</p>
<h2 class="western"><a name="installation-2"></a>133.1 Installation</h2>
<p class="first-paragraph">You may download the programs directly
from the website and place them in your <font face="Consolas, serif"><font size="2" style="font-size: 11pt">~/bin</font></font>
folder (or any other folder in your bash search path). Alternatively
you may install tools individually once you find the name that they
are distributed under with conda.</p>
<p>Search for all UCSC tools:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
search uscs</font></font></p>
<p class="first-paragraph">will produce a lenghty list of tool. To
find the <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedgraph</font></font>
to <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bigwig</font></font>
converter do a:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
search ucsc | grep bedgraph</font></font></p>
<p class="first-paragraph">this will produce the following:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">ucsc-bedgraphpack
               324               0  bioconda</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ucsc-bedgraphtobigwig
            308               0  bioconda</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ucsc-bedgraphtobigwig
            377      h8b15c51_0  bioconda</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">ucsc-bigwigtobedgraph
            324               0  bioconda</font></font></p>
<p class="first-paragraph">now install the tool of choice:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">conda
install ucsc-bedgraphtobigwig</font></font></p>
<p class="first-paragraph">test that the converter was installed
correctly:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedGraphToBigWig</font></font></p>
<p class="first-paragraph">will produce:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in"><font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedGraphToBigWig
v 4 - Convert a bedGraph file to bigWig format.</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">usage:</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
  bedGraphToBigWig in.bedGraph chrom.sizes out.bw</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">...</font></font></p>
<h2 class="western"><a name="manual-installation"></a>133.2 Manual
Installation</h2>
<p class="first-paragraph">The general way to install the tools is to
navigate to the directory <font color="#4f81bd"><a href="http://hgdownload.cse.ucsc.edu/admin/exe/">that
contains binaries</a></font> for your platform (Linux/Mac OSX/) then
download the tool.</p>
<p>For example <font face="Consolas, serif"><font size="2" style="font-size: 11pt">bedGraphToBigWig</font></font>
on MacOS X:</p>
<p style="margin-top: 0in; margin-bottom: 0.14in">    <font face="Consolas, serif"><font size="2" style="font-size: 11pt">#
Change this for Linux</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   URL=http://hgdownload.cse.ucsc.edu/admin/exe/macOSX.x86_64</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   # Get the files</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   curl $URL/bedGraphToBigWig &gt; ~/bin/bedGraphToBigWig</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   chmod +x ~/bin/bedGraphToBigWig</font></font><br/>
<br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   curl $URL/faToTwoBit &gt; ~/bin/faToTwoBit</font></font><br/>
<font face="Consolas, serif"><font size="2" style="font-size: 11pt">
   chmod +x ~/bin/faToTwoBit</font></font></p>
<p class="compact"><br/>
<br/>

</p>
</body>
</html>